{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "jqT-mhKOyUWz"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "pzmnivZlyehW"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "ogSYMCkyyekU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install docling"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "D4dhwksqyenT",
        "outputId": "36008d50-edd5-4211-9712-ed0e99cb564d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting docling\n",
            "  Downloading docling-2.31.0-py3-none-any.whl.metadata (10 kB)\n",
            "Requirement already satisfied: beautifulsoup4<5.0.0,>=4.12.3 in /usr/local/lib/python3.11/dist-packages (from docling) (4.13.4)\n",
            "Requirement already satisfied: certifi>=2024.7.4 in /usr/local/lib/python3.11/dist-packages (from docling) (2025.4.26)\n",
            "Collecting docling-core<3.0.0,>=2.26.0 (from docling-core[chunking]<3.0.0,>=2.26.0->docling)\n",
            "  Downloading docling_core-2.30.0-py3-none-any.whl.metadata (6.0 kB)\n",
            "Collecting docling-ibm-models<4.0.0,>=3.4.0 (from docling)\n",
            "  Downloading docling_ibm_models-3.4.3-py3-none-any.whl.metadata (7.6 kB)\n",
            "Collecting docling-parse<5.0.0,>=4.0.0 (from docling)\n",
            "  Downloading docling_parse-4.0.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (9.6 kB)\n",
            "Collecting easyocr<2.0,>=1.7 (from docling)\n",
            "  Downloading easyocr-1.7.2-py3-none-any.whl.metadata (10 kB)\n",
            "Collecting filetype<2.0.0,>=1.2.0 (from docling)\n",
            "  Downloading filetype-1.2.0-py2.py3-none-any.whl.metadata (6.5 kB)\n",
            "Requirement already satisfied: huggingface_hub<1,>=0.23 in /usr/local/lib/python3.11/dist-packages (from docling) (0.30.2)\n",
            "Requirement already satisfied: lxml<6.0.0,>=4.0.0 in /usr/local/lib/python3.11/dist-packages (from docling) (5.4.0)\n",
            "Collecting marko<3.0.0,>=2.1.2 (from docling)\n",
            "  Downloading marko-2.1.3-py3-none-any.whl.metadata (4.5 kB)\n",
            "Requirement already satisfied: openpyxl<4.0.0,>=3.1.5 in /usr/local/lib/python3.11/dist-packages (from docling) (3.1.5)\n",
            "Requirement already satisfied: pandas<3.0.0,>=2.1.4 in /usr/local/lib/python3.11/dist-packages (from docling) (2.2.2)\n",
            "Requirement already satisfied: pillow<12.0.0,>=10.0.0 in /usr/local/lib/python3.11/dist-packages (from docling) (11.2.1)\n",
            "Requirement already satisfied: pluggy<2.0.0,>=1.0.0 in /usr/local/lib/python3.11/dist-packages (from docling) (1.5.0)\n",
            "Requirement already satisfied: pydantic<3.0.0,>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from docling) (2.11.4)\n",
            "Collecting pydantic-settings<3.0.0,>=2.3.0 (from docling)\n",
            "  Downloading pydantic_settings-2.9.1-py3-none-any.whl.metadata (3.8 kB)\n",
            "Collecting pylatexenc<3.0,>=2.10 (from docling)\n",
            "  Downloading pylatexenc-2.10.tar.gz (162 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m162.6/162.6 kB\u001b[0m \u001b[31m10.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting pypdfium2<5.0.0,>=4.30.0 (from docling)\n",
            "  Downloading pypdfium2-4.30.1-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (48 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m48.2/48.2 kB\u001b[0m \u001b[31m1.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting python-docx<2.0.0,>=1.1.2 (from docling)\n",
            "  Downloading python_docx-1.1.2-py3-none-any.whl.metadata (2.0 kB)\n",
            "Collecting python-pptx<2.0.0,>=1.0.2 (from docling)\n",
            "  Downloading python_pptx-1.0.2-py3-none-any.whl.metadata (2.5 kB)\n",
            "Requirement already satisfied: requests<3.0.0,>=2.32.2 in /usr/local/lib/python3.11/dist-packages (from docling) (2.32.3)\n",
            "Collecting rtree<2.0.0,>=1.3.0 (from docling)\n",
            "  Downloading rtree-1.4.0-py3-none-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata (2.1 kB)\n",
            "Requirement already satisfied: scipy<2.0.0,>=1.6.0 in /usr/local/lib/python3.11/dist-packages (from docling) (1.15.2)\n",
            "Requirement already satisfied: tqdm<5.0.0,>=4.65.0 in /usr/local/lib/python3.11/dist-packages (from docling) (4.67.1)\n",
            "Requirement already satisfied: typer<0.16.0,>=0.12.5 in /usr/local/lib/python3.11/dist-packages (from docling) (0.15.3)\n",
            "Requirement already satisfied: soupsieve>1.2 in /usr/local/lib/python3.11/dist-packages (from beautifulsoup4<5.0.0,>=4.12.3->docling) (2.7)\n",
            "Requirement already satisfied: typing-extensions>=4.0.0 in /usr/local/lib/python3.11/dist-packages (from beautifulsoup4<5.0.0,>=4.12.3->docling) (4.13.2)\n",
            "Collecting jsonref<2.0.0,>=1.1.0 (from docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling)\n",
            "  Downloading jsonref-1.1.0-py3-none-any.whl.metadata (2.7 kB)\n",
            "Requirement already satisfied: jsonschema<5.0.0,>=4.16.0 in /usr/local/lib/python3.11/dist-packages (from docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (4.23.0)\n",
            "Collecting latex2mathml<4.0.0,>=3.77.0 (from docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling)\n",
            "  Downloading latex2mathml-3.78.0-py3-none-any.whl.metadata (14 kB)\n",
            "Requirement already satisfied: pyyaml<7.0.0,>=5.1 in /usr/local/lib/python3.11/dist-packages (from docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (6.0.2)\n",
            "Requirement already satisfied: tabulate<0.10.0,>=0.9.0 in /usr/local/lib/python3.11/dist-packages (from docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.9.0)\n",
            "Collecting semchunk<3.0.0,>=2.2.0 (from docling-core[chunking]<3.0.0,>=2.26.0->docling)\n",
            "  Downloading semchunk-2.2.2-py3-none-any.whl.metadata (10 kB)\n",
            "Requirement already satisfied: transformers<5.0.0,>=4.34.0 in /usr/local/lib/python3.11/dist-packages (from docling-core[chunking]<3.0.0,>=2.26.0->docling) (4.51.3)\n",
            "Collecting jsonlines<4.0.0,>=3.1.0 (from docling-ibm-models<4.0.0,>=3.4.0->docling)\n",
            "  Downloading jsonlines-3.1.0-py3-none-any.whl.metadata (1.7 kB)\n",
            "Requirement already satisfied: numpy<3.0.0,>=1.24.4 in /usr/local/lib/python3.11/dist-packages (from docling-ibm-models<4.0.0,>=3.4.0->docling) (2.0.2)\n",
            "Requirement already satisfied: opencv-python-headless<5.0.0.0,>=4.6.0.66 in /usr/local/lib/python3.11/dist-packages (from docling-ibm-models<4.0.0,>=3.4.0->docling) (4.11.0.86)\n",
            "Requirement already satisfied: safetensors<1,>=0.4.3 in /usr/local/lib/python3.11/dist-packages (from safetensors[torch]<1,>=0.4.3->docling-ibm-models<4.0.0,>=3.4.0->docling) (0.5.3)\n",
            "Requirement already satisfied: torch<3.0.0,>=2.2.2 in /usr/local/lib/python3.11/dist-packages (from docling-ibm-models<4.0.0,>=3.4.0->docling) (2.6.0+cu124)\n",
            "Requirement already satisfied: torchvision<1,>=0 in /usr/local/lib/python3.11/dist-packages (from docling-ibm-models<4.0.0,>=3.4.0->docling) (0.21.0+cu124)\n",
            "Requirement already satisfied: scikit-image in /usr/local/lib/python3.11/dist-packages (from easyocr<2.0,>=1.7->docling) (0.25.2)\n",
            "Collecting python-bidi (from easyocr<2.0,>=1.7->docling)\n",
            "  Downloading python_bidi-0.6.6-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (4.9 kB)\n",
            "Requirement already satisfied: Shapely in /usr/local/lib/python3.11/dist-packages (from easyocr<2.0,>=1.7->docling) (2.1.0)\n",
            "Collecting pyclipper (from easyocr<2.0,>=1.7->docling)\n",
            "  Downloading pyclipper-1.3.0.post6-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (9.0 kB)\n",
            "Collecting ninja (from easyocr<2.0,>=1.7->docling)\n",
            "  Downloading ninja-1.11.1.4-py3-none-manylinux_2_12_x86_64.manylinux2010_x86_64.whl.metadata (5.0 kB)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.11/dist-packages (from huggingface_hub<1,>=0.23->docling) (3.18.0)\n",
            "Requirement already satisfied: fsspec>=2023.5.0 in /usr/local/lib/python3.11/dist-packages (from huggingface_hub<1,>=0.23->docling) (2025.3.2)\n",
            "Requirement already satisfied: packaging>=20.9 in /usr/local/lib/python3.11/dist-packages (from huggingface_hub<1,>=0.23->docling) (24.2)\n",
            "Requirement already satisfied: et-xmlfile in /usr/local/lib/python3.11/dist-packages (from openpyxl<4.0.0,>=3.1.5->docling) (2.0.0)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas<3.0.0,>=2.1.4->docling) (2.9.0.post0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas<3.0.0,>=2.1.4->docling) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas<3.0.0,>=2.1.4->docling) (2025.2)\n",
            "Requirement already satisfied: annotated-types>=0.6.0 in /usr/local/lib/python3.11/dist-packages (from pydantic<3.0.0,>=2.0.0->docling) (0.7.0)\n",
            "Requirement already satisfied: pydantic-core==2.33.2 in /usr/local/lib/python3.11/dist-packages (from pydantic<3.0.0,>=2.0.0->docling) (2.33.2)\n",
            "Requirement already satisfied: typing-inspection>=0.4.0 in /usr/local/lib/python3.11/dist-packages (from pydantic<3.0.0,>=2.0.0->docling) (0.4.0)\n",
            "Collecting python-dotenv>=0.21.0 (from pydantic-settings<3.0.0,>=2.3.0->docling)\n",
            "  Downloading python_dotenv-1.1.0-py3-none-any.whl.metadata (24 kB)\n",
            "Collecting XlsxWriter>=0.5.7 (from python-pptx<2.0.0,>=1.0.2->docling)\n",
            "  Downloading XlsxWriter-3.2.3-py3-none-any.whl.metadata (2.7 kB)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests<3.0.0,>=2.32.2->docling) (3.4.1)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests<3.0.0,>=2.32.2->docling) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests<3.0.0,>=2.32.2->docling) (2.4.0)\n",
            "Requirement already satisfied: click>=8.0.0 in /usr/local/lib/python3.11/dist-packages (from typer<0.16.0,>=0.12.5->docling) (8.1.8)\n",
            "Requirement already satisfied: shellingham>=1.3.0 in /usr/local/lib/python3.11/dist-packages (from typer<0.16.0,>=0.12.5->docling) (1.5.4)\n",
            "Requirement already satisfied: rich>=10.11.0 in /usr/local/lib/python3.11/dist-packages (from typer<0.16.0,>=0.12.5->docling) (13.9.4)\n",
            "Requirement already satisfied: attrs>=19.2.0 in /usr/local/lib/python3.11/dist-packages (from jsonlines<4.0.0,>=3.1.0->docling-ibm-models<4.0.0,>=3.4.0->docling) (25.3.0)\n",
            "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /usr/local/lib/python3.11/dist-packages (from jsonschema<5.0.0,>=4.16.0->docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (2025.4.1)\n",
            "Requirement already satisfied: referencing>=0.28.4 in /usr/local/lib/python3.11/dist-packages (from jsonschema<5.0.0,>=4.16.0->docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.36.2)\n",
            "Requirement already satisfied: rpds-py>=0.7.1 in /usr/local/lib/python3.11/dist-packages (from jsonschema<5.0.0,>=4.16.0->docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.24.0)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas<3.0.0,>=2.1.4->docling) (1.17.0)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.11/dist-packages (from rich>=10.11.0->typer<0.16.0,>=0.12.5->docling) (3.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.11/dist-packages (from rich>=10.11.0->typer<0.16.0,>=0.12.5->docling) (2.19.1)\n",
            "Collecting mpire[dill] (from semchunk<3.0.0,>=2.2.0->docling-core[chunking]<3.0.0,>=2.26.0->docling)\n",
            "  Downloading mpire-2.10.2-py3-none-any.whl.metadata (14 kB)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (3.4.2)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (3.1.6)\n",
            "Collecting nvidia-cuda-nvrtc-cu12==12.4.127 (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling)\n",
            "  Downloading nvidia_cuda_nvrtc_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)\n",
            "Collecting nvidia-cuda-runtime-cu12==12.4.127 (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling)\n",
            "  Downloading nvidia_cuda_runtime_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)\n",
            "Collecting nvidia-cuda-cupti-cu12==12.4.127 (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling)\n",
            "  Downloading nvidia_cuda_cupti_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl.metadata (1.6 kB)\n",
            "Collecting nvidia-cudnn-cu12==9.1.0.70 (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling)\n",
            "  Downloading nvidia_cudnn_cu12-9.1.0.70-py3-none-manylinux2014_x86_64.whl.metadata (1.6 kB)\n",
            "Collecting nvidia-cublas-cu12==12.4.5.8 (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling)\n",
            "  Downloading nvidia_cublas_cu12-12.4.5.8-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)\n",
            "Collecting nvidia-cufft-cu12==11.2.1.3 (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling)\n",
            "  Downloading nvidia_cufft_cu12-11.2.1.3-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)\n",
            "Collecting nvidia-curand-cu12==10.3.5.147 (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling)\n",
            "  Downloading nvidia_curand_cu12-10.3.5.147-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)\n",
            "Collecting nvidia-cusolver-cu12==11.6.1.9 (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling)\n",
            "  Downloading nvidia_cusolver_cu12-11.6.1.9-py3-none-manylinux2014_x86_64.whl.metadata (1.6 kB)\n",
            "Collecting nvidia-cusparse-cu12==12.3.1.170 (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling)\n",
            "  Downloading nvidia_cusparse_cu12-12.3.1.170-py3-none-manylinux2014_x86_64.whl.metadata (1.6 kB)\n",
            "Requirement already satisfied: nvidia-cusparselt-cu12==0.6.2 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (0.6.2)\n",
            "Requirement already satisfied: nvidia-nccl-cu12==2.21.5 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (2.21.5)\n",
            "Requirement already satisfied: nvidia-nvtx-cu12==12.4.127 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (12.4.127)\n",
            "Collecting nvidia-nvjitlink-cu12==12.4.127 (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling)\n",
            "  Downloading nvidia_nvjitlink_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl.metadata (1.5 kB)\n",
            "Requirement already satisfied: triton==3.2.0 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (3.2.0)\n",
            "Requirement already satisfied: sympy==1.13.1 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (1.13.1)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from sympy==1.13.1->torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (1.3.0)\n",
            "Requirement already satisfied: regex!=2019.12.17 in /usr/local/lib/python3.11/dist-packages (from transformers<5.0.0,>=4.34.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (2024.11.6)\n",
            "Requirement already satisfied: tokenizers<0.22,>=0.21 in /usr/local/lib/python3.11/dist-packages (from transformers<5.0.0,>=4.34.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.21.1)\n",
            "Requirement already satisfied: imageio!=2.35.0,>=2.33 in /usr/local/lib/python3.11/dist-packages (from scikit-image->easyocr<2.0,>=1.7->docling) (2.37.0)\n",
            "Requirement already satisfied: tifffile>=2022.8.12 in /usr/local/lib/python3.11/dist-packages (from scikit-image->easyocr<2.0,>=1.7->docling) (2025.3.30)\n",
            "Requirement already satisfied: lazy-loader>=0.4 in /usr/local/lib/python3.11/dist-packages (from scikit-image->easyocr<2.0,>=1.7->docling) (0.4)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.11/dist-packages (from markdown-it-py>=2.2.0->rich>=10.11.0->typer<0.16.0,>=0.12.5->docling) (0.1.2)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from jinja2->torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (3.0.2)\n",
            "Collecting multiprocess>=0.70.15 (from mpire[dill]->semchunk<3.0.0,>=2.2.0->docling-core[chunking]<3.0.0,>=2.26.0->docling)\n",
            "  Downloading multiprocess-0.70.18-py311-none-any.whl.metadata (7.5 kB)\n",
            "Collecting dill>=0.4.0 (from multiprocess>=0.70.15->mpire[dill]->semchunk<3.0.0,>=2.2.0->docling-core[chunking]<3.0.0,>=2.26.0->docling)\n",
            "  Downloading dill-0.4.0-py3-none-any.whl.metadata (10 kB)\n",
            "Downloading docling-2.31.0-py3-none-any.whl (166 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m166.1/166.1 kB\u001b[0m \u001b[31m11.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading docling_core-2.30.0-py3-none-any.whl (142 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m142.3/142.3 kB\u001b[0m \u001b[31m11.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading docling_ibm_models-3.4.3-py3-none-any.whl (80 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m80.9/80.9 kB\u001b[0m \u001b[31m5.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading docling_parse-4.0.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (15.1 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m15.1/15.1 MB\u001b[0m \u001b[31m58.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading easyocr-1.7.2-py3-none-any.whl (2.9 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.9/2.9 MB\u001b[0m \u001b[31m41.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading filetype-1.2.0-py2.py3-none-any.whl (19 kB)\n",
            "Downloading marko-2.1.3-py3-none-any.whl (42 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m42.2/42.2 kB\u001b[0m \u001b[31m2.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pydantic_settings-2.9.1-py3-none-any.whl (44 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m44.4/44.4 kB\u001b[0m \u001b[31m2.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pypdfium2-4.30.1-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.9 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.9/2.9 MB\u001b[0m \u001b[31m66.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading python_docx-1.1.2-py3-none-any.whl (244 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m244.3/244.3 kB\u001b[0m \u001b[31m17.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading python_pptx-1.0.2-py3-none-any.whl (472 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m472.8/472.8 kB\u001b[0m \u001b[31m27.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading rtree-1.4.0-py3-none-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (541 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m541.1/541.1 kB\u001b[0m \u001b[31m30.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading jsonlines-3.1.0-py3-none-any.whl (8.6 kB)\n",
            "Downloading jsonref-1.1.0-py3-none-any.whl (9.4 kB)\n",
            "Downloading latex2mathml-3.78.0-py3-none-any.whl (73 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m73.7/73.7 kB\u001b[0m \u001b[31m5.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading python_dotenv-1.1.0-py3-none-any.whl (20 kB)\n",
            "Downloading semchunk-2.2.2-py3-none-any.whl (10 kB)\n",
            "Downloading nvidia_cublas_cu12-12.4.5.8-py3-none-manylinux2014_x86_64.whl (363.4 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m363.4/363.4 MB\u001b[0m \u001b[31m4.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cuda_cupti_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl (13.8 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m13.8/13.8 MB\u001b[0m \u001b[31m64.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cuda_nvrtc_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl (24.6 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m24.6/24.6 MB\u001b[0m \u001b[31m50.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cuda_runtime_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl (883 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m883.7/883.7 kB\u001b[0m \u001b[31m28.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cudnn_cu12-9.1.0.70-py3-none-manylinux2014_x86_64.whl (664.8 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m664.8/664.8 MB\u001b[0m \u001b[31m2.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cufft_cu12-11.2.1.3-py3-none-manylinux2014_x86_64.whl (211.5 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m211.5/211.5 MB\u001b[0m \u001b[31m5.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_curand_cu12-10.3.5.147-py3-none-manylinux2014_x86_64.whl (56.3 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.3/56.3 MB\u001b[0m \u001b[31m13.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cusolver_cu12-11.6.1.9-py3-none-manylinux2014_x86_64.whl (127.9 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m127.9/127.9 MB\u001b[0m \u001b[31m7.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_cusparse_cu12-12.3.1.170-py3-none-manylinux2014_x86_64.whl (207.5 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m207.5/207.5 MB\u001b[0m \u001b[31m5.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading nvidia_nvjitlink_cu12-12.4.127-py3-none-manylinux2014_x86_64.whl (21.1 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m21.1/21.1 MB\u001b[0m \u001b[31m61.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading XlsxWriter-3.2.3-py3-none-any.whl (169 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m169.4/169.4 kB\u001b[0m \u001b[31m12.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading ninja-1.11.1.4-py3-none-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (422 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m422.8/422.8 kB\u001b[0m \u001b[31m25.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pyclipper-1.3.0.post6-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (969 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m969.6/969.6 kB\u001b[0m \u001b[31m48.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading python_bidi-0.6.6-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (292 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m292.9/292.9 kB\u001b[0m \u001b[31m19.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading multiprocess-0.70.18-py311-none-any.whl (144 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m144.5/144.5 kB\u001b[0m \u001b[31m9.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading mpire-2.10.2-py3-none-any.whl (272 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m272.8/272.8 kB\u001b[0m \u001b[31m18.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading dill-0.4.0-py3-none-any.whl (119 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m119.7/119.7 kB\u001b[0m \u001b[31m9.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hBuilding wheels for collected packages: pylatexenc\n",
            "  Building wheel for pylatexenc (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pylatexenc: filename=pylatexenc-2.10-py3-none-any.whl size=136817 sha256=89eac35bdc1d2e0ba2cfe1537a4f271fc2db12a94a6b1bb6a0bace4b59c5072a\n",
            "  Stored in directory: /root/.cache/pip/wheels/b1/7a/33/9fdd892f784ed4afda62b685ae3703adf4c91aa0f524c28f03\n",
            "Successfully built pylatexenc\n",
            "Installing collected packages: python-bidi, pylatexenc, pyclipper, filetype, XlsxWriter, rtree, python-dotenv, python-docx, pypdfium2, nvidia-nvjitlink-cu12, nvidia-curand-cu12, nvidia-cufft-cu12, nvidia-cuda-runtime-cu12, nvidia-cuda-nvrtc-cu12, nvidia-cuda-cupti-cu12, nvidia-cublas-cu12, ninja, mpire, marko, latex2mathml, jsonref, jsonlines, dill, python-pptx, nvidia-cusparse-cu12, nvidia-cudnn-cu12, multiprocess, pydantic-settings, nvidia-cusolver-cu12, semchunk, docling-core, docling-parse, easyocr, docling-ibm-models, docling\n",
            "  Attempting uninstall: nvidia-nvjitlink-cu12\n",
            "    Found existing installation: nvidia-nvjitlink-cu12 12.5.82\n",
            "    Uninstalling nvidia-nvjitlink-cu12-12.5.82:\n",
            "      Successfully uninstalled nvidia-nvjitlink-cu12-12.5.82\n",
            "  Attempting uninstall: nvidia-curand-cu12\n",
            "    Found existing installation: nvidia-curand-cu12 10.3.6.82\n",
            "    Uninstalling nvidia-curand-cu12-10.3.6.82:\n",
            "      Successfully uninstalled nvidia-curand-cu12-10.3.6.82\n",
            "  Attempting uninstall: nvidia-cufft-cu12\n",
            "    Found existing installation: nvidia-cufft-cu12 11.2.3.61\n",
            "    Uninstalling nvidia-cufft-cu12-11.2.3.61:\n",
            "      Successfully uninstalled nvidia-cufft-cu12-11.2.3.61\n",
            "  Attempting uninstall: nvidia-cuda-runtime-cu12\n",
            "    Found existing installation: nvidia-cuda-runtime-cu12 12.5.82\n",
            "    Uninstalling nvidia-cuda-runtime-cu12-12.5.82:\n",
            "      Successfully uninstalled nvidia-cuda-runtime-cu12-12.5.82\n",
            "  Attempting uninstall: nvidia-cuda-nvrtc-cu12\n",
            "    Found existing installation: nvidia-cuda-nvrtc-cu12 12.5.82\n",
            "    Uninstalling nvidia-cuda-nvrtc-cu12-12.5.82:\n",
            "      Successfully uninstalled nvidia-cuda-nvrtc-cu12-12.5.82\n",
            "  Attempting uninstall: nvidia-cuda-cupti-cu12\n",
            "    Found existing installation: nvidia-cuda-cupti-cu12 12.5.82\n",
            "    Uninstalling nvidia-cuda-cupti-cu12-12.5.82:\n",
            "      Successfully uninstalled nvidia-cuda-cupti-cu12-12.5.82\n",
            "  Attempting uninstall: nvidia-cublas-cu12\n",
            "    Found existing installation: nvidia-cublas-cu12 12.5.3.2\n",
            "    Uninstalling nvidia-cublas-cu12-12.5.3.2:\n",
            "      Successfully uninstalled nvidia-cublas-cu12-12.5.3.2\n",
            "  Attempting uninstall: nvidia-cusparse-cu12\n",
            "    Found existing installation: nvidia-cusparse-cu12 12.5.1.3\n",
            "    Uninstalling nvidia-cusparse-cu12-12.5.1.3:\n",
            "      Successfully uninstalled nvidia-cusparse-cu12-12.5.1.3\n",
            "  Attempting uninstall: nvidia-cudnn-cu12\n",
            "    Found existing installation: nvidia-cudnn-cu12 9.3.0.75\n",
            "    Uninstalling nvidia-cudnn-cu12-9.3.0.75:\n",
            "      Successfully uninstalled nvidia-cudnn-cu12-9.3.0.75\n",
            "  Attempting uninstall: nvidia-cusolver-cu12\n",
            "    Found existing installation: nvidia-cusolver-cu12 11.6.3.83\n",
            "    Uninstalling nvidia-cusolver-cu12-11.6.3.83:\n",
            "      Successfully uninstalled nvidia-cusolver-cu12-11.6.3.83\n",
            "Successfully installed XlsxWriter-3.2.3 dill-0.4.0 docling-2.31.0 docling-core-2.30.0 docling-ibm-models-3.4.3 docling-parse-4.0.1 easyocr-1.7.2 filetype-1.2.0 jsonlines-3.1.0 jsonref-1.1.0 latex2mathml-3.78.0 marko-2.1.3 mpire-2.10.2 multiprocess-0.70.18 ninja-1.11.1.4 nvidia-cublas-cu12-12.4.5.8 nvidia-cuda-cupti-cu12-12.4.127 nvidia-cuda-nvrtc-cu12-12.4.127 nvidia-cuda-runtime-cu12-12.4.127 nvidia-cudnn-cu12-9.1.0.70 nvidia-cufft-cu12-11.2.1.3 nvidia-curand-cu12-10.3.5.147 nvidia-cusolver-cu12-11.6.1.9 nvidia-cusparse-cu12-12.3.1.170 nvidia-nvjitlink-cu12-12.4.127 pyclipper-1.3.0.post6 pydantic-settings-2.9.1 pylatexenc-2.10 pypdfium2-4.30.1 python-bidi-0.6.6 python-docx-1.1.2 python-dotenv-1.1.0 python-pptx-1.0.2 rtree-1.4.0 semchunk-2.2.2\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from docling.document_converter import DocumentConverter\n",
        "\n",
        "source = \"https://arxiv.org/pdf/2408.09869\"  # document per local path or URL\n",
        "converter = DocumentConverter()\n",
        "result = converter.convert(source)\n",
        "print(result.document.export_to_markdown())  # output: \"## Docling Technical Report[...]\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "LI7div4Syfud",
        "outputId": "3272efe5-401d-4311-a5bd-77888b214701"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:easyocr.easyocr:Downloading detection model, please wait. This may take several minutes depending upon your network connection.\n",
            "WARNING:easyocr.easyocr:Downloading recognition model, please wait. This may take several minutes depending upon your network connection.\n",
            "/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: \n",
            "The secret `HF_TOKEN` does not exist in your Colab secrets.\n",
            "To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.\n",
            "You will be able to reuse this secret in all of your notebooks.\n",
            "Please note that authentication is recommended but still optional to access public models or datasets.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "<!-- image -->\n",
            "\n",
            "## Docling Technical Report\n",
            "\n",
            "Version 1.0\n",
            "\n",
            "Christoph Auer Maksym Lysak Ahmed Nassar Michele Dolfi Nikolaos Livathinos Panos Vagenas Cesar Berrospi Ramis Matteo Omenetti Fabian Lindlbauer Kasper Dinkla Lokesh Mishra Yusik Kim Shubham Gupta Rafael Teixeira de Lima Valery Weber Lucas Morin Ingmar Meijer Viktor Kuropiatnyk Peter W. J. Staar\n",
            "\n",
            "AI4K Group, IBM Research R¨ uschlikon, Switzerland\n",
            "\n",
            "## Abstract\n",
            "\n",
            "This technical report introduces Docling , an easy to use, self-contained, MITlicensed open-source package for PDF document conversion. It is powered by state-of-the-art specialized AI models for layout analysis (DocLayNet) and table structure recognition (TableFormer), and runs efficiently on commodity hardware in a small resource budget. The code interface allows for easy extensibility and addition of new features and models.\n",
            "\n",
            "## 1 Introduction\n",
            "\n",
            "Converting PDF documents back into a machine-processable format has been a major challenge for decades due to their huge variability in formats, weak standardization and printing-optimized characteristic, which discards most structural features and metadata. With the advent of LLMs and popular application patterns such as retrieval-augmented generation (RAG), leveraging the rich content embedded in PDFs has become ever more relevant. In the past decade, several powerful document understanding solutions have emerged on the market, most of which are commercial software, cloud offerings [3] and most recently, multi-modal vision-language models. As of today, only a handful of open-source tools cover PDF conversion, leaving a significant feature and quality gap to proprietary solutions.\n",
            "\n",
            "With Docling , we open-source a very capable and efficient document conversion tool which builds on the powerful, specialized AI models and datasets for layout analysis and table structure recognition we developed and presented in the recent past [12, 13, 9]. Docling is designed as a simple, self-contained python library with permissive license, running entirely locally on commodity hardware. Its code architecture allows for easy extensibility and addition of new features and models.\n",
            "\n",
            "Here is what Docling delivers today:\n",
            "\n",
            "- · Converts PDF documents to JSON or Markdown format, stable and lightning fast\n",
            "- · Understands detailed page layout, reading order, locates figures and recovers table structures\n",
            "- · Extracts metadata from the document, such as title, authors, references and language\n",
            "- · Optionally applies OCR, e.g. for scanned PDFs\n",
            "- · Can be configured to be optimal for batch-mode (i.e high throughput, low time-to-solution) or interactive mode (compromise on efficiency, low time-to-solution)\n",
            "- · Can leverage different accelerators (GPU, MPS, etc).\n",
            "\n",
            "## 2 Getting Started\n",
            "\n",
            "To use Docling, you can simply install the docling package from PyPI. Documentation and examples are available in our GitHub repository at github.com/DS4SD/docling. All required model assets 1 are downloaded to a local huggingface datasets cache on first use, unless you choose to pre-install the model assets in advance.\n",
            "\n",
            "Docling provides an easy code interface to convert PDF documents from file system, URLs or binary streams, and retrieve the output in either JSON or Markdown format. For convenience, separate methods are offered to convert single documents or batches of documents. A basic usage example is illustrated below. Further examples are available in the Doclign code repository.\n",
            "\n",
            "```\n",
            "from docling.document_converter import DocumentConverter Large\n",
            "```\n",
            "\n",
            "```\n",
            "source = \"https://arxiv.org/pdf/2206.01062\" # PDF path or URL converter = DocumentConverter() result = converter.convert_single(source) print(result.render_as_markdown()) # output: \"## DocLayNet: A Human -Annotated Dataset for Document -Layout Analysis [...]\"\n",
            "```\n",
            "\n",
            "Optionally, you can configure custom pipeline features and runtime options, such as turning on or off features (e.g. OCR, table structure recognition), enforcing limits on the input document size, and defining the budget of CPU threads. Advanced usage examples and options are documented in the README file. Docling also provides a Dockerfile to demonstrate how to install and run it inside a container.\n",
            "\n",
            "## 3 Processing pipeline\n",
            "\n",
            "Docling implements a linear pipeline of operations, which execute sequentially on each given document (see Fig. 1). Each document is first parsed by a PDF backend, which retrieves the programmatic text tokens, consisting of string content and its coordinates on the page, and also renders a bitmap image of each page to support downstream operations. Then, the standard model pipeline applies a sequence of AI models independently on every page in the document to extract features and content, such as layout and table structures. Finally, the results from all pages are aggregated and passed through a post-processing stage, which augments metadata, detects the document language, infers reading-order and eventually assembles a typed document object which can be serialized to JSON or Markdown.\n",
            "\n",
            "## 3.1 PDF backends\n",
            "\n",
            "Two basic requirements to process PDF documents in our pipeline are a) to retrieve all text content and their geometric coordinates on each page and b) to render the visual representation of each page as it would appear in a PDF viewer. Both these requirements are encapsulated in Docling's PDF backend interface. While there are several open-source PDF parsing libraries available for python, we faced major obstacles with all of them for different reasons, among which were restrictive\n",
            "\n",
            "1 see huggingface.co/ds4sd/docling-models/\n",
            "\n",
            "Figure 1: Sketch of Docling's default processing pipeline. The inner part of the model pipeline is easily customizable and extensible.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "licensing (e.g. pymupdf [7]), poor speed or unrecoverable quality issues, such as merged text cells across far-apart text tokens or table columns (pypdfium, PyPDF) [15, 14].\n",
            "\n",
            "We therefore decided to provide multiple backend choices, and additionally open-source a custombuilt PDF parser, which is based on the low-level qpdf [4] library. It is made available in a separate package named docling-parse and powers the default PDF backend in Docling. As an alternative, we provide a PDF backend relying on pypdfium , which may be a safe backup choice in certain cases, e.g. if issues are seen with particular font encodings.\n",
            "\n",
            "## 3.2 AI models\n",
            "\n",
            "As part of Docling, we initially release two highly capable AI models to the open-source community, which have been developed and published recently by our team. The first model is a layout analysis model, an accurate object-detector for page elements [13]. The second model is TableFormer [12, 9], a state-of-the-art table structure recognition model. We provide the pre-trained weights (hosted on huggingface) and a separate package for the inference code as docling-ibm-models . Both models are also powering the open-access deepsearch-experience, our cloud-native service for knowledge exploration tasks.\n",
            "\n",
            "## Layout Analysis Model\n",
            "\n",
            "Our layout analysis model is an object-detector which predicts the bounding-boxes and classes of various elements on the image of a given page. Its architecture is derived from RT-DETR [16] and re-trained on DocLayNet [13], our popular human-annotated dataset for document-layout analysis, among other proprietary datasets. For inference, our implementation relies on the onnxruntime [5].\n",
            "\n",
            "The Docling pipeline feeds page images at 72 dpi resolution, which can be processed on a single CPU with sub-second latency. All predicted bounding-box proposals for document elements are post-processed to remove overlapping proposals based on confidence and size, and then intersected with the text tokens in the PDF to group them into meaningful and complete units such as paragraphs, section titles, list items, captions, figures or tables.\n",
            "\n",
            "## Table Structure Recognition\n",
            "\n",
            "The TableFormer model [12], first published in 2022 and since refined with a custom structure token language [9], is a vision-transformer model for table structure recovery. It can predict the logical row and column structure of a given table based on an input image, and determine which table cells belong to column headers, row headers or the table body. Compared to earlier approaches, TableFormer handles many characteristics of tables, such as partial or no borderlines, empty cells, rows or columns, cell spans and hierarchy both on column-heading or row-heading level, tables with inconsistent indentation or alignment and other complexities. For inference, our implementation relies on PyTorch [2].\n",
            "\n",
            "The Docling pipeline feeds all table objects detected in the layout analysis to the TableFormer model, by providing an image-crop of the table and the included text cells. TableFormer structure predictions are matched back to the PDF cells in post-processing to avoid expensive re-transcription text in the table image. Typical tables require between 2 and 6 seconds to be processed on a standard CPU, strongly depending on the amount of included table cells.\n",
            "\n",
            "## OCR\n",
            "\n",
            "Docling provides optional support for OCR, for example to cover scanned PDFs or content in bitmaps images embedded on a page. In our initial release, we rely on EasyOCR [1], a popular thirdparty OCR library with support for many languages. Docling, by default, feeds a high-resolution page image (216 dpi) to the OCR engine, to allow capturing small print detail in decent quality. While EasyOCR delivers reasonable transcription quality, we observe that it runs fairly slow on CPU (upwards of 30 seconds per page).\n",
            "\n",
            "We are actively seeking collaboration from the open-source community to extend Docling with additional OCR backends and speed improvements.\n",
            "\n",
            "## 3.3 Assembly\n",
            "\n",
            "In the final pipeline stage, Docling assembles all prediction results produced on each page into a well-defined datatype that encapsulates a converted document, as defined in the auxiliary package docling-core . The generated document object is passed through a post-processing model which leverages several algorithms to augment features, such as detection of the document language, correcting the reading order, matching figures with captions and labelling metadata such as title, authors and references. The final output can then be serialized to JSON or transformed into a Markdown representation at the users request.\n",
            "\n",
            "## 3.4 Extensibility\n",
            "\n",
            "Docling provides a straight-forward interface to extend its capabilities, namely the model pipeline. A model pipeline constitutes the central part in the processing, following initial document parsing and preceding output assembly, and can be fully customized by sub-classing from an abstract baseclass ( BaseModelPipeline ) or cloning the default model pipeline. This effectively allows to fully customize the chain of models, add or replace models, and introduce additional pipeline configuration parameters. To use a custom model pipeline, the custom pipeline class to instantiate can be provided as an argument to the main document conversion methods. We invite everyone in the community to propose additional or alternative models and improvements.\n",
            "\n",
            "Implementations of model classes must satisfy the python Callable interface. The \\_\\_call\\_\\_ method must accept an iterator over page objects, and produce another iterator over the page objects which were augmented with the additional features predicted by the model, by extending the provided PagePredictions data model accordingly.\n",
            "\n",
            "## 4 Performance\n",
            "\n",
            "In this section, we establish some reference numbers for the processing speed of Docling and the resource budget it requires. All tests in this section are run with default options on our standard test set distributed with Docling, which consists of three papers from arXiv and two IBM Redbooks, with a total of 225 pages. Measurements were taken using both available PDF backends on two different hardware systems: one MacBook Pro M3 Max, and one bare-metal server running Ubuntu 20.04 LTS on an Intel Xeon E5-2690 CPU. For reproducibility, we fixed the thread budget (through setting OMP NUM THREADS environment variable ) once to 4 (Docling default) and once to 16 (equal to full core count on the test hardware). All results are shown in Table 1.\n",
            "\n",
            "If you need to run Docling in very low-resource environments, please consider configuring the pypdfium backend. While it is faster and more memory efficient than the default docling-parse backend, it will come at the expense of worse quality results, especially in table structure recovery.\n",
            "\n",
            "Establishing GPU acceleration support for the AI models is currently work-in-progress and largely untested, but may work implicitly when CUDA is available and discovered by the onnxruntime and\n",
            "\n",
            "torch runtimes backing the Docling pipeline. We will deliver updates on this topic at in a future version of this report.\n",
            "\n",
            "Table 1: Runtime characteristics of Docling with the standard model pipeline and settings, on our test dataset of 225 pages, on two different systems. OCR is disabled. We show the time-to-solution (TTS), computed throughput in pages per second, and the peak memory used (resident set size) for both the Docling-native PDF backend and for the pypdfium backend, using 4 and 16 threads.\n",
            "\n",
            "| CPU                         | Thread budget   | native backend   | native backend   | native backend   | pypdfium backend   | pypdfium backend   | pypdfium backend   |\n",
            "|-----------------------------|-----------------|------------------|------------------|------------------|--------------------|--------------------|--------------------|\n",
            "|                             |                 | TTS              | Pages/s          | Mem              | TTS                | Pages/s            | Mem                |\n",
            "| Apple M3 Max                | 4               | 177 s 167 s      | 1.27 1.34        | 6.20 GB          | 103 s 92 s         | 2.18 2.45          | 2.56 GB            |\n",
            "| (16 cores) Intel(R) E5-2690 | 16 4 16         | 375 s 244 s      | 0.60 0.92        | 6.16 GB          | 239 s 143 s        | 0.94 1.57          | 2.42 GB            |\n",
            "\n",
            "## 5 Applications\n",
            "\n",
            "Thanks to the high-quality, richly structured document conversion achieved by Docling, its output qualifies for numerous downstream applications. For example, Docling can provide a base for detailed enterprise document search, passage retrieval or classification use-cases, or support knowledge extraction pipelines, allowing specific treatment of different structures in the document, such as tables, figures, section structure or references. For popular generative AI application patterns, such as retrieval-augmented generation (RAG), we provide quackling , an open-source package which capitalizes on Docling's feature-rich document output to enable document-native optimized vector embedding and chunking. It plugs in seamlessly with LLM frameworks such as LlamaIndex [8]. Since Docling is fast, stable and cheap to run, it also makes for an excellent choice to build document-derived datasets. With its powerful table structure recognition, it provides significant benefit to automated knowledge-base construction [11, 10]. Docling is also integrated within the open IBM data prep kit [6], which implements scalable data transforms to build large-scale multi-modal training datasets.\n",
            "\n",
            "## 6 Future work and contributions\n",
            "\n",
            "Docling is designed to allow easy extension of the model library and pipelines. In the future, we plan to extend Docling with several more models, such as a figure-classifier model, an equationrecognition model, a code-recognition model and more. This will help improve the quality of conversion for specific types of content, as well as augment extracted document metadata with additional information. Further investment into testing and optimizing GPU acceleration as well as improving the Docling-native PDF backend are on our roadmap, too.\n",
            "\n",
            "We encourage everyone to propose or implement additional features and models, and will gladly take your inputs and contributions under review . The codebase of Docling is open for use and contribution, under the MIT license agreement and in alignment with our contributing guidelines included in the Docling repository. If you use Docling in your projects, please consider citing this technical report.\n",
            "\n",
            "## References\n",
            "\n",
            "- [1] J. AI. Easyocr: Ready-to-use ocr with 80+ supported languages. https://github.com/ JaidedAI/EasyOCR , 2024. Version: 1.7.0.\n",
            "- [2] J. Ansel, E. Yang, H. He, N. Gimelshein, A. Jain, M. Voznesensky, B. Bao, P. Bell, D. Berard, E. Burovski, G. Chauhan, A. Chourdia, W. Constable, A. Desmaison, Z. DeVito, E. Ellison, W. Feng, J. Gong, M. Gschwind, B. Hirsh, S. Huang, K. Kalambarkar, L. Kirsch, M. Lazos, M. Lezcano, Y. Liang, J. Liang, Y. Lu, C. Luk, B. Maher, Y. Pan, C. Puhrsch, M. Reso, M. Saroufim, M. Y. Siraichi, H. Suk, M. Suo, P. Tillet, E. Wang, X. Wang, W. Wen, S. Zhang, X. Zhao, K. Zhou, R. Zou, A. Mathews, G. Chanan, P. Wu, and S. Chintala. Pytorch 2: Faster\n",
            "\n",
            "machine learning through dynamic python bytecode transformation and graph compilation. In Proceedings of the 29th ACM International Conference on Architectural Support for Programming Languages and Operating Systems, Volume 2 (ASPLOS '24) . ACM, 4 2024. doi: 10.1145/3620665.3640366. URL https://pytorch.org/assets/pytorch2-2.pdf .\n",
            "\n",
            "- [3] C. Auer, M. Dolfi, A. Carvalho, C. B. Ramis, and P. W. Staar. Delivering document conversion as a cloud service with high throughput and responsiveness. In 2022 IEEE 15th International Conference on Cloud Computing (CLOUD) , pages 363-373. IEEE, 2022.\n",
            "- [4] J. Berkenbilt. Qpdf: A content-preserving pdf document transformer, 2024. URL https: //github.com/qpdf/qpdf .\n",
            "- [5] O. R. developers. Onnx runtime. https://onnxruntime.ai/ , 2024. Version: 1.18.1.\n",
            "- [6] IBM. Data Prep Kit: a community project to democratize and accelerate unstructured data preparation for LLM app developers, 2024. URL https://github.com/IBM/ data-prep-kit .\n",
            "- [7] A. S. Inc. PyMuPDF, 2024. URL https://github.com/pymupdf/PyMuPDF .\n",
            "- [8] J. Liu. LlamaIndex, 11 2022. URL https://github.com/jerryjliu/llama\\_index .\n",
            "- [9] M. Lysak, A. Nassar, N. Livathinos, C. Auer, and P. Staar. Optimized Table Tokenization for Table Structure Recognition. In Document Analysis and Recognition - ICDAR 2023: 17th International Conference, San Jos´ e, CA, USA, August 21-26, 2023, Proceedings, Part II , pages 37-50, Berlin, Heidelberg, Aug. 2023. Springer-Verlag. ISBN 978-3-031-41678-1. doi: 10. 1007/978-3-031-41679-8 3. URL https://doi.org/10.1007/978-3-031-41679-8\\_3 .\n",
            "- [10] L. Mishra, S. Dhibi, Y. Kim, C. Berrospi Ramis, S. Gupta, M. Dolfi, and P. Staar. Statements: Universal information extraction from tables with large language models for ESG KPIs. In D. Stammbach, J. Ni, T. Schimanski, K. Dutia, A. Singh, J. Bingler, C. Christiaen, N. Kushwaha, V. Muccione, S. A. Vaghefi, and M. Leippold, editors, Proceedings of the 1st Workshop on Natural Language Processing Meets Climate Change (ClimateNLP 2024) , pages 193-214, Bangkok, Thailand, Aug. 2024. Association for Computational Linguistics. URL https://aclanthology.org/2024.climatenlp-1.15 .\n",
            "- [11] L. Morin, V. Weber, G. I. Meijer, F. Yu, and P. W. J. Staar. Patcid: an open-access dataset of chemical structures in patent documents. Nature Communications , 15(1):6532, August 2024. ISSN 2041-1723. doi: 10.1038/s41467-024-50779-y. URL https://doi.org/10.1038/ s41467-024-50779-y .\n",
            "- [12] A. Nassar, N. Livathinos, M. Lysak, and P. Staar. Tableformer: Table structure understanding with transformers. In Proceedings of the IEEE/CVF Conference on Computer Vision and Pattern Recognition , pages 4614-4623, 2022.\n",
            "- [13] B. Pfitzmann, C. Auer, M. Dolfi, A. S. Nassar, and P. Staar. Doclaynet: a large humanannotated dataset for document-layout segmentation. pages 3743-3751, 2022.\n",
            "- [14] pypdf Maintainers. pypdf: A Pure-Python PDF Library, 2024. URL https://github.com/ py-pdf/pypdf .\n",
            "- [15] P. Team. PyPDFium2: Python bindings for PDFium, 2024. URL https://github.com/ pypdfium2-team/pypdfium2 .\n",
            "- [16] Y. Zhao, W. Lv, S. Xu, J. Wei, G. Wang, Q. Dang, Y. Liu, and J. Chen. Detrs beat yolos on real-time object detection, 2023.\n",
            "\n",
            "## Appendix\n",
            "\n",
            "In this section, we illustrate a few examples of Docling's output in Markdown and JSON.\n",
            "\n",
            "## DocLayNet: A Large Human-Annotated Dataset for Document-Layout Analysis\n",
            "\n",
            "## DocLayNet: A Large Human-Annotated Dataset for Document-Layout Analysis\n",
            "\n",
            "Birgit Pfitzmann IBM Research Rueschlikon, Switzerland bpf@zurich.ibm.com\n",
            "\n",
            "Christoph Auer IBM Research Rueschlikon, Switzerland cau@zurich.ibm.com\n",
            "\n",
            "Michele Dolfi IBM Research Rueschlikon, Switzerland dol@zurich.ibm.com\n",
            "\n",
            "Ahmed S. Nassar IBM Research Rueschlikon, Switzerland ahn@zurich.ibm.com\n",
            "\n",
            "Peter Staar IBM Research Rueschlikon, Switzerland taa@zurich.ibm.com\n",
            "\n",
            "## ABSTRACT\n",
            "\n",
            "Accurate document layout analysis is a key requirement for highquality PDF document conversion. With the recent availability of public, large ground-truth datasets such as PubLayNet and DocBank, deep-learning models have proven to be very effective at layout detection and segmentation. While these datasets are of adequate size to train such models, they severely lack in layout variability since they are sourced from scientific article repositories such as PubMed and arXiv only. Consequently, the accuracy of the layout segmentation drops significantly when these models are applied on more challenging and diverse layouts. In this paper, we present DocLayNet , a new, publicly available, document-layout annotation dataset in COCO format. It contains 80863 manually annotated pages from diverse data sources to represent a wide variability in layouts. For each PDF page, the layout annotations provide labelled bounding-boxes with a choice of 11 distinct classes. DocLayNet also provides a subset of double- and triple-annotated pages to determine the inter-annotator agreement. In multiple experiments, we provide baseline accuracy scores (in mAP) for a set of popular object detection models. We also demonstrate that these models fall approximately 10% behind the inter-annotator agreement. Furthermore, we provide evidence that DocLayNet is of sufficient size. Lastly, we compare models trained on PubLayNet, DocBank and DocLayNet, showing that layout predictions of the DocLayNettrained models are more robust and thus the preferred choice for general-purpose document-layout analysis.\n",
            "\n",
            "## CCS CONCEPTS\n",
            "\n",
            "· Informationsystems → Documentstructure ; · Appliedcomputing → Document analysis ; · Computing methodologies → Machine learning Computer vision ; ; Object detection ;\n",
            "\n",
            "Permission to make digital or hard copies of part or all of this work for personal or classroom use is granted without fee provided that copies are not made or distributed for profit or commercial advantage and that copies bear this notice and the full citation on the first page. Copyrights for third-party components of this work must be honored. For all other uses, contact the owner/author(s). KDD '22, August 14-18, 2022, Washington, DC, USA © 2022 Copyright held by the owner/author(s). ACM ISBN 978-1-4503-9385-0/22/08. https://doi.org/10.1145/3534678.3539043\n",
            "\n",
            "Birgit Pfitzmann IBM Research Rueschlikon, Switzerland bpf@zurich.ibm.com\n",
            "\n",
            "Christoph Auer IBM Research Rueschlikon, Switzerland cau@zurich.ibm.com\n",
            "\n",
            "Michele Dolfi IBM Research Rueschlikon, Switzerland dol@zurich.ibm.com\n",
            "\n",
            "Ahmed S. Nassar IBM Research Rueschlikon, Switzerland ahn@zurich.ibm.com\n",
            "\n",
            "Peter Staar IBM Research Rueschlikon, Switzerland taa@zurich.ibm.com\n",
            "\n",
            "## ABSTRACT\n",
            "\n",
            "Accurate document layout analysis is a key requirement for highquality PDF document conversion. With the recent availability of public, large groundtruth datasets such as PubLayNet and DocBank, deep-learning models have proven to be very effective at layout detection and segmentation. While these datasets are of adequate size to train such models, they severely lack in layout variability since they are sourced from scientific article repositories such as PubMed and arXiv only. Consequently, the accuracy of the layout segmentation drops significantly when these models are applied on more challenging and diverse layouts. In this paper, we present DocLayNet , a new, publicly available, document-layout annotation dataset in COCO format. It contains 80863 manually annotated pages from diverse data sources to represent a wide variability in layouts. For each PDF page, the layout annotations provide labelled bounding-boxes with a choice of 11 distinct classes. DocLayNet also provides a subset of double- and triple-annotated pages to determine the inter-annotator agreement. In multiple experiments, we provide baseline accuracy scores (in mAP) for a set of popular object detection models. We also demonstrate that these models fall approximately 10% behind the inter-annotator agreement. Furthermore, we provide evidence that DocLayNet is of sufficient size. Lastly, we compare models trained on PubLayNet, DocBank and DocLayNet, showing that layout predictions of the DocLayNettrained models are more robust and thus the preferred choice for general-purpose document-layout analysis.\n",
            "\n",
            "## CCS CONCEPTS\n",
            "\n",
            "Æ Information systems → Document structure ; Æ Applied computing → Document analysis ; Æ Computing methodologies → Machine learning ; Computer vision ; Object detection ;\n",
            "\n",
            "Permission to make digital or hard copies of part or all of this work for personal or classroom use is granted without fee provided that copies are not made or distributed for profit or commercial advantage and that copies bear this notice and the full citation on the first page. Copyrights for third-party components of this work must be honored. For all other uses, contact the owner/author(s).\n",
            "\n",
            "KDD '22, August 14-18, 2022, Washington, DC, USA ' 2022 Copyright held by the owner/author(s). ACM ISBN 978-1-4503-9385-0/22/08. https://doi.org/10.1145/3534678.3539043\n",
            "\n",
            "Figure 1: Four examples of complex page layouts across different document categories\n",
            "\n",
            "## KEYWORDS\n",
            "\n",
            "PDF document conversion, layout segmentation, object-detection, data set, Machine Learning\n",
            "\n",
            "## ACM Reference Format:\n",
            "\n",
            "Birgit Pfitzmann, Christoph Auer, Michele Dolfi, Ahmed S. Nassar, and Peter Staar. 2022. DocLayNet: A Large Human-Annotated Dataset for DocumentLayout Analysis. In Proceedings of the 28th ACM SIGKDD Conference on Knowledge Discovery and Data Mining (KDD '22), August 14-18, 2022, Washington, DC, USA. ACM, New York, NY, USA, 9 pages. https://doi.org/10.1145/ 3534678.3539043\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "AGL Energy Limited  ABN 74 1\n",
            "\n",
            "5 061 375\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Figure 1: Four examples of complex page layouts across different document categories\n",
            "\n",
            "## KEYWORDS\n",
            "\n",
            "PDF document conversion, layout segmentation, object-detection, data set, Machine Learning\n",
            "\n",
            "## ACMReference Format:\n",
            "\n",
            "Birgit Pfitzmann, Christoph Auer, Michele Dolfi, Ahmed S. Nassar, and Peter Staar. 2022. DocLayNet: A Large Human-Annotated Dataset for DocumentLayout Analysis. In Proceedings of the 28th ACM SIGKDD Conference on Knowledge Discovery and Data Mining (KDD '22), August 14-18, 2022, Washington, DC, USA. ACM, New York, NY, USA, 9 pages. https://doi.org/10.1145/ 3534678.3539043\n",
            "\n",
            "1 INTRODUCTION\n",
            "\n",
            "Despite the substantial improvements achieved with machine-learning (ML) approaches and deep neural networks in recent years, document conversion remains a challenging problem, as demonstrated by the numerous public competitions held on this topic [1-4]. The challenge originates from the huge variability in PDF documents regarding layout, language and formats (scanned, programmatic or a combination of both). Engineering a single ML model that can be applied on all types of documents and provides high-quality layout segmentation remains to this day extremely challenging [5]. To highlight the variability in document layouts, we show a few example documents from the DocLayNet dataset in Figure 1. Figure 2: Title page of the DocLayNet paper (arxiv.org/pdf/2206.01062) - left PDF, right rendered Markdown. If recognized, metadata such as authors are appearing first under the title. Text content inside figures is currently dropped, the caption is retained and linked to the figure in the JSON representation (not shown).\n",
            "\n",
            "KDD '22, August 14-18, 2022, Washington, DC, USA Birgit Pfitzmann, Christoph Auer, Michele Dolfi, Ahmed S. Nassar, and Peter Staar\n",
            "\n",
            "Table 2: Prediction performance (mAP@0.5-0.95) of object detection networks on DocLayNet test set. The MRCNN (Mask R-CNN) and FRCNN (Faster R-CNN) models with ResNet-50 or ResNet-101 backbone were trained based on the network architectures from the detectron2 model zoo (Mask R-CNN R50, R101-FPN 3x, Faster R-CNN R101-FPN 3x), with default configurations. The YOLO implementation utilized was YOLOv5x6 [13]. All models were initialised using pre-trained weights from the COCO 2017 dataset.\n",
            "\n",
            "|                                                                                                        | human                                                                   | MRCNN R50 R101                                                                                                          | FRCNN R101                                                  | YOLO v5x6                                                   |\n",
            "|--------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------|-------------------------------------------------------------|\n",
            "| Caption Footnote Formula List-item Page-footer Page-header Picture Section-header Table Text Title All | 84-89 83-91 83-85 87-88 93-94 85-89 69-71 83-84 77-81 84-86 60-72 82-83 | 68.4 71.5 70.9 71.8 60.1 63.4 81.2 80.8 61.6 59.3 71.9 70.0 71.7 72.7 67.6 69.3 82.2 82.9 84.6 85.8 76.7 80.4 72.4 73.5 | 70.1 73.7 63.5 81.0 58.9 72.0 72.0 68.4 82.2 85.4 79.9 73.4 | 77.7 77.2 66.2 86.2 61.1 67.9 77.1 74.6 86.3 88.1 82.7 76.8 |\n",
            "\n",
            "to avoid this at any cost in order to have clear, unbiased baseline numbers for human document-layout annotation. Third, we introduced the feature of snapping boxes around text segments to obtain a pixel-accurate annotation and again reduce time and effort. The CCS annotation tool automatically shrinks every user-drawn box to the minimum bounding-box around the enclosed text-cells for all purely text-based segments, which excludes only Table and Picture . For the latter, we instructed annotation staff to minimise inclusion of surrounding whitespace while including all graphical lines. A downside of snapping boxes to enclosed text cells is that some wrongly parsed PDF pages cannot be annotated correctly and need to be skipped. Fourth, we established a way to flag pages as rejected for cases where no valid annotation according to the label guidelines could be achieved. Example cases for this would be PDF pages that render incorrectly or contain layouts that are impossible to capture with non-overlapping rectangles. Such rejected pages are not contained in the final dataset. With all these measures in place, experienced annotation staff managed to annotate a single page in a typical timeframe of 20s to 60s, depending on its complexity.\n",
            "\n",
            "## 5 EXPERIMENTS\n",
            "\n",
            "The primary goal of DocLayNet is to obtain high-quality ML models capable of accurate document-layout analysis on a wide variety of challenging layouts. As discussed in Section 2, object detection models are currently the easiest to use, due to the standardisation of ground-truth data in COCO format [16] and the availability of general frameworks such as detectron2 [17]. Furthermore, baseline numbers in PubLayNet and DocBank were obtained using standard object detection models such as Mask R-CNN and Faster R-CNN. As such, we will relate to these object detection methods in this\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Third, Ccs achienec Exanole\n",
            "\n",
            "## EXPERIMENTS\n",
            "\n",
            "ground-vuth dawa such WC\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Figure 5: Prediction performance (mAP@0.5-0.95) of a Mask R-CNNnetworkwithResNet50backbonetrainedonincreasing fractions of the DocLayNet dataset. The learning curve flattens around the 80% mark, indicating that increasing the size of the DocLayNet dataset with similar data will not yield significantly better predictions.\n",
            "\n",
            "paper and leave the detailed evaluation of more recent methods mentioned in Section 2 for future work.\n",
            "\n",
            "In this section, we will present several aspects related to the performance of object detection models on DocLayNet. Similarly as in PubLayNet, we will evaluate the quality of their predictions using mean average precision (mAP) with 10 overlaps that range from 0.5 to 0.95 in steps of 0.05 (mAP@0.5-0.95). These scores are computed by leveraging the evaluation code provided by the COCO API [16].\n",
            "\n",
            "## Baselines for Object Detection\n",
            "\n",
            "In Table 2, we present baseline experiments (given in mAP) on Mask R-CNN [12], Faster R-CNN [11], and YOLOv5 [13]. Both training and evaluation were performed on RGB images with dimensions of 1025 × 1025 pixels. For training, we only used one annotation in case of redundantly annotated pages. As one can observe, the variation in mAP between the models is rather low, but overall between 6 and 10% lower than the mAP computed from the pairwise human annotations on triple-annotated pages. This gives a good indication that the DocLayNet dataset poses a worthwhile challenge for the research community to close the gap between human recognition and ML approaches. It is interesting to see that Mask R-CNN and Faster R-CNN produce very comparable mAP scores, indicating that pixel-based image segmentation derived from bounding-boxes does not help to obtain better predictions. On the other hand, the more recent Yolov5x model does very well and even out-performs humans on selected labels such as Text , Table and Picture . This is not entirely surprising, as Text , Table and Picture are abundant and the most visually distinctive in a document.\n",
            "\n",
            "## Baselines for Object Detection\n",
            "\n",
            "mak enbrel\n",
            "\n",
            "Figure 3: Page 6 of the DocLayNet paper. If recognized, metadata such as authors are appearing first under the title. Elements recognized as page headers or footers are suppressed in Markdown to deliver uninterrupted content in reading order. Tables are inserted in reading order. The paragraph in '5. Experiments' wrapping over the column end is broken up in two and interrupted by the table.\n",
            "\n",
            "KDD '22, August 14-18, 2022, Washington, DC, USA\n",
            "\n",
            "Birgit Pfitzmann, Christoph Auer, Michele Dolfi, Ahmed S. Nassar, and Peter Staar\n",
            "\n",
            "Table 1: DocLayNet dataset overview. Along with the frequency of each class label, we present the relative occurrence (as %\n",
            "\n",
            "between pairwise annotations from the triple-annotated pages, from which we obtain accuracy ranges.\n",
            "\n",
            "of row 'Total') in the train, test and validation sets. The inter-annotator agreement is computed as the mAP@0.5-0.95 metric\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "| class label    | Count   | % of Total   | % of Total   | % of Total   | triple inter-annotator mAP @0.5-0.95 (%)   | triple inter-annotator mAP @0.5-0.95 (%)   | triple inter-annotator mAP @0.5-0.95 (%)   | triple inter-annotator mAP @0.5-0.95 (%)   | triple inter-annotator mAP @0.5-0.95 (%)   | triple inter-annotator mAP @0.5-0.95 (%)   | triple inter-annotator mAP @0.5-0.95 (%)   |\n",
            "|----------------|---------|--------------|--------------|--------------|--------------------------------------------|--------------------------------------------|--------------------------------------------|--------------------------------------------|--------------------------------------------|--------------------------------------------|--------------------------------------------|\n",
            "| class label    | Count   | Train        | Test         | Val          | All                                        | Fin                                        | Man                                        | Sci                                        | Law                                        | Pat                                        | Ten                                        |\n",
            "| Caption        | 22524   | 2.04         | 1.77         | 2.32         | 84-89                                      | 40-61                                      | 86-92                                      | 94-99                                      | 95-99                                      | 69-78                                      | n/a                                        |\n",
            "| Footnote       | 6318    | 0.60         | 0.31         | 0.58         | 83-91                                      | n/a                                        | 100                                        | 62-88                                      | 85-94                                      | n/a                                        | 82-97                                      |\n",
            "| Formula        | 25027   | 2.25         | 1.90         | 2.96         | 83-85                                      | n/a                                        | n/a                                        | 84-87                                      | 86-96                                      | n/a                                        | n/a                                        |\n",
            "| List-item      | 185660  | 17.19        | 13.34        | 15.82        | 87-88                                      | 74-83                                      | 90-92                                      | 97-97                                      | 81-85                                      | 75-88                                      | 93-95                                      |\n",
            "| Page-footer    | 70878   | 6.51         | 5.58         | 6.00         | 93-94                                      | 88-90                                      | 95-96                                      | 100                                        | 92-97                                      | 100                                        | 96-98                                      |\n",
            "| Page-header    | 58022   | 5.10         | 6.70         | 5.06         | 85-89                                      | 66-76                                      | 90-94                                      | 98-100                                     | 91-92                                      | 97-99                                      | 81-86                                      |\n",
            "| Picture        | 45976   | 4.21         | 2.78         | 5.31         | 69-71                                      | 56-59                                      | 82-86                                      | 69-82                                      | 80-95                                      | 66-71                                      | 59-76                                      |\n",
            "| Section-header | 142884  | 12.60        | 15.77        | 12.85        | 83-84                                      | 76-81                                      | 90-92                                      | 94-95                                      | 87-94                                      | 69-73                                      | 78-86                                      |\n",
            "| Table          | 34733   | 3.20         | 2.27         | 3.60         | 77-81                                      | 75-80                                      | 83-86                                      | 98-99                                      | 58-80                                      | 79-84                                      | 70-85                                      |\n",
            "| Text           | 510377  | 45.82        | 49.28        | 45.00        | 84-86                                      | 81-86                                      | 88-93                                      | 89-93                                      | 87-92                                      | 71-79                                      | 87-95                                      |\n",
            "| Title          | 5071    | 0.47         | 0.30         | 0.50         | 60-72                                      | 24-63                                      | 50-63                                      | 94-100                                     | 82-96                                      | 68-79                                      | 24-56                                      |\n",
            "| Total          | 1107470 | 941123       | 99816        | 66531        | 82-83                                      | 71-74                                      | 79-81                                      | 89-94                                      | 86-91                                      | 71-76                                      | 68-85                                      |\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "include publication repositories such as arXiv\n",
            "\n",
            "Table 1: DocLayNet dataset overview. Along with the frequency of each class label, we present the relative occurrence (as % of row \"Total\") in the train, test and validation sets. The inter-annotator agreement is computed as the mAP@0.5-0.95 metric between pairwise annotations from the triple-\n",
            "\n",
            "annotated pages, from which we obtain accuracy ranges.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "|                       |         | %of Total   | %of Total   | %of Total   | triple inter- annotator mAP @ 0.5-0.95 (%)   | triple inter- annotator mAP @ 0.5-0.95 (%)   | triple inter- annotator mAP @ 0.5-0.95 (%)   | triple inter- annotator mAP @ 0.5-0.95 (%)   | triple inter- annotator mAP @ 0.5-0.95 (%)   | triple inter- annotator mAP @ 0.5-0.95 (%)   | triple inter- annotator mAP @ 0.5-0.95 (%)   |\n",
            "|-----------------------|---------|-------------|-------------|-------------|----------------------------------------------|----------------------------------------------|----------------------------------------------|----------------------------------------------|----------------------------------------------|----------------------------------------------|----------------------------------------------|\n",
            "| class label           | Count   | Train       | Test        | Val         | All                                          | Fin                                          | Man                                          | Sci                                          | Law                                          | Pat                                          | Ten                                          |\n",
            "| Caption               | 22524   | 2.04        | 1.77        | 2.32        | 84-89                                        | 40-61                                        | 86-92                                        | 94-99                                        | 95-99                                        | 69-78                                        | n/a                                          |\n",
            "| Footnote              | 6318    | 0.60        | 0.31        | 0.58        | 83-91                                        | n/a                                          | 100                                          | 62-88                                        | 85-94                                        | n/a                                          | 82-97                                        |\n",
            "| Formula               | 25027   | 2.25        | 1.90        | 2.96        | 83-85                                        | n/a                                          | n/a                                          | 84-87                                        | 86-96                                        | n/a                                          | n/a                                          |\n",
            "| List-item             | 185660  | 17.19       | 13.34       | 15.82       | 87-88                                        | 74-83                                        | 90-92                                        | 97-97                                        | 81-85                                        | 75-88                                        | 93-95                                        |\n",
            "| Page- footer          | 70878   | 6.51        | 5.58        | 6.00        | 93-94                                        | 88-90                                        | 95-96                                        | 100                                          | 92-97                                        | 100                                          | 96-98                                        |\n",
            "| Page- header offices, | 58022   | 5.10        | 6.70        | 5.06        | 85-89                                        | 66-76                                        | 90-94                                        | 98-100                                       | 91-92                                        | 97-99                                        | 81-86                                        |\n",
            "| Picture               | 45976   | 4.21        | 2.78        | 5.31        | 69-71                                        | 56-59                                        | 82-86                                        | 69-82                                        | 80-95                                        | 66-71                                        | 59-76                                        |\n",
            "| Section- header not   | 142884  | 12.60       | 15.77       | 12.85       | 83-84                                        | 76-81                                        | 90-92                                        | 94-95                                        | 87-94                                        | 69-73                                        | 78-86                                        |\n",
            "| Table                 | 34733   | 3.20        | 2.27        | 3.60        | 77-81                                        | 75-80                                        | 83-86                                        | 98-99                                        | 58-80                                        | 79-84                                        | 70-85                                        |\n",
            "| Text                  | 510377  | 45.82       | 49.28       | 45.00       | 84-86                                        | 81-86                                        | 88-93                                        | 89-93                                        | 87-92                                        | 71-79                                        | 87-95                                        |\n",
            "| Title [22], a         | 5071    | 0.47        | 0.30        | 0.50        | 60-72                                        | 24-63                                        | 50-63                                        | 94-100                                       | 82-96                                        | 68-79                                        | 24-56                                        |\n",
            "| Total in-             | 1107470 | 941123      | 99816       | 66531       | 82-83                                        | 71-74                                        | 79-81                                        | 89-94                                        | 86-91                                        | 71-76                                        | 68-85                                        |\n",
            "\n",
            "3\n",
            "\n",
            ",\n",
            "\n",
            "government offices,\n",
            "\n",
            "We reviewed the col-\n",
            "\n",
            ",\n",
            "\n",
            "Page-\n",
            "\n",
            "Title and\n",
            "\n",
            ".\n",
            "\n",
            "page. Specificity ensures that the choice of label is not ambiguous,\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "we distributed the annotation workload and performed continuous be annotated. We refrained from class labels that are very specific\n",
            "\n",
            "only. For phases three and four, a group of 40 dedicated annotators while coverage ensures that all meaningful items on a page can\n",
            "\n",
            "quality controls. Phase one and two required a small team of experts to a document category, such as\n",
            "\n",
            "Abstract in the\n",
            "\n",
            "Scientific Articles were assembled and supervised.\n",
            "\n",
            "category. We also avoided class labels that are tightly linked to the\n",
            "\n",
            "Phase 1: Data selection and preparation.\n",
            "\n",
            "Our inclusion cri-\n",
            "\n",
            "Author\n",
            "\n",
            "Affiliation\n",
            "\n",
            "teria for documents were described in Section 3. A large effort went into ensuring that all documents are free to use. The data sources in DocBank, are often only distinguishable by discriminating on 3 https://arxiv.org/ Figure 4: Table 1 from the DocLayNet paper in the original PDF (A), as rendered Markdown (B) and in JSON representation (C). Spanning table cells, such as the multi-column header 'triple interannotator mAP@0.5-0.95 (%)', is repeated for each column in the Markdown representation (B), which guarantees that every data point can be traced back to row and column headings only by its grid coordinates in the table. In the JSON representation, the span information is reflected in the fields of each table cell (C).\n",
            "\n",
            "semantics of the text. Labels such as and\n",
            "\n",
            ",\n",
            "\n",
            "as seen\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from docling.document_converter import DocumentConverter\n",
        "\n",
        "source = \"/content/dol.pdf\"  # document per local path or URL\n",
        "converter = DocumentConverter()\n",
        "result = converter.convert(source)\n",
        "print(result.document.export_to_markdown())  # output: \"## Docling Technical Report[...]\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "V6ZRliKMy_Hc",
        "outputId": "c2326e43-af72-4aa8-d45c-6a6c0bd29244"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "https:/ /huggingface.co /ds4sd /SmolDocling-256M-preview /blob/main/READMEmd\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "SmolDocling is a multimodal Image-Text-to-Text model designed for efficient document conversion. It retains Docling's most popular features while ensuring full compatibility with\n",
            "\n",
            "Docling\n",
            "\n",
            "Features:\n",
            "\n",
            "for Efficient Tokenization\n",
            "\n",
            "Introduces DocTags an efficient and minimal representation for documents that is fully compatible with DoclingDocuments\n",
            "\n",
            "DocTags\n",
            "\n",
            "- OCR (Optical Character Recognition) Extracts text accurately from images.\n",
            "- Layout and Localization Preserves document structure and document element bounding boxes .\n",
            "\n",
            "Code Recognition\n",
            "\n",
            "Detects and formats code blocks including identation\n",
            "\n",
            "- Formula Recognition Identifies and processes mathematical expressions.\n",
            "- Chart Recognition Extracts and interprets chart data\n",
            "- Table Recognition Supports column and row headers for structured table extraction.\n",
            "\n",
            "Figure Classification\n",
            "\n",
            "Differentiates figures and graphical elements.\n",
            "\n",
            "- Caption Correspondence Links captions to relevant images and figures.\n",
            "- List Grouping Organizes and structures list elements correctly.\n",
            "- Full-Page Conversion\n",
            "- OCR with Bounding Boxes OCR regions bounding box. using\n",
            "- General Document Processing Trained for both scientific and non-scientific documents\n",
            "- Seamless Docling Integration Import into Docling and export in multiple formats.\n",
            "\n",
            "Fast inference\n",
            "\n",
            "VLLM\n",
            "\n",
            "of0.35 secs per page on AlOO GPU.\n",
            "\n",
            "using\n",
            "\n",
            "Coming soon!\n",
            "\n",
            "Better chart recognition\n",
            "\n",
            "One shot multi-page inference\n",
            "\n",
            "Chemical Recognition\n",
            "\n",
            "Datasets\n",
            "\n",
            "Get started (code examples)\n",
            "\n",
            "You can use transformers, vllm; or onnx to perform inference; and Docling to convert results to variety of output formats (md, html, etc:):\n",
            "\n",
            "Single page image inference using\n",
            "\n",
            "Tranformers through seamless support for DoclingDocuments\n",
            "\n",
            "Q\n",
            "\n",
            "{\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from docling.document_converter import DocumentConverter\n",
        "\n",
        "source = \"/content/1.pdf\"  # document per local path or URL\n",
        "converter = DocumentConverter()\n",
        "result = converter.convert(source)\n",
        "print(result.document.export_to_markdown())  # output: \"## Docling Technical Report[...]\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7UVWi9agQcrc",
        "outputId": "fdc376a4-ddcf-4af2-936c-1b695c997291"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:easyocr.easyocr:Downloading detection model, please wait. This may take several minutes depending upon your network connection.\n",
            "WARNING:easyocr.easyocr:Downloading recognition model, please wait. This may take several minutes depending upon your network connection.\n",
            "/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: \n",
            "The secret `HF_TOKEN` does not exist in your Colab secrets.\n",
            "To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.\n",
            "You will be able to reuse this secret in all of your notebooks.\n",
            "Please note that authentication is recommended but still optional to access public models or datasets.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "a single analyte. Instead, the membrane potential is proportional to the concentration of all  ions  in  the  sample  solution  capable  of  interacting  at  the  membrane's active sites. Equation 11.8 can be generalized to include the contribution of an interferent, I,\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where z A and z I are the charges of the analyte and interferent, and K A,I is a selectivity coefficient accounting for the relative response of the interferent.* The selectivity coefficient is defined as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where [A] E and [I] E are the concentrations of analyte and interferent yielding identical cell potentials. When the selectivity coefficient is 1.00, the membrane responds equally to the analyte and interferent. A membrane shows good selectivity for the analyte when K A,I is significantly less than 1.00.\n",
            "\n",
            "Selectivity coefficients for most commercially available ion-selective electrodes are provided by the manufacturer. If the selectivity coefficient is unknown, it can be determined experimentally. The easiest method for determining K A,I is to prepare a series of solutions, each of which contains the same concentration of interferent, [I]add, but a different concentration of analyte. A plot of cell potential versus the log of the analyte's concentration has two distinct linear regions (Figure 11.11). When the analyte's concentration is significantly larger than K A,I [I]add ,  the  potential  is  a  linear  function  of  log [A],  as given by equation 11.8. If K A,I [I]add is  significantly  larger  than  the  analyte's concentration, however, the cell potential remains constant. The concentration of analyte and interferent at the intersection of these two linear regions is used to calculate K A,I .\n",
            "\n",
            "Glass Ion-Selective Electrodes The first commercial glass electrodes were manufactured using Corning 015, a glass with a composition of approximately 22% Na2O, 6% CaO, and 72% SiO2. When immersed in an aqueous solution, the outer approximately 10 nm of the membrane becomes hydrated over the course of several  hours.  Hydration  of  the  glass  membrane  results  in  the  formation of negatively charged sites, G - , that are part of the glass membrane's silica framework. Sodium ions, which are able to move through the hydrated layer, serve as the counterions. Hydrogen ions from solution diffuse into the membrane and, since they bind more strongly to the glass than does Na + , displace the sodium ions\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "giving rise to the membrane's selectivity for H + . The transport of charge across the membrane is carried by the Na + ions. The potential of glass electrodes using Corning 015 obeys the equation\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "over a pH range of approximately 0.5-9. Above a pH of 9-10, the glass membrane may become more responsive to other cations, such as Na + and K + .\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.11\n",
            "\n",
            "Plot of cell potential versus the log of the analyte's concentration in the presence of a fixed concentration of interferent, showing the determination of the selectivity coefficient.\n",
            "\n",
            "## glass electrode\n",
            "\n",
            "An ion-selective electrode based on a glass membrane in which the potential develops from an ion-exchange reaction on the membrane's surface.\n",
            "\n",
            "## Figure 11.12\n",
            "\n",
            "Schematic diagram of a combination glass electrode for measuring pH.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11. 4\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "The selectivity coefficient K H + /Na + for Corning 015 is approximately 10 -11 . What error in pH is expected for a solution of 0.05 M NaOH.\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "A solution of 0.05 M NaOH has an actual H + , [H + ]act, concentration of 2 × 10 -13 M and a pH of 12.7. The electrode responds, however, to both H + and Na + , with the apparent concentration of H + , [H + ]app, given by\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The apparent concentration of H + corresponds to a pH of 12.2, giving an error of -0.5 pH units.\n",
            "\n",
            "Replacing Na2O and CaO with Li2O and BaO extends the useful pH range of glass membrane electrodes to pH levels greater than 12.\n",
            "\n",
            "Glass membrane pH electrodes are often available in a combination form that includes both the indicator and the reference electrode. The use of a single electrode greatly simplifies the measurement of pH. An example of a typical combination electrode is shown in Figure 11.12.\n",
            "\n",
            "| Table 11.1   | Representative Examples of Glass Membrane Ion-Selective Electrodes   | Representative Examples of Glass Membrane Ion-Selective Electrodes   |\n",
            "|--------------|----------------------------------------------------------------------|----------------------------------------------------------------------|\n",
            "| Analyte      | Membrane Composition                                                 | Selectivity Coefficients a                                           |\n",
            "| Na +         | 11% Na 2 O, 18% Al 2 O 3 , 71% SiO 2                                 | K Na + /H + = 1000 K Na + /K + = K Na + /Li + = 10 -3                |\n",
            "| Li +         | 15% Li 2 O, 25% Al 2 O 3 , 60% SiO 2                                 | K Li + /Na + = 0.3 K Li + /K + = 10 -3                               |\n",
            "| K +          | 27% Na 2 O, 5% Al 2 O 3 , 68% SiO 2                                  | K K + /Na + = 0.05                                                   |\n",
            "\n",
            "a Selectivity constants are approximate, and those found experimentally may vary substantially from the listed values. 3\n",
            "\n",
            "The response of the Corning 015 glass membrane to monovalent cations other than H + at high pH led to the development of glass membranes possessing a greater selectivity for other cations. For example, a glass membrane with a composition of 11% Na2O, 18% Al2O3, and 71% SiO2 is used as a Na + ion-selective electrode. Other glass electrodes have been developed for the analysis of Li + , K + , Rb + , Cs + , NH4 + , Ag + ,  and  Tl + .  Several  representative examples of glass membrane electrodes are listed in Table 11.1.\n",
            "\n",
            "Since the typical thickness of the glass membrane in an ion-selective electrode is about 50 µ m, they must be handled carefully to prevent the formation of cracks or breakage. Before a glass electrode can be used it must be conditioned by soaking for several hours in a solution containing the analyte. Glass electrodes should not be allowed to dry out, as this destroys the membrane's hydrated layer. If a glass electrode has been allowed to dry out, it must be reconditioned before it can be used. The composition of a glass membrane changes over time, affecting the electrode's performance. The average lifetime for a glass electrode is several years.\n",
            "\n",
            "Crystalline Solid-State Ion-Selective Electrodes Solid-state ion-selective electrodes use membranes fashioned from polycrystalline or single-crystal inorganic salts.  Polycrystalline ion-selective electrodes are made by forming a thin pellet of Ag2S, or a mixture of Ag2S and either a second silver salt or another metal sulfide. The pellet, which is 1-2 mm in thickness, is sealed into the end of a nonconducting plastic cylinder, and an internal solution containing the analyte and a reference electrode are placed in the cylinder. Charge is carried across the membrane by Ag + ions.\n",
            "\n",
            "The membrane potential for a Ag2S pellet develops as the result of a difference in the equilibrium position of the solubility reaction\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "on the two sides of the membrane. When used to monitor the concentration of Ag + ions, the cell potential is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The membrane also responds to the concentration of S 2- , with the cell potential given as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "solid-state ion-selective electrode An ion-selective electrode based on a sparingly soluble inorganic crystalline material.\n",
            "\n",
            "If a mixture of an insoluble silver salt and Ag2S is used to make the membrane, then the membrane potential also responds to the concentration of the anion of the added silver salt. Thus, pellets made from a mixture of Ag2S and AgCl can serve as a Cl ion-selective electrode, with a cell potential of\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Membranes fashioned from a mixture of Ag2S with CdS, CuS, or PbS are used to make ion-selective electrodes that respond to the concentration of Cd 2+ , Cu 2+ , or Pb 2+ . In this case the cell potential is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where [M 2+ ] is the concentration of the appropriate metal ion.\n",
            "\n",
            "Several examples of polycrystalline, Ag2S-based ion-selective electrodes are listed in Table 11.2. The selectivity of these ion-selective electrodes is determined by solubility. Thus, a Cl ion-selective electrode constructed using a Ag2S/AgCl membrane is more selective for Br -( K Cl - /Br - = 10 2 ) and I -( K Cl - /l - = 10 6 ) since AgBr and AgI are less soluble than AgCl. If the concentration of Br is  sufficiently high, the AgCl at the membrane-solution interface is replaced by AgBr, and the electrode's response to Cl decreases substantially. Most of the ion-selective electrodes listed in Table 11.2 can be used over an extended range of pH levels. The equilibrium between S 2and HS limits the analysis for S 2to a pH range of 13-14. Solutions of CN - , on the other hand, must be kept basic to avoid the release of HCN.\n",
            "\n",
            "The membrane of a F ion-selective electrode is fashioned from a single crystal of LaF3 that is usually doped with a small amount of EuF2 to enhance the membrane's conductivity. Since EuF2 provides only two F ions, compared with three for LaF3, each EuF2 produces a vacancy in the crystal lattice. Fluoride ions move through the membrane by moving into adjacent vacancies. The LaF3 membrane is sealed into the end of a nonconducting plastic tube, with a standard solution of F - , typically 0.1 M NaF, and a Ag/AgCl reference electrode.\n",
            "\n",
            "The membrane potential for a F ion-selective electrode results from a difference in the solubility of LaF3 on opposite sides of the membrane, with the potential given by\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "One advantage of the F ion-selective electrode is its freedom from interference. The only significant exception is OH -( K F - /OH = 0.1), which imposes a maximum pH limit for a successful analysis.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11. 5\n",
            "\n",
            "What is the maximum pH that can be tolerated in the analysis of a 1 × 10 -5 M solution of F if the error is to be less than 1%?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "To achieve an error of less than 1%, the difference between the actual, [F - ]act, and apparent, [F - ]app, concentrations of F must be less than 1%; thus\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving for [OH - ], and making appropriate substitutions gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "corresponding to a pH of less than 8.\n",
            "\n",
            "| Table 11.2   | Representative Examples of Ion-Selective Electrodes   | Polycrystalline                                                                                             |\n",
            "|--------------|-------------------------------------------------------|-------------------------------------------------------------------------------------------------------------|\n",
            "| Analyte      | Membrane Composition                                  | Selectivity Coefficients a                                                                                  |\n",
            "| Ag +         | Ag 2 S                                                | K Ag + /Cu 2+ = 10 -6 K Ag + /Pb 2+ = 10 -10 Hg 2+ interferes                                               |\n",
            "| Cd 2+        | CdS/Ag 2 S                                            | K Cd 2+ /Fe 2+ = 200 K Cd 2+ /Pb 2+ = 6 Ag + , Hg 2+ , Cu 2+ must be absent                                 |\n",
            "| Cu 2+        | CuS/Ag 2 S                                            | K Cu 2+ /Fe 3+ = 10 K Cu 2+ /Cu + = 1 Ag + , Hg 2+ must be absent                                           |\n",
            "| Pb 2+        | PbS/Ag 2 S                                            | K Pb 2+ /Fe 3+ = 1 K Pb 2+ /Cd 2+ = 1 Ag + , Hg 2+ , Cu 2+ must be absent                                   |\n",
            "| Br -         | AgBr/Ag 2 S                                           | K Br - /I - = 5000 K Br - /CN - = 100 K Br - /Cl - = 5 × 10 -3 K Br - /OH - = 1 × 10 -5 S 2- must be absent |\n",
            "| Cl -         | AgCl/Ag 2 S                                           | K Cl - /I - = 1 × 10 6 K Cl - /CN - = 1 × 10 4 K Cl - /Br - = 100 K Cl - /OH - = 0.01 S 2- must be absent   |\n",
            "| CN -         | AgI/Ag 2 S                                            | K CN - /I - = 100 K CN - /Br - = 1 × 10 -4 K CN - /Cl - = 1 × 10 -6 K CN - /OH - = 1 × 10 -8                |\n",
            "| I -          | AgI/Ag 2 S                                            | S must be absent K I - /S 2- = 30 K I - /CN - = 0.01 K I - /Br - = 1 × 10 -4 K = 1 × 10 -6                  |\n",
            "| SCN -        | AgSCN/Ag 2 S                                          | K I - /OH - = 1 × 10 -7 K SCN - /I - = 1000 K SCN - /Br - = 100 K SCN - /CN - = 100 K SCN - /Cl - = 0.1     |\n",
            "| S 2-         | Ag 2 S                                                | K SCN - /OH - = 0.01 S 2- must be absent Hg 2+ interferes                                                   |\n",
            "\n",
            "Figure 11.13\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Structure and formula of di-( n -decyl) phosphate.\n",
            "\n",
            "## Figure 11.14\n",
            "\n",
            "Schematic diagram of a Ca 2+ liquid-based ion-selective electrode.\n",
            "\n",
            "## ionophore\n",
            "\n",
            "A neutral ligand whose exterior is hydrophobic and whose interior is hydrophilic.\n",
            "\n",
            "liquid-based ion-selective electrode An ion-selective electrode in which a chelating agent is incorporated into a hydrophobic membrane.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Below a pH of 4 the predominate form of fluoride in solution is HF, which, unlike F , does not contribute to the membrane potential. For this reason, an analysis for total fluoride must be carried out at a pH greater than 4.\n",
            "\n",
            "Unlike ion-selective electrodes using glass membranes, crystalline solid-state ion-selective electrodes do not need to be conditioned before use and may be stored dry. The surface of the electrode is subject to poisoning, as described earlier for a Cl ISE in contact with an excessive concentration of Br - . When this happens, the electrode can be returned to its original condition by sanding and polishing the crystalline membrane.\n",
            "\n",
            "Liquid-Based Ion-Selective Electrodes Another approach to constructing an ion-selective electrode is to use a hydrophobic membrane containing a selective, liquid organic complexing agent. Three types of organic liquids have been used: cation exchangers, anion exchangers, and neutral ionophores. When the analyte's concentration on the two sides of the membrane is different, a membrane potential is the result. Current is carried through the membrane by the analyte.\n",
            "\n",
            "One example of a liquid-based ion-selective electrode is that for Ca 2+ ,  which uses a porous plastic membrane saturated with di-( n -decyl) phosphate (Figure 11.13). As shown in Figure 11.14, the membrane is placed at the end of a nonconducting cylindrical tube and is in contact with two reservoirs. The outer reservoir contains di-( n -decyl) phosphate in din -octylphenylphosphonate, which soaks into the porous membrane. The inner reservoir contains a standard aqueous solution of Ca 2+ and a Ag/AgCl reference electrode. Calcium ion-selective electrodes are also available in which the di-( n -decyl) phosphate is immobilized in a polyvinyl chloride\n",
            "\n",
            "| Table 11. 3   | Representative Examples of Liquid-Based Ion-Selective Electrodes   | Representative Examples of Liquid-Based Ion-Selective Electrodes                                                                       |\n",
            "|---------------|--------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------|\n",
            "| Analyte       | Membrane Composition                                               | Selectivity Coefficients a                                                                                                             |\n",
            "| Ca 2+         | di-( n -decyl) phosphate in PVC                                    | K Ca 2+ /Zn 2+ = 1-5 K Ca 2+ /Al 3+ = 0.90 K Ca 2+ /Mn 2+ = 0.38 K Ca 2+ /Cu 2+ = 0.070 K Ca 2+ /Mg 2+ = 0.032                         |\n",
            "| K +           | Valinomycin in PVC                                                 | K K + /Rb + = 1.9 K K + /Cs + = 0.38 K K + /Li + = 1 × 10 -4 K K + /Na + = 1 × 10 -5                                                   |\n",
            "| Li +          | ETH 149 in PVC                                                     | K Li + /H + = 1 K Li + /Na + = 0.05 K Li + /K + = 7 × 10 -3                                                                            |\n",
            "| NH 4 +        | Nonactin and monactin in PVC                                       | K NH 4 + /K + = 0.12 K NH 4 + /H + = 0.016 K NH 4 + /Li + = 4.2 × 10 -3 K NH 4 + /Na + = 2 × 10 -3                                     |\n",
            "| ClO 4 -       | Fe( o -phen) 3 3+ in p -nitrocymene with porous membrane           | K ClO 4 - /OH - = 1 K ClO 4 - /I - = 0.012 K ClO 4 - /NO 3 - = 1.5 × 10 -3 K ClO 4 - /Br - = 5.6 × 10 -4 K ClO 4 - /Cl - = 2.2 × 10 -4 |\n",
            "| NO 3 -        | tetradodecyl ammonium nitrate in PVC                               | K NO 3 - /Cl - = 6 × 10 -3 K NO 3 - /F - = 9 × 10 -4                                                                                   |\n",
            "\n",
            "a Selectivity constants are approximate, and those found experimentally may vary substantially from the listed values. 3\n",
            "\n",
            "(PVC) membrane, eliminating the need for a reservoir containing di-( n -decyl) phosphate.\n",
            "\n",
            "A membrane potential develops as the result of a difference in the equilibrium position of the complexation reaction\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "on the two sides of the membrane, where ( m ) indicates that the species is present in the membrane. The cell potential for the Ca 2+ ion-selective electrode is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The selectivity of the electrode for Ca 2+ is  very  good, with only Zn 2+ showing greater selectivity.\n",
            "\n",
            "The properties of several representative liquid-based ion-selective electrodes are presented in Table 11.3. An electrode using a liquid reservoir can be stored in a dilute solution of analyte and needs no additional conditioning before use. The lifetime of an electrode with a PVC membrane, however, is proportional to its exposure to aqueous solutions. For this reason these electrodes are best stored by covering the membrane with a cap containing a small amount of wetted gauze to\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.15\n",
            "\n",
            "Schematic diagram of a gas-sensing membrane electrode.\n",
            "\n",
            "## enzyme electrodes\n",
            "\n",
            "An electrode that responds to the concentration of a substrate by reacting the substrate with an immobilized enzyme, producing an ion that can be monitored with an ion-selective electrode.\n",
            "\n",
            "maintain a humid environment. The electrode must then be conditioned before use by soaking in a solution of analyte for 30-60 min.\n",
            "\n",
            "Gas-Sensing Electrodes A number of membrane electrodes have been developed that respond to the concentration of dissolved gases. The basic design of these electrodes is shown in Figure 11.15 and consists of a thin membrane separating the sample from an inner solution containing an ion-selective electrode. The membrane is permeable to the gaseous analyte, but is not permeable to nonvolatile components in the sample matrix. Once the gaseous analyte passes through the membrane, it reacts in the inner solution, producing a species whose concentration can be monitored by an appropriate ion-selective electrode. For example, in the CO2 electrode, CO2 reacts in the inner solution to produce H3O + .\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The change in the concentration of H3O + is monitored with a pH ion-selective electrode, for which the cell potential is given by equation 11.9. The relationship between the concentration of H3O + and CO2 is given by rearranging the equilibrium constant expression for reaction 11.10; thus\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where K is the equilibrium constant. If the amount of HCO3 -in the internal solution is sufficiently large, then its concentration is unaffected by the presence of CO2 and remains constant. Substituting equation 11.11 into equation 11.9 gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where K ′ is a constant that includes the constant for the pH ion-selective electrode, the equilibrium constant for reaction 11.10, and the concentration of HCO3 -.\n",
            "\n",
            "Gas-sensing electrodes have been developed for a variety of gases, the characteristics  for  which  are  listed  in  Table  11.4.  The  composition of the inner solution changes with use, and both it and the membrane must be replaced periodically. Gas-sensing electrodes are stored in a solution similar to the internal solution to minimize their exposure to atmospheric gases.\n",
            "\n",
            "Potentiometric Biosensors Potentiometric electrodes for the analysis of molecules of biochemical importance can be constructed in a fashion similar to that used for gas-sensing electrodes. The most common class of potentiometric biosensors are the so-called enzyme electrodes, in which an enzyme is trapped or immobilized at the surface of an ion-selective electrode. Reaction of the analyte with the enzyme produces a product whose concentration is monitored by the ion-selective electrode. Potentiometric biosensors have also been designed around other biologically  active  species,  including  antibodies,  bacterial  particles,  tissue,  and  hormone receptors.\n",
            "\n",
            "One example of an enzyme electrode is the urea electrode, which is based on the catalytic hydrolysis of urea by urease\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "In one version of the urea electrode, shown in Figure 11.16, an NH3 electrode is modified by adding a dialysis membrane that physically traps a pH 7.0 buffered solution of urease between the dialysis membrane and the gas-permeable\n",
            "\n",
            "| Table 11. 4   | Characteristics of Gas-Sensing Membrane Electrodes   | Characteristics of Gas-Sensing Membrane Electrodes   | Characteristics of Gas-Sensing Membrane Electrodes   |\n",
            "|---------------|------------------------------------------------------|------------------------------------------------------|------------------------------------------------------|\n",
            "| Analyte       | Reaction in Inner Solution                           | Inner Solution                                       | Ion-Selective Electrode                              |\n",
            "| CO 2          | CO 2 + 2H 2 O t HCO 3 - + H 3 O +                    | 0.01 MNaHCO 3 , 0.01 MNaCl                           | glass pH electrode                                   |\n",
            "| HCN           | HCN + H 2 O t CN - + H 3 O +                         | 0.01 MKAg(CN) 2                                      | Ag 2 S membrane electrode                            |\n",
            "| HF            | HF + H 2 O t F - + H 3 O +                           | 1 MH 3 O +                                           | F - electrode                                        |\n",
            "| H 2 S         | H 2 S + H 2 O t HS - + H 3 O +                       | pH 5 citrate buffer                                  | Ag 2 S membrane electrode                            |\n",
            "| NH 3          | NH 3 + H 2 O t NH 4 + + OH -                         | 0.01 MNH 4 Cl, 0.1 MKNO 3                            | glass pH electrode                                   |\n",
            "| NO 2          | 2NO 2 + 3H 2 O t NO 3 - + NO 2 - + 2H 3 O +          | 0.02 MNaNO 2 , 0.1 MKNO 3                            | glass pH electrode                                   |\n",
            "| SO 2          | SO 2 + 2H 2 O t HSO 3 - + H 3 O +                    | 0.001 MNaHSO 3 , pH 5                                | glass pH electrode                                   |\n",
            "\n",
            "Source: Data compiled from Cammann, K. Working with Ion-Selective Electrodes. Springer-Verlag: Berlin, 1977.\n",
            "\n",
            "membrane. 4 When immersed in the sample, urea diffuses through the dialysis membrane, where it reacts with the enzyme urease. The NH4 + that is produced is in equilibrium with NH3\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "which, in turn, diffuses through the gas-permeable membrane, where it is detected by a pH electrode. The response of the electrode to the concentration of urea is given by\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Another version of the urea electrode (Figure 11.17) immobilizes the enzyme in a polymer membrane formed directly on the tip of a glass pH electrode. 5 In this case, the electrode's response is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Few potentiometric biosensors are commercially available. As shown in Figures 11.16 and 11.17, however, available ion-selective and gas-sensing electrodes may be easily converted into biosensors. Several representative examples are described in Table 11.5, and additional examples can be found in several reviews listed in the suggested readings at the end of the chapter.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## 11B. 5 Quantitative Applications\n",
            "\n",
            "## Figure 11.16\n",
            "\n",
            "The potentiometric determination of an analyte's concentration is one of the most common quantitative analytical techniques. Perhaps the most frequently employed, routine quantitative measurement is the potentiometric determination of a solution's pH, a technique considered in more detail in the following discussion. Other areas in which potentiometric applications are important include clinical chemistry, environmental chemistry, and potentiometric titrations. Before considering these applications, however, we must first examine more closely the relationship between cell potential and the analyte's concentration, as well as methods for standardizing potentiometric measurements.\n",
            "\n",
            "Activity Versus Concentration In describing metallic and membrane indicator electrodes, the Nernst equation relates the measured cell potential to the concentration of analyte. In writing the Nernst equation, we often ignore an important detail-the\n",
            "\n",
            "Schematic diagram of an enzyme-based potentiometric biosensor for urea in which urease is trapped between two membranes.\n",
            "\n",
            "Urease immobilized in polymer membrane\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.17\n",
            "\n",
            "Schematic diagrams of a second enzymebased potentiometric biosensor for urea in which urease is immobilized in a polymer matrix.\n",
            "\n",
            "potential of an electrochemical cell is a function of activity, not concentration. Thus, the Nernst equation for a metallic electrode of the first kind is more appropriately written as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where a M n+ is the activity of the metal ion. As described in Chapter 6, the activity of an ion is equal to the product of its concentration, [M n+ ], and a matrix-dependent activity coefficient, γ M n+ .\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Substituting equation 11.15 into equation 11.14 and rearranging gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Equation 11.16 can be solved for the metal ion's concentration if its activity coefficient is known. This presents a serious complication since the activity coefficient may be difficult to determine. If, however, the standards and samples have an identical matrix, then γ M n+ remains constant, and equation 11.16 simplifies to\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where K ′ includes the activity coefficient.\n",
            "\n",
            "Quantitative Analysis Using External Standards To determine the concentration of analyte in a sample, it is necessary to standardize the electrode. If the electrode's response obeys the Nernst equation,\n",
            "\n",
            "Table 11. 5\n",
            "\n",
            "## Representative Examples of Potentiometric Biosensors\n",
            "\n",
            "| Analyte                                                                                                                                             | Biologically Active Phase a                                                                                                                                                                                                                                         | Substance Determined                                                    |\n",
            "|-----------------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------|\n",
            "| 5 ′ -adenosinemonophosphate (5 ′ -AMP) L -arginine asparagine L -cysteine L -glutamate L -glutamine oxalate penicillin L -phenylalanine sugars urea | AMP-deaminase (E) arginase + urease (E) asparaginase (E) Proteus morganii (B) yellow squash (T) Sarcina flava (B) oxalate decarboxylase (E) penicillinase (E) L -amino acid oxidase and horseradish peroxidase (E) bacteria from human dental plaque (B) urease (E) | NH 3 NH 3 NH 4 + H 2 S CO 2 NH 3 CO 2 H 3 O + I - H 3 O + NH 3 or H 3 O |\n",
            "\n",
            "- Source: Compiled from Cammann, K. Working with Ion-Selective Electrodes. Springer-Verlag: Berlin, 1977; and Lunte, C. E.; Heineman, W. R.\n",
            "- 'Electrochemical Techniques in Bioanalysis.' In Steckham, E., ed. Topics in Current Chemistry, Vol. 143, Springer-Verlag: Berlin, 1988, p. 8. 3,6\n",
            "- a Abbreviations: E = enzyme; B = bacterial particle; T = tissue.\n",
            "\n",
            "then only the constant K need be determined, and standardizing with a single external standard is possible. Since small deviations from the ideal 'Nerstian' slope of ± RT nF / or ± RT zF / are frequently observed, standardization is usually accomplished using two or more external standards.\n",
            "\n",
            "In most quantitative analyses we are interested in determining the concentration, not the activity, of the analyte. As noted earlier, however, the electrode's response is a function of the analyte's activity. In the absence of interferents, a calibration curve of potential versus activity is a straight line. A plot of potential  versus  concentration, however, may be curved at higher concentrations of  analyte due to changes in the analyte's activity coefficient. A curved calibration curve may still be used to determine the analyte's concentration if the standard's matrix matches that of the sample. When the exact composition of the sample matrix is unknown, which often is the case, matrix matching becomes impossible.\n",
            "\n",
            "Another approach to matrix matching, which does not rely on knowing the exact composition of the sample's matrix, is to add a high concentration of inert electrolyte to all samples and standards. If the concentration of added electrolyte is sufficient, any difference between the sample's matrix and that of the standards becomes trivial, and the activity coefficient remains essentially constant. The solution of inert electrolyte added to the sample and standards is called a total ionic strength adjustment buffer (TISAB).\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11.6\n",
            "\n",
            "The concentration of Ca 2+ in a water sample was determined by the method of external standards. The ionic strength of the samples and standards was maintained at a nearly constant level by making each solution 0.5 M in KNO3. The measured cell potentials for the external standards are shown in the following table.\n",
            "\n",
            "| [Ca 2+ ] (M)   |   E cell (V) |\n",
            "|----------------|--------------|\n",
            "| 1.00 × 10 -5   |       -0.125 |\n",
            "| 5.00 × 10 -5   |       -0.103 |\n",
            "| 1.00 × 10 -4   |       -0.093 |\n",
            "| 5.00 × 10 -4   |       -0.072 |\n",
            "| 1.00 × 10 -3   |       -0.065 |\n",
            "| 5.00 × 10 -3   |       -0.043 |\n",
            "| 1.00 × 10 -2   |       -0.033 |\n",
            "\n",
            "What is the concentration of Ca 2+ in a water sample if its cell potential is found to be -0.084 V?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "Linear regression gives the equation for the calibration curve as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Substituting the cell potential for the sample gives the concentration of Ca 2+ as 2.17 × 10 -4 M. Note that the slope of the calibration curve is slightly different from the ideal value of 0.05916/2 = 0.02958.\n",
            "\n",
            "total ionic strength adjustment buffer A solution containing a relatively high concentration of inert electrolytes such that its composition fixes the ionic concentration of all solutions to which it is added.\n",
            "\n",
            "Quantitative Analysis Using the Method of Standard Additions Because of the difficulty of maintaining a constant matrix for samples and standards, many quantitative potentiometric methods use the method of standard additions. A sample of volume, V X,  and analyte concentration, C X,  is  transferred  to  a  sample  cell,  and  the potential, ( E cell )X, measured. A standard addition is made by adding a small volume, V S ,  of  a  standard containing a known concentration of analyte, C S ,  to  the sample, and the potential, ( E cell )S ,  measured. Provided that V S is  significantly smaller than V X, the change in sample matrix is ignored, and the analyte's activity coefficient remains constant. Example 11.7 shows how a one-point standard addition can be used to determine the concentration of an analyte.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11. 7\n",
            "\n",
            "The concentration of Ca 2+ in a sample of sea water is determined using a Ca ion-selective electrode and a one-point standard addition. A 10.00-mL sample is transferred to a 100-mL volumetric flask and diluted to volume. A 50.00-mL aliquot of sample is placed in a beaker with the Ca ion-selective electrode and a reference electrode, and the potential is measured as -0.05290 V. A 1.00-mL aliquot of a 5.00 × 10 -2 M standard solution of Ca 2+ is added, and a potential of -0.04417 V is measured. What is the concentration of Ca 2+ in the sample of sea water?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "To begin, we write Nernst equations for the two measured cell potentials. The cell potential for the sample is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and that following the standard addition is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where V T is the total volume ( V S + V X) after the standard addition. Subtracting the first equation from the second equation gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Replacing ( E cell )S - ( E cell )X with ∆ E and rearranging yields\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Substituting known values for ∆ E, V X, V S , V T, and C S,\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and taking the inverse log of both sides gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Finally, solving for C X gives the concentration of Ca 2+ as 9.88 × 10 -4 M. Since the original sample of sea water was diluted by a factor of 10, the concentration of Ca 2+ in the sea water sample is 9.88 × 10 -3 M.\n",
            "\n",
            "Free Ions Versus Complexed Ions In discussing the F ion-selective electrode, we noted that the membrane potential is influenced by the concentration of F - , but not the concentration of HF. An analysis for fluoride, therefore, is pH-dependent. Below a pH of approximately 4, fluoride is present predominantly as HF, and a quantitative analysis for total fluoride is impossible. If the pH is increased to greater than 4, however, the equilibrium\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "shifts to the right, and a quantitative analysis for total fluoride is possible.\n",
            "\n",
            "Most potentiometric electrodes are selective for only the free, uncomplexed analyte and do not respond to complexed forms of the analyte. Solution conditions, therefore, must be carefully controlled if the purpose of the analysis is to determine the analyte's total concentration. On the other hand, this selectivity provides a significant advantage over other quantitative methods of analysis when it is necessary to determine the concentration of free ions. For example, calcium is present in urine both as free Ca 2+ ions and as protein-bound Ca 2+ ions. If a urine sample is analyzed by atomic absorption spectroscopy, the signal is proportional to the total concentration of Ca 2+ , since both free and bound calcium are atomized. Analysis with a Ca 2+ ISE, however, gives a signal that is a function of only free Ca 2+ ions since the protein-bound ions cannot interact with the electrode's membrane.\n",
            "\n",
            "Representative Method Ion-selective electrodes find application in numerous quantitative analyses, each of which has its own unique considerations. The following procedure for the analysis of fluoride in toothpaste provides an instructive example.\n",
            "\n",
            "## Method 11.1\n",
            "\n",
            "## Determination of Fluoride in Toothpaste 7\n",
            "\n",
            "Description of the Method. The concentration of fluoride in toothpastes containing soluble F may be determined with a F - ion-selective electrode, using a calibration curve prepared with external standards. Although the F - ISE is very selective (only OH with K F /OH - of 0.1 is a significant interferent), Fe 3+ and Al 3+ interfere with the analysis by forming soluble fluoride complexes that do not interact with the ion-selective electrode's membrane. This interference is minimized by reacting any Fe 3+ and Al 3+ with a suitable complexing agent.\n",
            "\n",
            "Procedure. Prepare 1 L of a standard solution of 1.00% w/v SnF2, and transfer to a plastic bottle for storage. Using this solution, prepare 100 mL each of standards containing 0.32%, 0.36%, 0.40%, 0.44%, and 0.48% w/v SnF2, adding\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Continued from page 489\n",
            "\n",
            "400 mg of malic acid to each solution as a stabilizer. Transfer the standards to plastic bottles for storage. Prepare a total ionic strength adjustment buffer (TISAB) by mixing 500 mL of water, 57 mL of glacial acetic acid, 58 g of NaCl, and 4 g of the disodium salt of DCTA ( trans -1,2-cyclohexanetetraacetic acid) in a 1-L beaker, stirring until dissolved. Cool the beaker in a water bath, and add 5 M NaOH until the pH is between 5 and 5.5. Transfer the contents of the beaker to a 1-L volumetric flask, and dilute to volume. Standards are prepared by placing approximately 1 g of a fluoride-free toothpaste, 30 mL of distilled water, and 1.00 mL of the standard into a 50-mL plastic beaker and stirring vigorously for 2 min with a stir bar. The resulting suspension is quantitatively transferred to a 100-mL volumetric flask along with 50 mL of TISAB and diluted to volume with distilled water. The entire standard solution is then transferred to a 250-mL plastic beaker until its potential is measured. Samples of toothpaste are prepared for analysis by using approximately 1-g portions and treating in the same manner as the standards. The cell potential for the standards and samples are measured using a F ion-selective electrode and an appropriate reference electrode. The solution is stirred during the measurement, and 2-3 min is allowed for equilibrium to be reached. The concentration of F - in the toothpaste is reported as %w/w SnF2.\n",
            "\n",
            "## Questions\n",
            "\n",
            "- 1. The total ionic strength adjustment buffer serves several purposes in this procedure. Identify these purposes.\n",
            "\n",
            "The composition of the TISAB accomplishes three things: (1) The high concentration of NaCl (approximately 1 M) ensures that the ionic strength of the standards and samples are essentially identical. Since the activity coefficient for fluoride will be the same in all solutions, the Nernst equation can be written using the concentration of F in place of its activity; (2) Glacial acetic acid and NaOH are used to prepare an acetic acid/acetate buffer of pH 5-5.5. The pH of this buffer is high enough to ensure that the predominant form of fluoride is F instead of HF; and (3) DCTA is added as a complexing agent for any Fe 3+ or Al 3+ that might be present, preventing the formation of FeF6 3- or AlF6 3- .\n",
            "\n",
            "- 2. Why is a fluoride-free toothpaste added to the standard solutions?\n",
            "- Fluoride-free toothpaste is added as a precaution against any matrix effects that might influence the ion-selective electrode's response. This assumes, of course, that the matrices of the two toothpastes are otherwise similar.\n",
            "- 3. The procedure specifies that the standard and sample solutions should be stored in plastic containers. Why is it not a good idea to store the solutions in glass containers?\n",
            "\n",
            "The fluoride ion is capable of reacting with glass to form SiF4.\n",
            "\n",
            "- 4. The slope of the calibration curve is found to be -57.98 mV per tenfold change in the concentration of F - , compared with the expected slope of -59.16 mV per tenfold change in concentration. What effect does this have on the quantitative analysis for %w/w SnF2 in the toothpaste samples?\n",
            "\n",
            "No effect at all-this is the reason for preparing a calibration curve with multiple standards.\n",
            "\n",
            "Measurement of pH With the availability of inexpensive glass pH electrodes and pH meters, the determination of pH has become one of the most frequent quantitative analytical measurements. The potentiometric determination of pH, however, is not without complications, several of which are discussed in this section.\n",
            "\n",
            "One complication is the meaning of pH. 8,9 The conventional definition of pH as presented in most introductory texts is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The pH of a solution, however, is defined by the response of an electrode to the H + ion and, therefore, is a measure of its activity.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Calculating the pH of a solution using equation 11.17 only approximates the true pH. Thus, a solution of 0.1 M HCl has a calculated pH of 1.00 using equation 11.17, but an actual pH of 1.1 as defined by equation 11.18. 8 The difference between the two values occurs because the activity coefficient for H + is not unity in a matrix of 0.1 M HCl. Obviously the true pH of a solution is affected by the composition of its matrix. As an extreme example, the pH of 0.01 M HCl in 5 m LiCl is 0.8, a value that is more acidic than that of 0.1 M HCl! 8\n",
            "\n",
            "A second complication in measuring pH results from uncertainties in the relationship between potential and activity. For a glass membrane electrode, the cell potential, E X, for a solution of unknown pH is given as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where K includes the potential of the reference electrode, the asymmetry potential of the glass membrane and any liquid junction potentials in the electrochemical cell. All the contributions to K are subject to uncertainty and may change from day to day, as well as between electrodes. For this reason a pH electrode must be calibrated using a standard buffer of known pH. The cell potential for the standard, E S , is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where pHS is  the  pH  of  the  standard.  Subtracting equation 11.20 from equation 11.19 and solving for pH gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "which is the operational definition of pH adopted by the International Union of Pure and Applied Chemistry.*\n",
            "\n",
            "*Equations 11.19-11.21 are defined for a potentiometric electrochemical cell in which the pH electrode is the cathode. In this case an increase in pH decreases the cell potential. Many pH meters are designed with the pH electrode as the anode so that an increase in pH increases the cell potential. The operational definition of pH then becomes\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "This difference, however, does not affect the operation of a pH meter.\n",
            "\n",
            "| Table 11.6       | pH Values for Selected NIST Primary Standard Buffers a   | pH Values for Selected NIST Primary Standard Buffers a   | pH Values for Selected NIST Primary Standard Buffers a   | pH Values for Selected NIST Primary Standard Buffers a   | pH Values for Selected NIST Primary Standard Buffers a   | pH Values for Selected NIST Primary Standard Buffers a   | pH Values for Selected NIST Primary Standard Buffers a   |\n",
            "|------------------|----------------------------------------------------------|----------------------------------------------------------|----------------------------------------------------------|----------------------------------------------------------|----------------------------------------------------------|----------------------------------------------------------|----------------------------------------------------------|\n",
            "| Temperature (°C) | Saturated (25 °C) KHC 4 H 4 O 6 (tartrate)               | 0.05 m KH 2 C 6 H 5 O 7 (citrate)                        | 0.05 m KHC 8 H 4 O 4 (phthalate)                         | 0.025 m KH 2 PO 4 , 0.025 m Na 2 HPO 4                   | 0.008695 m KH 2 PO 4 , 0.03043 m Na 2 HPO 4              | 0.01 m Na 4 B 4 O 7 (borax)                              | 0.025 m NaHCO 3 , 0.025 m Na 2 CO 3                      |\n",
            "| 0                | -                                                        | 3.863                                                    | 4.003                                                    | 6.984                                                    | 7.534                                                    | 9.464                                                    | 10.317                                                   |\n",
            "| 5                | -                                                        | 3.840                                                    | 3.999                                                    | 6.951                                                    | 7.500                                                    | 9.395                                                    | 10.245                                                   |\n",
            "| 10               | -                                                        | 3.820                                                    | 3.998                                                    | 6.923                                                    | 7.472                                                    | 9.332                                                    | 10.179                                                   |\n",
            "| 15               | -                                                        | 3.802                                                    | 3.999                                                    | 6.900                                                    | 7.448                                                    | 9.276                                                    | 10.118                                                   |\n",
            "| 20               | -                                                        | 3.788                                                    | 4.002                                                    | 6.881                                                    | 7.429                                                    | 9.225                                                    | 10.062                                                   |\n",
            "| 25               | 3.557                                                    | 3.776                                                    | 4.008                                                    | 6.865                                                    | 7.413                                                    | 9.180                                                    | 10.012                                                   |\n",
            "| 30               | 3.552                                                    | 3.766                                                    | 4.015                                                    | 6.854                                                    | 7.400                                                    | 9.139                                                    | 9.966                                                    |\n",
            "| 35               | 3.549                                                    | 3.759                                                    | 4.024                                                    | 6.844                                                    | 7.389                                                    | 9.102                                                    | 9.925                                                    |\n",
            "| 40               | 3.547                                                    | 3.753                                                    | 4.035                                                    | 6.838                                                    | 7.380                                                    | 9.068                                                    | 9.889                                                    |\n",
            "| 45               | 3.547                                                    | 3.750                                                    | 4.047                                                    | 6.834                                                    | 7.373                                                    | 9.038                                                    | 9.856                                                    |\n",
            "| 50               | 3.549                                                    | 3.749                                                    | 4.060                                                    | 6.833                                                    | 7.367                                                    | 9.011                                                    | 9.828                                                    |\n",
            "\n",
            "Source: Values taken from Bates, R. G. Determination of pH: Theory and Practice, 2nd ed. Wiley: New York, 1973. 10\n",
            "\n",
            "a Concentrations are given in molality (moles solute per kilograms solvent).\n",
            "\n",
            "Calibrating the electrode presents a third complication since a standard with an accurately known activity for H + needs to be used. Unfortunately, it is not possible to calculate rigorously the activity of a single ion. For this reason pH electrodes are calibrated using a standard buffer whose composition is chosen such that the defined pH is as close as possible to that given by equation 11.18. Table 11.6 gives pH values for several primary standard buffer solutions accepted by the National Institute of Standards and Technology.\n",
            "\n",
            "A pH electrode is normally standardized using two buffers: one near a pH of 7 and one that is more acidic or basic depending on the sample's expected pH. The pH electrode is immersed in the first buffer, and the 'standardize' or 'calibrate' control is adjusted until the meter reads the correct pH. The electrode is placed in the second buffer, and the 'slope' or 'temperature' control is adjusted to thebuffer's pH. Some pH meters are equipped with a temperature compensation feature, allowing the pH meter to correct the measured pH for any change in temperature. In this case a thermistor is placed in the sample and connected to the pH meter. The 'temperature' control is set to the solution's temperature, and the pH meter is calibrated using the 'calibrate' and 'slope' controls. If a change in the sample's temperature is indicated by the thermistor, the pH meter adjusts the slope of the calibration based on an assumed Nerstian response of 2.303 RT/F.\n",
            "\n",
            "Clinical Applications Perhaps the area in which ion-selective electrodes receive the widest use is in clinical analysis, where their selectivity for the analyte in a complex matrix provides a significant advantage over many other analytical methods. The most common analytes are electrolytes, such as Na + , K + , Ca 2+ , H + , and Cl , and dissolved gases, such as CO2. For extracellular fluids, such as blood and urine, the analysis can be made in vitro with conventional electrodes, provided that sufficient sample is available. Some clinical analyzers place a series of ion-selective electrodes in a flow\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "cell,  allowing several analytes to be monitored simultaneously. Standards, samples, and rinse solutions are pumped through the flow cell and across the surface of the electrodes. For smaller volumes of sample the analysis can be conducted using disposable ion-selective systems, such as the Kodak Ektachem analyzer for K + shown in Figure 11.18. The analyzer consists of separate electrodes for the sample and reference solutions. Each electrode is constructed from several thin films, consisting of a Ag/AgCl reference electrode, a salt bridge and an ion-selective membrane, deposited on a support base. The two electrodes are connected by a paper salt bridge saturated with the sample and reference solutions. The overall dimensions of the analyzer are 2.8 cm × 2.4 cm with a thickness of 150 µ m and require only 10 µ L each of sample and reference solution. Similar analyzers are available for the determination of Na + , Cl , and CO2.\n",
            "\n",
            "The analysis of intercellular fluids requires an ion-selective electrode that can be inserted directly into the desired cell. Liquid-based membrane microelectrodes with tip diameters of less than 1 µ m are constructed by heating and drawing out a hard-glass capillary tube with an initial diameter of approximately 1-2 mm (Figure 11.19). The tip of the microelectrode is made hydrophobic by dipping in dichlorodimethyl silane. An inner solution appropriate for the desired analyte and a Ag/AgCl wire reference electrode are placed within the microelectrode. The tip of the microelectrode is then dipped into a solution containing the liquid complexing agent. The small volume of liquid complexing agent entering the microelectrode is retained within the tip by capillary action, eliminating the need for a solid membrane. Potentiometric microelectrodes have been developed for a number of clinically important analytes, including H + , K + , Na + , Ca 2+ , Cl - , and I -.\n",
            "\n",
            "Figure 11.19 Schematic diagram of a liquid-based ionselective microelectrode.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Environmental Applications Although ion-selective electrodes find use in environmental analysis, their application is not as widespread as in clinical analysis. Standard methods have been developed for the analysis of CN - , F - , NH3, and NO3 -in water and wastewater. Except for F - , however, other analytical methods are considered superior. By incorporating the ion-selective electrode into a flow cell, the continuous monitoring of wastewater streams and other flow systems is possible. Such applications are limited, however, by the electrode's response to the analyte's activity, rather than its concentration. Considerable interest has been shown in the development of biosensors for the field screening and monitoring of environmental samples for a number of priority pollutants. 11\n",
            "\n",
            "Potentiometric Titrations In Chapter 9 we noted that one method for determining the equivalence point of an acid-base titration is to follow the change in pH with a pH electrode. The potentiometric determination of equivalence points is feasible for acid-base, complexation, redox, and precipitation titrations, as well as for titrations in  aqueous and nonaqueous solvents. Acid-base, complexation, and precipitation potentiometric titrations are usually monitored with an ion-selective electrode that is selective for the analyte, although an electrode that is selective for the titrant or a reaction product also can be used. A redox electrode, such as a Pt wire, and a reference electrode are used for potentiometric redox titrations. More details about potentiometric titrations are found in Chapter 9.\n",
            "\n",
            "## 11B.6 Evaluation\n",
            "\n",
            "Scale of Operation The working range for most ion-selective electrodes is from a maximum concentration of 0.1-1 M to a minimum concentration of 10 -5 - 10 -10 M. This broad working range extends from major to ultratrace analytes, and is significantly greater than many other analytical methods. For conventional ion-selective electrodes, macro-sized samples with minimum volumes of 0.05-10 mL are necessary. Microelectrodes and specially designed analyzers, such as the Kodak Ektachem analyzer for K + shown in Figure 11.18, may be used with ultramicro-sized samples provided that the sample taken for analysis is sufficiently large to be representative of the original sample.\n",
            "\n",
            "Accuracy The accuracy of a potentiometric analysis is limited by the measurement error for the cell's potential. Several factors contribute to this measurement error, including the contribution to the potential from interfering ions, the finite current drawn through the cell while measuring the potential, differences in the analyte's  activity  coefficient  in  the  sample  and  standard  solutions,  and  liquid junction potentials. Errors in accuracy due to interfering ions often can be eliminated by including a separation step before the potentiometric analysis. Modern high-impedance potentiometers minimize errors due to the passage of current through the electrochemical cell. Errors due to activity coefficients and liquid junction potentials are minimized by matching the matrix of the standards to that of the sample. Even in the best circumstances, however, a difference in potential of approximately ± 1 mV is observed for samples and standards at equal concentration.\n",
            "\n",
            "The effect of an uncertainty in potential on the accuracy of a potentiometric method of analysis is evaluated using a propagation of uncertainty. For a membrane ion-selective electrode the general expression for potential is given as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where z is the charge of the analyte. From Table 4.9 in Chapter 4, the error in the cell potential, ∆ E cell is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Rearranging and multiplying through by 100 gives the percent relative error in concentration as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The relative measurement error in concentration, therefore, is determined by the magnitude of the error in measuring the cell's potential and by the charge of the analyte. Representative values are shown in Table 11.7 for ions with charges of ± 1 and ± 2, at a temperature of 25 °C. Accuracies of 1-5% for monovalent ions and 2-10% for divalent ions are typical. Although equation 11.22 was developed for membrane electrodes, it also applies to metallic electrodes of the first and second kind when z is replaced by n.\n",
            "\n",
            "Precision The precision of a potentiometric measurement is limited by variations in  temperature and the sensitivity of the potentiometer. Under most conditions, and with simple, general-purpose potentiometers, the potential can be measured with a repeatability of ± 0.1 mV. From Table 11.7 this result corresponds to an uncertainty of ± 0.4% for monovalent analytes, and ± 0.8% for divalent analytes. The reproducibility of potentiometric measurements is about a factor of 10 poorer.\n",
            "\n",
            "Sensitivity The sensitivity of a potentiometric analysis is determined by the term RT/nF or RT/zF in the Nernst equation. Sensitivity is best for smaller values of n or z.\n",
            "\n",
            "| Table 11. 7                 | Relationship Between Measurement Error in Potential and Relative Error in Concentration   | Relationship Between Measurement Error in Potential and Relative Error in Concentration   |\n",
            "|-----------------------------|-------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------|\n",
            "| Relative Error in Potential | Relative Error in Potential                                                               | (%) z                                                                                     |\n",
            "| 0.1                         | 0.1                                                                                       | ± 0.8                                                                                     |\n",
            "| 0.5                         | 0.5                                                                                       | ± 3.9                                                                                     |\n",
            "| 1.0                         | 1.0                                                                                       | ± 7.8                                                                                     |\n",
            "| 1.5                         | 1.5                                                                                       | ± 11.1                                                                                    |\n",
            "| 2.0                         | 2.0                                                                                       | ± 15.6                                                                                    |\n",
            "\n",
            "## coulometry\n",
            "\n",
            "An electrochemical method in which the current required to exhaustively oxidize or reduce the analyte is measured.\n",
            "\n",
            "## Faraday's law\n",
            "\n",
            "The current or charge passed in a redox reaction is proportional to the moles of the reaction's reactants and products.\n",
            "\n",
            "## current efficiency\n",
            "\n",
            "The percentage of current that actually leads to the analyte's oxidation or reduction.\n",
            "\n",
            "Selectivity As described earlier, most ion-selective electrodes respond to more than one analyte. For many ion-selective electrodes, however, the selectivity for the analyte is significantly greater than for most interfering ions. Published selectivity coefficients for ion-selective electrodes (representative values are found in Tables 11.1 through 11.3) provide a useful guide in helping the analyst determine whether a potentiometric analysis is feasible for a given sample.\n",
            "\n",
            "Time, Cost, and Equipment In comparison with competing methods, potentiometry provides a rapid, relatively low-cost means for analyzing samples. Commercial instruments for measuring pH or potential are available in a variety of price ranges and include portable models for use in the field.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## 11C Coulometric Methods of Analysis\n",
            "\n",
            "In potentiometry, the potential of an electrochemical cell under static conditions is used to determine an analyte's concentration. As seen in the preceding section, potentiometry is an important and frequently used quantitative method of analysis. Dynamic electrochemical methods, such as coulometry, voltammetry, and amperometry, in which current passes through the electrochemical cell, also are important analytical techniques. In this section we consider coulometric methods of analysis. Voltammetry and amperometry are covered in Section 11D.\n",
            "\n",
            "Coulometric methods of analysis are based on an exhaustive electrolysis of the analyte. By exhaustive we mean that the analyte is quantitatively oxidized or reduced at the working electrode or reacts quantitatively with a reagent generated at the working electrode. There are two forms of coulometry: controlled-potential coulometry, in which a constant potential is applied to the electrochemical cell, and controlled-current coulometry, in which a constant current is passed through the electrochemical cell.\n",
            "\n",
            "The total charge, Q, in coulombs, passed during an electrolysis is related to the absolute amount of analyte by Faraday's law\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where n is  the number of electrons transferred per mole of analyte, F is  Faraday's constant (96487 C mol -1 ), and N is the moles of analyte. A coulomb is also equivalent to an A s; thus, for a constant current, ⋅ i, the charge is given as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where t e is the electrolysis time. If current varies with time, as it does in controlledpotential coulometry, then the total charge is given by\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "11.25\n",
            "\n",
            "In coulometry, current and time are measured, and equation 11.24 or equation 11.25 is used to calculate Q. Equation 11.23 is then used to determine the moles of analyte. To obtain an accurate value for N, therefore, all the current must result in the analyte's oxidation or reduction. In other words, coulometry requires 100% current efficiency (or an accurately measured current efficiency established using a standard), a factor that must be considered in designing a coulometric method of analysis.\n",
            "\n",
            "## 11C.1 Controlled-Potential Coulometry\n",
            "\n",
            "The easiest method for ensuring 100% current efficiency is to maintain the working electrode at a constant potential that allows for the analyte's quantitative oxidation or reduction, without simultaneously oxidizing or reducing an interfering species. The current flowing through an electrochemical cell under a constant potential is proportional to the analyte's concentration. As electrolysis progresses the analyte's concentration decreases, as does the current. The resulting current-versus-time profile  for  controlled-potential coulometry, which also is known as potentiostatic coulometry, is shown in Figure 11.20. Integrating the area under the curve (equation 11.25), from t = 0 until t = t e , gives the total charge. In this section we consider the experimental parameters and instrumentation needed to develop a controlledpotential coulometric method of analysis.\n",
            "\n",
            "Selecting a Constant Potential In controlled-potential coulometry, the potential is selected so that the desired oxidation or reduction reaction goes to completion without interference from redox reactions involving other components of the sample matrix. To see how an appropriate potential for the working electrode is selected, let's develop a constant-potential coulometric method for Cu 2+ based on its reduction to copper metal at a Pt cathode working electrode.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "A ladder diagram for a solution of Cu 2+ (Figure 11.21) provides a useful means for evaluating the solution's redox properties. From the ladder diagram we can see that reaction 11.26 is favored when the working electrode's potential is more negative than +0.342 V versus the SHE (+0.093 V versus the SCE). To maintain a 100% current efficiency, however, the potential must be selected so that the reduction of H3O + to H2 does not contribute significantly to the total charge passed at the electrode.\n",
            "\n",
            "The potential needed for a quantitative reduction of Cu 2+ can be calculated using the Nernst equation\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "If we define a quantitative reduction as one in which 99.99% of the Cu 2+ is reduced to Cu, then the concentration of Cu 2+ at the end of the electrolysis must be\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where [Cu 2+ ]0 is the initial concentration of Cu 2+ in the sample. Substituting equation 11.28 into equation 11.27 gives the desired potential electrode as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "If the initial concentration of Cu 2+ is 1.00 × 10 -4 M, for example, then the cathode's potential must be more negative than +0.105 V versus the SHE (-0.139 V versus the SCE) to achieve a quantitative reduction of Cu 2+ to Cu. Note that at this potential H3O + is  not  reduced to H2, maintaining a 100% current efficiency. Many of the published procedures for the controlled-potential coulometric analysis of Cu 2+ call for potentials that are more negative than that shown for the reduction of H3O + in Figure 11.21. 12 Such potentials can be used, however, because the slow kinetics for reducing H3O + results in a significant overpotential that shifts the potential of the H3O + /H2 redox couple to more negative potentials.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.20\n",
            "\n",
            "Current-time curve for controlled-potential coulometry.\n",
            "\n",
            "Figure 11.21\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Ladder diagram for aqueous solutions of Cu 2+ .\n",
            "\n",
            "## overpotential\n",
            "\n",
            "The difference between the potential actually required to initiate an oxidation or reduction reaction, and the potential predicted by the Nernst equation.\n",
            "\n",
            "Figure 11.22\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Charge-time curve obtained by integrating the current-time curve in Figure 11.20.\n",
            "\n",
            "Minimizing Electrolysis Time The current-time curve for controlled-potential coulometry in Figure 11.20 shows that the current decreases continuously throughout electrolysis. An exhaustive electrolysis, therefore, may require a long time. Since time is an important consideration in choosing and designing analytical methods, the factors that determine the analysis time need to be considered.\n",
            "\n",
            "The change in current as a function of time in controlled-potential coulometry is approximated by an exponential decay; thus, the current at time t is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where i 0 is the initial current, and k is a constant that is directly proportional to the area of the working electrode and the rate of stirring and inversely proportional to the volume of the solution. For an exhaustive electrolysis in which 99.99% of the analyte is oxidized or reduced, the current at the end of the analysis, t e , may be approximated as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Substituting equation 11.30 into equation 11.29 and solving for t e gives the minimum time for an exhaustive electrolysis as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "From this equation we see that increasing k leads to a shorter analysis time. For this reason controlled-potential coulometry is carried out in small-volume electrochemical  cells,  using  electrodes  with  large  surface  areas  and  with  high  stirring  rates.  A quantitative electrolysis typically requires approximately 30-60 min, although shorter or longer times are possible.\n",
            "\n",
            "Instrumentation The potential in controlled-potential coulometry is set using a three-electrode potentiostat. Two types of working electrodes are commonly used: a Pt electrode manufactured from platinum-gauze and fashioned into a cylindrical tube, and an Hg pool electrode. The large overpotential for reducing H3O + at mercury makes it the electrode of choice for analytes requiring negative potentials. For example, potentials more negative than -1 V versus the SCE are feasible at an Hg electrode (but not at a Pt electrode), even in very acidic solutions. The ease with which mercury is oxidized, however, prevents its use at potentials  that  are  positive  with  respect  to  the  SHE.  Platinum  working  electrodes are used when positive potentials are required. The auxiliary electrode, which is often a Pt wire, is separated by a salt bridge from the solution containing the analyte. This is necessary to prevent electrolysis products generated at the auxiliary electrode from reacting with the analyte and interfering in the analysis. A saturated calomel or Ag/AgCl electrode serves as the reference electrode.\n",
            "\n",
            "The other essential feature of instrumentation for controlled-potential coulometry is a means of determining the total charge passed during electrolysis. One method is to monitor the current as a function of time and determine the area under the curve (see Figure 11.20). Modern instruments, however, use electronic integration to monitor charge as a function of time. The total charge at the end of the electrolysis then can be read directly from a digital readout or from a plot of charge versus time (Figure 11.22).\n",
            "\n",
            "## 11C.2 Controlled-Current Coulometry\n",
            "\n",
            "A second approach to coulometry is to use a constant current in place of a constant potential (Figure 11.23). Controlled-current coulometry, also known as amperostatic coulometry or coulometric titrimetry, has two advantages over controlled-potential coulometry. First, using a constant current makes for a more rapid analysis since the current does not decrease over time. Thus, a typical analysis time for controlledcurrent coulometry is less than 10 min, as opposed to approximately 30-60 min for controlled-potential coulometry. Second, with a constant current the total charge is simply the product of current and time (equation 11.24). A method for integrating the current-time curve, therefore, is not necessary.\n",
            "\n",
            "Using a constant current does present two important experimental problems that must be solved if accurate results are to be obtained. First, as electrolysis occurs the analyte's concentration and, therefore, the current due to its oxidation or reduction steadily decreases. To maintain a constant current the cell potential must change until another oxidation or reduction reaction can occur at the working electrode. Unless the system is carefully designed, these secondary reactions will produce a current efficiency of less than 100%. The second problem is the  need  for  a  method  of  determining  when  the  analyte  has  been  exhaustively electrolyzed. In controlled-potential coulometry this is signaled by a decrease in the current to a constant background or residual current (see Figure 11.20). In controlled-current coulometry, however, a constant current continues to flow even when the analyte has been completely oxidized or reduced. A suitable means of determining the end-point of the reaction, t e , is needed.\n",
            "\n",
            "Maintaining Current Efficiency To illustrate why changing the working electrode's potential can lead to less than 100% current efficiency, let's consider the coulometric analysis for Fe 2+ based on its oxidation to Fe 3+ at a Pt working electrode in 1 M H2SO4.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The ladder diagram for this system is shown in Figure 11.24a. Initially the potential of the working electrode remains nearly constant at a level near the standard-state potential for the Fe 3+ /Fe 2+ redox couple. As the concentration of Fe 2+ decreases, however, the potential of the working electrode shifts toward more positive values until another oxidation reaction can provide the necessary current. Thus, in this case the potential eventually increases to a level at which the oxidation of H2O occurs.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Since the current due to the oxidation of H3O + does not contribute to the oxidation of Fe 2+ , the current efficiency of the analysis is less than 100%. To maintain a 100% current efficiency the products of any competing oxidation reactions must react both rapidly and quantitatively with the remaining Fe 2+ . This may be accomplished, for example, by adding an excess of Ce 3+ to the analytical solution (Figure 11.24b). When the potential of the working electrode shifts to a more positive potential, the first species to be oxidized is Ce 3+ .\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The Ce 4+ produced at the working electrode rapidly mixes with the solution, where it reacts with any available Fe 2+ .\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.23\n",
            "\n",
            "Current-time curve for controlled-current coulometry.\n",
            "\n",
            "## Figure 11.24\n",
            "\n",
            "Ladder diagrams for the controlled-current coulometric analysis of Fe 2+ (a) without the addition of Ce 3+ , and (b) with the addition of Ce 3+ . The matrix is 1 M H2SO4 in both cases.\n",
            "\n",
            "## mediator\n",
            "\n",
            "A species that transfers electrons from the electrode to the analyte.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Combining these reactions gives the desired overall reaction of\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "In this manner, a current efficiency of 100% is maintained. Furthermore, since the concentration of Ce 3+ remains at its initial level, the potential of the working electrode remains constant as long as any Fe 2+ is present. This prevents other oxidation reactions, such as that for H2O, from interfering with the analysis. A species, such as Ce 3+ , which is used to maintain 100% current efficiency, is called a mediator.\n",
            "\n",
            "End Point Determination Adding a mediator solves the problem of maintaining 100% current efficiency, but does not solve the problem of determining when the analyte's electrolysis is complete. Using the same example, once all the Fe 2+ has been oxidized current continues to flow as a result of the oxidation of Ce 3+ and, eventually, the oxidation of H2O. What is needed is a means of indicating when the oxidation of Fe 2+ is complete. In this respect it is convenient to treat a controlledcurrent coulometric analysis as if electrolysis of the analyte occurs only as a result of its reaction with the mediator. A reaction between an analyte and a mediator, such as that shown in reaction 11.31, is identical to that encountered in a redox titration. Thus, the same end points that are used in redox titrimetry (see Chapter 9), such as visual indicators, and potentiometric and conductometric measurements, may be used to signal the end point of a controlled-current coulometric analysis. For example, ferroin may be used to provide a visual end point for the Ce 3+ -mediated coulometric analysis for Fe 2+ .\n",
            "\n",
            "Instrumentation Controlled-current coulometry normally is carried out using a galvanostat and an electrochemical cell consisting of a working electrode and a counterelectrode. The working electrode, which often is constructed from Pt, is also\n",
            "\n",
            "called the generator electrode since it is where the mediator reacts to generate the species reacting with the analyte. The counterelectrode is isolated from the analytical solution by a salt bridge or porous frit to prevent its electrolysis products from reacting with the analyte. Alternatively, oxidizing or reducing the mediator can be carried out externally, and the appropriate products flushed into the analytical solution. Figure 11.25 shows one simple method by which oxidizing and reducing agents can be generated externally. A solution containing the mediator flows under the influence of gravity into a small-volume electrochemical cell. The products generated at the anode and cathode pass through separate tubes, and the appropriate oxidizing or reducing reagent can be selectively delivered to the analytical solution. For example, external generation of Ce 4+ can be obtained using an aqueous solution of Ce 3+ and the products generated at the anode.\n",
            "\n",
            "The other necessary instrumental component for controlled-current coulometry is an accurate clock for measuring the electrolysis time, t e , and a switch for starting and stopping the electrolysis. Analog clocks can read time to the nearest ± 0.01 s, but the need to frequently stop and start the electrolysis near the end point leads to a net uncertainty of ± 0.1 s. Digital clocks provide a more accurate measurement of time, with errors of ± 1 ms being possible. The switch must control the flow of current and the clock, so that an accurate determination of the electrolysis time is possible.\n",
            "\n",
            "Coulometric Titrations Controlled-current coulometric methods commonly are called coulometric titrations because of their similarity to conventional titrations. We already have noted, in discussing the controlled-current coulometric determination of Fe 2+ , that the oxidation of Fe 2+ by Ce 4+ is identical to the reaction used in a redox titration. Other similarities between the two techniques also exist. Combining equations 11.23 and 11.24 and solving for the moles of analyte gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Compare this equation with the relationship between the moles of strong acid, N, titrated with a strong base of known concentration.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The titrant in a conventional titration is replaced in a coulometric titration by a constant-current source whose current is analogous to the titrant's molarity. The time needed for an exhaustive electrolysis takes the place of the volume of titrant, and the switch for starting and stopping the electrolysis serves the same function as a buret's stopcock.\n",
            "\n",
            "## 11C. 3 Quantitative Applications\n",
            "\n",
            "Coulometry may be used for the quantitative analysis of both inorganic and organic compounds. Examples of controlled-potential and controlled-current coulometric methods are discussed in the following sections.\n",
            "\n",
            "Controlled-Potential Coulometry The majority of controlled-potential coulometric analyses involve the determination of inorganic cations and anions, including trace metals and halides. Table 11.8 provides a summary of several of these methods.\n",
            "\n",
            "The ability to control selectivity by carefully selecting the working electrode's potential, makes controlled-potential coulometry particularly useful for the analysis of alloys. For example, the composition of an alloy containing Ag, Bi, Cd, and Sb\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.25\n",
            "\n",
            "Method for the external generation of oxidizing and reducing agents in coulometric titrations.\n",
            "\n",
            "## coulometric titrations\n",
            "\n",
            "A titration in which the equivalence point is the time required for a constant current to completely oxidize or reduce the analyte.\n",
            "\n",
            "Table 11.8 Representative Examples for the ControlledPotential Coulometric Analysis of Inorganic Ions\n",
            "\n",
            "| Analyte   | Electrolytic Reaction a   | Electrode   |\n",
            "|-----------|---------------------------|-------------|\n",
            "| Antimony  | Sb(III) + 3 e - t Sb      | Pt          |\n",
            "| Arsenic   | As(III) t As(V) + 2 e -   | Pt          |\n",
            "| Cadmium   | Cd(II) + 2 e - t Cd       | Pt or Hg    |\n",
            "| Cobalt    | Co(II) + 2 e - t Co       | Pt or Hg    |\n",
            "| Copper    | Cu(II) + 2 e - t Cu       | Pt or Hg    |\n",
            "| Halides   | Ag + X - t AgX + e -      | Ag          |\n",
            "| Iron      | Fe(II) t Fe(III) + e -    | Pt          |\n",
            "| Lead      | Pb(II) + 2 e - t Pb       | Pt or Hg    |\n",
            "| Nickel    | Ni(II) + 2 e - t Ni       | Pt or Hg    |\n",
            "| Plutonium | Pu(III) t Pu(IV) + e -    | Pt          |\n",
            "| Silver    | Ag(I) + e - t Ag          | Pt          |\n",
            "| Tin       | Sn(II) + 2 e - t Sn       | Pt          |\n",
            "| Uranium   | U(VI) + 2 e - t U(IV)     | Pt or Hg    |\n",
            "| Zinc      | Zn(II) + 2 e - t Zn       | Pt or Hg    |\n",
            "\n",
            "Source: Compiled from Rechnitz, G. A. Controlled-Potential Analysis. Macmillan: New York, 1963, p. 49. a Electrolytic reactions are written in terms of the change in oxidation state. The actual species in solution depend on the composition of the sample matrix.\n",
            "\n",
            "can be determined by dissolving the sample and placing it in a matrix of 0.2 M H2SO4. A platinum working electrode is immersed in the solution and held at a constant potential of +0.40 V versus the SCE. At this potential Ag(I) deposits on the Pt electrode as Ag, and the other metal ions remain in solution. When electrolysis is complete, the total charge is used to determine the amount of silver in the alloy. The potential of the platinum electrode is then shifted to -0.08 V versus the SCE, depositing Bi on the working electrode. When the coulometric analysis for bismuth is complete, antimony is determined by shifting the working electrode's potential to -0.33 V versus the SCE, depositing Sb. Finally, cadmium is determined following its electrodeposition on the Pt electrode at a potential of -0.80 V versus the SCE.\n",
            "\n",
            "Another area where controlled-potential coulometry has found application is in nuclear chemistry, in which elements such as uranium and polonium can be determined at trace levels. For example, microgram quantities of uranium in a medium of H2SO4 can be determined by reducing U(VI) to U(IV) at a mercury working electrode.\n",
            "\n",
            "Controlled-potential coulometry also can be applied to the quantitative analysis of organic compounds, although the number of applications is significantly less than that for inorganic analytes. One example is the six-electron reduction of a nitro group, -NO2, to a primary amine, -NH2, at a mercury electrode. Solutions of picric acid, for instance, can be analyzed by reducing to triaminophenol.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "| Table 11. 9   | Representative Examples of Coulometric Redox Titrations   | Representative Examples of Coulometric Redox Titrations   | Representative Examples of Coulometric Redox Titrations      |\n",
            "|---------------|-----------------------------------------------------------|-----------------------------------------------------------|--------------------------------------------------------------|\n",
            "| Mediator      | Electrochemically Generated Reagent                       | Generator-Electrode Reaction                              | Representative Application a                                 |\n",
            "| Ag +          | Ag 2+                                                     | Ag + t Ag 2+ + e -                                        | H 2 C 2 O 4 + 2Ag 2+ + 2H 2 O t 2CO 2 + 2Ag + + 2H 3 O +     |\n",
            "| Br -          | Br 2                                                      | 2Br - t Br 2 + 2 e -                                      | H 2 S + Br 2 + 2H 2 O t S + 2Br - + 2H 3 O +                 |\n",
            "| Ce 3+         | Ce 4+                                                     | Ce 3+ t Ce 4+ + e -                                       | Fe(CN) 6 4- + Ce 4+ t Fe(CN) 6 3- + Ce 3+                    |\n",
            "| Cl -          | Cl 2                                                      | 2Cl - t Cl 2 + 2 e -                                      | Tl(I) + Cl 2 t Tl(III) + 2Cl -                               |\n",
            "| Fe 3+         | Fe 2+                                                     | Fe 3+ + e - t Fe 2+                                       | Cr 2 O 7 2- + 6Fe 2+ + 14H 3 O + t 2Cr 3+ + 6Fe 3+ + 21H 2 O |\n",
            "| I -           | I 3 -                                                     | 3I - t I 3 - + 2 e -                                      | 2S 2 O 3 2- + I 3 - t S 4 O 6 2- + 3I                        |\n",
            "| Mn 2+         | Mn 3+                                                     | Mn 2+ t Mn 3+ + e -                                       | As(III) + 2Mn 3+ t As(V) + 2Mn 2+                            |\n",
            "\n",
            "a The analyte is the underlined species in each reaction.\n",
            "\n",
            "Another example is the successive reduction of trichloroacetate to dichloroacetate, and of dichloroacetate to monochloroacetate\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Mixtures of trichloroacetate and dichloroacetate are analyzed by selecting an initial potential at which only the more easily reduced trichloroacetate is reduced. When its electrolysis is complete, the potential is switched to a more negative potential at which dichloroacetate is reduced. The total charge for the first electrolysis is used to determine the amount of trichloroacetate, and the difference in total charge between the first and second electrolyses gives the amount of dichloroacetate.\n",
            "\n",
            "Controlled-Current Coulometry The use of a mediator makes controlled-current coulometry a more versatile analytical method than controlled-potential coulometry. For example, the direct oxidation or reduction of a protein at the working electrode in controlled-potential coulometry is difficult if the protein's active redox site lies  deep  within its structure. The controlled-current coulometric analysis of the protein is made possible, however, by coupling its oxidation or reduction to a mediator that is reduced or oxidized at the working electrode. Controlled-current coulometric methods have been developed for many of the same analytes that may be determined by conventional redox titrimetry. These methods, several of which are summarized in Table 11.9, also are called coulometric redox titrations.\n",
            "\n",
            "Coupling the mediator's oxidation or reduction to an acid-base, precipitation, or complexation reaction involving the analyte allows for the coulometric titration of analytes that are not easily oxidized or reduced. For example, when using H2O as a mediator, oxidation at the anode produces H3O +\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "while reduction at the cathode produces OH - .\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "If  the  oxidation or reduction of H2O is carried out externally using the generator cell shown in Figure 11.25, then H3O + or OH can be dispensed selectively into a solution containing a basic or acidic analyte. The resulting reaction is identical to that in an acid-base titration. Coulometric acid-base titrations have been used for\n",
            "\n",
            "## Table 11.10\n",
            "\n",
            "## Representative Examples of Coulometric Titrations Using Acid-Base, Complexation, and Precipitation Reactions\n",
            "\n",
            "| Type of Reaction   | Mediator    | Electrochemically Generated Reagent   | Generator-Electrode Reaction                      | Representative Reaction a                           |\n",
            "|--------------------|-------------|---------------------------------------|---------------------------------------------------|-----------------------------------------------------|\n",
            "| Acid-Base          | H 2 O       | H 3 O +                               | 6H 2 O t 4H 3 O + + O 2 + 4 e -                   | OH - + H 3 O + t 2H 2 O                             |\n",
            "|                    | H 2 O       | OH -                                  | 2H 2 O+2 e - t 2OH - + H 2                        | H 3 O + + OH - t 2H 2 O                             |\n",
            "| Complexation       | HgNH 3 Y 2- | EDTA (HY 3- )                         | HgNH 3 Y 2- + NH 4 + + 2 e - t HY 3- + Hg + 2NH 3 | Ca 2+ + HY 3- +H 2 O t CaY 2- + H 3 O +             |\n",
            "| Precipitation      | Ag          | Ag +                                  | Ag t Ag + + e -                                   | I - + Ag + t AgI                                    |\n",
            "|                    | Hg          | Hg 2 2+                               | 2Hg t Hg 2 2+ + 2 e -                             | 2Cl - + Hg 2 2+ t Hg 2 Cl 2                         |\n",
            "|                    | Fe(CN) 6 3- | Fe(CN) 6 4-                           | Fe(CN) 6 3- + e - t Fe(CN) 6 4-                   | 3Zn 2+ + 2K + + 2Fe(CN) 6 4- t K 2 Zn 3 [Fe(CN) 6 ] |\n",
            "\n",
            "- a The analyte is the underlined species in each reaction.\n",
            "\n",
            "the analysis of strong and weak acids and bases, in both aqueous and nonaqueous matrices. Examples of coulometric titrations involving acid-base, complexation, and precipitation reactions are summarized in Table 11.10.\n",
            "\n",
            "In comparison with conventional titrimetry, there are several advantages to the coulometric titrations listed in Tables 11.9 and 11.10. One advantage is that the electrochemical generation of a 'titrant' that reacts immediately with the analyte allows the use of reagents whose instability prevents their preparation and storage as a standard solution. Thus, highly reactive reagents such as Ag 2+ and Mn 3+ can be used in coulometric titrations. Because it is relatively easy to measure small quantities of charge, coulometric titrations can be used to determine small quantities of analyte that cannot be measured accurately by a conventional titration.\n",
            "\n",
            "Quantitative Calculations The absolute amount of analyte in a coulometric analysis is determined by applying Faraday's law (equation 11.23) with the total charge during the electrolysis given by equation 11.24 or equation 11.25. Example 11.8 shows the calculations for a typical coulometric analysis.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11.8\n",
            "\n",
            "The purity of a sample of Na2S2O3 was determined by a coulometric redox titration using I as  a  mediator, and I3 -as  the  'titrant.'  A  sample  weighing 0.1342 g is transferred to a 100-mL volumetric flask and diluted to volume with distilled water. A 10.00-mL portion is transferred to an electrochemical cell  along  with  25  mL  of  1  M  KI,  75  mL  of  a  pH  7.0  phosphate  buffer,  and several drops of a starch indicator solution. Electrolysis at a constant current of 36.45 mA required 221.8 s to reach the starch indicator end point. Determine the purity of the sample.\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "From Table 11.9 we see that the coulometric titration of S2O3 2with I3 -is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Oxidizing S2O3 2to S4O6 2requires one electron per S2O3 2-( n = 1). Combining equations 11.23 and 11.24, and making an appropriate substitution for moles of Na2S2O3 gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving for the grams of Na2S2O3 gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "This represents the amount of Na2S2O3 in a 10.00-mL portion of a 100-mL sample, thus 0.1325 g of Na2S2O3 is present in the original sample. The purity of the sample, therefore, is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Note that the calculation is worked as if S2O3 2is  oxidized directly at the working electrode instead of in solution.\n",
            "\n",
            "Representative Method Every controlled-potential or controlled-current coulometric method has its own unique considerations. Nevertheless, the following procedure for the determination of dichromate by a coulometric redox titration provides an instructive example.\n",
            "\n",
            "Representative Methods\n",
            "\n",
            "## Method 11.2 Titration 13\n",
            "\n",
            "## Determination of Dichromate by a Coulometric Redox\n",
            "\n",
            "Description of the Method. The concentration of Cr2O7 2- in a sample is determined by a coulometric redox titration using Fe 3+ as a mediator and electrogenerated Fe 2+ as the 'titrant.' The end point of the coulometric redox titration is determined potentiometrically.\n",
            "\n",
            "Procedure. The electrochemical cell consists of a Pt working electrode and a Pt counterelectrode that is maintained in a separate compartment connected to the analytical solution by a porous glass disk. The counter electrode's compartment is filled with 0.2 M Na2SO4, the level of which must always be maintained above that of the solution in the electrochemical cell. Platinum and tungsten electrodes connected to a potentiometer are used to follow the change in potential during the analysis. A solution of approximately 0.3 M NH4Fe(SO4)2 is prepared for use as the mediator. The coulometric redox titration is carried out by adding 5.00 mL of the sample solution, 2 mL of 9 M H2SO4 and 10-25 mL of the NH4Fe(SO4)2 mediator solution. Distilled water is added as needed until the electrodes are covered. Pure N2 is bubbled through the solution for 15 min to remove traces of O2. A flow of N2 is maintained during the electrolysis, but is turned off momentarily when measuring the potential. A magnetic stir bar is used to stir the solution. Adjust the current to 15-50 mA, and begin the titration. Periodically stop the titration, and measure the potential of the solution. In this manner a titration\n",
            "\n",
            "-Continued\n",
            "\n",
            "Continued from page 505\n",
            "\n",
            "curve of potential versus time is recorded. The time needed to reach the equivalence point is read from the titration curve.\n",
            "\n",
            "## Questions\n",
            "\n",
            "- 1. Does the platinum working electrode serve as the cathode or the anode in this analysis?\n",
            "\n",
            "Reduction of Fe 3+ to Fe 2+ occurs at the working electrode, making it the cathode in the electrochemical cell.\n",
            "\n",
            "- 2. Pure N2 is bubbled through the solution to remove any trace of dissolved O2. Why is this necessary?\n",
            "\n",
            "If O2 is present, some of the electrogenerated Fe 2+ may be oxidized back to Fe 3+ by the reaction\n",
            "\n",
            "4Fe 2+ ( aq ) + O2( g ) + 4H3O + ( aq ) t 4Fe 3+ ( aq ) + 6H2O( l )\n",
            "\n",
            "Any Fe 2+ lost in this fashion must be replaced by the additional reduction of Fe 3+ , reducing the current efficiency and increasing the time needed to reach the titration's end point. The net result is that the reported concentration of Cr2O7 2is too large.\n",
            "\n",
            "- 3. What is the effect on the analysis if the NH4Fe(SO4)2 used to prepare the mediator solution is contaminated with trace amounts of Fe 2+ ?\n",
            "\n",
            "The Fe 2+ introduced when the mediator solution is first added will react with the Cr2O7 2before the electrolysis begins. As a result, the amount of Fe 2+ that must be generated is less than expected, decreasing the time needed to reach the end point of the titration. The current efficiency, therefore, is greater than 100%, and the reported concentration of Cr2O7 2- is too small. Trace amount of Fe 2+ can be removed from the mediator solution by adding H2O2 and heating at 50-70 °C until the evolution of O2 ceases. Alternatively, a blank titration can be used to correct for any Fe 2+ initially present in the mediator.\n",
            "\n",
            "## 11C. 4 Characterization Applications\n",
            "\n",
            "Studies aimed at characterizing the mechanisms of electrode reactions often make use of coulometry for determining the number of electrons involved in the reaction. To make such measurements a known amount of a pure compound is subject to a controlled-potential electrolysis. The coulombs of charge needed to complete the electrolysis are used to determine the value of n using Faraday's law (equation 11.23).\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11. 9\n",
            "\n",
            "A 0.3619-g sample of tetrachloropicolinic acid, C6HNO2Cl4, is dissolved in distilled water, transferred to a 1000-mL volumetric flask, and diluted to volume. An exhaustive controlled-potential electrolysis of a 10.00-mL portion of this solution at a spongy silver cathode requires 5.374 C of charge. What is the value of n for this reduction reaction?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "The 10.00-mL portion of sample contains 3.619 mg, or 1.39 × 10 -5 mol of tetrachloropicolinic acid. Solving equation 11.23 for n and making appropriate substitutions gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Thus, reducing a molecule of tetrachloropicolinic acid requires four electrons. The overall reaction, which results in the selective formation of 3,6dichloropicolinic acid, is\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## 11C. 5 Evaluation\n",
            "\n",
            "Scale of Operation Coulometric methods of analysis can be used to analyze small absolute amounts of analyte. In controlled-current coulometry, for example, the moles of analyte consumed during an exhaustive electrolysis is given by equation 11.32. An electrolysis carried out with a constant current of 100 µ A for 100 s, therefore, consumes only 1 × 10 -7 mol of analyte if n = 1. For an analyte with a molecular weight of 100 g/mol, 1 × 10 -7 mol corresponds to only 10 µ g. The concentration of analyte in the electrochemical cell, however, must be sufficient to allow an accurate determination of the end point. When using visual end points, coulometric titrations require solution concentrations greater than 10 -4 M and, as with conventional titrations, are limited to major and minor analytes. A coulometric titration to a  preset  potentiometric end point is feasible even with solution concentrations of 10 -7 M, making possible the analysis of trace analytes. 14\n",
            "\n",
            "Accuracy The accuracy of a controlled-current coulometric method of analysis is determined by the current efficiency, the accuracy with which current and time can be measured, and the accuracy of the end point. With modern instrumentation the maximum measurement error for current is about ± 0.01%, and that for time is approximately ± 0.1%. The maximum end point error for a coulometric titration is at least as good as that for conventional titrations and is often better when using small quantities of reagents. Taken together, these measurement errors suggest that accuracies of 0.1-0.3% are feasible. The limiting factor in many analyses, therefore, is current efficiency. Fortunately current efficiencies of greater than 99.5% are obtained routinely and often exceed 99.9%.\n",
            "\n",
            "In controlled-potential coulometry, accuracy is determined by current efficiency and the determination of charge. Provided that no interferents are present that are easier to oxidize or reduce than the analyte, current efficiencies of greater than 99.9% are easily obtained. When interferents are present, however, they can often be eliminated by applying a potential such that the exhaustive electrolysis of the interferents is possible without the simultaneous electrolysis of the analyte. Once the interferents have been removed the potential can be switched to a level at\n",
            "\n",
            "## voltammetry\n",
            "\n",
            "An electrochemical method in which we measure current as a function of the applied potential.\n",
            "\n",
            "## voltammogram\n",
            "\n",
            "A plot of current as a function of applied potential.\n",
            "\n",
            "which electrolysis of the analyte is feasible. The limiting factor in the accuracy of many controlled-potential coulometric methods of analysis is the determination of charge. With modern electronic integrators, the total charge can be determined with an accuracy of better than 0.5%.\n",
            "\n",
            "So what is to be done when an acceptable current efficiency is not feasible? If the analyte's oxidation or reduction leads to its deposition on the working electrode, it may be possible to determine the analyte's mass. In this case the working electrode is weighed before beginning the electrolysis and reweighed when electrolysis of the analyte is complete. The difference in the electrode's weight gives the analyte's mass. This technique is known as electrogravimetry. More details about this technique are provided in the suggested reading listed at the end of this chapter.\n",
            "\n",
            "Precision Precision is determined by the uncertainties of measuring current, time, and the end point in controlled-current coulometry and of measuring charge in controlled-potential coulometry. Precisions of ± 0.1-0.3% are routinely obtained for coulometric titrations, and precisions of ± 0.5% are typical for controlled-potential coulometry.\n",
            "\n",
            "Sensitivity For a coulometric method of analysis, the calibration sensitivity is equivalent to nF in equation 11.25. In general, coulometric methods in which the analyte's oxidation or reduction involves a larger value of n show a greater sensitivity.\n",
            "\n",
            "Selectivity Selectivity in controlled-potential and controlled-current coulometry is improved by carefully adjusting solution conditions and by properly selecting the electrolysis potential. In controlled-potential coulometry the potential is fixed by the potentiostat, whereas in controlled-current coulometry the potential is determined by the redox reaction involving the mediator. In either case, the ability to control the potential at which electrolysis occurs affords some measure of selectivity. By adjusting pH or adding a complexing agent, it may be possible to shift the potential at which an analyte or interferent undergoes oxidation or reduction. For example, the standard-state reduction potential for Zn 2+ is  -0.762 V versus the SHE, but shifts to -1.04 for Zn(NH3)4 2+ .  This  provides  an  additional means for controlling selectivity when an analyte and interferent undergo electrolysis at similar potentials.\n",
            "\n",
            "Time, Cost, and Equipment Controlled-potential coulometry is a relatively timeconsuming analysis, with a typical analysis requiring 30-60 min. Coulometric titrations, on the other hand, require only a few minutes and are easily adapted for automated analysis. Commercial instrumentation for both controlled-potential and controlled-current coulometry is available and is relatively inexpensive. Low-cost potentiostats and constant-current sources are available for less than $1000.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Voltammetric Methods of Analysis\n",
            "\n",
            "In voltammetry a time-dependent potential is applied to an electrochemical cell, and the current flowing through the cell is measured as a function of that potential. A plot of current as a function of applied potential is called a voltammogram and is the electrochemical equivalent of a spectrum in spectroscopy, providing quantitative and qualitative information about the species involved in the oxidation or reduction reaction. 15 The earliest voltammetric technique to be introduced was polarography, which was developed by Jaroslav Heyrovsky\n",
            "\n",
            "(1890-1967) in the early 1920s, for which he was awarded the Nobel Prize in chemistry in 1959. Since then, many different forms of voltammetry have been developed, a few of which are noted in Figure 11.1. Before examining these techniques and their applications in more detail, however, we must first consider the basic experimental design for making voltammetric measurements and the factors influencing the shape of the resulting voltammogram.\n",
            "\n",
            "## 11D.1 Voltammetric Measurements\n",
            "\n",
            "Although early voltammetric methods relied on the use of only two electrodes, modern voltammetry makes use of a three-electrode potentiostat, such as that shown in Figure 11.4. A time-dependent potential excitation signal is applied to the working electrode, changing its potential relative to the fixed potential of the reference electrode. The resulting current between the working and auxiliary electrodes is measured. The auxiliary electrode is generally a platinum wire, and the SCE and Ag/AgCl electrode are common reference electrodes.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.26\n",
            "\n",
            "Several different materials have been used as working electrodes, including mercury, platinum, gold, silver, and carbon. The earliest voltammetric techniques, including polarography, used mercury for the working electrode. Since mercury is a liquid, the working electrode often consists of a drop suspended from the end of a capillary tube (Figure 11.26). In the hanging mercury drop electrode, or HMDE, a drop of the desired size is formed by the action of a micrometer screw that pushes the mercury through a narrow capillary tube. In the dropping mercury electrode, or DME, mercury drops form at the end of the capillary tube as a result of gravity. Unlike the HMDE, the mercury drop of a DME grows continuously and has a finite lifetime of several seconds. At the end of its  lifetime  the  mercury drop is dislodged, either manually or by gravity, and replaced by a new drop. The static mercury drop electrode, or SMDE, uses a solenoid-driven plunger to control the flow of mercury. The SMDE can be used as either a hanging mercury drop electrode or as a dropping mercury electrode. A single activation of the solenoid momentarily lifts the plunger, allowing enough mercury to flow through the capillary to form a single drop. To obtain a dropping mercury electrode the solenoid is activated repeatedly. A mercury film electrode consists of a thin layer of mercury deposited on the surface of a solid carbon, platinum, or gold electrode. The solid electrode is placed in a solution of Hg 2+ and held at a potential at which the reduction of Hg 2+ to Hg is favorable, forming a thin mercury film.\n",
            "\n",
            "Mercury has several advantages as a working electrode. Perhaps its most important advantage is its high overpotential for the reduction of H3O + to H2, which allows for the application of potentials as negative as -1 V versus the SCE in acidic solutions, and -2 V versus the SCE in basic solutions. A species such as Zn 2+ , which is difficult to reduce at other electrodes without simultaneously reducing H3O + , is easily reduced at a mercury working electrode. Other advantages include the ability of metals to dissolve in the mercury, resulting in the formation of an amalgam, and the ability to easily renew the surface of the electrode by extruding a new drop. One limitation to its use as a working electrode is the ease with which Hg is oxidized. For this reason, mercury electrodes cannot be used as at potentials more positive than -0.3 V to +0.4 V versus the SCE, depending on the composition of the solution.\n",
            "\n",
            "Mercury electrodes: (a) hanging mercury drop electrode; (b) dropping mercury electrode; (c) static mercury drop electrode.\n",
            "\n",
            "## hanging mercury drop electrode\n",
            "\n",
            "An electrode in which a drop of Hg is suspended from a capillary tube.\n",
            "\n",
            "## dropping mercury electrode\n",
            "\n",
            "An electrode in which successive drops of Hg form at the end of a capillary tube as a result of gravity, with each drop providing a fresh electrode surface.\n",
            "\n",
            "## static mercury drop electrode\n",
            "\n",
            "An electrode in which successive drops of Hg form at the end of a capillary tube as the result of a mechanical plunger, with each drop providing a fresh electrode surface.\n",
            "\n",
            "## amalgam\n",
            "\n",
            "A metallic solution of mercury with another metal.\n",
            "\n",
            "Figure 11.27 Schematic diagram of a solid electrode.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## faradaic current\n",
            "\n",
            "Any current in an electrochemical cell due to an oxidation or reduction reaction.\n",
            "\n",
            "## cathodic current\n",
            "\n",
            "A faradaic current due to a reduction reaction.\n",
            "\n",
            "## anodic current\n",
            "\n",
            "A faradaic current due to an oxidation reaction.\n",
            "\n",
            "Solid electrodes constructed using platinum, gold, silver, or carbon may be used over a range of potentials, including potentials that are negative and positive with respect to the SCE. For example, the potential range for a Pt electrode extends from approximately +1.2 V to -0.2 V versus the SCE in acidic solutions and from +0.7 V to -1 V versus the SCE in basic solutions. Solid electrodes, therefore, can be used in place of mercury for many voltammetric analyses requiring negative potentials and for voltammetric analyses at positive potentials at which mercury electrodes cannot be used. Except for the carbon paste electrode, solid electrodes are fashioned into disks that are sealed into the end of an inert support and are in contact with an electrical lead (Figure 11.27). The carbon paste electrode is made by filling the cavity at the end of the inert support with a paste consisting of carbon particles and a viscous oil. Solid electrodes are not without problems, the most important of which is the ease with which the electrode's surface may be altered by the adsorption of solution species or the formation of oxide layers. For this reason solid electrodes need frequent reconditioning, either by applying an appropriate potential or by polishing.\n",
            "\n",
            "A typical arrangement for a voltammetric electrochemical cell is shown in Figure 11.28. Besides the working, reference, and auxiliary electrodes, the cell also includes a N2 purge line for removing dissolved O2 and an optional stir bar. Electrochemical cells are available in a variety of sizes, allowing for the analysis of solution volumes ranging from more than 100 mL to as small as 50 µ L.\n",
            "\n",
            "## 11D.2 Current in Voltammetry\n",
            "\n",
            "When an analyte is oxidized at the working electrode, a current passes electrons through the external electric circuitry to the auxiliary electrode, where reduction of the solvent or other components of the solution matrix occurs. Reducing an analyte at the working electrode requires a source of electrons, generating a current that flows from the auxiliary electrode to the cathode. In either case, a current resulting from redox reactions at the working and auxiliary electrodes is  called  a faradaic current. In  this  section  we  consider  the  factors  affecting the magnitude of this faradaic current, as well as the source of any nonfaradaic currents.\n",
            "\n",
            "Sign Conventions Since the reaction of interest occurs at the working electrode, the classification of current is based on this reaction. A current due to the analyte's reduction is called a cathodic current and, by convention, is considered positive. Anodic currents are due to oxidation reactions and carry a negative value.\n",
            "\n",
            "Influence of Applied Potential on the Faradaic Current As an example, let's consider the faradaic current when a solution of Fe(CN)6 3is reduced to Fe(CN)6 4at the working electrode. The relationship between the concentrations of Fe(CN)6 3-, Fe(CN)6 4- , and the potential of the working electrode is given by the Nernst equation; thus\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where +0.356 is the standard-state potential for the Fe(CN)6 3/Fe(CN)6 4redox couple, and x = 0 indicates that the concentrations of Fe(CN)6 3and Fe(CN)6 4are\n",
            "\n",
            "those at the surface of the electrode. Surface concentrations are used instead of bulk concentrations since the equilibrium position for the redox reaction\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "can only be established electrochemically at the electrode's surface.\n",
            "\n",
            "Let's assume that we have a solution for which the concentration of Fe(CN)6 3is 1.0 mM and in which Fe(CN)6 4is absent. A ladder diagram for this redox example is  shown in Figure 11.29. If a potential of +0.530 V is applied to the working electrode, the concentrations of Fe(CN)6 3and Fe(CN)6 4at the surface of the electrode are unaffected, and no faradaic current is observed. Switching the potential to +0.356 V, however, requires that\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "which is only possible if a portion of the Fe(CN)6 3at  the  electrode surface is reduced to Fe(CN)6 4- . If this was all that occurred after the potential was applied, the result would be a brief surge of faradaic current that would quickly return to zero. However, although the concentration of Fe(CN)6 4at the electrode surface is 0.50 mM, its concentration in the bulk of solution is zero. As a result, a concentration gradient exists between the solution at the electrode surface and the bulk solution. This concentration gradient creates a driving force that transports Fe(CN)6 4away from the electrode surface (Figure 11.30). The subsequent decrease in the concentration of Fe(CN)6 4at  the  electrode surface requires the further reduction of Fe(CN)6 3- , as well as its transport from bulk solution to the electrode surface. Thus, a faradaic current continues to flow until there is no difference between the concentrations of Fe(CN)6 3and Fe(CN)6 4at  the  electrode surface and their concentrations in the bulk of solution. *\n",
            "\n",
            "Although the applied potential at the working electrode determines if a faradaic current flows, the magnitude of the current is determined by the rate of the resulting oxidation or reduction reaction at the electrode surface. Two factors contribute to the rate of the electrochemical reaction: the rate at which the reactants and products are transported to and from the surface of the electrode, and the rate at which electrons pass between the electrode and the reactants and products in solution.\n",
            "\n",
            "Influence of Mass Transport on the Faradaic Current There are three modes of mass transport that influence the rate at which reactants and products are transported to and from the electrode surface: diffusion, migration, and convection. Diffusion from a region of high concentration to a region of low concentration occurs whenever the concentration of an ion or molecule at the surface of the electrode is different from that in bulk solution. When the potential applied to the working electrode is sufficient to reduce or oxidize the analyte at the electrode surface, a concentration gradient similar to that shown in Figure 11.31 is established. The volume of solution in which the concentration gradient exists is called the diffusion layer. Without other modes of mass transport, the width of the diffusion layer, δ ,  increases with time as the concentration of reactants near the electrode surface decreases. The contribution of diffusion to the rate of mass transport, therefore, is time-dependent.\n",
            "\n",
            "*In voltammetry the working electrode's surface area is significantly smaller than that used in coulometry. Consequently, very little analyte undergoes electrolysis, and the analyte's concentration in bulk solution remains essentially unchanged.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.28\n",
            "\n",
            "Typical electrochemical cell for use in voltammetry.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.29\n",
            "\n",
            "Ladder diagram for the Fe(CN)6 3- /Fe(CN)6 4redox half-reaction.\n",
            "\n",
            "## mass transport\n",
            "\n",
            "The movement of material toward or away from the electrode surface.\n",
            "\n",
            "## diffusion\n",
            "\n",
            "The movement of material in response to a concentration gradient.\n",
            "\n",
            "## diffusion layer\n",
            "\n",
            "The layer of solution adjacent to the electrode in which diffusion is the only means of mass transport.\n",
            "\n",
            "## convection\n",
            "\n",
            "The movement of material in response to a mechanical force, such as stirring a solution.\n",
            "\n",
            "## migration\n",
            "\n",
            "The movement of a cation or anion in response to an applied potential.\n",
            "\n",
            "Electrode\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.30\n",
            "\n",
            "Schematic showing transport of Fe(CN)6 3toward the electrode and Fe(CN)6 4away from the electrode following the reduction of Fe(CN)6 3- .\n",
            "\n",
            "Distance from electrode\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.31\n",
            "\n",
            "Concentration gradients for the analyte in the absence of convection, showing the time-dependent change in diffusion as a method of mass transport.\n",
            "\n",
            "## nonfaradaic current\n",
            "\n",
            "A current in an electrochemical cell that is not the result of a redox reaction.\n",
            "\n",
            "Convection occurs when a mechanical means is used to carry reactants toward the electrode and to remove products from the electrode. The most common means of convection is to stir the solution using a stir bar. Other methods include rotating the electrode and incorporating the electrode into a flow cell.\n",
            "\n",
            "The final mode of mass transport is migration, which occurs when charged particles in solution are attracted or repelled from an electrode that has a positive or negative surface charge. Thus, when the electrode is positively charged, negatively charged particles move toward the electrode, while positively charged particles move toward the bulk solution. Unlike diffusion and convection, migration only affects the mass transport of charged particles.\n",
            "\n",
            "The flux of material to and from the electrode surface is a complex function of all three modes of mass transport. In the limit in which diffusion is the only significant means for the mass transport of the reactants and products, the current in a voltammetric cell is given by\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where n is the number of electrons transferred in the redox reaction, F is Faraday's constant, A is the area of the electrode, D is the diffusion coefficient for the reactant or product, C bulk and Cx= 0 are the concentration of the analyte in bulk solution and at the electrode surface, and δ is the thickness of the diffusion layer.\n",
            "\n",
            "For equation 11.33 to be valid, convection and migration must not interfere with the formation of a diffusion layer between the electrode and the bulk of solution. Migration is eliminated by adding a high concentration of an inert supporting electrolyte to the analytical solution. Ions of similar charge are equally attracted or repelled from the surface of the electrode and, therefore, have an equal probability of undergoing migration. The large excess of inert ions, however, ensures that few reactant and product ions will move as a result of migration. Although convection may be easily eliminated by not physically agitating the solution, in some situations it is desirable either to stir the solution or to push the solution through an electrochemical flow cell. Fortunately, the dynamics of a fluid moving past an electrode results in a small diffusion layer, typically of 0.001 - 0.01-cm thickness, in which the rate of mass transport by convection drops to zero (Figure 11.32).\n",
            "\n",
            "Influence of the Kinetics of Electron Transfer on the Faradaic Current The rate of mass transport is one factor influencing the current in a voltammetric experiment. The ease with which electrons are transferred between the electrode and the reactants and products in solution also affects the current. When electron transfer kinetics are fast, the redox reaction is at equilibrium, and the concentrations of reactants and products at the electrode are those specified by the Nernst equation. Such systems are considered electrochemically reversible. In other systems, when electron transfer kinetics are sufficiently slow, the concentration of reactants and products at the electrode surface, and thus the current, differ from that predicted by the Nernst equation. In this case the system is electrochemically irreversible.\n",
            "\n",
            "Nonfaradaic Currents Faradaic currents result from a redox reaction at the electrode surface. Other currents may also exist in an electrochemical cell that are unrelated to any redox reaction. These currents are called nonfaradaic currents and must be accounted for if the faradaic component of the measured current is to be determined.\n",
            "\n",
            "The most important example of a nonfaradaic current occurs whenever the electrode's potential is changed. In discussing migration as a means of mass transport, we noted that negatively charged particles in solution migrate toward a positively charged electrode, and positively charged particles move away from the same electrode. When an inert electrolyte is responsible for migration, the result is a structured electrode-surface interface called the electrical double layer, or EDL, the exact structure of which is of no concern in the context of this text. The movement of charged particles in solution, however, gives rise to a short-lived, nonfaradaic charging current. Changing the potential of an electrode causes a change in the structure of the EDL, producing a small charging current.\n",
            "\n",
            "Residual Current Even in the absence of analyte, a small current inevitably flows through an electrochemical cell. This current, which is called the residual current, consists of two components: a faradaic current due to the oxidation or reduction of trace impurities, and the charging current. Methods for discriminating between the faradaic current due to the analyte and the residual current are discussed later in this chapter.\n",
            "\n",
            "## 11D. 3 Shape of Voltammograms\n",
            "\n",
            "The shape of a voltammogram is determined by several experimental factors, the most important of which are how the current is measured and whether convection is included as a means of mass transport. Despite an abundance of different voltammetric techniques, several of which are discussed in this chapter, only three shapes are common for voltammograms (Figure 11.33).\n",
            "\n",
            "The voltammogram in Figure 11.33a is characterized by a current that increases from the background residual current to a limiting current at potentials at which the analyte is oxidized or reduced. Since the magnitude of a faradaic current is inversely proportional to δ (equation 11.33), a limiting current implies that the thickness of the diffusion layer remains constant. The simplest method for obtaining a limiting current is to stir the solution (Figure 11.32), which can be accomplished with a magnetic stir bar, or by rotating the electrode. Voltammetric techniques that include convection by stirring are called hydrodynamic voltammetry. When convection is absent, the thickness of the diffusion layer increases with time (Figure 11.31), resulting in a peak current in place of a limiting current (Figure 11.33b).\n",
            "\n",
            "In the voltammograms in Figures 11.33a and 11.33b, the current is monitored as a function of the applied potential. Alternatively, the change in current following\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.33\n",
            "\n",
            "Three common shapes for voltammograms.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.32\n",
            "\n",
            "Concentration gradient for the analyte showing the effects of diffusion and convection as methods of mass transport.\n",
            "\n",
            "## electrical double layer\n",
            "\n",
            "The interface between a positively or negatively charged electrode and the negatively or positively charged layer of solution in contact with the electrode.\n",
            "\n",
            "## charging current\n",
            "\n",
            "A current in an electrochemical cell due to the electrical double layer's formation.\n",
            "\n",
            "## residual current\n",
            "\n",
            "The current in an electrochemical cell that is present even in the absence of the analyte.\n",
            "\n",
            "## hydrodynamic voltammetry\n",
            "\n",
            "A form of voltammetry in which the solution is mechanically stirred.\n",
            "\n",
            "a  change in potential may be measured. The resulting voltammogram, which is shown in Figure 11.33c, also is characterized by a peak current.\n",
            "\n",
            "## 11D. 4 Quantitative and Qualitative Aspects of Voltammetry\n",
            "\n",
            "Earlier we described a voltammogram as the electrochemical equivalent of a spectrum in spectroscopy. In this section we consider how quantitative and qualitative information may be extracted from a voltammogram. Quantitative information is obtained by relating current to the concentration of analyte in the bulk solution. Qualitative information is obtained from the voltammogram by extracting the standard-state potential for the redox reaction. For simplicity we only consider voltammograms similar to that shown in Figure 11.33a.\n",
            "\n",
            "Determining Concentration Let's assume that the redox reaction at the working electrode is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and that initially only O is present in the bulk solution. The current is determined by the rate at which O diffuses through the fixed diffusion layer (see Figure 11.32), and is given by equation 11.33, or\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where K O is a constant equal to nFAD O/ δ . When the limiting current is reached, the concentration of O at the electrode surface is zero, and this equation simplifies to\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Thus, the limiting current, i lim , is a linear function of the concentration of O in bulk solution, and a quantitative analysis is possible using any of the standardization methods discussed in Chapter 5. Equations similar to equation 11.35 can be developed for other forms of voltammetry, in which peak currents are related to the analyte's concentration in bulk solution.\n",
            "\n",
            "Determining the Standard-State Potential To extract the standard-state potential, or formal potential, for reaction 11.34 from a voltammogram, it is necessary to rewrite the Nernst equation\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "in terms of current instead of the concentration of O and R. Substituting equation 11.36 into equation 11.35 and rearranging gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "To derive a similar equation for the concentration of R at the electrode surface we note that\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Since the concentration of R in bulk solution is zero, this equation simplifies to\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Substituting equations 11.39 and 11.38 into equation 11.37 and rearranging gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "When the current is half that of the limiting current\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "equation 11.40 simplifies to\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where E 1/2 is  the  half-wave potential (Figure 11.34). If K O is  approximately equal to K R, which is often the case, then the half-wave potential is equal to the standard-state potential. Note that equation 11.41 is only valid if the redox reaction is electrochemically reversible. Voltammetric techniques giving peak potentials also can be used to determine a redox reaction's standardstate potential.\n",
            "\n",
            "Current\n",
            "\n",
            "## 11D. 5 Voltammetric Techniques\n",
            "\n",
            "A number of voltammetric experiments are routinely used in quantitative and qualitative analyses. Several of these methods are briefly described in this section.\n",
            "\n",
            "Polarography The earliest voltammetric experiment was normal polarography at a  dropping mercury electrode. In normal polarography the potential is linearly scanned, producing voltammograms such as that shown in Figure 11.35. Although polarography takes place in an unstirred solution, a limiting current is obtained because the falling Hg drops mix the solution. Each new Hg drop, therefore, grows in  a  solution  whose composition is identical to that of the initial bulk solution. Oscillations in the current are due to the growth of the Hg drop, which leads to a\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "(a)\n",
            "\n",
            "Potential\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.34\n",
            "\n",
            "Determination of limiting current and halfwave potential in linear scan hydrodynamic voltammetry.\n",
            "\n",
            "## polarography\n",
            "\n",
            "A form of voltammetry using a dropping mercury electrode or a static mercury drop electrode.\n",
            "\n",
            "## Figure 11.35\n",
            "\n",
            "Potential-excitation signal and voltammogram for normal polarography.\n",
            "\n",
            "## stripping voltammetry\n",
            "\n",
            "A form of voltammetry in which the analyte is first deposited on the electrode and then removed, or 'stripped,' electrochemically while monitoring the current as a function of the applied potential.\n",
            "\n",
            "time-dependent change in the area of the working electrode. The limiting current, which is also called the diffusion current, may be measured from the maximum current, i max, or from the average current, i avg .  The relationship between the concentration of analyte, C A, and the limiting current is given by the Ilikovic equation\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where n is  the  number of electrons transferred in the redox reaction, D is  the analyte's diffusion coefficient, m is the flow rate of the Hg, and t is the drop time. The half-wave potential, E 1/2 ,  provides qualitative information about the redox reaction (see Appendix 3E for a list of selected polarographic half-wave potentials).\n",
            "\n",
            "Normal polarography has been replaced by various forms of pulse polarography, 16 several examples of which are shown in Figure 11.36. Differential pulse polarography (Figure 11.36b), for example, uses a series of potential pulses characterized by a cycle of time τ , a pulse time of t p, a potential pulse of ∆ E p, and a potential step per cycle of ∆ E s .  Typical  experimental conditions for differential pulse polarography are τ ≈ 1 s, t p ≈ 50 ms, ∆ E p ≈ 50 mV, ∆ E s ≈ 2 mV. The current is measured twice, for approximately 17 ms before the forward pulse and for approximately 17 ms before the reverse pulse. The difference in the two currents gives rise to a peak-shaped voltammogram. Other forms of pulse polarography include normal pulse polarography (Figure 11.36a), staircase polarography (Figure 11.36c), and square-wave polarography (Figure 11.36d). Limiting and peak currents are directly proportional to the concentration of analyte, and half-wave and peak potentials can be used for qualitative purposes. The popularity of pulse polarography is due to a substantial improvement in sensitivity and detection limits from those in normal polarography.\n",
            "\n",
            "Polarography is used extensively for the analysis of metal ions and inorganic anions, such as IO3 -and NO3 - . Organic compounds containing easily reducible or oxidizable functional groups also can be studied polarographically. Functional groups that have been used include carbonyls, carboxylic acids, and carbon-carbon double bonds.\n",
            "\n",
            "Hydrodynamic Voltammetry In polarography a limiting current is obtained because each falling drop of mercury returns the solution near the electrode to its initial composition. As noted earlier, a limiting current is also obtained whenever the solution is stirred during the analysis. The simplest means of stirring the solution is with a magnetic stir bar. More commonly, however, stirring is achieved by rotating the electrode.\n",
            "\n",
            "In hydrodynamic voltammetry current is measured as a function of the potential applied to a solid working electrode. The same potential profiles used for polarography, such as a linear scan or a differential pulse, are used in hydrodynamic voltammetry. The resulting voltammograms are identical to those for polarography, except for the lack of current oscillations resulting from the growth of the mercury drops. Because hydrodynamic voltammetry is not limited to Hg electrodes, it is useful for the analysis of analytes that are reduced or oxidized at more positive potentials.\n",
            "\n",
            "Stripping Voltammetry One of the most important quantitative voltammetric techniques is stripping voltammetry, which is composed of three related techniques: anodic, cathodic, and adsorptive stripping voltammetry. Since anodic strip-\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "ping voltammetry has found the widest application, we consider it in the greatest detail.\n",
            "\n",
            "Anodic stripping voltammetry consists of two steps (Figure 11.37). The first is a controlled potential electrolysis in which the working electrode, usually a hanging mercury drop or mercury film, is held at a cathodic potential sufficient to deposit the metal ion on the electrode. For example, with Cu 2+ the deposition reaction is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "## Figure 11.36\n",
            "\n",
            "Potential-excitation signals and voltammograms for (a) normal pulse polarography, (b) differential pulse polarography, (c) staircase polarography, and (d) square-wave polarography. See text for an explanation of the symbols. Current is sampled at the time intervals indicated by the solid circles ( · ).\n",
            "\n",
            "## Figure 11.37\n",
            "\n",
            "Potential-excitation signal and voltammogram for anodic stripping voltammetry at a hanging mercury drop electrode.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "where Cu(Hg) indicates that the copper is amalgamated with the mercury. This step essentially serves as a means of preconcentrating the analyte from the larger volume of the solution to the smaller volume of the electrode. The solution is stirred during electrolysis to increase the rate of deposition. Near the end of the deposition time stirring is stopped, eliminating convection as a mode of mass transport. Deposition times of 1-30 min are common, with longer times being used for analytes at lower concentrations.\n",
            "\n",
            "In the second step, the potential is scanned anodically toward more positive potentials. When the potential of the working electrode is sufficiently positive the analyte is stripped from the electrode, returning to solution as its oxidized form\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The current during the stripping step is monitored as a function of potential, giving rise  to  peak-shaped voltammograms similar to that shown in Figure 11.37. The peak current is proportional to the analyte's concentration in the solution.\n",
            "\n",
            "Anodic stripping voltammetry is very sensitive to experimental conditions, which must be carefully controlled if results are to be accurate and precise. Key variables include the area of the mercury film electrode or the size of the Hg drop when using a hanging mercury drop electrode, the deposition time, the rest time, the rate of stirring, and the scan rate during the stripping step. Anodic stripping voltammetry is best used for metals that form amalgams with mercury, several examples of which are listed in Table 11.11.\n",
            "\n",
            "The experimental design for cathodic stripping voltammetry is similar to that for anodic stripping voltammetry with two exceptions. First, the deposition step in-\n",
            "\n",
            "Table 11.11\n",
            "\n",
            "## Representative Examples of Analytes Determined by Stripping Voltammetry\n",
            "\n",
            "| Anodic Stripping Voltammetry                                 | Cathodic Stripping Voltammetry                               | Absorptive Stripping Voltammetry                                        |\n",
            "|--------------------------------------------------------------|--------------------------------------------------------------|-------------------------------------------------------------------------|\n",
            "| bismuth cadmium copper gallium indium lead thallium tin zinc | bromide chloride iodide mercaptans (RSH) sulfide thiocyanate | bilirubin codeine cocaine digitoxin dopamine heme monensin testosterone |\n",
            "\n",
            "Source: Compiled from Peterson, W. M.; Wong, R. V. Am. Lab. November 1981, 116-128; and Wang, J. Am. Lab. May 1985, 41-50. 17,18\n",
            "\n",
            "volves the oxidation of the Hg electrode to Hg2 2+ , which then reacts with the analyte to form an insoluble film at the surface of the electrode. For example, when Cl is the analyte the deposition step is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Second, stripping is accomplished by scanning cathodically toward a more negative potential, reducing Hg2 2+ back to Hg and returning the analyte to solution.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Table 11.11 lists several analytes that have been analyzed successfully by cathodic stripping voltammetry.\n",
            "\n",
            "In adsorptive stripping voltammetry the deposition step occurs without electrolysis.  Instead,  the  analyte  adsorbs to the electrode's surface. During deposition the electrode is maintained at a potential that enhances adsorption. For example, adsorption of a neutral molecule on a Hg drop is enhanced if the electrode is held at -0.4 V versus the SCE, a potential at which the surface charge of mercury is approximately zero. When deposition is complete the potential is scanned in an anodic or cathodic direction depending on whether we wish to oxidize or reduce the analyte. Examples of compounds that have been analyzed by absorptive stripping voltammetry also are listed in Table 11.11.\n",
            "\n",
            "Amperometry The final voltammetric technique to be considered is amperometry, in which a constant potential is applied to the working electrode, and current is measured as a function of time. Since the potential is not scanned, amperometry does not lead to a voltammogram.\n",
            "\n",
            "One important application of amperometry is in the construction of chemical sensors. One of the first amperometric sensors to be developed was for dissolved O2 in blood, which was developed in 1956 by L. C. Clark. The design of the amperometric sensor is shown in Figure 11.38 and is similar to potentiometric membrane electrodes. A gas-permeable membrane is stretched across the end of the sensor and is separated from the working and counter electrodes by a thin solution of  KCl.  The working electrode is a Pt disk cathode, and an Ag ring anode is the\n",
            "\n",
            "## amperometry\n",
            "\n",
            "A form of voltammetry in which we measure current as a function of time while maintaining a constant potential.\n",
            "\n",
            "Figure 11.38\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Clark amperometric sensor for the determination of dissolved O2.\n",
            "\n",
            "counter electrode. Although several gases can diffuse across the membrane, including O2, N2, and CO2, only oxygen is reduced at the cathode.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Another example of an amperometric sensor is the glucose sensor. In this case the single membrane in Figure 11.38 is replaced with three membranes. The outermost membrane is of polycarbonate, which is permeable to glucose and O2. The second membrane contains an immobilized preparation of glucose oxidase that catalyzes the oxidation of glucose to gluconolactone and hydrogen peroxide.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The hydrogen peroxide then diffuses through the innermost membrane of cellulose acetate, where it is oxidized at a Pt anode.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Figure 11.39 summarizes the reactions taking place in this amperometric sensor. FAD is the oxidized form of flavin adenine nucleotide (the active site of the enzyme glucose oxidase), and FADH2 is  the  active  site's  reduced  form. Note that O2 serves as a mediator, carrying electrons to the electrode. Other mediators, such as Fe(CN)6 3, can be used in place of O2.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "By changing the enzyme and mediator, the amperometric sensor  in  Figure  11.39  is  easily  extended  to  the  analysis  of  other  substrates.  Other  bioselective  materials  may  be  incorporated  into  amperometric sensors. For example, a CO2 sensor has been developed using an amperometric O2 sensor with a two-layer membrane, one of which contains an immobilized preparation of autotrophic bacteria. 19 As CO2 diffuses through the membranes, it is converted to O2 by the bacteria, increasing the concentration of O2 at  the  Pt cathode.\n",
            "\n",
            "Electrode\n",
            "\n",
            "## Figure 11.39\n",
            "\n",
            "Schematic showing the reactions by which an amperometric biosensor responds to glucose.\n",
            "\n",
            "## 11D.6 Quantitative Applications\n",
            "\n",
            "Quantitative voltammetry has been applied to a wide variety of sample types, including environmental samples, clinical samples, pharmaceutical formulations, steels, gasoline, and oil.\n",
            "\n",
            "Selecting the Voltammetric Technique The choice of which voltammetric technique to use depends on the sample's characteristics, including the analyte's expected concentration and the location of the sample. Amperometry is best suited for use as a detector in flow systems or as a selective sensor for the rapid analysis of a single analyte. The portability of amperometric sensors, which are similar to potentiometric sensors, make them ideal for field studies.\n",
            "\n",
            "Pulse polarography and stripping voltammetry can frequently be used interchangeably, although each has its advantages and disadvantages. Pulse polarography is better for analyzing a wider range of inorganic and organic analytes because the need to preconcentrate the analyte at the electrode surface restricts the application of anodic and cathodic stripping voltammetry.\n",
            "\n",
            "When either pulse polarography or anodic stripping voltammetry can be used, the selection is often based on the analyte's expected concentration and the desired\n",
            "\n",
            "accuracy and precision. Detection limits for normal pulse polarography generally are on the order of 10 -6 -10 -7 M, whereas those for differential pulse polarography, staircase, and square-wave polarography are between 10 -7 M and 10 -8 M. Preconcentrating the analyte in stripping voltammetry lowers the detection limit for many analytes to as little as 10 -10 M. On the other hand, the current in stripping voltammetry is much more sensitive than pulse polarography to changes in experimental conditions, which may lead to poorer precision and accuracy.\n",
            "\n",
            "Anodic stripping voltammetry also suffers from occasional interferences when two metals, such as Cu and Zn, combine to form an intermetallic compound in the mercury amalgam. The deposition potential for Zn 2+ is sufficiently negative that any Cu 2+ present in the sample is also deposited. After deposition, intermetallic compounds such as CuZn and CuZn2 form within the mercury amalgam. During the stripping step, the zinc in the intermetallic compounds strips  at  potentials  near  that  of  copper,  decreasing  the  current  for  zinc  and  increasing the current for copper. This problem can often be overcome by adding a third element that forms a stronger intermetallic compound with the interfering metal. Thus, adding Ga 3+ minimizes this problem by forming an intermetallic compound of Cu and Ga.\n",
            "\n",
            "Correcting for Residual Current In any quantitative analysis the signal due to the analyte must be corrected for signals arising from other sources. The total measured current in any voltammetric experiment, i tot ,  consists of two parts: that due to the analyte's oxidation or reduction, i a , and a background, or residual, current, i r .\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The residual current, in turn, has two sources. One source is a faradaic current due to the oxidation or reduction of trace impurities in the sample, i i . The other source is the charging current, i ch , that is present whenever the working electrode's potential changes.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Faradaic currents due to impurities can usually be minimized by carefully preparing the sample. For example, one important impurity is dissolved O2, which is reduced first to H2O2 and then to H2O. Dissolved O2 is  removed by bubbling an inert gas such as N2 through the sample before the analysis.\n",
            "\n",
            "Two methods are commonly used to correct for the residual current. One method is to extrapolate the total measured current when the analyte's faradaic current is zero. This is the method shown in the voltammograms included in this chapter. The advantage of this method is that it does not require any additional data. On the other hand, extrapolation assumes that changes in the residual current with potential are predictable, which often is not the case. A second, and more rigorous, approach is to obtain a voltammogram for an appropriate blank. The blank's residual current is then subtracted from the total current obtained with the sample.\n",
            "\n",
            "Analysis for Single Components The analysis of samples containing only a single electroactive analyte is straightforward. Any of the standardization methods discussed in Chapter 5 can be used to establish the relationship between current and the concentration of analyte.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11.10\n",
            "\n",
            "The concentration of As(III) in water can be determined by differential pulse polarography in 1 M HCl. The initial potential is set to -0.1 V versus the SCE, and is scanned toward more negative potentials at a rate of 5 mV/s. Reduction of  As(III)  to  As(0)  occurs at a potential of approximately -0.44 V versus the SCE. The peak currents, corrected for the residual current, for a set of standard solutions are shown in the following table.\n",
            "\n",
            "| [As(III)] (M)   |   i p ( µ A) |\n",
            "|-----------------|--------------|\n",
            "| 1.00 × 10 -6    |        0.298 |\n",
            "| 3.00 × 10 -6    |        0.947 |\n",
            "| 6.00 × 10 -6    |        1.83  |\n",
            "| 9.00 × 10 -6    |        2.72  |\n",
            "\n",
            "What is the concentration of As(III) in a sample of water if the peak current under the same conditions is 1.37 µ A?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "Linear regression gives the equation for the calibration curve as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Substituting the sample's peak current into the regression equation gives the concentration of As(III) as 4.49 × 10 -6 M.\n",
            "\n",
            "## EXAMPLE 11.11\n",
            "\n",
            "The concentration of copper in a sample of sea water is determined by anodic stripping voltammetry using the method of standard additions. When a 50.0-mL sample is analyzed, the peak current is 0.886 µ A. A 5.00µ L spike of 10.0-ppm Cu 2+ is added, giving a peak current of 2.52 µ A. Calculate the parts per million of copper in the sample of sea water.\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "Peak currents in anodic stripping voltammetry are a linear function of concentration\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where k is a constant. Thus, for the sample we write\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and for the standard addition\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving the first equation for k, substituting into the second equation, and simplifying gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving gives the concentration of Cu 2+ as 5.42 10 -4 ppm, or 0.542 ppb.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Multicomponent Analysis One advantage of voltammetry as a quantitative method of analysis is its capability for analyzing two or more analytes in a single sample. As long as the components behave independently, the resulting voltammogram for a multicomponent mixture is a summation of their respective individual voltammograms. If the separation between the half-wave potentials or peak potentials is sufficient, each component can be determined independently as if it were the only component in the sample (Figure 11.40). The minimum separation between the half-wave potentials or peak potentials for the independent analysis of two components depends on several factors, including the type of electrode and the potential-excitation signal. For normal polarography the separation must be at least ± 0.2-0.3 V, and for differential pulse voltammetry a minimum separation of ± 0.04-0.05 V is needed.\n",
            "\n",
            "When the overlap between the voltammograms for two components prevents their independent analysis, a simultaneous analysis similar to that used in spectrophotometry may be possible. An example of this approach is outlined in Example 11.12.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11.12\n",
            "\n",
            "The differential pulse polarographic analysis of mixtures of indium and cadmium in 0.1 M HCl is complicated by the overlap of their respective voltammograms. 20 The peak potential for indium is at -0.557 V, and that for cadmium occurs at a potential of -0.597 V. When a 0.800-ppm indium standard is analyzed, the peak current (in arbitrary units) is found to be 200.5 at -0.557 V and 87.5 at -0.597 V. A standard solution of 0.793-ppm cadmium gives peak currents of 58.5 at -0.557 V and 128.5 at -0.597 V. What is the concentration of indium and cadmium in a sample if the peak current is 167.0 at a potential of -0.557 V and 99.5 at a potential of -0.597 V?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "Peak currents in differential pulse polarography are a linear function of the concentration of analyte; thus\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where k is  a  constant that depends on the analyte and the applied potential. To determine the concentrations of indium and cadmium in the sample, we must first find the values of k for each analyte at each potential. For simplicity we will identify the potential of -0.557 V as E 1, and that for -0.597 V as E 2. The values of k are\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Figure 11.40\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Voltammograms showing the independent analysis of two components.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Next, we write simultaneous equations for the current at the two potentials\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "i E 2 = 109.4 ppm -1 (ppm In) + 162.0 ppm -1 (ppm Cd) = 99.5\n",
            "\n",
            "Solving the simultaneous equations, which is left as an exercise, gives the concentration of indium as 0.606 ppm and the concentration of cadmium as 0.205 ppm.\n",
            "\n",
            "Environmental Samples One area in which quantitative voltammetry has had a significant effect is in the analysis of trace metals in environmental samples. The most common samples are natural waters, including groundwater, lakes, rivers and streams, sea water, rain, and snow. Concentrations of trace metals at the parts-perbillion level can be determined using differential pulse polarography, whereas with anodic stripping voltammetry the determination of trace metals at the pptr (partsper-trillion) level is possible. The combination of low detection limits and the capability for the simultaneous analysis of several analytes makes differential pulse polarography and anodic stripping voltammetry ideally suited for such samples.\n",
            "\n",
            "One interesting application of anodic stripping voltammetry to the analysis of natural waters is the determination of the speciation, or chemical form, of the trace metals. The speciation of a trace metal is important because its bioavailability, toxicity, and ease of transport through the environment often depend on its chemical form. For example, trace metals strongly bound to colloidal particles are generally not available to aquatic lifeforms and, therefore, are not toxic. Unfortunately, anodic stripping voltammetry cannot distinguish the exact chemical form of a trace metal as closely related species, such as Pb 2+ and PbCl + , yield only a single stripping peak. Instead, trace metals are divided into several 'operationally defined' categories that have environmental significance.\n",
            "\n",
            "Several speciation schemes have been developed, 21 but we consider only the speciation scheme proposed by Batley and Florence, 22 which uses a combination of anodic stripping voltammetry with ion-exchange and UV irradiation to divide soluble trace metals into seven classes. Anodic stripping voltammetry in a pH 4.8 acetic acid buffer is used to distinguish labile metals present as hydrated ions, weakly bound complexes, or weakly adsorbed on colloidal surfaces from those metals that are bound in stronger complexes or strongly adsorbed. Only those metals that are hydrated, weakly bound, or weakly adsorbed deposit at the electrode. Ion exchange and UV irradiation are used to further subdivide the trace metals. A Chelex-100 ion-exchange resin is used to distinguish between ionic metals and strongly bound metals, whereas UV radiation is used to separate metals bound to organic and inorganic phases. Table 11.12 shows how trace metals are divided into seven classes using these three experimental techniques. The analysis of sea water samples, for example, showed that cadmium, copper, and lead were primarily present as labile  organic complexes or as labile adsorbates on organic colloids (see group II, Table 11.12).\n",
            "\n",
            "Differential pulse polarography and stripping voltammetry have been applied to the analysis of trace metals in airborne particulates, incinerator fly ash, rocks,\n",
            "\n",
            "## Table 11.12 Speciation of Soluble Trace Metals According to the Scheme of Batley and Florence 22\n",
            "\n",
            "| Method         | Speciation of Soluble Metals                                              | Speciation of Soluble Metals                                    | Speciation of Soluble Metals                                        | Speciation of Soluble Metals   | Speciation of Soluble Metals   | Speciation of Soluble Metals                                          | Speciation of Soluble Metals                                              |\n",
            "|----------------|---------------------------------------------------------------------------|-----------------------------------------------------------------|---------------------------------------------------------------------|--------------------------------|--------------------------------|-----------------------------------------------------------------------|---------------------------------------------------------------------------|\n",
            "| ASV            | Labile Metals                                                             | Labile Metals                                                   | Labile Metals                                                       | Nonlabile or Bound Metals      | Nonlabile or Bound Metals      | Nonlabile or Bound Metals                                             | Nonlabile or Bound Metals                                                 |\n",
            "| Ion-Exchange   | Removed                                                                   | Not Removed                                                     | Not Removed                                                         | Removed                        | Removed                        | Not Removed                                                           | Not Removed                                                               |\n",
            "| UV Irradiation |                                                                           | Released                                                        | Not Released                                                        | Released                       | Not Released                   | Released                                                              | Not Released                                                              |\n",
            "| Description    | free metal ions, labile organic complexes, and labile inorganic complexes | labile organic complexes and labile metals absorbed on organics | labile inorganic complexes and labile metals absorbed on inorganics | nonlabile organic complexes    | nonlabile inorganic complexes  | nonlabile organic complexes and nonlabile metals absorbed on organics | nonlabile inorganic complexes and nonlabile metals absorbed on inorganics |\n",
            "| Group          | I                                                                         | II                                                              | III                                                                 | IV                             | V                              | VI                                                                    | VII                                                                       |\n",
            "\n",
            "minerals, and sediments. The trace metals, of course, must be brought into solution by digesting or extracting before the voltammetric analysis.\n",
            "\n",
            "Amperometric sensors also are used to analyze environmental samples. For example, the dissolved O2 sensor described earlier is routinely used for the determination of dissolved oxygen and biochemical oxygen demand, or BOD, in waters and wastewaters. The latter test, which is a measure of the amount of oxygen required by aquatic bacteria during the decomposition of organic matter, is of importance in evaluating the efficiency of wastewater treatment plants and in monitoring organic pollution in natural waters. A high BOD corresponds to a high concentration of organic material that may seriously deplete the level of dissolved oxygen in the water. Other amperometric sensors have been developed to monitor anionic surfactants in water, and CO2, H2SO4, and NH3 in atmospheric gases. 19\n",
            "\n",
            "Clinical Samples Differential pulse polarography and stripping voltammetry have been used to determine the concentration of trace metals in a variety of matrices, including blood, urine, and tissue samples. The determination of lead in blood is of considerable interest due to concerns about lead poisoning. Because the concentration of lead in blood is so small, anodic stripping voltammetry frequently is the method of choice. The analysis is complicated, however, by the presence of proteins that may adsorb at the surface of the mercury electrode, inhibiting either the deposition or stripping of lead. In addition, proteins may prevent the electrodeposition of lead through the formation of stable, nonlabile complexes. For these reasons samples of whole blood must be digested or ashed before the analysis. Differential  pulse  polarography is one of the few techniques that can be used for the routine quantitative analysis of drugs in biological fluids at concentrations of less than 10 -6 M. 23 Amperometric sensors based on enzyme catalysts also have wide applicability. 24 Table 11.13 provides a partial list of enzymatic amperometric sensors.\n",
            "\n",
            "Miscellaneous Samples Besides environmental and clinical samples, differential pulse polarography and stripping voltammetry have been used for the analysis of trace metals in other samples, including food, steels and other alloys, gasoline, gunpowder residues, and pharmaceuticals. Voltammetry is also an important tool for\n",
            "\n",
            "## Table 11.1 3\n",
            "\n",
            "## Representative Examples of Amperometric Biosensors\n",
            "\n",
            "| Analyte                                                                            | Enzyme                                                                                                                                                                                       | Species Detected                                                     |\n",
            "|------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------|\n",
            "| choline ethanol formaldehyde glucose glutamine glycerol lactate phenol inorganic P | choline oxidase alcohol oxidase formaldehyde dehydrogenase glucose oxidase glutaminase, glutamate oxidase glycerol dehydrogenase lactate oxidase polyphenol oxidase nucleoside phosphorylase | H 2 O 2 H 2 O 2 NADH a H 2 O 2 H 2 O 2 NADH, O 2 H 2 O 2 quinone O 2 |\n",
            "\n",
            "Source: Compiled from Cammann, K.; Lemke, U.; Rohen, A.; et al. Angew. Chem. Int. Ed. Engl. 1991, 30, 516-539.\n",
            "\n",
            "a NADH is the reduced form of nicotinamide adenine dinucleotide.\n",
            "\n",
            "the quantitative analysis of organics, particularly in the pharmaceutical industry, in which it is used to determine the concentration of drugs and vitamins in formulations. 23 For example, voltammetric methods have been developed for the quantitative analysis of vitamin A, niacinamide, and riboflavin. When the compound of interest is not electroactive, it often can be derivatized to an electroactive form. One example is the differential pulse polarographic determination of sulfanilamide, in which it is converted into an electroactive azo dye by coupling with sulfamic acid and 1-naphthol.\n",
            "\n",
            "Representative Methods\n",
            "\n",
            "## Method 11. 3\n",
            "\n",
            "## Determination of Chlorpromazine 25\n",
            "\n",
            "Description of Method. The amount of chlorpromazine in a pharmaceutical formulation is determined voltammetrically at a graphite working electrode in a nonstirred solution. Calibration is achieved using the method of standard additions.\n",
            "\n",
            "Procedure. Place 10.00 mL of a solution consisting of 0.01 M HCl and 0.1 M KCl in the electrochemical cell. Place a graphite working electrode, a Pt auxiliary electrode, and a SCE reference electrode in the cell, and record the voltammogram from +0.2 V to 2.0 V at a scan rate of 50 mV/s. Weigh out an appropriate amount of the pharmaceutical formulation, and dissolve it in a small amount of the electrolyte. Transfer the solution to a 100-mL volumetric flask, and dilute to volume with the electrolyte. Filter a small amount of the diluted solution, and transfer 1.00 mL of the filtrate to the voltammetric cell. Mix the contents of the voltammetric cell, and allow the solution to sit for 10 s before recording the voltammogram. Return the potential to +0.2 V, add 1.00 mL of a chlorpromazine standard, and record the voltammogram. Report the %w/w chlorpromazine in the formulation.\n",
            "\n",
            "## Questions\n",
            "\n",
            "## 1. Is chlorpromazine oxidized or reduced at the graphite working electrode?\n",
            "\n",
            "Since the direction of the scan is toward more positive potentials, the chlorpromazine is oxidized at the graphite working electrode.\n",
            "\n",
            "-Continued\n",
            "\n",
            "- 2. Why is it not necessary to remove dissolved O2 from the solution before recording the voltammogram?\n",
            "\n",
            "Dissolved O2 is a problem when scanning toward more negative potentials where its reduction can lead to a significant cathodic current. In this procedure we are scanning toward more positive potentials and generating anodic currents; thus, dissolved O2 is not an interferent.\n",
            "\n",
            "- 3. What is the purpose of waiting 10 s after mixing the contents of the electrochemical cell before recording the voltammogram?\n",
            "\n",
            "Mixing the solution leads to convection, but the delay allows convection to cease.\n",
            "\n",
            "- 4. In preparing the sample for analysis the initial solution is filtered. Why is it not necessary to collect the entire filtrate before proceeding?\n",
            "\n",
            "This analysis is an example of a concentration technique. Once the original sample is brought to volume in the 100-mL volumetric flask, any portion of the sample solution, even that obtained on filtering, may be used for the analysis.\n",
            "\n",
            "## 11D. 7 Characterization Applications\n",
            "\n",
            "In the previous section we saw how voltammetry can be used to determine the concentration of an analyte. Voltammetry also can be used to obtain additional information, including verifying electrochemical reversibility, determining the number of electrons transferred in a redox reaction, and determining equilibrium constants for coupled chemical reactions. Our discussion of these applications is limited to the use of voltammetric techniques that give limiting currents, although other voltammetric techniques also can be used to obtain the same information.\n",
            "\n",
            "Electrochemical Reversibility and Determination of n In deriving a relationship between E 1/2 and the standard-state potential for a redox couple (11.41), we noted that the redox reaction must be reversible. How can we tell if a redox reaction is reversible from its voltammogram? For a reversible reaction, equation 11.40 describes the voltammogram.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "A plot of E versus log( / i i lim -i )  for  a  reversible  reaction,  therefore,  should be a straight line with a slope of -0.05916/ n. In addition, the slope should yield an integer value for n.\n",
            "\n",
            "## EXAMPLE 11.1 3\n",
            "\n",
            "The following data were obtained from the linear scan hydrodynamic voltammogram of a reversible reduction reaction\n",
            "\n",
            "|   Potential (V vs. SCE) |   Current ( µ A) |\n",
            "|-------------------------|------------------|\n",
            "|                  -0.358 |             0.37 |\n",
            "|                  -0.372 |             0.95 |\n",
            "|                  -0.382 |             1.71 |\n",
            "|                  -0.4   |             3.48 |\n",
            "|                  -0.41  |             4.2  |\n",
            "|                  -0.435 |             4.97 |\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.42\n",
            "\n",
            "Effect of complexation on voltammograms.\n",
            "\n",
            "The limiting current was 5.15 µ A. Show that the reduction reaction is reversible, and determine values for n and E 1/2 .\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "If  the  reaction is reversible, then a plot of E versus log( / i i lim -i )  will  be  a straight line with a slope that yields an integer value for n. As shown in Figure 11.41, plotting the data in this manner gives a straight line. Linear regression gives the equation for the straight line as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Knowing that\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "gives a value for n of 1.97, or 2 electrons. Combining equations 11.40 and 11.41 shows that the y -intercept for a plot of E versus log( / i i lim -i ) is equal to the halfwave potential. Thus, E 1/2 for the reduction reaction is -0.391 V versus the SCE.\n",
            "\n",
            "Figure 11.41\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Graphical determination of electrochemical reversibility, n, and half-wave potential in linear scan hydrodynamic voltammetry.\n",
            "\n",
            "Determining Equilibrium Constants for Coupled Chemical Reactions Another important application of voltammetry is the determination of equilibrium constants for solution reactions that are coupled to a redox reaction occurring at the electrode. The presence of the solution reaction affects the ease of electron transfer, shifting the potential to more negative or more positive potentials. Consider, for example, the reduction of O to R\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "the voltammogram for which is shown in Figure 11.42. If a ligand, L, capable of forming a strong complex with O, is present, then the reaction\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "also must be considered. The overall reaction, therefore, is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Because of its stability, reduction of the OL p complex is less favorable than the reduction of O. As shown in Figure 11.42, the voltammograms for OL p occur at potentials  more negative than those for O. Furthermore, the shift in the voltammogram depends on the ligand's concentration.\n",
            "\n",
            "The shift in the voltammogram for a metal ion in the presence of a ligand may be used to determine both the metal-ligand complex's stoichiometry and its formation constant. To derive a relationship between the relevant variables we begin with two equations: the Nernst equation for the reduction of O\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and the stability constant for the complex at the electrode surface\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "In the absence of ligand the half-wave potential occurs when [R] x =0 and [O] x= 0 are equal; thus, from the Nernst equation we have\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where the subscript 'nc' signifies that no complex is present.\n",
            "\n",
            "When ligand is present we must account for its effect on the concentration of O. Solving equation 11.42 for [O] x =0 and substituting into the Nernst equation gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "If  the  ligand  is  present  in  excess  and  the  formation constant is sufficiently large, such that all of O is present as the complex, then [R] x =0 and [OL ] p x =0 are equal at the half-wave potential, and equation 11.44 simplifies to\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where the subscript 'c' indicates that the complex is present. Defining ∆ E 1/2 as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and substituting equations 11.44 and 11.45 gives, after rearranging\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "A plot of ∆ E 1/2 versus log [L], therefore, gives a straight line. The slope of the line is used to determine the stoichiometric coefficient, p, and the y -intercept gives the value for the formation constant β p .\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11.1 4\n",
            "\n",
            "A voltammogram for the two-electron reduction of M has a half-wave potential of -0.226 V versus the SCE. In the presence of an excess of the ligand L, the following half-wave potentials are recorded\n",
            "\n",
            "|   [L] (M) |   ( E 1/2 ) c (V vs. SCE) |\n",
            "|-----------|---------------------------|\n",
            "|      0.02 |                    -0.494 |\n",
            "|      0.04 |                    -0.512 |\n",
            "|      0.06 |                    -0.523 |\n",
            "|      0.08 |                    -0.53  |\n",
            "|      0.1  |                    -0.536 |\n",
            "\n",
            "Determine values for the stoichiometry of the complex and the formation constant of the complex.\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "First we must calculate values of ∆ E 1/2 using equation 11.46; thus\n",
            "\n",
            "|   [L] (M) |   ∆ E 1/2 (V vs. SCE) |\n",
            "|-----------|-----------------------|\n",
            "|      0.02 |                -0.268 |\n",
            "|      0.04 |                -0.286 |\n",
            "|      0.06 |                -0.297 |\n",
            "|      0.08 |                -0.304 |\n",
            "|      0.1  |                -0.31  |\n",
            "\n",
            "The plot of ∆ E 1/2 as a function of the log of the ligand concentration is shown in Figure 11.43. Linear regression gives the equation for the straight line as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "From the slope and intercept we know that\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving these equations gives the stoichiometry of the complex as ML2 ( p = 2.03), with a formation constant of β 2 = 3.5 × 10 12 .\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.43\n",
            "\n",
            "Graphical determination of the stoichiometry and formation constant for a complexation reaction.\n",
            "\n",
            "## 11D.8 Evaluation\n",
            "\n",
            "Scale  of  Operation Voltammetry is routinely used to analyze samples at the parts-per-million level and, in some cases, can be used to detect analytes at the parts-per-billion or parts-per-trillion level. Most analyses are carried out in conventional electrochemical cells using macro samples; however, microcells are available that require as little as 50 µ L of sample. Microelectrodes, with diameters  as  small  as  2 µ m, allow voltammetric measurements to be made on even smaller samples. For example, the concentration of glucose in 200µ m pond snail neurons has been successfully monitored using a 2µ m amperometric glucose electrode. 26\n",
            "\n",
            "Accuracy The accuracy of a voltammetric analysis often is limited by the ability to correct for residual currents, particularly those due to charging. For analytes at the parts-per-million level, accuracies of ± 1-3% are easily obtained. As expected, a decrease in accuracy is experienced when analyzing samples with significantly smaller concentrations of analyte.\n",
            "\n",
            "Precision Precision is generally limited by the uncertainty in measuring the limiting or peak current. Under most experimental conditions, precisions of ± 1-3% can be reasonably expected. One exception is the analysis of ultratrace analytes in complex matrices by stripping voltammetry, for which precisions as poor as ± 25% are possible.\n",
            "\n",
            "Sensitivity In many voltammetric experiments, sensitivity can be improved by adjusting the experimental conditions. For example, in stripping voltammetry, sensitivity  is  improved by increasing the deposition time, by increasing the rate of the linear potential scan, or by using a differential-pulse technique. One reason for the popularity of potential pulse techniques is an increase in current relative to that obtained with a linear potential scan.\n",
            "\n",
            "Selectivity Selectivity in voltammetry is determined by the difference between half-wave potentials or peak potentials, with minimum differences of ± 0.2-0.3 V required for a linear potential scan, and ± 0.04-0.05 V for differential pulse voltammetry. Selectivity can be improved by adjusting solution conditions. As we have seen, the presence of a complexing ligand can substantially shift the potential at which an analyte is oxidized or reduced. Other solution parameters, such as pH, also can be used to improve selectivity.\n",
            "\n",
            "Time, Cost, and Equipment Commercial instrumentation for voltammetry ranges from less than $1000 for simple instruments to as much as $20,000 for more sophisticated instruments. In general, less expensive instrumentation is limited to linear potential scans, and the more expensive instruments allow for more complex potential-excitation signals using potential pulses. Except for stripping voltammetry, which uses long deposition times, voltammetric analyses are relatively rapid.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## 11E KEY TERMS\n",
            "\n",
            "| amalgam ( p. 509 )                      |\n",
            "|-----------------------------------------|\n",
            "| amperometry ( p. 519 )                  |\n",
            "| anode ( p. 466 )                        |\n",
            "| anodic current ( p. 510 )               |\n",
            "| asymmetry potential ( p. 476 )          |\n",
            "| auxiliary electrode ( p. 463 )          |\n",
            "| cathode ( p. 466 )                      |\n",
            "| cathodic current ( p. 510 )             |\n",
            "| charging current ( p. 513 )             |\n",
            "| convection ( p. 512 )                   |\n",
            "| coulometric titrations ( p. 501 )       |\n",
            "| coulometry ( p. 496 )                   |\n",
            "| counter electrode ( p. 462 )            |\n",
            "| current efficiency ( p. 496 )           |\n",
            "| diffusion ( p. 511 )                    |\n",
            "| dropping mercury electrode p.           |\n",
            "| (                                       |\n",
            "| electrical double layer ( p. 513 )      |\n",
            "| electrode of the first kind ( p. 474 )  |\n",
            "| electrode of the second kind ( p. 475 ) |\n",
            "\n",
            "| enzyme electrodes ( p.                          | enzyme electrodes ( p.                          | enzyme electrodes ( p.                          | 484 )                                           |\n",
            "|-------------------------------------------------|-------------------------------------------------|-------------------------------------------------|-------------------------------------------------|\n",
            "| faradaic current ( p. Faraday's law ( p. 496    | hydrodynamic                                    | faradaic current ( p. Faraday's law ( p. 496    | voltammetry ( p. 513 )                          |\n",
            "| liquid-based ion-selective electrode ( p. 482 ) | liquid-based ion-selective electrode ( p. 482 ) | liquid-based ion-selective electrode ( p. 482 ) | liquid-based ion-selective electrode ( p. 482 ) |\n",
            "| liquid junction potential transport ( p. 511    |                                                 | ( )                                             | p. 470 )                                        |\n",
            "| mass mediator                                   | ( p. 500                                        | ( p.                                            | )                                               |\n",
            "| membrane potential 475 )                        | membrane potential 475 )                        | membrane potential 475 )                        | membrane potential 475 )                        |\n",
            "| migration                                       | ( p. 512 )                                      |                                                 |                                                 |\n",
            "| Ohm's law                                       | nonfaradaic current                             | ( p.                                            | 512 )                                           |\n",
            "| ( p. 463 )                                      | ( p. 463 )                                      | ( p. 463 )                                      | ( p. 463 )                                      |\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## 11F SUMMARY\n",
            "\n",
            "Electrochemical methods covered in this chapter include potentiometry, coulometry, and voltammetry. Potentiometric methods are based on the measurement of an electrochemical cell's potential when only a negligible current is allowed to flow. In principle the Nernst equation can be used to calculate the concentration of species in the electrochemical cell by measuring its potential and solving the Nernst equation; the presence of liquid junction potentials,  however, necessitates the use of an external standardization or the use of standard additions.\n",
            "\n",
            "Potentiometric electrodes are divided into two classes: metallic electrodes and membrane electrodes. The smaller of these classes are the metallic electrodes. Electrodes of the first kind respond to the concentration of their cation in solution; thus the potential of an Ag wire is determined by the concentration of Ag + in solution. When another species is present in solution and in equilibrium with the metal ion, then the electrode's potential will respond to the concentration of that ion. For example, an Ag wire in contact with a solution of Cl will  respond to the concentration of Cl since the relative concentrations of Ag + and Cl are fixed by the solubility product for AgCl. Such electrodes are called electrodes of the second kind.\n",
            "\n",
            "The potential of a membrane electrode is determined by a difference in the composition of the solution on either side of the membrane. Electrodes using a glass membrane respond to ions that bind to negatively charged sites on the glass membrane's surface. A pH electrode is one example of a glass membrane electrode. Other kinds of membrane electrodes include those using insoluble crystalline solids and liquid ion exchangers incorporated in a hydrophobic membrane. The F ion-selective electrode, which uses a single crystal of LaF3 as  the  ion-selective  membrane, is an example of a solid-state electrode. The Ca 2+ ion-selective electrode, in which a chelating ligand such as di-( n -decyl)phosphate is immobilized in a PVC membrane, is an example of a liquid-based ion-selective electrode.\n",
            "\n",
            "Potentiometric electrodes also can be designed to respond to molecules by incorporating a reaction producing an ion whose concentration can be determined using a traditional ion-selective electrode. Gas-sensing electrodes, for example, include a gas-permeable membrane that isolates the ion-selective electrode from the solution containing the analyte. Diffusion of a dissolved gas across the membrane alters the composition of the inner solution in a manner that can be followed with an ion-selective electrode. Enzyme electrodes operate in the same way.\n",
            "\n",
            "Coulometric methods are based on Faraday's law that the total charge or current passed during an electrolysis is proportional to the amount of reactants and products in the redox reaction. If the electrolysis is 100% efficient, in that only the analyte is oxidized or reduced, then the total charge or current can be used to determine\n",
            "\n",
            "| overpotential ( p. 497 ) polarography ( p. 515 )                                                                                                              |\n",
            "|---------------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
            "| potentiometer ( p. 464 ) potentiostat ( p. 465 ) redox electrode ( p. 475 ) reference electrode ( p. 462 ) residual current ( p. 513 ) salt bridge ( p. 466 ) |\n",
            "| saturated calomel electrode ( p. 472 ) silver/silver chloride electrode ( p. 473 )                                                                            |\n",
            "| solid-state ion-selective electrode ( p. 479 )                                                                                                                |\n",
            "| standard hydrogen electrode ( p. 471 ) static mercury drop electrode ( p. 509 )                                                                               |\n",
            "| total ionic strength                                                                                                                                          |\n",
            "| adjustment buffer ( p. 487 )                                                                                                                                  |\n",
            "| voltammetry ( p. 508 )                                                                                                                                        |\n",
            "| )                                                                                                                                                             |\n",
            "| voltammogram ( p. 508                                                                                                                                         |\n",
            "\n",
            "the amount of analyte in a sample. In controlled-potential coulometry, a constant potential is applied and the current is measured as a function of time, whereas in controlled-current coulometry, the current is held constant and the time required to completely oxidize or reduce the analyte is measured.\n",
            "\n",
            "In voltammetry we measure the current in an electrochemical cell as a function of the applied potential. Individual voltammetric methods differ in terms of the type of electrode used, how the applied potential is changed, and whether the transport of material to the electrode's surface is enhanced by stirring.\n",
            "\n",
            "Polarography is a voltammetric experiment conducted at a mercury electrode under conditions in which the solution is not stirred.  Normal polarography uses a dropping mercury electrode (or a static mercury drop electrode) and a linear potential scan in an unstirred solution. Other forms of polarography include normal pulse polarography, differential pulse polarography, staircase polarography, and square-wave polarography, all of which apply a series of potential pulses to the mercury electrode.\n",
            "\n",
            "11G\n",
            "\n",
            "In hydrodynamic voltammetry the solution is stirred either by using a magnetic stir bar or by rotating the electrode. Because the solution is stirred, a dropping mercury electrode cannot be used and is replaced with a solid electrode. Both linear potential scans or potential pulses can be applied.\n",
            "\n",
            "In stripping voltammetry the analyte is first deposited on the electrode, usually as the result of an oxidation or reduction reaction. The potential is then scanned, either linearly or by using potential pulses, in a direction that removes the analyte by a reduction or oxidation reaction.\n",
            "\n",
            "Amperometry is a voltammetric method in which a constant potential is applied to the electrode and the resulting current is measured. Amperometry is most often used in the construction of chemical sensors that, as with potentiometric sensors, are used for the quantitative analysis of single analytes. One important example, for instance, is the Clark O2 electrode, which responds to the concentration of dissolved O2 in solutions such as blood and water.\n",
            "\n",
            "Suggested\n",
            "\n",
            "## EXPERIMENTS\n",
            "\n",
            "## Experiments\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "The following set of suggested experiments describes the preparation of solid-state and liquid ion-exchange ionselective electrodes, as well as potentiometric biosensors.\n",
            "\n",
            "Chan, W. H.; Wong, M. S.; Yip, C. W. 'Ion-Selective Electrode in Organic Analysis: A Salicylate Electrode,' J. Chem. Educ. 1986, 63, 915-916.\n",
            "\n",
            "The preparation of an ion-selective electrode for salicylate is described. The electrode incorporates an ion-pair of crystal violet and salicylate in a PVC matrix as the ion-selective membrane. Its use for the determination of acetylsalicylic acid in aspirin tablets is described. A similar experiment is described by Creager, S. E.; Lawrence, K. D.; Tibbets, C. R. in 'An Easily Constructed Salicylate-Ion-Selective Electrode for Use in the Instructional Laboratory,' J. Chem. Educ. 1995, 72, 274-276.\n",
            "\n",
            "Christopoulos, T. K.; Diamandis, E. P. 'Use of a Sintered Glass Crucible for Easy Construction of Liquid-Membrane Ion-Selective Electrodes,' J. Chem. Educ. 1988, 65, 648.\n",
            "\n",
            "Liquid-membrane ion-selective electrodes are prepared by removing the bottom below the glass filter, cleaning the filter, and making it hydrophobic using trimethylchlorosilane. Suitable liquid ion exchangers are added and allowed to penetrate the hydrophobic glass membrane. Adding an appropriate internal reference solution and reference electrode completes the ion-selective electrode.\n",
            "\n",
            "Fricke, G. H.; Kuntz, M. J. 'Inexpensive Solid-State IonSelective Electrodes for Student Use,' J. Chem. Educ. 1977, 54, 517-520.\n",
            "\n",
            "Instructions are provided in this experiment for the preparation of solid-state electrodes for the quantitative analysis of Ag + , Cl -, Cu 2+ , and Pb 2+ . All electrodes are based on AgCl, AgCl/Ag2S, CuS/Ag2S, and PbS/Ag2S preparations. Electrodes are prepared by (a) mixing the solid with silicon grease and coating on an Ag wire; (b) mixing the solid with silicone grease and packing into the end of a glass tube; or (c) preparing a pressed pellet.\n",
            "\n",
            "Lloyd, B. W.; O'Brien, F. L.; Wilson, W. D. 'Student Preparation and Analysis of Chloride and Calcium Ion Selective Electrodes,' J. Chem. Educ. 1976, 53, 328-330.\n",
            "\n",
            "This experiment describes the preparation of liquid ionexchange electrodes for Cl and Ca 2+ . The liquid ionexchange solutions are incorporated into PVC membranes and fixed to the end of glass tubing. The internal solutions are either NaCl or CaCl2, and a Ag/AgCl reference electrode is situated in the internal solution.\n",
            "\n",
            "Martínez-Fàbregas, E.; Alegret, S. 'A Practical Approach to Chemical Sensors through Potentiometric Transducers: Determination of Urea in Serum by Means of a Biosensor,' J. Chem. Educ. 1994, 71, A67-A70.\n",
            "\n",
            "Directions are provided for constructing and characterizing an ammonium ion-selective electrode. The electrode is then modified to respond to urea by adding a few milligrams of urease and covering with a section of dialysis membrane. Directions for determining urea in serum also are provided.\n",
            "\n",
            "## Experiments\n",
            "\n",
            "Continued from page 533\n",
            "\n",
            "Mifflin, T. E.; Andriano, K. M.; Robbins, W. B. 'Determination of Penicillin Using an Immobilized Enzyme Electrode,' J. Chem. Educ. 1984, 61, 638-639.\n",
            "\n",
            "Directions for preparing a potentiometric biosensor for penicillin are provided in this experiment. The enzyme penicillinase is immobilized in a polyacrylamide polymer formed on the surface of a glass pH electrode. The electrode shows a linear response to penicillin G over a concentration range of 10 -5 M to 10 -3 M.\n",
            "\n",
            "Palanivel, A.; Riyazuddin, P. 'Fabrication of an Inexpensive Ion-Selective Electrode,' J. Chem. Educ. 1984, 61, 290.\n",
            "\n",
            "This experiment describes the preparation of a Ag2S/CuS electrode using the graphite rod from a dry cell D-type battery. After obtaining the graphite rod, it is inserted and sealed in a polyethylene tube, leaving a 2-cm portion exposed. The protruding portion of the graphite rod, is placed in a saturated solution of Na2S for 1 h, and in a saturated solution of CuS for an additional hour. In a related experiment (Riyazuddin, P.; Devika, D. 'Potentiometric Acid-Base Titrations with Activated Graphite Electrodes,' J. Chem. Educ. 1997, 74, 1198-1199.), activated graphite electrodes are used for the potentiometric titration of HCl with NaOH.\n",
            "\n",
            "Ramaley, L; Wedge, P. J.; Crain, S. M. 'Inexpensive Instrumental Analysis: Part 1. Ion-Selective Electrodes,' J. Chem. Educ. 1994, 71, 164-167.\n",
            "\n",
            "This experiment describes the preparation and evaluation of two liquid-membrane Na + ion-selective electrodes, using either the sodium salt of monensin or a hemisodium ionophore as ion exchangers incorporated into a PVC matrix. Electrodes prepared using monensin performed poorly, but those prepared using hemisodium showed a linear response over a range of 0.1 M to 3 × 10 -5 M Na + with slopes close to the theoretical value.\n",
            "\n",
            "The following set of experiments illustrate several applications of potentiometric electrodes.\n",
            "\n",
            "Harris, T. M. 'Potentiometric Measurement in a Freshwater Aquarium,' J. Chem. Educ. 1993, 70, 340-341.\n",
            "\n",
            "This experiment describes a semester-long project in which the concentration of several ions in a fresh water aquarium are monitored. Ions that are monitored using potentiometric electrodes include H + (pH electrode), Cl (chloride electrode), HCO3 -(CO2 electrode), NH4 + (NH3 electrode), and NO3 - (NH3 electrode). Nitrate concentrations were determined following its conversion to ammonia.\n",
            "\n",
            "The titration of Al 3+ with F in a water-acetonitrile mixture is followed potentiometrically using a fluoride ion-selective electrode. The cumulative formation constant for AlF6 3is determined from the titration curve.\n",
            "\n",
            "Selig, W. S. 'Potentiometric Titrations Using Pencil and Graphite Sensors,' J. Chem. Educ. 1984, 61, 80-81.\n",
            "\n",
            "Radic, N.; Komijenovic, J. 'Potentiometric Determination of an Overall Formation Constant Using an Ion-Selective Membrane Electrode,' J. Chem. Educ. 1993, 70, 509-511.\n",
            "\n",
            "This experiment describes the use of coated graphite electrodes for the potentiometric monitoring of precipitation, acid-base, complexation, and redox titrations.\n",
            "\n",
            "The following four experiments describe useful and accessible applications of coulometry.\n",
            "\n",
            "Bertotti, M.; Vaz, J. M.; Telles, R. 'Ascorbic Acid Determination in Natural Orange Juice,' J. Chem. Educ. 1995, 72, 445-447.\n",
            "\n",
            "Lötz, A. 'A Variety of Electrochemical Methods in a Coulometric Titration Experiment,' J. Chem. Educ. 1998, 75, 775-777.\n",
            "\n",
            "The titration of ascorbic acid using coulometrically generated I 2 and Br2 is described in this experiment. Details are also given for the polarographic analysis of ascorbic acid.\n",
            "\n",
            "Kalbus, G. E.; Lieu, V. T. 'Dietary Fat and Health: An Experiment on the Determination of Iodine Number of Fats and Oils by Coulometric Titration,' J. Chem. Educ. 1991, 68, 64-65.\n",
            "\n",
            "The iodine number of fats and oils provides a quantitative measurement of the degree of unsaturation. A solution containing a 100% excess of ICl is added to the sample, reacting across the double-bonded sites of unsaturation. The excess ICl is converted to I2 by adding KI. The resulting I2 is reacted with a known excess of Na2S2O3. To complete the analysis the excess S2O3 2is back titrated with coulometrically generated I2.\n",
            "\n",
            "This experiment describes three coulometric titrations for the quantitative analysis of mixtures containing KI and HCl. Iodide is determined by titrating with coulometrically generated Br2. Chloride is determined indirectly by the coulometric titration of H + . Finally, iodide and chloride are determined together by precipitating AgI and AgCl following the anodic dissolution of an Ag wire.\n",
            "\n",
            "Swim, J.; Earps, E.; Reed, L. M.; et al. 'Constant-Current Coulometric Titration of Hydrochloric Acid,' J. Chem. Educ. 1996, 73, 679-683.\n",
            "\n",
            "Directions are provided for constructing an inexpensive constant-current source and demonstrate its use in determining the concentration of HCl coulometrically.\n",
            "\n",
            "-Continued\n",
            "\n",
            "## Experiments\n",
            "\n",
            "The final set of experiments provide examples of voltammetric and amperometric analyses.\n",
            "\n",
            "García-Armada, P.; Losada, J.; de Vicente-Pérez, S. 'Cation Analysis Scheme by Differential Pulse Polarography,' J. Chem. Educ. 1996, 73, 544-547.\n",
            "\n",
            "unknowns, serum, or plasma can be determined using a normal calibration curve.\n",
            "\n",
            "Directions are provided for the simultaneous qualitative and quantitative analysis of Bi(III), Cd(II), Co(II), Cu(II), Cr(III), Mn(II), Ni(II), Pb(II), Sb(III), and Zn(II). The analysis is accomplished using differential pulse polarography in three separate supporting electrolytes.\n",
            "\n",
            "Herrera-Melián, J. A.; Doña-Rodríguez, J. M.; HernándezBrito, J.; et al. 'Voltammetric Determination of Ni and Co in Water Samples,' J. Chem. Educ. 1997, 74, 1444-1445.\n",
            "\n",
            "This experiment describes the application of cathodic stripping voltammetry to the analysis of Ni and Co in fresh water and sea water.\n",
            "\n",
            "Marin, D.; Mendicuti, F. 'Polarographic Determination of Composition and Thermodynamic Stability Constant of a Complex Metal Ion,' J. Chem. Educ. 1988, 65, 916-918.\n",
            "\n",
            "This experiment describes the determination of the stability (cumulative formation) constant for the formation of Pb(OH)3 - by measuring the shift in the half-wave potential for the reduction of Pb 2+ as a function of the concentration of OH - . The influence of ionic strength is also considered, and results are extrapolated to zero ionic strength to determine the thermodynamic formation constant.\n",
            "\n",
            "Sittampalam, G.; Wilson, G. S. 'Amperometric Determination of Glucose at Parts Per Million Levels with Immobilized Glucose Oxidase,' J. Chem. Educ. 1982, 59, 70-73.\n",
            "\n",
            "This experiment describes the use of a commercially available amperometric biosensor for glucose that utilizes the enzyme glucose oxidase. The concentration of glucose in artificial\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## 11H PROBLEMS\n",
            "\n",
            "- 1. Identify the anode and cathode for the following electrochemical cells, and write the oxidation or reduction reaction occurring at each electrode. (a) Pt | FeCl2 (0.015 M), FeCl3 (0.045 M) || AgNO3 (0.1 M) | Ag (b) Ag | AgBr(s) | NaBr (1.0 M) || CdCl2 (0.05 M) | Cd (c) Pb, PbSO4 | H2SO4 (1.5 M) || H2SO4 (2.0 M) | PbSO4, PbO2\n",
            "- 2. Calculate the potential for the electrochemical cells in problem 1.\n",
            "- 3. Calculate the molar concentration for the underlined component in the following cell if the cell potential is measured at +0.294 V\n",
            "\n",
            "Ag | AgCl (sat'd) | NaCl (0.1 M) || KI ( x M) | I2(s) | Pt\n",
            "\n",
            "Town, J. L.; MacLaren, F.; Dewald, H. D. 'Rotating Disk Voltammetry Experiment,' J. Chem. Educ. 1991, 68, 352-354.\n",
            "\n",
            "This experiment introduces hydrodynamic voltammetry using a rotating working electrode. Its application for the quantitative analysis of K4Fe(CN)6 is demonstrated.\n",
            "\n",
            "Wang, J. 'Sensitive Electroanalysis Using Solid Electrodes,' J. Chem. Educ. 1982, 59, 691-692.\n",
            "\n",
            "The technique of hydrodynamic modulation voltammetry (HMV), in which the rate of stirring is pulsed between high and low values, is demonstrated in this experiment. The application of HMV for the quantitative analysis of ascorbic acid in vitamin C tablets using the method of standard additions also is outlined.\n",
            "\n",
            "Wang, J. 'Anodic Stripping Voltammetry,' J. Chem. Educ. 1983, 60, 1074-1075.\n",
            "\n",
            "The application of anodic stripping voltammetry for the quantitative analysis of Cd, Pb, and Cu in natural waters is described in this experiment.\n",
            "\n",
            "Wang, J.; Maccà, C. 'Use of Blood-Glucose Test Strips for Introducing Enzyme Electrodes and Modern Biosensors,' J. Chem. Educ. 1996, 73, 797-800.\n",
            "\n",
            "Commercially available kits for monitoring blood-glucose use an amperometric biosensor incorporating the enzyme glucose oxidase. This experiment describes how such monitors can be adapted to the quantitative analysis of glucose in beverages.\n",
            "\n",
            "- 4. What reaction prevents Zn from being used as an electrode of the first kind in acidic solution? Which other electrodes of the first kind would you expect to behave in the same manner as Zn when immersed in acidic solutions?\n",
            "- 5. Creager and colleagues designed a salicylate ion-selective electrode using a PVC membrane impregnated with tetraalkylammonium salicylate. 27 To determine the ionselective electrode's selectivity coefficient for benzoate, a set of salicylate calibration standards was prepared in which the concentration of benzoate was held constant at 0.10 M. Using the following data, determine the value of the selectivity coefficient.\n",
            "\n",
            "| [Salicylate] (M)   |   Potential (mV) |\n",
            "|--------------------|------------------|\n",
            "| 1.0                |             20.2 |\n",
            "| 1.0 × 10 -1        |             73.5 |\n",
            "| 1.0 × 10 -2        |            126   |\n",
            "| 1.0 × 10 -3        |            168   |\n",
            "| 1.0 × 10 -4        |            182   |\n",
            "| 1.0 × 10 -5        |            182   |\n",
            "| 1.0 × 10 -6        |            177   |\n",
            "\n",
            "What is the maximum acceptable concentration of benzoate if this ion-selective electrode is to be used for analyzing samples containing as little as 10 -5 M salicylate with an accuracy of better than 1%?\n",
            "\n",
            "- 6. Watanabe and co-workers described a new membrane electrode for the determination of cocaine, which is a weak base alkaloid with a p K a of 8.64. 28 The response of the electrode for a fixed concentration of cocaine was found to be independent of pH in the range of 1-8, but decreased sharply above a pH of 8. Offer an explanation for the source of this pH dependency.\n",
            "- 7. Show that equation 11.12 for the urea electrode is correct.\n",
            "- 8. The pH-based urea electrode has the response shown in equation 11.13. Rewrite the equation to show the relationship between potential, E, and the concentration of urea. Comment on your result.\n",
            "- 9. Mifflin and associates described a membrane electrode for the quantitative analysis of penicillin in which the enzyme penicillinase is immobilized in a polyacrylamide gel that is coated on a glass pH electrode. 29 The following data were collected for a series of penicillin standards.\n",
            "\n",
            "| [Penicillin] (M)   |   Potential (mV) |\n",
            "|--------------------|------------------|\n",
            "| 1.0 × 10 -2        |              220 |\n",
            "| 2.0 × 10 -3        |              204 |\n",
            "| 1.0 × 10 -3        |              190 |\n",
            "| 2.0 × 10 -4        |              153 |\n",
            "| 1.0 × 10 -4        |              135 |\n",
            "| 1.0 × 10 -5        |               96 |\n",
            "| 1.0 × 10 -6        |               80 |\n",
            "\n",
            "Construct a calibration curve for the electrode, and report (a) the range of concentrations in which a linear response is observed, (b) the equation for the calibration curve in this range, and (c) the concentration of penicillin in a sample that yields a potential of 142 mV.\n",
            "\n",
            "- 10. Ion-selective electrodes can be incorporated in flow cells to monitor the concentration of an analyte in standards and samples that are pumped through the flow cell. As the analyte passes through the cell, a potential spike is recorded instead of a steady-state potential. The concentration of K + in serum has been determined in this fashion, using standards prepared in a matrix of 0.014 M NaCl. 30\n",
            "\n",
            "|   [K + ] (mM) |   Potential (arb. units) |\n",
            "|---------------|--------------------------|\n",
            "|           0.1 |                     25.5 |\n",
            "|           0.2 |                     37.2 |\n",
            "|           0.4 |                     50.8 |\n",
            "|           0.6 |                     58.7 |\n",
            "|           0.8 |                     64   |\n",
            "|           1   |                     66.8 |\n",
            "\n",
            "A 1.00-mL sample of serum is diluted to volume in a 10-mL volumetric flask and analyzed, giving a potential of 51.1. Report the concentration of K + in the sample of serum.\n",
            "\n",
            "- 11. Wang and Taha described an interesting application of potentiometry called batch injection. 31 As shown in the following figure, an ion-selective electrode is placed in an inverted position in a large-volume tank, and a fixed volume of a sample or standard solution is injected toward the electrode's surface using a micropipet.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "The response of the electrode is a peak whose height is proportional to the analyte's concentration. The following data were collected using a pH electrode\n",
            "\n",
            "|   pH |   Potential (mV) |\n",
            "|------|------------------|\n",
            "|    2 |             +300 |\n",
            "|    3 |             +240 |\n",
            "|    4 |             +168 |\n",
            "|    5 |              +81 |\n",
            "|    6 |              +35 |\n",
            "|    8 |              -92 |\n",
            "|    9 |             -168 |\n",
            "|   10 |             -235 |\n",
            "|   11 |             -279 |\n",
            "\n",
            "Determine the pH for each of the following samples, given the recorded peak potential: (a) tomato juice, +167 mV; (b) tap water, -27 mV; (c) coffee, 122 mV.\n",
            "\n",
            "- 12. The concentration of NO3 -in a water sample is determined by a one-point standard addition using an NO3 -ion-selective electrode. A 25.00-mL sample is placed in a beaker, and a potential of +0.102 V is measured. A 1.00-mL aliquot of a 200.0 ppm standard solution of NO3 -is added, after which the potential is found to be +0.089 V. Report the concentration of NO3 -in parts per million.\n",
            "- 13. The following data were collected for the analysis of fluoride in tap water and in toothpaste. (a) For the analysis of tap water, three 25.0-mL samples were each mixed with 25.0 mL of TISAB, and the potential was measured with an F ISE relative to a saturated calomel electrode. Five 1.00-mL additions of a standard solution of 100.0-ppm F were added to each, measuring the potential following each addition.\n",
            "\n",
            "| Milliliters of F - added   | Potential (mV)   | Potential (mV)   | Potential (mV)   |\n",
            "|----------------------------|------------------|------------------|------------------|\n",
            "| standard                   | Trial 1          | Trial 2          | Trial 3          |\n",
            "| 0.00                       | -79              | -82              | -81              |\n",
            "| 1.00                       | -119             | -119             | -118             |\n",
            "| 2.00                       | -133             | -133             | -133             |\n",
            "| 3.00                       | -142             | -142             | -142             |\n",
            "| 4.00                       | -149             | -148             | -148             |\n",
            "| 5.00                       | -154             | -153             | -153             |\n",
            "\n",
            "Determine the parts per million of F in the tap water. (b) For the analysis of toothpaste a 0.3619-g sample was transferred to a 100-mL volumetric flask along with 50.0 mL of TISAB and diluted to volume with distilled water. Three 20.0-mL aliquots were removed, and the potential was measured with an F ion-selective electrode using a saturated calomel electrode as a reference. Five separate 1.00-mL additions of a 100.0-ppm solution of F were added to each, measuring the potential following each addition.\n",
            "\n",
            "| Milliliters of standard F - added   | Potential (mV)   | Potential (mV)   | Potential (mV)   |\n",
            "|-------------------------------------|------------------|------------------|------------------|\n",
            "| 0.00                                | Trial 1          | Trial 2          | Trial 3          |\n",
            "|                                     | -55              | -54              | -55              |\n",
            "| 1.00                                | -82              | -82              | -83              |\n",
            "| 2.00                                | -94              | -94              | -94              |\n",
            "| 3.00                                | -102             | -103             | -102             |\n",
            "| 4.00                                | -108             | -108             | -109             |\n",
            "| 5.00                                | -112             | -112             | -113             |\n",
            "\n",
            "Report the parts per million of F in the sample of toothpaste.\n",
            "\n",
            "- 14. You are responsible for determining the amount of KI in iodized salt and decide to use an I ion-selective electrode. Describe how you would perform this analysis using (a) external standards and (b) the method of standard additions.\n",
            "- 15. The purity of a sample of picric acid, C6H3N3O7, is determined by controlled-potential coulometry, converting the picric acid to triaminophenol, C6H9N3O. A 0.2917-g sample of picric acid is placed in a 1000-mL volumetric flask\n",
            "\n",
            "and diluted to volume. A 10.00-mL portion of this solution is transferred to a coulometric cell and diluted till the Pt cathode is immersed. The exhaustive electrolysis of the sample requires 21.67 C of charge. Report the purity of the picric acid.\n",
            "\n",
            "- 16. The concentration of H2S in the drainage from an abandoned mine can be determined by a coulometric titration using KI as a mediator and I3 -as the 'titrant.'\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "A 50.00-mL sample of water is placed in a coulometric cell, along with an excess of KI and a small amount of starch as an indicator. Electrolysis is carried out at a constant current of 84.6 mA, requiring 386 s to reach the starch end point. Report the concentration of H2S in the sample in parts per million.\n",
            "\n",
            "- 17. One method for the determination of H3AsO3 is by a coulometric titration using I3 -as a 'titrant.' The relevant reactions and standard-state potentials are summarized as follows.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Explain why the coulometric titration must be done in neutral solutions (pH ≈ 7), instead of in strongly acidic solutions (pH &lt; 0).\n",
            "\n",
            "- 18. The reduction of acrylonitrile, CH2=CHCN, to adiponitrile, NC(CH2)4CN, is an important industrial process. A 0.594-g sample of acrylonitrile is placed in a 1-L volumetric flask and diluted to volume. An exhaustive controlled-potential electrolysis of a 1.00-mL portion of the diluted acrylonitrile requires 1.080 C of charge. What is the value of n for this reduction?\n",
            "- 19. A linear-potential scan hydrodynamic voltammogram for a mixture of Fe 2+ and Fe 3+ is shown in the figure, where i l,a and\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "i l,c are the anodic and cathodic limiting currents. (a) Show that the potential is given by\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "- (b) What is the potential when i = 0 for a solution that is 0.1 mM Fe 3+ and 0.05 mM Fe 2+ ? You may assume that K Fe 3+ ≈ K Fe 2+ .\n",
            "\n",
            "- 20. The amount of sulfur in aromatic monomers can be determined by differential pulse polarography. Standard solutions are prepared for analysis by dissolving 1.000 mL of the purified monomer in 25.00 mL of an electrolytic solvent, adding a known amount of S, deaerating, and measuring the peak current. The following results were obtained for a set of calibration standards\n",
            "\n",
            "|   Micrograms of added S |   Peak current ( µ A) |\n",
            "|-------------------------|-----------------------|\n",
            "|                       0 |                  0.14 |\n",
            "|                      28 |                  0.7  |\n",
            "|                      56 |                  1.23 |\n",
            "|                     112 |                  2.41 |\n",
            "|                     168 |                  3.42 |\n",
            "\n",
            "Analysis of a 1.000-mL sample, treated in the same manner as the standards, gives a peak current of 1.77 µ A. Report the amount of sulfur present in the sample in milligrams per milliliter.\n",
            "\n",
            "- 21. The purity of a sample of K3Fe(CN)6 was determined using linear-potential scan hydrodynamic voltammetry at a glassy carbon electrode using the method of external standards. The following data were obtained for a set of calibration standards.\n",
            "\n",
            "|   [K 3 Fe(CN) 6 ] (mM) |   Limiting current ( µ A) |\n",
            "|------------------------|---------------------------|\n",
            "|                      2 |                       127 |\n",
            "|                      4 |                       252 |\n",
            "|                      6 |                       376 |\n",
            "|                      8 |                       500 |\n",
            "|                     10 |                       624 |\n",
            "\n",
            "A sample was prepared for analysis by diluting a 0.246-g sample to volume in a 100-mL volumetric flask. The limiting current for the sample was found to be 444 µ A. Report the purity of this sample of K3Fe(CN)6.\n",
            "\n",
            "- 22. Anodic stripping voltammetry at a mercury film electrode can be used to determine whether an individual has recently fired a gun by looking for traces of antimony in residue collected from the individual's hands. In a typical analysis a sample is collected with a cotton-tipped swab that had been wetted with 5% v/v HNO3. When returned to the lab, the swab is placed in a vial containing 5.00 mL of 4 M HCl that is 0.02 M in hydrazine sulfate. After allowing the swab to soak overnight, 4.00 mL of the solution is transferred to an electrochemical cell, and 100 µ L of 1.00 × 10 -2 M HgCl2 is added. Anodic stripping voltammetry of the sample gives a peak current of 0.38 µ A. After a standard addition of 100 µ L of 5.00 × 10 -2 ppb Sb was added, anodic stripping voltammetry is repeated, giving a peak current of 1.14 µ A. How many nanograms of Sb is collected from the individual's hand?\n",
            "- 23. Zinc can be used as an internal standard in the analysis of thallium by differential pulse polarography. A standard\n",
            "- solution containing 5.00 × 10 -5 M Zn 2+ and 2.50 × 10 -5 M Tl + gave peak currents of 5.71 µ A and 3.19 µ A, respectively. An 8.713-g sample of an alloy known to be free of zinc was dissolved in acid, transferred to a 500-mL volumetric flask, and diluted to volume. A 25.0-mL portion of this solution was mixed with 25.0 mL of a 5.00 × 10 -4 M solution of Zn 2+ . Analysis of this solution gave a peak current for Zn 2+ of 12.3 µ A, and for Tl + of 20.2 µ A. Report the Tl% w/w in the alloy.\n",
            "- 24. Differential pulse voltammetry at a carbon working electrode can be used to determine the concentrations of ascorbic acid and caffeine in drug formulations. 32 In a typical analysis, a 0.9183-g tablet is crushed and ground into a fine powder. A 0.5630-g sample of this powder is transferred to a 100-mL volumetric flask, brought into solution, and diluted to volume. A 0.500-mL portion is then transferred to a voltammetric cell containing 20.00 mL of a suitable supporting electrolyte. The resulting voltammogram gives peak currents of 1.40 µ A and 3.88 µ A for ascorbic acid and caffeine, respectively. A 0.500-mL aliquot of a standard solution containing 250.0 ppm ascorbic acid and 200.0 ppm caffeine is then added. A voltammogram of this solution gives peak currents of 2.80 µ A and 8.02 µ A for ascorbic acid and caffeine, respectively. Report the number of milligrams of ascorbic acid and of caffeine in the tablet.\n",
            "- 25. Ratana-ohpas and co-workers describe a stripping analysis method for determining levels of tin in canned fruit juices. 33 Standards containing 50.0 ppb Sn 4+ , 100.0 ppb Sn 4+ , and 150.0 ppb Sn 4+ were analyzed, giving signals of 83.0, 171.6, and 260.2, respectively. A 2.00-mL sample of lychee juice was mixed with 20.00 mL of 1:1 HCl/HNO3. A 0.500-mL portion of this mixture was added to 10 mL of 6 M HCl and the volume adjusted to 30.00 mL. Analysis of this diluted sample gave a signal of 128.2. Report the parts per million of Sn 4+ in the original sample of lychee juice.\n",
            "- 26. Sittampalam and Wilson described the preparation and use of an amperometric sensor for glucose. 34 The sensor is calibrated by measuring the steady-state current when it is immersed in standard solutions of glucose. A typical set of calibration data is shown in the following table.\n",
            "\n",
            "|   Milligrams of glucose/100 mL |   Current (arb. units) |\n",
            "|--------------------------------|------------------------|\n",
            "|                              2 |                   17.2 |\n",
            "|                              4 |                   32.9 |\n",
            "|                              6 |                   52.1 |\n",
            "|                              8 |                   68   |\n",
            "|                             10 |                   85.8 |\n",
            "\n",
            "A 2.00-mL sample of a solution containing an unknown amount of glucose is diluted to 10 mL in a volumetric flask, and a steady-state current of 23.6 is measured. What is the concentration of glucose in the sample in milligrams per 100 mL?\n",
            "\n",
            "- 27. Differential pulse polarography is used to determine the concentrations of lead, thallium, and indium in a mixture. 20\n",
            "\n",
            "Since the polarographic peaks for lead and thallium, and for thallium and indium overlap, a simultaneous analysis is necessary. Peak currents (in arbitrary units) at potentials of -0.385 V, -0.455 V, and -0.557 V were measured for a single standard solution and for a sample, giving the results shown in the following table.\n",
            "\n",
            "| Analyte   | Concentration ( µ g/mL)   |   i peak at -0.385 V |   i peak at -0.455V |   i peak at -0.557 V |\n",
            "|-----------|---------------------------|----------------------|---------------------|----------------------|\n",
            "| Pb 2+     | 1.0                       |                 26.1 |                 2.9 |                  0   |\n",
            "| Tl +      | 2.0                       |                  7.8 |                23.5 |                  3.2 |\n",
            "| In 3+     | 0.4                       |                  0   |                 0   |                 22.9 |\n",
            "| sample    |                           |                 60.6 |                28.8 |                 54.1 |\n",
            "\n",
            "Report the concentrations of Pb 2+ , Tl + , and In 3+ in the sample.\n",
            "\n",
            "28. Abass and colleagues developed an amperometric biosensor for NH4 + that uses the enzyme glutamate dehydrogenase to catalyze the following reaction.\n",
            "\n",
            "2-oxyglutarate( aq ) + NH4 + ( aq ) + NADH( aq ) → glutamate( aq ) + NAD + ( aq ) + H2O( l )\n",
            "\n",
            "where NADH is the reduced form of nicotinamide adenine dinucleotide. 35 The biosensor actually responds to the concentration of NADH; however, the rate of the reaction depends on the concentration of NH4 + . If the initial concentrations of 2-oxyglutarate and NADH are the same for all samples and standards, then the signal is proportional to the concentration of NH4 + . As shown in the following table, the sensitivity of the method is dependent on pH.\n",
            "\n",
            "|   pH | Sensitivity (nA s -1 M -1 )   |\n",
            "|------|-------------------------------|\n",
            "| 6.2  | 1.67 × 10 3                   |\n",
            "| 6.75 | 5.00 × 10 3                   |\n",
            "| 7.3  | 9.33 × 10 3                   |\n",
            "| 7.7  | 1.03 × 10 4                   |\n",
            "| 8.3  | 1.27 × 10 4                   |\n",
            "| 9.3  | 2.67 × 10 3                   |\n",
            "\n",
            "Two possible explanations for the effect of pH on the sensitivity of this analysis are the acid-base chemistry of NH4 + , and, the acid-base chemistry of the enzyme. Given that the p K a for NH4 + is 9.244, explain the source of this pHdependent sensitivity.\n",
            "\n",
            "- 29. The speciation scheme of Batley and Florence 22 requires eight measurements on four samples. After removing insoluble particulates by filtration, the solution is analyzed for the concentration of anodic stripping voltammetry (ASV) labile metal and the total concentration of metal. A portion of the filtered solution is passed through an ion-exchange column, and the concentrations of ASV metal and total metal are determined. A second portion of the filtered solution is irradiated with UV light, and the concentrations of ASV metal\n",
            "\n",
            "and total metal are measured. Finally, a third portion of the filtered solution is irradiated with UV light, passed through an ion-exchange column, and the concentrations of ASV-labile metal and total metal are again determined. The groups that are included in each measurement are summarized in the following table (see Table 11.12).\n",
            "\n",
            "| Separation Step                 | Groups Remaining in Solution   | Groups Contributing to Concentration of ASV- Labile Metal   | Groups Contributing to Concentration of Total Metal   |\n",
            "|---------------------------------|--------------------------------|-------------------------------------------------------------|-------------------------------------------------------|\n",
            "| 1. filtration                   | I, II, III, IV, V, VI, VII     | I, II, III                                                  | I, II, III, IV, V, VI, VII                            |\n",
            "| 2. ion- exchange                | II, III, VI, VII               | II, III                                                     | II, III, VI, VII                                      |\n",
            "| 3. UV irradiation               | I, II, III, IV, V, VI, VII     | I, II, III, IV, VI                                          | I, II, III, IV, V, VI, VII                            |\n",
            "| 4. UV irradiation/ ion-exchange | III, VII                       | III                                                         | III, VII                                              |\n",
            "\n",
            "(a) Show how these eight measurements can be used to determine the concentration of metal present in each group.\n",
            "\n",
            "(b) Batley and Florence 22a report the following results for the speciation of cadmium, lead, and copper in a sample of sea water\n",
            "\n",
            "| Measurement              |   ppb Cd 2+ |   ppb Pb 2+ |   ppb Cu 2+ |\n",
            "|--------------------------|-------------|-------------|-------------|\n",
            "| filtration: ASV-labile   |        0.24 |        0.39 |        0.26 |\n",
            "| filtration: total        |        0.28 |        0.5  |        0.4  |\n",
            "| ion-exchange: ASV-labile |        0.21 |        0.33 |        0.17 |\n",
            "| ion-exchange: total      |        0.26 |        0.43 |        0.24 |\n",
            "| UV: ASV-labile           |        0.26 |        0.37 |        0.33 |\n",
            "| UV: total                |        0.28 |        0.5  |        0.43 |\n",
            "| ion-exchange/UV: labile  |        0    |        0    |        0    |\n",
            "| ion-exchange/UV: total   |        0.02 |        0.12 |        0.1  |\n",
            "\n",
            "Determine the speciation of each metal in this sample of sea water.\n",
            "\n",
            "30. The polarographic half-wave potentials (versus the SCE) for Pb 2+ and Tl + in 1 M HCl are, respectively, -0.44 V and -0.45 V. In an electrolyte of 1 M NaOH, however, the half-wave potentials are -0.76 V for Pb 2+ and -0.48 V for Tl + . Why does the change in electrolyte have such a great effect on the half-wave potential for Pb 2+ , but not on the half-wave potential for Tl + ?\n",
            "\n",
            "- 31. The following data were collected for the reduction of Pb 2+ by normal pulse polarography\n",
            "\n",
            "|   Potential (V vs. SCE) |   Current ( µ A) |\n",
            "|-------------------------|------------------|\n",
            "|                  -0.345 |             0.16 |\n",
            "|                  -0.37  |             0.98 |\n",
            "|                  -0.383 |             2.05 |\n",
            "|                  -0.393 |             3.13 |\n",
            "|                  -0.409 |             4.62 |\n",
            "\n",
            "-0.420\n",
            "\n",
            "5.16\n",
            "\n",
            "## Modern Analytical Chemistry\n",
            "\n",
            "The limiting current was 5.67 µ A. Verify that the reduction reaction is reversible, and determine values for n and E 1/2 . The half-wave potentials for the normal pulse polarograms of Pb 2+ in the presence of several different concentrations of OH are shown in the following table.\n",
            "\n",
            "|   [OH - ] (M) |   E 1/2 (V vs. SCE) |\n",
            "|---------------|---------------------|\n",
            "|          0.05 |              -0.646 |\n",
            "|          0.1  |              -0.673 |\n",
            "|          0.15 |              -0.689 |\n",
            "|          0.3  |              -0.715 |\n",
            "\n",
            "Determine the stoichiometry of the Pb-hydroxide complex and its formation constant.\n",
            "\n",
            "- 32. The concentration of Cu 2+ in sea water may be determined by anodic stripping voltammetry at a hanging mercury drop electrode. To determine the total amount of Cu 2+ in a sample of sea water, it is necessary to free any copper bound in organic complexes. To a 20.00-mL sample of sea water is added 1 mL of 0.05 M HNO3 and 1 mL of 0.1% H2O2. The sample is irradiated with UV light for 8 h and then diluted to volume in a 25-mL volumetric flask. Deposition of Cu 2+ takes place at -0.3 V for 10 min, producing a peak current of 26.1 (arbitrary units). A second 20.00-mL sample of the sea water is treated identically, except that 0.1 mL of a 5.00 µ M solution of Cu 2+ is added, producing a peak current of 38.4. Report the concentration of Cu 2+ in the sea water in parts per billion.\n",
            "- 33. Thioamide drugs, such as that shown here, can be determined by cathodic stripping analysis. Deposition occurs at +0.05 V\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## 11I SUGGESTED READINGS\n",
            "\n",
            "The following general references provide a broad introduction to electrochemical methods of analysis.\n",
            "\n",
            "Bard, A. J.; Faulkner, L. R. Electrochemical Methods. Wiley: New York, 1980.\n",
            "\n",
            "Faulkner, L. R. 'Electrochemical Characterization of Chemical Systems.' In Kuwana, T. E., ed., Physical Methods in Modern Chemical Analysis, Vol. 3. Academic Press: New York, 1983, pp. 137-248.\n",
            "\n",
            "Kissinger, P. T.; Heineman, W. R. Laboratory Techniques in Electroanalytical Chemistry. 2nd ed. Marcel Dekker: New York, 1996.\n",
            "\n",
            "Lingane, J. J. Electroanalytical Chemistry. 2nd ed. Interscience: New York, 1958.\n",
            "\n",
            "versus the SCE. During the stripping step, the potential is scanned cathodically, and a stripping peak is observed at -0.55 V. In a typical application a 2.00-mL sample of urine was mixed with 2.00 mL of a pH 4.78 buffer. Following a 2-minute deposition, a peak current of 0.562 µ A was measured. A 0.10-mL addition of a 5.00 µ M solution of the thioamide drug was added to the same solution. A peak current of 0.837 µ A was recorded following the same 2-min deposition time. Report the drug's molar concentration in the urine sample.\n",
            "\n",
            "- 34. The concentration of vanadium (V) in sea water can be determined by adsorptive stripping voltammetry after forming a complex with catechol. 36 The catechol-V(V) complex is deposited on a hanging mercury drop electrode at a potential of -0.1 V (vs. Ag/AgCl). A cathodic potential scan gives a stripping peak that is proportional to the concentration of V(V) in the sample of sea water. The following standard additions were used to analyze a sample of sea water.\n",
            "\n",
            "| [V (V)] add (M)   |   i peak (nA) |\n",
            "|-------------------|---------------|\n",
            "| 2.0 × 10 -8       |            24 |\n",
            "| 4.0 × 10 -8       |            33 |\n",
            "| 8.0 × 10 -8       |            52 |\n",
            "| 1.2 × 10 -7       |            69 |\n",
            "| 1.8 × 10 -7       |            97 |\n",
            "| 2.8 × 10 -7       |           140 |\n",
            "\n",
            "Determine the molar concentration of V (V) in the sample of sea water, assuming that the standard additions result in a negligible change in the sample's volume.\n",
            "\n",
            "Sawyer, D. T.; Roberts, J. L., Jr. Experimental Electrochemistry for Chemists. Wiley-Interscience: New York, 1974.\n",
            "\n",
            "Vassos, B. H.; Ewing, G. W. Electroanalytical Chemistry. Wiley- Interscience: New York, 1983.\n",
            "\n",
            "Three short articles providing a good understanding of important general principles of electrochemistry follow.\n",
            "\n",
            "Faulkner, L. R. 'Understanding Electrochemistry: Some Distinctive Concepts,' J. Chem. Educ. 1983, 60, 262-264.\n",
            "\n",
            "Maloy, J. T. 'Factors Affecting the Shape of Current-Potential Curves,' J. Chem. Educ. 1983, 60, 285-289.\n",
            "\n",
            "Zoski, C. G. 'Charging Current Discrimination in Analytical Voltammetry,' J. Chem. Educ. 1986, 63, 910-914.\n",
            "\n",
            "Additional information on potentiometry and ion-selective electrodes can be found in the following sources.\n",
            "\n",
            "Bates, R. G. Determination of pH: Theory and Practice. 2nd ed. Wiley: New York, 1973.\n",
            "\n",
            "Buck, R. P. 'Potentiometry: pH Measurements and Ion Selective Electrodes.' In Weissberger, A., ed., Physical Methods of Organic Chemistry, Vol. 1, Part IIA. Wiley: New York, 1971, pp. 61-162.\n",
            "\n",
            "Cammann, K. Working with Ion-Selective Electrodes. Springer-Verlag: Berlin, 1977.\n",
            "\n",
            "Evans, A. Potentiometry and Ion-Selective Electrodes. Wiley: New York, 1987.\n",
            "\n",
            "Frant, M. S. 'Where Did Ion Selective Electrodes Come From?' J. Chem. Educ. 1997, 74, 159-166.\n",
            "\n",
            "Light, T. S. 'Industrial Use and Application of Ion-Selective Electrodes,' J. Chem. Educ. 1997, 74, 171-177.\n",
            "\n",
            "Rechnitz, G. A. 'Ion and Bio-Selective Membrane Electrodes,' J. Chem. Educ. 1983, 60, 282-284.\n",
            "\n",
            "Ruzicka, J. 'The Seventies-Golden Age for Ion-Selective Electrodes,' J. Chem. Educ. 1997, 74, 167-170.\n",
            "\n",
            "- Young, C. C. 'Evolution of Blood Chemistry Analyzers Based on Ion Selective Electrodes,' J. Chem. Educ. 1997, 74, 177-182.\n",
            "\n",
            "The following sources provide additional information on electrochemical biosensors.\n",
            "\n",
            "Alvarez-Icasa, M.; Bilitewski, U. 'Mass Production of Biosensors,' Anal. Chem. 1993, 65, 525A-533A.\n",
            "\n",
            "Meyerhoff, M. E.; Fu, B.; Bakker, E.; et al. 'Polyion-Sensitive Membrane Electrodes for Biomedical Analysis,' Anal. Chem. 1996, 68, 168A-175A.\n",
            "\n",
            "Nicolini, C.; Adami, M; Antolini, F.; et al. 'Biosensors: A Step to Bioelectronics,' Phys. World, May 1992, 30-34.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## REFERENCES\n",
            "\n",
            "- 1. Stork, J. T. Anal. Chem. 1993, 65, 344A-351A.\n",
            "- 2. Sawyer, D. T.; Roberts, J. L., Jr. Experimental Electrochemistry for Chemists. Wiley-Interscience: New York, 1974, p. 22.\n",
            "- 3. Cammann, K. Working with Ion-Selective Electrodes. Springer-Verlag: Berlin, 1977.\n",
            "- 4. (a) Papastathopoulos, D. S.; Rechnitz, G. A. Anal. Chim. Acta 1975, 79, 17; (b) Riechel, T. L. J. Chem. Educ. 1984, 61, 640-642.\n",
            "- 5. Tor, R.; Freeman, A. Anal. Chem. 1986, 58, 1042-1046.\n",
            "- 6. Lunte, C. E.; Heineman, W. R. 'Electrochemical Techniques in Bioanalysis' in Steckhan, E., ed. Topics in Current Chemistry. Vol. 143. Springer-Verlag: Berlin, 1988, p. 8.\n",
            "- 7. Procedure adapted from Kennedy, J. H. Analytical ChemistryPractice. Harcourt Brace Jovanovich: San Diego, 1984, p. 117-118.\n",
            "- 8. Hawkes, S. J. J. Chem. Educ. 1994, 71, 747-749.\n",
            "- 9. Kristensen, H. B.; Saloman, A.; Kokholm, G. Anal. Chem. 1991, 63, 885A-891A.\n",
            "- 10. Bates, R. G. Determination of pH: Theory and Practice. 2nd ed. Wiley: New York, 1973.\n",
            "- 11. Rogers, K. R.; Williams, L. R. Trends Anal. Chem. 1995, 14, 289-294.\n",
            "- 12. Rechnitz, G. A. Controlled-Potential Analysis. Macmillan: New York, 1963, p. 49.\n",
            "\n",
            "## Chapter 11 Electrochemical Methods of Analysis\n",
            "\n",
            "Schultz, J. S. 'Biosensors,' Sci. Am. August 1991, 64-69.\n",
            "\n",
            "Thompson, M.; Krull, U. 'Biosensors and the Transduction of Molecular Recognition,' Anal. Chem. 1991, 63, 393A-405A.\n",
            "\n",
            "249-252.\n",
            "\n",
            "Vadgama, P. 'Designing Biosensors,' Chem. Brit. 1992, 28, The following is a good source covering the clinical application of electrochemistry.\n",
            "\n",
            "Wang, J. Electroanalytical Techniques in Clinical Chemistry and Laboratory Medicine. VCH: New York, 1998.\n",
            "\n",
            "Coulometry is covered in the following texts.\n",
            "\n",
            "Milner, G. W. C.; Phillips, G. Coulometry in Analytical Chemistry. Pergamon: New York, 1967.\n",
            "\n",
            "Rechnitz, G. A. Controlled-Potential Analysis. Macmillan: New York, 1963.\n",
            "\n",
            "For a description of electrogravimetry, see the following resource.\n",
            "\n",
            "Tanaka, N. 'Electrodeposition,' In Kolthoff, I. M.; Elving, P. J., eds. Treatise on Analytical Chemistry, Part I: Theory and Practice, Vol. 4. Interscience: New York, 1963.\n",
            "\n",
            "The following sources provide additional information on polarography and pulse polarography.\n",
            "\n",
            "Flato, J. B. 'The Renaissance in Polarographic and Voltammetric Analysis,' Anal. Chem. 1972, 44(11), 75A-87A.\n",
            "\n",
            "- Kolthoff, I. M.; Lingane, J. J. Polarography. Interscience: New York, 1952.\n",
            "\n",
            "Osteryoung, J. 'Pulse Voltammetry,' J. Chem. Educ. 1983, 60, 296-298. Additional information on stripping voltammetry is available in the following text.\n",
            "\n",
            "Wang, J. Stripping Analysis. VCH Publishers: Deerfield Beach, FL, 1985, pp. 116-124.\n",
            "\n",
            "- 13. Procedure adapted from Bassett, J.; Denney, R. C.; Jeffery, G. H.; et al. Vogel's Textbook of Quantitative Inorganic Analysis. Longman: London, 1978, pp. 559-560.\n",
            "- 14. Curran, D. J. 'Constant-Current Coulometry.' Chapter 20 in Kissinger, P. T.; Heineman, W. R., eds. Laboratory Techniques in Electroanalytical Chemistry. Marcel Dekker, Inc.: New York, 1984, pp. 539-568.\n",
            "- 15. Maloy, J. T. J. Chem. Educ. 1983, 60, 285-289.\n",
            "- 16. Osteryoung, J. J. Chem. Educ. 1983, 60, 296-298.\n",
            "- 17. Peterson, W. M.; Wong, R. V. Am. Lab. November 1981, 116-128.\n",
            "- 18. Wang, J. Am. Lab. May 1985, 41-50.\n",
            "- 19. Karube, I.; Nomura, Y.; Arikawa, Y. Trends in Anal. Chem. 1995, 14, 295-299.\n",
            "- 20. Lanza, P. J. Chem. Educ. 1990, 67, 704-705.\n",
            "- 21. Wang, J. Stripping Analysis. VCH Publishers: Deerfield Beach, FL, 1985, pp. 116-124.\n",
            "- 22. (a) Batley, G. E.; Florence, T. M. Anal. Lett. 1976, 9, 379-388; (b) Batley, G. E.; Florence, T. M. Talanta 1977, 24, 151; (c) Batley, G. E.; Florence, T. M. Anal. Chem. 1980, 52, 1962-1963; (d) Florence, T. M.; Batley, G. E. CRC Crit. Rev. Anal. Chem. 1980, 9, 219-296.\n",
            "- 23. Brooks, M. A. 'Application of Electrochemistry to Pharmaceutical Analysis' Chapter 21. In Kissinger, P. T.; Heineman, W. R., eds.\n",
            "\n",
            "## 542 Modern Analytical Chemistry\n",
            "\n",
            "Laboratory Techniques in Electroanalytical Chemistry. Marcel Dekker, Inc.: New York, 1984, pp. 569-607.\n",
            "\n",
            "- 24. Cammann, K.; Lemke, U.; Rohen, A.; et al. Angew. Chem. Int. Ed. Engl. 1991, 30, 516-539.\n",
            "- 25. Procedure adapted from Pungor, E. A Practical Guide to Instrumental Analysis. CRC Press: Boca Raton, FL, 1995, pp. 34-37.\n",
            "- 26. Abe, T.; Lau, L. L.; Ewing, A. G. J. Am. Chem. Soc. 1991, 113, 7421-7423.\n",
            "- 27. Creager, S. E.; Lawrence, K. D.; Tibbets, C. R. J. Chem. Educ. 1995, 72, 274-276.\n",
            "- 28. Watanabe, K.; Okada, K.; Oda, H.; et al. Anal. Chim. Acta 1995, 316, 371-375.\n",
            "- 29. Mifflin, T. E.; Andriano, K. M.; Robbins, W. B. J. Chem. Educ. 1984, 61, 638-639.\n",
            "- 30. Meyerhoff, M. E.; Kovach, P. M. J. Chem. Educ. 1983, 9, 766-768.\n",
            "- 31. Wang, J.; Taha, Z. Anal. Chim. Acta 1991, 252, 215-221.\n",
            "- 32. Lau, O.; Luk, S.; Cheung, Y. Analyst, 1989, 114, 1047-1051.\n",
            "- 33. Ratana-ohpas, R; Kanatharana, P.; Ratana-ohpas, W.; et al. Anal. Chim. Acta 1996, 333, 115-118.\n",
            "- 34. Sittampalam, G; Wilson, G. S. J. Chem. Educ. 1982, 59, 70-73.\n",
            "- 35. Abass, A. K.; Hart, J. P.; Cowell, D. C.; et al. Anal. Chim. Acta 1998, 373, 1-8.\n",
            "- 36. van der Berg, C. M. G.; Huang, Z. Q. Anal. Chem. 1984, 56, 2383-2386.\n",
            "\n",
            "## Chapter 12\n",
            "\n",
            "## Chromatographic and Electrophoretic Methods\n",
            "\n",
            "D rawing from an arsenal of analytical techniques, many of which were the subject of the preceding four chapters, analytical chemists have designed methods for the analysis of analytes at increasingly lower concentrations and in increasingly more complex matrices. Despite the power of these techniques, they often suffer from a lack of selectivity. For this reason, many analytical procedures include a step to separate the analyte from potential interferents. Several separation methods, such as liquid-liquid extractions and solid-phase microextractions, were discussed in Chapter 7. In this chapter we consider two additional separation methods that combine separation and analysis: chromatography and electrophoresis.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Overview of Analytical Separations\n",
            "\n",
            "In Chapter 7 we examined several methods for separating an analyte from potential interferents. For example, in a liquid-liquid extraction the analyte and interferent are initially present in a single liquid phase. A second, immiscible liquid phase is introduced, and the two phases are thoroughly mixed by shaking. During this process the analyte and interferents partition themselves between the two phases to different extents, affecting their separation. Despite the power of these separation techniques, there are some significant limitations.\n",
            "\n",
            "## 12A.1 The Problem with Simple Separations\n",
            "\n",
            "Suppose we have a sample containing an analyte in a matrix that is incompatible with our analytical method. To determine the analyte's concentration we first separate it  from the matrix using, for example, a liquid-liquid extraction. If there are additional analytes, we may need to use additional extractions to isolate them from the analyte's matrix. For a complex mixture of analytes this quickly becomes a tedious process.\n",
            "\n",
            "Furthermore, the extent to which we can effect a separation depends on the distribution ratio of each species in the sample. To separate an analyte from its matrix, its distribution ratio must be significantly greater than that for all other components in the matrix. When the analyte's distribution ratio is similar to that of another species, then a separation becomes impossible. For example, let's assume that an analyte, A, and a matrix interferent, I, have distribution ratios of 5 and 0.5, respectively. In an attempt to separate the analyte from its matrix, a simple liquidliquid extraction is carried out using equal volumes of sample and a suitable extraction solvent. Following the treatment outlined in Chapter 7, it is easy to show that a single extraction removes approximately 83% of the analyte and 33% of the interferent. Although it is possible to remove 99% of A with three extractions, 70% of I is also removed. In fact, there is no practical combination of number of extractions or volume ratio of sample and extracting phases that produce an acceptable separation of the analyte and interferent by a simple liquid-liquid extraction.\n",
            "\n",
            "## 12A.2 A Better Way to Separate Mixtures\n",
            "\n",
            "The problem with a simple extraction is that the separation only occurs in one direction. In a liquid-liquid extraction, for example, we extract a solute from its initial  phase  into  the  extracting  phase.  Consider, again, the separation of an analyte and a matrix interferent with distribution ratios of 5 and 0.5, respectively. A single liquid-liquid extraction transfers 83% of the analyte and 33% of the interferent to the extracting phase (Figure 12.1). If the concentrations of A and I in the sample were identical, then their concentration ratio in the extracting phase after one extraction is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Thus, a single extraction improves the separation of the solutes by a factor of 2.5. As shown in Figure 12.1, a second extraction actually leads to a poorer separation. After combining the two portions of the extracting phase, the concentration ratio decreases to\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "A\n",
            "\n",
            "I\n",
            "\n",
            "We can improve the separation by first extracting the solutes into the extracting phase, and then extracting them back into a fresh portion of the initial phase (Figure 12.2). Because solute A has the larger distribution ratio, it is extracted to a greater extent during the first extraction and to a lesser extent during the second extraction. In this case the final concentration ratio of\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "## Figure 12.1\n",
            "\n",
            "Progress of a liquid-liquid extraction using two identical extractions of a sample (initial phase) with fresh portions of the extracting phase. All numbers are fractions of solute in the phases; A = analyte, I = interferent.\n",
            "\n",
            "## Figure 12.2\n",
            "\n",
            "Progress of a liquid-liquid extraction in which the solutes are first extracted into the extracting phase and then extracted back into a fresh portion of the initial phase. All numbers are fractions of solute in the phases; A = analyte, I = interferent.\n",
            "\n",
            "## countercurrent extraction\n",
            "\n",
            "A liquid-liquid extraction in which solutes are extracted back and forth between fresh portions of two extracting phases.\n",
            "\n",
            "## mobile phase\n",
            "\n",
            "In chromatography, the extracting phase that moves through the system.\n",
            "\n",
            "## stationary phase\n",
            "\n",
            "In chromatography, the extracting phase that remains in a fixed position.\n",
            "\n",
            "## chromatography\n",
            "\n",
            "A separation in which solutes partition between a mobile and stationary phase.\n",
            "\n",
            "## column chromatography\n",
            "\n",
            "A form of chromatography in which the stationary phase is retained in a column.\n",
            "\n",
            "## planar chromatography\n",
            "\n",
            "A form of chromatography in which the stationary phase is immobilized on a flat surface.\n",
            "\n",
            "in the extracting phase is significantly greater. The process of extracting the solutes back and forth between fresh portions of the two phases, which is called a countercurrent extraction, was developed by Craig in the 1940s. 1* The same phenomenon forms the basis of modern chromatography.\n",
            "\n",
            "Chromatographic separations are accomplished by continuously passing one sample-free phase, called a mobile phase, over a second sample-free phase that remains fixed, or stationary. The sample is injected, or placed, into the mobile phase. As it moves with the mobile phase, the sample's components partition themselves between the mobile and stationary phases. Those components whose distribution ratio favors the stationary phase require a longer time to pass through the system. Given sufficient time, and sufficient stationary and mobile phase, solutes with similar distribution ratios can be separated.\n",
            "\n",
            "The history of modern chromatography can be traced to the turn of the century when the Russian botanist Mikhail Tswett (1872-1919) used a column packed with a stationary phase of calcium carbonate to separate colored pigments from plant extracts. The sample was placed at the top of the column and carried through the stationary phase using a mobile phase of petroleum ether. As the sample moved through the column, the pigments in the plant extract separated into individual colored bands. Once the pigments were adequately separated, the calcium carbonate was removed from the column, sectioned, and the pigments recovered by extraction. Tswett named the technique chromatography, combining the Greek words for 'color' and 'to write.' There was little interest in Tswett's technique until 1931 when chromatography was reintroduced as an analytical technique for biochemical separations. Pioneering work by Martin and Synge in 1941 2 established the importance of liquid-liquid partition chromatography and led to the development of a theory for chromatographic separations; they were awarded the 1952 Nobel Prize in chemistry for this work. Since then, chromatography in its many forms has become the most important and widely used separation technique. Other separation methods, such as electrophoresis, effect a separation without the use of a stationary phase.\n",
            "\n",
            "## 12A. 3 Classifying Analytical Separations\n",
            "\n",
            "Analytical separations may be classified in three ways: by the physical state of the mobile phase and stationary phase; by the method of contact between the mobile phase and stationary phase; or by the chemical or physical mechanism responsible for separating the sample's constituents. The mobile phase is usually a liquid or a gas, and the stationary phase, when present, is a solid or a liquid film coated on a solid surface. Chromatographic techniques are often named by listing the type of mobile phase, followed by the type of stationary phase. Thus, in gas-liquid chromatography the mobile phase is a gas and the stationary phase is a liquid. If only one phase is indicated, as in gas chromatography, it is assumed to be the mobile phase.\n",
            "\n",
            "Two common approaches are used to bring the mobile phase and stationary phase into contact. In column chromatography, the stationary phase is placed in a narrow column through which the mobile phase moves under the influence of gravity or pressure. The stationary phase is either a solid or a thin, liquid film coating on a solid particulate packing material or the column's walls. In planar chromatography the  stationary  phase  coats  a  flat  glass,  metal,  or  plastic  plate\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "and is placed in a developing chamber. A reservoir containing the mobile phase is  placed in contact with the stationary phase, and the mobile phase moves by capillary action.\n",
            "\n",
            "The mechanism by which solutes separate provides a third means for characterizing a separation (Figure 12.3). In adsorption chromatography, solutes separate based on their ability to adsorb to a solid stationary phase. In partition chromatography, a thin liquid film coating a solid support serves as the stationary phase. Separation is based on a difference in the equilibrium partitioning of solutes between the liquid stationary phase and the mobile phase. Stationary phases consisting of a solid support with covalently attached anionic (e.g., -SO3 -) or cationic (e.g.,  -N(CH3)3 + )  functional  groups  are  used  in  ion-exchange  chromatography. Ionic solutes are attracted to the stationary phase by electrostatic forces.  Porous gels are used as stationary phases in size-exclusion chromatography, in which separation is due to differences in the size of the solutes. Large solutes are unable to penetrate into the porous stationary phase and so quickly pass through the column. Smaller solutes enter into the porous stationary phase, increasing the time spent on the column. Not all separation methods require a stationary phase. In an electrophoretic separation, for example, charged solutes migrate under the influence of an applied potential field. Differences in the mobility of the ions account for their separation.\n",
            "\n",
            "## 12B General Theory of Column Chromatography\n",
            "\n",
            "Of the two methods for bringing the stationary and mobile phases into contact, the  more important is column chromatography. In this section we develop a  general  theory  that  we  may  apply  to  any  form  of  column  chromatography. With appropriate modifications, this theory also can be applied to planar chromatography.\n",
            "\n",
            "A typical column chromatography experiment is outlined in Figure 12.4. Although the figure depicts a liquid-solid chromatographic experiment similar to that first used by Tswett, the design of the column and the physical state of the\n",
            "\n",
            "## Figure 12.4\n",
            "\n",
            "Progress of a column chromatographic separation showing the separation of two solute bands.\n",
            "\n",
            "## Figure 12.5\n",
            "\n",
            "Another view of the progress of a column chromatographic separation showing the separation of two solute bands.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 12.6\n",
            "\n",
            "Typical chromatogram of detector response as a function of retention time.\n",
            "\n",
            "## chromatogram\n",
            "\n",
            "A plot of the detector's signal as function of elution time or volume.\n",
            "\n",
            "## retention time\n",
            "\n",
            "The time a solute takes to move from the point of injection to the detector ( t r ).\n",
            "\n",
            "## retention volume\n",
            "\n",
            "The volume of mobile phase needed to move a solute from its point of injection to the detector ( V r ).\n",
            "\n",
            "## baseline width\n",
            "\n",
            "The width of a solute's chromatographic band measured at the baseline ( w ).\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "stationary and mobile phases may vary. The sample is introduced at the top of the column as a narrow band. Ideally, the solute's initial concentration profile is rectangular (Figure 12.5a). As the sample moves down the column the solutes begin to  separate,  and the individual solute bands begin to broaden and develop a Gaussian profile (Figures 12.5b,c). If the strength of each solute's interaction with the stationary phase is sufficiently different, then the solutes separate into individual bands (Figure 12.5d). The progress of a chromatographic separation is monitored with a suitable detector situated at the end of the column. A plot of the detector's signal as a function of time or volume of eluted mobile phase is known as a chromatogram (Figure 12.6) and consists of a peak for each of the separated solute bands.\n",
            "\n",
            "A chromatographic peak may be characterized in many ways, two of which are shown in Figure 12.7. The retention time, t r , is the elapsed time from the introduction of the solute to the peak maximum. The retention time also can be measured indirectly as the volume of mobile phase eluting between the solute's introduction and the appearance of the solute's peak maximum. This is known as the retention volume, V r . Dividing the retention volume by the mobile phase's flow rate, u, gives the retention time.\n",
            "\n",
            "The second important parameter is the chromatographic peak's width at the baseline, w. As shown in Figure 12.7, baseline width is determined by the intersection with the baseline of tangent lines drawn through the inflection points on either side of the chromatographic peak. Baseline width is measured in units of time or volume, depending on whether the retention time or retention volume is of interest.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Besides the solute peak, Figure 12.7 also shows a small peak eluted soon after the sample is injected into the mobile phase. This peak results from solutes that move through the column at the same rate as the mobile phase. Since these solutes do not interact with the stationary phase, they are considered nonretained. The time or volume of mobile phase required to elute nonretained components is called the column's void time, t m, or void volume.\n",
            "\n",
            "## 12B.1 Chromatographic Resolution\n",
            "\n",
            "The goal of chromatography is to separate a sample into a series of chromatographic peaks, each representing a single component of the sample. Resolution is a quantitative measure of the degree of separation between two chromatographic peaks, A and B, and is defined as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "As shown in Figure 12.8, the degree of separation between two chromatographic peaks improves with an increase in R. For two peaks of equal size, a resolution of 1.5 corresponds to an overlap in area of only 0.13%. Because resolution is a quantitative measure of a separation's success, it provides a useful way to determine if a change in experimental conditions leads to a better separation.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 12.1\n",
            "\n",
            "In a chromatographic analysis of lemon oil a peak for limonene has a retention time of 8.36 min with a baseline width of 0.96 min. γ -Terpinene elutes at 9.54 min, with a baseline width of 0.64 min. What is the resolution between the two peaks?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "Using equation 12.1, we find that the resolution is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "## Figure 12.7\n",
            "\n",
            "Measurement of the column's void time, t m, and the retention time, t r , and baseline width, w, for a solute.\n",
            "\n",
            "## void time\n",
            "\n",
            "The time required for unretained solutes to move from the point of injection to the detector ( t m).\n",
            "\n",
            "## void volume\n",
            "\n",
            "The volume of mobile phase needed to move an unretained solute from the point of injection to the detector.\n",
            "\n",
            "## resolution\n",
            "\n",
            "The separation between two chromatographic bands ( R ).\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 12.8\n",
            "\n",
            "Three examples of chromatographic resolution.\n",
            "\n",
            "## Figure 12.9\n",
            "\n",
            "Two methods for improving chromatographic resolution: (a) Original separation showing a pair of poorly resolved solutes; (b) Improvement in resolution due to an increase in column efficiency; (c) Improvement in resolution due to a change in column selectivity.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "From equation 12.1 it is clear that resolution may be improved either by increasing ∆ t r or by decreasing w A or w B (Figure 12.9). We can increase ∆ t r by enhancing the interaction of the solutes with the column or by increasing the column's selectivity for one of the solutes. Peak width is a kinetic effect associated with the solute's movement within and between the mobile phase and stationary phase. The effect is governed by several factors that are collectively called column efficiency. Each of these factors is considered in more detail in the following sections.\n",
            "\n",
            "## 12B.2 Capacity Factor\n",
            "\n",
            "The distribution of a solute, S, between the mobile phase and stationary phase can be represented by an equilibrium reaction\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and its associated partition coefficient, K D, and distribution ratio, D,\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where the subscripts m and s refer to the mobile phase and stationary phase, respectively. As long as the solute is not involved in any additional equilibria in either the mobile phase or stationary phase, the equilibrium partition coefficient and the distribution ratio will be the same.\n",
            "\n",
            "Conservation of mass requires that the total moles of solute remain constant throughout the separation, thus\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving equation 12.3 for the moles of solute in the stationary phase and substituting into equation 12.2 gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where V m and V s are the volumes of the mobile and stationary phases. Rearranging and solving for the fraction of solute in the mobile phase, f m, gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Note that this equation is identical to that describing the extraction of a solute in a liquid-liquid extraction (equation 7.25 in Chapter 7). Since the volumes of the stationary and mobile phase may not be known, equation 12.4 is simplified by dividing both the numerator and denominator by V m; thus\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "is the solute's capacity factor.\n",
            "\n",
            "A solute's capacity factor can be determined from a chromatogram by measuring the column's void time, t m, and the solute's retention time, t r (see Figure 12.7). The mobile phase's average linear velocity, u, is equal to the length of the column, L, divided by the time required to elute a nonretained solute.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "By the same reasoning, the solute's average linear velocity, v, is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The solute can only move through the column when it is in the mobile phase. Its average linear velocity, therefore, is simply the product of the mobile phase's average linear velocity and the fraction of solute present in the mobile phase.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Substituting equations 12.5, 12.7, and 12.8 into equation 12.9 gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Finally, solving this equation for k ′ gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where t r ′ is known as the adjusted retention time.\n",
            "\n",
            "## capacity factor\n",
            "\n",
            "A measure of how strongly a solute is retained by the stationary phase ( k ′ ).\n",
            "\n",
            "## adjusted retention time\n",
            "\n",
            "The difference between a solute's retention time and column's void time ( t r ′ ).\n",
            "\n",
            "## selectivity factor\n",
            "\n",
            "The ratio of capacity factors for two solutes showing the column's selectivity for one of the solutes ( α ).\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 12.2\n",
            "\n",
            "In a chromatographic analysis of low-molecular-weight acids, butyric acid elutes with a retention time of 7.63 min. The column's void time is 0.31 min. Calculate the capacity factor for butyric acid.\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "## 12B. 3 Column Selectivity\n",
            "\n",
            "The relative selectivity of a chromatographic column for a pair of solutes is given by the selectivity factor, α , which is defined as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The identities of the solutes are defined such that solute A always has the smaller retention time. Accordingly, the selectivity factor is equal to 1 when the solutes elute with identical retention times, and is greater than 1 when t r,B is  greater than t r,A .\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 12. 3\n",
            "\n",
            "In the same chromatographic analysis for low-molecular-weight acids considered in Example 12.2, the retention time for isobutyric acid is 5.98 min. What is the selectivity factor for isobutyric acid and butyric acid?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "First we must calculate the capacity factor for isobutyric acid. Using the void time from Example 12.2, this is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The selectivity factor, therefore, is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "## 12B. 4 Column Efficiency\n",
            "\n",
            "At the beginning of a chromatographic separation the solute occupies a narrow band of finite width. As the solute passes through the column, the width of its band\n",
            "\n",
            "continually increases in a process called band broadening. Column efficiency provides a quantitative measure of the extent of band broadening.\n",
            "\n",
            "In their original theoretical model of chromatography, Martin and Synge 2 treated the chromatographic column as though it consists of discrete sections at which partitioning of the solute between the stationary and mobile phases occurs. They called each section a theoretical plate and defined column efficiency in terms of the number of theoretical plates, N, or the height of a theoretical plate, H; where\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "A column's efficiency improves with an increase in the number of theoretical plates or a decrease in the height of a theoretical plate.\n",
            "\n",
            "Assuming a Gaussian profile, the extent of band broadening is measured by the variance or standard deviation of a chromatographic peak. The height of a theoretical plate is defined as the variance per unit length of the column\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where the variance, σ 2 ,  has  units  of  distance  squared.  Because retention time and peak width are usually measured in seconds or minutes, it is more convenient to express the standard deviation in units of time, τ , by dividing σ by the mobile phase's average linear velocity.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "When a chromatographic peak has a Gaussian shape, its width at the baseline, w, is four times its standard deviation, τ .\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Combining equations 12.13 through 12.15 gives the height of a theoretical plate in terms of the easily measured chromatographic parameters t r and w.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The number of theoretical plates in a chromatographic column is obtained by combining equations 12.12 and 12.16.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Alternatively, the number of theoretical plates can be approximated as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where w 1/2 is the width of the chromatographic peak at half its height.\n",
            "\n",
            "## band broadening\n",
            "\n",
            "The increase in a solute's baseline width as it moves from the point of injection to the detector.\n",
            "\n",
            "## theoretical plate\n",
            "\n",
            "A quantitative means of evaluating column efficiency that treats the column as though it consists of a series of small zones, or plates, in which partitioning between the mobile and stationary phases occurs.\n",
            "\n",
            "## peak capacity\n",
            "\n",
            "The maximum number of solutes that can be resolved on a particular column ( n c ).\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 12. 4\n",
            "\n",
            "A chromatographic analysis for the chlorinated pesticide Dieldrin gives a peak with a retention time of 8.68 min and a baseline width of 0.29 min. How many theoretical plates are involved in this separation? Given that the column used in this analysis is 2.0 meters long, what is the height of a theoretical plate?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "Using equation 12.17, the number of theoretical plates is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving equation 12.12 for H gives the average height of a theoretical plate as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "It is important to remember that a theoretical plate is an artificial construct and that no such plates exist in a chromatographic column. In fact, the number of theoretical plates depends on both the properties of the column and the solute. As a result, the number of theoretical plates for a column is not fixed and may vary from solute to solute.\n",
            "\n",
            "## 12B. 5 Peak Capacity\n",
            "\n",
            "Another important consideration is the number of solutes that can be baseline resolved on a given column. An estimate of a column's peak capacity, n c , is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where V min and V max are the smallest and largest volumes of mobile phase in which a solute can be eluted and detected. 3 A column with 10,000 theoretical plates, for example, can resolve no more than\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "if  the  minimum and maximum volumes of mobile phase in which the solutes can elute are 1 mL and 30 mL. This estimate provides an upper bound on the number of solutes that might be separated and may help to exclude from consideration columns that do not have enough theoretical plates to separate a complex mixture. Just because a column's theoretical peak capacity is larger than the number of solutes to be separated, however, does not mean that the separation will be feasible. In most situations the peak capacity obtained is less\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "than the estimated value because the retention characteristics of some solutes are too similar to effect their separation. Nevertheless, columns with more theoretical plates, or a greater range of possible elution volumes, are more likely to separate a complex mixture.\n",
            "\n",
            "## 12B.6 Nonideal Behavior\n",
            "\n",
            "The treatment of chromatography outlined in Section 12B assumes that a solute elutes as a symmetrical band, such as that shown in Figure 12.7. This ideal behavior occurs when the solute's partition coefficient, K D, is constant for all concentrations of solute. In some situations, chromatographic peaks show nonideal behavior, leading to asymmetrical peaks, similar to those shown in Figure 12.10. The chromatographic peak in Figure 12.10a is an example of ' fronting ' and is most often the result of overloading the column with sample. Figure 12.10b, which is an example of ' tailing, '  occurs when some sites on the stationary phase retain the solutes more strongly than other sites.\n",
            "\n",
            "## Figure 12.10\n",
            "\n",
            "Nonideal asymmetrical chromatographic bands showing (a) fronting and (b) tailing. Also depicted are the corresponding sorption isotherms showing the relationship between the concentration of solute in the stationary phase as a function of its concentration in the mobile phase.\n",
            "\n",
            "## fronting\n",
            "\n",
            "A tail at the beginning of a chromatographic peak, usually due to injecting too much sample.\n",
            "\n",
            "## tailing\n",
            "\n",
            "A tail at the end of a chromatographic peak, usually due to the presence of highly active sites in the stationary phase.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Optimizing Chromatographic Separations\n",
            "\n",
            "Now that we have defined capacity factor, selectivity, and column efficiency we consider their relationship to chromatographic resolution. Since we are only interested in the resolution between solutes eluting with similar retention times, it is safe to assume that the peak widths for the two solutes are approximately the same. Equation 12.1, therefore, is written as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving equation 12.17 for w B and substituting into equation 12.19 gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The retention times for solutes A and B are replaced with their respective capacity factors by rearranging equation 12.10\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and substituting into equation 12.20.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Finally, solute A's capacity factor is eliminated using equation 12.11. After rearranging, the equation for the resolution between the chromatographic peaks for solutes A and B is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Besides resolution, another important factor in chromatography is the amount of time required to elute a pair of solutes. The time needed to elute solute B is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Equations 12.21 and 12.22 contain terms corresponding to column efficiency, column selectivity, and capacity factor. These terms can be varied, more or less independently, to obtain the desired resolution and analysis time for a pair of solutes. The first term, which is a function of the number of theoretical plates or the height of a theoretical plate, accounts for the effect of column efficiency. The second term is a function of α and accounts for the influence of column selectivity. Finally, the third term in both equations is a function of k B, and accounts for the effect of solute ′ B's capacity factor. Manipulating these parameters to improve resolution is the subject of the remainder of this section.\n",
            "\n",
            "## 12C.1 Using the Capacity Factor to Optimize Resolution\n",
            "\n",
            "One of the simplest ways to improve resolution is to adjust the capacity factor for solute B. If all other terms in equation 12.21 remain constant, increasing k B ′ improves resolution. As shown in Figure 12.11, however, the effect is greatest when the\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "original capacity factor is small. Furthermore, large increases in k B ′ do not lead to proportionally larger increases in resolution. For example, when the original value of k B ′ is 1, increasing its value to 10 gives an 82% improvement in resolution; a further increase to 15 provides a net improvement in resolution of only 87.5%.\n",
            "\n",
            "Any improvement in resolution obtained by increasing k B ′ generally comes at the expense of a longer analysis time. This is also indicated in Figure 12.11, which shows the relative change in retention time as a function of the new capacity factor. Note that a minimum in the retention time curve occurs when k B ′ is equal to 2, and that retention time increases in either direction. Increasing k B ′ from 2 to 10, for example, approximately doubles solute B's retention time.\n",
            "\n",
            "The relationship between capacity factor and analysis time can be advantageous when a separation produces an acceptable resolution with a large k B. In this case it ′ may be possible to decrease k B ′ with little loss in resolution while significantly shortening the analysis time.\n",
            "\n",
            "A solute's capacity factor is directly proportional to its distribution ratio (equation 12.6), which, in turn, is proportional to the solute's equilibrium distribution constant. To increase k B ′ without significantly changing α ,  which also is a function of k B,  it  is  necessary  to  alter  chromatographic conditions in a way that leads to a ′ general, nonselective increase in the capacity factor for both solutes. In gas chromatography, this is usually accomplished by decreasing the column's temperature. At a lower temperature a solute's vapor pressure decreases, ensuring that it spends more time in the stationary phase increasing its capacity factor. In liquid chromatography, changing the mobile phase's solvent strength is the easiest way to change a solute's capacity factor. When the mobile phase has a low solvent strength, solutes spend proportionally more time in the stationary phase, thereby increasing their capacity factors. Additionally, equation 12.6 shows that the capacity factor is proportional to the volume of stationary phase. Increasing the volume of stationary phase, therefore, also leads to an increase in kB. ′\n",
            "\n",
            "Adjusting the capacity factor to improve resolution between one pair of solutes may lead to an unacceptably long retention time for other solutes. For example, improving resolution for solutes with short retention times by increasing\n",
            "\n",
            "## Figure 12.11\n",
            "\n",
            "′\n",
            "\n",
            "Effect of a change in k B on resolution and retention time. The original value of k B ′ is assumed to be 1.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "zutrpzr0x254"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "source": [
        "!pip install python-docx"
      ],
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PW3JNcEzx4Hy",
        "outputId": "fb84195e-007d-4455-9e3f-93ec7689295a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: python-docx in /usr/local/lib/python3.11/dist-packages (1.1.2)\n",
            "Requirement already satisfied: lxml>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from python-docx) (5.4.0)\n",
            "Requirement already satisfied: typing-extensions>=4.9.0 in /usr/local/lib/python3.11/dist-packages (from python-docx) (4.13.2)\n"
          ]
        }
      ]
    },
    {
      "source": [
        "from docling.document_converter import DocumentConverter\n",
        "from docx import Document\n",
        "\n",
        "# ... (كود تحويل الملفات كما هو) ...\n",
        "\n",
        "# إنشاء مستند وورد جديد\n",
        "document = Document()\n",
        "\n",
        "# إضافة نتائج التحويل إلى المستند\n",
        "document.add_paragraph(result.document.export_to_markdown())\n",
        "\n",
        "# حفظ المستند\n",
        "document.save(\"output.docx\")"
      ],
      "cell_type": "code",
      "metadata": {
        "id": "GP2_WiXrx5I5"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from docling.document_converter import DocumentConverter\n",
        "from docx import Document\n",
        "\n",
        "\n",
        "\n",
        "from docling.document_converter import DocumentConverter\n",
        "\n",
        "source = \"/content/1.pdf\"  # document per local path or URL\n",
        "converter = DocumentConverter()\n",
        "result = converter.convert(source)\n",
        "print(result.document.export_to_markdown())  # output: \"## Docling Technical Report[...]\"\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "document = Document()\n",
        "\n",
        "# إضافة نتائج التحويل إلى المستند\n",
        "document.add_paragraph(result.document.export_to_markdown())\n",
        "\n",
        "# حفظ المستند\n",
        "document.save(\"output.docx\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fESSJ-3yx225",
        "outputId": "84845d8a-fdf6-4d67-b891-04cedf65f8aa"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "a single analyte. Instead, the membrane potential is proportional to the concentration of all  ions  in  the  sample  solution  capable  of  interacting  at  the  membrane's active sites. Equation 11.8 can be generalized to include the contribution of an interferent, I,\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where z A and z I are the charges of the analyte and interferent, and K A,I is a selectivity coefficient accounting for the relative response of the interferent.* The selectivity coefficient is defined as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where [A] E and [I] E are the concentrations of analyte and interferent yielding identical cell potentials. When the selectivity coefficient is 1.00, the membrane responds equally to the analyte and interferent. A membrane shows good selectivity for the analyte when K A,I is significantly less than 1.00.\n",
            "\n",
            "Selectivity coefficients for most commercially available ion-selective electrodes are provided by the manufacturer. If the selectivity coefficient is unknown, it can be determined experimentally. The easiest method for determining K A,I is to prepare a series of solutions, each of which contains the same concentration of interferent, [I]add, but a different concentration of analyte. A plot of cell potential versus the log of the analyte's concentration has two distinct linear regions (Figure 11.11). When the analyte's concentration is significantly larger than K A,I [I]add ,  the  potential  is  a  linear  function  of  log [A],  as given by equation 11.8. If K A,I [I]add is  significantly  larger  than  the  analyte's concentration, however, the cell potential remains constant. The concentration of analyte and interferent at the intersection of these two linear regions is used to calculate K A,I .\n",
            "\n",
            "Glass Ion-Selective Electrodes The first commercial glass electrodes were manufactured using Corning 015, a glass with a composition of approximately 22% Na2O, 6% CaO, and 72% SiO2. When immersed in an aqueous solution, the outer approximately 10 nm of the membrane becomes hydrated over the course of several  hours.  Hydration  of  the  glass  membrane  results  in  the  formation of negatively charged sites, G - , that are part of the glass membrane's silica framework. Sodium ions, which are able to move through the hydrated layer, serve as the counterions. Hydrogen ions from solution diffuse into the membrane and, since they bind more strongly to the glass than does Na + , displace the sodium ions\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "giving rise to the membrane's selectivity for H + . The transport of charge across the membrane is carried by the Na + ions. The potential of glass electrodes using Corning 015 obeys the equation\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "over a pH range of approximately 0.5-9. Above a pH of 9-10, the glass membrane may become more responsive to other cations, such as Na + and K + .\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.11\n",
            "\n",
            "Plot of cell potential versus the log of the analyte's concentration in the presence of a fixed concentration of interferent, showing the determination of the selectivity coefficient.\n",
            "\n",
            "## glass electrode\n",
            "\n",
            "An ion-selective electrode based on a glass membrane in which the potential develops from an ion-exchange reaction on the membrane's surface.\n",
            "\n",
            "## Figure 11.12\n",
            "\n",
            "Schematic diagram of a combination glass electrode for measuring pH.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11. 4\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "The selectivity coefficient K H + /Na + for Corning 015 is approximately 10 -11 . What error in pH is expected for a solution of 0.05 M NaOH.\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "A solution of 0.05 M NaOH has an actual H + , [H + ]act, concentration of 2 × 10 -13 M and a pH of 12.7. The electrode responds, however, to both H + and Na + , with the apparent concentration of H + , [H + ]app, given by\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The apparent concentration of H + corresponds to a pH of 12.2, giving an error of -0.5 pH units.\n",
            "\n",
            "Replacing Na2O and CaO with Li2O and BaO extends the useful pH range of glass membrane electrodes to pH levels greater than 12.\n",
            "\n",
            "Glass membrane pH electrodes are often available in a combination form that includes both the indicator and the reference electrode. The use of a single electrode greatly simplifies the measurement of pH. An example of a typical combination electrode is shown in Figure 11.12.\n",
            "\n",
            "| Table 11.1   | Representative Examples of Glass Membrane Ion-Selective Electrodes   | Representative Examples of Glass Membrane Ion-Selective Electrodes   |\n",
            "|--------------|----------------------------------------------------------------------|----------------------------------------------------------------------|\n",
            "| Analyte      | Membrane Composition                                                 | Selectivity Coefficients a                                           |\n",
            "| Na +         | 11% Na 2 O, 18% Al 2 O 3 , 71% SiO 2                                 | K Na + /H + = 1000 K Na + /K + = K Na + /Li + = 10 -3                |\n",
            "| Li +         | 15% Li 2 O, 25% Al 2 O 3 , 60% SiO 2                                 | K Li + /Na + = 0.3 K Li + /K + = 10 -3                               |\n",
            "| K +          | 27% Na 2 O, 5% Al 2 O 3 , 68% SiO 2                                  | K K + /Na + = 0.05                                                   |\n",
            "\n",
            "a Selectivity constants are approximate, and those found experimentally may vary substantially from the listed values. 3\n",
            "\n",
            "The response of the Corning 015 glass membrane to monovalent cations other than H + at high pH led to the development of glass membranes possessing a greater selectivity for other cations. For example, a glass membrane with a composition of 11% Na2O, 18% Al2O3, and 71% SiO2 is used as a Na + ion-selective electrode. Other glass electrodes have been developed for the analysis of Li + , K + , Rb + , Cs + , NH4 + , Ag + ,  and  Tl + .  Several  representative examples of glass membrane electrodes are listed in Table 11.1.\n",
            "\n",
            "Since the typical thickness of the glass membrane in an ion-selective electrode is about 50 µ m, they must be handled carefully to prevent the formation of cracks or breakage. Before a glass electrode can be used it must be conditioned by soaking for several hours in a solution containing the analyte. Glass electrodes should not be allowed to dry out, as this destroys the membrane's hydrated layer. If a glass electrode has been allowed to dry out, it must be reconditioned before it can be used. The composition of a glass membrane changes over time, affecting the electrode's performance. The average lifetime for a glass electrode is several years.\n",
            "\n",
            "Crystalline Solid-State Ion-Selective Electrodes Solid-state ion-selective electrodes use membranes fashioned from polycrystalline or single-crystal inorganic salts.  Polycrystalline ion-selective electrodes are made by forming a thin pellet of Ag2S, or a mixture of Ag2S and either a second silver salt or another metal sulfide. The pellet, which is 1-2 mm in thickness, is sealed into the end of a nonconducting plastic cylinder, and an internal solution containing the analyte and a reference electrode are placed in the cylinder. Charge is carried across the membrane by Ag + ions.\n",
            "\n",
            "The membrane potential for a Ag2S pellet develops as the result of a difference in the equilibrium position of the solubility reaction\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "on the two sides of the membrane. When used to monitor the concentration of Ag + ions, the cell potential is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The membrane also responds to the concentration of S 2- , with the cell potential given as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "solid-state ion-selective electrode An ion-selective electrode based on a sparingly soluble inorganic crystalline material.\n",
            "\n",
            "If a mixture of an insoluble silver salt and Ag2S is used to make the membrane, then the membrane potential also responds to the concentration of the anion of the added silver salt. Thus, pellets made from a mixture of Ag2S and AgCl can serve as a Cl ion-selective electrode, with a cell potential of\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Membranes fashioned from a mixture of Ag2S with CdS, CuS, or PbS are used to make ion-selective electrodes that respond to the concentration of Cd 2+ , Cu 2+ , or Pb 2+ . In this case the cell potential is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where [M 2+ ] is the concentration of the appropriate metal ion.\n",
            "\n",
            "Several examples of polycrystalline, Ag2S-based ion-selective electrodes are listed in Table 11.2. The selectivity of these ion-selective electrodes is determined by solubility. Thus, a Cl ion-selective electrode constructed using a Ag2S/AgCl membrane is more selective for Br -( K Cl - /Br - = 10 2 ) and I -( K Cl - /l - = 10 6 ) since AgBr and AgI are less soluble than AgCl. If the concentration of Br is  sufficiently high, the AgCl at the membrane-solution interface is replaced by AgBr, and the electrode's response to Cl decreases substantially. Most of the ion-selective electrodes listed in Table 11.2 can be used over an extended range of pH levels. The equilibrium between S 2and HS limits the analysis for S 2to a pH range of 13-14. Solutions of CN - , on the other hand, must be kept basic to avoid the release of HCN.\n",
            "\n",
            "The membrane of a F ion-selective electrode is fashioned from a single crystal of LaF3 that is usually doped with a small amount of EuF2 to enhance the membrane's conductivity. Since EuF2 provides only two F ions, compared with three for LaF3, each EuF2 produces a vacancy in the crystal lattice. Fluoride ions move through the membrane by moving into adjacent vacancies. The LaF3 membrane is sealed into the end of a nonconducting plastic tube, with a standard solution of F - , typically 0.1 M NaF, and a Ag/AgCl reference electrode.\n",
            "\n",
            "The membrane potential for a F ion-selective electrode results from a difference in the solubility of LaF3 on opposite sides of the membrane, with the potential given by\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "One advantage of the F ion-selective electrode is its freedom from interference. The only significant exception is OH -( K F - /OH = 0.1), which imposes a maximum pH limit for a successful analysis.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11. 5\n",
            "\n",
            "What is the maximum pH that can be tolerated in the analysis of a 1 × 10 -5 M solution of F if the error is to be less than 1%?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "To achieve an error of less than 1%, the difference between the actual, [F - ]act, and apparent, [F - ]app, concentrations of F must be less than 1%; thus\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving for [OH - ], and making appropriate substitutions gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "corresponding to a pH of less than 8.\n",
            "\n",
            "| Table 11.2   | Representative Examples of Ion-Selective Electrodes   | Polycrystalline                                                                                             |\n",
            "|--------------|-------------------------------------------------------|-------------------------------------------------------------------------------------------------------------|\n",
            "| Analyte      | Membrane Composition                                  | Selectivity Coefficients a                                                                                  |\n",
            "| Ag +         | Ag 2 S                                                | K Ag + /Cu 2+ = 10 -6 K Ag + /Pb 2+ = 10 -10 Hg 2+ interferes                                               |\n",
            "| Cd 2+        | CdS/Ag 2 S                                            | K Cd 2+ /Fe 2+ = 200 K Cd 2+ /Pb 2+ = 6 Ag + , Hg 2+ , Cu 2+ must be absent                                 |\n",
            "| Cu 2+        | CuS/Ag 2 S                                            | K Cu 2+ /Fe 3+ = 10 K Cu 2+ /Cu + = 1 Ag + , Hg 2+ must be absent                                           |\n",
            "| Pb 2+        | PbS/Ag 2 S                                            | K Pb 2+ /Fe 3+ = 1 K Pb 2+ /Cd 2+ = 1 Ag + , Hg 2+ , Cu 2+ must be absent                                   |\n",
            "| Br -         | AgBr/Ag 2 S                                           | K Br - /I - = 5000 K Br - /CN - = 100 K Br - /Cl - = 5 × 10 -3 K Br - /OH - = 1 × 10 -5 S 2- must be absent |\n",
            "| Cl -         | AgCl/Ag 2 S                                           | K Cl - /I - = 1 × 10 6 K Cl - /CN - = 1 × 10 4 K Cl - /Br - = 100 K Cl - /OH - = 0.01 S 2- must be absent   |\n",
            "| CN -         | AgI/Ag 2 S                                            | K CN - /I - = 100 K CN - /Br - = 1 × 10 -4 K CN - /Cl - = 1 × 10 -6 K CN - /OH - = 1 × 10 -8                |\n",
            "| I -          | AgI/Ag 2 S                                            | S must be absent K I - /S 2- = 30 K I - /CN - = 0.01 K I - /Br - = 1 × 10 -4 K = 1 × 10 -6                  |\n",
            "| SCN -        | AgSCN/Ag 2 S                                          | K I - /OH - = 1 × 10 -7 K SCN - /I - = 1000 K SCN - /Br - = 100 K SCN - /CN - = 100 K SCN - /Cl - = 0.1     |\n",
            "| S 2-         | Ag 2 S                                                | K SCN - /OH - = 0.01 S 2- must be absent Hg 2+ interferes                                                   |\n",
            "\n",
            "Figure 11.13\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Structure and formula of di-( n -decyl) phosphate.\n",
            "\n",
            "## Figure 11.14\n",
            "\n",
            "Schematic diagram of a Ca 2+ liquid-based ion-selective electrode.\n",
            "\n",
            "## ionophore\n",
            "\n",
            "A neutral ligand whose exterior is hydrophobic and whose interior is hydrophilic.\n",
            "\n",
            "liquid-based ion-selective electrode An ion-selective electrode in which a chelating agent is incorporated into a hydrophobic membrane.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Below a pH of 4 the predominate form of fluoride in solution is HF, which, unlike F , does not contribute to the membrane potential. For this reason, an analysis for total fluoride must be carried out at a pH greater than 4.\n",
            "\n",
            "Unlike ion-selective electrodes using glass membranes, crystalline solid-state ion-selective electrodes do not need to be conditioned before use and may be stored dry. The surface of the electrode is subject to poisoning, as described earlier for a Cl ISE in contact with an excessive concentration of Br - . When this happens, the electrode can be returned to its original condition by sanding and polishing the crystalline membrane.\n",
            "\n",
            "Liquid-Based Ion-Selective Electrodes Another approach to constructing an ion-selective electrode is to use a hydrophobic membrane containing a selective, liquid organic complexing agent. Three types of organic liquids have been used: cation exchangers, anion exchangers, and neutral ionophores. When the analyte's concentration on the two sides of the membrane is different, a membrane potential is the result. Current is carried through the membrane by the analyte.\n",
            "\n",
            "One example of a liquid-based ion-selective electrode is that for Ca 2+ ,  which uses a porous plastic membrane saturated with di-( n -decyl) phosphate (Figure 11.13). As shown in Figure 11.14, the membrane is placed at the end of a nonconducting cylindrical tube and is in contact with two reservoirs. The outer reservoir contains di-( n -decyl) phosphate in din -octylphenylphosphonate, which soaks into the porous membrane. The inner reservoir contains a standard aqueous solution of Ca 2+ and a Ag/AgCl reference electrode. Calcium ion-selective electrodes are also available in which the di-( n -decyl) phosphate is immobilized in a polyvinyl chloride\n",
            "\n",
            "| Table 11. 3   | Representative Examples of Liquid-Based Ion-Selective Electrodes   | Representative Examples of Liquid-Based Ion-Selective Electrodes                                                                       |\n",
            "|---------------|--------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------|\n",
            "| Analyte       | Membrane Composition                                               | Selectivity Coefficients a                                                                                                             |\n",
            "| Ca 2+         | di-( n -decyl) phosphate in PVC                                    | K Ca 2+ /Zn 2+ = 1-5 K Ca 2+ /Al 3+ = 0.90 K Ca 2+ /Mn 2+ = 0.38 K Ca 2+ /Cu 2+ = 0.070 K Ca 2+ /Mg 2+ = 0.032                         |\n",
            "| K +           | Valinomycin in PVC                                                 | K K + /Rb + = 1.9 K K + /Cs + = 0.38 K K + /Li + = 1 × 10 -4 K K + /Na + = 1 × 10 -5                                                   |\n",
            "| Li +          | ETH 149 in PVC                                                     | K Li + /H + = 1 K Li + /Na + = 0.05 K Li + /K + = 7 × 10 -3                                                                            |\n",
            "| NH 4 +        | Nonactin and monactin in PVC                                       | K NH 4 + /K + = 0.12 K NH 4 + /H + = 0.016 K NH 4 + /Li + = 4.2 × 10 -3 K NH 4 + /Na + = 2 × 10 -3                                     |\n",
            "| ClO 4 -       | Fe( o -phen) 3 3+ in p -nitrocymene with porous membrane           | K ClO 4 - /OH - = 1 K ClO 4 - /I - = 0.012 K ClO 4 - /NO 3 - = 1.5 × 10 -3 K ClO 4 - /Br - = 5.6 × 10 -4 K ClO 4 - /Cl - = 2.2 × 10 -4 |\n",
            "| NO 3 -        | tetradodecyl ammonium nitrate in PVC                               | K NO 3 - /Cl - = 6 × 10 -3 K NO 3 - /F - = 9 × 10 -4                                                                                   |\n",
            "\n",
            "a Selectivity constants are approximate, and those found experimentally may vary substantially from the listed values. 3\n",
            "\n",
            "(PVC) membrane, eliminating the need for a reservoir containing di-( n -decyl) phosphate.\n",
            "\n",
            "A membrane potential develops as the result of a difference in the equilibrium position of the complexation reaction\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "on the two sides of the membrane, where ( m ) indicates that the species is present in the membrane. The cell potential for the Ca 2+ ion-selective electrode is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The selectivity of the electrode for Ca 2+ is  very  good, with only Zn 2+ showing greater selectivity.\n",
            "\n",
            "The properties of several representative liquid-based ion-selective electrodes are presented in Table 11.3. An electrode using a liquid reservoir can be stored in a dilute solution of analyte and needs no additional conditioning before use. The lifetime of an electrode with a PVC membrane, however, is proportional to its exposure to aqueous solutions. For this reason these electrodes are best stored by covering the membrane with a cap containing a small amount of wetted gauze to\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.15\n",
            "\n",
            "Schematic diagram of a gas-sensing membrane electrode.\n",
            "\n",
            "## enzyme electrodes\n",
            "\n",
            "An electrode that responds to the concentration of a substrate by reacting the substrate with an immobilized enzyme, producing an ion that can be monitored with an ion-selective electrode.\n",
            "\n",
            "maintain a humid environment. The electrode must then be conditioned before use by soaking in a solution of analyte for 30-60 min.\n",
            "\n",
            "Gas-Sensing Electrodes A number of membrane electrodes have been developed that respond to the concentration of dissolved gases. The basic design of these electrodes is shown in Figure 11.15 and consists of a thin membrane separating the sample from an inner solution containing an ion-selective electrode. The membrane is permeable to the gaseous analyte, but is not permeable to nonvolatile components in the sample matrix. Once the gaseous analyte passes through the membrane, it reacts in the inner solution, producing a species whose concentration can be monitored by an appropriate ion-selective electrode. For example, in the CO2 electrode, CO2 reacts in the inner solution to produce H3O + .\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The change in the concentration of H3O + is monitored with a pH ion-selective electrode, for which the cell potential is given by equation 11.9. The relationship between the concentration of H3O + and CO2 is given by rearranging the equilibrium constant expression for reaction 11.10; thus\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where K is the equilibrium constant. If the amount of HCO3 -in the internal solution is sufficiently large, then its concentration is unaffected by the presence of CO2 and remains constant. Substituting equation 11.11 into equation 11.9 gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where K ′ is a constant that includes the constant for the pH ion-selective electrode, the equilibrium constant for reaction 11.10, and the concentration of HCO3 -.\n",
            "\n",
            "Gas-sensing electrodes have been developed for a variety of gases, the characteristics  for  which  are  listed  in  Table  11.4.  The  composition of the inner solution changes with use, and both it and the membrane must be replaced periodically. Gas-sensing electrodes are stored in a solution similar to the internal solution to minimize their exposure to atmospheric gases.\n",
            "\n",
            "Potentiometric Biosensors Potentiometric electrodes for the analysis of molecules of biochemical importance can be constructed in a fashion similar to that used for gas-sensing electrodes. The most common class of potentiometric biosensors are the so-called enzyme electrodes, in which an enzyme is trapped or immobilized at the surface of an ion-selective electrode. Reaction of the analyte with the enzyme produces a product whose concentration is monitored by the ion-selective electrode. Potentiometric biosensors have also been designed around other biologically  active  species,  including  antibodies,  bacterial  particles,  tissue,  and  hormone receptors.\n",
            "\n",
            "One example of an enzyme electrode is the urea electrode, which is based on the catalytic hydrolysis of urea by urease\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "In one version of the urea electrode, shown in Figure 11.16, an NH3 electrode is modified by adding a dialysis membrane that physically traps a pH 7.0 buffered solution of urease between the dialysis membrane and the gas-permeable\n",
            "\n",
            "| Table 11. 4   | Characteristics of Gas-Sensing Membrane Electrodes   | Characteristics of Gas-Sensing Membrane Electrodes   | Characteristics of Gas-Sensing Membrane Electrodes   |\n",
            "|---------------|------------------------------------------------------|------------------------------------------------------|------------------------------------------------------|\n",
            "| Analyte       | Reaction in Inner Solution                           | Inner Solution                                       | Ion-Selective Electrode                              |\n",
            "| CO 2          | CO 2 + 2H 2 O t HCO 3 - + H 3 O +                    | 0.01 MNaHCO 3 , 0.01 MNaCl                           | glass pH electrode                                   |\n",
            "| HCN           | HCN + H 2 O t CN - + H 3 O +                         | 0.01 MKAg(CN) 2                                      | Ag 2 S membrane electrode                            |\n",
            "| HF            | HF + H 2 O t F - + H 3 O +                           | 1 MH 3 O +                                           | F - electrode                                        |\n",
            "| H 2 S         | H 2 S + H 2 O t HS - + H 3 O +                       | pH 5 citrate buffer                                  | Ag 2 S membrane electrode                            |\n",
            "| NH 3          | NH 3 + H 2 O t NH 4 + + OH -                         | 0.01 MNH 4 Cl, 0.1 MKNO 3                            | glass pH electrode                                   |\n",
            "| NO 2          | 2NO 2 + 3H 2 O t NO 3 - + NO 2 - + 2H 3 O +          | 0.02 MNaNO 2 , 0.1 MKNO 3                            | glass pH electrode                                   |\n",
            "| SO 2          | SO 2 + 2H 2 O t HSO 3 - + H 3 O +                    | 0.001 MNaHSO 3 , pH 5                                | glass pH electrode                                   |\n",
            "\n",
            "Source: Data compiled from Cammann, K. Working with Ion-Selective Electrodes. Springer-Verlag: Berlin, 1977.\n",
            "\n",
            "membrane. 4 When immersed in the sample, urea diffuses through the dialysis membrane, where it reacts with the enzyme urease. The NH4 + that is produced is in equilibrium with NH3\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "which, in turn, diffuses through the gas-permeable membrane, where it is detected by a pH electrode. The response of the electrode to the concentration of urea is given by\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Another version of the urea electrode (Figure 11.17) immobilizes the enzyme in a polymer membrane formed directly on the tip of a glass pH electrode. 5 In this case, the electrode's response is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Few potentiometric biosensors are commercially available. As shown in Figures 11.16 and 11.17, however, available ion-selective and gas-sensing electrodes may be easily converted into biosensors. Several representative examples are described in Table 11.5, and additional examples can be found in several reviews listed in the suggested readings at the end of the chapter.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## 11B. 5 Quantitative Applications\n",
            "\n",
            "## Figure 11.16\n",
            "\n",
            "The potentiometric determination of an analyte's concentration is one of the most common quantitative analytical techniques. Perhaps the most frequently employed, routine quantitative measurement is the potentiometric determination of a solution's pH, a technique considered in more detail in the following discussion. Other areas in which potentiometric applications are important include clinical chemistry, environmental chemistry, and potentiometric titrations. Before considering these applications, however, we must first examine more closely the relationship between cell potential and the analyte's concentration, as well as methods for standardizing potentiometric measurements.\n",
            "\n",
            "Activity Versus Concentration In describing metallic and membrane indicator electrodes, the Nernst equation relates the measured cell potential to the concentration of analyte. In writing the Nernst equation, we often ignore an important detail-the\n",
            "\n",
            "Schematic diagram of an enzyme-based potentiometric biosensor for urea in which urease is trapped between two membranes.\n",
            "\n",
            "Urease immobilized in polymer membrane\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.17\n",
            "\n",
            "Schematic diagrams of a second enzymebased potentiometric biosensor for urea in which urease is immobilized in a polymer matrix.\n",
            "\n",
            "potential of an electrochemical cell is a function of activity, not concentration. Thus, the Nernst equation for a metallic electrode of the first kind is more appropriately written as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where a M n+ is the activity of the metal ion. As described in Chapter 6, the activity of an ion is equal to the product of its concentration, [M n+ ], and a matrix-dependent activity coefficient, γ M n+ .\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Substituting equation 11.15 into equation 11.14 and rearranging gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Equation 11.16 can be solved for the metal ion's concentration if its activity coefficient is known. This presents a serious complication since the activity coefficient may be difficult to determine. If, however, the standards and samples have an identical matrix, then γ M n+ remains constant, and equation 11.16 simplifies to\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where K ′ includes the activity coefficient.\n",
            "\n",
            "Quantitative Analysis Using External Standards To determine the concentration of analyte in a sample, it is necessary to standardize the electrode. If the electrode's response obeys the Nernst equation,\n",
            "\n",
            "Table 11. 5\n",
            "\n",
            "## Representative Examples of Potentiometric Biosensors\n",
            "\n",
            "| Analyte                                                                                                                                             | Biologically Active Phase a                                                                                                                                                                                                                                         | Substance Determined                                                    |\n",
            "|-----------------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------|\n",
            "| 5 ′ -adenosinemonophosphate (5 ′ -AMP) L -arginine asparagine L -cysteine L -glutamate L -glutamine oxalate penicillin L -phenylalanine sugars urea | AMP-deaminase (E) arginase + urease (E) asparaginase (E) Proteus morganii (B) yellow squash (T) Sarcina flava (B) oxalate decarboxylase (E) penicillinase (E) L -amino acid oxidase and horseradish peroxidase (E) bacteria from human dental plaque (B) urease (E) | NH 3 NH 3 NH 4 + H 2 S CO 2 NH 3 CO 2 H 3 O + I - H 3 O + NH 3 or H 3 O |\n",
            "\n",
            "- Source: Compiled from Cammann, K. Working with Ion-Selective Electrodes. Springer-Verlag: Berlin, 1977; and Lunte, C. E.; Heineman, W. R.\n",
            "- 'Electrochemical Techniques in Bioanalysis.' In Steckham, E., ed. Topics in Current Chemistry, Vol. 143, Springer-Verlag: Berlin, 1988, p. 8. 3,6\n",
            "- a Abbreviations: E = enzyme; B = bacterial particle; T = tissue.\n",
            "\n",
            "then only the constant K need be determined, and standardizing with a single external standard is possible. Since small deviations from the ideal 'Nerstian' slope of ± RT nF / or ± RT zF / are frequently observed, standardization is usually accomplished using two or more external standards.\n",
            "\n",
            "In most quantitative analyses we are interested in determining the concentration, not the activity, of the analyte. As noted earlier, however, the electrode's response is a function of the analyte's activity. In the absence of interferents, a calibration curve of potential versus activity is a straight line. A plot of potential  versus  concentration, however, may be curved at higher concentrations of  analyte due to changes in the analyte's activity coefficient. A curved calibration curve may still be used to determine the analyte's concentration if the standard's matrix matches that of the sample. When the exact composition of the sample matrix is unknown, which often is the case, matrix matching becomes impossible.\n",
            "\n",
            "Another approach to matrix matching, which does not rely on knowing the exact composition of the sample's matrix, is to add a high concentration of inert electrolyte to all samples and standards. If the concentration of added electrolyte is sufficient, any difference between the sample's matrix and that of the standards becomes trivial, and the activity coefficient remains essentially constant. The solution of inert electrolyte added to the sample and standards is called a total ionic strength adjustment buffer (TISAB).\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11.6\n",
            "\n",
            "The concentration of Ca 2+ in a water sample was determined by the method of external standards. The ionic strength of the samples and standards was maintained at a nearly constant level by making each solution 0.5 M in KNO3. The measured cell potentials for the external standards are shown in the following table.\n",
            "\n",
            "| [Ca 2+ ] (M)   |   E cell (V) |\n",
            "|----------------|--------------|\n",
            "| 1.00 × 10 -5   |       -0.125 |\n",
            "| 5.00 × 10 -5   |       -0.103 |\n",
            "| 1.00 × 10 -4   |       -0.093 |\n",
            "| 5.00 × 10 -4   |       -0.072 |\n",
            "| 1.00 × 10 -3   |       -0.065 |\n",
            "| 5.00 × 10 -3   |       -0.043 |\n",
            "| 1.00 × 10 -2   |       -0.033 |\n",
            "\n",
            "What is the concentration of Ca 2+ in a water sample if its cell potential is found to be -0.084 V?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "Linear regression gives the equation for the calibration curve as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Substituting the cell potential for the sample gives the concentration of Ca 2+ as 2.17 × 10 -4 M. Note that the slope of the calibration curve is slightly different from the ideal value of 0.05916/2 = 0.02958.\n",
            "\n",
            "total ionic strength adjustment buffer A solution containing a relatively high concentration of inert electrolytes such that its composition fixes the ionic concentration of all solutions to which it is added.\n",
            "\n",
            "Quantitative Analysis Using the Method of Standard Additions Because of the difficulty of maintaining a constant matrix for samples and standards, many quantitative potentiometric methods use the method of standard additions. A sample of volume, V X,  and analyte concentration, C X,  is  transferred  to  a  sample  cell,  and  the potential, ( E cell )X, measured. A standard addition is made by adding a small volume, V S ,  of  a  standard containing a known concentration of analyte, C S ,  to  the sample, and the potential, ( E cell )S ,  measured. Provided that V S is  significantly smaller than V X, the change in sample matrix is ignored, and the analyte's activity coefficient remains constant. Example 11.7 shows how a one-point standard addition can be used to determine the concentration of an analyte.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11. 7\n",
            "\n",
            "The concentration of Ca 2+ in a sample of sea water is determined using a Ca ion-selective electrode and a one-point standard addition. A 10.00-mL sample is transferred to a 100-mL volumetric flask and diluted to volume. A 50.00-mL aliquot of sample is placed in a beaker with the Ca ion-selective electrode and a reference electrode, and the potential is measured as -0.05290 V. A 1.00-mL aliquot of a 5.00 × 10 -2 M standard solution of Ca 2+ is added, and a potential of -0.04417 V is measured. What is the concentration of Ca 2+ in the sample of sea water?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "To begin, we write Nernst equations for the two measured cell potentials. The cell potential for the sample is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and that following the standard addition is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where V T is the total volume ( V S + V X) after the standard addition. Subtracting the first equation from the second equation gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Replacing ( E cell )S - ( E cell )X with ∆ E and rearranging yields\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Substituting known values for ∆ E, V X, V S , V T, and C S,\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and taking the inverse log of both sides gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Finally, solving for C X gives the concentration of Ca 2+ as 9.88 × 10 -4 M. Since the original sample of sea water was diluted by a factor of 10, the concentration of Ca 2+ in the sea water sample is 9.88 × 10 -3 M.\n",
            "\n",
            "Free Ions Versus Complexed Ions In discussing the F ion-selective electrode, we noted that the membrane potential is influenced by the concentration of F - , but not the concentration of HF. An analysis for fluoride, therefore, is pH-dependent. Below a pH of approximately 4, fluoride is present predominantly as HF, and a quantitative analysis for total fluoride is impossible. If the pH is increased to greater than 4, however, the equilibrium\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "shifts to the right, and a quantitative analysis for total fluoride is possible.\n",
            "\n",
            "Most potentiometric electrodes are selective for only the free, uncomplexed analyte and do not respond to complexed forms of the analyte. Solution conditions, therefore, must be carefully controlled if the purpose of the analysis is to determine the analyte's total concentration. On the other hand, this selectivity provides a significant advantage over other quantitative methods of analysis when it is necessary to determine the concentration of free ions. For example, calcium is present in urine both as free Ca 2+ ions and as protein-bound Ca 2+ ions. If a urine sample is analyzed by atomic absorption spectroscopy, the signal is proportional to the total concentration of Ca 2+ , since both free and bound calcium are atomized. Analysis with a Ca 2+ ISE, however, gives a signal that is a function of only free Ca 2+ ions since the protein-bound ions cannot interact with the electrode's membrane.\n",
            "\n",
            "Representative Method Ion-selective electrodes find application in numerous quantitative analyses, each of which has its own unique considerations. The following procedure for the analysis of fluoride in toothpaste provides an instructive example.\n",
            "\n",
            "## Method 11.1\n",
            "\n",
            "## Determination of Fluoride in Toothpaste 7\n",
            "\n",
            "Description of the Method. The concentration of fluoride in toothpastes containing soluble F may be determined with a F - ion-selective electrode, using a calibration curve prepared with external standards. Although the F - ISE is very selective (only OH with K F /OH - of 0.1 is a significant interferent), Fe 3+ and Al 3+ interfere with the analysis by forming soluble fluoride complexes that do not interact with the ion-selective electrode's membrane. This interference is minimized by reacting any Fe 3+ and Al 3+ with a suitable complexing agent.\n",
            "\n",
            "Procedure. Prepare 1 L of a standard solution of 1.00% w/v SnF2, and transfer to a plastic bottle for storage. Using this solution, prepare 100 mL each of standards containing 0.32%, 0.36%, 0.40%, 0.44%, and 0.48% w/v SnF2, adding\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Continued from page 489\n",
            "\n",
            "400 mg of malic acid to each solution as a stabilizer. Transfer the standards to plastic bottles for storage. Prepare a total ionic strength adjustment buffer (TISAB) by mixing 500 mL of water, 57 mL of glacial acetic acid, 58 g of NaCl, and 4 g of the disodium salt of DCTA ( trans -1,2-cyclohexanetetraacetic acid) in a 1-L beaker, stirring until dissolved. Cool the beaker in a water bath, and add 5 M NaOH until the pH is between 5 and 5.5. Transfer the contents of the beaker to a 1-L volumetric flask, and dilute to volume. Standards are prepared by placing approximately 1 g of a fluoride-free toothpaste, 30 mL of distilled water, and 1.00 mL of the standard into a 50-mL plastic beaker and stirring vigorously for 2 min with a stir bar. The resulting suspension is quantitatively transferred to a 100-mL volumetric flask along with 50 mL of TISAB and diluted to volume with distilled water. The entire standard solution is then transferred to a 250-mL plastic beaker until its potential is measured. Samples of toothpaste are prepared for analysis by using approximately 1-g portions and treating in the same manner as the standards. The cell potential for the standards and samples are measured using a F ion-selective electrode and an appropriate reference electrode. The solution is stirred during the measurement, and 2-3 min is allowed for equilibrium to be reached. The concentration of F - in the toothpaste is reported as %w/w SnF2.\n",
            "\n",
            "## Questions\n",
            "\n",
            "- 1. The total ionic strength adjustment buffer serves several purposes in this procedure. Identify these purposes.\n",
            "\n",
            "The composition of the TISAB accomplishes three things: (1) The high concentration of NaCl (approximately 1 M) ensures that the ionic strength of the standards and samples are essentially identical. Since the activity coefficient for fluoride will be the same in all solutions, the Nernst equation can be written using the concentration of F in place of its activity; (2) Glacial acetic acid and NaOH are used to prepare an acetic acid/acetate buffer of pH 5-5.5. The pH of this buffer is high enough to ensure that the predominant form of fluoride is F instead of HF; and (3) DCTA is added as a complexing agent for any Fe 3+ or Al 3+ that might be present, preventing the formation of FeF6 3- or AlF6 3- .\n",
            "\n",
            "- 2. Why is a fluoride-free toothpaste added to the standard solutions?\n",
            "- Fluoride-free toothpaste is added as a precaution against any matrix effects that might influence the ion-selective electrode's response. This assumes, of course, that the matrices of the two toothpastes are otherwise similar.\n",
            "- 3. The procedure specifies that the standard and sample solutions should be stored in plastic containers. Why is it not a good idea to store the solutions in glass containers?\n",
            "\n",
            "The fluoride ion is capable of reacting with glass to form SiF4.\n",
            "\n",
            "- 4. The slope of the calibration curve is found to be -57.98 mV per tenfold change in the concentration of F - , compared with the expected slope of -59.16 mV per tenfold change in concentration. What effect does this have on the quantitative analysis for %w/w SnF2 in the toothpaste samples?\n",
            "\n",
            "No effect at all-this is the reason for preparing a calibration curve with multiple standards.\n",
            "\n",
            "Measurement of pH With the availability of inexpensive glass pH electrodes and pH meters, the determination of pH has become one of the most frequent quantitative analytical measurements. The potentiometric determination of pH, however, is not without complications, several of which are discussed in this section.\n",
            "\n",
            "One complication is the meaning of pH. 8,9 The conventional definition of pH as presented in most introductory texts is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The pH of a solution, however, is defined by the response of an electrode to the H + ion and, therefore, is a measure of its activity.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Calculating the pH of a solution using equation 11.17 only approximates the true pH. Thus, a solution of 0.1 M HCl has a calculated pH of 1.00 using equation 11.17, but an actual pH of 1.1 as defined by equation 11.18. 8 The difference between the two values occurs because the activity coefficient for H + is not unity in a matrix of 0.1 M HCl. Obviously the true pH of a solution is affected by the composition of its matrix. As an extreme example, the pH of 0.01 M HCl in 5 m LiCl is 0.8, a value that is more acidic than that of 0.1 M HCl! 8\n",
            "\n",
            "A second complication in measuring pH results from uncertainties in the relationship between potential and activity. For a glass membrane electrode, the cell potential, E X, for a solution of unknown pH is given as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where K includes the potential of the reference electrode, the asymmetry potential of the glass membrane and any liquid junction potentials in the electrochemical cell. All the contributions to K are subject to uncertainty and may change from day to day, as well as between electrodes. For this reason a pH electrode must be calibrated using a standard buffer of known pH. The cell potential for the standard, E S , is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where pHS is  the  pH  of  the  standard.  Subtracting equation 11.20 from equation 11.19 and solving for pH gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "which is the operational definition of pH adopted by the International Union of Pure and Applied Chemistry.*\n",
            "\n",
            "*Equations 11.19-11.21 are defined for a potentiometric electrochemical cell in which the pH electrode is the cathode. In this case an increase in pH decreases the cell potential. Many pH meters are designed with the pH electrode as the anode so that an increase in pH increases the cell potential. The operational definition of pH then becomes\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "This difference, however, does not affect the operation of a pH meter.\n",
            "\n",
            "| Table 11.6       | pH Values for Selected NIST Primary Standard Buffers a   | pH Values for Selected NIST Primary Standard Buffers a   | pH Values for Selected NIST Primary Standard Buffers a   | pH Values for Selected NIST Primary Standard Buffers a   | pH Values for Selected NIST Primary Standard Buffers a   | pH Values for Selected NIST Primary Standard Buffers a   | pH Values for Selected NIST Primary Standard Buffers a   |\n",
            "|------------------|----------------------------------------------------------|----------------------------------------------------------|----------------------------------------------------------|----------------------------------------------------------|----------------------------------------------------------|----------------------------------------------------------|----------------------------------------------------------|\n",
            "| Temperature (°C) | Saturated (25 °C) KHC 4 H 4 O 6 (tartrate)               | 0.05 m KH 2 C 6 H 5 O 7 (citrate)                        | 0.05 m KHC 8 H 4 O 4 (phthalate)                         | 0.025 m KH 2 PO 4 , 0.025 m Na 2 HPO 4                   | 0.008695 m KH 2 PO 4 , 0.03043 m Na 2 HPO 4              | 0.01 m Na 4 B 4 O 7 (borax)                              | 0.025 m NaHCO 3 , 0.025 m Na 2 CO 3                      |\n",
            "| 0                | -                                                        | 3.863                                                    | 4.003                                                    | 6.984                                                    | 7.534                                                    | 9.464                                                    | 10.317                                                   |\n",
            "| 5                | -                                                        | 3.840                                                    | 3.999                                                    | 6.951                                                    | 7.500                                                    | 9.395                                                    | 10.245                                                   |\n",
            "| 10               | -                                                        | 3.820                                                    | 3.998                                                    | 6.923                                                    | 7.472                                                    | 9.332                                                    | 10.179                                                   |\n",
            "| 15               | -                                                        | 3.802                                                    | 3.999                                                    | 6.900                                                    | 7.448                                                    | 9.276                                                    | 10.118                                                   |\n",
            "| 20               | -                                                        | 3.788                                                    | 4.002                                                    | 6.881                                                    | 7.429                                                    | 9.225                                                    | 10.062                                                   |\n",
            "| 25               | 3.557                                                    | 3.776                                                    | 4.008                                                    | 6.865                                                    | 7.413                                                    | 9.180                                                    | 10.012                                                   |\n",
            "| 30               | 3.552                                                    | 3.766                                                    | 4.015                                                    | 6.854                                                    | 7.400                                                    | 9.139                                                    | 9.966                                                    |\n",
            "| 35               | 3.549                                                    | 3.759                                                    | 4.024                                                    | 6.844                                                    | 7.389                                                    | 9.102                                                    | 9.925                                                    |\n",
            "| 40               | 3.547                                                    | 3.753                                                    | 4.035                                                    | 6.838                                                    | 7.380                                                    | 9.068                                                    | 9.889                                                    |\n",
            "| 45               | 3.547                                                    | 3.750                                                    | 4.047                                                    | 6.834                                                    | 7.373                                                    | 9.038                                                    | 9.856                                                    |\n",
            "| 50               | 3.549                                                    | 3.749                                                    | 4.060                                                    | 6.833                                                    | 7.367                                                    | 9.011                                                    | 9.828                                                    |\n",
            "\n",
            "Source: Values taken from Bates, R. G. Determination of pH: Theory and Practice, 2nd ed. Wiley: New York, 1973. 10\n",
            "\n",
            "a Concentrations are given in molality (moles solute per kilograms solvent).\n",
            "\n",
            "Calibrating the electrode presents a third complication since a standard with an accurately known activity for H + needs to be used. Unfortunately, it is not possible to calculate rigorously the activity of a single ion. For this reason pH electrodes are calibrated using a standard buffer whose composition is chosen such that the defined pH is as close as possible to that given by equation 11.18. Table 11.6 gives pH values for several primary standard buffer solutions accepted by the National Institute of Standards and Technology.\n",
            "\n",
            "A pH electrode is normally standardized using two buffers: one near a pH of 7 and one that is more acidic or basic depending on the sample's expected pH. The pH electrode is immersed in the first buffer, and the 'standardize' or 'calibrate' control is adjusted until the meter reads the correct pH. The electrode is placed in the second buffer, and the 'slope' or 'temperature' control is adjusted to thebuffer's pH. Some pH meters are equipped with a temperature compensation feature, allowing the pH meter to correct the measured pH for any change in temperature. In this case a thermistor is placed in the sample and connected to the pH meter. The 'temperature' control is set to the solution's temperature, and the pH meter is calibrated using the 'calibrate' and 'slope' controls. If a change in the sample's temperature is indicated by the thermistor, the pH meter adjusts the slope of the calibration based on an assumed Nerstian response of 2.303 RT/F.\n",
            "\n",
            "Clinical Applications Perhaps the area in which ion-selective electrodes receive the widest use is in clinical analysis, where their selectivity for the analyte in a complex matrix provides a significant advantage over many other analytical methods. The most common analytes are electrolytes, such as Na + , K + , Ca 2+ , H + , and Cl , and dissolved gases, such as CO2. For extracellular fluids, such as blood and urine, the analysis can be made in vitro with conventional electrodes, provided that sufficient sample is available. Some clinical analyzers place a series of ion-selective electrodes in a flow\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "cell,  allowing several analytes to be monitored simultaneously. Standards, samples, and rinse solutions are pumped through the flow cell and across the surface of the electrodes. For smaller volumes of sample the analysis can be conducted using disposable ion-selective systems, such as the Kodak Ektachem analyzer for K + shown in Figure 11.18. The analyzer consists of separate electrodes for the sample and reference solutions. Each electrode is constructed from several thin films, consisting of a Ag/AgCl reference electrode, a salt bridge and an ion-selective membrane, deposited on a support base. The two electrodes are connected by a paper salt bridge saturated with the sample and reference solutions. The overall dimensions of the analyzer are 2.8 cm × 2.4 cm with a thickness of 150 µ m and require only 10 µ L each of sample and reference solution. Similar analyzers are available for the determination of Na + , Cl , and CO2.\n",
            "\n",
            "The analysis of intercellular fluids requires an ion-selective electrode that can be inserted directly into the desired cell. Liquid-based membrane microelectrodes with tip diameters of less than 1 µ m are constructed by heating and drawing out a hard-glass capillary tube with an initial diameter of approximately 1-2 mm (Figure 11.19). The tip of the microelectrode is made hydrophobic by dipping in dichlorodimethyl silane. An inner solution appropriate for the desired analyte and a Ag/AgCl wire reference electrode are placed within the microelectrode. The tip of the microelectrode is then dipped into a solution containing the liquid complexing agent. The small volume of liquid complexing agent entering the microelectrode is retained within the tip by capillary action, eliminating the need for a solid membrane. Potentiometric microelectrodes have been developed for a number of clinically important analytes, including H + , K + , Na + , Ca 2+ , Cl - , and I -.\n",
            "\n",
            "Figure 11.19 Schematic diagram of a liquid-based ionselective microelectrode.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Environmental Applications Although ion-selective electrodes find use in environmental analysis, their application is not as widespread as in clinical analysis. Standard methods have been developed for the analysis of CN - , F - , NH3, and NO3 -in water and wastewater. Except for F - , however, other analytical methods are considered superior. By incorporating the ion-selective electrode into a flow cell, the continuous monitoring of wastewater streams and other flow systems is possible. Such applications are limited, however, by the electrode's response to the analyte's activity, rather than its concentration. Considerable interest has been shown in the development of biosensors for the field screening and monitoring of environmental samples for a number of priority pollutants. 11\n",
            "\n",
            "Potentiometric Titrations In Chapter 9 we noted that one method for determining the equivalence point of an acid-base titration is to follow the change in pH with a pH electrode. The potentiometric determination of equivalence points is feasible for acid-base, complexation, redox, and precipitation titrations, as well as for titrations in  aqueous and nonaqueous solvents. Acid-base, complexation, and precipitation potentiometric titrations are usually monitored with an ion-selective electrode that is selective for the analyte, although an electrode that is selective for the titrant or a reaction product also can be used. A redox electrode, such as a Pt wire, and a reference electrode are used for potentiometric redox titrations. More details about potentiometric titrations are found in Chapter 9.\n",
            "\n",
            "## 11B.6 Evaluation\n",
            "\n",
            "Scale of Operation The working range for most ion-selective electrodes is from a maximum concentration of 0.1-1 M to a minimum concentration of 10 -5 - 10 -10 M. This broad working range extends from major to ultratrace analytes, and is significantly greater than many other analytical methods. For conventional ion-selective electrodes, macro-sized samples with minimum volumes of 0.05-10 mL are necessary. Microelectrodes and specially designed analyzers, such as the Kodak Ektachem analyzer for K + shown in Figure 11.18, may be used with ultramicro-sized samples provided that the sample taken for analysis is sufficiently large to be representative of the original sample.\n",
            "\n",
            "Accuracy The accuracy of a potentiometric analysis is limited by the measurement error for the cell's potential. Several factors contribute to this measurement error, including the contribution to the potential from interfering ions, the finite current drawn through the cell while measuring the potential, differences in the analyte's  activity  coefficient  in  the  sample  and  standard  solutions,  and  liquid junction potentials. Errors in accuracy due to interfering ions often can be eliminated by including a separation step before the potentiometric analysis. Modern high-impedance potentiometers minimize errors due to the passage of current through the electrochemical cell. Errors due to activity coefficients and liquid junction potentials are minimized by matching the matrix of the standards to that of the sample. Even in the best circumstances, however, a difference in potential of approximately ± 1 mV is observed for samples and standards at equal concentration.\n",
            "\n",
            "The effect of an uncertainty in potential on the accuracy of a potentiometric method of analysis is evaluated using a propagation of uncertainty. For a membrane ion-selective electrode the general expression for potential is given as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where z is the charge of the analyte. From Table 4.9 in Chapter 4, the error in the cell potential, ∆ E cell is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Rearranging and multiplying through by 100 gives the percent relative error in concentration as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The relative measurement error in concentration, therefore, is determined by the magnitude of the error in measuring the cell's potential and by the charge of the analyte. Representative values are shown in Table 11.7 for ions with charges of ± 1 and ± 2, at a temperature of 25 °C. Accuracies of 1-5% for monovalent ions and 2-10% for divalent ions are typical. Although equation 11.22 was developed for membrane electrodes, it also applies to metallic electrodes of the first and second kind when z is replaced by n.\n",
            "\n",
            "Precision The precision of a potentiometric measurement is limited by variations in  temperature and the sensitivity of the potentiometer. Under most conditions, and with simple, general-purpose potentiometers, the potential can be measured with a repeatability of ± 0.1 mV. From Table 11.7 this result corresponds to an uncertainty of ± 0.4% for monovalent analytes, and ± 0.8% for divalent analytes. The reproducibility of potentiometric measurements is about a factor of 10 poorer.\n",
            "\n",
            "Sensitivity The sensitivity of a potentiometric analysis is determined by the term RT/nF or RT/zF in the Nernst equation. Sensitivity is best for smaller values of n or z.\n",
            "\n",
            "| Table 11. 7                 | Relationship Between Measurement Error in Potential and Relative Error in Concentration   | Relationship Between Measurement Error in Potential and Relative Error in Concentration   |\n",
            "|-----------------------------|-------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------|\n",
            "| Relative Error in Potential | Relative Error in Potential                                                               | (%) z                                                                                     |\n",
            "| 0.1                         | 0.1                                                                                       | ± 0.8                                                                                     |\n",
            "| 0.5                         | 0.5                                                                                       | ± 3.9                                                                                     |\n",
            "| 1.0                         | 1.0                                                                                       | ± 7.8                                                                                     |\n",
            "| 1.5                         | 1.5                                                                                       | ± 11.1                                                                                    |\n",
            "| 2.0                         | 2.0                                                                                       | ± 15.6                                                                                    |\n",
            "\n",
            "## coulometry\n",
            "\n",
            "An electrochemical method in which the current required to exhaustively oxidize or reduce the analyte is measured.\n",
            "\n",
            "## Faraday's law\n",
            "\n",
            "The current or charge passed in a redox reaction is proportional to the moles of the reaction's reactants and products.\n",
            "\n",
            "## current efficiency\n",
            "\n",
            "The percentage of current that actually leads to the analyte's oxidation or reduction.\n",
            "\n",
            "Selectivity As described earlier, most ion-selective electrodes respond to more than one analyte. For many ion-selective electrodes, however, the selectivity for the analyte is significantly greater than for most interfering ions. Published selectivity coefficients for ion-selective electrodes (representative values are found in Tables 11.1 through 11.3) provide a useful guide in helping the analyst determine whether a potentiometric analysis is feasible for a given sample.\n",
            "\n",
            "Time, Cost, and Equipment In comparison with competing methods, potentiometry provides a rapid, relatively low-cost means for analyzing samples. Commercial instruments for measuring pH or potential are available in a variety of price ranges and include portable models for use in the field.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## 11C Coulometric Methods of Analysis\n",
            "\n",
            "In potentiometry, the potential of an electrochemical cell under static conditions is used to determine an analyte's concentration. As seen in the preceding section, potentiometry is an important and frequently used quantitative method of analysis. Dynamic electrochemical methods, such as coulometry, voltammetry, and amperometry, in which current passes through the electrochemical cell, also are important analytical techniques. In this section we consider coulometric methods of analysis. Voltammetry and amperometry are covered in Section 11D.\n",
            "\n",
            "Coulometric methods of analysis are based on an exhaustive electrolysis of the analyte. By exhaustive we mean that the analyte is quantitatively oxidized or reduced at the working electrode or reacts quantitatively with a reagent generated at the working electrode. There are two forms of coulometry: controlled-potential coulometry, in which a constant potential is applied to the electrochemical cell, and controlled-current coulometry, in which a constant current is passed through the electrochemical cell.\n",
            "\n",
            "The total charge, Q, in coulombs, passed during an electrolysis is related to the absolute amount of analyte by Faraday's law\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where n is  the number of electrons transferred per mole of analyte, F is  Faraday's constant (96487 C mol -1 ), and N is the moles of analyte. A coulomb is also equivalent to an A s; thus, for a constant current, ⋅ i, the charge is given as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where t e is the electrolysis time. If current varies with time, as it does in controlledpotential coulometry, then the total charge is given by\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "11.25\n",
            "\n",
            "In coulometry, current and time are measured, and equation 11.24 or equation 11.25 is used to calculate Q. Equation 11.23 is then used to determine the moles of analyte. To obtain an accurate value for N, therefore, all the current must result in the analyte's oxidation or reduction. In other words, coulometry requires 100% current efficiency (or an accurately measured current efficiency established using a standard), a factor that must be considered in designing a coulometric method of analysis.\n",
            "\n",
            "## 11C.1 Controlled-Potential Coulometry\n",
            "\n",
            "The easiest method for ensuring 100% current efficiency is to maintain the working electrode at a constant potential that allows for the analyte's quantitative oxidation or reduction, without simultaneously oxidizing or reducing an interfering species. The current flowing through an electrochemical cell under a constant potential is proportional to the analyte's concentration. As electrolysis progresses the analyte's concentration decreases, as does the current. The resulting current-versus-time profile  for  controlled-potential coulometry, which also is known as potentiostatic coulometry, is shown in Figure 11.20. Integrating the area under the curve (equation 11.25), from t = 0 until t = t e , gives the total charge. In this section we consider the experimental parameters and instrumentation needed to develop a controlledpotential coulometric method of analysis.\n",
            "\n",
            "Selecting a Constant Potential In controlled-potential coulometry, the potential is selected so that the desired oxidation or reduction reaction goes to completion without interference from redox reactions involving other components of the sample matrix. To see how an appropriate potential for the working electrode is selected, let's develop a constant-potential coulometric method for Cu 2+ based on its reduction to copper metal at a Pt cathode working electrode.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "A ladder diagram for a solution of Cu 2+ (Figure 11.21) provides a useful means for evaluating the solution's redox properties. From the ladder diagram we can see that reaction 11.26 is favored when the working electrode's potential is more negative than +0.342 V versus the SHE (+0.093 V versus the SCE). To maintain a 100% current efficiency, however, the potential must be selected so that the reduction of H3O + to H2 does not contribute significantly to the total charge passed at the electrode.\n",
            "\n",
            "The potential needed for a quantitative reduction of Cu 2+ can be calculated using the Nernst equation\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "If we define a quantitative reduction as one in which 99.99% of the Cu 2+ is reduced to Cu, then the concentration of Cu 2+ at the end of the electrolysis must be\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where [Cu 2+ ]0 is the initial concentration of Cu 2+ in the sample. Substituting equation 11.28 into equation 11.27 gives the desired potential electrode as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "If the initial concentration of Cu 2+ is 1.00 × 10 -4 M, for example, then the cathode's potential must be more negative than +0.105 V versus the SHE (-0.139 V versus the SCE) to achieve a quantitative reduction of Cu 2+ to Cu. Note that at this potential H3O + is  not  reduced to H2, maintaining a 100% current efficiency. Many of the published procedures for the controlled-potential coulometric analysis of Cu 2+ call for potentials that are more negative than that shown for the reduction of H3O + in Figure 11.21. 12 Such potentials can be used, however, because the slow kinetics for reducing H3O + results in a significant overpotential that shifts the potential of the H3O + /H2 redox couple to more negative potentials.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.20\n",
            "\n",
            "Current-time curve for controlled-potential coulometry.\n",
            "\n",
            "Figure 11.21\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Ladder diagram for aqueous solutions of Cu 2+ .\n",
            "\n",
            "## overpotential\n",
            "\n",
            "The difference between the potential actually required to initiate an oxidation or reduction reaction, and the potential predicted by the Nernst equation.\n",
            "\n",
            "Figure 11.22\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Charge-time curve obtained by integrating the current-time curve in Figure 11.20.\n",
            "\n",
            "Minimizing Electrolysis Time The current-time curve for controlled-potential coulometry in Figure 11.20 shows that the current decreases continuously throughout electrolysis. An exhaustive electrolysis, therefore, may require a long time. Since time is an important consideration in choosing and designing analytical methods, the factors that determine the analysis time need to be considered.\n",
            "\n",
            "The change in current as a function of time in controlled-potential coulometry is approximated by an exponential decay; thus, the current at time t is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where i 0 is the initial current, and k is a constant that is directly proportional to the area of the working electrode and the rate of stirring and inversely proportional to the volume of the solution. For an exhaustive electrolysis in which 99.99% of the analyte is oxidized or reduced, the current at the end of the analysis, t e , may be approximated as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Substituting equation 11.30 into equation 11.29 and solving for t e gives the minimum time for an exhaustive electrolysis as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "From this equation we see that increasing k leads to a shorter analysis time. For this reason controlled-potential coulometry is carried out in small-volume electrochemical  cells,  using  electrodes  with  large  surface  areas  and  with  high  stirring  rates.  A quantitative electrolysis typically requires approximately 30-60 min, although shorter or longer times are possible.\n",
            "\n",
            "Instrumentation The potential in controlled-potential coulometry is set using a three-electrode potentiostat. Two types of working electrodes are commonly used: a Pt electrode manufactured from platinum-gauze and fashioned into a cylindrical tube, and an Hg pool electrode. The large overpotential for reducing H3O + at mercury makes it the electrode of choice for analytes requiring negative potentials. For example, potentials more negative than -1 V versus the SCE are feasible at an Hg electrode (but not at a Pt electrode), even in very acidic solutions. The ease with which mercury is oxidized, however, prevents its use at potentials  that  are  positive  with  respect  to  the  SHE.  Platinum  working  electrodes are used when positive potentials are required. The auxiliary electrode, which is often a Pt wire, is separated by a salt bridge from the solution containing the analyte. This is necessary to prevent electrolysis products generated at the auxiliary electrode from reacting with the analyte and interfering in the analysis. A saturated calomel or Ag/AgCl electrode serves as the reference electrode.\n",
            "\n",
            "The other essential feature of instrumentation for controlled-potential coulometry is a means of determining the total charge passed during electrolysis. One method is to monitor the current as a function of time and determine the area under the curve (see Figure 11.20). Modern instruments, however, use electronic integration to monitor charge as a function of time. The total charge at the end of the electrolysis then can be read directly from a digital readout or from a plot of charge versus time (Figure 11.22).\n",
            "\n",
            "## 11C.2 Controlled-Current Coulometry\n",
            "\n",
            "A second approach to coulometry is to use a constant current in place of a constant potential (Figure 11.23). Controlled-current coulometry, also known as amperostatic coulometry or coulometric titrimetry, has two advantages over controlled-potential coulometry. First, using a constant current makes for a more rapid analysis since the current does not decrease over time. Thus, a typical analysis time for controlledcurrent coulometry is less than 10 min, as opposed to approximately 30-60 min for controlled-potential coulometry. Second, with a constant current the total charge is simply the product of current and time (equation 11.24). A method for integrating the current-time curve, therefore, is not necessary.\n",
            "\n",
            "Using a constant current does present two important experimental problems that must be solved if accurate results are to be obtained. First, as electrolysis occurs the analyte's concentration and, therefore, the current due to its oxidation or reduction steadily decreases. To maintain a constant current the cell potential must change until another oxidation or reduction reaction can occur at the working electrode. Unless the system is carefully designed, these secondary reactions will produce a current efficiency of less than 100%. The second problem is the  need  for  a  method  of  determining  when  the  analyte  has  been  exhaustively electrolyzed. In controlled-potential coulometry this is signaled by a decrease in the current to a constant background or residual current (see Figure 11.20). In controlled-current coulometry, however, a constant current continues to flow even when the analyte has been completely oxidized or reduced. A suitable means of determining the end-point of the reaction, t e , is needed.\n",
            "\n",
            "Maintaining Current Efficiency To illustrate why changing the working electrode's potential can lead to less than 100% current efficiency, let's consider the coulometric analysis for Fe 2+ based on its oxidation to Fe 3+ at a Pt working electrode in 1 M H2SO4.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The ladder diagram for this system is shown in Figure 11.24a. Initially the potential of the working electrode remains nearly constant at a level near the standard-state potential for the Fe 3+ /Fe 2+ redox couple. As the concentration of Fe 2+ decreases, however, the potential of the working electrode shifts toward more positive values until another oxidation reaction can provide the necessary current. Thus, in this case the potential eventually increases to a level at which the oxidation of H2O occurs.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Since the current due to the oxidation of H3O + does not contribute to the oxidation of Fe 2+ , the current efficiency of the analysis is less than 100%. To maintain a 100% current efficiency the products of any competing oxidation reactions must react both rapidly and quantitatively with the remaining Fe 2+ . This may be accomplished, for example, by adding an excess of Ce 3+ to the analytical solution (Figure 11.24b). When the potential of the working electrode shifts to a more positive potential, the first species to be oxidized is Ce 3+ .\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The Ce 4+ produced at the working electrode rapidly mixes with the solution, where it reacts with any available Fe 2+ .\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.23\n",
            "\n",
            "Current-time curve for controlled-current coulometry.\n",
            "\n",
            "## Figure 11.24\n",
            "\n",
            "Ladder diagrams for the controlled-current coulometric analysis of Fe 2+ (a) without the addition of Ce 3+ , and (b) with the addition of Ce 3+ . The matrix is 1 M H2SO4 in both cases.\n",
            "\n",
            "## mediator\n",
            "\n",
            "A species that transfers electrons from the electrode to the analyte.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Combining these reactions gives the desired overall reaction of\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "In this manner, a current efficiency of 100% is maintained. Furthermore, since the concentration of Ce 3+ remains at its initial level, the potential of the working electrode remains constant as long as any Fe 2+ is present. This prevents other oxidation reactions, such as that for H2O, from interfering with the analysis. A species, such as Ce 3+ , which is used to maintain 100% current efficiency, is called a mediator.\n",
            "\n",
            "End Point Determination Adding a mediator solves the problem of maintaining 100% current efficiency, but does not solve the problem of determining when the analyte's electrolysis is complete. Using the same example, once all the Fe 2+ has been oxidized current continues to flow as a result of the oxidation of Ce 3+ and, eventually, the oxidation of H2O. What is needed is a means of indicating when the oxidation of Fe 2+ is complete. In this respect it is convenient to treat a controlledcurrent coulometric analysis as if electrolysis of the analyte occurs only as a result of its reaction with the mediator. A reaction between an analyte and a mediator, such as that shown in reaction 11.31, is identical to that encountered in a redox titration. Thus, the same end points that are used in redox titrimetry (see Chapter 9), such as visual indicators, and potentiometric and conductometric measurements, may be used to signal the end point of a controlled-current coulometric analysis. For example, ferroin may be used to provide a visual end point for the Ce 3+ -mediated coulometric analysis for Fe 2+ .\n",
            "\n",
            "Instrumentation Controlled-current coulometry normally is carried out using a galvanostat and an electrochemical cell consisting of a working electrode and a counterelectrode. The working electrode, which often is constructed from Pt, is also\n",
            "\n",
            "called the generator electrode since it is where the mediator reacts to generate the species reacting with the analyte. The counterelectrode is isolated from the analytical solution by a salt bridge or porous frit to prevent its electrolysis products from reacting with the analyte. Alternatively, oxidizing or reducing the mediator can be carried out externally, and the appropriate products flushed into the analytical solution. Figure 11.25 shows one simple method by which oxidizing and reducing agents can be generated externally. A solution containing the mediator flows under the influence of gravity into a small-volume electrochemical cell. The products generated at the anode and cathode pass through separate tubes, and the appropriate oxidizing or reducing reagent can be selectively delivered to the analytical solution. For example, external generation of Ce 4+ can be obtained using an aqueous solution of Ce 3+ and the products generated at the anode.\n",
            "\n",
            "The other necessary instrumental component for controlled-current coulometry is an accurate clock for measuring the electrolysis time, t e , and a switch for starting and stopping the electrolysis. Analog clocks can read time to the nearest ± 0.01 s, but the need to frequently stop and start the electrolysis near the end point leads to a net uncertainty of ± 0.1 s. Digital clocks provide a more accurate measurement of time, with errors of ± 1 ms being possible. The switch must control the flow of current and the clock, so that an accurate determination of the electrolysis time is possible.\n",
            "\n",
            "Coulometric Titrations Controlled-current coulometric methods commonly are called coulometric titrations because of their similarity to conventional titrations. We already have noted, in discussing the controlled-current coulometric determination of Fe 2+ , that the oxidation of Fe 2+ by Ce 4+ is identical to the reaction used in a redox titration. Other similarities between the two techniques also exist. Combining equations 11.23 and 11.24 and solving for the moles of analyte gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Compare this equation with the relationship between the moles of strong acid, N, titrated with a strong base of known concentration.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The titrant in a conventional titration is replaced in a coulometric titration by a constant-current source whose current is analogous to the titrant's molarity. The time needed for an exhaustive electrolysis takes the place of the volume of titrant, and the switch for starting and stopping the electrolysis serves the same function as a buret's stopcock.\n",
            "\n",
            "## 11C. 3 Quantitative Applications\n",
            "\n",
            "Coulometry may be used for the quantitative analysis of both inorganic and organic compounds. Examples of controlled-potential and controlled-current coulometric methods are discussed in the following sections.\n",
            "\n",
            "Controlled-Potential Coulometry The majority of controlled-potential coulometric analyses involve the determination of inorganic cations and anions, including trace metals and halides. Table 11.8 provides a summary of several of these methods.\n",
            "\n",
            "The ability to control selectivity by carefully selecting the working electrode's potential, makes controlled-potential coulometry particularly useful for the analysis of alloys. For example, the composition of an alloy containing Ag, Bi, Cd, and Sb\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.25\n",
            "\n",
            "Method for the external generation of oxidizing and reducing agents in coulometric titrations.\n",
            "\n",
            "## coulometric titrations\n",
            "\n",
            "A titration in which the equivalence point is the time required for a constant current to completely oxidize or reduce the analyte.\n",
            "\n",
            "Table 11.8 Representative Examples for the ControlledPotential Coulometric Analysis of Inorganic Ions\n",
            "\n",
            "| Analyte   | Electrolytic Reaction a   | Electrode   |\n",
            "|-----------|---------------------------|-------------|\n",
            "| Antimony  | Sb(III) + 3 e - t Sb      | Pt          |\n",
            "| Arsenic   | As(III) t As(V) + 2 e -   | Pt          |\n",
            "| Cadmium   | Cd(II) + 2 e - t Cd       | Pt or Hg    |\n",
            "| Cobalt    | Co(II) + 2 e - t Co       | Pt or Hg    |\n",
            "| Copper    | Cu(II) + 2 e - t Cu       | Pt or Hg    |\n",
            "| Halides   | Ag + X - t AgX + e -      | Ag          |\n",
            "| Iron      | Fe(II) t Fe(III) + e -    | Pt          |\n",
            "| Lead      | Pb(II) + 2 e - t Pb       | Pt or Hg    |\n",
            "| Nickel    | Ni(II) + 2 e - t Ni       | Pt or Hg    |\n",
            "| Plutonium | Pu(III) t Pu(IV) + e -    | Pt          |\n",
            "| Silver    | Ag(I) + e - t Ag          | Pt          |\n",
            "| Tin       | Sn(II) + 2 e - t Sn       | Pt          |\n",
            "| Uranium   | U(VI) + 2 e - t U(IV)     | Pt or Hg    |\n",
            "| Zinc      | Zn(II) + 2 e - t Zn       | Pt or Hg    |\n",
            "\n",
            "Source: Compiled from Rechnitz, G. A. Controlled-Potential Analysis. Macmillan: New York, 1963, p. 49. a Electrolytic reactions are written in terms of the change in oxidation state. The actual species in solution depend on the composition of the sample matrix.\n",
            "\n",
            "can be determined by dissolving the sample and placing it in a matrix of 0.2 M H2SO4. A platinum working electrode is immersed in the solution and held at a constant potential of +0.40 V versus the SCE. At this potential Ag(I) deposits on the Pt electrode as Ag, and the other metal ions remain in solution. When electrolysis is complete, the total charge is used to determine the amount of silver in the alloy. The potential of the platinum electrode is then shifted to -0.08 V versus the SCE, depositing Bi on the working electrode. When the coulometric analysis for bismuth is complete, antimony is determined by shifting the working electrode's potential to -0.33 V versus the SCE, depositing Sb. Finally, cadmium is determined following its electrodeposition on the Pt electrode at a potential of -0.80 V versus the SCE.\n",
            "\n",
            "Another area where controlled-potential coulometry has found application is in nuclear chemistry, in which elements such as uranium and polonium can be determined at trace levels. For example, microgram quantities of uranium in a medium of H2SO4 can be determined by reducing U(VI) to U(IV) at a mercury working electrode.\n",
            "\n",
            "Controlled-potential coulometry also can be applied to the quantitative analysis of organic compounds, although the number of applications is significantly less than that for inorganic analytes. One example is the six-electron reduction of a nitro group, -NO2, to a primary amine, -NH2, at a mercury electrode. Solutions of picric acid, for instance, can be analyzed by reducing to triaminophenol.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "| Table 11. 9   | Representative Examples of Coulometric Redox Titrations   | Representative Examples of Coulometric Redox Titrations   | Representative Examples of Coulometric Redox Titrations      |\n",
            "|---------------|-----------------------------------------------------------|-----------------------------------------------------------|--------------------------------------------------------------|\n",
            "| Mediator      | Electrochemically Generated Reagent                       | Generator-Electrode Reaction                              | Representative Application a                                 |\n",
            "| Ag +          | Ag 2+                                                     | Ag + t Ag 2+ + e -                                        | H 2 C 2 O 4 + 2Ag 2+ + 2H 2 O t 2CO 2 + 2Ag + + 2H 3 O +     |\n",
            "| Br -          | Br 2                                                      | 2Br - t Br 2 + 2 e -                                      | H 2 S + Br 2 + 2H 2 O t S + 2Br - + 2H 3 O +                 |\n",
            "| Ce 3+         | Ce 4+                                                     | Ce 3+ t Ce 4+ + e -                                       | Fe(CN) 6 4- + Ce 4+ t Fe(CN) 6 3- + Ce 3+                    |\n",
            "| Cl -          | Cl 2                                                      | 2Cl - t Cl 2 + 2 e -                                      | Tl(I) + Cl 2 t Tl(III) + 2Cl -                               |\n",
            "| Fe 3+         | Fe 2+                                                     | Fe 3+ + e - t Fe 2+                                       | Cr 2 O 7 2- + 6Fe 2+ + 14H 3 O + t 2Cr 3+ + 6Fe 3+ + 21H 2 O |\n",
            "| I -           | I 3 -                                                     | 3I - t I 3 - + 2 e -                                      | 2S 2 O 3 2- + I 3 - t S 4 O 6 2- + 3I                        |\n",
            "| Mn 2+         | Mn 3+                                                     | Mn 2+ t Mn 3+ + e -                                       | As(III) + 2Mn 3+ t As(V) + 2Mn 2+                            |\n",
            "\n",
            "a The analyte is the underlined species in each reaction.\n",
            "\n",
            "Another example is the successive reduction of trichloroacetate to dichloroacetate, and of dichloroacetate to monochloroacetate\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Mixtures of trichloroacetate and dichloroacetate are analyzed by selecting an initial potential at which only the more easily reduced trichloroacetate is reduced. When its electrolysis is complete, the potential is switched to a more negative potential at which dichloroacetate is reduced. The total charge for the first electrolysis is used to determine the amount of trichloroacetate, and the difference in total charge between the first and second electrolyses gives the amount of dichloroacetate.\n",
            "\n",
            "Controlled-Current Coulometry The use of a mediator makes controlled-current coulometry a more versatile analytical method than controlled-potential coulometry. For example, the direct oxidation or reduction of a protein at the working electrode in controlled-potential coulometry is difficult if the protein's active redox site lies  deep  within its structure. The controlled-current coulometric analysis of the protein is made possible, however, by coupling its oxidation or reduction to a mediator that is reduced or oxidized at the working electrode. Controlled-current coulometric methods have been developed for many of the same analytes that may be determined by conventional redox titrimetry. These methods, several of which are summarized in Table 11.9, also are called coulometric redox titrations.\n",
            "\n",
            "Coupling the mediator's oxidation or reduction to an acid-base, precipitation, or complexation reaction involving the analyte allows for the coulometric titration of analytes that are not easily oxidized or reduced. For example, when using H2O as a mediator, oxidation at the anode produces H3O +\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "while reduction at the cathode produces OH - .\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "If  the  oxidation or reduction of H2O is carried out externally using the generator cell shown in Figure 11.25, then H3O + or OH can be dispensed selectively into a solution containing a basic or acidic analyte. The resulting reaction is identical to that in an acid-base titration. Coulometric acid-base titrations have been used for\n",
            "\n",
            "## Table 11.10\n",
            "\n",
            "## Representative Examples of Coulometric Titrations Using Acid-Base, Complexation, and Precipitation Reactions\n",
            "\n",
            "| Type of Reaction   | Mediator    | Electrochemically Generated Reagent   | Generator-Electrode Reaction                      | Representative Reaction a                           |\n",
            "|--------------------|-------------|---------------------------------------|---------------------------------------------------|-----------------------------------------------------|\n",
            "| Acid-Base          | H 2 O       | H 3 O +                               | 6H 2 O t 4H 3 O + + O 2 + 4 e -                   | OH - + H 3 O + t 2H 2 O                             |\n",
            "|                    | H 2 O       | OH -                                  | 2H 2 O+2 e - t 2OH - + H 2                        | H 3 O + + OH - t 2H 2 O                             |\n",
            "| Complexation       | HgNH 3 Y 2- | EDTA (HY 3- )                         | HgNH 3 Y 2- + NH 4 + + 2 e - t HY 3- + Hg + 2NH 3 | Ca 2+ + HY 3- +H 2 O t CaY 2- + H 3 O +             |\n",
            "| Precipitation      | Ag          | Ag +                                  | Ag t Ag + + e -                                   | I - + Ag + t AgI                                    |\n",
            "|                    | Hg          | Hg 2 2+                               | 2Hg t Hg 2 2+ + 2 e -                             | 2Cl - + Hg 2 2+ t Hg 2 Cl 2                         |\n",
            "|                    | Fe(CN) 6 3- | Fe(CN) 6 4-                           | Fe(CN) 6 3- + e - t Fe(CN) 6 4-                   | 3Zn 2+ + 2K + + 2Fe(CN) 6 4- t K 2 Zn 3 [Fe(CN) 6 ] |\n",
            "\n",
            "- a The analyte is the underlined species in each reaction.\n",
            "\n",
            "the analysis of strong and weak acids and bases, in both aqueous and nonaqueous matrices. Examples of coulometric titrations involving acid-base, complexation, and precipitation reactions are summarized in Table 11.10.\n",
            "\n",
            "In comparison with conventional titrimetry, there are several advantages to the coulometric titrations listed in Tables 11.9 and 11.10. One advantage is that the electrochemical generation of a 'titrant' that reacts immediately with the analyte allows the use of reagents whose instability prevents their preparation and storage as a standard solution. Thus, highly reactive reagents such as Ag 2+ and Mn 3+ can be used in coulometric titrations. Because it is relatively easy to measure small quantities of charge, coulometric titrations can be used to determine small quantities of analyte that cannot be measured accurately by a conventional titration.\n",
            "\n",
            "Quantitative Calculations The absolute amount of analyte in a coulometric analysis is determined by applying Faraday's law (equation 11.23) with the total charge during the electrolysis given by equation 11.24 or equation 11.25. Example 11.8 shows the calculations for a typical coulometric analysis.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11.8\n",
            "\n",
            "The purity of a sample of Na2S2O3 was determined by a coulometric redox titration using I as  a  mediator, and I3 -as  the  'titrant.'  A  sample  weighing 0.1342 g is transferred to a 100-mL volumetric flask and diluted to volume with distilled water. A 10.00-mL portion is transferred to an electrochemical cell  along  with  25  mL  of  1  M  KI,  75  mL  of  a  pH  7.0  phosphate  buffer,  and several drops of a starch indicator solution. Electrolysis at a constant current of 36.45 mA required 221.8 s to reach the starch indicator end point. Determine the purity of the sample.\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "From Table 11.9 we see that the coulometric titration of S2O3 2with I3 -is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Oxidizing S2O3 2to S4O6 2requires one electron per S2O3 2-( n = 1). Combining equations 11.23 and 11.24, and making an appropriate substitution for moles of Na2S2O3 gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving for the grams of Na2S2O3 gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "This represents the amount of Na2S2O3 in a 10.00-mL portion of a 100-mL sample, thus 0.1325 g of Na2S2O3 is present in the original sample. The purity of the sample, therefore, is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Note that the calculation is worked as if S2O3 2is  oxidized directly at the working electrode instead of in solution.\n",
            "\n",
            "Representative Method Every controlled-potential or controlled-current coulometric method has its own unique considerations. Nevertheless, the following procedure for the determination of dichromate by a coulometric redox titration provides an instructive example.\n",
            "\n",
            "Representative Methods\n",
            "\n",
            "## Method 11.2 Titration 13\n",
            "\n",
            "## Determination of Dichromate by a Coulometric Redox\n",
            "\n",
            "Description of the Method. The concentration of Cr2O7 2- in a sample is determined by a coulometric redox titration using Fe 3+ as a mediator and electrogenerated Fe 2+ as the 'titrant.' The end point of the coulometric redox titration is determined potentiometrically.\n",
            "\n",
            "Procedure. The electrochemical cell consists of a Pt working electrode and a Pt counterelectrode that is maintained in a separate compartment connected to the analytical solution by a porous glass disk. The counter electrode's compartment is filled with 0.2 M Na2SO4, the level of which must always be maintained above that of the solution in the electrochemical cell. Platinum and tungsten electrodes connected to a potentiometer are used to follow the change in potential during the analysis. A solution of approximately 0.3 M NH4Fe(SO4)2 is prepared for use as the mediator. The coulometric redox titration is carried out by adding 5.00 mL of the sample solution, 2 mL of 9 M H2SO4 and 10-25 mL of the NH4Fe(SO4)2 mediator solution. Distilled water is added as needed until the electrodes are covered. Pure N2 is bubbled through the solution for 15 min to remove traces of O2. A flow of N2 is maintained during the electrolysis, but is turned off momentarily when measuring the potential. A magnetic stir bar is used to stir the solution. Adjust the current to 15-50 mA, and begin the titration. Periodically stop the titration, and measure the potential of the solution. In this manner a titration\n",
            "\n",
            "-Continued\n",
            "\n",
            "Continued from page 505\n",
            "\n",
            "curve of potential versus time is recorded. The time needed to reach the equivalence point is read from the titration curve.\n",
            "\n",
            "## Questions\n",
            "\n",
            "- 1. Does the platinum working electrode serve as the cathode or the anode in this analysis?\n",
            "\n",
            "Reduction of Fe 3+ to Fe 2+ occurs at the working electrode, making it the cathode in the electrochemical cell.\n",
            "\n",
            "- 2. Pure N2 is bubbled through the solution to remove any trace of dissolved O2. Why is this necessary?\n",
            "\n",
            "If O2 is present, some of the electrogenerated Fe 2+ may be oxidized back to Fe 3+ by the reaction\n",
            "\n",
            "4Fe 2+ ( aq ) + O2( g ) + 4H3O + ( aq ) t 4Fe 3+ ( aq ) + 6H2O( l )\n",
            "\n",
            "Any Fe 2+ lost in this fashion must be replaced by the additional reduction of Fe 3+ , reducing the current efficiency and increasing the time needed to reach the titration's end point. The net result is that the reported concentration of Cr2O7 2is too large.\n",
            "\n",
            "- 3. What is the effect on the analysis if the NH4Fe(SO4)2 used to prepare the mediator solution is contaminated with trace amounts of Fe 2+ ?\n",
            "\n",
            "The Fe 2+ introduced when the mediator solution is first added will react with the Cr2O7 2before the electrolysis begins. As a result, the amount of Fe 2+ that must be generated is less than expected, decreasing the time needed to reach the end point of the titration. The current efficiency, therefore, is greater than 100%, and the reported concentration of Cr2O7 2- is too small. Trace amount of Fe 2+ can be removed from the mediator solution by adding H2O2 and heating at 50-70 °C until the evolution of O2 ceases. Alternatively, a blank titration can be used to correct for any Fe 2+ initially present in the mediator.\n",
            "\n",
            "## 11C. 4 Characterization Applications\n",
            "\n",
            "Studies aimed at characterizing the mechanisms of electrode reactions often make use of coulometry for determining the number of electrons involved in the reaction. To make such measurements a known amount of a pure compound is subject to a controlled-potential electrolysis. The coulombs of charge needed to complete the electrolysis are used to determine the value of n using Faraday's law (equation 11.23).\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11. 9\n",
            "\n",
            "A 0.3619-g sample of tetrachloropicolinic acid, C6HNO2Cl4, is dissolved in distilled water, transferred to a 1000-mL volumetric flask, and diluted to volume. An exhaustive controlled-potential electrolysis of a 10.00-mL portion of this solution at a spongy silver cathode requires 5.374 C of charge. What is the value of n for this reduction reaction?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "The 10.00-mL portion of sample contains 3.619 mg, or 1.39 × 10 -5 mol of tetrachloropicolinic acid. Solving equation 11.23 for n and making appropriate substitutions gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Thus, reducing a molecule of tetrachloropicolinic acid requires four electrons. The overall reaction, which results in the selective formation of 3,6dichloropicolinic acid, is\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## 11C. 5 Evaluation\n",
            "\n",
            "Scale of Operation Coulometric methods of analysis can be used to analyze small absolute amounts of analyte. In controlled-current coulometry, for example, the moles of analyte consumed during an exhaustive electrolysis is given by equation 11.32. An electrolysis carried out with a constant current of 100 µ A for 100 s, therefore, consumes only 1 × 10 -7 mol of analyte if n = 1. For an analyte with a molecular weight of 100 g/mol, 1 × 10 -7 mol corresponds to only 10 µ g. The concentration of analyte in the electrochemical cell, however, must be sufficient to allow an accurate determination of the end point. When using visual end points, coulometric titrations require solution concentrations greater than 10 -4 M and, as with conventional titrations, are limited to major and minor analytes. A coulometric titration to a  preset  potentiometric end point is feasible even with solution concentrations of 10 -7 M, making possible the analysis of trace analytes. 14\n",
            "\n",
            "Accuracy The accuracy of a controlled-current coulometric method of analysis is determined by the current efficiency, the accuracy with which current and time can be measured, and the accuracy of the end point. With modern instrumentation the maximum measurement error for current is about ± 0.01%, and that for time is approximately ± 0.1%. The maximum end point error for a coulometric titration is at least as good as that for conventional titrations and is often better when using small quantities of reagents. Taken together, these measurement errors suggest that accuracies of 0.1-0.3% are feasible. The limiting factor in many analyses, therefore, is current efficiency. Fortunately current efficiencies of greater than 99.5% are obtained routinely and often exceed 99.9%.\n",
            "\n",
            "In controlled-potential coulometry, accuracy is determined by current efficiency and the determination of charge. Provided that no interferents are present that are easier to oxidize or reduce than the analyte, current efficiencies of greater than 99.9% are easily obtained. When interferents are present, however, they can often be eliminated by applying a potential such that the exhaustive electrolysis of the interferents is possible without the simultaneous electrolysis of the analyte. Once the interferents have been removed the potential can be switched to a level at\n",
            "\n",
            "## voltammetry\n",
            "\n",
            "An electrochemical method in which we measure current as a function of the applied potential.\n",
            "\n",
            "## voltammogram\n",
            "\n",
            "A plot of current as a function of applied potential.\n",
            "\n",
            "which electrolysis of the analyte is feasible. The limiting factor in the accuracy of many controlled-potential coulometric methods of analysis is the determination of charge. With modern electronic integrators, the total charge can be determined with an accuracy of better than 0.5%.\n",
            "\n",
            "So what is to be done when an acceptable current efficiency is not feasible? If the analyte's oxidation or reduction leads to its deposition on the working electrode, it may be possible to determine the analyte's mass. In this case the working electrode is weighed before beginning the electrolysis and reweighed when electrolysis of the analyte is complete. The difference in the electrode's weight gives the analyte's mass. This technique is known as electrogravimetry. More details about this technique are provided in the suggested reading listed at the end of this chapter.\n",
            "\n",
            "Precision Precision is determined by the uncertainties of measuring current, time, and the end point in controlled-current coulometry and of measuring charge in controlled-potential coulometry. Precisions of ± 0.1-0.3% are routinely obtained for coulometric titrations, and precisions of ± 0.5% are typical for controlled-potential coulometry.\n",
            "\n",
            "Sensitivity For a coulometric method of analysis, the calibration sensitivity is equivalent to nF in equation 11.25. In general, coulometric methods in which the analyte's oxidation or reduction involves a larger value of n show a greater sensitivity.\n",
            "\n",
            "Selectivity Selectivity in controlled-potential and controlled-current coulometry is improved by carefully adjusting solution conditions and by properly selecting the electrolysis potential. In controlled-potential coulometry the potential is fixed by the potentiostat, whereas in controlled-current coulometry the potential is determined by the redox reaction involving the mediator. In either case, the ability to control the potential at which electrolysis occurs affords some measure of selectivity. By adjusting pH or adding a complexing agent, it may be possible to shift the potential at which an analyte or interferent undergoes oxidation or reduction. For example, the standard-state reduction potential for Zn 2+ is  -0.762 V versus the SHE, but shifts to -1.04 for Zn(NH3)4 2+ .  This  provides  an  additional means for controlling selectivity when an analyte and interferent undergo electrolysis at similar potentials.\n",
            "\n",
            "Time, Cost, and Equipment Controlled-potential coulometry is a relatively timeconsuming analysis, with a typical analysis requiring 30-60 min. Coulometric titrations, on the other hand, require only a few minutes and are easily adapted for automated analysis. Commercial instrumentation for both controlled-potential and controlled-current coulometry is available and is relatively inexpensive. Low-cost potentiostats and constant-current sources are available for less than $1000.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Voltammetric Methods of Analysis\n",
            "\n",
            "In voltammetry a time-dependent potential is applied to an electrochemical cell, and the current flowing through the cell is measured as a function of that potential. A plot of current as a function of applied potential is called a voltammogram and is the electrochemical equivalent of a spectrum in spectroscopy, providing quantitative and qualitative information about the species involved in the oxidation or reduction reaction. 15 The earliest voltammetric technique to be introduced was polarography, which was developed by Jaroslav Heyrovsky\n",
            "\n",
            "(1890-1967) in the early 1920s, for which he was awarded the Nobel Prize in chemistry in 1959. Since then, many different forms of voltammetry have been developed, a few of which are noted in Figure 11.1. Before examining these techniques and their applications in more detail, however, we must first consider the basic experimental design for making voltammetric measurements and the factors influencing the shape of the resulting voltammogram.\n",
            "\n",
            "## 11D.1 Voltammetric Measurements\n",
            "\n",
            "Although early voltammetric methods relied on the use of only two electrodes, modern voltammetry makes use of a three-electrode potentiostat, such as that shown in Figure 11.4. A time-dependent potential excitation signal is applied to the working electrode, changing its potential relative to the fixed potential of the reference electrode. The resulting current between the working and auxiliary electrodes is measured. The auxiliary electrode is generally a platinum wire, and the SCE and Ag/AgCl electrode are common reference electrodes.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.26\n",
            "\n",
            "Several different materials have been used as working electrodes, including mercury, platinum, gold, silver, and carbon. The earliest voltammetric techniques, including polarography, used mercury for the working electrode. Since mercury is a liquid, the working electrode often consists of a drop suspended from the end of a capillary tube (Figure 11.26). In the hanging mercury drop electrode, or HMDE, a drop of the desired size is formed by the action of a micrometer screw that pushes the mercury through a narrow capillary tube. In the dropping mercury electrode, or DME, mercury drops form at the end of the capillary tube as a result of gravity. Unlike the HMDE, the mercury drop of a DME grows continuously and has a finite lifetime of several seconds. At the end of its  lifetime  the  mercury drop is dislodged, either manually or by gravity, and replaced by a new drop. The static mercury drop electrode, or SMDE, uses a solenoid-driven plunger to control the flow of mercury. The SMDE can be used as either a hanging mercury drop electrode or as a dropping mercury electrode. A single activation of the solenoid momentarily lifts the plunger, allowing enough mercury to flow through the capillary to form a single drop. To obtain a dropping mercury electrode the solenoid is activated repeatedly. A mercury film electrode consists of a thin layer of mercury deposited on the surface of a solid carbon, platinum, or gold electrode. The solid electrode is placed in a solution of Hg 2+ and held at a potential at which the reduction of Hg 2+ to Hg is favorable, forming a thin mercury film.\n",
            "\n",
            "Mercury has several advantages as a working electrode. Perhaps its most important advantage is its high overpotential for the reduction of H3O + to H2, which allows for the application of potentials as negative as -1 V versus the SCE in acidic solutions, and -2 V versus the SCE in basic solutions. A species such as Zn 2+ , which is difficult to reduce at other electrodes without simultaneously reducing H3O + , is easily reduced at a mercury working electrode. Other advantages include the ability of metals to dissolve in the mercury, resulting in the formation of an amalgam, and the ability to easily renew the surface of the electrode by extruding a new drop. One limitation to its use as a working electrode is the ease with which Hg is oxidized. For this reason, mercury electrodes cannot be used as at potentials more positive than -0.3 V to +0.4 V versus the SCE, depending on the composition of the solution.\n",
            "\n",
            "Mercury electrodes: (a) hanging mercury drop electrode; (b) dropping mercury electrode; (c) static mercury drop electrode.\n",
            "\n",
            "## hanging mercury drop electrode\n",
            "\n",
            "An electrode in which a drop of Hg is suspended from a capillary tube.\n",
            "\n",
            "## dropping mercury electrode\n",
            "\n",
            "An electrode in which successive drops of Hg form at the end of a capillary tube as a result of gravity, with each drop providing a fresh electrode surface.\n",
            "\n",
            "## static mercury drop electrode\n",
            "\n",
            "An electrode in which successive drops of Hg form at the end of a capillary tube as the result of a mechanical plunger, with each drop providing a fresh electrode surface.\n",
            "\n",
            "## amalgam\n",
            "\n",
            "A metallic solution of mercury with another metal.\n",
            "\n",
            "Figure 11.27 Schematic diagram of a solid electrode.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## faradaic current\n",
            "\n",
            "Any current in an electrochemical cell due to an oxidation or reduction reaction.\n",
            "\n",
            "## cathodic current\n",
            "\n",
            "A faradaic current due to a reduction reaction.\n",
            "\n",
            "## anodic current\n",
            "\n",
            "A faradaic current due to an oxidation reaction.\n",
            "\n",
            "Solid electrodes constructed using platinum, gold, silver, or carbon may be used over a range of potentials, including potentials that are negative and positive with respect to the SCE. For example, the potential range for a Pt electrode extends from approximately +1.2 V to -0.2 V versus the SCE in acidic solutions and from +0.7 V to -1 V versus the SCE in basic solutions. Solid electrodes, therefore, can be used in place of mercury for many voltammetric analyses requiring negative potentials and for voltammetric analyses at positive potentials at which mercury electrodes cannot be used. Except for the carbon paste electrode, solid electrodes are fashioned into disks that are sealed into the end of an inert support and are in contact with an electrical lead (Figure 11.27). The carbon paste electrode is made by filling the cavity at the end of the inert support with a paste consisting of carbon particles and a viscous oil. Solid electrodes are not without problems, the most important of which is the ease with which the electrode's surface may be altered by the adsorption of solution species or the formation of oxide layers. For this reason solid electrodes need frequent reconditioning, either by applying an appropriate potential or by polishing.\n",
            "\n",
            "A typical arrangement for a voltammetric electrochemical cell is shown in Figure 11.28. Besides the working, reference, and auxiliary electrodes, the cell also includes a N2 purge line for removing dissolved O2 and an optional stir bar. Electrochemical cells are available in a variety of sizes, allowing for the analysis of solution volumes ranging from more than 100 mL to as small as 50 µ L.\n",
            "\n",
            "## 11D.2 Current in Voltammetry\n",
            "\n",
            "When an analyte is oxidized at the working electrode, a current passes electrons through the external electric circuitry to the auxiliary electrode, where reduction of the solvent or other components of the solution matrix occurs. Reducing an analyte at the working electrode requires a source of electrons, generating a current that flows from the auxiliary electrode to the cathode. In either case, a current resulting from redox reactions at the working and auxiliary electrodes is  called  a faradaic current. In  this  section  we  consider  the  factors  affecting the magnitude of this faradaic current, as well as the source of any nonfaradaic currents.\n",
            "\n",
            "Sign Conventions Since the reaction of interest occurs at the working electrode, the classification of current is based on this reaction. A current due to the analyte's reduction is called a cathodic current and, by convention, is considered positive. Anodic currents are due to oxidation reactions and carry a negative value.\n",
            "\n",
            "Influence of Applied Potential on the Faradaic Current As an example, let's consider the faradaic current when a solution of Fe(CN)6 3is reduced to Fe(CN)6 4at the working electrode. The relationship between the concentrations of Fe(CN)6 3-, Fe(CN)6 4- , and the potential of the working electrode is given by the Nernst equation; thus\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where +0.356 is the standard-state potential for the Fe(CN)6 3/Fe(CN)6 4redox couple, and x = 0 indicates that the concentrations of Fe(CN)6 3and Fe(CN)6 4are\n",
            "\n",
            "those at the surface of the electrode. Surface concentrations are used instead of bulk concentrations since the equilibrium position for the redox reaction\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "can only be established electrochemically at the electrode's surface.\n",
            "\n",
            "Let's assume that we have a solution for which the concentration of Fe(CN)6 3is 1.0 mM and in which Fe(CN)6 4is absent. A ladder diagram for this redox example is  shown in Figure 11.29. If a potential of +0.530 V is applied to the working electrode, the concentrations of Fe(CN)6 3and Fe(CN)6 4at the surface of the electrode are unaffected, and no faradaic current is observed. Switching the potential to +0.356 V, however, requires that\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "which is only possible if a portion of the Fe(CN)6 3at  the  electrode surface is reduced to Fe(CN)6 4- . If this was all that occurred after the potential was applied, the result would be a brief surge of faradaic current that would quickly return to zero. However, although the concentration of Fe(CN)6 4at the electrode surface is 0.50 mM, its concentration in the bulk of solution is zero. As a result, a concentration gradient exists between the solution at the electrode surface and the bulk solution. This concentration gradient creates a driving force that transports Fe(CN)6 4away from the electrode surface (Figure 11.30). The subsequent decrease in the concentration of Fe(CN)6 4at  the  electrode surface requires the further reduction of Fe(CN)6 3- , as well as its transport from bulk solution to the electrode surface. Thus, a faradaic current continues to flow until there is no difference between the concentrations of Fe(CN)6 3and Fe(CN)6 4at  the  electrode surface and their concentrations in the bulk of solution. *\n",
            "\n",
            "Although the applied potential at the working electrode determines if a faradaic current flows, the magnitude of the current is determined by the rate of the resulting oxidation or reduction reaction at the electrode surface. Two factors contribute to the rate of the electrochemical reaction: the rate at which the reactants and products are transported to and from the surface of the electrode, and the rate at which electrons pass between the electrode and the reactants and products in solution.\n",
            "\n",
            "Influence of Mass Transport on the Faradaic Current There are three modes of mass transport that influence the rate at which reactants and products are transported to and from the electrode surface: diffusion, migration, and convection. Diffusion from a region of high concentration to a region of low concentration occurs whenever the concentration of an ion or molecule at the surface of the electrode is different from that in bulk solution. When the potential applied to the working electrode is sufficient to reduce or oxidize the analyte at the electrode surface, a concentration gradient similar to that shown in Figure 11.31 is established. The volume of solution in which the concentration gradient exists is called the diffusion layer. Without other modes of mass transport, the width of the diffusion layer, δ ,  increases with time as the concentration of reactants near the electrode surface decreases. The contribution of diffusion to the rate of mass transport, therefore, is time-dependent.\n",
            "\n",
            "*In voltammetry the working electrode's surface area is significantly smaller than that used in coulometry. Consequently, very little analyte undergoes electrolysis, and the analyte's concentration in bulk solution remains essentially unchanged.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.28\n",
            "\n",
            "Typical electrochemical cell for use in voltammetry.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.29\n",
            "\n",
            "Ladder diagram for the Fe(CN)6 3- /Fe(CN)6 4redox half-reaction.\n",
            "\n",
            "## mass transport\n",
            "\n",
            "The movement of material toward or away from the electrode surface.\n",
            "\n",
            "## diffusion\n",
            "\n",
            "The movement of material in response to a concentration gradient.\n",
            "\n",
            "## diffusion layer\n",
            "\n",
            "The layer of solution adjacent to the electrode in which diffusion is the only means of mass transport.\n",
            "\n",
            "## convection\n",
            "\n",
            "The movement of material in response to a mechanical force, such as stirring a solution.\n",
            "\n",
            "## migration\n",
            "\n",
            "The movement of a cation or anion in response to an applied potential.\n",
            "\n",
            "Electrode\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.30\n",
            "\n",
            "Schematic showing transport of Fe(CN)6 3toward the electrode and Fe(CN)6 4away from the electrode following the reduction of Fe(CN)6 3- .\n",
            "\n",
            "Distance from electrode\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.31\n",
            "\n",
            "Concentration gradients for the analyte in the absence of convection, showing the time-dependent change in diffusion as a method of mass transport.\n",
            "\n",
            "## nonfaradaic current\n",
            "\n",
            "A current in an electrochemical cell that is not the result of a redox reaction.\n",
            "\n",
            "Convection occurs when a mechanical means is used to carry reactants toward the electrode and to remove products from the electrode. The most common means of convection is to stir the solution using a stir bar. Other methods include rotating the electrode and incorporating the electrode into a flow cell.\n",
            "\n",
            "The final mode of mass transport is migration, which occurs when charged particles in solution are attracted or repelled from an electrode that has a positive or negative surface charge. Thus, when the electrode is positively charged, negatively charged particles move toward the electrode, while positively charged particles move toward the bulk solution. Unlike diffusion and convection, migration only affects the mass transport of charged particles.\n",
            "\n",
            "The flux of material to and from the electrode surface is a complex function of all three modes of mass transport. In the limit in which diffusion is the only significant means for the mass transport of the reactants and products, the current in a voltammetric cell is given by\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where n is the number of electrons transferred in the redox reaction, F is Faraday's constant, A is the area of the electrode, D is the diffusion coefficient for the reactant or product, C bulk and Cx= 0 are the concentration of the analyte in bulk solution and at the electrode surface, and δ is the thickness of the diffusion layer.\n",
            "\n",
            "For equation 11.33 to be valid, convection and migration must not interfere with the formation of a diffusion layer between the electrode and the bulk of solution. Migration is eliminated by adding a high concentration of an inert supporting electrolyte to the analytical solution. Ions of similar charge are equally attracted or repelled from the surface of the electrode and, therefore, have an equal probability of undergoing migration. The large excess of inert ions, however, ensures that few reactant and product ions will move as a result of migration. Although convection may be easily eliminated by not physically agitating the solution, in some situations it is desirable either to stir the solution or to push the solution through an electrochemical flow cell. Fortunately, the dynamics of a fluid moving past an electrode results in a small diffusion layer, typically of 0.001 - 0.01-cm thickness, in which the rate of mass transport by convection drops to zero (Figure 11.32).\n",
            "\n",
            "Influence of the Kinetics of Electron Transfer on the Faradaic Current The rate of mass transport is one factor influencing the current in a voltammetric experiment. The ease with which electrons are transferred between the electrode and the reactants and products in solution also affects the current. When electron transfer kinetics are fast, the redox reaction is at equilibrium, and the concentrations of reactants and products at the electrode are those specified by the Nernst equation. Such systems are considered electrochemically reversible. In other systems, when electron transfer kinetics are sufficiently slow, the concentration of reactants and products at the electrode surface, and thus the current, differ from that predicted by the Nernst equation. In this case the system is electrochemically irreversible.\n",
            "\n",
            "Nonfaradaic Currents Faradaic currents result from a redox reaction at the electrode surface. Other currents may also exist in an electrochemical cell that are unrelated to any redox reaction. These currents are called nonfaradaic currents and must be accounted for if the faradaic component of the measured current is to be determined.\n",
            "\n",
            "The most important example of a nonfaradaic current occurs whenever the electrode's potential is changed. In discussing migration as a means of mass transport, we noted that negatively charged particles in solution migrate toward a positively charged electrode, and positively charged particles move away from the same electrode. When an inert electrolyte is responsible for migration, the result is a structured electrode-surface interface called the electrical double layer, or EDL, the exact structure of which is of no concern in the context of this text. The movement of charged particles in solution, however, gives rise to a short-lived, nonfaradaic charging current. Changing the potential of an electrode causes a change in the structure of the EDL, producing a small charging current.\n",
            "\n",
            "Residual Current Even in the absence of analyte, a small current inevitably flows through an electrochemical cell. This current, which is called the residual current, consists of two components: a faradaic current due to the oxidation or reduction of trace impurities, and the charging current. Methods for discriminating between the faradaic current due to the analyte and the residual current are discussed later in this chapter.\n",
            "\n",
            "## 11D. 3 Shape of Voltammograms\n",
            "\n",
            "The shape of a voltammogram is determined by several experimental factors, the most important of which are how the current is measured and whether convection is included as a means of mass transport. Despite an abundance of different voltammetric techniques, several of which are discussed in this chapter, only three shapes are common for voltammograms (Figure 11.33).\n",
            "\n",
            "The voltammogram in Figure 11.33a is characterized by a current that increases from the background residual current to a limiting current at potentials at which the analyte is oxidized or reduced. Since the magnitude of a faradaic current is inversely proportional to δ (equation 11.33), a limiting current implies that the thickness of the diffusion layer remains constant. The simplest method for obtaining a limiting current is to stir the solution (Figure 11.32), which can be accomplished with a magnetic stir bar, or by rotating the electrode. Voltammetric techniques that include convection by stirring are called hydrodynamic voltammetry. When convection is absent, the thickness of the diffusion layer increases with time (Figure 11.31), resulting in a peak current in place of a limiting current (Figure 11.33b).\n",
            "\n",
            "In the voltammograms in Figures 11.33a and 11.33b, the current is monitored as a function of the applied potential. Alternatively, the change in current following\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.33\n",
            "\n",
            "Three common shapes for voltammograms.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.32\n",
            "\n",
            "Concentration gradient for the analyte showing the effects of diffusion and convection as methods of mass transport.\n",
            "\n",
            "## electrical double layer\n",
            "\n",
            "The interface between a positively or negatively charged electrode and the negatively or positively charged layer of solution in contact with the electrode.\n",
            "\n",
            "## charging current\n",
            "\n",
            "A current in an electrochemical cell due to the electrical double layer's formation.\n",
            "\n",
            "## residual current\n",
            "\n",
            "The current in an electrochemical cell that is present even in the absence of the analyte.\n",
            "\n",
            "## hydrodynamic voltammetry\n",
            "\n",
            "A form of voltammetry in which the solution is mechanically stirred.\n",
            "\n",
            "a  change in potential may be measured. The resulting voltammogram, which is shown in Figure 11.33c, also is characterized by a peak current.\n",
            "\n",
            "## 11D. 4 Quantitative and Qualitative Aspects of Voltammetry\n",
            "\n",
            "Earlier we described a voltammogram as the electrochemical equivalent of a spectrum in spectroscopy. In this section we consider how quantitative and qualitative information may be extracted from a voltammogram. Quantitative information is obtained by relating current to the concentration of analyte in the bulk solution. Qualitative information is obtained from the voltammogram by extracting the standard-state potential for the redox reaction. For simplicity we only consider voltammograms similar to that shown in Figure 11.33a.\n",
            "\n",
            "Determining Concentration Let's assume that the redox reaction at the working electrode is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and that initially only O is present in the bulk solution. The current is determined by the rate at which O diffuses through the fixed diffusion layer (see Figure 11.32), and is given by equation 11.33, or\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where K O is a constant equal to nFAD O/ δ . When the limiting current is reached, the concentration of O at the electrode surface is zero, and this equation simplifies to\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Thus, the limiting current, i lim , is a linear function of the concentration of O in bulk solution, and a quantitative analysis is possible using any of the standardization methods discussed in Chapter 5. Equations similar to equation 11.35 can be developed for other forms of voltammetry, in which peak currents are related to the analyte's concentration in bulk solution.\n",
            "\n",
            "Determining the Standard-State Potential To extract the standard-state potential, or formal potential, for reaction 11.34 from a voltammogram, it is necessary to rewrite the Nernst equation\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "in terms of current instead of the concentration of O and R. Substituting equation 11.36 into equation 11.35 and rearranging gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "To derive a similar equation for the concentration of R at the electrode surface we note that\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Since the concentration of R in bulk solution is zero, this equation simplifies to\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Substituting equations 11.39 and 11.38 into equation 11.37 and rearranging gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "When the current is half that of the limiting current\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "equation 11.40 simplifies to\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where E 1/2 is  the  half-wave potential (Figure 11.34). If K O is  approximately equal to K R, which is often the case, then the half-wave potential is equal to the standard-state potential. Note that equation 11.41 is only valid if the redox reaction is electrochemically reversible. Voltammetric techniques giving peak potentials also can be used to determine a redox reaction's standardstate potential.\n",
            "\n",
            "Current\n",
            "\n",
            "## 11D. 5 Voltammetric Techniques\n",
            "\n",
            "A number of voltammetric experiments are routinely used in quantitative and qualitative analyses. Several of these methods are briefly described in this section.\n",
            "\n",
            "Polarography The earliest voltammetric experiment was normal polarography at a  dropping mercury electrode. In normal polarography the potential is linearly scanned, producing voltammograms such as that shown in Figure 11.35. Although polarography takes place in an unstirred solution, a limiting current is obtained because the falling Hg drops mix the solution. Each new Hg drop, therefore, grows in  a  solution  whose composition is identical to that of the initial bulk solution. Oscillations in the current are due to the growth of the Hg drop, which leads to a\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "(a)\n",
            "\n",
            "Potential\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.34\n",
            "\n",
            "Determination of limiting current and halfwave potential in linear scan hydrodynamic voltammetry.\n",
            "\n",
            "## polarography\n",
            "\n",
            "A form of voltammetry using a dropping mercury electrode or a static mercury drop electrode.\n",
            "\n",
            "## Figure 11.35\n",
            "\n",
            "Potential-excitation signal and voltammogram for normal polarography.\n",
            "\n",
            "## stripping voltammetry\n",
            "\n",
            "A form of voltammetry in which the analyte is first deposited on the electrode and then removed, or 'stripped,' electrochemically while monitoring the current as a function of the applied potential.\n",
            "\n",
            "time-dependent change in the area of the working electrode. The limiting current, which is also called the diffusion current, may be measured from the maximum current, i max, or from the average current, i avg .  The relationship between the concentration of analyte, C A, and the limiting current is given by the Ilikovic equation\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where n is  the  number of electrons transferred in the redox reaction, D is  the analyte's diffusion coefficient, m is the flow rate of the Hg, and t is the drop time. The half-wave potential, E 1/2 ,  provides qualitative information about the redox reaction (see Appendix 3E for a list of selected polarographic half-wave potentials).\n",
            "\n",
            "Normal polarography has been replaced by various forms of pulse polarography, 16 several examples of which are shown in Figure 11.36. Differential pulse polarography (Figure 11.36b), for example, uses a series of potential pulses characterized by a cycle of time τ , a pulse time of t p, a potential pulse of ∆ E p, and a potential step per cycle of ∆ E s .  Typical  experimental conditions for differential pulse polarography are τ ≈ 1 s, t p ≈ 50 ms, ∆ E p ≈ 50 mV, ∆ E s ≈ 2 mV. The current is measured twice, for approximately 17 ms before the forward pulse and for approximately 17 ms before the reverse pulse. The difference in the two currents gives rise to a peak-shaped voltammogram. Other forms of pulse polarography include normal pulse polarography (Figure 11.36a), staircase polarography (Figure 11.36c), and square-wave polarography (Figure 11.36d). Limiting and peak currents are directly proportional to the concentration of analyte, and half-wave and peak potentials can be used for qualitative purposes. The popularity of pulse polarography is due to a substantial improvement in sensitivity and detection limits from those in normal polarography.\n",
            "\n",
            "Polarography is used extensively for the analysis of metal ions and inorganic anions, such as IO3 -and NO3 - . Organic compounds containing easily reducible or oxidizable functional groups also can be studied polarographically. Functional groups that have been used include carbonyls, carboxylic acids, and carbon-carbon double bonds.\n",
            "\n",
            "Hydrodynamic Voltammetry In polarography a limiting current is obtained because each falling drop of mercury returns the solution near the electrode to its initial composition. As noted earlier, a limiting current is also obtained whenever the solution is stirred during the analysis. The simplest means of stirring the solution is with a magnetic stir bar. More commonly, however, stirring is achieved by rotating the electrode.\n",
            "\n",
            "In hydrodynamic voltammetry current is measured as a function of the potential applied to a solid working electrode. The same potential profiles used for polarography, such as a linear scan or a differential pulse, are used in hydrodynamic voltammetry. The resulting voltammograms are identical to those for polarography, except for the lack of current oscillations resulting from the growth of the mercury drops. Because hydrodynamic voltammetry is not limited to Hg electrodes, it is useful for the analysis of analytes that are reduced or oxidized at more positive potentials.\n",
            "\n",
            "Stripping Voltammetry One of the most important quantitative voltammetric techniques is stripping voltammetry, which is composed of three related techniques: anodic, cathodic, and adsorptive stripping voltammetry. Since anodic strip-\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "ping voltammetry has found the widest application, we consider it in the greatest detail.\n",
            "\n",
            "Anodic stripping voltammetry consists of two steps (Figure 11.37). The first is a controlled potential electrolysis in which the working electrode, usually a hanging mercury drop or mercury film, is held at a cathodic potential sufficient to deposit the metal ion on the electrode. For example, with Cu 2+ the deposition reaction is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "## Figure 11.36\n",
            "\n",
            "Potential-excitation signals and voltammograms for (a) normal pulse polarography, (b) differential pulse polarography, (c) staircase polarography, and (d) square-wave polarography. See text for an explanation of the symbols. Current is sampled at the time intervals indicated by the solid circles ( · ).\n",
            "\n",
            "## Figure 11.37\n",
            "\n",
            "Potential-excitation signal and voltammogram for anodic stripping voltammetry at a hanging mercury drop electrode.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "where Cu(Hg) indicates that the copper is amalgamated with the mercury. This step essentially serves as a means of preconcentrating the analyte from the larger volume of the solution to the smaller volume of the electrode. The solution is stirred during electrolysis to increase the rate of deposition. Near the end of the deposition time stirring is stopped, eliminating convection as a mode of mass transport. Deposition times of 1-30 min are common, with longer times being used for analytes at lower concentrations.\n",
            "\n",
            "In the second step, the potential is scanned anodically toward more positive potentials. When the potential of the working electrode is sufficiently positive the analyte is stripped from the electrode, returning to solution as its oxidized form\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The current during the stripping step is monitored as a function of potential, giving rise  to  peak-shaped voltammograms similar to that shown in Figure 11.37. The peak current is proportional to the analyte's concentration in the solution.\n",
            "\n",
            "Anodic stripping voltammetry is very sensitive to experimental conditions, which must be carefully controlled if results are to be accurate and precise. Key variables include the area of the mercury film electrode or the size of the Hg drop when using a hanging mercury drop electrode, the deposition time, the rest time, the rate of stirring, and the scan rate during the stripping step. Anodic stripping voltammetry is best used for metals that form amalgams with mercury, several examples of which are listed in Table 11.11.\n",
            "\n",
            "The experimental design for cathodic stripping voltammetry is similar to that for anodic stripping voltammetry with two exceptions. First, the deposition step in-\n",
            "\n",
            "Table 11.11\n",
            "\n",
            "## Representative Examples of Analytes Determined by Stripping Voltammetry\n",
            "\n",
            "| Anodic Stripping Voltammetry                                 | Cathodic Stripping Voltammetry                               | Absorptive Stripping Voltammetry                                        |\n",
            "|--------------------------------------------------------------|--------------------------------------------------------------|-------------------------------------------------------------------------|\n",
            "| bismuth cadmium copper gallium indium lead thallium tin zinc | bromide chloride iodide mercaptans (RSH) sulfide thiocyanate | bilirubin codeine cocaine digitoxin dopamine heme monensin testosterone |\n",
            "\n",
            "Source: Compiled from Peterson, W. M.; Wong, R. V. Am. Lab. November 1981, 116-128; and Wang, J. Am. Lab. May 1985, 41-50. 17,18\n",
            "\n",
            "volves the oxidation of the Hg electrode to Hg2 2+ , which then reacts with the analyte to form an insoluble film at the surface of the electrode. For example, when Cl is the analyte the deposition step is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Second, stripping is accomplished by scanning cathodically toward a more negative potential, reducing Hg2 2+ back to Hg and returning the analyte to solution.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Table 11.11 lists several analytes that have been analyzed successfully by cathodic stripping voltammetry.\n",
            "\n",
            "In adsorptive stripping voltammetry the deposition step occurs without electrolysis.  Instead,  the  analyte  adsorbs to the electrode's surface. During deposition the electrode is maintained at a potential that enhances adsorption. For example, adsorption of a neutral molecule on a Hg drop is enhanced if the electrode is held at -0.4 V versus the SCE, a potential at which the surface charge of mercury is approximately zero. When deposition is complete the potential is scanned in an anodic or cathodic direction depending on whether we wish to oxidize or reduce the analyte. Examples of compounds that have been analyzed by absorptive stripping voltammetry also are listed in Table 11.11.\n",
            "\n",
            "Amperometry The final voltammetric technique to be considered is amperometry, in which a constant potential is applied to the working electrode, and current is measured as a function of time. Since the potential is not scanned, amperometry does not lead to a voltammogram.\n",
            "\n",
            "One important application of amperometry is in the construction of chemical sensors. One of the first amperometric sensors to be developed was for dissolved O2 in blood, which was developed in 1956 by L. C. Clark. The design of the amperometric sensor is shown in Figure 11.38 and is similar to potentiometric membrane electrodes. A gas-permeable membrane is stretched across the end of the sensor and is separated from the working and counter electrodes by a thin solution of  KCl.  The working electrode is a Pt disk cathode, and an Ag ring anode is the\n",
            "\n",
            "## amperometry\n",
            "\n",
            "A form of voltammetry in which we measure current as a function of time while maintaining a constant potential.\n",
            "\n",
            "Figure 11.38\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Clark amperometric sensor for the determination of dissolved O2.\n",
            "\n",
            "counter electrode. Although several gases can diffuse across the membrane, including O2, N2, and CO2, only oxygen is reduced at the cathode.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Another example of an amperometric sensor is the glucose sensor. In this case the single membrane in Figure 11.38 is replaced with three membranes. The outermost membrane is of polycarbonate, which is permeable to glucose and O2. The second membrane contains an immobilized preparation of glucose oxidase that catalyzes the oxidation of glucose to gluconolactone and hydrogen peroxide.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The hydrogen peroxide then diffuses through the innermost membrane of cellulose acetate, where it is oxidized at a Pt anode.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Figure 11.39 summarizes the reactions taking place in this amperometric sensor. FAD is the oxidized form of flavin adenine nucleotide (the active site of the enzyme glucose oxidase), and FADH2 is  the  active  site's  reduced  form. Note that O2 serves as a mediator, carrying electrons to the electrode. Other mediators, such as Fe(CN)6 3, can be used in place of O2.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "By changing the enzyme and mediator, the amperometric sensor  in  Figure  11.39  is  easily  extended  to  the  analysis  of  other  substrates.  Other  bioselective  materials  may  be  incorporated  into  amperometric sensors. For example, a CO2 sensor has been developed using an amperometric O2 sensor with a two-layer membrane, one of which contains an immobilized preparation of autotrophic bacteria. 19 As CO2 diffuses through the membranes, it is converted to O2 by the bacteria, increasing the concentration of O2 at  the  Pt cathode.\n",
            "\n",
            "Electrode\n",
            "\n",
            "## Figure 11.39\n",
            "\n",
            "Schematic showing the reactions by which an amperometric biosensor responds to glucose.\n",
            "\n",
            "## 11D.6 Quantitative Applications\n",
            "\n",
            "Quantitative voltammetry has been applied to a wide variety of sample types, including environmental samples, clinical samples, pharmaceutical formulations, steels, gasoline, and oil.\n",
            "\n",
            "Selecting the Voltammetric Technique The choice of which voltammetric technique to use depends on the sample's characteristics, including the analyte's expected concentration and the location of the sample. Amperometry is best suited for use as a detector in flow systems or as a selective sensor for the rapid analysis of a single analyte. The portability of amperometric sensors, which are similar to potentiometric sensors, make them ideal for field studies.\n",
            "\n",
            "Pulse polarography and stripping voltammetry can frequently be used interchangeably, although each has its advantages and disadvantages. Pulse polarography is better for analyzing a wider range of inorganic and organic analytes because the need to preconcentrate the analyte at the electrode surface restricts the application of anodic and cathodic stripping voltammetry.\n",
            "\n",
            "When either pulse polarography or anodic stripping voltammetry can be used, the selection is often based on the analyte's expected concentration and the desired\n",
            "\n",
            "accuracy and precision. Detection limits for normal pulse polarography generally are on the order of 10 -6 -10 -7 M, whereas those for differential pulse polarography, staircase, and square-wave polarography are between 10 -7 M and 10 -8 M. Preconcentrating the analyte in stripping voltammetry lowers the detection limit for many analytes to as little as 10 -10 M. On the other hand, the current in stripping voltammetry is much more sensitive than pulse polarography to changes in experimental conditions, which may lead to poorer precision and accuracy.\n",
            "\n",
            "Anodic stripping voltammetry also suffers from occasional interferences when two metals, such as Cu and Zn, combine to form an intermetallic compound in the mercury amalgam. The deposition potential for Zn 2+ is sufficiently negative that any Cu 2+ present in the sample is also deposited. After deposition, intermetallic compounds such as CuZn and CuZn2 form within the mercury amalgam. During the stripping step, the zinc in the intermetallic compounds strips  at  potentials  near  that  of  copper,  decreasing  the  current  for  zinc  and  increasing the current for copper. This problem can often be overcome by adding a third element that forms a stronger intermetallic compound with the interfering metal. Thus, adding Ga 3+ minimizes this problem by forming an intermetallic compound of Cu and Ga.\n",
            "\n",
            "Correcting for Residual Current In any quantitative analysis the signal due to the analyte must be corrected for signals arising from other sources. The total measured current in any voltammetric experiment, i tot ,  consists of two parts: that due to the analyte's oxidation or reduction, i a , and a background, or residual, current, i r .\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The residual current, in turn, has two sources. One source is a faradaic current due to the oxidation or reduction of trace impurities in the sample, i i . The other source is the charging current, i ch , that is present whenever the working electrode's potential changes.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Faradaic currents due to impurities can usually be minimized by carefully preparing the sample. For example, one important impurity is dissolved O2, which is reduced first to H2O2 and then to H2O. Dissolved O2 is  removed by bubbling an inert gas such as N2 through the sample before the analysis.\n",
            "\n",
            "Two methods are commonly used to correct for the residual current. One method is to extrapolate the total measured current when the analyte's faradaic current is zero. This is the method shown in the voltammograms included in this chapter. The advantage of this method is that it does not require any additional data. On the other hand, extrapolation assumes that changes in the residual current with potential are predictable, which often is not the case. A second, and more rigorous, approach is to obtain a voltammogram for an appropriate blank. The blank's residual current is then subtracted from the total current obtained with the sample.\n",
            "\n",
            "Analysis for Single Components The analysis of samples containing only a single electroactive analyte is straightforward. Any of the standardization methods discussed in Chapter 5 can be used to establish the relationship between current and the concentration of analyte.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11.10\n",
            "\n",
            "The concentration of As(III) in water can be determined by differential pulse polarography in 1 M HCl. The initial potential is set to -0.1 V versus the SCE, and is scanned toward more negative potentials at a rate of 5 mV/s. Reduction of  As(III)  to  As(0)  occurs at a potential of approximately -0.44 V versus the SCE. The peak currents, corrected for the residual current, for a set of standard solutions are shown in the following table.\n",
            "\n",
            "| [As(III)] (M)   |   i p ( µ A) |\n",
            "|-----------------|--------------|\n",
            "| 1.00 × 10 -6    |        0.298 |\n",
            "| 3.00 × 10 -6    |        0.947 |\n",
            "| 6.00 × 10 -6    |        1.83  |\n",
            "| 9.00 × 10 -6    |        2.72  |\n",
            "\n",
            "What is the concentration of As(III) in a sample of water if the peak current under the same conditions is 1.37 µ A?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "Linear regression gives the equation for the calibration curve as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Substituting the sample's peak current into the regression equation gives the concentration of As(III) as 4.49 × 10 -6 M.\n",
            "\n",
            "## EXAMPLE 11.11\n",
            "\n",
            "The concentration of copper in a sample of sea water is determined by anodic stripping voltammetry using the method of standard additions. When a 50.0-mL sample is analyzed, the peak current is 0.886 µ A. A 5.00µ L spike of 10.0-ppm Cu 2+ is added, giving a peak current of 2.52 µ A. Calculate the parts per million of copper in the sample of sea water.\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "Peak currents in anodic stripping voltammetry are a linear function of concentration\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where k is a constant. Thus, for the sample we write\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and for the standard addition\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving the first equation for k, substituting into the second equation, and simplifying gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving gives the concentration of Cu 2+ as 5.42 10 -4 ppm, or 0.542 ppb.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Multicomponent Analysis One advantage of voltammetry as a quantitative method of analysis is its capability for analyzing two or more analytes in a single sample. As long as the components behave independently, the resulting voltammogram for a multicomponent mixture is a summation of their respective individual voltammograms. If the separation between the half-wave potentials or peak potentials is sufficient, each component can be determined independently as if it were the only component in the sample (Figure 11.40). The minimum separation between the half-wave potentials or peak potentials for the independent analysis of two components depends on several factors, including the type of electrode and the potential-excitation signal. For normal polarography the separation must be at least ± 0.2-0.3 V, and for differential pulse voltammetry a minimum separation of ± 0.04-0.05 V is needed.\n",
            "\n",
            "When the overlap between the voltammograms for two components prevents their independent analysis, a simultaneous analysis similar to that used in spectrophotometry may be possible. An example of this approach is outlined in Example 11.12.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11.12\n",
            "\n",
            "The differential pulse polarographic analysis of mixtures of indium and cadmium in 0.1 M HCl is complicated by the overlap of their respective voltammograms. 20 The peak potential for indium is at -0.557 V, and that for cadmium occurs at a potential of -0.597 V. When a 0.800-ppm indium standard is analyzed, the peak current (in arbitrary units) is found to be 200.5 at -0.557 V and 87.5 at -0.597 V. A standard solution of 0.793-ppm cadmium gives peak currents of 58.5 at -0.557 V and 128.5 at -0.597 V. What is the concentration of indium and cadmium in a sample if the peak current is 167.0 at a potential of -0.557 V and 99.5 at a potential of -0.597 V?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "Peak currents in differential pulse polarography are a linear function of the concentration of analyte; thus\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where k is  a  constant that depends on the analyte and the applied potential. To determine the concentrations of indium and cadmium in the sample, we must first find the values of k for each analyte at each potential. For simplicity we will identify the potential of -0.557 V as E 1, and that for -0.597 V as E 2. The values of k are\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Figure 11.40\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Voltammograms showing the independent analysis of two components.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Next, we write simultaneous equations for the current at the two potentials\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "i E 2 = 109.4 ppm -1 (ppm In) + 162.0 ppm -1 (ppm Cd) = 99.5\n",
            "\n",
            "Solving the simultaneous equations, which is left as an exercise, gives the concentration of indium as 0.606 ppm and the concentration of cadmium as 0.205 ppm.\n",
            "\n",
            "Environmental Samples One area in which quantitative voltammetry has had a significant effect is in the analysis of trace metals in environmental samples. The most common samples are natural waters, including groundwater, lakes, rivers and streams, sea water, rain, and snow. Concentrations of trace metals at the parts-perbillion level can be determined using differential pulse polarography, whereas with anodic stripping voltammetry the determination of trace metals at the pptr (partsper-trillion) level is possible. The combination of low detection limits and the capability for the simultaneous analysis of several analytes makes differential pulse polarography and anodic stripping voltammetry ideally suited for such samples.\n",
            "\n",
            "One interesting application of anodic stripping voltammetry to the analysis of natural waters is the determination of the speciation, or chemical form, of the trace metals. The speciation of a trace metal is important because its bioavailability, toxicity, and ease of transport through the environment often depend on its chemical form. For example, trace metals strongly bound to colloidal particles are generally not available to aquatic lifeforms and, therefore, are not toxic. Unfortunately, anodic stripping voltammetry cannot distinguish the exact chemical form of a trace metal as closely related species, such as Pb 2+ and PbCl + , yield only a single stripping peak. Instead, trace metals are divided into several 'operationally defined' categories that have environmental significance.\n",
            "\n",
            "Several speciation schemes have been developed, 21 but we consider only the speciation scheme proposed by Batley and Florence, 22 which uses a combination of anodic stripping voltammetry with ion-exchange and UV irradiation to divide soluble trace metals into seven classes. Anodic stripping voltammetry in a pH 4.8 acetic acid buffer is used to distinguish labile metals present as hydrated ions, weakly bound complexes, or weakly adsorbed on colloidal surfaces from those metals that are bound in stronger complexes or strongly adsorbed. Only those metals that are hydrated, weakly bound, or weakly adsorbed deposit at the electrode. Ion exchange and UV irradiation are used to further subdivide the trace metals. A Chelex-100 ion-exchange resin is used to distinguish between ionic metals and strongly bound metals, whereas UV radiation is used to separate metals bound to organic and inorganic phases. Table 11.12 shows how trace metals are divided into seven classes using these three experimental techniques. The analysis of sea water samples, for example, showed that cadmium, copper, and lead were primarily present as labile  organic complexes or as labile adsorbates on organic colloids (see group II, Table 11.12).\n",
            "\n",
            "Differential pulse polarography and stripping voltammetry have been applied to the analysis of trace metals in airborne particulates, incinerator fly ash, rocks,\n",
            "\n",
            "## Table 11.12 Speciation of Soluble Trace Metals According to the Scheme of Batley and Florence 22\n",
            "\n",
            "| Method         | Speciation of Soluble Metals                                              | Speciation of Soluble Metals                                    | Speciation of Soluble Metals                                        | Speciation of Soluble Metals   | Speciation of Soluble Metals   | Speciation of Soluble Metals                                          | Speciation of Soluble Metals                                              |\n",
            "|----------------|---------------------------------------------------------------------------|-----------------------------------------------------------------|---------------------------------------------------------------------|--------------------------------|--------------------------------|-----------------------------------------------------------------------|---------------------------------------------------------------------------|\n",
            "| ASV            | Labile Metals                                                             | Labile Metals                                                   | Labile Metals                                                       | Nonlabile or Bound Metals      | Nonlabile or Bound Metals      | Nonlabile or Bound Metals                                             | Nonlabile or Bound Metals                                                 |\n",
            "| Ion-Exchange   | Removed                                                                   | Not Removed                                                     | Not Removed                                                         | Removed                        | Removed                        | Not Removed                                                           | Not Removed                                                               |\n",
            "| UV Irradiation |                                                                           | Released                                                        | Not Released                                                        | Released                       | Not Released                   | Released                                                              | Not Released                                                              |\n",
            "| Description    | free metal ions, labile organic complexes, and labile inorganic complexes | labile organic complexes and labile metals absorbed on organics | labile inorganic complexes and labile metals absorbed on inorganics | nonlabile organic complexes    | nonlabile inorganic complexes  | nonlabile organic complexes and nonlabile metals absorbed on organics | nonlabile inorganic complexes and nonlabile metals absorbed on inorganics |\n",
            "| Group          | I                                                                         | II                                                              | III                                                                 | IV                             | V                              | VI                                                                    | VII                                                                       |\n",
            "\n",
            "minerals, and sediments. The trace metals, of course, must be brought into solution by digesting or extracting before the voltammetric analysis.\n",
            "\n",
            "Amperometric sensors also are used to analyze environmental samples. For example, the dissolved O2 sensor described earlier is routinely used for the determination of dissolved oxygen and biochemical oxygen demand, or BOD, in waters and wastewaters. The latter test, which is a measure of the amount of oxygen required by aquatic bacteria during the decomposition of organic matter, is of importance in evaluating the efficiency of wastewater treatment plants and in monitoring organic pollution in natural waters. A high BOD corresponds to a high concentration of organic material that may seriously deplete the level of dissolved oxygen in the water. Other amperometric sensors have been developed to monitor anionic surfactants in water, and CO2, H2SO4, and NH3 in atmospheric gases. 19\n",
            "\n",
            "Clinical Samples Differential pulse polarography and stripping voltammetry have been used to determine the concentration of trace metals in a variety of matrices, including blood, urine, and tissue samples. The determination of lead in blood is of considerable interest due to concerns about lead poisoning. Because the concentration of lead in blood is so small, anodic stripping voltammetry frequently is the method of choice. The analysis is complicated, however, by the presence of proteins that may adsorb at the surface of the mercury electrode, inhibiting either the deposition or stripping of lead. In addition, proteins may prevent the electrodeposition of lead through the formation of stable, nonlabile complexes. For these reasons samples of whole blood must be digested or ashed before the analysis. Differential  pulse  polarography is one of the few techniques that can be used for the routine quantitative analysis of drugs in biological fluids at concentrations of less than 10 -6 M. 23 Amperometric sensors based on enzyme catalysts also have wide applicability. 24 Table 11.13 provides a partial list of enzymatic amperometric sensors.\n",
            "\n",
            "Miscellaneous Samples Besides environmental and clinical samples, differential pulse polarography and stripping voltammetry have been used for the analysis of trace metals in other samples, including food, steels and other alloys, gasoline, gunpowder residues, and pharmaceuticals. Voltammetry is also an important tool for\n",
            "\n",
            "## Table 11.1 3\n",
            "\n",
            "## Representative Examples of Amperometric Biosensors\n",
            "\n",
            "| Analyte                                                                            | Enzyme                                                                                                                                                                                       | Species Detected                                                     |\n",
            "|------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------|\n",
            "| choline ethanol formaldehyde glucose glutamine glycerol lactate phenol inorganic P | choline oxidase alcohol oxidase formaldehyde dehydrogenase glucose oxidase glutaminase, glutamate oxidase glycerol dehydrogenase lactate oxidase polyphenol oxidase nucleoside phosphorylase | H 2 O 2 H 2 O 2 NADH a H 2 O 2 H 2 O 2 NADH, O 2 H 2 O 2 quinone O 2 |\n",
            "\n",
            "Source: Compiled from Cammann, K.; Lemke, U.; Rohen, A.; et al. Angew. Chem. Int. Ed. Engl. 1991, 30, 516-539.\n",
            "\n",
            "a NADH is the reduced form of nicotinamide adenine dinucleotide.\n",
            "\n",
            "the quantitative analysis of organics, particularly in the pharmaceutical industry, in which it is used to determine the concentration of drugs and vitamins in formulations. 23 For example, voltammetric methods have been developed for the quantitative analysis of vitamin A, niacinamide, and riboflavin. When the compound of interest is not electroactive, it often can be derivatized to an electroactive form. One example is the differential pulse polarographic determination of sulfanilamide, in which it is converted into an electroactive azo dye by coupling with sulfamic acid and 1-naphthol.\n",
            "\n",
            "Representative Methods\n",
            "\n",
            "## Method 11. 3\n",
            "\n",
            "## Determination of Chlorpromazine 25\n",
            "\n",
            "Description of Method. The amount of chlorpromazine in a pharmaceutical formulation is determined voltammetrically at a graphite working electrode in a nonstirred solution. Calibration is achieved using the method of standard additions.\n",
            "\n",
            "Procedure. Place 10.00 mL of a solution consisting of 0.01 M HCl and 0.1 M KCl in the electrochemical cell. Place a graphite working electrode, a Pt auxiliary electrode, and a SCE reference electrode in the cell, and record the voltammogram from +0.2 V to 2.0 V at a scan rate of 50 mV/s. Weigh out an appropriate amount of the pharmaceutical formulation, and dissolve it in a small amount of the electrolyte. Transfer the solution to a 100-mL volumetric flask, and dilute to volume with the electrolyte. Filter a small amount of the diluted solution, and transfer 1.00 mL of the filtrate to the voltammetric cell. Mix the contents of the voltammetric cell, and allow the solution to sit for 10 s before recording the voltammogram. Return the potential to +0.2 V, add 1.00 mL of a chlorpromazine standard, and record the voltammogram. Report the %w/w chlorpromazine in the formulation.\n",
            "\n",
            "## Questions\n",
            "\n",
            "## 1. Is chlorpromazine oxidized or reduced at the graphite working electrode?\n",
            "\n",
            "Since the direction of the scan is toward more positive potentials, the chlorpromazine is oxidized at the graphite working electrode.\n",
            "\n",
            "-Continued\n",
            "\n",
            "- 2. Why is it not necessary to remove dissolved O2 from the solution before recording the voltammogram?\n",
            "\n",
            "Dissolved O2 is a problem when scanning toward more negative potentials where its reduction can lead to a significant cathodic current. In this procedure we are scanning toward more positive potentials and generating anodic currents; thus, dissolved O2 is not an interferent.\n",
            "\n",
            "- 3. What is the purpose of waiting 10 s after mixing the contents of the electrochemical cell before recording the voltammogram?\n",
            "\n",
            "Mixing the solution leads to convection, but the delay allows convection to cease.\n",
            "\n",
            "- 4. In preparing the sample for analysis the initial solution is filtered. Why is it not necessary to collect the entire filtrate before proceeding?\n",
            "\n",
            "This analysis is an example of a concentration technique. Once the original sample is brought to volume in the 100-mL volumetric flask, any portion of the sample solution, even that obtained on filtering, may be used for the analysis.\n",
            "\n",
            "## 11D. 7 Characterization Applications\n",
            "\n",
            "In the previous section we saw how voltammetry can be used to determine the concentration of an analyte. Voltammetry also can be used to obtain additional information, including verifying electrochemical reversibility, determining the number of electrons transferred in a redox reaction, and determining equilibrium constants for coupled chemical reactions. Our discussion of these applications is limited to the use of voltammetric techniques that give limiting currents, although other voltammetric techniques also can be used to obtain the same information.\n",
            "\n",
            "Electrochemical Reversibility and Determination of n In deriving a relationship between E 1/2 and the standard-state potential for a redox couple (11.41), we noted that the redox reaction must be reversible. How can we tell if a redox reaction is reversible from its voltammogram? For a reversible reaction, equation 11.40 describes the voltammogram.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "A plot of E versus log( / i i lim -i )  for  a  reversible  reaction,  therefore,  should be a straight line with a slope of -0.05916/ n. In addition, the slope should yield an integer value for n.\n",
            "\n",
            "## EXAMPLE 11.1 3\n",
            "\n",
            "The following data were obtained from the linear scan hydrodynamic voltammogram of a reversible reduction reaction\n",
            "\n",
            "|   Potential (V vs. SCE) |   Current ( µ A) |\n",
            "|-------------------------|------------------|\n",
            "|                  -0.358 |             0.37 |\n",
            "|                  -0.372 |             0.95 |\n",
            "|                  -0.382 |             1.71 |\n",
            "|                  -0.4   |             3.48 |\n",
            "|                  -0.41  |             4.2  |\n",
            "|                  -0.435 |             4.97 |\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.42\n",
            "\n",
            "Effect of complexation on voltammograms.\n",
            "\n",
            "The limiting current was 5.15 µ A. Show that the reduction reaction is reversible, and determine values for n and E 1/2 .\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "If  the  reaction is reversible, then a plot of E versus log( / i i lim -i )  will  be  a straight line with a slope that yields an integer value for n. As shown in Figure 11.41, plotting the data in this manner gives a straight line. Linear regression gives the equation for the straight line as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Knowing that\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "gives a value for n of 1.97, or 2 electrons. Combining equations 11.40 and 11.41 shows that the y -intercept for a plot of E versus log( / i i lim -i ) is equal to the halfwave potential. Thus, E 1/2 for the reduction reaction is -0.391 V versus the SCE.\n",
            "\n",
            "Figure 11.41\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Graphical determination of electrochemical reversibility, n, and half-wave potential in linear scan hydrodynamic voltammetry.\n",
            "\n",
            "Determining Equilibrium Constants for Coupled Chemical Reactions Another important application of voltammetry is the determination of equilibrium constants for solution reactions that are coupled to a redox reaction occurring at the electrode. The presence of the solution reaction affects the ease of electron transfer, shifting the potential to more negative or more positive potentials. Consider, for example, the reduction of O to R\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "the voltammogram for which is shown in Figure 11.42. If a ligand, L, capable of forming a strong complex with O, is present, then the reaction\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "also must be considered. The overall reaction, therefore, is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Because of its stability, reduction of the OL p complex is less favorable than the reduction of O. As shown in Figure 11.42, the voltammograms for OL p occur at potentials  more negative than those for O. Furthermore, the shift in the voltammogram depends on the ligand's concentration.\n",
            "\n",
            "The shift in the voltammogram for a metal ion in the presence of a ligand may be used to determine both the metal-ligand complex's stoichiometry and its formation constant. To derive a relationship between the relevant variables we begin with two equations: the Nernst equation for the reduction of O\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and the stability constant for the complex at the electrode surface\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "In the absence of ligand the half-wave potential occurs when [R] x =0 and [O] x= 0 are equal; thus, from the Nernst equation we have\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where the subscript 'nc' signifies that no complex is present.\n",
            "\n",
            "When ligand is present we must account for its effect on the concentration of O. Solving equation 11.42 for [O] x =0 and substituting into the Nernst equation gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "If  the  ligand  is  present  in  excess  and  the  formation constant is sufficiently large, such that all of O is present as the complex, then [R] x =0 and [OL ] p x =0 are equal at the half-wave potential, and equation 11.44 simplifies to\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where the subscript 'c' indicates that the complex is present. Defining ∆ E 1/2 as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and substituting equations 11.44 and 11.45 gives, after rearranging\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "A plot of ∆ E 1/2 versus log [L], therefore, gives a straight line. The slope of the line is used to determine the stoichiometric coefficient, p, and the y -intercept gives the value for the formation constant β p .\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 11.1 4\n",
            "\n",
            "A voltammogram for the two-electron reduction of M has a half-wave potential of -0.226 V versus the SCE. In the presence of an excess of the ligand L, the following half-wave potentials are recorded\n",
            "\n",
            "|   [L] (M) |   ( E 1/2 ) c (V vs. SCE) |\n",
            "|-----------|---------------------------|\n",
            "|      0.02 |                    -0.494 |\n",
            "|      0.04 |                    -0.512 |\n",
            "|      0.06 |                    -0.523 |\n",
            "|      0.08 |                    -0.53  |\n",
            "|      0.1  |                    -0.536 |\n",
            "\n",
            "Determine values for the stoichiometry of the complex and the formation constant of the complex.\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "First we must calculate values of ∆ E 1/2 using equation 11.46; thus\n",
            "\n",
            "|   [L] (M) |   ∆ E 1/2 (V vs. SCE) |\n",
            "|-----------|-----------------------|\n",
            "|      0.02 |                -0.268 |\n",
            "|      0.04 |                -0.286 |\n",
            "|      0.06 |                -0.297 |\n",
            "|      0.08 |                -0.304 |\n",
            "|      0.1  |                -0.31  |\n",
            "\n",
            "The plot of ∆ E 1/2 as a function of the log of the ligand concentration is shown in Figure 11.43. Linear regression gives the equation for the straight line as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "From the slope and intercept we know that\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving these equations gives the stoichiometry of the complex as ML2 ( p = 2.03), with a formation constant of β 2 = 3.5 × 10 12 .\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 11.43\n",
            "\n",
            "Graphical determination of the stoichiometry and formation constant for a complexation reaction.\n",
            "\n",
            "## 11D.8 Evaluation\n",
            "\n",
            "Scale  of  Operation Voltammetry is routinely used to analyze samples at the parts-per-million level and, in some cases, can be used to detect analytes at the parts-per-billion or parts-per-trillion level. Most analyses are carried out in conventional electrochemical cells using macro samples; however, microcells are available that require as little as 50 µ L of sample. Microelectrodes, with diameters  as  small  as  2 µ m, allow voltammetric measurements to be made on even smaller samples. For example, the concentration of glucose in 200µ m pond snail neurons has been successfully monitored using a 2µ m amperometric glucose electrode. 26\n",
            "\n",
            "Accuracy The accuracy of a voltammetric analysis often is limited by the ability to correct for residual currents, particularly those due to charging. For analytes at the parts-per-million level, accuracies of ± 1-3% are easily obtained. As expected, a decrease in accuracy is experienced when analyzing samples with significantly smaller concentrations of analyte.\n",
            "\n",
            "Precision Precision is generally limited by the uncertainty in measuring the limiting or peak current. Under most experimental conditions, precisions of ± 1-3% can be reasonably expected. One exception is the analysis of ultratrace analytes in complex matrices by stripping voltammetry, for which precisions as poor as ± 25% are possible.\n",
            "\n",
            "Sensitivity In many voltammetric experiments, sensitivity can be improved by adjusting the experimental conditions. For example, in stripping voltammetry, sensitivity  is  improved by increasing the deposition time, by increasing the rate of the linear potential scan, or by using a differential-pulse technique. One reason for the popularity of potential pulse techniques is an increase in current relative to that obtained with a linear potential scan.\n",
            "\n",
            "Selectivity Selectivity in voltammetry is determined by the difference between half-wave potentials or peak potentials, with minimum differences of ± 0.2-0.3 V required for a linear potential scan, and ± 0.04-0.05 V for differential pulse voltammetry. Selectivity can be improved by adjusting solution conditions. As we have seen, the presence of a complexing ligand can substantially shift the potential at which an analyte is oxidized or reduced. Other solution parameters, such as pH, also can be used to improve selectivity.\n",
            "\n",
            "Time, Cost, and Equipment Commercial instrumentation for voltammetry ranges from less than $1000 for simple instruments to as much as $20,000 for more sophisticated instruments. In general, less expensive instrumentation is limited to linear potential scans, and the more expensive instruments allow for more complex potential-excitation signals using potential pulses. Except for stripping voltammetry, which uses long deposition times, voltammetric analyses are relatively rapid.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## 11E KEY TERMS\n",
            "\n",
            "| amalgam ( p. 509 )                      |\n",
            "|-----------------------------------------|\n",
            "| amperometry ( p. 519 )                  |\n",
            "| anode ( p. 466 )                        |\n",
            "| anodic current ( p. 510 )               |\n",
            "| asymmetry potential ( p. 476 )          |\n",
            "| auxiliary electrode ( p. 463 )          |\n",
            "| cathode ( p. 466 )                      |\n",
            "| cathodic current ( p. 510 )             |\n",
            "| charging current ( p. 513 )             |\n",
            "| convection ( p. 512 )                   |\n",
            "| coulometric titrations ( p. 501 )       |\n",
            "| coulometry ( p. 496 )                   |\n",
            "| counter electrode ( p. 462 )            |\n",
            "| current efficiency ( p. 496 )           |\n",
            "| diffusion ( p. 511 )                    |\n",
            "| dropping mercury electrode p.           |\n",
            "| (                                       |\n",
            "| electrical double layer ( p. 513 )      |\n",
            "| electrode of the first kind ( p. 474 )  |\n",
            "| electrode of the second kind ( p. 475 ) |\n",
            "\n",
            "| enzyme electrodes ( p.                          | enzyme electrodes ( p.                          | enzyme electrodes ( p.                          | 484 )                                           |\n",
            "|-------------------------------------------------|-------------------------------------------------|-------------------------------------------------|-------------------------------------------------|\n",
            "| faradaic current ( p. Faraday's law ( p. 496    | hydrodynamic                                    | faradaic current ( p. Faraday's law ( p. 496    | voltammetry ( p. 513 )                          |\n",
            "| liquid-based ion-selective electrode ( p. 482 ) | liquid-based ion-selective electrode ( p. 482 ) | liquid-based ion-selective electrode ( p. 482 ) | liquid-based ion-selective electrode ( p. 482 ) |\n",
            "| liquid junction potential transport ( p. 511    |                                                 | ( )                                             | p. 470 )                                        |\n",
            "| mass mediator                                   | ( p. 500                                        | ( p.                                            | )                                               |\n",
            "| membrane potential 475 )                        | membrane potential 475 )                        | membrane potential 475 )                        | membrane potential 475 )                        |\n",
            "| migration                                       | ( p. 512 )                                      |                                                 |                                                 |\n",
            "| Ohm's law                                       | nonfaradaic current                             | ( p.                                            | 512 )                                           |\n",
            "| ( p. 463 )                                      | ( p. 463 )                                      | ( p. 463 )                                      | ( p. 463 )                                      |\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## 11F SUMMARY\n",
            "\n",
            "Electrochemical methods covered in this chapter include potentiometry, coulometry, and voltammetry. Potentiometric methods are based on the measurement of an electrochemical cell's potential when only a negligible current is allowed to flow. In principle the Nernst equation can be used to calculate the concentration of species in the electrochemical cell by measuring its potential and solving the Nernst equation; the presence of liquid junction potentials,  however, necessitates the use of an external standardization or the use of standard additions.\n",
            "\n",
            "Potentiometric electrodes are divided into two classes: metallic electrodes and membrane electrodes. The smaller of these classes are the metallic electrodes. Electrodes of the first kind respond to the concentration of their cation in solution; thus the potential of an Ag wire is determined by the concentration of Ag + in solution. When another species is present in solution and in equilibrium with the metal ion, then the electrode's potential will respond to the concentration of that ion. For example, an Ag wire in contact with a solution of Cl will  respond to the concentration of Cl since the relative concentrations of Ag + and Cl are fixed by the solubility product for AgCl. Such electrodes are called electrodes of the second kind.\n",
            "\n",
            "The potential of a membrane electrode is determined by a difference in the composition of the solution on either side of the membrane. Electrodes using a glass membrane respond to ions that bind to negatively charged sites on the glass membrane's surface. A pH electrode is one example of a glass membrane electrode. Other kinds of membrane electrodes include those using insoluble crystalline solids and liquid ion exchangers incorporated in a hydrophobic membrane. The F ion-selective electrode, which uses a single crystal of LaF3 as  the  ion-selective  membrane, is an example of a solid-state electrode. The Ca 2+ ion-selective electrode, in which a chelating ligand such as di-( n -decyl)phosphate is immobilized in a PVC membrane, is an example of a liquid-based ion-selective electrode.\n",
            "\n",
            "Potentiometric electrodes also can be designed to respond to molecules by incorporating a reaction producing an ion whose concentration can be determined using a traditional ion-selective electrode. Gas-sensing electrodes, for example, include a gas-permeable membrane that isolates the ion-selective electrode from the solution containing the analyte. Diffusion of a dissolved gas across the membrane alters the composition of the inner solution in a manner that can be followed with an ion-selective electrode. Enzyme electrodes operate in the same way.\n",
            "\n",
            "Coulometric methods are based on Faraday's law that the total charge or current passed during an electrolysis is proportional to the amount of reactants and products in the redox reaction. If the electrolysis is 100% efficient, in that only the analyte is oxidized or reduced, then the total charge or current can be used to determine\n",
            "\n",
            "| overpotential ( p. 497 ) polarography ( p. 515 )                                                                                                              |\n",
            "|---------------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
            "| potentiometer ( p. 464 ) potentiostat ( p. 465 ) redox electrode ( p. 475 ) reference electrode ( p. 462 ) residual current ( p. 513 ) salt bridge ( p. 466 ) |\n",
            "| saturated calomel electrode ( p. 472 ) silver/silver chloride electrode ( p. 473 )                                                                            |\n",
            "| solid-state ion-selective electrode ( p. 479 )                                                                                                                |\n",
            "| standard hydrogen electrode ( p. 471 ) static mercury drop electrode ( p. 509 )                                                                               |\n",
            "| total ionic strength                                                                                                                                          |\n",
            "| adjustment buffer ( p. 487 )                                                                                                                                  |\n",
            "| voltammetry ( p. 508 )                                                                                                                                        |\n",
            "| )                                                                                                                                                             |\n",
            "| voltammogram ( p. 508                                                                                                                                         |\n",
            "\n",
            "the amount of analyte in a sample. In controlled-potential coulometry, a constant potential is applied and the current is measured as a function of time, whereas in controlled-current coulometry, the current is held constant and the time required to completely oxidize or reduce the analyte is measured.\n",
            "\n",
            "In voltammetry we measure the current in an electrochemical cell as a function of the applied potential. Individual voltammetric methods differ in terms of the type of electrode used, how the applied potential is changed, and whether the transport of material to the electrode's surface is enhanced by stirring.\n",
            "\n",
            "Polarography is a voltammetric experiment conducted at a mercury electrode under conditions in which the solution is not stirred.  Normal polarography uses a dropping mercury electrode (or a static mercury drop electrode) and a linear potential scan in an unstirred solution. Other forms of polarography include normal pulse polarography, differential pulse polarography, staircase polarography, and square-wave polarography, all of which apply a series of potential pulses to the mercury electrode.\n",
            "\n",
            "11G\n",
            "\n",
            "In hydrodynamic voltammetry the solution is stirred either by using a magnetic stir bar or by rotating the electrode. Because the solution is stirred, a dropping mercury electrode cannot be used and is replaced with a solid electrode. Both linear potential scans or potential pulses can be applied.\n",
            "\n",
            "In stripping voltammetry the analyte is first deposited on the electrode, usually as the result of an oxidation or reduction reaction. The potential is then scanned, either linearly or by using potential pulses, in a direction that removes the analyte by a reduction or oxidation reaction.\n",
            "\n",
            "Amperometry is a voltammetric method in which a constant potential is applied to the electrode and the resulting current is measured. Amperometry is most often used in the construction of chemical sensors that, as with potentiometric sensors, are used for the quantitative analysis of single analytes. One important example, for instance, is the Clark O2 electrode, which responds to the concentration of dissolved O2 in solutions such as blood and water.\n",
            "\n",
            "Suggested\n",
            "\n",
            "## EXPERIMENTS\n",
            "\n",
            "## Experiments\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "The following set of suggested experiments describes the preparation of solid-state and liquid ion-exchange ionselective electrodes, as well as potentiometric biosensors.\n",
            "\n",
            "Chan, W. H.; Wong, M. S.; Yip, C. W. 'Ion-Selective Electrode in Organic Analysis: A Salicylate Electrode,' J. Chem. Educ. 1986, 63, 915-916.\n",
            "\n",
            "The preparation of an ion-selective electrode for salicylate is described. The electrode incorporates an ion-pair of crystal violet and salicylate in a PVC matrix as the ion-selective membrane. Its use for the determination of acetylsalicylic acid in aspirin tablets is described. A similar experiment is described by Creager, S. E.; Lawrence, K. D.; Tibbets, C. R. in 'An Easily Constructed Salicylate-Ion-Selective Electrode for Use in the Instructional Laboratory,' J. Chem. Educ. 1995, 72, 274-276.\n",
            "\n",
            "Christopoulos, T. K.; Diamandis, E. P. 'Use of a Sintered Glass Crucible for Easy Construction of Liquid-Membrane Ion-Selective Electrodes,' J. Chem. Educ. 1988, 65, 648.\n",
            "\n",
            "Liquid-membrane ion-selective electrodes are prepared by removing the bottom below the glass filter, cleaning the filter, and making it hydrophobic using trimethylchlorosilane. Suitable liquid ion exchangers are added and allowed to penetrate the hydrophobic glass membrane. Adding an appropriate internal reference solution and reference electrode completes the ion-selective electrode.\n",
            "\n",
            "Fricke, G. H.; Kuntz, M. J. 'Inexpensive Solid-State IonSelective Electrodes for Student Use,' J. Chem. Educ. 1977, 54, 517-520.\n",
            "\n",
            "Instructions are provided in this experiment for the preparation of solid-state electrodes for the quantitative analysis of Ag + , Cl -, Cu 2+ , and Pb 2+ . All electrodes are based on AgCl, AgCl/Ag2S, CuS/Ag2S, and PbS/Ag2S preparations. Electrodes are prepared by (a) mixing the solid with silicon grease and coating on an Ag wire; (b) mixing the solid with silicone grease and packing into the end of a glass tube; or (c) preparing a pressed pellet.\n",
            "\n",
            "Lloyd, B. W.; O'Brien, F. L.; Wilson, W. D. 'Student Preparation and Analysis of Chloride and Calcium Ion Selective Electrodes,' J. Chem. Educ. 1976, 53, 328-330.\n",
            "\n",
            "This experiment describes the preparation of liquid ionexchange electrodes for Cl and Ca 2+ . The liquid ionexchange solutions are incorporated into PVC membranes and fixed to the end of glass tubing. The internal solutions are either NaCl or CaCl2, and a Ag/AgCl reference electrode is situated in the internal solution.\n",
            "\n",
            "Martínez-Fàbregas, E.; Alegret, S. 'A Practical Approach to Chemical Sensors through Potentiometric Transducers: Determination of Urea in Serum by Means of a Biosensor,' J. Chem. Educ. 1994, 71, A67-A70.\n",
            "\n",
            "Directions are provided for constructing and characterizing an ammonium ion-selective electrode. The electrode is then modified to respond to urea by adding a few milligrams of urease and covering with a section of dialysis membrane. Directions for determining urea in serum also are provided.\n",
            "\n",
            "## Experiments\n",
            "\n",
            "Continued from page 533\n",
            "\n",
            "Mifflin, T. E.; Andriano, K. M.; Robbins, W. B. 'Determination of Penicillin Using an Immobilized Enzyme Electrode,' J. Chem. Educ. 1984, 61, 638-639.\n",
            "\n",
            "Directions for preparing a potentiometric biosensor for penicillin are provided in this experiment. The enzyme penicillinase is immobilized in a polyacrylamide polymer formed on the surface of a glass pH electrode. The electrode shows a linear response to penicillin G over a concentration range of 10 -5 M to 10 -3 M.\n",
            "\n",
            "Palanivel, A.; Riyazuddin, P. 'Fabrication of an Inexpensive Ion-Selective Electrode,' J. Chem. Educ. 1984, 61, 290.\n",
            "\n",
            "This experiment describes the preparation of a Ag2S/CuS electrode using the graphite rod from a dry cell D-type battery. After obtaining the graphite rod, it is inserted and sealed in a polyethylene tube, leaving a 2-cm portion exposed. The protruding portion of the graphite rod, is placed in a saturated solution of Na2S for 1 h, and in a saturated solution of CuS for an additional hour. In a related experiment (Riyazuddin, P.; Devika, D. 'Potentiometric Acid-Base Titrations with Activated Graphite Electrodes,' J. Chem. Educ. 1997, 74, 1198-1199.), activated graphite electrodes are used for the potentiometric titration of HCl with NaOH.\n",
            "\n",
            "Ramaley, L; Wedge, P. J.; Crain, S. M. 'Inexpensive Instrumental Analysis: Part 1. Ion-Selective Electrodes,' J. Chem. Educ. 1994, 71, 164-167.\n",
            "\n",
            "This experiment describes the preparation and evaluation of two liquid-membrane Na + ion-selective electrodes, using either the sodium salt of monensin or a hemisodium ionophore as ion exchangers incorporated into a PVC matrix. Electrodes prepared using monensin performed poorly, but those prepared using hemisodium showed a linear response over a range of 0.1 M to 3 × 10 -5 M Na + with slopes close to the theoretical value.\n",
            "\n",
            "The following set of experiments illustrate several applications of potentiometric electrodes.\n",
            "\n",
            "Harris, T. M. 'Potentiometric Measurement in a Freshwater Aquarium,' J. Chem. Educ. 1993, 70, 340-341.\n",
            "\n",
            "This experiment describes a semester-long project in which the concentration of several ions in a fresh water aquarium are monitored. Ions that are monitored using potentiometric electrodes include H + (pH electrode), Cl (chloride electrode), HCO3 -(CO2 electrode), NH4 + (NH3 electrode), and NO3 - (NH3 electrode). Nitrate concentrations were determined following its conversion to ammonia.\n",
            "\n",
            "The titration of Al 3+ with F in a water-acetonitrile mixture is followed potentiometrically using a fluoride ion-selective electrode. The cumulative formation constant for AlF6 3is determined from the titration curve.\n",
            "\n",
            "Selig, W. S. 'Potentiometric Titrations Using Pencil and Graphite Sensors,' J. Chem. Educ. 1984, 61, 80-81.\n",
            "\n",
            "Radic, N.; Komijenovic, J. 'Potentiometric Determination of an Overall Formation Constant Using an Ion-Selective Membrane Electrode,' J. Chem. Educ. 1993, 70, 509-511.\n",
            "\n",
            "This experiment describes the use of coated graphite electrodes for the potentiometric monitoring of precipitation, acid-base, complexation, and redox titrations.\n",
            "\n",
            "The following four experiments describe useful and accessible applications of coulometry.\n",
            "\n",
            "Bertotti, M.; Vaz, J. M.; Telles, R. 'Ascorbic Acid Determination in Natural Orange Juice,' J. Chem. Educ. 1995, 72, 445-447.\n",
            "\n",
            "Lötz, A. 'A Variety of Electrochemical Methods in a Coulometric Titration Experiment,' J. Chem. Educ. 1998, 75, 775-777.\n",
            "\n",
            "The titration of ascorbic acid using coulometrically generated I 2 and Br2 is described in this experiment. Details are also given for the polarographic analysis of ascorbic acid.\n",
            "\n",
            "Kalbus, G. E.; Lieu, V. T. 'Dietary Fat and Health: An Experiment on the Determination of Iodine Number of Fats and Oils by Coulometric Titration,' J. Chem. Educ. 1991, 68, 64-65.\n",
            "\n",
            "The iodine number of fats and oils provides a quantitative measurement of the degree of unsaturation. A solution containing a 100% excess of ICl is added to the sample, reacting across the double-bonded sites of unsaturation. The excess ICl is converted to I2 by adding KI. The resulting I2 is reacted with a known excess of Na2S2O3. To complete the analysis the excess S2O3 2is back titrated with coulometrically generated I2.\n",
            "\n",
            "This experiment describes three coulometric titrations for the quantitative analysis of mixtures containing KI and HCl. Iodide is determined by titrating with coulometrically generated Br2. Chloride is determined indirectly by the coulometric titration of H + . Finally, iodide and chloride are determined together by precipitating AgI and AgCl following the anodic dissolution of an Ag wire.\n",
            "\n",
            "Swim, J.; Earps, E.; Reed, L. M.; et al. 'Constant-Current Coulometric Titration of Hydrochloric Acid,' J. Chem. Educ. 1996, 73, 679-683.\n",
            "\n",
            "Directions are provided for constructing an inexpensive constant-current source and demonstrate its use in determining the concentration of HCl coulometrically.\n",
            "\n",
            "-Continued\n",
            "\n",
            "## Experiments\n",
            "\n",
            "The final set of experiments provide examples of voltammetric and amperometric analyses.\n",
            "\n",
            "García-Armada, P.; Losada, J.; de Vicente-Pérez, S. 'Cation Analysis Scheme by Differential Pulse Polarography,' J. Chem. Educ. 1996, 73, 544-547.\n",
            "\n",
            "unknowns, serum, or plasma can be determined using a normal calibration curve.\n",
            "\n",
            "Directions are provided for the simultaneous qualitative and quantitative analysis of Bi(III), Cd(II), Co(II), Cu(II), Cr(III), Mn(II), Ni(II), Pb(II), Sb(III), and Zn(II). The analysis is accomplished using differential pulse polarography in three separate supporting electrolytes.\n",
            "\n",
            "Herrera-Melián, J. A.; Doña-Rodríguez, J. M.; HernándezBrito, J.; et al. 'Voltammetric Determination of Ni and Co in Water Samples,' J. Chem. Educ. 1997, 74, 1444-1445.\n",
            "\n",
            "This experiment describes the application of cathodic stripping voltammetry to the analysis of Ni and Co in fresh water and sea water.\n",
            "\n",
            "Marin, D.; Mendicuti, F. 'Polarographic Determination of Composition and Thermodynamic Stability Constant of a Complex Metal Ion,' J. Chem. Educ. 1988, 65, 916-918.\n",
            "\n",
            "This experiment describes the determination of the stability (cumulative formation) constant for the formation of Pb(OH)3 - by measuring the shift in the half-wave potential for the reduction of Pb 2+ as a function of the concentration of OH - . The influence of ionic strength is also considered, and results are extrapolated to zero ionic strength to determine the thermodynamic formation constant.\n",
            "\n",
            "Sittampalam, G.; Wilson, G. S. 'Amperometric Determination of Glucose at Parts Per Million Levels with Immobilized Glucose Oxidase,' J. Chem. Educ. 1982, 59, 70-73.\n",
            "\n",
            "This experiment describes the use of a commercially available amperometric biosensor for glucose that utilizes the enzyme glucose oxidase. The concentration of glucose in artificial\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## 11H PROBLEMS\n",
            "\n",
            "- 1. Identify the anode and cathode for the following electrochemical cells, and write the oxidation or reduction reaction occurring at each electrode. (a) Pt | FeCl2 (0.015 M), FeCl3 (0.045 M) || AgNO3 (0.1 M) | Ag (b) Ag | AgBr(s) | NaBr (1.0 M) || CdCl2 (0.05 M) | Cd (c) Pb, PbSO4 | H2SO4 (1.5 M) || H2SO4 (2.0 M) | PbSO4, PbO2\n",
            "- 2. Calculate the potential for the electrochemical cells in problem 1.\n",
            "- 3. Calculate the molar concentration for the underlined component in the following cell if the cell potential is measured at +0.294 V\n",
            "\n",
            "Ag | AgCl (sat'd) | NaCl (0.1 M) || KI ( x M) | I2(s) | Pt\n",
            "\n",
            "Town, J. L.; MacLaren, F.; Dewald, H. D. 'Rotating Disk Voltammetry Experiment,' J. Chem. Educ. 1991, 68, 352-354.\n",
            "\n",
            "This experiment introduces hydrodynamic voltammetry using a rotating working electrode. Its application for the quantitative analysis of K4Fe(CN)6 is demonstrated.\n",
            "\n",
            "Wang, J. 'Sensitive Electroanalysis Using Solid Electrodes,' J. Chem. Educ. 1982, 59, 691-692.\n",
            "\n",
            "The technique of hydrodynamic modulation voltammetry (HMV), in which the rate of stirring is pulsed between high and low values, is demonstrated in this experiment. The application of HMV for the quantitative analysis of ascorbic acid in vitamin C tablets using the method of standard additions also is outlined.\n",
            "\n",
            "Wang, J. 'Anodic Stripping Voltammetry,' J. Chem. Educ. 1983, 60, 1074-1075.\n",
            "\n",
            "The application of anodic stripping voltammetry for the quantitative analysis of Cd, Pb, and Cu in natural waters is described in this experiment.\n",
            "\n",
            "Wang, J.; Maccà, C. 'Use of Blood-Glucose Test Strips for Introducing Enzyme Electrodes and Modern Biosensors,' J. Chem. Educ. 1996, 73, 797-800.\n",
            "\n",
            "Commercially available kits for monitoring blood-glucose use an amperometric biosensor incorporating the enzyme glucose oxidase. This experiment describes how such monitors can be adapted to the quantitative analysis of glucose in beverages.\n",
            "\n",
            "- 4. What reaction prevents Zn from being used as an electrode of the first kind in acidic solution? Which other electrodes of the first kind would you expect to behave in the same manner as Zn when immersed in acidic solutions?\n",
            "- 5. Creager and colleagues designed a salicylate ion-selective electrode using a PVC membrane impregnated with tetraalkylammonium salicylate. 27 To determine the ionselective electrode's selectivity coefficient for benzoate, a set of salicylate calibration standards was prepared in which the concentration of benzoate was held constant at 0.10 M. Using the following data, determine the value of the selectivity coefficient.\n",
            "\n",
            "| [Salicylate] (M)   |   Potential (mV) |\n",
            "|--------------------|------------------|\n",
            "| 1.0                |             20.2 |\n",
            "| 1.0 × 10 -1        |             73.5 |\n",
            "| 1.0 × 10 -2        |            126   |\n",
            "| 1.0 × 10 -3        |            168   |\n",
            "| 1.0 × 10 -4        |            182   |\n",
            "| 1.0 × 10 -5        |            182   |\n",
            "| 1.0 × 10 -6        |            177   |\n",
            "\n",
            "What is the maximum acceptable concentration of benzoate if this ion-selective electrode is to be used for analyzing samples containing as little as 10 -5 M salicylate with an accuracy of better than 1%?\n",
            "\n",
            "- 6. Watanabe and co-workers described a new membrane electrode for the determination of cocaine, which is a weak base alkaloid with a p K a of 8.64. 28 The response of the electrode for a fixed concentration of cocaine was found to be independent of pH in the range of 1-8, but decreased sharply above a pH of 8. Offer an explanation for the source of this pH dependency.\n",
            "- 7. Show that equation 11.12 for the urea electrode is correct.\n",
            "- 8. The pH-based urea electrode has the response shown in equation 11.13. Rewrite the equation to show the relationship between potential, E, and the concentration of urea. Comment on your result.\n",
            "- 9. Mifflin and associates described a membrane electrode for the quantitative analysis of penicillin in which the enzyme penicillinase is immobilized in a polyacrylamide gel that is coated on a glass pH electrode. 29 The following data were collected for a series of penicillin standards.\n",
            "\n",
            "| [Penicillin] (M)   |   Potential (mV) |\n",
            "|--------------------|------------------|\n",
            "| 1.0 × 10 -2        |              220 |\n",
            "| 2.0 × 10 -3        |              204 |\n",
            "| 1.0 × 10 -3        |              190 |\n",
            "| 2.0 × 10 -4        |              153 |\n",
            "| 1.0 × 10 -4        |              135 |\n",
            "| 1.0 × 10 -5        |               96 |\n",
            "| 1.0 × 10 -6        |               80 |\n",
            "\n",
            "Construct a calibration curve for the electrode, and report (a) the range of concentrations in which a linear response is observed, (b) the equation for the calibration curve in this range, and (c) the concentration of penicillin in a sample that yields a potential of 142 mV.\n",
            "\n",
            "- 10. Ion-selective electrodes can be incorporated in flow cells to monitor the concentration of an analyte in standards and samples that are pumped through the flow cell. As the analyte passes through the cell, a potential spike is recorded instead of a steady-state potential. The concentration of K + in serum has been determined in this fashion, using standards prepared in a matrix of 0.014 M NaCl. 30\n",
            "\n",
            "|   [K + ] (mM) |   Potential (arb. units) |\n",
            "|---------------|--------------------------|\n",
            "|           0.1 |                     25.5 |\n",
            "|           0.2 |                     37.2 |\n",
            "|           0.4 |                     50.8 |\n",
            "|           0.6 |                     58.7 |\n",
            "|           0.8 |                     64   |\n",
            "|           1   |                     66.8 |\n",
            "\n",
            "A 1.00-mL sample of serum is diluted to volume in a 10-mL volumetric flask and analyzed, giving a potential of 51.1. Report the concentration of K + in the sample of serum.\n",
            "\n",
            "- 11. Wang and Taha described an interesting application of potentiometry called batch injection. 31 As shown in the following figure, an ion-selective electrode is placed in an inverted position in a large-volume tank, and a fixed volume of a sample or standard solution is injected toward the electrode's surface using a micropipet.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "The response of the electrode is a peak whose height is proportional to the analyte's concentration. The following data were collected using a pH electrode\n",
            "\n",
            "|   pH |   Potential (mV) |\n",
            "|------|------------------|\n",
            "|    2 |             +300 |\n",
            "|    3 |             +240 |\n",
            "|    4 |             +168 |\n",
            "|    5 |              +81 |\n",
            "|    6 |              +35 |\n",
            "|    8 |              -92 |\n",
            "|    9 |             -168 |\n",
            "|   10 |             -235 |\n",
            "|   11 |             -279 |\n",
            "\n",
            "Determine the pH for each of the following samples, given the recorded peak potential: (a) tomato juice, +167 mV; (b) tap water, -27 mV; (c) coffee, 122 mV.\n",
            "\n",
            "- 12. The concentration of NO3 -in a water sample is determined by a one-point standard addition using an NO3 -ion-selective electrode. A 25.00-mL sample is placed in a beaker, and a potential of +0.102 V is measured. A 1.00-mL aliquot of a 200.0 ppm standard solution of NO3 -is added, after which the potential is found to be +0.089 V. Report the concentration of NO3 -in parts per million.\n",
            "- 13. The following data were collected for the analysis of fluoride in tap water and in toothpaste. (a) For the analysis of tap water, three 25.0-mL samples were each mixed with 25.0 mL of TISAB, and the potential was measured with an F ISE relative to a saturated calomel electrode. Five 1.00-mL additions of a standard solution of 100.0-ppm F were added to each, measuring the potential following each addition.\n",
            "\n",
            "| Milliliters of F - added   | Potential (mV)   | Potential (mV)   | Potential (mV)   |\n",
            "|----------------------------|------------------|------------------|------------------|\n",
            "| standard                   | Trial 1          | Trial 2          | Trial 3          |\n",
            "| 0.00                       | -79              | -82              | -81              |\n",
            "| 1.00                       | -119             | -119             | -118             |\n",
            "| 2.00                       | -133             | -133             | -133             |\n",
            "| 3.00                       | -142             | -142             | -142             |\n",
            "| 4.00                       | -149             | -148             | -148             |\n",
            "| 5.00                       | -154             | -153             | -153             |\n",
            "\n",
            "Determine the parts per million of F in the tap water. (b) For the analysis of toothpaste a 0.3619-g sample was transferred to a 100-mL volumetric flask along with 50.0 mL of TISAB and diluted to volume with distilled water. Three 20.0-mL aliquots were removed, and the potential was measured with an F ion-selective electrode using a saturated calomel electrode as a reference. Five separate 1.00-mL additions of a 100.0-ppm solution of F were added to each, measuring the potential following each addition.\n",
            "\n",
            "| Milliliters of standard F - added   | Potential (mV)   | Potential (mV)   | Potential (mV)   |\n",
            "|-------------------------------------|------------------|------------------|------------------|\n",
            "| 0.00                                | Trial 1          | Trial 2          | Trial 3          |\n",
            "|                                     | -55              | -54              | -55              |\n",
            "| 1.00                                | -82              | -82              | -83              |\n",
            "| 2.00                                | -94              | -94              | -94              |\n",
            "| 3.00                                | -102             | -103             | -102             |\n",
            "| 4.00                                | -108             | -108             | -109             |\n",
            "| 5.00                                | -112             | -112             | -113             |\n",
            "\n",
            "Report the parts per million of F in the sample of toothpaste.\n",
            "\n",
            "- 14. You are responsible for determining the amount of KI in iodized salt and decide to use an I ion-selective electrode. Describe how you would perform this analysis using (a) external standards and (b) the method of standard additions.\n",
            "- 15. The purity of a sample of picric acid, C6H3N3O7, is determined by controlled-potential coulometry, converting the picric acid to triaminophenol, C6H9N3O. A 0.2917-g sample of picric acid is placed in a 1000-mL volumetric flask\n",
            "\n",
            "and diluted to volume. A 10.00-mL portion of this solution is transferred to a coulometric cell and diluted till the Pt cathode is immersed. The exhaustive electrolysis of the sample requires 21.67 C of charge. Report the purity of the picric acid.\n",
            "\n",
            "- 16. The concentration of H2S in the drainage from an abandoned mine can be determined by a coulometric titration using KI as a mediator and I3 -as the 'titrant.'\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "A 50.00-mL sample of water is placed in a coulometric cell, along with an excess of KI and a small amount of starch as an indicator. Electrolysis is carried out at a constant current of 84.6 mA, requiring 386 s to reach the starch end point. Report the concentration of H2S in the sample in parts per million.\n",
            "\n",
            "- 17. One method for the determination of H3AsO3 is by a coulometric titration using I3 -as a 'titrant.' The relevant reactions and standard-state potentials are summarized as follows.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Explain why the coulometric titration must be done in neutral solutions (pH ≈ 7), instead of in strongly acidic solutions (pH &lt; 0).\n",
            "\n",
            "- 18. The reduction of acrylonitrile, CH2=CHCN, to adiponitrile, NC(CH2)4CN, is an important industrial process. A 0.594-g sample of acrylonitrile is placed in a 1-L volumetric flask and diluted to volume. An exhaustive controlled-potential electrolysis of a 1.00-mL portion of the diluted acrylonitrile requires 1.080 C of charge. What is the value of n for this reduction?\n",
            "- 19. A linear-potential scan hydrodynamic voltammogram for a mixture of Fe 2+ and Fe 3+ is shown in the figure, where i l,a and\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "i l,c are the anodic and cathodic limiting currents. (a) Show that the potential is given by\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "- (b) What is the potential when i = 0 for a solution that is 0.1 mM Fe 3+ and 0.05 mM Fe 2+ ? You may assume that K Fe 3+ ≈ K Fe 2+ .\n",
            "\n",
            "- 20. The amount of sulfur in aromatic monomers can be determined by differential pulse polarography. Standard solutions are prepared for analysis by dissolving 1.000 mL of the purified monomer in 25.00 mL of an electrolytic solvent, adding a known amount of S, deaerating, and measuring the peak current. The following results were obtained for a set of calibration standards\n",
            "\n",
            "|   Micrograms of added S |   Peak current ( µ A) |\n",
            "|-------------------------|-----------------------|\n",
            "|                       0 |                  0.14 |\n",
            "|                      28 |                  0.7  |\n",
            "|                      56 |                  1.23 |\n",
            "|                     112 |                  2.41 |\n",
            "|                     168 |                  3.42 |\n",
            "\n",
            "Analysis of a 1.000-mL sample, treated in the same manner as the standards, gives a peak current of 1.77 µ A. Report the amount of sulfur present in the sample in milligrams per milliliter.\n",
            "\n",
            "- 21. The purity of a sample of K3Fe(CN)6 was determined using linear-potential scan hydrodynamic voltammetry at a glassy carbon electrode using the method of external standards. The following data were obtained for a set of calibration standards.\n",
            "\n",
            "|   [K 3 Fe(CN) 6 ] (mM) |   Limiting current ( µ A) |\n",
            "|------------------------|---------------------------|\n",
            "|                      2 |                       127 |\n",
            "|                      4 |                       252 |\n",
            "|                      6 |                       376 |\n",
            "|                      8 |                       500 |\n",
            "|                     10 |                       624 |\n",
            "\n",
            "A sample was prepared for analysis by diluting a 0.246-g sample to volume in a 100-mL volumetric flask. The limiting current for the sample was found to be 444 µ A. Report the purity of this sample of K3Fe(CN)6.\n",
            "\n",
            "- 22. Anodic stripping voltammetry at a mercury film electrode can be used to determine whether an individual has recently fired a gun by looking for traces of antimony in residue collected from the individual's hands. In a typical analysis a sample is collected with a cotton-tipped swab that had been wetted with 5% v/v HNO3. When returned to the lab, the swab is placed in a vial containing 5.00 mL of 4 M HCl that is 0.02 M in hydrazine sulfate. After allowing the swab to soak overnight, 4.00 mL of the solution is transferred to an electrochemical cell, and 100 µ L of 1.00 × 10 -2 M HgCl2 is added. Anodic stripping voltammetry of the sample gives a peak current of 0.38 µ A. After a standard addition of 100 µ L of 5.00 × 10 -2 ppb Sb was added, anodic stripping voltammetry is repeated, giving a peak current of 1.14 µ A. How many nanograms of Sb is collected from the individual's hand?\n",
            "- 23. Zinc can be used as an internal standard in the analysis of thallium by differential pulse polarography. A standard\n",
            "- solution containing 5.00 × 10 -5 M Zn 2+ and 2.50 × 10 -5 M Tl + gave peak currents of 5.71 µ A and 3.19 µ A, respectively. An 8.713-g sample of an alloy known to be free of zinc was dissolved in acid, transferred to a 500-mL volumetric flask, and diluted to volume. A 25.0-mL portion of this solution was mixed with 25.0 mL of a 5.00 × 10 -4 M solution of Zn 2+ . Analysis of this solution gave a peak current for Zn 2+ of 12.3 µ A, and for Tl + of 20.2 µ A. Report the Tl% w/w in the alloy.\n",
            "- 24. Differential pulse voltammetry at a carbon working electrode can be used to determine the concentrations of ascorbic acid and caffeine in drug formulations. 32 In a typical analysis, a 0.9183-g tablet is crushed and ground into a fine powder. A 0.5630-g sample of this powder is transferred to a 100-mL volumetric flask, brought into solution, and diluted to volume. A 0.500-mL portion is then transferred to a voltammetric cell containing 20.00 mL of a suitable supporting electrolyte. The resulting voltammogram gives peak currents of 1.40 µ A and 3.88 µ A for ascorbic acid and caffeine, respectively. A 0.500-mL aliquot of a standard solution containing 250.0 ppm ascorbic acid and 200.0 ppm caffeine is then added. A voltammogram of this solution gives peak currents of 2.80 µ A and 8.02 µ A for ascorbic acid and caffeine, respectively. Report the number of milligrams of ascorbic acid and of caffeine in the tablet.\n",
            "- 25. Ratana-ohpas and co-workers describe a stripping analysis method for determining levels of tin in canned fruit juices. 33 Standards containing 50.0 ppb Sn 4+ , 100.0 ppb Sn 4+ , and 150.0 ppb Sn 4+ were analyzed, giving signals of 83.0, 171.6, and 260.2, respectively. A 2.00-mL sample of lychee juice was mixed with 20.00 mL of 1:1 HCl/HNO3. A 0.500-mL portion of this mixture was added to 10 mL of 6 M HCl and the volume adjusted to 30.00 mL. Analysis of this diluted sample gave a signal of 128.2. Report the parts per million of Sn 4+ in the original sample of lychee juice.\n",
            "- 26. Sittampalam and Wilson described the preparation and use of an amperometric sensor for glucose. 34 The sensor is calibrated by measuring the steady-state current when it is immersed in standard solutions of glucose. A typical set of calibration data is shown in the following table.\n",
            "\n",
            "|   Milligrams of glucose/100 mL |   Current (arb. units) |\n",
            "|--------------------------------|------------------------|\n",
            "|                              2 |                   17.2 |\n",
            "|                              4 |                   32.9 |\n",
            "|                              6 |                   52.1 |\n",
            "|                              8 |                   68   |\n",
            "|                             10 |                   85.8 |\n",
            "\n",
            "A 2.00-mL sample of a solution containing an unknown amount of glucose is diluted to 10 mL in a volumetric flask, and a steady-state current of 23.6 is measured. What is the concentration of glucose in the sample in milligrams per 100 mL?\n",
            "\n",
            "- 27. Differential pulse polarography is used to determine the concentrations of lead, thallium, and indium in a mixture. 20\n",
            "\n",
            "Since the polarographic peaks for lead and thallium, and for thallium and indium overlap, a simultaneous analysis is necessary. Peak currents (in arbitrary units) at potentials of -0.385 V, -0.455 V, and -0.557 V were measured for a single standard solution and for a sample, giving the results shown in the following table.\n",
            "\n",
            "| Analyte   | Concentration ( µ g/mL)   |   i peak at -0.385 V |   i peak at -0.455V |   i peak at -0.557 V |\n",
            "|-----------|---------------------------|----------------------|---------------------|----------------------|\n",
            "| Pb 2+     | 1.0                       |                 26.1 |                 2.9 |                  0   |\n",
            "| Tl +      | 2.0                       |                  7.8 |                23.5 |                  3.2 |\n",
            "| In 3+     | 0.4                       |                  0   |                 0   |                 22.9 |\n",
            "| sample    |                           |                 60.6 |                28.8 |                 54.1 |\n",
            "\n",
            "Report the concentrations of Pb 2+ , Tl + , and In 3+ in the sample.\n",
            "\n",
            "28. Abass and colleagues developed an amperometric biosensor for NH4 + that uses the enzyme glutamate dehydrogenase to catalyze the following reaction.\n",
            "\n",
            "2-oxyglutarate( aq ) + NH4 + ( aq ) + NADH( aq ) → glutamate( aq ) + NAD + ( aq ) + H2O( l )\n",
            "\n",
            "where NADH is the reduced form of nicotinamide adenine dinucleotide. 35 The biosensor actually responds to the concentration of NADH; however, the rate of the reaction depends on the concentration of NH4 + . If the initial concentrations of 2-oxyglutarate and NADH are the same for all samples and standards, then the signal is proportional to the concentration of NH4 + . As shown in the following table, the sensitivity of the method is dependent on pH.\n",
            "\n",
            "|   pH | Sensitivity (nA s -1 M -1 )   |\n",
            "|------|-------------------------------|\n",
            "| 6.2  | 1.67 × 10 3                   |\n",
            "| 6.75 | 5.00 × 10 3                   |\n",
            "| 7.3  | 9.33 × 10 3                   |\n",
            "| 7.7  | 1.03 × 10 4                   |\n",
            "| 8.3  | 1.27 × 10 4                   |\n",
            "| 9.3  | 2.67 × 10 3                   |\n",
            "\n",
            "Two possible explanations for the effect of pH on the sensitivity of this analysis are the acid-base chemistry of NH4 + , and, the acid-base chemistry of the enzyme. Given that the p K a for NH4 + is 9.244, explain the source of this pHdependent sensitivity.\n",
            "\n",
            "- 29. The speciation scheme of Batley and Florence 22 requires eight measurements on four samples. After removing insoluble particulates by filtration, the solution is analyzed for the concentration of anodic stripping voltammetry (ASV) labile metal and the total concentration of metal. A portion of the filtered solution is passed through an ion-exchange column, and the concentrations of ASV metal and total metal are determined. A second portion of the filtered solution is irradiated with UV light, and the concentrations of ASV metal\n",
            "\n",
            "and total metal are measured. Finally, a third portion of the filtered solution is irradiated with UV light, passed through an ion-exchange column, and the concentrations of ASV-labile metal and total metal are again determined. The groups that are included in each measurement are summarized in the following table (see Table 11.12).\n",
            "\n",
            "| Separation Step                 | Groups Remaining in Solution   | Groups Contributing to Concentration of ASV- Labile Metal   | Groups Contributing to Concentration of Total Metal   |\n",
            "|---------------------------------|--------------------------------|-------------------------------------------------------------|-------------------------------------------------------|\n",
            "| 1. filtration                   | I, II, III, IV, V, VI, VII     | I, II, III                                                  | I, II, III, IV, V, VI, VII                            |\n",
            "| 2. ion- exchange                | II, III, VI, VII               | II, III                                                     | II, III, VI, VII                                      |\n",
            "| 3. UV irradiation               | I, II, III, IV, V, VI, VII     | I, II, III, IV, VI                                          | I, II, III, IV, V, VI, VII                            |\n",
            "| 4. UV irradiation/ ion-exchange | III, VII                       | III                                                         | III, VII                                              |\n",
            "\n",
            "(a) Show how these eight measurements can be used to determine the concentration of metal present in each group.\n",
            "\n",
            "(b) Batley and Florence 22a report the following results for the speciation of cadmium, lead, and copper in a sample of sea water\n",
            "\n",
            "| Measurement              |   ppb Cd 2+ |   ppb Pb 2+ |   ppb Cu 2+ |\n",
            "|--------------------------|-------------|-------------|-------------|\n",
            "| filtration: ASV-labile   |        0.24 |        0.39 |        0.26 |\n",
            "| filtration: total        |        0.28 |        0.5  |        0.4  |\n",
            "| ion-exchange: ASV-labile |        0.21 |        0.33 |        0.17 |\n",
            "| ion-exchange: total      |        0.26 |        0.43 |        0.24 |\n",
            "| UV: ASV-labile           |        0.26 |        0.37 |        0.33 |\n",
            "| UV: total                |        0.28 |        0.5  |        0.43 |\n",
            "| ion-exchange/UV: labile  |        0    |        0    |        0    |\n",
            "| ion-exchange/UV: total   |        0.02 |        0.12 |        0.1  |\n",
            "\n",
            "Determine the speciation of each metal in this sample of sea water.\n",
            "\n",
            "30. The polarographic half-wave potentials (versus the SCE) for Pb 2+ and Tl + in 1 M HCl are, respectively, -0.44 V and -0.45 V. In an electrolyte of 1 M NaOH, however, the half-wave potentials are -0.76 V for Pb 2+ and -0.48 V for Tl + . Why does the change in electrolyte have such a great effect on the half-wave potential for Pb 2+ , but not on the half-wave potential for Tl + ?\n",
            "\n",
            "- 31. The following data were collected for the reduction of Pb 2+ by normal pulse polarography\n",
            "\n",
            "|   Potential (V vs. SCE) |   Current ( µ A) |\n",
            "|-------------------------|------------------|\n",
            "|                  -0.345 |             0.16 |\n",
            "|                  -0.37  |             0.98 |\n",
            "|                  -0.383 |             2.05 |\n",
            "|                  -0.393 |             3.13 |\n",
            "|                  -0.409 |             4.62 |\n",
            "\n",
            "-0.420\n",
            "\n",
            "5.16\n",
            "\n",
            "## Modern Analytical Chemistry\n",
            "\n",
            "The limiting current was 5.67 µ A. Verify that the reduction reaction is reversible, and determine values for n and E 1/2 . The half-wave potentials for the normal pulse polarograms of Pb 2+ in the presence of several different concentrations of OH are shown in the following table.\n",
            "\n",
            "|   [OH - ] (M) |   E 1/2 (V vs. SCE) |\n",
            "|---------------|---------------------|\n",
            "|          0.05 |              -0.646 |\n",
            "|          0.1  |              -0.673 |\n",
            "|          0.15 |              -0.689 |\n",
            "|          0.3  |              -0.715 |\n",
            "\n",
            "Determine the stoichiometry of the Pb-hydroxide complex and its formation constant.\n",
            "\n",
            "- 32. The concentration of Cu 2+ in sea water may be determined by anodic stripping voltammetry at a hanging mercury drop electrode. To determine the total amount of Cu 2+ in a sample of sea water, it is necessary to free any copper bound in organic complexes. To a 20.00-mL sample of sea water is added 1 mL of 0.05 M HNO3 and 1 mL of 0.1% H2O2. The sample is irradiated with UV light for 8 h and then diluted to volume in a 25-mL volumetric flask. Deposition of Cu 2+ takes place at -0.3 V for 10 min, producing a peak current of 26.1 (arbitrary units). A second 20.00-mL sample of the sea water is treated identically, except that 0.1 mL of a 5.00 µ M solution of Cu 2+ is added, producing a peak current of 38.4. Report the concentration of Cu 2+ in the sea water in parts per billion.\n",
            "- 33. Thioamide drugs, such as that shown here, can be determined by cathodic stripping analysis. Deposition occurs at +0.05 V\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## 11I SUGGESTED READINGS\n",
            "\n",
            "The following general references provide a broad introduction to electrochemical methods of analysis.\n",
            "\n",
            "Bard, A. J.; Faulkner, L. R. Electrochemical Methods. Wiley: New York, 1980.\n",
            "\n",
            "Faulkner, L. R. 'Electrochemical Characterization of Chemical Systems.' In Kuwana, T. E., ed., Physical Methods in Modern Chemical Analysis, Vol. 3. Academic Press: New York, 1983, pp. 137-248.\n",
            "\n",
            "Kissinger, P. T.; Heineman, W. R. Laboratory Techniques in Electroanalytical Chemistry. 2nd ed. Marcel Dekker: New York, 1996.\n",
            "\n",
            "Lingane, J. J. Electroanalytical Chemistry. 2nd ed. Interscience: New York, 1958.\n",
            "\n",
            "versus the SCE. During the stripping step, the potential is scanned cathodically, and a stripping peak is observed at -0.55 V. In a typical application a 2.00-mL sample of urine was mixed with 2.00 mL of a pH 4.78 buffer. Following a 2-minute deposition, a peak current of 0.562 µ A was measured. A 0.10-mL addition of a 5.00 µ M solution of the thioamide drug was added to the same solution. A peak current of 0.837 µ A was recorded following the same 2-min deposition time. Report the drug's molar concentration in the urine sample.\n",
            "\n",
            "- 34. The concentration of vanadium (V) in sea water can be determined by adsorptive stripping voltammetry after forming a complex with catechol. 36 The catechol-V(V) complex is deposited on a hanging mercury drop electrode at a potential of -0.1 V (vs. Ag/AgCl). A cathodic potential scan gives a stripping peak that is proportional to the concentration of V(V) in the sample of sea water. The following standard additions were used to analyze a sample of sea water.\n",
            "\n",
            "| [V (V)] add (M)   |   i peak (nA) |\n",
            "|-------------------|---------------|\n",
            "| 2.0 × 10 -8       |            24 |\n",
            "| 4.0 × 10 -8       |            33 |\n",
            "| 8.0 × 10 -8       |            52 |\n",
            "| 1.2 × 10 -7       |            69 |\n",
            "| 1.8 × 10 -7       |            97 |\n",
            "| 2.8 × 10 -7       |           140 |\n",
            "\n",
            "Determine the molar concentration of V (V) in the sample of sea water, assuming that the standard additions result in a negligible change in the sample's volume.\n",
            "\n",
            "Sawyer, D. T.; Roberts, J. L., Jr. Experimental Electrochemistry for Chemists. Wiley-Interscience: New York, 1974.\n",
            "\n",
            "Vassos, B. H.; Ewing, G. W. Electroanalytical Chemistry. Wiley- Interscience: New York, 1983.\n",
            "\n",
            "Three short articles providing a good understanding of important general principles of electrochemistry follow.\n",
            "\n",
            "Faulkner, L. R. 'Understanding Electrochemistry: Some Distinctive Concepts,' J. Chem. Educ. 1983, 60, 262-264.\n",
            "\n",
            "Maloy, J. T. 'Factors Affecting the Shape of Current-Potential Curves,' J. Chem. Educ. 1983, 60, 285-289.\n",
            "\n",
            "Zoski, C. G. 'Charging Current Discrimination in Analytical Voltammetry,' J. Chem. Educ. 1986, 63, 910-914.\n",
            "\n",
            "Additional information on potentiometry and ion-selective electrodes can be found in the following sources.\n",
            "\n",
            "Bates, R. G. Determination of pH: Theory and Practice. 2nd ed. Wiley: New York, 1973.\n",
            "\n",
            "Buck, R. P. 'Potentiometry: pH Measurements and Ion Selective Electrodes.' In Weissberger, A., ed., Physical Methods of Organic Chemistry, Vol. 1, Part IIA. Wiley: New York, 1971, pp. 61-162.\n",
            "\n",
            "Cammann, K. Working with Ion-Selective Electrodes. Springer-Verlag: Berlin, 1977.\n",
            "\n",
            "Evans, A. Potentiometry and Ion-Selective Electrodes. Wiley: New York, 1987.\n",
            "\n",
            "Frant, M. S. 'Where Did Ion Selective Electrodes Come From?' J. Chem. Educ. 1997, 74, 159-166.\n",
            "\n",
            "Light, T. S. 'Industrial Use and Application of Ion-Selective Electrodes,' J. Chem. Educ. 1997, 74, 171-177.\n",
            "\n",
            "Rechnitz, G. A. 'Ion and Bio-Selective Membrane Electrodes,' J. Chem. Educ. 1983, 60, 282-284.\n",
            "\n",
            "Ruzicka, J. 'The Seventies-Golden Age for Ion-Selective Electrodes,' J. Chem. Educ. 1997, 74, 167-170.\n",
            "\n",
            "- Young, C. C. 'Evolution of Blood Chemistry Analyzers Based on Ion Selective Electrodes,' J. Chem. Educ. 1997, 74, 177-182.\n",
            "\n",
            "The following sources provide additional information on electrochemical biosensors.\n",
            "\n",
            "Alvarez-Icasa, M.; Bilitewski, U. 'Mass Production of Biosensors,' Anal. Chem. 1993, 65, 525A-533A.\n",
            "\n",
            "Meyerhoff, M. E.; Fu, B.; Bakker, E.; et al. 'Polyion-Sensitive Membrane Electrodes for Biomedical Analysis,' Anal. Chem. 1996, 68, 168A-175A.\n",
            "\n",
            "Nicolini, C.; Adami, M; Antolini, F.; et al. 'Biosensors: A Step to Bioelectronics,' Phys. World, May 1992, 30-34.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## REFERENCES\n",
            "\n",
            "- 1. Stork, J. T. Anal. Chem. 1993, 65, 344A-351A.\n",
            "- 2. Sawyer, D. T.; Roberts, J. L., Jr. Experimental Electrochemistry for Chemists. Wiley-Interscience: New York, 1974, p. 22.\n",
            "- 3. Cammann, K. Working with Ion-Selective Electrodes. Springer-Verlag: Berlin, 1977.\n",
            "- 4. (a) Papastathopoulos, D. S.; Rechnitz, G. A. Anal. Chim. Acta 1975, 79, 17; (b) Riechel, T. L. J. Chem. Educ. 1984, 61, 640-642.\n",
            "- 5. Tor, R.; Freeman, A. Anal. Chem. 1986, 58, 1042-1046.\n",
            "- 6. Lunte, C. E.; Heineman, W. R. 'Electrochemical Techniques in Bioanalysis' in Steckhan, E., ed. Topics in Current Chemistry. Vol. 143. Springer-Verlag: Berlin, 1988, p. 8.\n",
            "- 7. Procedure adapted from Kennedy, J. H. Analytical ChemistryPractice. Harcourt Brace Jovanovich: San Diego, 1984, p. 117-118.\n",
            "- 8. Hawkes, S. J. J. Chem. Educ. 1994, 71, 747-749.\n",
            "- 9. Kristensen, H. B.; Saloman, A.; Kokholm, G. Anal. Chem. 1991, 63, 885A-891A.\n",
            "- 10. Bates, R. G. Determination of pH: Theory and Practice. 2nd ed. Wiley: New York, 1973.\n",
            "- 11. Rogers, K. R.; Williams, L. R. Trends Anal. Chem. 1995, 14, 289-294.\n",
            "- 12. Rechnitz, G. A. Controlled-Potential Analysis. Macmillan: New York, 1963, p. 49.\n",
            "\n",
            "## Chapter 11 Electrochemical Methods of Analysis\n",
            "\n",
            "Schultz, J. S. 'Biosensors,' Sci. Am. August 1991, 64-69.\n",
            "\n",
            "Thompson, M.; Krull, U. 'Biosensors and the Transduction of Molecular Recognition,' Anal. Chem. 1991, 63, 393A-405A.\n",
            "\n",
            "249-252.\n",
            "\n",
            "Vadgama, P. 'Designing Biosensors,' Chem. Brit. 1992, 28, The following is a good source covering the clinical application of electrochemistry.\n",
            "\n",
            "Wang, J. Electroanalytical Techniques in Clinical Chemistry and Laboratory Medicine. VCH: New York, 1998.\n",
            "\n",
            "Coulometry is covered in the following texts.\n",
            "\n",
            "Milner, G. W. C.; Phillips, G. Coulometry in Analytical Chemistry. Pergamon: New York, 1967.\n",
            "\n",
            "Rechnitz, G. A. Controlled-Potential Analysis. Macmillan: New York, 1963.\n",
            "\n",
            "For a description of electrogravimetry, see the following resource.\n",
            "\n",
            "Tanaka, N. 'Electrodeposition,' In Kolthoff, I. M.; Elving, P. J., eds. Treatise on Analytical Chemistry, Part I: Theory and Practice, Vol. 4. Interscience: New York, 1963.\n",
            "\n",
            "The following sources provide additional information on polarography and pulse polarography.\n",
            "\n",
            "Flato, J. B. 'The Renaissance in Polarographic and Voltammetric Analysis,' Anal. Chem. 1972, 44(11), 75A-87A.\n",
            "\n",
            "- Kolthoff, I. M.; Lingane, J. J. Polarography. Interscience: New York, 1952.\n",
            "\n",
            "Osteryoung, J. 'Pulse Voltammetry,' J. Chem. Educ. 1983, 60, 296-298. Additional information on stripping voltammetry is available in the following text.\n",
            "\n",
            "Wang, J. Stripping Analysis. VCH Publishers: Deerfield Beach, FL, 1985, pp. 116-124.\n",
            "\n",
            "- 13. Procedure adapted from Bassett, J.; Denney, R. C.; Jeffery, G. H.; et al. Vogel's Textbook of Quantitative Inorganic Analysis. Longman: London, 1978, pp. 559-560.\n",
            "- 14. Curran, D. J. 'Constant-Current Coulometry.' Chapter 20 in Kissinger, P. T.; Heineman, W. R., eds. Laboratory Techniques in Electroanalytical Chemistry. Marcel Dekker, Inc.: New York, 1984, pp. 539-568.\n",
            "- 15. Maloy, J. T. J. Chem. Educ. 1983, 60, 285-289.\n",
            "- 16. Osteryoung, J. J. Chem. Educ. 1983, 60, 296-298.\n",
            "- 17. Peterson, W. M.; Wong, R. V. Am. Lab. November 1981, 116-128.\n",
            "- 18. Wang, J. Am. Lab. May 1985, 41-50.\n",
            "- 19. Karube, I.; Nomura, Y.; Arikawa, Y. Trends in Anal. Chem. 1995, 14, 295-299.\n",
            "- 20. Lanza, P. J. Chem. Educ. 1990, 67, 704-705.\n",
            "- 21. Wang, J. Stripping Analysis. VCH Publishers: Deerfield Beach, FL, 1985, pp. 116-124.\n",
            "- 22. (a) Batley, G. E.; Florence, T. M. Anal. Lett. 1976, 9, 379-388; (b) Batley, G. E.; Florence, T. M. Talanta 1977, 24, 151; (c) Batley, G. E.; Florence, T. M. Anal. Chem. 1980, 52, 1962-1963; (d) Florence, T. M.; Batley, G. E. CRC Crit. Rev. Anal. Chem. 1980, 9, 219-296.\n",
            "- 23. Brooks, M. A. 'Application of Electrochemistry to Pharmaceutical Analysis' Chapter 21. In Kissinger, P. T.; Heineman, W. R., eds.\n",
            "\n",
            "## 542 Modern Analytical Chemistry\n",
            "\n",
            "Laboratory Techniques in Electroanalytical Chemistry. Marcel Dekker, Inc.: New York, 1984, pp. 569-607.\n",
            "\n",
            "- 24. Cammann, K.; Lemke, U.; Rohen, A.; et al. Angew. Chem. Int. Ed. Engl. 1991, 30, 516-539.\n",
            "- 25. Procedure adapted from Pungor, E. A Practical Guide to Instrumental Analysis. CRC Press: Boca Raton, FL, 1995, pp. 34-37.\n",
            "- 26. Abe, T.; Lau, L. L.; Ewing, A. G. J. Am. Chem. Soc. 1991, 113, 7421-7423.\n",
            "- 27. Creager, S. E.; Lawrence, K. D.; Tibbets, C. R. J. Chem. Educ. 1995, 72, 274-276.\n",
            "- 28. Watanabe, K.; Okada, K.; Oda, H.; et al. Anal. Chim. Acta 1995, 316, 371-375.\n",
            "- 29. Mifflin, T. E.; Andriano, K. M.; Robbins, W. B. J. Chem. Educ. 1984, 61, 638-639.\n",
            "- 30. Meyerhoff, M. E.; Kovach, P. M. J. Chem. Educ. 1983, 9, 766-768.\n",
            "- 31. Wang, J.; Taha, Z. Anal. Chim. Acta 1991, 252, 215-221.\n",
            "- 32. Lau, O.; Luk, S.; Cheung, Y. Analyst, 1989, 114, 1047-1051.\n",
            "- 33. Ratana-ohpas, R; Kanatharana, P.; Ratana-ohpas, W.; et al. Anal. Chim. Acta 1996, 333, 115-118.\n",
            "- 34. Sittampalam, G; Wilson, G. S. J. Chem. Educ. 1982, 59, 70-73.\n",
            "- 35. Abass, A. K.; Hart, J. P.; Cowell, D. C.; et al. Anal. Chim. Acta 1998, 373, 1-8.\n",
            "- 36. van der Berg, C. M. G.; Huang, Z. Q. Anal. Chem. 1984, 56, 2383-2386.\n",
            "\n",
            "## Chapter 12\n",
            "\n",
            "## Chromatographic and Electrophoretic Methods\n",
            "\n",
            "D rawing from an arsenal of analytical techniques, many of which were the subject of the preceding four chapters, analytical chemists have designed methods for the analysis of analytes at increasingly lower concentrations and in increasingly more complex matrices. Despite the power of these techniques, they often suffer from a lack of selectivity. For this reason, many analytical procedures include a step to separate the analyte from potential interferents. Several separation methods, such as liquid-liquid extractions and solid-phase microextractions, were discussed in Chapter 7. In this chapter we consider two additional separation methods that combine separation and analysis: chromatography and electrophoresis.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Overview of Analytical Separations\n",
            "\n",
            "In Chapter 7 we examined several methods for separating an analyte from potential interferents. For example, in a liquid-liquid extraction the analyte and interferent are initially present in a single liquid phase. A second, immiscible liquid phase is introduced, and the two phases are thoroughly mixed by shaking. During this process the analyte and interferents partition themselves between the two phases to different extents, affecting their separation. Despite the power of these separation techniques, there are some significant limitations.\n",
            "\n",
            "## 12A.1 The Problem with Simple Separations\n",
            "\n",
            "Suppose we have a sample containing an analyte in a matrix that is incompatible with our analytical method. To determine the analyte's concentration we first separate it  from the matrix using, for example, a liquid-liquid extraction. If there are additional analytes, we may need to use additional extractions to isolate them from the analyte's matrix. For a complex mixture of analytes this quickly becomes a tedious process.\n",
            "\n",
            "Furthermore, the extent to which we can effect a separation depends on the distribution ratio of each species in the sample. To separate an analyte from its matrix, its distribution ratio must be significantly greater than that for all other components in the matrix. When the analyte's distribution ratio is similar to that of another species, then a separation becomes impossible. For example, let's assume that an analyte, A, and a matrix interferent, I, have distribution ratios of 5 and 0.5, respectively. In an attempt to separate the analyte from its matrix, a simple liquidliquid extraction is carried out using equal volumes of sample and a suitable extraction solvent. Following the treatment outlined in Chapter 7, it is easy to show that a single extraction removes approximately 83% of the analyte and 33% of the interferent. Although it is possible to remove 99% of A with three extractions, 70% of I is also removed. In fact, there is no practical combination of number of extractions or volume ratio of sample and extracting phases that produce an acceptable separation of the analyte and interferent by a simple liquid-liquid extraction.\n",
            "\n",
            "## 12A.2 A Better Way to Separate Mixtures\n",
            "\n",
            "The problem with a simple extraction is that the separation only occurs in one direction. In a liquid-liquid extraction, for example, we extract a solute from its initial  phase  into  the  extracting  phase.  Consider, again, the separation of an analyte and a matrix interferent with distribution ratios of 5 and 0.5, respectively. A single liquid-liquid extraction transfers 83% of the analyte and 33% of the interferent to the extracting phase (Figure 12.1). If the concentrations of A and I in the sample were identical, then their concentration ratio in the extracting phase after one extraction is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Thus, a single extraction improves the separation of the solutes by a factor of 2.5. As shown in Figure 12.1, a second extraction actually leads to a poorer separation. After combining the two portions of the extracting phase, the concentration ratio decreases to\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "A\n",
            "\n",
            "I\n",
            "\n",
            "We can improve the separation by first extracting the solutes into the extracting phase, and then extracting them back into a fresh portion of the initial phase (Figure 12.2). Because solute A has the larger distribution ratio, it is extracted to a greater extent during the first extraction and to a lesser extent during the second extraction. In this case the final concentration ratio of\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "## Figure 12.1\n",
            "\n",
            "Progress of a liquid-liquid extraction using two identical extractions of a sample (initial phase) with fresh portions of the extracting phase. All numbers are fractions of solute in the phases; A = analyte, I = interferent.\n",
            "\n",
            "## Figure 12.2\n",
            "\n",
            "Progress of a liquid-liquid extraction in which the solutes are first extracted into the extracting phase and then extracted back into a fresh portion of the initial phase. All numbers are fractions of solute in the phases; A = analyte, I = interferent.\n",
            "\n",
            "## countercurrent extraction\n",
            "\n",
            "A liquid-liquid extraction in which solutes are extracted back and forth between fresh portions of two extracting phases.\n",
            "\n",
            "## mobile phase\n",
            "\n",
            "In chromatography, the extracting phase that moves through the system.\n",
            "\n",
            "## stationary phase\n",
            "\n",
            "In chromatography, the extracting phase that remains in a fixed position.\n",
            "\n",
            "## chromatography\n",
            "\n",
            "A separation in which solutes partition between a mobile and stationary phase.\n",
            "\n",
            "## column chromatography\n",
            "\n",
            "A form of chromatography in which the stationary phase is retained in a column.\n",
            "\n",
            "## planar chromatography\n",
            "\n",
            "A form of chromatography in which the stationary phase is immobilized on a flat surface.\n",
            "\n",
            "in the extracting phase is significantly greater. The process of extracting the solutes back and forth between fresh portions of the two phases, which is called a countercurrent extraction, was developed by Craig in the 1940s. 1* The same phenomenon forms the basis of modern chromatography.\n",
            "\n",
            "Chromatographic separations are accomplished by continuously passing one sample-free phase, called a mobile phase, over a second sample-free phase that remains fixed, or stationary. The sample is injected, or placed, into the mobile phase. As it moves with the mobile phase, the sample's components partition themselves between the mobile and stationary phases. Those components whose distribution ratio favors the stationary phase require a longer time to pass through the system. Given sufficient time, and sufficient stationary and mobile phase, solutes with similar distribution ratios can be separated.\n",
            "\n",
            "The history of modern chromatography can be traced to the turn of the century when the Russian botanist Mikhail Tswett (1872-1919) used a column packed with a stationary phase of calcium carbonate to separate colored pigments from plant extracts. The sample was placed at the top of the column and carried through the stationary phase using a mobile phase of petroleum ether. As the sample moved through the column, the pigments in the plant extract separated into individual colored bands. Once the pigments were adequately separated, the calcium carbonate was removed from the column, sectioned, and the pigments recovered by extraction. Tswett named the technique chromatography, combining the Greek words for 'color' and 'to write.' There was little interest in Tswett's technique until 1931 when chromatography was reintroduced as an analytical technique for biochemical separations. Pioneering work by Martin and Synge in 1941 2 established the importance of liquid-liquid partition chromatography and led to the development of a theory for chromatographic separations; they were awarded the 1952 Nobel Prize in chemistry for this work. Since then, chromatography in its many forms has become the most important and widely used separation technique. Other separation methods, such as electrophoresis, effect a separation without the use of a stationary phase.\n",
            "\n",
            "## 12A. 3 Classifying Analytical Separations\n",
            "\n",
            "Analytical separations may be classified in three ways: by the physical state of the mobile phase and stationary phase; by the method of contact between the mobile phase and stationary phase; or by the chemical or physical mechanism responsible for separating the sample's constituents. The mobile phase is usually a liquid or a gas, and the stationary phase, when present, is a solid or a liquid film coated on a solid surface. Chromatographic techniques are often named by listing the type of mobile phase, followed by the type of stationary phase. Thus, in gas-liquid chromatography the mobile phase is a gas and the stationary phase is a liquid. If only one phase is indicated, as in gas chromatography, it is assumed to be the mobile phase.\n",
            "\n",
            "Two common approaches are used to bring the mobile phase and stationary phase into contact. In column chromatography, the stationary phase is placed in a narrow column through which the mobile phase moves under the influence of gravity or pressure. The stationary phase is either a solid or a thin, liquid film coating on a solid particulate packing material or the column's walls. In planar chromatography the  stationary  phase  coats  a  flat  glass,  metal,  or  plastic  plate\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "and is placed in a developing chamber. A reservoir containing the mobile phase is  placed in contact with the stationary phase, and the mobile phase moves by capillary action.\n",
            "\n",
            "The mechanism by which solutes separate provides a third means for characterizing a separation (Figure 12.3). In adsorption chromatography, solutes separate based on their ability to adsorb to a solid stationary phase. In partition chromatography, a thin liquid film coating a solid support serves as the stationary phase. Separation is based on a difference in the equilibrium partitioning of solutes between the liquid stationary phase and the mobile phase. Stationary phases consisting of a solid support with covalently attached anionic (e.g., -SO3 -) or cationic (e.g.,  -N(CH3)3 + )  functional  groups  are  used  in  ion-exchange  chromatography. Ionic solutes are attracted to the stationary phase by electrostatic forces.  Porous gels are used as stationary phases in size-exclusion chromatography, in which separation is due to differences in the size of the solutes. Large solutes are unable to penetrate into the porous stationary phase and so quickly pass through the column. Smaller solutes enter into the porous stationary phase, increasing the time spent on the column. Not all separation methods require a stationary phase. In an electrophoretic separation, for example, charged solutes migrate under the influence of an applied potential field. Differences in the mobility of the ions account for their separation.\n",
            "\n",
            "## 12B General Theory of Column Chromatography\n",
            "\n",
            "Of the two methods for bringing the stationary and mobile phases into contact, the  more important is column chromatography. In this section we develop a  general  theory  that  we  may  apply  to  any  form  of  column  chromatography. With appropriate modifications, this theory also can be applied to planar chromatography.\n",
            "\n",
            "A typical column chromatography experiment is outlined in Figure 12.4. Although the figure depicts a liquid-solid chromatographic experiment similar to that first used by Tswett, the design of the column and the physical state of the\n",
            "\n",
            "## Figure 12.4\n",
            "\n",
            "Progress of a column chromatographic separation showing the separation of two solute bands.\n",
            "\n",
            "## Figure 12.5\n",
            "\n",
            "Another view of the progress of a column chromatographic separation showing the separation of two solute bands.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 12.6\n",
            "\n",
            "Typical chromatogram of detector response as a function of retention time.\n",
            "\n",
            "## chromatogram\n",
            "\n",
            "A plot of the detector's signal as function of elution time or volume.\n",
            "\n",
            "## retention time\n",
            "\n",
            "The time a solute takes to move from the point of injection to the detector ( t r ).\n",
            "\n",
            "## retention volume\n",
            "\n",
            "The volume of mobile phase needed to move a solute from its point of injection to the detector ( V r ).\n",
            "\n",
            "## baseline width\n",
            "\n",
            "The width of a solute's chromatographic band measured at the baseline ( w ).\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "stationary and mobile phases may vary. The sample is introduced at the top of the column as a narrow band. Ideally, the solute's initial concentration profile is rectangular (Figure 12.5a). As the sample moves down the column the solutes begin to  separate,  and the individual solute bands begin to broaden and develop a Gaussian profile (Figures 12.5b,c). If the strength of each solute's interaction with the stationary phase is sufficiently different, then the solutes separate into individual bands (Figure 12.5d). The progress of a chromatographic separation is monitored with a suitable detector situated at the end of the column. A plot of the detector's signal as a function of time or volume of eluted mobile phase is known as a chromatogram (Figure 12.6) and consists of a peak for each of the separated solute bands.\n",
            "\n",
            "A chromatographic peak may be characterized in many ways, two of which are shown in Figure 12.7. The retention time, t r , is the elapsed time from the introduction of the solute to the peak maximum. The retention time also can be measured indirectly as the volume of mobile phase eluting between the solute's introduction and the appearance of the solute's peak maximum. This is known as the retention volume, V r . Dividing the retention volume by the mobile phase's flow rate, u, gives the retention time.\n",
            "\n",
            "The second important parameter is the chromatographic peak's width at the baseline, w. As shown in Figure 12.7, baseline width is determined by the intersection with the baseline of tangent lines drawn through the inflection points on either side of the chromatographic peak. Baseline width is measured in units of time or volume, depending on whether the retention time or retention volume is of interest.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Besides the solute peak, Figure 12.7 also shows a small peak eluted soon after the sample is injected into the mobile phase. This peak results from solutes that move through the column at the same rate as the mobile phase. Since these solutes do not interact with the stationary phase, they are considered nonretained. The time or volume of mobile phase required to elute nonretained components is called the column's void time, t m, or void volume.\n",
            "\n",
            "## 12B.1 Chromatographic Resolution\n",
            "\n",
            "The goal of chromatography is to separate a sample into a series of chromatographic peaks, each representing a single component of the sample. Resolution is a quantitative measure of the degree of separation between two chromatographic peaks, A and B, and is defined as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "As shown in Figure 12.8, the degree of separation between two chromatographic peaks improves with an increase in R. For two peaks of equal size, a resolution of 1.5 corresponds to an overlap in area of only 0.13%. Because resolution is a quantitative measure of a separation's success, it provides a useful way to determine if a change in experimental conditions leads to a better separation.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 12.1\n",
            "\n",
            "In a chromatographic analysis of lemon oil a peak for limonene has a retention time of 8.36 min with a baseline width of 0.96 min. γ -Terpinene elutes at 9.54 min, with a baseline width of 0.64 min. What is the resolution between the two peaks?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "Using equation 12.1, we find that the resolution is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "## Figure 12.7\n",
            "\n",
            "Measurement of the column's void time, t m, and the retention time, t r , and baseline width, w, for a solute.\n",
            "\n",
            "## void time\n",
            "\n",
            "The time required for unretained solutes to move from the point of injection to the detector ( t m).\n",
            "\n",
            "## void volume\n",
            "\n",
            "The volume of mobile phase needed to move an unretained solute from the point of injection to the detector.\n",
            "\n",
            "## resolution\n",
            "\n",
            "The separation between two chromatographic bands ( R ).\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Figure 12.8\n",
            "\n",
            "Three examples of chromatographic resolution.\n",
            "\n",
            "## Figure 12.9\n",
            "\n",
            "Two methods for improving chromatographic resolution: (a) Original separation showing a pair of poorly resolved solutes; (b) Improvement in resolution due to an increase in column efficiency; (c) Improvement in resolution due to a change in column selectivity.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "From equation 12.1 it is clear that resolution may be improved either by increasing ∆ t r or by decreasing w A or w B (Figure 12.9). We can increase ∆ t r by enhancing the interaction of the solutes with the column or by increasing the column's selectivity for one of the solutes. Peak width is a kinetic effect associated with the solute's movement within and between the mobile phase and stationary phase. The effect is governed by several factors that are collectively called column efficiency. Each of these factors is considered in more detail in the following sections.\n",
            "\n",
            "## 12B.2 Capacity Factor\n",
            "\n",
            "The distribution of a solute, S, between the mobile phase and stationary phase can be represented by an equilibrium reaction\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and its associated partition coefficient, K D, and distribution ratio, D,\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where the subscripts m and s refer to the mobile phase and stationary phase, respectively. As long as the solute is not involved in any additional equilibria in either the mobile phase or stationary phase, the equilibrium partition coefficient and the distribution ratio will be the same.\n",
            "\n",
            "Conservation of mass requires that the total moles of solute remain constant throughout the separation, thus\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving equation 12.3 for the moles of solute in the stationary phase and substituting into equation 12.2 gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where V m and V s are the volumes of the mobile and stationary phases. Rearranging and solving for the fraction of solute in the mobile phase, f m, gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Note that this equation is identical to that describing the extraction of a solute in a liquid-liquid extraction (equation 7.25 in Chapter 7). Since the volumes of the stationary and mobile phase may not be known, equation 12.4 is simplified by dividing both the numerator and denominator by V m; thus\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "is the solute's capacity factor.\n",
            "\n",
            "A solute's capacity factor can be determined from a chromatogram by measuring the column's void time, t m, and the solute's retention time, t r (see Figure 12.7). The mobile phase's average linear velocity, u, is equal to the length of the column, L, divided by the time required to elute a nonretained solute.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "By the same reasoning, the solute's average linear velocity, v, is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The solute can only move through the column when it is in the mobile phase. Its average linear velocity, therefore, is simply the product of the mobile phase's average linear velocity and the fraction of solute present in the mobile phase.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Substituting equations 12.5, 12.7, and 12.8 into equation 12.9 gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Finally, solving this equation for k ′ gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where t r ′ is known as the adjusted retention time.\n",
            "\n",
            "## capacity factor\n",
            "\n",
            "A measure of how strongly a solute is retained by the stationary phase ( k ′ ).\n",
            "\n",
            "## adjusted retention time\n",
            "\n",
            "The difference between a solute's retention time and column's void time ( t r ′ ).\n",
            "\n",
            "## selectivity factor\n",
            "\n",
            "The ratio of capacity factors for two solutes showing the column's selectivity for one of the solutes ( α ).\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 12.2\n",
            "\n",
            "In a chromatographic analysis of low-molecular-weight acids, butyric acid elutes with a retention time of 7.63 min. The column's void time is 0.31 min. Calculate the capacity factor for butyric acid.\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "## 12B. 3 Column Selectivity\n",
            "\n",
            "The relative selectivity of a chromatographic column for a pair of solutes is given by the selectivity factor, α , which is defined as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The identities of the solutes are defined such that solute A always has the smaller retention time. Accordingly, the selectivity factor is equal to 1 when the solutes elute with identical retention times, and is greater than 1 when t r,B is  greater than t r,A .\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 12. 3\n",
            "\n",
            "In the same chromatographic analysis for low-molecular-weight acids considered in Example 12.2, the retention time for isobutyric acid is 5.98 min. What is the selectivity factor for isobutyric acid and butyric acid?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "First we must calculate the capacity factor for isobutyric acid. Using the void time from Example 12.2, this is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The selectivity factor, therefore, is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "## 12B. 4 Column Efficiency\n",
            "\n",
            "At the beginning of a chromatographic separation the solute occupies a narrow band of finite width. As the solute passes through the column, the width of its band\n",
            "\n",
            "continually increases in a process called band broadening. Column efficiency provides a quantitative measure of the extent of band broadening.\n",
            "\n",
            "In their original theoretical model of chromatography, Martin and Synge 2 treated the chromatographic column as though it consists of discrete sections at which partitioning of the solute between the stationary and mobile phases occurs. They called each section a theoretical plate and defined column efficiency in terms of the number of theoretical plates, N, or the height of a theoretical plate, H; where\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "A column's efficiency improves with an increase in the number of theoretical plates or a decrease in the height of a theoretical plate.\n",
            "\n",
            "Assuming a Gaussian profile, the extent of band broadening is measured by the variance or standard deviation of a chromatographic peak. The height of a theoretical plate is defined as the variance per unit length of the column\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where the variance, σ 2 ,  has  units  of  distance  squared.  Because retention time and peak width are usually measured in seconds or minutes, it is more convenient to express the standard deviation in units of time, τ , by dividing σ by the mobile phase's average linear velocity.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "When a chromatographic peak has a Gaussian shape, its width at the baseline, w, is four times its standard deviation, τ .\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Combining equations 12.13 through 12.15 gives the height of a theoretical plate in terms of the easily measured chromatographic parameters t r and w.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The number of theoretical plates in a chromatographic column is obtained by combining equations 12.12 and 12.16.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Alternatively, the number of theoretical plates can be approximated as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where w 1/2 is the width of the chromatographic peak at half its height.\n",
            "\n",
            "## band broadening\n",
            "\n",
            "The increase in a solute's baseline width as it moves from the point of injection to the detector.\n",
            "\n",
            "## theoretical plate\n",
            "\n",
            "A quantitative means of evaluating column efficiency that treats the column as though it consists of a series of small zones, or plates, in which partitioning between the mobile and stationary phases occurs.\n",
            "\n",
            "## peak capacity\n",
            "\n",
            "The maximum number of solutes that can be resolved on a particular column ( n c ).\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## EXAMPLE 12. 4\n",
            "\n",
            "A chromatographic analysis for the chlorinated pesticide Dieldrin gives a peak with a retention time of 8.68 min and a baseline width of 0.29 min. How many theoretical plates are involved in this separation? Given that the column used in this analysis is 2.0 meters long, what is the height of a theoretical plate?\n",
            "\n",
            "## SOLUTION\n",
            "\n",
            "Using equation 12.17, the number of theoretical plates is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving equation 12.12 for H gives the average height of a theoretical plate as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "It is important to remember that a theoretical plate is an artificial construct and that no such plates exist in a chromatographic column. In fact, the number of theoretical plates depends on both the properties of the column and the solute. As a result, the number of theoretical plates for a column is not fixed and may vary from solute to solute.\n",
            "\n",
            "## 12B. 5 Peak Capacity\n",
            "\n",
            "Another important consideration is the number of solutes that can be baseline resolved on a given column. An estimate of a column's peak capacity, n c , is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "where V min and V max are the smallest and largest volumes of mobile phase in which a solute can be eluted and detected. 3 A column with 10,000 theoretical plates, for example, can resolve no more than\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "if  the  minimum and maximum volumes of mobile phase in which the solutes can elute are 1 mL and 30 mL. This estimate provides an upper bound on the number of solutes that might be separated and may help to exclude from consideration columns that do not have enough theoretical plates to separate a complex mixture. Just because a column's theoretical peak capacity is larger than the number of solutes to be separated, however, does not mean that the separation will be feasible. In most situations the peak capacity obtained is less\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "than the estimated value because the retention characteristics of some solutes are too similar to effect their separation. Nevertheless, columns with more theoretical plates, or a greater range of possible elution volumes, are more likely to separate a complex mixture.\n",
            "\n",
            "## 12B.6 Nonideal Behavior\n",
            "\n",
            "The treatment of chromatography outlined in Section 12B assumes that a solute elutes as a symmetrical band, such as that shown in Figure 12.7. This ideal behavior occurs when the solute's partition coefficient, K D, is constant for all concentrations of solute. In some situations, chromatographic peaks show nonideal behavior, leading to asymmetrical peaks, similar to those shown in Figure 12.10. The chromatographic peak in Figure 12.10a is an example of ' fronting ' and is most often the result of overloading the column with sample. Figure 12.10b, which is an example of ' tailing, '  occurs when some sites on the stationary phase retain the solutes more strongly than other sites.\n",
            "\n",
            "## Figure 12.10\n",
            "\n",
            "Nonideal asymmetrical chromatographic bands showing (a) fronting and (b) tailing. Also depicted are the corresponding sorption isotherms showing the relationship between the concentration of solute in the stationary phase as a function of its concentration in the mobile phase.\n",
            "\n",
            "## fronting\n",
            "\n",
            "A tail at the beginning of a chromatographic peak, usually due to injecting too much sample.\n",
            "\n",
            "## tailing\n",
            "\n",
            "A tail at the end of a chromatographic peak, usually due to the presence of highly active sites in the stationary phase.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Optimizing Chromatographic Separations\n",
            "\n",
            "Now that we have defined capacity factor, selectivity, and column efficiency we consider their relationship to chromatographic resolution. Since we are only interested in the resolution between solutes eluting with similar retention times, it is safe to assume that the peak widths for the two solutes are approximately the same. Equation 12.1, therefore, is written as\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Solving equation 12.17 for w B and substituting into equation 12.19 gives\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "The retention times for solutes A and B are replaced with their respective capacity factors by rearranging equation 12.10\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "and substituting into equation 12.20.\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Finally, solute A's capacity factor is eliminated using equation 12.11. After rearranging, the equation for the resolution between the chromatographic peaks for solutes A and B is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Besides resolution, another important factor in chromatography is the amount of time required to elute a pair of solutes. The time needed to elute solute B is\n",
            "\n",
            "<!-- formula-not-decoded -->\n",
            "\n",
            "Equations 12.21 and 12.22 contain terms corresponding to column efficiency, column selectivity, and capacity factor. These terms can be varied, more or less independently, to obtain the desired resolution and analysis time for a pair of solutes. The first term, which is a function of the number of theoretical plates or the height of a theoretical plate, accounts for the effect of column efficiency. The second term is a function of α and accounts for the influence of column selectivity. Finally, the third term in both equations is a function of k B, and accounts for the effect of solute ′ B's capacity factor. Manipulating these parameters to improve resolution is the subject of the remainder of this section.\n",
            "\n",
            "## 12C.1 Using the Capacity Factor to Optimize Resolution\n",
            "\n",
            "One of the simplest ways to improve resolution is to adjust the capacity factor for solute B. If all other terms in equation 12.21 remain constant, increasing k B ′ improves resolution. As shown in Figure 12.11, however, the effect is greatest when the\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "original capacity factor is small. Furthermore, large increases in k B ′ do not lead to proportionally larger increases in resolution. For example, when the original value of k B ′ is 1, increasing its value to 10 gives an 82% improvement in resolution; a further increase to 15 provides a net improvement in resolution of only 87.5%.\n",
            "\n",
            "Any improvement in resolution obtained by increasing k B ′ generally comes at the expense of a longer analysis time. This is also indicated in Figure 12.11, which shows the relative change in retention time as a function of the new capacity factor. Note that a minimum in the retention time curve occurs when k B ′ is equal to 2, and that retention time increases in either direction. Increasing k B ′ from 2 to 10, for example, approximately doubles solute B's retention time.\n",
            "\n",
            "The relationship between capacity factor and analysis time can be advantageous when a separation produces an acceptable resolution with a large k B. In this case it ′ may be possible to decrease k B ′ with little loss in resolution while significantly shortening the analysis time.\n",
            "\n",
            "A solute's capacity factor is directly proportional to its distribution ratio (equation 12.6), which, in turn, is proportional to the solute's equilibrium distribution constant. To increase k B ′ without significantly changing α ,  which also is a function of k B,  it  is  necessary  to  alter  chromatographic conditions in a way that leads to a ′ general, nonselective increase in the capacity factor for both solutes. In gas chromatography, this is usually accomplished by decreasing the column's temperature. At a lower temperature a solute's vapor pressure decreases, ensuring that it spends more time in the stationary phase increasing its capacity factor. In liquid chromatography, changing the mobile phase's solvent strength is the easiest way to change a solute's capacity factor. When the mobile phase has a low solvent strength, solutes spend proportionally more time in the stationary phase, thereby increasing their capacity factors. Additionally, equation 12.6 shows that the capacity factor is proportional to the volume of stationary phase. Increasing the volume of stationary phase, therefore, also leads to an increase in kB. ′\n",
            "\n",
            "Adjusting the capacity factor to improve resolution between one pair of solutes may lead to an unacceptably long retention time for other solutes. For example, improving resolution for solutes with short retention times by increasing\n",
            "\n",
            "## Figure 12.11\n",
            "\n",
            "′\n",
            "\n",
            "Effect of a change in k B on resolution and retention time. The original value of k B ′ is assumed to be 1.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "ByEcVOjrx2z6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from docling.document_converter import DocumentConverter\n",
        "from docx import Document\n",
        "\n",
        "\n",
        "\n",
        "from docling.document_converter import DocumentConverter\n",
        "\n",
        "source = \"/content/2408.09869v5.pdf\"  # document per local path or URL\n",
        "converter = DocumentConverter()\n",
        "result = converter.convert(source)\n",
        "print(result.document.export_to_markdown())  # output: \"## Docling Technical Report[...]\"\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "document = Document()\n",
        "\n",
        "# إضافة نتائج التحويل إلى المستند\n",
        "document.add_paragraph(result.document.export_to_markdown())\n",
        "\n",
        "# حفظ المستند\n",
        "document.save(\"output.docx\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "uhRO_e7Sx2wU",
        "outputId": "4e41b233-1836-4304-ed3a-f0250f50e2bb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "<!-- image -->\n",
            "\n",
            "## Docling Technical Report\n",
            "\n",
            "Version 1.0\n",
            "\n",
            "Christoph Auer Maksym Lysak Ahmed Nassar Michele Dolfi Nikolaos Livathinos Panos Vagenas Cesar Berrospi Ramis Matteo Omenetti Fabian Lindlbauer Kasper Dinkla Lokesh Mishra Yusik Kim Shubham Gupta Rafael Teixeira de Lima Valery Weber Lucas Morin Ingmar Meijer Viktor Kuropiatnyk Peter W. J. Staar\n",
            "\n",
            "AI4K Group, IBM Research R¨ uschlikon, Switzerland\n",
            "\n",
            "## Abstract\n",
            "\n",
            "This technical report introduces Docling , an easy to use, self-contained, MITlicensed open-source package for PDF document conversion. It is powered by state-of-the-art specialized AI models for layout analysis (DocLayNet) and table structure recognition (TableFormer), and runs efficiently on commodity hardware in a small resource budget. The code interface allows for easy extensibility and addition of new features and models.\n",
            "\n",
            "## 1 Introduction\n",
            "\n",
            "Converting PDF documents back into a machine-processable format has been a major challenge for decades due to their huge variability in formats, weak standardization and printing-optimized characteristic, which discards most structural features and metadata. With the advent of LLMs and popular application patterns such as retrieval-augmented generation (RAG), leveraging the rich content embedded in PDFs has become ever more relevant. In the past decade, several powerful document understanding solutions have emerged on the market, most of which are commercial software, cloud offerings [3] and most recently, multi-modal vision-language models. As of today, only a handful of open-source tools cover PDF conversion, leaving a significant feature and quality gap to proprietary solutions.\n",
            "\n",
            "With Docling , we open-source a very capable and efficient document conversion tool which builds on the powerful, specialized AI models and datasets for layout analysis and table structure recognition we developed and presented in the recent past [12, 13, 9]. Docling is designed as a simple, self-contained python library with permissive license, running entirely locally on commodity hardware. Its code architecture allows for easy extensibility and addition of new features and models.\n",
            "\n",
            "Here is what Docling delivers today:\n",
            "\n",
            "- · Converts PDF documents to JSON or Markdown format, stable and lightning fast\n",
            "- · Understands detailed page layout, reading order, locates figures and recovers table structures\n",
            "- · Extracts metadata from the document, such as title, authors, references and language\n",
            "- · Optionally applies OCR, e.g. for scanned PDFs\n",
            "- · Can be configured to be optimal for batch-mode (i.e high throughput, low time-to-solution) or interactive mode (compromise on efficiency, low time-to-solution)\n",
            "- · Can leverage different accelerators (GPU, MPS, etc).\n",
            "\n",
            "## 2 Getting Started\n",
            "\n",
            "To use Docling, you can simply install the docling package from PyPI. Documentation and examples are available in our GitHub repository at github.com/DS4SD/docling. All required model assets 1 are downloaded to a local huggingface datasets cache on first use, unless you choose to pre-install the model assets in advance.\n",
            "\n",
            "Docling provides an easy code interface to convert PDF documents from file system, URLs or binary streams, and retrieve the output in either JSON or Markdown format. For convenience, separate methods are offered to convert single documents or batches of documents. A basic usage example is illustrated below. Further examples are available in the Doclign code repository.\n",
            "\n",
            "```\n",
            "from docling.document_converter import DocumentConverter Large\n",
            "```\n",
            "\n",
            "```\n",
            "source = \"https://arxiv.org/pdf/2206.01062\" # PDF path or URL converter = DocumentConverter() result = converter.convert_single(source) print(result.render_as_markdown()) # output: \"## DocLayNet: A Human -Annotated Dataset for Document -Layout Analysis [...]\"\n",
            "```\n",
            "\n",
            "Optionally, you can configure custom pipeline features and runtime options, such as turning on or off features (e.g. OCR, table structure recognition), enforcing limits on the input document size, and defining the budget of CPU threads. Advanced usage examples and options are documented in the README file. Docling also provides a Dockerfile to demonstrate how to install and run it inside a container.\n",
            "\n",
            "## 3 Processing pipeline\n",
            "\n",
            "Docling implements a linear pipeline of operations, which execute sequentially on each given document (see Fig. 1). Each document is first parsed by a PDF backend, which retrieves the programmatic text tokens, consisting of string content and its coordinates on the page, and also renders a bitmap image of each page to support downstream operations. Then, the standard model pipeline applies a sequence of AI models independently on every page in the document to extract features and content, such as layout and table structures. Finally, the results from all pages are aggregated and passed through a post-processing stage, which augments metadata, detects the document language, infers reading-order and eventually assembles a typed document object which can be serialized to JSON or Markdown.\n",
            "\n",
            "## 3.1 PDF backends\n",
            "\n",
            "Two basic requirements to process PDF documents in our pipeline are a) to retrieve all text content and their geometric coordinates on each page and b) to render the visual representation of each page as it would appear in a PDF viewer. Both these requirements are encapsulated in Docling's PDF backend interface. While there are several open-source PDF parsing libraries available for python, we faced major obstacles with all of them for different reasons, among which were restrictive\n",
            "\n",
            "1 see huggingface.co/ds4sd/docling-models/\n",
            "\n",
            "Figure 1: Sketch of Docling's default processing pipeline. The inner part of the model pipeline is easily customizable and extensible.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "licensing (e.g. pymupdf [7]), poor speed or unrecoverable quality issues, such as merged text cells across far-apart text tokens or table columns (pypdfium, PyPDF) [15, 14].\n",
            "\n",
            "We therefore decided to provide multiple backend choices, and additionally open-source a custombuilt PDF parser, which is based on the low-level qpdf [4] library. It is made available in a separate package named docling-parse and powers the default PDF backend in Docling. As an alternative, we provide a PDF backend relying on pypdfium , which may be a safe backup choice in certain cases, e.g. if issues are seen with particular font encodings.\n",
            "\n",
            "## 3.2 AI models\n",
            "\n",
            "As part of Docling, we initially release two highly capable AI models to the open-source community, which have been developed and published recently by our team. The first model is a layout analysis model, an accurate object-detector for page elements [13]. The second model is TableFormer [12, 9], a state-of-the-art table structure recognition model. We provide the pre-trained weights (hosted on huggingface) and a separate package for the inference code as docling-ibm-models . Both models are also powering the open-access deepsearch-experience, our cloud-native service for knowledge exploration tasks.\n",
            "\n",
            "## Layout Analysis Model\n",
            "\n",
            "Our layout analysis model is an object-detector which predicts the bounding-boxes and classes of various elements on the image of a given page. Its architecture is derived from RT-DETR [16] and re-trained on DocLayNet [13], our popular human-annotated dataset for document-layout analysis, among other proprietary datasets. For inference, our implementation relies on the onnxruntime [5].\n",
            "\n",
            "The Docling pipeline feeds page images at 72 dpi resolution, which can be processed on a single CPU with sub-second latency. All predicted bounding-box proposals for document elements are post-processed to remove overlapping proposals based on confidence and size, and then intersected with the text tokens in the PDF to group them into meaningful and complete units such as paragraphs, section titles, list items, captions, figures or tables.\n",
            "\n",
            "## Table Structure Recognition\n",
            "\n",
            "The TableFormer model [12], first published in 2022 and since refined with a custom structure token language [9], is a vision-transformer model for table structure recovery. It can predict the logical row and column structure of a given table based on an input image, and determine which table cells belong to column headers, row headers or the table body. Compared to earlier approaches, TableFormer handles many characteristics of tables, such as partial or no borderlines, empty cells, rows or columns, cell spans and hierarchy both on column-heading or row-heading level, tables with inconsistent indentation or alignment and other complexities. For inference, our implementation relies on PyTorch [2].\n",
            "\n",
            "The Docling pipeline feeds all table objects detected in the layout analysis to the TableFormer model, by providing an image-crop of the table and the included text cells. TableFormer structure predictions are matched back to the PDF cells in post-processing to avoid expensive re-transcription text in the table image. Typical tables require between 2 and 6 seconds to be processed on a standard CPU, strongly depending on the amount of included table cells.\n",
            "\n",
            "## OCR\n",
            "\n",
            "Docling provides optional support for OCR, for example to cover scanned PDFs or content in bitmaps images embedded on a page. In our initial release, we rely on EasyOCR [1], a popular thirdparty OCR library with support for many languages. Docling, by default, feeds a high-resolution page image (216 dpi) to the OCR engine, to allow capturing small print detail in decent quality. While EasyOCR delivers reasonable transcription quality, we observe that it runs fairly slow on CPU (upwards of 30 seconds per page).\n",
            "\n",
            "We are actively seeking collaboration from the open-source community to extend Docling with additional OCR backends and speed improvements.\n",
            "\n",
            "## 3.3 Assembly\n",
            "\n",
            "In the final pipeline stage, Docling assembles all prediction results produced on each page into a well-defined datatype that encapsulates a converted document, as defined in the auxiliary package docling-core . The generated document object is passed through a post-processing model which leverages several algorithms to augment features, such as detection of the document language, correcting the reading order, matching figures with captions and labelling metadata such as title, authors and references. The final output can then be serialized to JSON or transformed into a Markdown representation at the users request.\n",
            "\n",
            "## 3.4 Extensibility\n",
            "\n",
            "Docling provides a straight-forward interface to extend its capabilities, namely the model pipeline. A model pipeline constitutes the central part in the processing, following initial document parsing and preceding output assembly, and can be fully customized by sub-classing from an abstract baseclass ( BaseModelPipeline ) or cloning the default model pipeline. This effectively allows to fully customize the chain of models, add or replace models, and introduce additional pipeline configuration parameters. To use a custom model pipeline, the custom pipeline class to instantiate can be provided as an argument to the main document conversion methods. We invite everyone in the community to propose additional or alternative models and improvements.\n",
            "\n",
            "Implementations of model classes must satisfy the python Callable interface. The \\_\\_call\\_\\_ method must accept an iterator over page objects, and produce another iterator over the page objects which were augmented with the additional features predicted by the model, by extending the provided PagePredictions data model accordingly.\n",
            "\n",
            "## 4 Performance\n",
            "\n",
            "In this section, we establish some reference numbers for the processing speed of Docling and the resource budget it requires. All tests in this section are run with default options on our standard test set distributed with Docling, which consists of three papers from arXiv and two IBM Redbooks, with a total of 225 pages. Measurements were taken using both available PDF backends on two different hardware systems: one MacBook Pro M3 Max, and one bare-metal server running Ubuntu 20.04 LTS on an Intel Xeon E5-2690 CPU. For reproducibility, we fixed the thread budget (through setting OMP NUM THREADS environment variable ) once to 4 (Docling default) and once to 16 (equal to full core count on the test hardware). All results are shown in Table 1.\n",
            "\n",
            "If you need to run Docling in very low-resource environments, please consider configuring the pypdfium backend. While it is faster and more memory efficient than the default docling-parse backend, it will come at the expense of worse quality results, especially in table structure recovery.\n",
            "\n",
            "Establishing GPU acceleration support for the AI models is currently work-in-progress and largely untested, but may work implicitly when CUDA is available and discovered by the onnxruntime and\n",
            "\n",
            "torch runtimes backing the Docling pipeline. We will deliver updates on this topic at in a future version of this report.\n",
            "\n",
            "Table 1: Runtime characteristics of Docling with the standard model pipeline and settings, on our test dataset of 225 pages, on two different systems. OCR is disabled. We show the time-to-solution (TTS), computed throughput in pages per second, and the peak memory used (resident set size) for both the Docling-native PDF backend and for the pypdfium backend, using 4 and 16 threads.\n",
            "\n",
            "| CPU                         | Thread budget   | native backend   | native backend   | native backend   | pypdfium backend   | pypdfium backend   | pypdfium backend   |\n",
            "|-----------------------------|-----------------|------------------|------------------|------------------|--------------------|--------------------|--------------------|\n",
            "|                             |                 | TTS              | Pages/s          | Mem              | TTS                | Pages/s            | Mem                |\n",
            "| Apple M3 Max                | 4               | 177 s 167 s      | 1.27 1.34        | 6.20 GB          | 103 s 92 s         | 2.18 2.45          | 2.56 GB            |\n",
            "| (16 cores) Intel(R) E5-2690 | 16 4 16         | 375 s 244 s      | 0.60 0.92        | 6.16 GB          | 239 s 143 s        | 0.94 1.57          | 2.42 GB            |\n",
            "\n",
            "## 5 Applications\n",
            "\n",
            "Thanks to the high-quality, richly structured document conversion achieved by Docling, its output qualifies for numerous downstream applications. For example, Docling can provide a base for detailed enterprise document search, passage retrieval or classification use-cases, or support knowledge extraction pipelines, allowing specific treatment of different structures in the document, such as tables, figures, section structure or references. For popular generative AI application patterns, such as retrieval-augmented generation (RAG), we provide quackling , an open-source package which capitalizes on Docling's feature-rich document output to enable document-native optimized vector embedding and chunking. It plugs in seamlessly with LLM frameworks such as LlamaIndex [8]. Since Docling is fast, stable and cheap to run, it also makes for an excellent choice to build document-derived datasets. With its powerful table structure recognition, it provides significant benefit to automated knowledge-base construction [11, 10]. Docling is also integrated within the open IBM data prep kit [6], which implements scalable data transforms to build large-scale multi-modal training datasets.\n",
            "\n",
            "## 6 Future work and contributions\n",
            "\n",
            "Docling is designed to allow easy extension of the model library and pipelines. In the future, we plan to extend Docling with several more models, such as a figure-classifier model, an equationrecognition model, a code-recognition model and more. This will help improve the quality of conversion for specific types of content, as well as augment extracted document metadata with additional information. Further investment into testing and optimizing GPU acceleration as well as improving the Docling-native PDF backend are on our roadmap, too.\n",
            "\n",
            "We encourage everyone to propose or implement additional features and models, and will gladly take your inputs and contributions under review . The codebase of Docling is open for use and contribution, under the MIT license agreement and in alignment with our contributing guidelines included in the Docling repository. If you use Docling in your projects, please consider citing this technical report.\n",
            "\n",
            "## References\n",
            "\n",
            "- [1] J. AI. Easyocr: Ready-to-use ocr with 80+ supported languages. https://github.com/ JaidedAI/EasyOCR , 2024. Version: 1.7.0.\n",
            "- [2] J. Ansel, E. Yang, H. He, N. Gimelshein, A. Jain, M. Voznesensky, B. Bao, P. Bell, D. Berard, E. Burovski, G. Chauhan, A. Chourdia, W. Constable, A. Desmaison, Z. DeVito, E. Ellison, W. Feng, J. Gong, M. Gschwind, B. Hirsh, S. Huang, K. Kalambarkar, L. Kirsch, M. Lazos, M. Lezcano, Y. Liang, J. Liang, Y. Lu, C. Luk, B. Maher, Y. Pan, C. Puhrsch, M. Reso, M. Saroufim, M. Y. Siraichi, H. Suk, M. Suo, P. Tillet, E. Wang, X. Wang, W. Wen, S. Zhang, X. Zhao, K. Zhou, R. Zou, A. Mathews, G. Chanan, P. Wu, and S. Chintala. Pytorch 2: Faster\n",
            "\n",
            "machine learning through dynamic python bytecode transformation and graph compilation. In Proceedings of the 29th ACM International Conference on Architectural Support for Programming Languages and Operating Systems, Volume 2 (ASPLOS '24) . ACM, 4 2024. doi: 10.1145/3620665.3640366. URL https://pytorch.org/assets/pytorch2-2.pdf .\n",
            "\n",
            "- [3] C. Auer, M. Dolfi, A. Carvalho, C. B. Ramis, and P. W. Staar. Delivering document conversion as a cloud service with high throughput and responsiveness. In 2022 IEEE 15th International Conference on Cloud Computing (CLOUD) , pages 363-373. IEEE, 2022.\n",
            "- [4] J. Berkenbilt. Qpdf: A content-preserving pdf document transformer, 2024. URL https: //github.com/qpdf/qpdf .\n",
            "- [5] O. R. developers. Onnx runtime. https://onnxruntime.ai/ , 2024. Version: 1.18.1.\n",
            "- [6] IBM. Data Prep Kit: a community project to democratize and accelerate unstructured data preparation for LLM app developers, 2024. URL https://github.com/IBM/ data-prep-kit .\n",
            "- [7] A. S. Inc. PyMuPDF, 2024. URL https://github.com/pymupdf/PyMuPDF .\n",
            "- [8] J. Liu. LlamaIndex, 11 2022. URL https://github.com/jerryjliu/llama\\_index .\n",
            "- [9] M. Lysak, A. Nassar, N. Livathinos, C. Auer, and P. Staar. Optimized Table Tokenization for Table Structure Recognition. In Document Analysis and Recognition - ICDAR 2023: 17th International Conference, San Jos´ e, CA, USA, August 21-26, 2023, Proceedings, Part II , pages 37-50, Berlin, Heidelberg, Aug. 2023. Springer-Verlag. ISBN 978-3-031-41678-1. doi: 10. 1007/978-3-031-41679-8 3. URL https://doi.org/10.1007/978-3-031-41679-8\\_3 .\n",
            "- [10] L. Mishra, S. Dhibi, Y. Kim, C. Berrospi Ramis, S. Gupta, M. Dolfi, and P. Staar. Statements: Universal information extraction from tables with large language models for ESG KPIs. In D. Stammbach, J. Ni, T. Schimanski, K. Dutia, A. Singh, J. Bingler, C. Christiaen, N. Kushwaha, V. Muccione, S. A. Vaghefi, and M. Leippold, editors, Proceedings of the 1st Workshop on Natural Language Processing Meets Climate Change (ClimateNLP 2024) , pages 193-214, Bangkok, Thailand, Aug. 2024. Association for Computational Linguistics. URL https://aclanthology.org/2024.climatenlp-1.15 .\n",
            "- [11] L. Morin, V. Weber, G. I. Meijer, F. Yu, and P. W. J. Staar. Patcid: an open-access dataset of chemical structures in patent documents. Nature Communications , 15(1):6532, August 2024. ISSN 2041-1723. doi: 10.1038/s41467-024-50779-y. URL https://doi.org/10.1038/ s41467-024-50779-y .\n",
            "- [12] A. Nassar, N. Livathinos, M. Lysak, and P. Staar. Tableformer: Table structure understanding with transformers. In Proceedings of the IEEE/CVF Conference on Computer Vision and Pattern Recognition , pages 4614-4623, 2022.\n",
            "- [13] B. Pfitzmann, C. Auer, M. Dolfi, A. S. Nassar, and P. Staar. Doclaynet: a large humanannotated dataset for document-layout segmentation. pages 3743-3751, 2022.\n",
            "- [14] pypdf Maintainers. pypdf: A Pure-Python PDF Library, 2024. URL https://github.com/ py-pdf/pypdf .\n",
            "- [15] P. Team. PyPDFium2: Python bindings for PDFium, 2024. URL https://github.com/ pypdfium2-team/pypdfium2 .\n",
            "- [16] Y. Zhao, W. Lv, S. Xu, J. Wei, G. Wang, Q. Dang, Y. Liu, and J. Chen. Detrs beat yolos on real-time object detection, 2023.\n",
            "\n",
            "## Appendix\n",
            "\n",
            "In this section, we illustrate a few examples of Docling's output in Markdown and JSON.\n",
            "\n",
            "## DocLayNet: A Large Human-Annotated Dataset for Document-Layout Analysis\n",
            "\n",
            "## DocLayNet: A Large Human-Annotated Dataset for Document-Layout Analysis\n",
            "\n",
            "Birgit Pfitzmann IBM Research Rueschlikon, Switzerland bpf@zurich.ibm.com\n",
            "\n",
            "Christoph Auer IBM Research Rueschlikon, Switzerland cau@zurich.ibm.com\n",
            "\n",
            "Michele Dolfi IBM Research Rueschlikon, Switzerland dol@zurich.ibm.com\n",
            "\n",
            "Ahmed S. Nassar IBM Research Rueschlikon, Switzerland ahn@zurich.ibm.com\n",
            "\n",
            "Peter Staar IBM Research Rueschlikon, Switzerland taa@zurich.ibm.com\n",
            "\n",
            "## ABSTRACT\n",
            "\n",
            "Accurate document layout analysis is a key requirement for highquality PDF document conversion. With the recent availability of public, large ground-truth datasets such as PubLayNet and DocBank, deep-learning models have proven to be very effective at layout detection and segmentation. While these datasets are of adequate size to train such models, they severely lack in layout variability since they are sourced from scientific article repositories such as PubMed and arXiv only. Consequently, the accuracy of the layout segmentation drops significantly when these models are applied on more challenging and diverse layouts. In this paper, we present DocLayNet , a new, publicly available, document-layout annotation dataset in COCO format. It contains 80863 manually annotated pages from diverse data sources to represent a wide variability in layouts. For each PDF page, the layout annotations provide labelled bounding-boxes with a choice of 11 distinct classes. DocLayNet also provides a subset of double- and triple-annotated pages to determine the inter-annotator agreement. In multiple experiments, we provide baseline accuracy scores (in mAP) for a set of popular object detection models. We also demonstrate that these models fall approximately 10% behind the inter-annotator agreement. Furthermore, we provide evidence that DocLayNet is of sufficient size. Lastly, we compare models trained on PubLayNet, DocBank and DocLayNet, showing that layout predictions of the DocLayNettrained models are more robust and thus the preferred choice for general-purpose document-layout analysis.\n",
            "\n",
            "## CCS CONCEPTS\n",
            "\n",
            "· Informationsystems → Documentstructure ; · Appliedcomputing → Document analysis ; · Computing methodologies → Machine learning Computer vision ; ; Object detection ;\n",
            "\n",
            "Permission to make digital or hard copies of part or all of this work for personal or classroom use is granted without fee provided that copies are not made or distributed for profit or commercial advantage and that copies bear this notice and the full citation on the first page. Copyrights for third-party components of this work must be honored. For all other uses, contact the owner/author(s). KDD '22, August 14-18, 2022, Washington, DC, USA © 2022 Copyright held by the owner/author(s). ACM ISBN 978-1-4503-9385-0/22/08. https://doi.org/10.1145/3534678.3539043\n",
            "\n",
            "Birgit Pfitzmann IBM Research Rueschlikon, Switzerland bpf@zurich.ibm.com\n",
            "\n",
            "Christoph Auer IBM Research Rueschlikon, Switzerland cau@zurich.ibm.com\n",
            "\n",
            "Michele Dolfi IBM Research Rueschlikon, Switzerland dol@zurich.ibm.com\n",
            "\n",
            "Ahmed S. Nassar IBM Research Rueschlikon, Switzerland ahn@zurich.ibm.com\n",
            "\n",
            "Peter Staar IBM Research Rueschlikon, Switzerland taa@zurich.ibm.com\n",
            "\n",
            "## ABSTRACT\n",
            "\n",
            "Accurate document layout analysis is a key requirement for highquality PDF document conversion. With the recent availability of public, large groundtruth datasets such as PubLayNet and DocBank, deep-learning models have proven to be very effective at layout detection and segmentation. While these datasets are of adequate size to train such models, they severely lack in layout variability since they are sourced from scientific article repositories such as PubMed and arXiv only. Consequently, the accuracy of the layout segmentation drops significantly when these models are applied on more challenging and diverse layouts. In this paper, we present DocLayNet , a new, publicly available, document-layout annotation dataset in COCO format. It contains 80863 manually annotated pages from diverse data sources to represent a wide variability in layouts. For each PDF page, the layout annotations provide labelled bounding-boxes with a choice of 11 distinct classes. DocLayNet also provides a subset of double- and triple-annotated pages to determine the inter-annotator agreement. In multiple experiments, we provide baseline accuracy scores (in mAP) for a set of popular object detection models. We also demonstrate that these models fall approximately 10% behind the inter-annotator agreement. Furthermore, we provide evidence that DocLayNet is of sufficient size. Lastly, we compare models trained on PubLayNet, DocBank and DocLayNet, showing that layout predictions of the DocLayNettrained models are more robust and thus the preferred choice for general-purpose document-layout analysis.\n",
            "\n",
            "## CCS CONCEPTS\n",
            "\n",
            "Æ Information systems → Document structure ; Æ Applied computing → Document analysis ; Æ Computing methodologies → Machine learning ; Computer vision ; Object detection ;\n",
            "\n",
            "Permission to make digital or hard copies of part or all of this work for personal or classroom use is granted without fee provided that copies are not made or distributed for profit or commercial advantage and that copies bear this notice and the full citation on the first page. Copyrights for third-party components of this work must be honored. For all other uses, contact the owner/author(s).\n",
            "\n",
            "KDD '22, August 14-18, 2022, Washington, DC, USA ' 2022 Copyright held by the owner/author(s). ACM ISBN 978-1-4503-9385-0/22/08. https://doi.org/10.1145/3534678.3539043\n",
            "\n",
            "Figure 1: Four examples of complex page layouts across different document categories\n",
            "\n",
            "## KEYWORDS\n",
            "\n",
            "PDF document conversion, layout segmentation, object-detection, data set, Machine Learning\n",
            "\n",
            "## ACM Reference Format:\n",
            "\n",
            "Birgit Pfitzmann, Christoph Auer, Michele Dolfi, Ahmed S. Nassar, and Peter Staar. 2022. DocLayNet: A Large Human-Annotated Dataset for DocumentLayout Analysis. In Proceedings of the 28th ACM SIGKDD Conference on Knowledge Discovery and Data Mining (KDD '22), August 14-18, 2022, Washington, DC, USA. ACM, New York, NY, USA, 9 pages. https://doi.org/10.1145/ 3534678.3539043\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "AGL Energy Limited  ABN 74 1\n",
            "\n",
            "5 061 375\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Figure 1: Four examples of complex page layouts across different document categories\n",
            "\n",
            "## KEYWORDS\n",
            "\n",
            "PDF document conversion, layout segmentation, object-detection, data set, Machine Learning\n",
            "\n",
            "## ACMReference Format:\n",
            "\n",
            "Birgit Pfitzmann, Christoph Auer, Michele Dolfi, Ahmed S. Nassar, and Peter Staar. 2022. DocLayNet: A Large Human-Annotated Dataset for DocumentLayout Analysis. In Proceedings of the 28th ACM SIGKDD Conference on Knowledge Discovery and Data Mining (KDD '22), August 14-18, 2022, Washington, DC, USA. ACM, New York, NY, USA, 9 pages. https://doi.org/10.1145/ 3534678.3539043\n",
            "\n",
            "1 INTRODUCTION\n",
            "\n",
            "Despite the substantial improvements achieved with machine-learning (ML) approaches and deep neural networks in recent years, document conversion remains a challenging problem, as demonstrated by the numerous public competitions held on this topic [1-4]. The challenge originates from the huge variability in PDF documents regarding layout, language and formats (scanned, programmatic or a combination of both). Engineering a single ML model that can be applied on all types of documents and provides high-quality layout segmentation remains to this day extremely challenging [5]. To highlight the variability in document layouts, we show a few example documents from the DocLayNet dataset in Figure 1. Figure 2: Title page of the DocLayNet paper (arxiv.org/pdf/2206.01062) - left PDF, right rendered Markdown. If recognized, metadata such as authors are appearing first under the title. Text content inside figures is currently dropped, the caption is retained and linked to the figure in the JSON representation (not shown).\n",
            "\n",
            "KDD '22, August 14-18, 2022, Washington, DC, USA Birgit Pfitzmann, Christoph Auer, Michele Dolfi, Ahmed S. Nassar, and Peter Staar\n",
            "\n",
            "Table 2: Prediction performance (mAP@0.5-0.95) of object detection networks on DocLayNet test set. The MRCNN (Mask R-CNN) and FRCNN (Faster R-CNN) models with ResNet-50 or ResNet-101 backbone were trained based on the network architectures from the detectron2 model zoo (Mask R-CNN R50, R101-FPN 3x, Faster R-CNN R101-FPN 3x), with default configurations. The YOLO implementation utilized was YOLOv5x6 [13]. All models were initialised using pre-trained weights from the COCO 2017 dataset.\n",
            "\n",
            "|                                                                                                        | human                                                                   | MRCNN R50 R101                                                                                                          | FRCNN R101                                                  | YOLO v5x6                                                   |\n",
            "|--------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------|-------------------------------------------------------------|\n",
            "| Caption Footnote Formula List-item Page-footer Page-header Picture Section-header Table Text Title All | 84-89 83-91 83-85 87-88 93-94 85-89 69-71 83-84 77-81 84-86 60-72 82-83 | 68.4 71.5 70.9 71.8 60.1 63.4 81.2 80.8 61.6 59.3 71.9 70.0 71.7 72.7 67.6 69.3 82.2 82.9 84.6 85.8 76.7 80.4 72.4 73.5 | 70.1 73.7 63.5 81.0 58.9 72.0 72.0 68.4 82.2 85.4 79.9 73.4 | 77.7 77.2 66.2 86.2 61.1 67.9 77.1 74.6 86.3 88.1 82.7 76.8 |\n",
            "\n",
            "to avoid this at any cost in order to have clear, unbiased baseline numbers for human document-layout annotation. Third, we introduced the feature of snapping boxes around text segments to obtain a pixel-accurate annotation and again reduce time and effort. The CCS annotation tool automatically shrinks every user-drawn box to the minimum bounding-box around the enclosed text-cells for all purely text-based segments, which excludes only Table and Picture . For the latter, we instructed annotation staff to minimise inclusion of surrounding whitespace while including all graphical lines. A downside of snapping boxes to enclosed text cells is that some wrongly parsed PDF pages cannot be annotated correctly and need to be skipped. Fourth, we established a way to flag pages as rejected for cases where no valid annotation according to the label guidelines could be achieved. Example cases for this would be PDF pages that render incorrectly or contain layouts that are impossible to capture with non-overlapping rectangles. Such rejected pages are not contained in the final dataset. With all these measures in place, experienced annotation staff managed to annotate a single page in a typical timeframe of 20s to 60s, depending on its complexity.\n",
            "\n",
            "## 5 EXPERIMENTS\n",
            "\n",
            "The primary goal of DocLayNet is to obtain high-quality ML models capable of accurate document-layout analysis on a wide variety of challenging layouts. As discussed in Section 2, object detection models are currently the easiest to use, due to the standardisation of ground-truth data in COCO format [16] and the availability of general frameworks such as detectron2 [17]. Furthermore, baseline numbers in PubLayNet and DocBank were obtained using standard object detection models such as Mask R-CNN and Faster R-CNN. As such, we will relate to these object detection methods in this\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Third, Ccs achienec Exanole\n",
            "\n",
            "## EXPERIMENTS\n",
            "\n",
            "ground-vuth dawa such WC\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Figure 5: Prediction performance (mAP@0.5-0.95) of a Mask R-CNNnetworkwithResNet50backbonetrainedonincreasing fractions of the DocLayNet dataset. The learning curve flattens around the 80% mark, indicating that increasing the size of the DocLayNet dataset with similar data will not yield significantly better predictions.\n",
            "\n",
            "paper and leave the detailed evaluation of more recent methods mentioned in Section 2 for future work.\n",
            "\n",
            "In this section, we will present several aspects related to the performance of object detection models on DocLayNet. Similarly as in PubLayNet, we will evaluate the quality of their predictions using mean average precision (mAP) with 10 overlaps that range from 0.5 to 0.95 in steps of 0.05 (mAP@0.5-0.95). These scores are computed by leveraging the evaluation code provided by the COCO API [16].\n",
            "\n",
            "## Baselines for Object Detection\n",
            "\n",
            "In Table 2, we present baseline experiments (given in mAP) on Mask R-CNN [12], Faster R-CNN [11], and YOLOv5 [13]. Both training and evaluation were performed on RGB images with dimensions of 1025 × 1025 pixels. For training, we only used one annotation in case of redundantly annotated pages. As one can observe, the variation in mAP between the models is rather low, but overall between 6 and 10% lower than the mAP computed from the pairwise human annotations on triple-annotated pages. This gives a good indication that the DocLayNet dataset poses a worthwhile challenge for the research community to close the gap between human recognition and ML approaches. It is interesting to see that Mask R-CNN and Faster R-CNN produce very comparable mAP scores, indicating that pixel-based image segmentation derived from bounding-boxes does not help to obtain better predictions. On the other hand, the more recent Yolov5x model does very well and even out-performs humans on selected labels such as Text , Table and Picture . This is not entirely surprising, as Text , Table and Picture are abundant and the most visually distinctive in a document.\n",
            "\n",
            "## Baselines for Object Detection\n",
            "\n",
            "mak enbrel\n",
            "\n",
            "Figure 3: Page 6 of the DocLayNet paper. If recognized, metadata such as authors are appearing first under the title. Elements recognized as page headers or footers are suppressed in Markdown to deliver uninterrupted content in reading order. Tables are inserted in reading order. The paragraph in '5. Experiments' wrapping over the column end is broken up in two and interrupted by the table.\n",
            "\n",
            "KDD '22, August 14-18, 2022, Washington, DC, USA\n",
            "\n",
            "Birgit Pfitzmann, Christoph Auer, Michele Dolfi, Ahmed S. Nassar, and Peter Staar\n",
            "\n",
            "Table 1: DocLayNet dataset overview. Along with the frequency of each class label, we present the relative occurrence (as %\n",
            "\n",
            "between pairwise annotations from the triple-annotated pages, from which we obtain accuracy ranges.\n",
            "\n",
            "of row 'Total') in the train, test and validation sets. The inter-annotator agreement is computed as the mAP@0.5-0.95 metric\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "| class label    | Count   | % of Total   | % of Total   | % of Total   | triple inter-annotator mAP @0.5-0.95 (%)   | triple inter-annotator mAP @0.5-0.95 (%)   | triple inter-annotator mAP @0.5-0.95 (%)   | triple inter-annotator mAP @0.5-0.95 (%)   | triple inter-annotator mAP @0.5-0.95 (%)   | triple inter-annotator mAP @0.5-0.95 (%)   | triple inter-annotator mAP @0.5-0.95 (%)   |\n",
            "|----------------|---------|--------------|--------------|--------------|--------------------------------------------|--------------------------------------------|--------------------------------------------|--------------------------------------------|--------------------------------------------|--------------------------------------------|--------------------------------------------|\n",
            "| class label    | Count   | Train        | Test         | Val          | All                                        | Fin                                        | Man                                        | Sci                                        | Law                                        | Pat                                        | Ten                                        |\n",
            "| Caption        | 22524   | 2.04         | 1.77         | 2.32         | 84-89                                      | 40-61                                      | 86-92                                      | 94-99                                      | 95-99                                      | 69-78                                      | n/a                                        |\n",
            "| Footnote       | 6318    | 0.60         | 0.31         | 0.58         | 83-91                                      | n/a                                        | 100                                        | 62-88                                      | 85-94                                      | n/a                                        | 82-97                                      |\n",
            "| Formula        | 25027   | 2.25         | 1.90         | 2.96         | 83-85                                      | n/a                                        | n/a                                        | 84-87                                      | 86-96                                      | n/a                                        | n/a                                        |\n",
            "| List-item      | 185660  | 17.19        | 13.34        | 15.82        | 87-88                                      | 74-83                                      | 90-92                                      | 97-97                                      | 81-85                                      | 75-88                                      | 93-95                                      |\n",
            "| Page-footer    | 70878   | 6.51         | 5.58         | 6.00         | 93-94                                      | 88-90                                      | 95-96                                      | 100                                        | 92-97                                      | 100                                        | 96-98                                      |\n",
            "| Page-header    | 58022   | 5.10         | 6.70         | 5.06         | 85-89                                      | 66-76                                      | 90-94                                      | 98-100                                     | 91-92                                      | 97-99                                      | 81-86                                      |\n",
            "| Picture        | 45976   | 4.21         | 2.78         | 5.31         | 69-71                                      | 56-59                                      | 82-86                                      | 69-82                                      | 80-95                                      | 66-71                                      | 59-76                                      |\n",
            "| Section-header | 142884  | 12.60        | 15.77        | 12.85        | 83-84                                      | 76-81                                      | 90-92                                      | 94-95                                      | 87-94                                      | 69-73                                      | 78-86                                      |\n",
            "| Table          | 34733   | 3.20         | 2.27         | 3.60         | 77-81                                      | 75-80                                      | 83-86                                      | 98-99                                      | 58-80                                      | 79-84                                      | 70-85                                      |\n",
            "| Text           | 510377  | 45.82        | 49.28        | 45.00        | 84-86                                      | 81-86                                      | 88-93                                      | 89-93                                      | 87-92                                      | 71-79                                      | 87-95                                      |\n",
            "| Title          | 5071    | 0.47         | 0.30         | 0.50         | 60-72                                      | 24-63                                      | 50-63                                      | 94-100                                     | 82-96                                      | 68-79                                      | 24-56                                      |\n",
            "| Total          | 1107470 | 941123       | 99816        | 66531        | 82-83                                      | 71-74                                      | 79-81                                      | 89-94                                      | 86-91                                      | 71-76                                      | 68-85                                      |\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "include publication repositories such as arXiv\n",
            "\n",
            "Table 1: DocLayNet dataset overview. Along with the frequency of each class label, we present the relative occurrence (as % of row \"Total\") in the train, test and validation sets. The inter-annotator agreement is computed as the mAP@0.5-0.95 metric between pairwise annotations from the triple-\n",
            "\n",
            "annotated pages, from which we obtain accuracy ranges.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "|                       |         | %of Total   | %of Total   | %of Total   | triple inter- annotator mAP @ 0.5-0.95 (%)   | triple inter- annotator mAP @ 0.5-0.95 (%)   | triple inter- annotator mAP @ 0.5-0.95 (%)   | triple inter- annotator mAP @ 0.5-0.95 (%)   | triple inter- annotator mAP @ 0.5-0.95 (%)   | triple inter- annotator mAP @ 0.5-0.95 (%)   | triple inter- annotator mAP @ 0.5-0.95 (%)   |\n",
            "|-----------------------|---------|-------------|-------------|-------------|----------------------------------------------|----------------------------------------------|----------------------------------------------|----------------------------------------------|----------------------------------------------|----------------------------------------------|----------------------------------------------|\n",
            "| class label           | Count   | Train       | Test        | Val         | All                                          | Fin                                          | Man                                          | Sci                                          | Law                                          | Pat                                          | Ten                                          |\n",
            "| Caption               | 22524   | 2.04        | 1.77        | 2.32        | 84-89                                        | 40-61                                        | 86-92                                        | 94-99                                        | 95-99                                        | 69-78                                        | n/a                                          |\n",
            "| Footnote              | 6318    | 0.60        | 0.31        | 0.58        | 83-91                                        | n/a                                          | 100                                          | 62-88                                        | 85-94                                        | n/a                                          | 82-97                                        |\n",
            "| Formula               | 25027   | 2.25        | 1.90        | 2.96        | 83-85                                        | n/a                                          | n/a                                          | 84-87                                        | 86-96                                        | n/a                                          | n/a                                          |\n",
            "| List-item             | 185660  | 17.19       | 13.34       | 15.82       | 87-88                                        | 74-83                                        | 90-92                                        | 97-97                                        | 81-85                                        | 75-88                                        | 93-95                                        |\n",
            "| Page- footer          | 70878   | 6.51        | 5.58        | 6.00        | 93-94                                        | 88-90                                        | 95-96                                        | 100                                          | 92-97                                        | 100                                          | 96-98                                        |\n",
            "| Page- header offices, | 58022   | 5.10        | 6.70        | 5.06        | 85-89                                        | 66-76                                        | 90-94                                        | 98-100                                       | 91-92                                        | 97-99                                        | 81-86                                        |\n",
            "| Picture               | 45976   | 4.21        | 2.78        | 5.31        | 69-71                                        | 56-59                                        | 82-86                                        | 69-82                                        | 80-95                                        | 66-71                                        | 59-76                                        |\n",
            "| Section- header not   | 142884  | 12.60       | 15.77       | 12.85       | 83-84                                        | 76-81                                        | 90-92                                        | 94-95                                        | 87-94                                        | 69-73                                        | 78-86                                        |\n",
            "| Table                 | 34733   | 3.20        | 2.27        | 3.60        | 77-81                                        | 75-80                                        | 83-86                                        | 98-99                                        | 58-80                                        | 79-84                                        | 70-85                                        |\n",
            "| Text                  | 510377  | 45.82       | 49.28       | 45.00       | 84-86                                        | 81-86                                        | 88-93                                        | 89-93                                        | 87-92                                        | 71-79                                        | 87-95                                        |\n",
            "| Title [22], a         | 5071    | 0.47        | 0.30        | 0.50        | 60-72                                        | 24-63                                        | 50-63                                        | 94-100                                       | 82-96                                        | 68-79                                        | 24-56                                        |\n",
            "| Total in-             | 1107470 | 941123      | 99816       | 66531       | 82-83                                        | 71-74                                        | 79-81                                        | 89-94                                        | 86-91                                        | 71-76                                        | 68-85                                        |\n",
            "\n",
            "3\n",
            "\n",
            ",\n",
            "\n",
            "government offices,\n",
            "\n",
            "We reviewed the col-\n",
            "\n",
            ",\n",
            "\n",
            "Page-\n",
            "\n",
            "Title and\n",
            "\n",
            ".\n",
            "\n",
            "page. Specificity ensures that the choice of label is not ambiguous,\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "we distributed the annotation workload and performed continuous be annotated. We refrained from class labels that are very specific\n",
            "\n",
            "only. For phases three and four, a group of 40 dedicated annotators while coverage ensures that all meaningful items on a page can\n",
            "\n",
            "quality controls. Phase one and two required a small team of experts to a document category, such as\n",
            "\n",
            "Abstract in the\n",
            "\n",
            "Scientific Articles were assembled and supervised.\n",
            "\n",
            "category. We also avoided class labels that are tightly linked to the\n",
            "\n",
            "Phase 1: Data selection and preparation.\n",
            "\n",
            "Our inclusion cri-\n",
            "\n",
            "Author\n",
            "\n",
            "Affiliation\n",
            "\n",
            "teria for documents were described in Section 3. A large effort went into ensuring that all documents are free to use. The data sources in DocBank, are often only distinguishable by discriminating on 3 https://arxiv.org/ Figure 4: Table 1 from the DocLayNet paper in the original PDF (A), as rendered Markdown (B) and in JSON representation (C). Spanning table cells, such as the multi-column header 'triple interannotator mAP@0.5-0.95 (%)', is repeated for each column in the Markdown representation (B), which guarantees that every data point can be traced back to row and column headings only by its grid coordinates in the table. In the JSON representation, the span information is reflected in the fields of each table cell (C).\n",
            "\n",
            "semantics of the text. Labels such as and\n",
            "\n",
            ",\n",
            "\n",
            "as seen\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from docling.document_converter import DocumentConverter\n",
        "from docx import Document\n",
        "\n",
        "\n",
        "\n",
        "from docling.document_converter import DocumentConverter\n",
        "\n",
        "source = \"/content/2408.09869v5.pdf\"  # document per local path or URL\n",
        "converter = DocumentConverter()\n",
        "result = converter.convert(source)\n",
        "print(result.document.export_to_markdown())  # output: \"## Docling Technical Report[...]\"\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "document = Document()\n",
        "\n",
        "# إضافة نتائج التحويل إلى المستند\n",
        "document.add_paragraph(result.document.export_to_markdown())\n",
        "\n",
        "# حفظ المستند\n",
        "document.save(\"output.docx\")"
      ],
      "metadata": {
        "id": "r_CmzwsX9hi7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "source": [
        "with open(\"output.md\", \"w\") as file:\n",
        "    file.write(result.document.export_to_markdown())"
      ],
      "cell_type": "code",
      "metadata": {
        "id": "rvppwOyq9vWz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "source": [
        "from docling.document_converter import DocumentConverter\n",
        "\n",
        "source = \"/content/2408.09869v5.pdf\"  # استبدل your_file.pdf باسم ملفك\n",
        "converter = DocumentConverter()\n",
        "result = converter.convert(source)\n",
        "\n",
        "with open(\"output.md\", \"w\") as file:\n",
        "    file.write(result.document.export_to_markdown())"
      ],
      "cell_type": "code",
      "metadata": {
        "id": "iMqDm-zK9z7r"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## حفظ المخرجات الى ملف markdown"
      ],
      "metadata": {
        "id": "qgirSt_oCCFf"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from docling.document_converter import DocumentConverter\n",
        "\n",
        "source = \"/content/1.pdf\"  # استبدل your_file.pdf باسم ملفك\n",
        "converter = DocumentConverter()\n",
        "result = converter.convert(source)\n",
        "\n",
        "with open(\"output1.md\", \"w\") as file:\n",
        "    file.write(result.document.export_to_markdown())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hvCSxgGD-6FT",
        "outputId": "3c4a6676-cdc2-4b8d-f7de-8f44e622f1cd"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: \n",
            "The secret `HF_TOKEN` does not exist in your Colab secrets.\n",
            "To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.\n",
            "You will be able to reuse this secret in all of your notebooks.\n",
            "Please note that authentication is recommended but still optional to access public models or datasets.\n",
            "  warnings.warn(\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "برنامج مايكروسوفت وورد يستخدم عدة آليات وتقنيات للحفاظ على العلاقة بين النص والرسم (أو أي كائن آخر مثل صورة أو شكل) عند حفظ الملف أو تحويله (مثلاً إلى PDF). إليك كيفية عمل ذلك:\n",
        "نقاط الارتساء (Anchoring):\n",
        "المفهوم الأساسي: كل كائن عائم (غير \"سطري مع النص\") في وورد يتم \"إرساؤه\" أو \"ربطه\" بفقرة معينة في النص. هذه هي الآلية الأساسية.\n",
        "كيف يعمل: عندما تُدرج صورة وتختار لها خيار التفاف نص (مثل \"مربع\"، \"ضيق\"، \"أمام النص\"، \"خلف النص\")، يقوم وورد بربط هذه الصورة بفقرة. إذا تحركت هذه الفقرة (بسبب إضافة أو حذف نص قبلها)، تتحرك الصورة معها لتحافظ على موقعها النسبي بالنسبة لتلك الفقرة.\n",
        "التحكم: يمكنك رؤية علامة المرساة إذا قمت بتفعيل خيار \"إظهار علامات التنسيق\" ثم تحديد الصورة. يمكنك حتى سحب هذه المرساة لربط الصورة بفقرة مختلفة.\n",
        "خيارات تخطيط الكائن (Object Layout Options):\n",
        "التفاف النص (Text Wrapping):\n",
        "سطري مع النص (In Line with Text): يُعامل الكائن كحرف كبير ضمن النص. يتحرك مع النص تمامًا. هذا هو أبسط أشكال العلاقة.\n",
        "مربع (Square)، ضيق (Tight)، عبر (Through): يلتف النص حول حدود الكائن.\n",
        "أعلى وأسفل (Top and Bottom): يأخذ الكائن عرض الصفحة (أو العمود) ويظهر النص فوقه وأسفله.\n",
        "خلف النص (Behind Text)، أمام النص (In Front of Text): يظهر الكائن بشكل مستقل عن تدفق النص، إما خلفه أو أمامه.\n",
        "الموضع (Position):\n",
        "الموضع النسبي: يمكنك تحديد موضع الكائن بالنسبة للفقرة المرتبط بها، الهامش، الصفحة، العمود، إلخ. (مثلاً: \"في المنتصف بالنسبة للهامش الأيسر\").\n",
        "الموضع المطلق: يمكنك تحديد موضع دقيق على الصفحة.\n",
        "خيارات إضافية: مثل \"تحريك الكائن مع النص\" (وهو مرتبط بالارتساء) و \"تثبيت الموضع على الصفحة\" (يلغي تحرك الكائن مع الفقرة المرتبط بها، ويثبته في مكان معين على الصفحة بغض النظر عن تغير النص).\n",
        "تنسيق الملف (File Format - DOCX):\n",
        "عندما تحفظ المستند بتنسيق DOCX (وهو تنسيق مبني على XML)، يتم تخزين كل هذه المعلومات (نوع الارتساء، خيارات الالتفاف، إحداثيات الموضع، حجم الكائن، بيانات الصورة نفسها أو رابط إليها) بشكل منظم داخل ملف XML.\n",
        "عند فتح الملف، يقرأ وورد هذه البيانات ويعيد بناء الصفحة كما كانت، محافظًا على العلاقات بين النص والكائنات.\n",
        "عند التحويل إلى PDF:\n",
        "عندما تقوم بـ \"حفظ باسم\" أو \"تصدير\" المستند إلى PDF، يقوم وورد بعملية \"ترجمة\" للتخطيط.\n",
        "الـ PDF مصمم للحفاظ على المظهر المرئي للمستند كما هو تمامًا.\n",
        "يتم تحويل النص إلى كائنات نصية ذات مواضع ثابتة، والصور إلى كائنات صور ذات مواضع ثابتة على الصفحة.\n",
        "العلاقات الديناميكية (مثل الارتساء بفقرة قد تتحرك) يتم \"تجميدها\" في الـ PDF. بمعنى أن الـ PDF يلتقط \"لقطة شاشة\" متقدمة للتخطيط.\n",
        "يحاول وورد قدر الإمكان أن يجعل مظهر الـ PDF مطابقًا لما تراه على الشاشة في وورد.\n",
        "التجميع (Grouping):\n",
        "إذا كان لديك رسم يتكون من عدة أشكال ونصوص (مربعات نص)، يمكنك \"تجميعها\" معًا. هذا يجعلها تُعامل كوحدة واحدة، مما يسهل الحفاظ على علاقاتها الداخلية عند تحريكها أو تغيير حجمها.\n",
        "باختصار:\n",
        "يعتمد وورد بشكل أساسي على \"الارتساء\" لربط الكائنات (الرسوم والصور) بالفقرات النصية، بالإضافة إلى \"خيارات التخطيط\" التي تحدد كيفية تفاعل النص مع الكائن وموضع الكائن. يتم حفظ كل هذه المعلومات في بنية ملف DOCX، وعند التحويل إلى PDF، يتم تحويل هذا التخطيط الديناميكي إلى تمثيل مرئي ثابت.\n",
        "نصائح للحفاظ على التنسيق بشكل أفضل:\n",
        "استخدم \"سطري مع النص\" قدر الإمكان إذا كان الرسم صغيرًا ويمكن أن يتناسب مع تدفق النص. هذا هو الخيار الأكثر استقرارًا.\n",
        "عند استخدام التفاف النص: تأكد من أن الصورة مرساة بالفقرة الصحيحة.\n",
        "لتخطيطات معقدة: فكر في استخدام الجداول (مع إخفاء حدودها) أو مربعات النص لتنظيم المحتوى، ثم ضع الصور داخلها.\n",
        "اختبر قبل التحويل النهائي: قم دائمًا بمراجعة ملف PDF الناتج للتأكد من أن كل شيء في مكانه الصحيح."
      ],
      "metadata": {
        "id": "yQazdFszCvHt"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "JF1O5dTtCwBM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Yv7RvTgPEnq0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "zCNx_vJtEnoV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "pTuocS9TEnjx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from PIL import Image\n",
        "import pytesseract\n",
        "from docx import Document\n",
        "from docx.shared import Inches\n",
        "import os\n",
        "\n",
        "# --- إعدادات Tesseract (قد تحتاج لتعديل المسار) ---\n",
        "# مثال لنظام ويندوز، إذا لم يكن Tesseract في PATH\n",
        "# pytesseract.pytesseract.tesseract_cmd = r'C:\\Program Files\\Tesseract-OCR\\tesseract.exe'\n",
        "# مثال لنظام لينكس أو ماك (عادة لا تحتاج لهذا السطر إذا كان Tesseract في PATH)\n",
        "# pytesseract.pytesseract.tesseract_cmd = r'/usr/local/bin/tesseract' # أو /usr/bin/tesseract\n",
        "\n",
        "def extract_text_from_image(image_path, lang='ara+eng'):\n",
        "    \"\"\"\n",
        "    يستخرج النص من الصورة باستخدام Tesseract OCR.\n",
        "    \"\"\"\n",
        "    try:\n",
        "        img = Image.open(image_path)\n",
        "        text = pytesseract.image_to_string(img, lang=lang)\n",
        "        return text\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في استخلاص النص: {e}\")\n",
        "        return \"\"\n",
        "\n",
        "def create_word_document(image_path, extracted_text, output_filename=\"output88.docx\"):\n",
        "    \"\"\"\n",
        "    ينشئ ملف وورد يحتوي على الصورة والنص المستخرج.\n",
        "    \"العلاقة\" هنا هي أن كلاهما موجود في نفس المستند.\n",
        "    \"\"\"\n",
        "    try:\n",
        "        doc = Document()\n",
        "        doc.add_heading('الصورة الأصلية والنص المستخرج', level=1)\n",
        "\n",
        "        # إضافة الصورة الأصلية\n",
        "        # يمكنك التحكم في حجم الصورة إذا أردت\n",
        "        # للحفاظ على \"الشكل\" كما هو في الصورة، ندرج الصورة كاملة.\n",
        "        doc.add_picture(image_path, width=Inches(6.0)) # عرض 6 بوصات كمثال\n",
        "\n",
        "        doc.add_heading('النص المستخرج من الصورة:', level=2)\n",
        "        doc.add_paragraph(extracted_text)\n",
        "\n",
        "        doc.save(output_filename)\n",
        "        print(f\"تم حفظ ملف الوورد: {output_filename}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في إنشاء ملف الوورد: {e}\")\n",
        "\n",
        "def create_markdown_file(image_path, extracted_text, output_filename=\"output88.md\"):\n",
        "    \"\"\"\n",
        "    ينشئ ملف ماركداون يحتوي على الصورة والنص المستخرج.\n",
        "    العلاقة هنا هي أن الصورة تسبق النص في الملف.\n",
        "    \"\"\"\n",
        "    try:\n",
        "        # يجب أن تكون الصورة في نفس مسار ملف الماركداون أو مسار نسبي صحيح\n",
        "        image_filename = os.path.basename(image_path)\n",
        "\n",
        "        md_content = f\"# الصورة الأصلية والنص المستخرج\\n\\n\"\n",
        "        md_content += f\"## الصورة الأصلية\\n\"\n",
        "        md_content += f\"![الرسم من الصورة]({image_filename})\\n\\n\" # يفترض أن الصورة في نفس المجلد\n",
        "        md_content += f\"## النص المستخرج\\n\"\n",
        "        md_content += f\"```\\n{extracted_text}\\n```\\n\"\n",
        "\n",
        "        with open(output_filename, \"w\", encoding=\"utf-8\") as f:\n",
        "            f.write(md_content)\n",
        "        print(f\"تم حفظ ملف الماركداون: {output_filename}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في إنشاء ملف الماركداون: {e}\")\n",
        "\n",
        "# --- البرنامج الرئيسي ---\n",
        "if __name__ == \"__main__\":\n",
        "    # ضع مسار الصورة التي تريد معالجتها هنا\n",
        "    # تأكد أن الصورة تحتوي على نص عربي واضح وأشكال\n",
        "    input_image_path = \"/content/PVC-Structure-PhotoRoom.png-PhotoRoom.png\" # <--- قم بتغيير هذا المسار\n",
        "\n",
        "    if not os.path.exists(input_image_path):\n",
        "        print(f\"خطأ: ملف الصورة '{input_image_path}' غير موجود.\")\n",
        "    else:\n",
        "        # 1. استخلاص النص من الصورة\n",
        "        # استخدم 'ara' للغة العربية، 'eng' للإنجليزية، 'ara+eng' لكليهما\n",
        "        text_content = extract_text_from_image(input_image_path, lang='ara')\n",
        "\n",
        "        if text_content:\n",
        "            print(\"\\n--- النص المستخرج ---\")\n",
        "            print(text_content)\n",
        "            print(\"---------------------\\n\")\n",
        "\n",
        "            # 2. إنشاء ملف وورد\n",
        "            # (افترض أن الصورة يجب أن تكون في نفس مجلد السكربت أو توفير المسار الكامل)\n",
        "            word_output_path = \"extracted_content.docx\"\n",
        "            create_word_document(input_image_path, text_content, word_output_path)\n",
        "\n",
        "            # 3. إنشاء ملف ماركداون\n",
        "            # (افترض أن الصورة يجب أن تكون في نفس مجلد السكربت أو توفير المسار الكامل)\n",
        "            # لكي تعمل الصورة في الماركداون بشكل صحيح، قد تحتاج لنسخ الصورة إلى نفس مجلد الماركداون\n",
        "            # أو استخدام مسار نسبي صحيح إذا كان ملف الماركداون في مكان مختلف.\n",
        "            # هنا نفترض أن الصورة وملف الماركداون سيكونان في نفس مكان تشغيل السكربت.\n",
        "            # إذا لم يكن كذلك، انسخ الصورة يدوياً أو برمجياً إلى جانب ملف الماركداون الناتج.\n",
        "\n",
        "            # نسخ الصورة إلى مجلد الإخراج إذا كان ملف الماركداون سيوضع هناك\n",
        "            # output_dir = os.path.dirname(os.path.abspath(__file__)) # مجلد السكربت الحالي\n",
        "            # shutil.copy(input_image_path, os.path.join(output_dir, os.path.basename(input_image_path)))\n",
        "\n",
        "            markdown_output_path = \"extracted_content.md\"\n",
        "            create_markdown_file(input_image_path, text_content, markdown_output_path)\n",
        "        else:\n",
        "            print(\"لم يتم استخلاص أي نص من الصورة.\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 384
        },
        "id": "rwA1Ql-uEngd",
        "outputId": "64c94188-48d0-4c82-ec44-dec7c143e20c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "ModuleNotFoundError",
          "evalue": "No module named 'pytesseract'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-2-c0bc2f179582>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mPIL\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mImage\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mpytesseract\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mdocx\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mDocument\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mdocx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshared\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mInches\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'pytesseract'",
            "",
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0;32m\nNOTE: If your import is failing due to a missing package, you can\nmanually install dependencies using either !pip or !apt.\n\nTo view examples of installing some common dependencies, click the\n\"Open Examples\" button below.\n\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n"
          ],
          "errorDetails": {
            "actions": [
              {
                "action": "open_url",
                "actionText": "Open Examples",
                "url": "/notebooks/snippets/importing_libraries.ipynb"
              }
            ]
          }
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install python-docx Pillow lxml # lxml لتحليل hOCR إذا لم نستخدم parser docling الداخلي"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xDOx9F8qF1uP",
        "outputId": "42b8d6f0-68a9-4f89-a1be-0e8132aa94ea"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: python-docx in /usr/local/lib/python3.11/dist-packages (1.1.2)\n",
            "Requirement already satisfied: Pillow in /usr/local/lib/python3.11/dist-packages (11.2.1)\n",
            "Requirement already satisfied: lxml in /usr/local/lib/python3.11/dist-packages (5.4.0)\n",
            "Requirement already satisfied: typing-extensions>=4.9.0 in /usr/local/lib/python3.11/dist-packages (from python-docx) (4.13.2)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install lxml # lxml لتحليل hOCR إذا لم نستخدم parser docling الداخلي"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "omjanlwCGGXt",
        "outputId": "7d9a2986-7ae4-48e6-ac99-fffa5b988b6f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: lxml in /usr/local/lib/python3.11/dist-packages (5.4.0)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import subprocess\n",
        "import os\n",
        "import shutil\n",
        "from PIL import Image\n",
        "from docx import Document\n",
        "from docx.shared import Inches\n",
        "#from docling.parser import HOCRParser # Import HOCRParser from docling.parser\n",
        "\n",
        "# --- إعدادات ---\n",
        "DOCLING_OUTPUT_DIR = \"docling_output\"\n",
        "\n",
        "def run_docling_ocr(image_path):\n",
        "    \"\"\"\n",
        "    ينفذ أمر docling ocr على الصورة ويخزن المخرجات.\n",
        "    \"\"\"\n",
        "    if os.path.exists(DOCLING_OUTPUT_DIR):\n",
        "        shutil.rmtree(DOCLING_OUTPUT_DIR) # حذف المخرجات القديمة\n",
        "    os.makedirs(DOCLING_OUTPUT_DIR, exist_ok=True)\n",
        "\n",
        "    # أمر docling الأساسي للـ OCR.\n",
        "    # قد تحتاج لتعديل هذا الأمر بناءً على كيفية تثبيت docling\n",
        "    # أو إذا كنت تريد تحديد محرك OCR معين يدعمه docling.\n",
        "    # بدون تحديد محرك، قد يستخدم docling الإعداد الافتراضي (الذي قد يكون Tesseract).\n",
        "    cmd = [\"docling\", \"ocr\", image_path, DOCLING_OUTPUT_DIR]\n",
        "\n",
        "    print(f\"تنفيذ أمر Docling: {' '.join(cmd)}\")\n",
        "    try:\n",
        "        process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
        "        stdout, stderr = process.communicate()\n",
        "\n",
        "        if process.returncode != 0:\n",
        "            print(f\"خطأ أثناء تنفيذ Docling:\")\n",
        "            print(f\"STDOUT: {stdout.decode('utf-8', errors='ignore')}\")\n",
        "            print(f\"STDERR: {stderr.decode('utf-8', errors='ignore')}\")\n",
        "            return None\n",
        "        else:\n",
        "            print(f\"Docling STDOUT: {stdout.decode('utf-8', errors='ignore')}\")\n",
        "            print(f\"Docling أنجز بنجاح. المخرجات في: {DOCLING_OUTPUT_DIR}\")\n",
        "            # البحث عن ملف hOCR الناتج\n",
        "            for fname in os.listdir(DOCLING_OUTPUT_DIR):\n",
        "                if fname.endswith(\".hocr\"):\n",
        "                    return os.path.join(DOCLING_OUTPUT_DIR, fname)\n",
        "            print(\"لم يتم العثور على ملف hOCR في المخرجات.\")\n",
        "            return None\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(\"خطأ: لم يتم العثور على أمر 'docling'. هل هو مثبت وفي PATH؟\")\n",
        "        return None\n",
        "    except Exception as e:\n",
        "        print(f\"حدث خطأ غير متوقع أثناء تشغيل Docling: {e}\")\n",
        "        return None\n",
        "\n",
        "def parse_hocr_with_docling(hocr_file_path):\n",
        "    \"\"\"\n",
        "    يحلل ملف hOCR باستخدام HOCRParser من docling ويستخرج النصوص وإحداثياتها.\n",
        "    \"\"\"\n",
        "    if not hocr_file_path or not os.path.exists(hocr_file_path):\n",
        "        print(f\"ملف HOCR غير موجود: {hocr_file_path}\")\n",
        "        return []\n",
        "\n",
        "    extracted_elements = [] # قائمة لتخزين (text, bbox)\n",
        "    try:\n",
        "        parser = HOCRParser()\n",
        "        doc_layout = parser.parse(hocr_file_path)\n",
        "\n",
        "        for page in doc_layout.pages:\n",
        "            # يمكننا استخلاص على مستوى الكلمات، الأسطر، أو الكتل\n",
        "            # لاستخلاص كتل نصية أكبر، قد يكون من الأفضل التجميع على مستوى البلوك\n",
        "            for block in page.blocks: # Blocks can be text or image\n",
        "                if block.block_type == \"text\": # أو أي نوع نصي آخر يحدده docling\n",
        "                    block_text = \"\"\n",
        "                    # تجميع النص من الأسطر والكلمات داخل البلوك\n",
        "                    for line in block.lines:\n",
        "                        line_text_parts = []\n",
        "                        for word in line.words:\n",
        "                            line_text_parts.append(word.text)\n",
        "                        block_text += \" \".join(line_text_parts) + \"\\n\" # سطر جديد لكل سطر في البلوك\n",
        "\n",
        "                    if block_text.strip():\n",
        "                         # bbox هو (x1, y1, x2, y2)\n",
        "                        extracted_elements.append({\n",
        "                            \"type\": \"text\",\n",
        "                            \"text\": block_text.strip(),\n",
        "                            \"bbox\": block.bbox\n",
        "                        })\n",
        "                elif block.block_type == \"image\" or block.block_type == \"figure\": # أو أنواع صور أخرى\n",
        "                    # docling قد يحفظ الصور المكتشفة في مجلد المخرجات\n",
        "                    # يمكننا هنا فقط تسجيل إحداثياتها لاستخدامها لاحقًا\n",
        "                    # إذا أردنا قصها من الصورة الأصلية\n",
        "                     extracted_elements.append({\n",
        "                        \"type\": \"image_region\",\n",
        "                        \"bbox\": block.bbox,\n",
        "                        # قد تحتاج لإيجاد مسار الصورة المقصوصة إذا قام docling بحفظها\n",
        "                        # \"image_path\": os.path.join(DOCLING_OUTPUT_DIR, \"images\", ...)\n",
        "                    })\n",
        "\n",
        "\n",
        "        # ترتيب العناصر بناءً على موضعها الرأسي (y1) ثم الأفقي (x1)\n",
        "        extracted_elements.sort(key=lambda el: (el[\"bbox\"][1], el[\"bbox\"][0]))\n",
        "        return extracted_elements\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في تحليل ملف hOCR: {e}\")\n",
        "        return []\n",
        "\n",
        "\n",
        "def create_word_from_elements(original_image_path, elements, output_filename=\"docling_output.docx\"):\n",
        "    \"\"\"\n",
        "    ينشئ ملف وورد. يحاول وضع الصورة الأصلية والنصوص المستخرجة.\n",
        "    الحفاظ على \"العلاقة\" هنا سيكون تقريبيًا.\n",
        "    \"\"\"\n",
        "    doc = Document()\n",
        "    doc.add_heading('مستند تم إنشاؤه بواسطة Docling', level=1)\n",
        "\n",
        "    # الخيار 1: إضافة الصورة الأصلية كاملة أولاً\n",
        "    try:\n",
        "        # الحصول على أبعاد الصورة الأصلية لمعرفة مقياس الإحداثيات\n",
        "        with Image.open(original_image_path) as img:\n",
        "            img_width_px, img_height_px = img.size\n",
        "\n",
        "        doc.add_paragraph(\"الصورة الأصلية:\")\n",
        "        # قد تحتاج لتعديل حجم الصورة ليتناسب مع الصفحة\n",
        "        doc.add_picture(original_image_path, width=Inches(6.0))\n",
        "        doc.add_page_break()\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في إضافة الصورة الأصلية إلى الوورد: {e}\")\n",
        "\n",
        "    doc.add_heading('العناصر المستخرجة بواسطة Docling:', level=2)\n",
        "\n",
        "    # الخيار 2: محاولة إضافة النصوص (و ربما أجزاء الصورة لاحقًا)\n",
        "    # هذا الجزء يعتمد بشدة على كيفية رغبتك في تمثيل \"العلاقة\".\n",
        "    # python-docx لا يسهل وضع العناصر في إحداثيات مطلقة.\n",
        "    # الطريقة الأبسط هي إضافة النصوص كفقرات متتالية.\n",
        "\n",
        "    text_content_combined = \"\"\n",
        "    for element in elements:\n",
        "        if element[\"type\"] == \"text\":\n",
        "            text_content_combined += element[\"text\"] + \"\\n\\n\" # فصل بين الكتل النصية\n",
        "        # التعامل مع element[\"type\"] == \"image_region\" سيكون أكثر تعقيدًا:\n",
        "        # ستحتاج لقص هذا الجزء من الصورة الأصلية وإضافته.\n",
        "        # مثال:\n",
        "        # elif element[\"type\"] == \"image_region\":\n",
        "        #     try:\n",
        "        #         with Image.open(original_image_path) as img:\n",
        "        #             cropped_img = img.crop(element[\"bbox\"])\n",
        "        #             # حفظ الصورة المقصوصة مؤقتًا لإضافتها\n",
        "        #             temp_cropped_path = os.path.join(DOCLING_OUTPUT_DIR, f\"cropped_{element['bbox']}.png\")\n",
        "        #             cropped_img.save(temp_cropped_path)\n",
        "        #             doc.add_picture(temp_cropped_path, width=Inches(2)) # مثال للحجم\n",
        "        #             os.remove(temp_cropped_path) # حذف المؤقتة\n",
        "        #     except Exception as e:\n",
        "        #         print(f\"خطأ في معالجة منطقة الصورة: {e}\")\n",
        "\n",
        "\n",
        "    if text_content_combined:\n",
        "        doc.add_paragraph(text_content_combined)\n",
        "    else:\n",
        "        doc.add_paragraph(\"لم يتم استخراج نصوص أو تم استخراج نصوص فارغة.\")\n",
        "\n",
        "    try:\n",
        "        doc.save(output_filename)\n",
        "        print(f\"تم حفظ ملف الوورد: {output_filename}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في حفظ ملف الوورد: {e}\")\n",
        "\n",
        "\n",
        "def create_markdown_from_elements(original_image_path, elements, output_filename=\"docling_output.md\"):\n",
        "    \"\"\"\n",
        "    ينشئ ملف ماركداون. يدرج الصورة الأصلية والنصوص.\n",
        "    \"\"\"\n",
        "    md_content = f\"# مستند تم إنشاؤه بواسطة Docling\\n\\n\"\n",
        "\n",
        "    # يفترض أن الصورة ستكون في نفس مسار ملف الماركداون أو يتم توفير مسار نسبي صحيح.\n",
        "    # لتبسيط الأمر، سنستخدم اسم الملف الأصلي.\n",
        "    original_image_filename = os.path.basename(original_image_path)\n",
        "\n",
        "    md_content += f\"## الصورة الأصلية\\n\"\n",
        "    md_content += f\"![الصورة الأصلية]({original_image_filename})\\n\\n\"\n",
        "\n",
        "    md_content += f\"## النصوص المستخرجة\\n\"\n",
        "    text_for_md = \"\"\n",
        "    for element in elements:\n",
        "        if element[\"type\"] == \"text\":\n",
        "            # في الماركداون، عرض النصوص ككتل متتالية هو الأسهل\n",
        "            text_for_md += element[\"text\"] + \"\\n\\n---\\n\\n\" # فصل بفاصل أفقي\n",
        "        # يمكن أيضًا التعامل مع image_region هنا إذا تم قصها وحفظها كملفات منفصلة\n",
        "        # elif element[\"type\"] == \"image_region\" and element.get(\"image_path\"):\n",
        "        #     md_content += f\"![جزء من الصورة]({os.path.basename(element['image_path'])})\\n\\n\"\n",
        "\n",
        "\n",
        "    if text_for_md:\n",
        "        md_content += text_for_md\n",
        "    else:\n",
        "        md_content += \"لم يتم استخراج نصوص.\\n\"\n",
        "\n",
        "    try:\n",
        "        with open(output_filename, \"w\", encoding=\"utf-8\") as f:\n",
        "            f.write(md_content)\n",
        "        print(f\"تم حفظ ملف الماركداون: {output_filename}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في حفظ ملف الماركداون: {e}\")\n",
        "\n",
        "# --- البرنامج الرئيسي ---\n",
        "if __name__ == \"__main__\":\n",
        "    # ضع مسار الصورة التي تريد معالجتها هنا\n",
        "    input_image_path = \"/content/PVC-Structure-PhotoRoom.png-PhotoRoom.png\" # <--- قم بتغيير هذا المسار\n",
        "\n",
        "    if not os.path.exists(input_image_path):\n",
        "        print(f\"خطأ: ملف الصورة '{input_image_path}' غير موجود.\")\n",
        "    else:\n",
        "        # 1. تنفيذ Docling OCR\n",
        "        hocr_output_file = run_docling_ocr(input_image_path)\n",
        "\n",
        "        if hocr_output_file:\n",
        "            # 2. تحليل مخرجات hOCR\n",
        "            # إذا كنت تريد استخدام محلل hOCR الخاص بـ docling:\n",
        "            extracted_data = parse_hocr_with_docling(hocr_output_file)\n",
        "\n",
        "            if not extracted_data:\n",
        "                print(\"لم يتم استخلاص أية عناصر من ملف hOCR.\")\n",
        "            else:\n",
        "                print(f\"\\n--- تم استخلاص {len(extracted_data)} عنصرًا ---\")\n",
        "                for i, element in enumerate(extracted_data):\n",
        "                    print(f\"العنصر {i+1}: النوع={element['type']}, الإحداثيات={element['bbox']}\")\n",
        "                    if element['type'] == 'text':\n",
        "                        print(f\"  النص (مقتطف): {element['text'][:100]}...\") # طباعة جزء من النص\n",
        "                print(\"------------------------------\\n\")\n",
        "\n",
        "                # 3. إنشاء ملف وورد\n",
        "                word_output_filename = \"output_from_docling.docx\"\n",
        "                create_word_from_elements(input_image_path, extracted_data, word_output_filename)\n",
        "\n",
        "                # 4. إنشاء ملف ماركداون\n",
        "                # لكي تعمل الصورة في الماركداون، تأكد من نسخها إلى جانب ملف الماركداون\n",
        "                # أو عدّل المسار في create_markdown_from_elements\n",
        "                # مثال لنسخ الصورة:\n",
        "                # target_md_image_path = os.path.join(os.getcwd(), os.path.basename(input_image_path))\n",
        "                # if not os.path.exists(target_md_image_path):\n",
        "                #    shutil.copy(input_image_path, target_md_image_path)\n",
        "\n",
        "                markdown_output_filename = \"output_from_docling.md\"\n",
        "                create_markdown_from_elements(input_image_path, extracted_data, markdown_output_filename)\n",
        "        else:\n",
        "            print(\"فشل Docling OCR أو لم يتم إنتاج ملف hOCR.\")\n",
        "\n",
        "    # تنظيف مجلد المخرجات إذا أردت\n",
        "    # if os.path.exists(DOCLING_OUTPUT_DIR):\n",
        "    #     shutil.rmtree(DOCLING_OUTPUT_DIR)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "dHyeouOhF6ug",
        "outputId": "6ac6bb0e-e487-40be-f540-8129f8eca38b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "تنفيذ أمر Docling: docling ocr /content/PVC-Structure-PhotoRoom.png-PhotoRoom.png docling_output\n",
            "خطأ أثناء تنفيذ Docling:\n",
            "STDOUT: \n",
            "STDERR: 2025-05-10 21:30:16.275754: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746912616.301071   29864 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746912616.310147   29864 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "2025-05-10 21:30:16.335290: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
            "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
            "Error: The input file ocr does not exist.\n",
            "Aborted.\n",
            "\n",
            "فشل Docling OCR أو لم يتم إنتاج ملف hOCR.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install docling --upgrade  # Upgrade to the latest version of docling\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "CXoyh1zvGSH-",
        "outputId": "04351e7a-9f43-4a2f-eefa-25a4be88968a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: docling in /usr/local/lib/python3.11/dist-packages (2.31.0)\n",
            "Requirement already satisfied: beautifulsoup4<5.0.0,>=4.12.3 in /usr/local/lib/python3.11/dist-packages (from docling) (4.13.4)\n",
            "Requirement already satisfied: certifi>=2024.7.4 in /usr/local/lib/python3.11/dist-packages (from docling) (2025.4.26)\n",
            "Requirement already satisfied: docling-core<3.0.0,>=2.26.0 in /usr/local/lib/python3.11/dist-packages (from docling-core[chunking]<3.0.0,>=2.26.0->docling) (2.30.0)\n",
            "Requirement already satisfied: docling-ibm-models<4.0.0,>=3.4.0 in /usr/local/lib/python3.11/dist-packages (from docling) (3.4.3)\n",
            "Requirement already satisfied: docling-parse<5.0.0,>=4.0.0 in /usr/local/lib/python3.11/dist-packages (from docling) (4.0.1)\n",
            "Requirement already satisfied: easyocr<2.0,>=1.7 in /usr/local/lib/python3.11/dist-packages (from docling) (1.7.2)\n",
            "Requirement already satisfied: filetype<2.0.0,>=1.2.0 in /usr/local/lib/python3.11/dist-packages (from docling) (1.2.0)\n",
            "Requirement already satisfied: huggingface_hub<1,>=0.23 in /usr/local/lib/python3.11/dist-packages (from docling) (0.30.2)\n",
            "Requirement already satisfied: lxml<6.0.0,>=4.0.0 in /usr/local/lib/python3.11/dist-packages (from docling) (5.4.0)\n",
            "Requirement already satisfied: marko<3.0.0,>=2.1.2 in /usr/local/lib/python3.11/dist-packages (from docling) (2.1.3)\n",
            "Requirement already satisfied: openpyxl<4.0.0,>=3.1.5 in /usr/local/lib/python3.11/dist-packages (from docling) (3.1.5)\n",
            "Requirement already satisfied: pandas<3.0.0,>=2.1.4 in /usr/local/lib/python3.11/dist-packages (from docling) (2.2.2)\n",
            "Requirement already satisfied: pillow<12.0.0,>=10.0.0 in /usr/local/lib/python3.11/dist-packages (from docling) (11.2.1)\n",
            "Requirement already satisfied: pluggy<2.0.0,>=1.0.0 in /usr/local/lib/python3.11/dist-packages (from docling) (1.5.0)\n",
            "Requirement already satisfied: pydantic<3.0.0,>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from docling) (2.11.4)\n",
            "Requirement already satisfied: pydantic-settings<3.0.0,>=2.3.0 in /usr/local/lib/python3.11/dist-packages (from docling) (2.9.1)\n",
            "Requirement already satisfied: pylatexenc<3.0,>=2.10 in /usr/local/lib/python3.11/dist-packages (from docling) (2.10)\n",
            "Requirement already satisfied: pypdfium2<5.0.0,>=4.30.0 in /usr/local/lib/python3.11/dist-packages (from docling) (4.30.1)\n",
            "Requirement already satisfied: python-docx<2.0.0,>=1.1.2 in /usr/local/lib/python3.11/dist-packages (from docling) (1.1.2)\n",
            "Requirement already satisfied: python-pptx<2.0.0,>=1.0.2 in /usr/local/lib/python3.11/dist-packages (from docling) (1.0.2)\n",
            "Requirement already satisfied: requests<3.0.0,>=2.32.2 in /usr/local/lib/python3.11/dist-packages (from docling) (2.32.3)\n",
            "Requirement already satisfied: rtree<2.0.0,>=1.3.0 in /usr/local/lib/python3.11/dist-packages (from docling) (1.4.0)\n",
            "Requirement already satisfied: scipy<2.0.0,>=1.6.0 in /usr/local/lib/python3.11/dist-packages (from docling) (1.15.2)\n",
            "Requirement already satisfied: tqdm<5.0.0,>=4.65.0 in /usr/local/lib/python3.11/dist-packages (from docling) (4.67.1)\n",
            "Requirement already satisfied: typer<0.16.0,>=0.12.5 in /usr/local/lib/python3.11/dist-packages (from docling) (0.15.3)\n",
            "Requirement already satisfied: soupsieve>1.2 in /usr/local/lib/python3.11/dist-packages (from beautifulsoup4<5.0.0,>=4.12.3->docling) (2.7)\n",
            "Requirement already satisfied: typing-extensions>=4.0.0 in /usr/local/lib/python3.11/dist-packages (from beautifulsoup4<5.0.0,>=4.12.3->docling) (4.13.2)\n",
            "Requirement already satisfied: jsonref<2.0.0,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (1.1.0)\n",
            "Requirement already satisfied: jsonschema<5.0.0,>=4.16.0 in /usr/local/lib/python3.11/dist-packages (from docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (4.23.0)\n",
            "Requirement already satisfied: latex2mathml<4.0.0,>=3.77.0 in /usr/local/lib/python3.11/dist-packages (from docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (3.78.0)\n",
            "Requirement already satisfied: pyyaml<7.0.0,>=5.1 in /usr/local/lib/python3.11/dist-packages (from docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (6.0.2)\n",
            "Requirement already satisfied: tabulate<0.10.0,>=0.9.0 in /usr/local/lib/python3.11/dist-packages (from docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.9.0)\n",
            "Requirement already satisfied: semchunk<3.0.0,>=2.2.0 in /usr/local/lib/python3.11/dist-packages (from docling-core[chunking]<3.0.0,>=2.26.0->docling) (2.2.2)\n",
            "Requirement already satisfied: transformers<5.0.0,>=4.34.0 in /usr/local/lib/python3.11/dist-packages (from docling-core[chunking]<3.0.0,>=2.26.0->docling) (4.51.3)\n",
            "Requirement already satisfied: jsonlines<4.0.0,>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from docling-ibm-models<4.0.0,>=3.4.0->docling) (3.1.0)\n",
            "Requirement already satisfied: numpy<3.0.0,>=1.24.4 in /usr/local/lib/python3.11/dist-packages (from docling-ibm-models<4.0.0,>=3.4.0->docling) (2.0.2)\n",
            "Requirement already satisfied: opencv-python-headless<5.0.0.0,>=4.6.0.66 in /usr/local/lib/python3.11/dist-packages (from docling-ibm-models<4.0.0,>=3.4.0->docling) (4.11.0.86)\n",
            "Requirement already satisfied: safetensors<1,>=0.4.3 in /usr/local/lib/python3.11/dist-packages (from safetensors[torch]<1,>=0.4.3->docling-ibm-models<4.0.0,>=3.4.0->docling) (0.5.3)\n",
            "Requirement already satisfied: torch<3.0.0,>=2.2.2 in /usr/local/lib/python3.11/dist-packages (from docling-ibm-models<4.0.0,>=3.4.0->docling) (2.6.0+cu124)\n",
            "Requirement already satisfied: torchvision<1,>=0 in /usr/local/lib/python3.11/dist-packages (from docling-ibm-models<4.0.0,>=3.4.0->docling) (0.21.0+cu124)\n",
            "Requirement already satisfied: scikit-image in /usr/local/lib/python3.11/dist-packages (from easyocr<2.0,>=1.7->docling) (0.25.2)\n",
            "Requirement already satisfied: python-bidi in /usr/local/lib/python3.11/dist-packages (from easyocr<2.0,>=1.7->docling) (0.6.6)\n",
            "Requirement already satisfied: Shapely in /usr/local/lib/python3.11/dist-packages (from easyocr<2.0,>=1.7->docling) (2.1.0)\n",
            "Requirement already satisfied: pyclipper in /usr/local/lib/python3.11/dist-packages (from easyocr<2.0,>=1.7->docling) (1.3.0.post6)\n",
            "Requirement already satisfied: ninja in /usr/local/lib/python3.11/dist-packages (from easyocr<2.0,>=1.7->docling) (1.11.1.4)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.11/dist-packages (from huggingface_hub<1,>=0.23->docling) (3.18.0)\n",
            "Requirement already satisfied: fsspec>=2023.5.0 in /usr/local/lib/python3.11/dist-packages (from huggingface_hub<1,>=0.23->docling) (2025.3.2)\n",
            "Requirement already satisfied: packaging>=20.9 in /usr/local/lib/python3.11/dist-packages (from huggingface_hub<1,>=0.23->docling) (24.2)\n",
            "Requirement already satisfied: et-xmlfile in /usr/local/lib/python3.11/dist-packages (from openpyxl<4.0.0,>=3.1.5->docling) (2.0.0)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas<3.0.0,>=2.1.4->docling) (2.9.0.post0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas<3.0.0,>=2.1.4->docling) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas<3.0.0,>=2.1.4->docling) (2025.2)\n",
            "Requirement already satisfied: annotated-types>=0.6.0 in /usr/local/lib/python3.11/dist-packages (from pydantic<3.0.0,>=2.0.0->docling) (0.7.0)\n",
            "Requirement already satisfied: pydantic-core==2.33.2 in /usr/local/lib/python3.11/dist-packages (from pydantic<3.0.0,>=2.0.0->docling) (2.33.2)\n",
            "Requirement already satisfied: typing-inspection>=0.4.0 in /usr/local/lib/python3.11/dist-packages (from pydantic<3.0.0,>=2.0.0->docling) (0.4.0)\n",
            "Requirement already satisfied: python-dotenv>=0.21.0 in /usr/local/lib/python3.11/dist-packages (from pydantic-settings<3.0.0,>=2.3.0->docling) (1.1.0)\n",
            "Requirement already satisfied: XlsxWriter>=0.5.7 in /usr/local/lib/python3.11/dist-packages (from python-pptx<2.0.0,>=1.0.2->docling) (3.2.3)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests<3.0.0,>=2.32.2->docling) (3.4.1)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests<3.0.0,>=2.32.2->docling) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests<3.0.0,>=2.32.2->docling) (2.4.0)\n",
            "Requirement already satisfied: click>=8.0.0 in /usr/local/lib/python3.11/dist-packages (from typer<0.16.0,>=0.12.5->docling) (8.1.8)\n",
            "Requirement already satisfied: shellingham>=1.3.0 in /usr/local/lib/python3.11/dist-packages (from typer<0.16.0,>=0.12.5->docling) (1.5.4)\n",
            "Requirement already satisfied: rich>=10.11.0 in /usr/local/lib/python3.11/dist-packages (from typer<0.16.0,>=0.12.5->docling) (13.9.4)\n",
            "Requirement already satisfied: attrs>=19.2.0 in /usr/local/lib/python3.11/dist-packages (from jsonlines<4.0.0,>=3.1.0->docling-ibm-models<4.0.0,>=3.4.0->docling) (25.3.0)\n",
            "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /usr/local/lib/python3.11/dist-packages (from jsonschema<5.0.0,>=4.16.0->docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (2025.4.1)\n",
            "Requirement already satisfied: referencing>=0.28.4 in /usr/local/lib/python3.11/dist-packages (from jsonschema<5.0.0,>=4.16.0->docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.36.2)\n",
            "Requirement already satisfied: rpds-py>=0.7.1 in /usr/local/lib/python3.11/dist-packages (from jsonschema<5.0.0,>=4.16.0->docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.24.0)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas<3.0.0,>=2.1.4->docling) (1.17.0)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.11/dist-packages (from rich>=10.11.0->typer<0.16.0,>=0.12.5->docling) (3.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.11/dist-packages (from rich>=10.11.0->typer<0.16.0,>=0.12.5->docling) (2.19.1)\n",
            "Requirement already satisfied: mpire[dill] in /usr/local/lib/python3.11/dist-packages (from semchunk<3.0.0,>=2.2.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (2.10.2)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (3.4.2)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (3.1.6)\n",
            "Requirement already satisfied: nvidia-cuda-nvrtc-cu12==12.4.127 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (12.4.127)\n",
            "Requirement already satisfied: nvidia-cuda-runtime-cu12==12.4.127 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (12.4.127)\n",
            "Requirement already satisfied: nvidia-cuda-cupti-cu12==12.4.127 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (12.4.127)\n",
            "Requirement already satisfied: nvidia-cudnn-cu12==9.1.0.70 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (9.1.0.70)\n",
            "Requirement already satisfied: nvidia-cublas-cu12==12.4.5.8 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (12.4.5.8)\n",
            "Requirement already satisfied: nvidia-cufft-cu12==11.2.1.3 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (11.2.1.3)\n",
            "Requirement already satisfied: nvidia-curand-cu12==10.3.5.147 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (10.3.5.147)\n",
            "Requirement already satisfied: nvidia-cusolver-cu12==11.6.1.9 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (11.6.1.9)\n",
            "Requirement already satisfied: nvidia-cusparse-cu12==12.3.1.170 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (12.3.1.170)\n",
            "Requirement already satisfied: nvidia-cusparselt-cu12==0.6.2 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (0.6.2)\n",
            "Requirement already satisfied: nvidia-nccl-cu12==2.21.5 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (2.21.5)\n",
            "Requirement already satisfied: nvidia-nvtx-cu12==12.4.127 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (12.4.127)\n",
            "Requirement already satisfied: nvidia-nvjitlink-cu12==12.4.127 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (12.4.127)\n",
            "Requirement already satisfied: triton==3.2.0 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (3.2.0)\n",
            "Requirement already satisfied: sympy==1.13.1 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (1.13.1)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from sympy==1.13.1->torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (1.3.0)\n",
            "Requirement already satisfied: regex!=2019.12.17 in /usr/local/lib/python3.11/dist-packages (from transformers<5.0.0,>=4.34.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (2024.11.6)\n",
            "Requirement already satisfied: tokenizers<0.22,>=0.21 in /usr/local/lib/python3.11/dist-packages (from transformers<5.0.0,>=4.34.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.21.1)\n",
            "Requirement already satisfied: imageio!=2.35.0,>=2.33 in /usr/local/lib/python3.11/dist-packages (from scikit-image->easyocr<2.0,>=1.7->docling) (2.37.0)\n",
            "Requirement already satisfied: tifffile>=2022.8.12 in /usr/local/lib/python3.11/dist-packages (from scikit-image->easyocr<2.0,>=1.7->docling) (2025.3.30)\n",
            "Requirement already satisfied: lazy-loader>=0.4 in /usr/local/lib/python3.11/dist-packages (from scikit-image->easyocr<2.0,>=1.7->docling) (0.4)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.11/dist-packages (from markdown-it-py>=2.2.0->rich>=10.11.0->typer<0.16.0,>=0.12.5->docling) (0.1.2)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from jinja2->torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (3.0.2)\n",
            "Requirement already satisfied: multiprocess>=0.70.15 in /usr/local/lib/python3.11/dist-packages (from mpire[dill]->semchunk<3.0.0,>=2.2.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.70.18)\n",
            "Requirement already satisfied: dill>=0.4.0 in /usr/local/lib/python3.11/dist-packages (from multiprocess>=0.70.15->mpire[dill]->semchunk<3.0.0,>=2.2.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.4.0)\n"
          ]
        }
      ]
    },
    {
      "source": [
        "\n",
        "import subprocess\n",
        "import os\n",
        "import shutil\n",
        "from PIL import Image\n",
        "from docx import Document\n",
        "from docx.shared import Inches\n",
        "from docling.parser import HOCRParser # Import HOCRParser from docling.parser\n",
        "\n",
        "# ... (Rest of the code remains the same) ..."
      ],
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 384
        },
        "id": "1CA2ONXMGP7u",
        "outputId": "9727a5f5-09e1-4085-c29a-d568db77dc15"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "ModuleNotFoundError",
          "evalue": "No module named 'docling.parser'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-4-fb20ff847bed>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mdocx\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mDocument\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mdocx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshared\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mInches\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mdocling\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparser\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mHOCRParser\u001b[0m \u001b[0;31m# Import HOCRParser from docling.parser\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;31m# ... (Rest of the code remains the same) ...\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'docling.parser'",
            "",
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0;32m\nNOTE: If your import is failing due to a missing package, you can\nmanually install dependencies using either !pip or !apt.\n\nTo view examples of installing some common dependencies, click the\n\"Open Examples\" button below.\n\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n"
          ],
          "errorDetails": {
            "actions": [
              {
                "action": "open_url",
                "actionText": "Open Examples",
                "url": "/notebooks/snippets/importing_libraries.ipynb"
              }
            ]
          }
        }
      ]
    },
    {
      "source": [
        "!pip install docling --upgrade  # Upgrade to the latest version of docling\n",
        "\n",
        "import subprocess\n",
        "import os\n",
        "import shutil\n",
        "from PIL import Image\n",
        "from docx import Document\n",
        "from docx.shared import Inches\n",
        "from docling import HOCRParser # Import HOCRParser from docling directly\n",
        "\n",
        "# --- إعدادات ---\n",
        "DOCLING_OUTPUT_DIR = \"docling_output\"\n",
        "# ... (Rest of your code remains the same) ..."
      ],
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "cQvrzPfIGoq4",
        "outputId": "76f6f692-3f40-4cad-eee6-b87925e9d7d3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: docling in /usr/local/lib/python3.11/dist-packages (2.31.0)\n",
            "Requirement already satisfied: beautifulsoup4<5.0.0,>=4.12.3 in /usr/local/lib/python3.11/dist-packages (from docling) (4.13.4)\n",
            "Requirement already satisfied: certifi>=2024.7.4 in /usr/local/lib/python3.11/dist-packages (from docling) (2025.4.26)\n",
            "Requirement already satisfied: docling-core<3.0.0,>=2.26.0 in /usr/local/lib/python3.11/dist-packages (from docling-core[chunking]<3.0.0,>=2.26.0->docling) (2.30.0)\n",
            "Requirement already satisfied: docling-ibm-models<4.0.0,>=3.4.0 in /usr/local/lib/python3.11/dist-packages (from docling) (3.4.3)\n",
            "Requirement already satisfied: docling-parse<5.0.0,>=4.0.0 in /usr/local/lib/python3.11/dist-packages (from docling) (4.0.1)\n",
            "Requirement already satisfied: easyocr<2.0,>=1.7 in /usr/local/lib/python3.11/dist-packages (from docling) (1.7.2)\n",
            "Requirement already satisfied: filetype<2.0.0,>=1.2.0 in /usr/local/lib/python3.11/dist-packages (from docling) (1.2.0)\n",
            "Requirement already satisfied: huggingface_hub<1,>=0.23 in /usr/local/lib/python3.11/dist-packages (from docling) (0.30.2)\n",
            "Requirement already satisfied: lxml<6.0.0,>=4.0.0 in /usr/local/lib/python3.11/dist-packages (from docling) (5.4.0)\n",
            "Requirement already satisfied: marko<3.0.0,>=2.1.2 in /usr/local/lib/python3.11/dist-packages (from docling) (2.1.3)\n",
            "Requirement already satisfied: openpyxl<4.0.0,>=3.1.5 in /usr/local/lib/python3.11/dist-packages (from docling) (3.1.5)\n",
            "Requirement already satisfied: pandas<3.0.0,>=2.1.4 in /usr/local/lib/python3.11/dist-packages (from docling) (2.2.2)\n",
            "Requirement already satisfied: pillow<12.0.0,>=10.0.0 in /usr/local/lib/python3.11/dist-packages (from docling) (11.2.1)\n",
            "Requirement already satisfied: pluggy<2.0.0,>=1.0.0 in /usr/local/lib/python3.11/dist-packages (from docling) (1.5.0)\n",
            "Requirement already satisfied: pydantic<3.0.0,>=2.0.0 in /usr/local/lib/python3.11/dist-packages (from docling) (2.11.4)\n",
            "Requirement already satisfied: pydantic-settings<3.0.0,>=2.3.0 in /usr/local/lib/python3.11/dist-packages (from docling) (2.9.1)\n",
            "Requirement already satisfied: pylatexenc<3.0,>=2.10 in /usr/local/lib/python3.11/dist-packages (from docling) (2.10)\n",
            "Requirement already satisfied: pypdfium2<5.0.0,>=4.30.0 in /usr/local/lib/python3.11/dist-packages (from docling) (4.30.1)\n",
            "Requirement already satisfied: python-docx<2.0.0,>=1.1.2 in /usr/local/lib/python3.11/dist-packages (from docling) (1.1.2)\n",
            "Requirement already satisfied: python-pptx<2.0.0,>=1.0.2 in /usr/local/lib/python3.11/dist-packages (from docling) (1.0.2)\n",
            "Requirement already satisfied: requests<3.0.0,>=2.32.2 in /usr/local/lib/python3.11/dist-packages (from docling) (2.32.3)\n",
            "Requirement already satisfied: rtree<2.0.0,>=1.3.0 in /usr/local/lib/python3.11/dist-packages (from docling) (1.4.0)\n",
            "Requirement already satisfied: scipy<2.0.0,>=1.6.0 in /usr/local/lib/python3.11/dist-packages (from docling) (1.15.2)\n",
            "Requirement already satisfied: tqdm<5.0.0,>=4.65.0 in /usr/local/lib/python3.11/dist-packages (from docling) (4.67.1)\n",
            "Requirement already satisfied: typer<0.16.0,>=0.12.5 in /usr/local/lib/python3.11/dist-packages (from docling) (0.15.3)\n",
            "Requirement already satisfied: soupsieve>1.2 in /usr/local/lib/python3.11/dist-packages (from beautifulsoup4<5.0.0,>=4.12.3->docling) (2.7)\n",
            "Requirement already satisfied: typing-extensions>=4.0.0 in /usr/local/lib/python3.11/dist-packages (from beautifulsoup4<5.0.0,>=4.12.3->docling) (4.13.2)\n",
            "Requirement already satisfied: jsonref<2.0.0,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (1.1.0)\n",
            "Requirement already satisfied: jsonschema<5.0.0,>=4.16.0 in /usr/local/lib/python3.11/dist-packages (from docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (4.23.0)\n",
            "Requirement already satisfied: latex2mathml<4.0.0,>=3.77.0 in /usr/local/lib/python3.11/dist-packages (from docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (3.78.0)\n",
            "Requirement already satisfied: pyyaml<7.0.0,>=5.1 in /usr/local/lib/python3.11/dist-packages (from docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (6.0.2)\n",
            "Requirement already satisfied: tabulate<0.10.0,>=0.9.0 in /usr/local/lib/python3.11/dist-packages (from docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.9.0)\n",
            "Requirement already satisfied: semchunk<3.0.0,>=2.2.0 in /usr/local/lib/python3.11/dist-packages (from docling-core[chunking]<3.0.0,>=2.26.0->docling) (2.2.2)\n",
            "Requirement already satisfied: transformers<5.0.0,>=4.34.0 in /usr/local/lib/python3.11/dist-packages (from docling-core[chunking]<3.0.0,>=2.26.0->docling) (4.51.3)\n",
            "Requirement already satisfied: jsonlines<4.0.0,>=3.1.0 in /usr/local/lib/python3.11/dist-packages (from docling-ibm-models<4.0.0,>=3.4.0->docling) (3.1.0)\n",
            "Requirement already satisfied: numpy<3.0.0,>=1.24.4 in /usr/local/lib/python3.11/dist-packages (from docling-ibm-models<4.0.0,>=3.4.0->docling) (2.0.2)\n",
            "Requirement already satisfied: opencv-python-headless<5.0.0.0,>=4.6.0.66 in /usr/local/lib/python3.11/dist-packages (from docling-ibm-models<4.0.0,>=3.4.0->docling) (4.11.0.86)\n",
            "Requirement already satisfied: safetensors<1,>=0.4.3 in /usr/local/lib/python3.11/dist-packages (from safetensors[torch]<1,>=0.4.3->docling-ibm-models<4.0.0,>=3.4.0->docling) (0.5.3)\n",
            "Requirement already satisfied: torch<3.0.0,>=2.2.2 in /usr/local/lib/python3.11/dist-packages (from docling-ibm-models<4.0.0,>=3.4.0->docling) (2.6.0+cu124)\n",
            "Requirement already satisfied: torchvision<1,>=0 in /usr/local/lib/python3.11/dist-packages (from docling-ibm-models<4.0.0,>=3.4.0->docling) (0.21.0+cu124)\n",
            "Requirement already satisfied: scikit-image in /usr/local/lib/python3.11/dist-packages (from easyocr<2.0,>=1.7->docling) (0.25.2)\n",
            "Requirement already satisfied: python-bidi in /usr/local/lib/python3.11/dist-packages (from easyocr<2.0,>=1.7->docling) (0.6.6)\n",
            "Requirement already satisfied: Shapely in /usr/local/lib/python3.11/dist-packages (from easyocr<2.0,>=1.7->docling) (2.1.0)\n",
            "Requirement already satisfied: pyclipper in /usr/local/lib/python3.11/dist-packages (from easyocr<2.0,>=1.7->docling) (1.3.0.post6)\n",
            "Requirement already satisfied: ninja in /usr/local/lib/python3.11/dist-packages (from easyocr<2.0,>=1.7->docling) (1.11.1.4)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.11/dist-packages (from huggingface_hub<1,>=0.23->docling) (3.18.0)\n",
            "Requirement already satisfied: fsspec>=2023.5.0 in /usr/local/lib/python3.11/dist-packages (from huggingface_hub<1,>=0.23->docling) (2025.3.2)\n",
            "Requirement already satisfied: packaging>=20.9 in /usr/local/lib/python3.11/dist-packages (from huggingface_hub<1,>=0.23->docling) (24.2)\n",
            "Requirement already satisfied: et-xmlfile in /usr/local/lib/python3.11/dist-packages (from openpyxl<4.0.0,>=3.1.5->docling) (2.0.0)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas<3.0.0,>=2.1.4->docling) (2.9.0.post0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas<3.0.0,>=2.1.4->docling) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas<3.0.0,>=2.1.4->docling) (2025.2)\n",
            "Requirement already satisfied: annotated-types>=0.6.0 in /usr/local/lib/python3.11/dist-packages (from pydantic<3.0.0,>=2.0.0->docling) (0.7.0)\n",
            "Requirement already satisfied: pydantic-core==2.33.2 in /usr/local/lib/python3.11/dist-packages (from pydantic<3.0.0,>=2.0.0->docling) (2.33.2)\n",
            "Requirement already satisfied: typing-inspection>=0.4.0 in /usr/local/lib/python3.11/dist-packages (from pydantic<3.0.0,>=2.0.0->docling) (0.4.0)\n",
            "Requirement already satisfied: python-dotenv>=0.21.0 in /usr/local/lib/python3.11/dist-packages (from pydantic-settings<3.0.0,>=2.3.0->docling) (1.1.0)\n",
            "Requirement already satisfied: XlsxWriter>=0.5.7 in /usr/local/lib/python3.11/dist-packages (from python-pptx<2.0.0,>=1.0.2->docling) (3.2.3)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests<3.0.0,>=2.32.2->docling) (3.4.1)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests<3.0.0,>=2.32.2->docling) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests<3.0.0,>=2.32.2->docling) (2.4.0)\n",
            "Requirement already satisfied: click>=8.0.0 in /usr/local/lib/python3.11/dist-packages (from typer<0.16.0,>=0.12.5->docling) (8.1.8)\n",
            "Requirement already satisfied: shellingham>=1.3.0 in /usr/local/lib/python3.11/dist-packages (from typer<0.16.0,>=0.12.5->docling) (1.5.4)\n",
            "Requirement already satisfied: rich>=10.11.0 in /usr/local/lib/python3.11/dist-packages (from typer<0.16.0,>=0.12.5->docling) (13.9.4)\n",
            "Requirement already satisfied: attrs>=19.2.0 in /usr/local/lib/python3.11/dist-packages (from jsonlines<4.0.0,>=3.1.0->docling-ibm-models<4.0.0,>=3.4.0->docling) (25.3.0)\n",
            "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /usr/local/lib/python3.11/dist-packages (from jsonschema<5.0.0,>=4.16.0->docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (2025.4.1)\n",
            "Requirement already satisfied: referencing>=0.28.4 in /usr/local/lib/python3.11/dist-packages (from jsonschema<5.0.0,>=4.16.0->docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.36.2)\n",
            "Requirement already satisfied: rpds-py>=0.7.1 in /usr/local/lib/python3.11/dist-packages (from jsonschema<5.0.0,>=4.16.0->docling-core<3.0.0,>=2.26.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.24.0)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas<3.0.0,>=2.1.4->docling) (1.17.0)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.11/dist-packages (from rich>=10.11.0->typer<0.16.0,>=0.12.5->docling) (3.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.11/dist-packages (from rich>=10.11.0->typer<0.16.0,>=0.12.5->docling) (2.19.1)\n",
            "Requirement already satisfied: mpire[dill] in /usr/local/lib/python3.11/dist-packages (from semchunk<3.0.0,>=2.2.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (2.10.2)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (3.4.2)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (3.1.6)\n",
            "Requirement already satisfied: nvidia-cuda-nvrtc-cu12==12.4.127 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (12.4.127)\n",
            "Requirement already satisfied: nvidia-cuda-runtime-cu12==12.4.127 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (12.4.127)\n",
            "Requirement already satisfied: nvidia-cuda-cupti-cu12==12.4.127 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (12.4.127)\n",
            "Requirement already satisfied: nvidia-cudnn-cu12==9.1.0.70 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (9.1.0.70)\n",
            "Requirement already satisfied: nvidia-cublas-cu12==12.4.5.8 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (12.4.5.8)\n",
            "Requirement already satisfied: nvidia-cufft-cu12==11.2.1.3 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (11.2.1.3)\n",
            "Requirement already satisfied: nvidia-curand-cu12==10.3.5.147 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (10.3.5.147)\n",
            "Requirement already satisfied: nvidia-cusolver-cu12==11.6.1.9 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (11.6.1.9)\n",
            "Requirement already satisfied: nvidia-cusparse-cu12==12.3.1.170 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (12.3.1.170)\n",
            "Requirement already satisfied: nvidia-cusparselt-cu12==0.6.2 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (0.6.2)\n",
            "Requirement already satisfied: nvidia-nccl-cu12==2.21.5 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (2.21.5)\n",
            "Requirement already satisfied: nvidia-nvtx-cu12==12.4.127 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (12.4.127)\n",
            "Requirement already satisfied: nvidia-nvjitlink-cu12==12.4.127 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (12.4.127)\n",
            "Requirement already satisfied: triton==3.2.0 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (3.2.0)\n",
            "Requirement already satisfied: sympy==1.13.1 in /usr/local/lib/python3.11/dist-packages (from torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (1.13.1)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.11/dist-packages (from sympy==1.13.1->torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (1.3.0)\n",
            "Requirement already satisfied: regex!=2019.12.17 in /usr/local/lib/python3.11/dist-packages (from transformers<5.0.0,>=4.34.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (2024.11.6)\n",
            "Requirement already satisfied: tokenizers<0.22,>=0.21 in /usr/local/lib/python3.11/dist-packages (from transformers<5.0.0,>=4.34.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.21.1)\n",
            "Requirement already satisfied: imageio!=2.35.0,>=2.33 in /usr/local/lib/python3.11/dist-packages (from scikit-image->easyocr<2.0,>=1.7->docling) (2.37.0)\n",
            "Requirement already satisfied: tifffile>=2022.8.12 in /usr/local/lib/python3.11/dist-packages (from scikit-image->easyocr<2.0,>=1.7->docling) (2025.3.30)\n",
            "Requirement already satisfied: lazy-loader>=0.4 in /usr/local/lib/python3.11/dist-packages (from scikit-image->easyocr<2.0,>=1.7->docling) (0.4)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.11/dist-packages (from markdown-it-py>=2.2.0->rich>=10.11.0->typer<0.16.0,>=0.12.5->docling) (0.1.2)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.11/dist-packages (from jinja2->torch<3.0.0,>=2.2.2->docling-ibm-models<4.0.0,>=3.4.0->docling) (3.0.2)\n",
            "Requirement already satisfied: multiprocess>=0.70.15 in /usr/local/lib/python3.11/dist-packages (from mpire[dill]->semchunk<3.0.0,>=2.2.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.70.18)\n",
            "Requirement already satisfied: dill>=0.4.0 in /usr/local/lib/python3.11/dist-packages (from multiprocess>=0.70.15->mpire[dill]->semchunk<3.0.0,>=2.2.0->docling-core[chunking]<3.0.0,>=2.26.0->docling) (0.4.0)\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "ImportError",
          "evalue": "cannot import name 'HOCRParser' from 'docling' (/usr/local/lib/python3.11/dist-packages/docling/__init__.py)",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-5-d6ac8b7868e4>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mdocx\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mDocument\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mdocx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshared\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mInches\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mdocling\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mHOCRParser\u001b[0m \u001b[0;31m# Import HOCRParser from docling directly\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;31m# --- إعدادات ---\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mImportError\u001b[0m: cannot import name 'HOCRParser' from 'docling' (/usr/local/lib/python3.11/dist-packages/docling/__init__.py)",
            "",
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0;32m\nNOTE: If your import is failing due to a missing package, you can\nmanually install dependencies using either !pip or !apt.\n\nTo view examples of installing some common dependencies, click the\n\"Open Examples\" button below.\n\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n"
          ],
          "errorDetails": {
            "actions": [
              {
                "action": "open_url",
                "actionText": "Open Examples",
                "url": "/notebooks/snippets/importing_libraries.ipynb"
              }
            ]
          }
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def run_docling_ocr(image_path):\n",
        "    \"\"\"\n",
        "    ينفذ أمر docling ocr على الصورة ويخزن المخرجات.\n",
        "    \"\"\"\n",
        "    if not os.path.exists(image_path): # تحقق إضافي للتأكد من وجود ملف الصورة\n",
        "        print(f\"خطأ: ملف الصورة المدخل '{image_path}' غير موجود.\")\n",
        "        return None\n",
        "\n",
        "    if os.path.exists(DOCLING_OUTPUT_DIR):\n",
        "        shutil.rmtree(DOCLING_OUTPUT_DIR)\n",
        "    os.makedirs(DOCLING_OUTPUT_DIR, exist_ok=True)\n",
        "\n",
        "    # --- التعديل هنا ---\n",
        "    # استدعاء docling كوحدة بايثون\n",
        "    cmd = [\"python\", \"-m\", \"docling\", \"ocr\", image_path, DOCLING_OUTPUT_DIR]\n",
        "    # --------------------\n",
        "\n",
        "    print(f\"تنفيذ أمر Docling: {' '.join(cmd)}\")\n",
        "    try:\n",
        "        # استخدام subprocess.run للحصول على تحكم أفضل وانتظار اكتمال الأمر\n",
        "        process = subprocess.run(cmd, capture_output=True, text=True, check=False, encoding='utf-8', errors='ignore')\n",
        "\n",
        "        if process.returncode != 0:\n",
        "            print(f\"خطأ أثناء تنفيذ Docling (رمز الخروج: {process.returncode}):\")\n",
        "            if process.stdout:\n",
        "                print(f\"STDOUT:\\n{process.stdout}\")\n",
        "            if process.stderr:\n",
        "                print(f\"STDERR:\\n{process.stderr}\") # stderr غالبًا ما يحتوي على معلومات الخطأ الفعلية\n",
        "            return None\n",
        "        else:\n",
        "            if process.stdout:\n",
        "                 print(f\"Docling STDOUT:\\n{process.stdout}\")\n",
        "            if process.stderr: # حتى مع النجاح، قد تكون هناك تحذيرات مفيدة في stderr\n",
        "                print(f\"Docling STDERR (تحذيرات محتملة):\\n{process.stderr}\")\n",
        "            print(f\"Docling أنجز بنجاح. المخرجات في: {DOCLING_OUTPUT_DIR}\")\n",
        "\n",
        "            # البحث عن ملف hOCR الناتج\n",
        "            for fname in os.listdir(DOCLING_OUTPUT_DIR):\n",
        "                if fname.endswith(\".hocr\"):\n",
        "                    return os.path.join(DOCLING_OUTPUT_DIR, fname)\n",
        "            print(\"لم يتم العثور على ملف hOCR في المخرجات.\")\n",
        "            return None\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        # هذا الخطأ يحدث إذا لم يتم العثور على \"python\" نفسه، وهو أمر غير مرجح،\n",
        "        # أو إذا كان \"docling\" غير مثبت كوحدة بشكل صحيح.\n",
        "        print(\"خطأ: لم يتم العثور على أمر 'python' أو أن وحدة 'docling' غير مثبتة بشكل صحيح.\")\n",
        "        print(\"تأكد من تثبيت docling في بيئة بايثون الحالية: pip install docling\")\n",
        "        return None\n",
        "    except Exception as e:\n",
        "        print(f\"حدث خطأ غير متوقع أثناء تشغيل Docling: {e}\")\n",
        "        return None"
      ],
      "metadata": {
        "id": "syMEPBaxIqAt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import subprocess\n",
        "import os\n",
        "import shutil\n",
        "from PIL import Image\n",
        "from docx import Document\n",
        "from docx.shared import Inches\n",
        "#from docling.parser import HOCRParser # Import HOCRParser from docling.parser\n",
        "\n",
        "# --- إعدادات ---\n",
        "DOCLING_OUTPUT_DIR = \"docling_output\"\n",
        "\n",
        "def run_docling_ocr(image_path):\n",
        "    \"\"\"\n",
        "    ينفذ أمر docling ocr على الصورة ويخزن المخرجات.\n",
        "    \"\"\"\n",
        "    if os.path.exists(DOCLING_OUTPUT_DIR):\n",
        "        shutil.rmtree(DOCLING_OUTPUT_DIR) # حذف المخرجات القديمة\n",
        "    os.makedirs(DOCLING_OUTPUT_DIR, exist_ok=True)\n",
        "\n",
        "    # أمر docling الأساسي للـ OCR.\n",
        "    # قد تحتاج لتعديل هذا الأمر بناءً على كيفية تثبيت docling\n",
        "    # أو إذا كنت تريد تحديد محرك OCR معين يدعمه docling.\n",
        "    # بدون تحديد محرك، قد يستخدم docling الإعداد الافتراضي (الذي قد يكون Tesseract).\n",
        "    cmd = [\"docling\", \"ocr\", image_path, DOCLING_OUTPUT_DIR]\n",
        "\n",
        "    print(f\"تنفيذ أمر Docling: {' '.join(cmd)}\")\n",
        "    try:\n",
        "        process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
        "        stdout, stderr = process.communicate()\n",
        "\n",
        "        if process.returncode != 0:\n",
        "            print(f\"خطأ أثناء تنفيذ Docling:\")\n",
        "            print(f\"STDOUT: {stdout.decode('utf-8', errors='ignore')}\")\n",
        "            print(f\"STDERR: {stderr.decode('utf-8', errors='ignore')}\")\n",
        "            return None\n",
        "        else:\n",
        "            print(f\"Docling STDOUT: {stdout.decode('utf-8', errors='ignore')}\")\n",
        "            print(f\"Docling أنجز بنجاح. المخرجات في: {DOCLING_OUTPUT_DIR}\")\n",
        "            # البحث عن ملف hOCR الناتج\n",
        "            for fname in os.listdir(DOCLING_OUTPUT_DIR):\n",
        "                if fname.endswith(\".hocr\"):\n",
        "                    return os.path.join(DOCLING_OUTPUT_DIR, fname)\n",
        "            print(\"لم يتم العثور على ملف hOCR في المخرجات.\")\n",
        "            return None\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(\"خطأ: لم يتم العثور على أمر 'docling'. هل هو مثبت وفي PATH؟\")\n",
        "        return None\n",
        "    except Exception as e:\n",
        "        print(f\"حدث خطأ غير متوقع أثناء تشغيل Docling: {e}\")\n",
        "        return None\n",
        "\n",
        "def parse_hocr_with_docling(hocr_file_path):\n",
        "    \"\"\"\n",
        "    يحلل ملف hOCR باستخدام HOCRParser من docling ويستخرج النصوص وإحداثياتها.\n",
        "    \"\"\"\n",
        "    if not hocr_file_path or not os.path.exists(hocr_file_path):\n",
        "        print(f\"ملف HOCR غير موجود: {hocr_file_path}\")\n",
        "        return []\n",
        "\n",
        "    extracted_elements = [] # قائمة لتخزين (text, bbox)\n",
        "    try:\n",
        "        parser = HOCRParser()\n",
        "        doc_layout = parser.parse(hocr_file_path)\n",
        "\n",
        "        for page in doc_layout.pages:\n",
        "            # يمكننا استخلاص على مستوى الكلمات، الأسطر، أو الكتل\n",
        "            # لاستخلاص كتل نصية أكبر، قد يكون من الأفضل التجميع على مستوى البلوك\n",
        "            for block in page.blocks: # Blocks can be text or image\n",
        "                if block.block_type == \"text\": # أو أي نوع نصي آخر يحدده docling\n",
        "                    block_text = \"\"\n",
        "                    # تجميع النص من الأسطر والكلمات داخل البلوك\n",
        "                    for line in block.lines:\n",
        "                        line_text_parts = []\n",
        "                        for word in line.words:\n",
        "                            line_text_parts.append(word.text)\n",
        "                        block_text += \" \".join(line_text_parts) + \"\\n\" # سطر جديد لكل سطر في البلوك\n",
        "\n",
        "                    if block_text.strip():\n",
        "                         # bbox هو (x1, y1, x2, y2)\n",
        "                        extracted_elements.append({\n",
        "                            \"type\": \"text\",\n",
        "                            \"text\": block_text.strip(),\n",
        "                            \"bbox\": block.bbox\n",
        "                        })\n",
        "                elif block.block_type == \"image\" or block.block_type == \"figure\": # أو أنواع صور أخرى\n",
        "                    # docling قد يحفظ الصور المكتشفة في مجلد المخرجات\n",
        "                    # يمكننا هنا فقط تسجيل إحداثياتها لاستخدامها لاحقًا\n",
        "                    # إذا أردنا قصها من الصورة الأصلية\n",
        "                     extracted_elements.append({\n",
        "                        \"type\": \"image_region\",\n",
        "                        \"bbox\": block.bbox,\n",
        "                        # قد تحتاج لإيجاد مسار الصورة المقصوصة إذا قام docling بحفظها\n",
        "                        # \"image_path\": os.path.join(DOCLING_OUTPUT_DIR, \"images\", ...)\n",
        "                    })\n",
        "\n",
        "\n",
        "        # ترتيب العناصر بناءً على موضعها الرأسي (y1) ثم الأفقي (x1)\n",
        "        extracted_elements.sort(key=lambda el: (el[\"bbox\"][1], el[\"bbox\"][0]))\n",
        "        return extracted_elements\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في تحليل ملف hOCR: {e}\")\n",
        "        return []\n",
        "\n",
        "\n",
        "def create_word_from_elements(original_image_path, elements, output_filename=\"docling_output.docx\"):\n",
        "    \"\"\"\n",
        "    ينشئ ملف وورد. يحاول وضع الصورة الأصلية والنصوص المستخرجة.\n",
        "    الحفاظ على \"العلاقة\" هنا سيكون تقريبيًا.\n",
        "    \"\"\"\n",
        "    doc = Document()\n",
        "    doc.add_heading('مستند تم إنشاؤه بواسطة Docling', level=1)\n",
        "\n",
        "    # الخيار 1: إضافة الصورة الأصلية كاملة أولاً\n",
        "    try:\n",
        "        # الحصول على أبعاد الصورة الأصلية لمعرفة مقياس الإحداثيات\n",
        "        with Image.open(original_image_path) as img:\n",
        "            img_width_px, img_height_px = img.size\n",
        "\n",
        "        doc.add_paragraph(\"الصورة الأصلية:\")\n",
        "        # قد تحتاج لتعديل حجم الصورة ليتناسب مع الصفحة\n",
        "        doc.add_picture(original_image_path, width=Inches(6.0))\n",
        "        doc.add_page_break()\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في إضافة الصورة الأصلية إلى الوورد: {e}\")\n",
        "\n",
        "    doc.add_heading('العناصر المستخرجة بواسطة Docling:', level=2)\n",
        "\n",
        "    # الخيار 2: محاولة إضافة النصوص (و ربما أجزاء الصورة لاحقًا)\n",
        "    # هذا الجزء يعتمد بشدة على كيفية رغبتك في تمثيل \"العلاقة\".\n",
        "    # python-docx لا يسهل وضع العناصر في إحداثيات مطلقة.\n",
        "    # الطريقة الأبسط هي إضافة النصوص كفقرات متتالية.\n",
        "\n",
        "    text_content_combined = \"\"\n",
        "    for element in elements:\n",
        "        if element[\"type\"] == \"text\":\n",
        "            text_content_combined += element[\"text\"] + \"\\n\\n\" # فصل بين الكتل النصية\n",
        "        # التعامل مع element[\"type\"] == \"image_region\" سيكون أكثر تعقيدًا:\n",
        "        # ستحتاج لقص هذا الجزء من الصورة الأصلية وإضافته.\n",
        "        # مثال:\n",
        "        # elif element[\"type\"] == \"image_region\":\n",
        "        #     try:\n",
        "        #         with Image.open(original_image_path) as img:\n",
        "        #             cropped_img = img.crop(element[\"bbox\"])\n",
        "        #             # حفظ الصورة المقصوصة مؤقتًا لإضافتها\n",
        "        #             temp_cropped_path = os.path.join(DOCLING_OUTPUT_DIR, f\"cropped_{element['bbox']}.png\")\n",
        "        #             cropped_img.save(temp_cropped_path)\n",
        "        #             doc.add_picture(temp_cropped_path, width=Inches(2)) # مثال للحجم\n",
        "        #             os.remove(temp_cropped_path) # حذف المؤقتة\n",
        "        #     except Exception as e:\n",
        "        #         print(f\"خطأ في معالجة منطقة الصورة: {e}\")\n",
        "\n",
        "\n",
        "    if text_content_combined:\n",
        "        doc.add_paragraph(text_content_combined)\n",
        "    else:\n",
        "        doc.add_paragraph(\"لم يتم استخراج نصوص أو تم استخراج نصوص فارغة.\")\n",
        "\n",
        "    try:\n",
        "        doc.save(output_filename)\n",
        "        print(f\"تم حفظ ملف الوورد: {output_filename}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في حفظ ملف الوورد: {e}\")\n",
        "\n",
        "\n",
        "def create_markdown_from_elements(original_image_path, elements, output_filename=\"docling_output.md\"):\n",
        "    \"\"\"\n",
        "    ينشئ ملف ماركداون. يدرج الصورة الأصلية والنصوص.\n",
        "    \"\"\"\n",
        "    md_content = f\"# مستند تم إنشاؤه بواسطة Docling\\n\\n\"\n",
        "\n",
        "    # يفترض أن الصورة ستكون في نفس مسار ملف الماركداون أو يتم توفير مسار نسبي صحيح.\n",
        "    # لتبسيط الأمر، سنستخدم اسم الملف الأصلي.\n",
        "    original_image_filename = os.path.basename(original_image_path)\n",
        "\n",
        "    md_content += f\"## الصورة الأصلية\\n\"\n",
        "    md_content += f\"![الصورة الأصلية]({original_image_filename})\\n\\n\"\n",
        "\n",
        "    md_content += f\"## النصوص المستخرجة\\n\"\n",
        "    text_for_md = \"\"\n",
        "    for element in elements:\n",
        "        if element[\"type\"] == \"text\":\n",
        "            # في الماركداون، عرض النصوص ككتل متتالية هو الأسهل\n",
        "            text_for_md += element[\"text\"] + \"\\n\\n---\\n\\n\" # فصل بفاصل أفقي\n",
        "        # يمكن أيضًا التعامل مع image_region هنا إذا تم قصها وحفظها كملفات منفصلة\n",
        "        # elif element[\"type\"] == \"image_region\" and element.get(\"image_path\"):\n",
        "        #     md_content += f\"![جزء من الصورة]({os.path.basename(element['image_path'])})\\n\\n\"\n",
        "\n",
        "\n",
        "    if text_for_md:\n",
        "        md_content += text_for_md\n",
        "    else:\n",
        "        md_content += \"لم يتم استخراج نصوص.\\n\"\n",
        "\n",
        "    try:\n",
        "        with open(output_filename, \"w\", encoding=\"utf-8\") as f:\n",
        "            f.write(md_content)\n",
        "        print(f\"تم حفظ ملف الماركداون: {output_filename}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في حفظ ملف الماركداون: {e}\")\n",
        "\n",
        "# --- البرنامج الرئيسي ---\n",
        "if __name__ == \"__main__\":\n",
        "    # ضع مسار الصورة التي تريد معالجتها هنا\n",
        "    input_image_path = \"/content/PVC-Structure-PhotoRoom.png-PhotoRoom.png\" # <--- قم بتغيير هذا المسار\n",
        "\n",
        "    if not os.path.exists(input_image_path):\n",
        "        print(f\"خطأ: ملف الصورة '{input_image_path}' غير موجود.\")\n",
        "    else:\n",
        "        # 1. تنفيذ Docling OCR\n",
        "        hocr_output_file = run_docling_ocr(input_image_path)\n",
        "\n",
        "        if hocr_output_file:\n",
        "            # 2. تحليل مخرجات hOCR\n",
        "            # إذا كنت تريد استخدام محلل hOCR الخاص بـ docling:\n",
        "            extracted_data = parse_hocr_with_docling(hocr_output_file)\n",
        "\n",
        "            if not extracted_data:\n",
        "                print(\"لم يتم استخلاص أية عناصر من ملف hOCR.\")\n",
        "            else:\n",
        "                print(f\"\\n--- تم استخلاص {len(extracted_data)} عنصرًا ---\")\n",
        "                for i, element in enumerate(extracted_data):\n",
        "                    print(f\"العنصر {i+1}: النوع={element['type']}, الإحداثيات={element['bbox']}\")\n",
        "                    if element['type'] == 'text':\n",
        "                        print(f\"  النص (مقتطف): {element['text'][:100]}...\") # طباعة جزء من النص\n",
        "                print(\"------------------------------\\n\")\n",
        "\n",
        "                # 3. إنشاء ملف وورد\n",
        "                word_output_filename = \"output_from_docling.docx\"\n",
        "                create_word_from_elements(input_image_path, extracted_data, word_output_filename)\n",
        "\n",
        "                # 4. إنشاء ملف ماركداون\n",
        "                # لكي تعمل الصورة في الماركداون، تأكد من نسخها إلى جانب ملف الماركداون\n",
        "                # أو عدّل المسار في create_markdown_from_elements\n",
        "                # مثال لنسخ الصورة:\n",
        "                # target_md_image_path = os.path.join(os.getcwd(), os.path.basename(input_image_path))\n",
        "                # if not os.path.exists(target_md_image_path):\n",
        "                #    shutil.copy(input_image_path, target_md_image_path)\n",
        "\n",
        "                markdown_output_filename = \"output_from_docling.md\"\n",
        "                create_markdown_from_elements(input_image_path, extracted_data, markdown_output_filename)\n",
        "        else:\n",
        "            print(\"فشل Docling OCR أو لم يتم إنتاج ملف hOCR.\")\n",
        "\n",
        "    # تنظيف مجلد المخرجات إذا أردت\n",
        "    # if os.path.exists(DOCLING_OUTPUT_DIR):\n",
        "    #     shutil.rmtree(DOCLING_OUTPUT_DIR)"
      ],
      "metadata": {
        "id": "J04v9d7mIzK9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "7zLMXODfJUjN"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Q4aGH1fkJUgV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import subprocess\n",
        "import os\n",
        "import shutil\n",
        "from PIL import Image\n",
        "from docx import Document\n",
        "from docx.shared import Inches\n",
        "from docling import HOCRParser # لاستخدام محلل hOCR الخاص بـ docling\n",
        "\n",
        "# --- إعدادات ---\n",
        "DOCLING_OUTPUT_DIR = \"docling_output\"\n",
        "\n",
        "def run_docling_ocr(image_path):\n",
        "    \"\"\"\n",
        "    ينفذ أمر docling ocr على الصورة ويخزن المخرجات.\n",
        "    يستخدم \"python -m docling\" لضمان التنفيذ الصحيح.\n",
        "    \"\"\"\n",
        "    if not os.path.exists(image_path):\n",
        "        print(f\"خطأ: ملف الصورة المدخل '{image_path}' غير موجود.\")\n",
        "        return None\n",
        "\n",
        "    if os.path.exists(DOCLING_OUTPUT_DIR):\n",
        "        shutil.rmtree(DOCLING_OUTPUT_DIR) # حذف المخرجات القديمة\n",
        "    os.makedirs(DOCLING_OUTPUT_DIR, exist_ok=True)\n",
        "\n",
        "    # استدعاء docling كوحدة بايثون\n",
        "    cmd = [\"python\", \"-m\", \"docling\", \"ocr\", image_path, DOCLING_OUTPUT_DIR]\n",
        "\n",
        "    print(f\"تنفيذ أمر Docling: {' '.join(cmd)}\")\n",
        "    try:\n",
        "        # استخدام subprocess.run للحصول على تحكم أفضل وانتظار اكتمال الأمر\n",
        "        process = subprocess.run(cmd, capture_output=True, text=True, check=False, encoding='utf-8', errors='ignore')\n",
        "\n",
        "        if process.returncode != 0:\n",
        "            print(f\"خطأ أثناء تنفيذ Docling (رمز الخروج: {process.returncode}):\")\n",
        "            if process.stdout:\n",
        "                print(f\"STDOUT:\\n{process.stdout}\")\n",
        "            if process.stderr:\n",
        "                print(f\"STDERR:\\n{process.stderr}\") # stderr غالبًا ما يحتوي على معلومات الخطأ الفعلية\n",
        "            return None\n",
        "        else:\n",
        "            # حتى مع النجاح، قد تكون هناك تحذيرات مفيدة في stderr\n",
        "            if process.stdout and process.stdout.strip(): # طباعة stdout إذا لم يكن فارغًا\n",
        "                 print(f\"Docling STDOUT:\\n{process.stdout}\")\n",
        "            if process.stderr and process.stderr.strip(): # طباعة stderr إذا لم يكن فارغًا ويحتوي على تحذيرات\n",
        "                print(f\"Docling STDERR (تحذيرات محتملة):\\n{process.stderr}\")\n",
        "            print(f\"Docling أنجز بنجاح. المخرجات في: {DOCLING_OUTPUT_DIR}\")\n",
        "\n",
        "            # البحث عن ملف hOCR الناتج\n",
        "            for fname in os.listdir(DOCLING_OUTPUT_DIR):\n",
        "                if fname.endswith(\".hocr\"):\n",
        "                    return os.path.join(DOCLING_OUTPUT_DIR, fname)\n",
        "            print(\"لم يتم العثور على ملف hOCR في المخرجات.\")\n",
        "            return None\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(\"خطأ: لم يتم العثور على أمر 'python' أو أن وحدة 'docling' غير مثبتة بشكل صحيح.\")\n",
        "        print(\"تأكد من تثبيت docling في بيئة بايثون الحالية: pip install docling\")\n",
        "        return None\n",
        "    except Exception as e:\n",
        "        print(f\"حدث خطأ غير متوقع أثناء تشغيل Docling: {e}\")\n",
        "        return None\n",
        "\n",
        "def parse_hocr_with_docling(hocr_file_path):\n",
        "    \"\"\"\n",
        "    يحلل ملف hOCR باستخدام HOCRParser من docling ويستخرج النصوص وإحداثياتها.\n",
        "    \"\"\"\n",
        "    if not hocr_file_path or not os.path.exists(hocr_file_path):\n",
        "        print(f\"ملف HOCR غير موجود: {hocr_file_path}\")\n",
        "        return []\n",
        "\n",
        "    extracted_elements = [] # قائمة لتخزين العناصر المستخرجة\n",
        "    try:\n",
        "        parser = HOCRParser()\n",
        "        doc_layout = parser.parse(hocr_file_path)\n",
        "\n",
        "        for page in doc_layout.pages:\n",
        "            for block in page.blocks:\n",
        "                if block.block_type == \"text\":\n",
        "                    block_text = \"\"\n",
        "                    for line in block.lines:\n",
        "                        line_text_parts = [word.text for word in line.words if word.text]\n",
        "                        if line_text_parts:\n",
        "                            block_text += \" \".join(line_text_parts) + \"\\n\"\n",
        "\n",
        "                    if block_text.strip():\n",
        "                        extracted_elements.append({\n",
        "                            \"type\": \"text\",\n",
        "                            \"text\": block_text.strip(),\n",
        "                            \"bbox\": block.bbox\n",
        "                        })\n",
        "                elif block.block_type in [\"image\", \"figure\", \"picture\"]: # توسيع أنواع الصور المحتملة\n",
        "                    # تحديد المسار المحتمل للصورة المكتشفة إذا قام docling بحفظها\n",
        "                    # هذا جزء تخميني، قد تحتاج لتكييفه بناءً على مخرجات docling الفعلية\n",
        "                    potential_image_filename = f\"{os.path.splitext(os.path.basename(hocr_file_path))[0]}_img_{len(extracted_elements)}.png\" # مثال\n",
        "                    potential_image_path = os.path.join(DOCLING_OUTPUT_DIR, \"images\", potential_image_filename) # أو مجلد آخر يستخدمه docling\n",
        "\n",
        "                    extracted_elements.append({\n",
        "                        \"type\": \"image_region\",\n",
        "                        \"bbox\": block.bbox,\n",
        "                        \"image_path\": potential_image_path if os.path.exists(potential_image_path) else None\n",
        "                    })\n",
        "\n",
        "        # ترتيب العناصر بناءً على موضعها الرأسي (y1) ثم الأفقي (x1)\n",
        "        extracted_elements.sort(key=lambda el: (el[\"bbox\"][1], el[\"bbox\"][0]))\n",
        "        return extracted_elements\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في تحليل ملف hOCR: {e}\")\n",
        "        return []\n",
        "\n",
        "\n",
        "def create_word_from_elements(original_image_path, elements, output_filename=\"docling_output.docx\"):\n",
        "    \"\"\"\n",
        "    ينشئ ملف وورد. يحاول وضع الصورة الأصلية والنصوص المستخرجة.\n",
        "    الحفاظ على \"العلاقة\" هنا سيكون تقريبيًا.\n",
        "    \"\"\"\n",
        "    doc = Document()\n",
        "    doc.add_heading('مستند تم إنشاؤه بواسطة Docling', level=1)\n",
        "\n",
        "    try:\n",
        "        with Image.open(original_image_path) as img:\n",
        "            img_width_px, img_height_px = img.size\n",
        "\n",
        "        doc.add_paragraph(\"الصورة الأصلية:\")\n",
        "        doc.add_picture(original_image_path, width=Inches(6.0))\n",
        "        doc.add_page_break()\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في إضافة الصورة الأصلية إلى الوورد: {e}\")\n",
        "\n",
        "    doc.add_heading('العناصر المستخرجة بواسطة Docling:', level=2)\n",
        "\n",
        "    for i, element in enumerate(elements):\n",
        "        if element[\"type\"] == \"text\":\n",
        "            doc.add_paragraph(f\"كتلة نص ({i+1}):\\n\" + element[\"text\"])\n",
        "            doc.add_paragraph(f\"(الإحداثيات: {element['bbox']})\") # اختياري: إضافة الإحداثيات\n",
        "            doc.add_paragraph(\"---\")\n",
        "        elif element[\"type\"] == \"image_region\":\n",
        "            doc.add_paragraph(f\"منطقة صورة مكتشفة ({i+1}): (الإحداثيات: {element['bbox']})\")\n",
        "            if element.get(\"image_path\") and os.path.exists(element[\"image_path\"]):\n",
        "                try:\n",
        "                    doc.add_picture(element[\"image_path\"], width=Inches(3)) # مثال للحجم\n",
        "                    doc.add_paragraph(f\"(من ملف: {os.path.basename(element['image_path'])})\")\n",
        "                except Exception as e_img:\n",
        "                    print(f\"خطأ في إضافة الصورة المكتشفة '{element['image_path']}' إلى الوورد: {e_img}\")\n",
        "            else:\n",
        "                # إذا لم يتم العثور على الصورة المقصوصة، يمكن قصها من الأصلية\n",
        "                try:\n",
        "                    with Image.open(original_image_path) as img_orig:\n",
        "                        # التأكد من أن الإحداثيات صحيحة (قد تكون نسبية أو مطلقة)\n",
        "                        # docling عادة ما يعطي إحداثيات مطلقة بالنسبة للصورة\n",
        "                        bbox = element['bbox']\n",
        "                        # التأكد من أن الإحداثيات ضمن حدود الصورة\n",
        "                        bbox = (\n",
        "                            max(0, bbox[0]),\n",
        "                            max(0, bbox[1]),\n",
        "                            min(img_width_px, bbox[2]),\n",
        "                            min(img_height_px, bbox[3])\n",
        "                        )\n",
        "                        if bbox[0] < bbox[2] and bbox[1] < bbox[3]: # التحقق من أن الصندوق صالح\n",
        "                            cropped_img = img_orig.crop(bbox)\n",
        "                            temp_cropped_path = os.path.join(DOCLING_OUTPUT_DIR, f\"temp_crop_{i}.png\")\n",
        "                            cropped_img.save(temp_cropped_path)\n",
        "                            doc.add_picture(temp_cropped_path, width=Inches(3))\n",
        "                            os.remove(temp_cropped_path)\n",
        "                        else:\n",
        "                            doc.add_paragraph(\"(إحداثيات المنطقة غير صالحة للقص)\")\n",
        "                except Exception as e_crop:\n",
        "                     print(f\"خطأ في قص وإضافة منطقة الصورة من الأصلية: {e_crop}\")\n",
        "            doc.add_paragraph(\"---\")\n",
        "\n",
        "    if not elements:\n",
        "        doc.add_paragraph(\"لم يتم استخراج أي عناصر.\")\n",
        "\n",
        "    try:\n",
        "        doc.save(output_filename)\n",
        "        print(f\"تم حفظ ملف الوورد: {output_filename}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في حفظ ملف الوورد: {e}\")\n",
        "\n",
        "\n",
        "def create_markdown_from_elements(original_image_path, elements, output_filename=\"docling_output.md\"):\n",
        "    \"\"\"\n",
        "    ينشئ ملف ماركداون. يدرج الصورة الأصلية والنصوص والمناطق المكتشفة.\n",
        "    \"\"\"\n",
        "    md_content = f\"# مستند تم إنشاؤه بواسطة Docling\\n\\n\"\n",
        "\n",
        "    original_image_filename_for_md = os.path.basename(original_image_path)\n",
        "    # قد تحتاج لنسخ الصورة إلى جانب ملف الماركداون أو استخدام مسار نسبي\n",
        "    # نفترض هنا أنه سيكون في نفس المجلد\n",
        "\n",
        "    md_content += f\"## الصورة الأصلية\\n\"\n",
        "    md_content += f\"![الصورة الأصلية]({original_image_filename_for_md})\\n\\n\"\n",
        "\n",
        "    md_content += f\"## العناصر المستخرجة\\n\\n\"\n",
        "    if not elements:\n",
        "        md_content += \"لم يتم استخراج أي عناصر.\\n\"\n",
        "\n",
        "    for i, element in enumerate(elements):\n",
        "        md_content += f\"### عنصر {i+1}: {element['type']}\\n\"\n",
        "        md_content += f\"*الإحداثيات: `{element['bbox']}`*\\n\\n\"\n",
        "        if element[\"type\"] == \"text\":\n",
        "            md_content += f\"```text\\n{element['text']}\\n```\\n\\n\"\n",
        "        elif element[\"type\"] == \"image_region\":\n",
        "            if element.get(\"image_path\") and os.path.exists(element[\"image_path\"]):\n",
        "                img_region_filename_for_md = os.path.basename(element[\"image_path\"])\n",
        "                # تأكد من نسخ هذه الصورة أيضًا إلى جانب ملف الماركداون\n",
        "                md_content += f\"![منطقة صورة مكتشفة]({img_region_filename_for_md})\\n\\n\"\n",
        "            else:\n",
        "                # يمكن إضافة ملاحظة بأنه تم اكتشاف منطقة صورة ولكن لم يتم العثور على ملفها\n",
        "                md_content += \"(تم اكتشاف منطقة صورة، لكن لم يتم العثور على ملف صورة مقصوصة مرتبط بها. يمكن قصها من الصورة الأصلية باستخدام الإحداثيات أعلاه.)\\n\\n\"\n",
        "        md_content += \"---\\n\\n\"\n",
        "\n",
        "    try:\n",
        "        with open(output_filename, \"w\", encoding=\"utf-8\") as f:\n",
        "            f.write(md_content)\n",
        "        print(f\"تم حفظ ملف الماركداون: {output_filename}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في حفظ ملف الماركداون: {e}\")\n",
        "\n",
        "# --- البرنامج الرئيسي ---\n",
        "if __name__ == \"__main__\":\n",
        "    # ضع مسار الصورة التي تريد معالجتها هنا\n",
        "    input_image_path = \"path/to/your/image_with_text_and_shape.png\" # <--- قم بتغيير هذا المسار\n",
        "    # مثال: input_image_path = \"pvc_image.png\" # إذا كانت الصورة في نفس مجلد السكربت\n",
        "\n",
        "    if not os.path.exists(input_image_path):\n",
        "        print(f\"خطأ: ملف الصورة '{input_image_path}' غير موجود.\")\n",
        "    else:\n",
        "        # 1. تنفيذ Docling OCR\n",
        "        hocr_output_file = run_docling_ocr(input_image_path)\n",
        "\n",
        "        if hocr_output_file:\n",
        "            # 2. تحليل مخرجات hOCR\n",
        "            extracted_data = parse_hocr_with_docling(hocr_output_file)\n",
        "\n",
        "            if not extracted_data:\n",
        "                print(\"لم يتم استخلاص أية عناصر من ملف hOCR.\")\n",
        "            else:\n",
        "                print(f\"\\n--- تم استخلاص {len(extracted_data)} عنصرًا ---\")\n",
        "                for i_el, element in enumerate(extracted_data):\n",
        "                    print(f\"العنصر {i_el+1}: النوع={element['type']}, الإحداثيات={element['bbox']}\")\n",
        "                    if element['type'] == 'text':\n",
        "                        print(f\"  النص (مقتطف): {element['text'][:80].replace(chr(10), ' ')}...\")\n",
        "                    elif element['type'] == 'image_region' and element.get('image_path'):\n",
        "                        print(f\"  مسار الصورة المكتشفة: {element['image_path']}\")\n",
        "                print(\"------------------------------\\n\")\n",
        "\n",
        "                # 3. إنشاء ملف وورد\n",
        "                word_output_filename = \"output_from_docling.docx\"\n",
        "                create_word_from_elements(input_image_path, extracted_data, word_output_filename)\n",
        "\n",
        "                # 4. إنشاء ملف ماركداون\n",
        "                # لكي تعمل الصور في الماركداون، تأكد من نسخها إلى جانب ملف الماركداون\n",
        "                # أو عدّل المسار في create_markdown_from_elements\n",
        "                # نسخ الصورة الأصلية والصور المقصوصة (إذا وجدت) إلى جانب ملف الماركداون\n",
        "                markdown_dir = os.getcwd() # أو أي مجلد آخر تختاره لملف الماركداون\n",
        "\n",
        "                # نسخ الصورة الأصلية\n",
        "                try:\n",
        "                    shutil.copy(input_image_path, os.path.join(markdown_dir, os.path.basename(input_image_path)))\n",
        "                except Exception as e_copy:\n",
        "                    print(f\"تنبيه: لم يتم نسخ الصورة الأصلية لملف الماركداون: {e_copy}\")\n",
        "\n",
        "                # نسخ الصور المكتشفة بواسطة docling (إذا تم حفظها بالفعل)\n",
        "                for element in extracted_data:\n",
        "                    if element['type'] == 'image_region' and element.get('image_path') and os.path.exists(element['image_path']):\n",
        "                        try:\n",
        "                            shutil.copy(element['image_path'], os.path.join(markdown_dir, os.path.basename(element['image_path'])))\n",
        "                        except Exception as e_copy_region:\n",
        "                            print(f\"تنبيه: لم يتم نسخ منطقة الصورة '{element['image_path']}' لملف الماركداون: {e_copy_region}\")\n",
        "\n",
        "                markdown_output_filename = \"output_from_docling.md\"\n",
        "                create_markdown_from_elements(input_image_path, extracted_data, markdown_output_filename)\n",
        "        else:\n",
        "            print(\"فشل Docling OCR أو لم يتم إنتاج ملف hOCR.\")\n",
        "\n",
        "    # تنظيف مجلد المخرجات إذا أردت، ولكن من الأفضل إبقاؤه للتحقق من مخرجات docling\n",
        "    # if os.path.exists(DOCLING_OUTPUT_DIR):\n",
        "    #     shutil.rmtree(DOCLING_OUTPUT_DIR)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 402
        },
        "id": "KlP2ZPEhJUc9",
        "outputId": "47f9a0b2-27da-4615-9508-c7ce13cc0e29"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "error",
          "ename": "ImportError",
          "evalue": "cannot import name 'HOCRParser' from 'docling' (/usr/local/lib/python3.11/dist-packages/docling/__init__.py)",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-7-0a876a39d043>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mdocx\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mDocument\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mdocx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshared\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mInches\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mdocling\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mHOCRParser\u001b[0m \u001b[0;31m# لاستخدام محلل hOCR الخاص بـ docling\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;31m# --- إعدادات ---\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mImportError\u001b[0m: cannot import name 'HOCRParser' from 'docling' (/usr/local/lib/python3.11/dist-packages/docling/__init__.py)",
            "",
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0;32m\nNOTE: If your import is failing due to a missing package, you can\nmanually install dependencies using either !pip or !apt.\n\nTo view examples of installing some common dependencies, click the\n\"Open Examples\" button below.\n\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n"
          ],
          "errorDetails": {
            "actions": [
              {
                "action": "open_url",
                "actionText": "Open Examples",
                "url": "/notebooks/snippets/importing_libraries.ipynb"
              }
            ]
          }
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "https://github.com/docling-project/docling/blob/main/docs/v2.md"
      ],
      "metadata": {
        "id": "gul6Lf59N7kn"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import subprocess\n",
        "import os\n",
        "import shutil\n",
        "from PIL import Image\n",
        "from docx import Document\n",
        "from docx.shared import Inches\n",
        "#from docling import HOCRParser # لاستخدام محلل hOCR الخاص بـ docling\n",
        "#from .hocr_node import HOCRNode\n",
        "import docling\n",
        "#from docling.parsers import HOCRParser\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "# --- إعدادات ---\n",
        "DOCLING_OUTPUT_DIR = \"docling_output\"\n",
        "\n",
        "def run_docling_ocr(image_path):\n",
        "    \"\"\"\n",
        "    ينفذ أمر docling ocr على الصورة ويخزن المخرجات.\n",
        "    يستخدم \"python -m docling\" لضمان التنفيذ الصحيح.\n",
        "    \"\"\"\n",
        "    if not os.path.exists(image_path):\n",
        "        print(f\"خطأ: ملف الصورة المدخل '{image_path}' غير موجود.\")\n",
        "        return None\n",
        "\n",
        "    if os.path.exists(DOCLING_OUTPUT_DIR):\n",
        "        shutil.rmtree(DOCLING_OUTPUT_DIR) # حذف المخرجات القديمة\n",
        "    os.makedirs(DOCLING_OUTPUT_DIR, exist_ok=True)\n",
        "\n",
        "    # استدعاء docling كوحدة بايثون\n",
        "    cmd = [\"python\", \"-m\", \"docling\", \"ocr\", image_path, DOCLING_OUTPUT_DIR]\n",
        "\n",
        "    print(f\"تنفيذ أمر Docling: {' '.join(cmd)}\")\n",
        "    try:\n",
        "        # استخدام subprocess.run للحصول على تحكم أفضل وانتظار اكتمال الأمر\n",
        "        process = subprocess.run(cmd, capture_output=True, text=True, check=False, encoding='utf-8', errors='ignore')\n",
        "\n",
        "        if process.returncode != 0:\n",
        "            print(f\"خطأ أثناء تنفيذ Docling (رمز الخروج: {process.returncode}):\")\n",
        "            if process.stdout:\n",
        "                print(f\"STDOUT:\\n{process.stdout}\")\n",
        "            if process.stderr:\n",
        "                print(f\"STDERR:\\n{process.stderr}\") # stderr غالبًا ما يحتوي على معلومات الخطأ الفعلية\n",
        "            return None\n",
        "        else:\n",
        "            # حتى مع النجاح، قد تكون هناك تحذيرات مفيدة في stderr\n",
        "            if process.stdout and process.stdout.strip(): # طباعة stdout إذا لم يكن فارغًا\n",
        "                 print(f\"Docling STDOUT:\\n{process.stdout}\")\n",
        "            if process.stderr and process.stderr.strip(): # طباعة stderr إذا لم يكن فارغًا ويحتوي على تحذيرات\n",
        "                print(f\"Docling STDERR (تحذيرات محتملة):\\n{process.stderr}\")\n",
        "            print(f\"Docling أنجز بنجاح. المخرجات في: {DOCLING_OUTPUT_DIR}\")\n",
        "\n",
        "            # البحث عن ملف hOCR الناتج\n",
        "            for fname in os.listdir(DOCLING_OUTPUT_DIR):\n",
        "                if fname.endswith(\".hocr\"):\n",
        "                    return os.path.join(DOCLING_OUTPUT_DIR, fname)\n",
        "            print(\"لم يتم العثور على ملف hOCR في المخرجات.\")\n",
        "            return None\n",
        "\n",
        "    except FileNotFoundError:\n",
        "        print(\"خطأ: لم يتم العثور على أمر 'python' أو أن وحدة 'docling' غير مثبتة بشكل صحيح.\")\n",
        "        print(\"تأكد من تثبيت docling في بيئة بايثون الحالية: pip install docling\")\n",
        "        return None\n",
        "    except Exception as e:\n",
        "        print(f\"حدث خطأ غير متوقع أثناء تشغيل Docling: {e}\")\n",
        "        return None\n",
        "\n",
        "def parse_hocr_with_docling(hocr_file_path):\n",
        "    \"\"\"\n",
        "    يحلل ملف hOCR باستخدام HOCRParser من docling ويستخرج النصوص وإحداثياتها.\n",
        "    \"\"\"\n",
        "    if not hocr_file_path or not os.path.exists(hocr_file_path):\n",
        "        print(f\"ملف HOCR غير موجود: {hocr_file_path}\")\n",
        "        return []\n",
        "\n",
        "    extracted_elements = [] # قائمة لتخزين العناصر المستخرجة\n",
        "    try:\n",
        "        parser = HOCRParser()\n",
        "        doc_layout = parser.parse(hocr_file_path)\n",
        "\n",
        "        for page in doc_layout.pages:\n",
        "            for block in page.blocks:\n",
        "                if block.block_type == \"text\":\n",
        "                    block_text = \"\"\n",
        "                    for line in block.lines:\n",
        "                        line_text_parts = [word.text for word in line.words if word.text]\n",
        "                        if line_text_parts:\n",
        "                            block_text += \" \".join(line_text_parts) + \"\\n\"\n",
        "\n",
        "                    if block_text.strip():\n",
        "                        extracted_elements.append({\n",
        "                            \"type\": \"text\",\n",
        "                            \"text\": block_text.strip(),\n",
        "                            \"bbox\": block.bbox\n",
        "                        })\n",
        "                elif block.block_type in [\"image\", \"figure\", \"picture\"]: # توسيع أنواع الصور المحتملة\n",
        "                    # تحديد المسار المحتمل للصورة المكتشفة إذا قام docling بحفظها\n",
        "                    # هذا جزء تخميني، قد تحتاج لتكييفه بناءً على مخرجات docling الفعلية\n",
        "                    potential_image_filename = f\"{os.path.splitext(os.path.basename(hocr_file_path))[0]}_img_{len(extracted_elements)}.png\" # مثال\n",
        "                    potential_image_path = os.path.join(DOCLING_OUTPUT_DIR, \"images\", potential_image_filename) # أو مجلد آخر يستخدمه docling\n",
        "\n",
        "                    extracted_elements.append({\n",
        "                        \"type\": \"image_region\",\n",
        "                        \"bbox\": block.bbox,\n",
        "                        \"image_path\": potential_image_path if os.path.exists(potential_image_path) else None\n",
        "                    })\n",
        "\n",
        "        # ترتيب العناصر بناءً على موضعها الرأسي (y1) ثم الأفقي (x1)\n",
        "        extracted_elements.sort(key=lambda el: (el[\"bbox\"][1], el[\"bbox\"][0]))\n",
        "        return extracted_elements\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في تحليل ملف hOCR: {e}\")\n",
        "        return []\n",
        "\n",
        "\n",
        "def create_word_from_elements(original_image_path, elements, output_filename=\"docling_output.docx\"):\n",
        "    \"\"\"\n",
        "    ينشئ ملف وورد. يحاول وضع الصورة الأصلية والنصوص المستخرجة.\n",
        "    الحفاظ على \"العلاقة\" هنا سيكون تقريبيًا.\n",
        "    \"\"\"\n",
        "    doc = Document()\n",
        "    doc.add_heading('مستند تم إنشاؤه بواسطة Docling', level=1)\n",
        "\n",
        "    try:\n",
        "        with Image.open(original_image_path) as img:\n",
        "            img_width_px, img_height_px = img.size\n",
        "\n",
        "        doc.add_paragraph(\"الصورة الأصلية:\")\n",
        "        doc.add_picture(original_image_path, width=Inches(6.0))\n",
        "        doc.add_page_break()\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في إضافة الصورة الأصلية إلى الوورد: {e}\")\n",
        "\n",
        "    doc.add_heading('العناصر المستخرجة بواسطة Docling:', level=2)\n",
        "\n",
        "    for i, element in enumerate(elements):\n",
        "        if element[\"type\"] == \"text\":\n",
        "            doc.add_paragraph(f\"كتلة نص ({i+1}):\\n\" + element[\"text\"])\n",
        "            doc.add_paragraph(f\"(الإحداثيات: {element['bbox']})\") # اختياري: إضافة الإحداثيات\n",
        "            doc.add_paragraph(\"---\")\n",
        "        elif element[\"type\"] == \"image_region\":\n",
        "            doc.add_paragraph(f\"منطقة صورة مكتشفة ({i+1}): (الإحداثيات: {element['bbox']})\")\n",
        "            if element.get(\"image_path\") and os.path.exists(element[\"image_path\"]):\n",
        "                try:\n",
        "                    doc.add_picture(element[\"image_path\"], width=Inches(3)) # مثال للحجم\n",
        "                    doc.add_paragraph(f\"(من ملف: {os.path.basename(element['image_path'])})\")\n",
        "                except Exception as e_img:\n",
        "                    print(f\"خطأ في إضافة الصورة المكتشفة '{element['image_path']}' إلى الوورد: {e_img}\")\n",
        "            else:\n",
        "                # إذا لم يتم العثور على الصورة المقصوصة، يمكن قصها من الأصلية\n",
        "                try:\n",
        "                    with Image.open(original_image_path) as img_orig:\n",
        "                        # التأكد من أن الإحداثيات صحيحة (قد تكون نسبية أو مطلقة)\n",
        "                        # docling عادة ما يعطي إحداثيات مطلقة بالنسبة للصورة\n",
        "                        bbox = element['bbox']\n",
        "                        # التأكد من أن الإحداثيات ضمن حدود الصورة\n",
        "                        bbox = (\n",
        "                            max(0, bbox[0]),\n",
        "                            max(0, bbox[1]),\n",
        "                            min(img_width_px, bbox[2]),\n",
        "                            min(img_height_px, bbox[3])\n",
        "                        )\n",
        "                        if bbox[0] < bbox[2] and bbox[1] < bbox[3]: # التحقق من أن الصندوق صالح\n",
        "                            cropped_img = img_orig.crop(bbox)\n",
        "                            temp_cropped_path = os.path.join(DOCLING_OUTPUT_DIR, f\"temp_crop_{i}.png\")\n",
        "                            cropped_img.save(temp_cropped_path)\n",
        "                            doc.add_picture(temp_cropped_path, width=Inches(3))\n",
        "                            os.remove(temp_cropped_path)\n",
        "                        else:\n",
        "                            doc.add_paragraph(\"(إحداثيات المنطقة غير صالحة للقص)\")\n",
        "                except Exception as e_crop:\n",
        "                     print(f\"خطأ في قص وإضافة منطقة الصورة من الأصلية: {e_crop}\")\n",
        "            doc.add_paragraph(\"---\")\n",
        "\n",
        "    if not elements:\n",
        "        doc.add_paragraph(\"لم يتم استخراج أي عناصر.\")\n",
        "\n",
        "    try:\n",
        "        doc.save(output_filename)\n",
        "        print(f\"تم حفظ ملف الوورد: {output_filename}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في حفظ ملف الوورد: {e}\")\n",
        "\n",
        "\n",
        "def create_markdown_from_elements(original_image_path, elements, output_filename=\"docling_output.md\"):\n",
        "    \"\"\"\n",
        "    ينشئ ملف ماركداون. يدرج الصورة الأصلية والنصوص والمناطق المكتشفة.\n",
        "    \"\"\"\n",
        "    md_content = f\"# مستند تم إنشاؤه بواسطة Docling\\n\\n\"\n",
        "\n",
        "    original_image_filename_for_md = os.path.basename(original_image_path)\n",
        "    # قد تحتاج لنسخ الصورة إلى جانب ملف الماركداون أو استخدام مسار نسبي\n",
        "    # نفترض هنا أنه سيكون في نفس المجلد\n",
        "\n",
        "    md_content += f\"## الصورة الأصلية\\n\"\n",
        "    md_content += f\"![الصورة الأصلية]({original_image_filename_for_md})\\n\\n\"\n",
        "\n",
        "    md_content += f\"## العناصر المستخرجة\\n\\n\"\n",
        "    if not elements:\n",
        "        md_content += \"لم يتم استخراج أي عناصر.\\n\"\n",
        "\n",
        "    for i, element in enumerate(elements):\n",
        "        md_content += f\"### عنصر {i+1}: {element['type']}\\n\"\n",
        "        md_content += f\"*الإحداثيات: `{element['bbox']}`*\\n\\n\"\n",
        "        if element[\"type\"] == \"text\":\n",
        "            md_content += f\"```text\\n{element['text']}\\n```\\n\\n\"\n",
        "        elif element[\"type\"] == \"image_region\":\n",
        "            if element.get(\"image_path\") and os.path.exists(element[\"image_path\"]):\n",
        "                img_region_filename_for_md = os.path.basename(element[\"image_path\"])\n",
        "                # تأكد من نسخ هذه الصورة أيضًا إلى جانب ملف الماركداون\n",
        "                md_content += f\"![منطقة صورة مكتشفة]({img_region_filename_for_md})\\n\\n\"\n",
        "            else:\n",
        "                # يمكن إضافة ملاحظة بأنه تم اكتشاف منطقة صورة ولكن لم يتم العثور على ملفها\n",
        "                md_content += \"(تم اكتشاف منطقة صورة، لكن لم يتم العثور على ملف صورة مقصوصة مرتبط بها. يمكن قصها من الصورة الأصلية باستخدام الإحداثيات أعلاه.)\\n\\n\"\n",
        "        md_content += \"---\\n\\n\"\n",
        "\n",
        "    try:\n",
        "        with open(output_filename, \"w\", encoding=\"utf-8\") as f:\n",
        "            f.write(md_content)\n",
        "        print(f\"تم حفظ ملف الماركداون: {output_filename}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في حفظ ملف الماركداون: {e}\")\n",
        "\n",
        "# --- البرنامج الرئيسي ---\n",
        "if __name__ == \"__main__\":\n",
        "    # ضع مسار الصورة التي تريد معالجتها هنا\n",
        "    input_image_path = \"/content/PVC-Structure-PhotoRoom.png-PhotoRoom.png\" # <--- قم بتغيير هذا المسار\n",
        "    # مثال: input_image_path = \"pvc_image.png\" # إذا كانت الصورة في نفس مجلد السكربت\n",
        "\n",
        "    if not os.path.exists(input_image_path):\n",
        "        print(f\"خطأ: ملف الصورة '{input_image_path}' غير موجود.\")\n",
        "    else:\n",
        "        # 1. تنفيذ Docling OCR\n",
        "        hocr_output_file = run_docling_ocr(input_image_path)\n",
        "\n",
        "        if hocr_output_file:\n",
        "            # 2. تحليل مخرجات hOCR\n",
        "            extracted_data = parse_hocr_with_docling(hocr_output_file)\n",
        "\n",
        "            if not extracted_data:\n",
        "                print(\"لم يتم استخلاص أية عناصر من ملف hOCR.\")\n",
        "            else:\n",
        "                print(f\"\\n--- تم استخلاص {len(extracted_data)} عنصرًا ---\")\n",
        "                for i_el, element in enumerate(extracted_data):\n",
        "                    print(f\"العنصر {i_el+1}: النوع={element['type']}, الإحداثيات={element['bbox']}\")\n",
        "                    if element['type'] == 'text':\n",
        "                        print(f\"  النص (مقتطف): {element['text'][:80].replace(chr(10), ' ')}...\")\n",
        "                    elif element['type'] == 'image_region' and element.get('image_path'):\n",
        "                        print(f\"  مسار الصورة المكتشفة: {element['image_path']}\")\n",
        "                print(\"------------------------------\\n\")\n",
        "\n",
        "                # 3. إنشاء ملف وورد\n",
        "                word_output_filename = \"output_from_docling.docx\"\n",
        "                create_word_from_elements(input_image_path, extracted_data, word_output_filename)\n",
        "\n",
        "                # 4. إنشاء ملف ماركداون\n",
        "                # لكي تعمل الصور في الماركداون، تأكد من نسخها إلى جانب ملف الماركداون\n",
        "                # أو عدّل المسار في create_markdown_from_elements\n",
        "                # نسخ الصورة الأصلية والصور المقصوصة (إذا وجدت) إلى جانب ملف الماركداون\n",
        "                markdown_dir = os.getcwd() # أو أي مجلد آخر تختاره لملف الماركداون\n",
        "\n",
        "                # نسخ الصورة الأصلية\n",
        "                try:\n",
        "                    shutil.copy(input_image_path, os.path.join(markdown_dir, os.path.basename(input_image_path)))\n",
        "                except Exception as e_copy:\n",
        "                    print(f\"تنبيه: لم يتم نسخ الصورة الأصلية لملف الماركداون: {e_copy}\")\n",
        "\n",
        "                # نسخ الصور المكتشفة بواسطة docling (إذا تم حفظها بالفعل)\n",
        "                for element in extracted_data:\n",
        "                    if element['type'] == 'image_region' and element.get('image_path') and os.path.exists(element['image_path']):\n",
        "                        try:\n",
        "                            shutil.copy(element['image_path'], os.path.join(markdown_dir, os.path.basename(element['image_path'])))\n",
        "                        except Exception as e_copy_region:\n",
        "                            print(f\"تنبيه: لم يتم نسخ منطقة الصورة '{element['image_path']}' لملف الماركداون: {e_copy_region}\")\n",
        "\n",
        "                markdown_output_filename = \"output_from_docling.md\"\n",
        "                create_markdown_from_elements(input_image_path, extracted_data, markdown_output_filename)\n",
        "        else:\n",
        "            print(\"فشل Docling OCR أو لم يتم إنتاج ملف hOCR.\")\n",
        "\n",
        "    # تنظيف مجلد المخرجات إذا أردت، ولكن من الأفضل إبقاؤه للتحقق من مخرجات docling\n",
        "    # if os.path.exists(DOCLING_OUTPUT_DIR):\n",
        "    #     shutil.rmtree(DOCLING_OUTPUT_DIR)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PlZgtD8rM9-O",
        "outputId": "0f8eca9a-641d-4af6-c124-9c4359df25f1"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "تنفيذ أمر Docling: python -m docling ocr /content/PVC-Structure-PhotoRoom.png-PhotoRoom.png docling_output\n",
            "خطأ أثناء تنفيذ Docling (رمز الخروج: 1):\n",
            "STDERR:\n",
            "/usr/bin/python3: No module named docling.__main__; 'docling' is a package and cannot be directly executed\n",
            "\n",
            "فشل Docling OCR أو لم يتم إنتاج ملف hOCR.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "https://github.com/jlieth/hocr-parser\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "https://docling-project.github.io/docling/usage/#using-remote-services"
      ],
      "metadata": {
        "id": "2n3n22cJNtFv"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install git+https://github.com/jlieth/hocr-parser"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "VJJomkK5M96f",
        "outputId": "26a29a2d-5dc2-479c-dca2-e0f695e9054c"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting git+https://github.com/jlieth/hocr-parser\n",
            "  Cloning https://github.com/jlieth/hocr-parser to /tmp/pip-req-build-7485mwow\n",
            "  Running command git clone --filter=blob:none --quiet https://github.com/jlieth/hocr-parser /tmp/pip-req-build-7485mwow\n",
            "  Resolved https://github.com/jlieth/hocr-parser to commit ed12231387bbc0e5841222152d996df4323931bc\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: lxml in /usr/local/lib/python3.11/dist-packages (from hocr-parser==0.3.0) (5.4.0)\n",
            "Collecting cssselect (from hocr-parser==0.3.0)\n",
            "  Downloading cssselect-1.3.0-py3-none-any.whl.metadata (2.6 kB)\n",
            "Downloading cssselect-1.3.0-py3-none-any.whl (18 kB)\n",
            "Building wheels for collected packages: hocr-parser\n",
            "  Building wheel for hocr-parser (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for hocr-parser: filename=hocr_parser-0.3.0-py2.py3-none-any.whl size=9242 sha256=a9192f50b9a7e12700b7e326210e860ccb77074cd76d01b52cfe87a89cd3a1bc\n",
            "  Stored in directory: /tmp/pip-ephem-wheel-cache-9kbipjio/wheels/ec/0b/12/97571c3133c6eaad69e85ea2a4a03dda81306b39c88e74aea8\n",
            "Successfully built hocr-parser\n",
            "Installing collected packages: cssselect, hocr-parser\n",
            "Successfully installed cssselect-1.3.0 hocr-parser-0.3.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import docling\n",
        "# dir(docling) # سيعرض الأسماء المتاحة مباشرة تحت docling\n",
        "# help(docling) # قد يعطي معلومات أكثر تفصيلاً\n",
        "# يمكنك محاولة استيراد وحدات فرعية محتملة إذا كنت تعرف أسماءها\n",
        "# import docling.parsers # كمثال\n",
        "# dir(docling.parsers)"
      ],
      "metadata": {
        "id": "HA5-wMPvM93F"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "CGAcwtedPUPY"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Hul_W3QCPUNB"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Yqw8tYX7PUJO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "IXK_KugAPUFH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "https://docling-project.github.io/docling/usage/#limit-resource-usage"
      ],
      "metadata": {
        "id": "J7zbQHV4PWHf"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import urllib.request\n",
        "from io import BytesIO\n",
        "from docling.backend.html_backend import HTMLDocumentBackend\n",
        "from docling.datamodel.base_models import InputFormat\n",
        "from docling.datamodel.document import InputDocument\n",
        "\n",
        "url = \"https://en.wikipedia.org/wiki/Duck\"\n",
        "text = urllib.request.urlopen(url).read()\n",
        "in_doc = InputDocument(\n",
        "    path_or_stream=BytesIO(text),\n",
        "    format=InputFormat.HTML,\n",
        "    backend=HTMLDocumentBackend,\n",
        "    filename=\"duck.html\",\n",
        ")\n",
        "backend = HTMLDocumentBackend(in_doc=in_doc, path_or_stream=BytesIO(text))\n",
        "dl_doc = backend.convert()\n",
        "print(dl_doc.export_to_markdown())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "v0PfryDWObcR",
        "outputId": "5a329b44-00f3-4660-ce07-f39f46fba297"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "## Contents\n",
            "\n",
            "- (Top)\n",
            "- 1 Etymology\n",
            "- 2 Taxonomy\n",
            "- 3 Morphology\n",
            "- 4 Distribution and habitat\n",
            "- 5 Behaviour Toggle Behaviour subsection\n",
            "    - 5.1 Feeding\n",
            "    - 5.2 Breeding\n",
            "    - 5.3 Communication\n",
            "    - 5.4 Predators\n",
            "- 6 Relationship with humans Toggle Relationship with humans subsection\n",
            "    - 6.1 Hunting\n",
            "    - 6.2 Domestication\n",
            "    - 6.3 Heraldry\n",
            "    - 6.4 Cultural references\n",
            "- 7 See also\n",
            "- 8 Notes Toggle Notes subsection\n",
            "    - 8.1 Citations\n",
            "    - 8.2 Sources\n",
            "- 9 External links\n",
            "\n",
            "# Duck\n",
            "\n",
            "- Acèh\n",
            "- Afrikaans\n",
            "- Alemannisch\n",
            "- አማርኛ\n",
            "- Ænglisc\n",
            "- العربية\n",
            "- Aragonés\n",
            "- ܐܪܡܝܐ\n",
            "- Armãneashti\n",
            "- Asturianu\n",
            "- Atikamekw\n",
            "- Авар\n",
            "- Aymar aru\n",
            "- تۆرکجه\n",
            "- Basa Bali\n",
            "- বাংলা\n",
            "- 閩南語 / Bân-lâm-gú\n",
            "- Беларуская\n",
            "- Беларуская (тарашкевіца)\n",
            "- Bikol Central\n",
            "- Български\n",
            "- Brezhoneg\n",
            "- Буряад\n",
            "- Català\n",
            "- Чӑвашла\n",
            "- Čeština\n",
            "- ChiShona\n",
            "- Cymraeg\n",
            "- Dagbanli\n",
            "- Dansk\n",
            "- Deitsch\n",
            "- Deutsch\n",
            "- डोटेली\n",
            "- Ελληνικά\n",
            "- Emiliàn e rumagnòl\n",
            "- Español\n",
            "- Esperanto\n",
            "- Euskara\n",
            "- فارسی\n",
            "- Français\n",
            "- Gaeilge\n",
            "- Galego\n",
            "- ГӀалгӀай\n",
            "- 贛語\n",
            "- گیلکی\n",
            "- 𐌲𐌿𐍄𐌹𐍃𐌺\n",
            "- गोंयची कोंकणी / Gõychi Konknni\n",
            "- 客家語 / Hak-kâ-ngî\n",
            "- 한국어\n",
            "- Hausa\n",
            "- Հայերեն\n",
            "- हिन्दी\n",
            "- Hrvatski\n",
            "- Ido\n",
            "- Bahasa Indonesia\n",
            "- Iñupiatun\n",
            "- Ирон\n",
            "- Íslenska\n",
            "- Italiano\n",
            "- עברית\n",
            "- Jawa\n",
            "- ಕನ್ನಡ\n",
            "- Kapampangan\n",
            "- ქართული\n",
            "- कॉशुर / کٲشُر\n",
            "- Қазақша\n",
            "- Ikirundi\n",
            "- Kongo\n",
            "- Kreyòl ayisyen\n",
            "- Кырык мары\n",
            "- ລາວ\n",
            "- Latina\n",
            "- Latviešu\n",
            "- Lietuvių\n",
            "- Li Niha\n",
            "- Ligure\n",
            "- Limburgs\n",
            "- Lingála\n",
            "- Malagasy\n",
            "- മലയാളം\n",
            "- मराठी\n",
            "- مازِرونی\n",
            "- Bahasa Melayu\n",
            "- ꯃꯤꯇꯩ ꯂꯣꯟ\n",
            "- 閩東語 / Mìng-dĕ̤ng-ngṳ̄\n",
            "- Мокшень\n",
            "- Монгол\n",
            "- မြန်မာဘာသာ\n",
            "- Nederlands\n",
            "- Nedersaksies\n",
            "- नेपाली\n",
            "- नेपाल भाषा\n",
            "- 日本語\n",
            "- Нохчийн\n",
            "- Norsk nynorsk\n",
            "- Occitan\n",
            "- Oromoo\n",
            "- ਪੰਜਾਬੀ\n",
            "- Picard\n",
            "- Plattdüütsch\n",
            "- Polski\n",
            "- Português\n",
            "- Qırımtatarca\n",
            "- Română\n",
            "- Русский\n",
            "- Саха тыла\n",
            "- ᱥᱟᱱᱛᱟᱲᱤ\n",
            "- Sardu\n",
            "- Scots\n",
            "- Seeltersk\n",
            "- Shqip\n",
            "- Sicilianu\n",
            "- සිංහල\n",
            "- Simple English\n",
            "- سنڌي\n",
            "- کوردی\n",
            "- Српски / srpski\n",
            "- Srpskohrvatski / српскохрватски\n",
            "- Sunda\n",
            "- Svenska\n",
            "- Tagalog\n",
            "- தமிழ்\n",
            "- Taqbaylit\n",
            "- Татарча / tatarça\n",
            "- తెలుగు\n",
            "- ไทย\n",
            "- Türkçe\n",
            "- Українська\n",
            "- ئۇيغۇرچە / Uyghurche\n",
            "- Vahcuengh\n",
            "- Tiếng Việt\n",
            "- Walon\n",
            "- 文言\n",
            "- Winaray\n",
            "- 吴语\n",
            "- 粵語\n",
            "- Žemaitėška\n",
            "- 中文\n",
            "- Jaku Iban\n",
            "\n",
            "- Article\n",
            "- Talk\n",
            "\n",
            "- Read\n",
            "- View source\n",
            "- View history\n",
            "\n",
            "Tools\n",
            "\n",
            "Actions\n",
            "\n",
            "- Read\n",
            "- View source\n",
            "- View history\n",
            "\n",
            "General\n",
            "\n",
            "- What links here\n",
            "- Related changes\n",
            "- Upload file\n",
            "- Permanent link\n",
            "- Page information\n",
            "- Cite this page\n",
            "- Get shortened URL\n",
            "- Download QR code\n",
            "\n",
            "Print/export\n",
            "\n",
            "- Download as PDF\n",
            "- Printable version\n",
            "\n",
            "In other projects\n",
            "\n",
            "- Wikimedia Commons\n",
            "- Wikiquote\n",
            "- Wikidata item\n",
            "\n",
            "Appearance\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "From Wikipedia, the free encyclopedia\n",
            "\n",
            "Common name for many species of bird\n",
            "\n",
            "This article is about the bird. For duck as a food, see . For other uses, see .\n",
            "\n",
            "\"Duckling\" redirects here. For other uses, see .\n",
            "\n",
            "| Duck                           | Duck                           |\n",
            "|--------------------------------|--------------------------------|\n",
            "|                                |                                |\n",
            "| Bufflehead (Bucephala albeola) | Bufflehead (Bucephala albeola) |\n",
            "| Scientific classification      | Scientific classification      |\n",
            "| Domain:                        | Eukaryota                      |\n",
            "| Kingdom:                       | Animalia                       |\n",
            "| Phylum:                        | Chordata                       |\n",
            "| Class:                         | Aves                           |\n",
            "| Order:                         | Anseriformes                   |\n",
            "| Superfamily:                   | Anatoidea                      |\n",
            "| Family:                        | Anatidae                       |\n",
            "| Subfamilies                    | Subfamilies                    |\n",
            "| See text                       | See text                       |\n",
            "\n",
            "Duck is the common name for numerous species of waterfowl in the family Anatidae. Ducks are generally smaller and shorter-necked than swans and geese, which are members of the same family. Divided among several subfamilies, they are a form taxon; they do not represent a monophyletic group (the group of all descendants of a single common ancestral species), since swans and geese are not considered ducks. Ducks are mostly aquatic birds, and may be found in both fresh water and sea water.\n",
            "\n",
            "Ducks are sometimes confused with several types of unrelated water birds with similar forms, such as loons or divers, grebes, gallinules and coots.\n",
            "\n",
            "## Etymology\n",
            "\n",
            "The word duck comes from Old English dūce 'diver', a derivative of the verb *dūcan 'to duck, bend down low as if to get under something, or dive', because of the way many species in the dabbling duck group feed by upending; compare with Dutch duiken and German tauchen 'to dive'.\n",
            "\n",
            "Pacific black duck displaying the characteristic upending \"duck\"\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "This word replaced Old English ened /ænid 'duck', possibly to avoid confusion with other words, such as ende 'end' with similar forms. Other Germanic languages still have similar words for duck, for example, Dutch eend, German Ente and Norwegian and. The word ened /ænid was inherited from Proto-Indo-European; cf. Latin anas \"duck\", Lithuanian ántis 'duck', Ancient Greek νῆσσα /νῆττα (nēssa /nētta) 'duck', and Sanskrit ātí 'water bird', among others.\n",
            "\n",
            "A duckling is a young duck in downy plumage[1] or baby duck,[2] but in the food trade a young domestic duck which has just reached adult size and bulk and its meat is still fully tender, is sometimes labelled as a duckling.\n",
            "\n",
            "A male is called a drake and the female is called a duck, or in ornithology a hen.[3][4]\n",
            "\n",
            "Male mallard.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Wood ducks.\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "## Taxonomy\n",
            "\n",
            "All ducks belong to the biological order Anseriformes, a group that contains the ducks, geese and swans, as well as the screamers, and the magpie goose.[5] All except the screamers belong to the biological family Anatidae.[5] Within the family, ducks are split into a variety of subfamilies and 'tribes'. The number and composition of these subfamilies and tribes is the cause of considerable disagreement among taxonomists.[5] Some base their decisions on morphological characteristics, others on shared behaviours or genetic studies.[6][7] The number of suggested subfamilies containing ducks ranges from two to five.[8][9] The significant level of hybridisation that occurs among wild ducks complicates efforts to tease apart the relationships between various species.[9]\n",
            "\n",
            "Mallard landing in approach\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "In most modern classifications, the so-called 'true ducks' belong to the subfamily Anatinae, which is further split into a varying number of tribes.[10] The largest of these, the Anatini, contains the 'dabbling' or 'river' ducks – named for their method of feeding primarily at the surface of fresh water.[11] The 'diving ducks', also named for their primary feeding method, make up the tribe Aythyini.[12] The 'sea ducks' of the tribe Mergini are diving ducks which specialise on fish and shellfish and spend a majority of their lives in saltwater.[13] The tribe Oxyurini contains the 'stifftails', diving ducks notable for their small size and stiff, upright tails.[14]\n",
            "\n",
            "A number of other species called ducks are not considered to be 'true ducks', and are typically placed in other subfamilies or tribes. The whistling ducks are assigned either to a tribe (Dendrocygnini) in the subfamily Anatinae or the subfamily Anserinae,[15] or to their own subfamily (Dendrocygninae) or family (Dendrocyganidae).[9][16] The freckled duck of Australia is either the sole member of the tribe Stictonettini in the subfamily Anserinae,[15] or in its own family, the Stictonettinae.[9] The shelducks make up the tribe Tadornini in the family Anserinae in some classifications,[15] and their own subfamily, Tadorninae, in others,[17] while the steamer ducks are either placed in the family Anserinae in the tribe Tachyerini[15] or lumped with the shelducks in the tribe Tadorini.[9] The perching ducks make up in the tribe Cairinini in the subfamily Anserinae in some classifications, while that tribe is eliminated in other classifications and its members assigned to the tribe Anatini.[9] The torrent duck is generally included in the subfamily Anserinae in the monotypic tribe Merganettini,[15] but is sometimes included in the tribe Tadornini.[18] The pink-eared duck is sometimes included as a true duck either in the tribe Anatini[15] or the tribe Malacorhynchini,[19] and other times is included with the shelducks in the tribe Tadornini.[15]\n",
            "\n",
            "## Morphology\n",
            "\n",
            "Male Mandarin duck\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "The overall body plan of ducks is elongated and broad, and they are also relatively long-necked, albeit not as long-necked as the geese and swans. The body shape of diving ducks varies somewhat from this in being more rounded. The bill is usually broad and contains serrated pectens, which are particularly well defined in the filter-feeding species. In the case of some fishing species the bill is long and strongly serrated. The scaled legs are strong and well developed, and generally set far back on the body, more so in the highly aquatic species, which typically feature webbed feet. The wings are very strong and are generally short and pointed, and the flight of ducks requires fast continuous strokes, requiring in turn strong wing muscles. Three species of steamer duck are almost flightless, however. Many species of duck are temporarily flightless while moulting; they seek out protected habitat with good food supplies during this period. This moult typically precedes migration.\n",
            "\n",
            "The drakes of northern species often have extravagant plumage, but that is moulted in summer to give a more female-like appearance, the \"eclipse\" plumage. Southern resident species typically show less sexual dimorphism, although there are exceptions such as the paradise shelduck of New Zealand, which is both strikingly sexually dimorphic and in which the female's plumage is brighter than that of the male. The plumage of juvenile birds generally resembles that of the female. Female ducks have evolved to have a corkscrew shaped vagina to prevent rape.\n",
            "\n",
            "## Distribution and habitat\n",
            "\n",
            "Flying steamer ducks in Ushuaia, Argentina\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Ducks have a cosmopolitan distribution, and are found on every continent except Antarctica.[5] Several species manage to live on subantarctic islands, including South Georgia and the Auckland Islands.[20] Ducks have reached a number of isolated oceanic islands, including the Hawaiian Islands, Micronesia and the Galápagos Islands, where they are often vagrants and less often residents.[21][22] A handful are endemic to such far-flung islands.[21]\n",
            "\n",
            "Female mallard in Cornwall, England\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Some duck species, mainly those breeding in the temperate and Arctic Northern Hemisphere, are migratory; those in the tropics are generally not. Some ducks, particularly in Australia where rainfall is erratic, are nomadic, seeking out the temporary lakes and pools that form after localised heavy rain.[23]\n",
            "\n",
            "## Behaviour\n",
            "\n",
            "### Feeding\n",
            "\n",
            "Pecten along the bill\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Mallard duckling preening\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Ducks eat food sources such as grasses, aquatic plants, fish, insects, small amphibians, worms, and small molluscs.\n",
            "\n",
            "Dabbling ducks feed on the surface of water or on land, or as deep as they can reach by up-ending without completely submerging.[24] Along the edge of the bill, there is a comb-like structure called a pecten. This strains the water squirting from the side of the bill and traps any food. The pecten is also used to preen feathers and to hold slippery food items.\n",
            "\n",
            "Diving ducks and sea ducks forage deep underwater. To be able to submerge more easily, the diving ducks are heavier than dabbling ducks, and therefore have more difficulty taking off to fly.\n",
            "\n",
            "A few specialized species such as the mergansers are adapted to catch and swallow large fish.\n",
            "\n",
            "The others have the characteristic wide flat bill adapted to dredging-type jobs such as pulling up waterweed, pulling worms and small molluscs out of mud, searching for insect larvae, and bulk jobs such as dredging out, holding, turning head first, and swallowing a squirming frog. To avoid injury when digging into sediment it has no cere, but the nostrils come out through hard horn.\n",
            "\n",
            "The Guardian published an article advising that ducks should not be fed with bread because it damages the health of the ducks and pollutes waterways.[25]\n",
            "\n",
            "### Breeding\n",
            "\n",
            "A Muscovy duckling\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Ducks generally only have one partner at a time, although the partnership usually only lasts one year.[26] Larger species and the more sedentary species (like fast-river specialists) tend to have pair-bonds that last numerous years.[27] Most duck species breed once a year, choosing to do so in favourable conditions (spring/summer or wet seasons). Ducks also tend to make a nest before breeding, and, after hatching, lead their ducklings to water. Mother ducks are very caring and protective of their young, but may abandon some of their ducklings if they are physically stuck in an area they cannot get out of (such as nesting in an enclosed courtyard) or are not prospering due to genetic defects or sickness brought about by hypothermia, starvation, or disease. Ducklings can also be orphaned by inconsistent late hatching where a few eggs hatch after the mother has abandoned the nest and led her ducklings to water.[28]\n",
            "\n",
            "### Communication\n",
            "\n",
            "Female mallard ducks (as well as several other species in the genus Anas, such as the American and Pacific black ducks, spot-billed duck, northern pintail and common teal) make the classic \"quack\" sound while males make a similar but raspier sound that is sometimes written as \"breeeeze\",[29][self-published source?] but, despite widespread misconceptions, most species of duck do not \"quack\".[30] In general, ducks make a range of calls, including whistles, cooing, yodels and grunts. For example, the scaup – which are diving ducks – make a noise like \"scaup\" (hence their name). Calls may be loud displaying calls or quieter contact calls.\n",
            "\n",
            "A common urban legend claims that duck quacks do not echo; however, this has been proven to be false. This myth was first debunked by the Acoustics Research Centre at the University of Salford in 2003 as part of the British Association's Festival of Science.[31] It was also debunked in one of the earlier episodes of the popular Discovery Channel television show MythBusters.[32]\n",
            "\n",
            "### Predators\n",
            "\n",
            "Ringed teal\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Ducks have many predators. Ducklings are particularly vulnerable, since their inability to fly makes them easy prey not only for predatory birds but also for large fish like pike, crocodilians, predatory testudines such as the alligator snapping turtle, and other aquatic hunters, including fish-eating birds such as herons. Ducks' nests are raided by land-based predators, and brooding females may be caught unaware on the nest by mammals, such as foxes, or large birds, such as hawks or owls.\n",
            "\n",
            "Adult ducks are fast fliers, but may be caught on the water by large aquatic predators including big fish such as the North American muskie and the European pike. In flight, ducks are safe from all but a few predators such as humans and the peregrine falcon, which uses its speed and strength to catch ducks.\n",
            "\n",
            "## Relationship with humans\n",
            "\n",
            "### Hunting\n",
            "\n",
            "Humans have hunted ducks since prehistoric times. Excavations of middens in California dating to 7800 – 6400 BP have turned up bones of ducks, including at least one now-extinct flightless species.[33] Ducks were captured in \"significant numbers\" by Holocene inhabitants of the lower Ohio River valley, suggesting they took advantage of the seasonal bounty provided by migrating waterfowl.[34] Neolithic hunters in locations as far apart as the Caribbean,[35] Scandinavia,[36] Egypt,[37] Switzerland,[38] and China relied on ducks as a source of protein for some or all of the year.[39] Archeological evidence shows that Māori people in New Zealand hunted the flightless Finsch's duck, possibly to extinction, though rat predation may also have contributed to its fate.[40] A similar end awaited the Chatham duck, a species with reduced flying capabilities which went extinct shortly after its island was colonised by Polynesian settlers.[41] It is probable that duck eggs were gathered by Neolithic hunter-gathers as well, though hard evidence of this is uncommon.[35][42]\n",
            "\n",
            "In many areas, wild ducks (including ducks farmed and released into the wild) are hunted for food or sport,[43] by shooting, or by being trapped using duck decoys. Because an idle floating duck or a duck squatting on land cannot react to fly or move quickly, \"a sitting duck\" has come to mean \"an easy target\". These ducks may be contaminated by pollutants such as PCBs.[44]\n",
            "\n",
            "### Domestication\n",
            "\n",
            "Indian Runner ducks, a common breed of domestic ducks\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Ducks have many economic uses, being farmed for their meat, eggs, and feathers (particularly their down). Approximately 3 billion ducks are slaughtered each year for meat worldwide.[45] They are also kept and bred by aviculturists and often displayed in zoos. Almost all the varieties of domestic ducks are descended from the mallard (Anas platyrhynchos), apart from the Muscovy duck (Cairina moschata).[46][47] The Call duck is another example of a domestic duck breed. Its name comes from its original use established by hunters, as a decoy to attract wild mallards from the sky, into traps set for them on the ground. The call duck is the world's smallest domestic duck breed, as it weighs less than 1 kg (2.2 lb).[48]\n",
            "\n",
            "### Heraldry\n",
            "\n",
            "Three black-colored ducks in the coat of arms of Maaninka[49]\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Ducks appear on several coats of arms, including the coat of arms of Lubāna (Latvia)[50] and the coat of arms of Föglö (Åland).[51]\n",
            "\n",
            "### Cultural references\n",
            "\n",
            "In 2002, psychologist Richard Wiseman and colleagues at the University of Hertfordshire, UK, finished a year-long LaughLab experiment, concluding that of all animals, ducks attract the most humor and silliness; he said, \"If you're going to tell a joke involving an animal, make it a duck.\"[52] The word \"duck\" may have become an inherently funny word in many languages, possibly because ducks are seen as silly in their looks or behavior. Of the many ducks in fiction, many are cartoon characters, such as Walt Disney's Donald Duck, and Warner Bros.' Daffy Duck. Howard the Duck started as a comic book character in 1973[53][54] and was made into a movie in 1986.\n",
            "\n",
            "The 1992 Disney film The Mighty Ducks, starring Emilio Estevez, chose the duck as the mascot for the fictional youth hockey team who are protagonists of the movie, based on the duck being described as a fierce fighter. This led to the duck becoming the nickname and mascot for the eventual National Hockey League professional team of the Anaheim Ducks, who were founded with the name the Mighty Ducks of Anaheim.[citation needed] The duck is also the nickname of the University of Oregon sports teams as well as the Long Island Ducks minor league baseball team.[55]\n",
            "\n",
            "## See also\n",
            "\n",
            "- Birds portal\n",
            "\n",
            "- Domestic duck\n",
            "- Duck as food\n",
            "- Duck test\n",
            "- Duck breeds\n",
            "- Fictional ducks\n",
            "- Rubber duck\n",
            "\n",
            "## Notes\n",
            "\n",
            "### Citations\n",
            "\n",
            "1. ^ \"Duckling\". The American Heritage Dictionary of the English Language, Fourth Edition. Houghton Mifflin Company. 2006. Retrieved 2015-05-22.\n",
            "2. ^ \"Duckling\". Kernerman English Multilingual Dictionary (Beta Version). K. Dictionaries Ltd. 2000–2006. Retrieved 2015-05-22.\n",
            "3. ^ Dohner, Janet Vorwald (2001). The Encyclopedia of Historic and Endangered Livestock and Poultry Breeds. Yale University Press. ISBN 978-0300138139.\n",
            "4. ^ Visca, Curt; Visca, Kelley (2003). How to Draw Cartoon Birds. The Rosen Publishing Group. ISBN 9780823961566.\n",
            "5. ^ a b c d Carboneras 1992, p. 536.\n",
            "6. ^ Livezey 1986, pp. 737–738.\n",
            "7. ^ Madsen, McHugh &amp; de Kloet 1988, p. 452.\n",
            "8. ^ Donne-Goussé, Laudet &amp; Hänni 2002, pp. 353–354.\n",
            "9. ^ a b c d e f Carboneras 1992, p. 540.\n",
            "10. ^ Elphick, Dunning &amp; Sibley 2001, p. 191.\n",
            "11. ^ Kear 2005, p. 448.\n",
            "12. ^ Kear 2005, p. 622–623.\n",
            "13. ^ Kear 2005, p. 686.\n",
            "14. ^ Elphick, Dunning &amp; Sibley 2001, p. 193.\n",
            "15. ^ a b c d e f g Carboneras 1992, p. 537.\n",
            "16. ^ American Ornithologists' Union 1998, p. xix.\n",
            "17. ^ American Ornithologists' Union 1998.\n",
            "18. ^ Carboneras 1992, p. 538.\n",
            "19. ^ Christidis &amp; Boles 2008, p. 62.\n",
            "20. ^ Shirihai 2008, pp. 239, 245.\n",
            "21. ^ a b Pratt, Bruner &amp; Berrett 1987, pp. 98–107.\n",
            "22. ^ Fitter, Fitter &amp; Hosking 2000, pp. 52–3.\n",
            "23. ^ \"Pacific Black Duck\". www.wiresnr.org. Archived from the original on May 1, 2008. Retrieved 2018-04-27.\n",
            "24. ^ Ogden, Evans. \"Dabbling Ducks\". CWE. Retrieved 2006-11-02.\n",
            "25. ^ Karl Mathiesen (16 March 2015). \"Don't feed the ducks bread, say conservationists\". The Guardian. Retrieved 13 November 2016.\n",
            "26. ^ Rohwer, Frank C.; Anderson, Michael G. (1988). \"Female-Biased Philopatry, Monogamy, and the Timing of Pair Formation in Migratory Waterfowl\". Current Ornithology. pp. 187–221. doi:10.1007/978-1-4615-6787-5\\_4. ISBN 978-1-4615-6789-9.\n",
            "27. ^ Smith, Cyndi M.; Cooke, Fred; Robertson, Gregory J.; Goudie, R. Ian; Boyd, W. Sean (2000). \"Long-Term Pair Bonds in Harlequin Ducks\". The Condor. 102 (1): 201–205. doi:10.1093/condor/102.1.201. hdl:10315/13797.\n",
            "28. ^ \"If You Find An Orphaned Duckling - Wildlife Rehabber\". wildliferehabber.com. Archived from the original on 2018-09-23. Retrieved 2018-12-22.\n",
            "29. ^ Carver, Heather (2011). The Duck Bible. Lulu.com. ISBN 9780557901562.[self-published source]\n",
            "30. ^ Titlow, Budd (2013-09-03). Bird Brains: Inside the Strange Minds of Our Fine Feathered Friends. Rowman &amp; Littlefield. ISBN 9780762797707.\n",
            "31. ^ Amos, Jonathan (2003-09-08). \"Sound science is quackers\". BBC News. Retrieved 2006-11-02.\n",
            "32. ^ \"Mythbusters Episode 8\". 12 December 2003. Archived from the original on June 26, 2007.\n",
            "33. ^ Erlandson 1994, p. 171.\n",
            "34. ^ Jeffries 2008, pp. 168, 243.\n",
            "35. ^ a b Sued-Badillo 2003, p. 65.\n",
            "36. ^ Thorpe 1996, p. 68.\n",
            "37. ^ Maisels 1999, p. 42.\n",
            "38. ^ Rau 1876, p. 133.\n",
            "39. ^ Higman 2012, p. 23.\n",
            "40. ^ Hume 2012, p. 53.\n",
            "41. ^ Hume 2012, p. 52.\n",
            "42. ^ Fieldhouse 2002, p. 167.\n",
            "43. ^ Livingston, A. D. (1998-01-01). Guide to Edible Plants and Animals. Wordsworth Editions, Limited. ISBN 9781853263774.\n",
            "44. ^ \"Study plan for waterfowl injury assessment: Determining PCB concentrations in Hudson river resident waterfowl\" (PDF). New York State Department of Environmental Conservation. US Department of Commerce. December 2008. p. 3. Archived (PDF) from the original on 2022-10-09. Retrieved 2 July 2019.\n",
            "45. ^ \"FAOSTAT\". www.fao.org. Retrieved 2019-10-25.\n",
            "46. ^ \"Anas platyrhynchos, Domestic Duck; DigiMorph Staff - The University of Texas at Austin\". Digimorph.org. Retrieved 2012-12-23.\n",
            "47. ^ Sy Montgomery. \"Mallard; Encyclopædia Britannica\". Britannica.com. Retrieved 2012-12-23.\n",
            "48. ^ Glenday, Craig (2014). Guinness World Records. Guinness World Records Limited. pp. 135. ISBN 978-1-908843-15-9.\n",
            "49. ^ Suomen kunnallisvaakunat (in Finnish). Suomen Kunnallisliitto. 1982. p. 147. ISBN 951-773-085-3.\n",
            "50. ^ \"Lubānas simbolika\" (in Latvian). Retrieved September 9, 2021.\n",
            "51. ^ \"Föglö\" (in Swedish). Retrieved September 9, 2021.\n",
            "52. ^ Young, Emma. \"World's funniest joke revealed\". New Scientist. Retrieved 7 January 2019.\n",
            "53. ^ \"Howard the Duck (character)\". Grand Comics Database.\n",
            "54. ^ Sanderson, Peter; Gilbert, Laura (2008). \"1970s\". Marvel Chronicle A Year by Year History. London, United Kingdom: Dorling Kindersley. p. 161. ISBN 978-0756641238. December saw the debut of the cigar-smoking Howard the Duck. In this story by writer Steve Gerber and artist Val Mayerik, various beings from different realities had begun turning up in the Man-Thing's Florida swamp, including this bad-tempered talking duck.\n",
            "55. ^ \"The Duck\". University of Oregon Athletics. Retrieved 2022-01-20.\n",
            "\n",
            "### Sources\n",
            "\n",
            "- American Ornithologists' Union (1998). Checklist of North American Birds (PDF). Washington, DC: American Ornithologists' Union. ISBN 978-1-891276-00-2. Archived (PDF) from the original on 2022-10-09.\n",
            "- Carboneras, Carlos (1992). del Hoyo, Josep; Elliott, Andrew; Sargatal, Jordi (eds.). Handbook of the Birds of the World. Vol. 1: Ostrich to Ducks. Barcelona: Lynx Edicions. ISBN 978-84-87334-10-8.\n",
            "- Christidis, Les; Boles, Walter E., eds. (2008). Systematics and Taxonomy of Australian Birds. Collingwood, VIC: Csiro Publishing. ISBN 978-0-643-06511-6.\n",
            "- Donne-Goussé, Carole; Laudet, Vincent; Hänni, Catherine (July 2002). \"A molecular phylogeny of Anseriformes based on mitochondrial DNA analysis\". Molecular Phylogenetics and Evolution. 23 (3): 339–356. Bibcode:2002MolPE..23..339D. doi:10.1016/S1055-7903(02)00019-2. PMID 12099792.\n",
            "- Elphick, Chris; Dunning, John B. Jr.; Sibley, David, eds. (2001). The Sibley Guide to Bird Life and Behaviour. London: Christopher Helm. ISBN 978-0-7136-6250-4.\n",
            "- Erlandson, Jon M. (1994). Early Hunter-Gatherers of the California Coast. New York, NY: Springer Science &amp; Business Media. ISBN 978-1-4419-3231-0.\n",
            "- Fieldhouse, Paul (2002). Food, Feasts, and Faith: An Encyclopedia of Food Culture in World Religions. Vol. I: A–K. Santa Barbara: ABC-CLIO. ISBN 978-1-61069-412-4.\n",
            "- Fitter, Julian; Fitter, Daniel; Hosking, David (2000). Wildlife of the Galápagos. Princeton, NJ: Princeton University Press. ISBN 978-0-691-10295-5.\n",
            "- Higman, B. W. (2012). How Food Made History. Chichester, UK: John Wiley &amp; Sons. ISBN 978-1-4051-8947-7.\n",
            "- Hume, Julian H. (2012). Extinct Birds. London: Christopher Helm. ISBN 978-1-4729-3744-5.\n",
            "- Jeffries, Richard (2008). Holocene Hunter-Gatherers of the Lower Ohio River Valley. Tuscaloosa: University of Alabama Press. ISBN 978-0-8173-1658-7.\n",
            "- Kear, Janet, ed. (2005). Ducks, Geese and Swans: Species Accounts (Cairina to Mergus). Bird Families of the World. Oxford: Oxford University Press. ISBN 978-0-19-861009-0.\n",
            "- Livezey, Bradley C. (October 1986). \"A phylogenetic analysis of recent Anseriform genera using morphological characters\" (PDF). The Auk. 103 (4): 737–754. doi:10.1093/auk/103.4.737. Archived (PDF) from the original on 2022-10-09.\n",
            "- Madsen, Cort S.; McHugh, Kevin P.; de Kloet, Siwo R. (July 1988). \"A partial classification of waterfowl (Anatidae) based on single-copy DNA\" (PDF). The Auk. 105 (3): 452–459. doi:10.1093/auk/105.3.452. Archived (PDF) from the original on 2022-10-09.\n",
            "- Maisels, Charles Keith (1999). Early Civilizations of the Old World. London: Routledge. ISBN 978-0-415-10975-8.\n",
            "- Pratt, H. Douglas; Bruner, Phillip L.; Berrett, Delwyn G. (1987). A Field Guide to the Birds of Hawaii and the Tropical Pacific. Princeton, NJ: Princeton University Press. ISBN 0-691-02399-9.\n",
            "- Rau, Charles (1876). Early Man in Europe. New York: Harper &amp; Brothers. LCCN 05040168.\n",
            "- Shirihai, Hadoram (2008). A Complete Guide to Antarctic Wildlife. Princeton, NJ, US: Princeton University Press. ISBN 978-0-691-13666-0.\n",
            "- Sued-Badillo, Jalil (2003). Autochthonous Societies. General History of the Caribbean. Paris: UNESCO. ISBN 978-92-3-103832-7.\n",
            "- Thorpe, I. J. (1996). The Origins of Agriculture in Europe. New York: Routledge. ISBN 978-0-415-08009-5.\n",
            "\n",
            "## External links\n",
            "\n",
            "- Definitions from Wiktionary\n",
            "- Media from Commons\n",
            "- Quotations from Wikiquote\n",
            "- Recipes from Wikibooks\n",
            "- Taxa from Wikispecies\n",
            "- Data from Wikidata\n",
            "\n",
            "- list of books (useful looking abstracts)\n",
            "- Ducks on postage stamps Archived 2013-05-13 at the Wayback Machine\n",
            "- Ducks at a Distance, by Rob Hines at Project Gutenberg - A modern illustrated guide to identification of US waterfowl\n",
            "\n",
            "| Authority control databases    | Authority control databases                  |\n",
            "|--------------------------------|----------------------------------------------|\n",
            "| National                       | United StatesFranceBnF dataJapanLatviaIsrael |\n",
            "| Other                          | IdRef                                        |\n",
            "\n",
            "<!-- image -->\n",
            "\n",
            "Retrieved from \"\"\n",
            "\n",
            ":\n",
            "\n",
            "- Ducks\n",
            "- Game birds\n",
            "- Bird common names\n",
            "\n",
            "Hidden categories:\n",
            "\n",
            "- CS1: unfit URL\n",
            "- All accuracy disputes\n",
            "- Accuracy disputes from February 2020\n",
            "- CS1 Finnish-language sources (fi)\n",
            "- CS1 Latvian-language sources (lv)\n",
            "- CS1 Swedish-language sources (sv)\n",
            "- Articles with short description\n",
            "- Short description is different from Wikidata\n",
            "- Wikipedia indefinitely move-protected pages\n",
            "- Wikipedia indefinitely semi-protected pages\n",
            "- Articles with 'species' microformats\n",
            "- Articles containing Old English (ca. 450-1100)-language text\n",
            "- Articles containing Dutch-language text\n",
            "- Articles containing German-language text\n",
            "- Articles containing Norwegian-language text\n",
            "- Articles containing Lithuanian-language text\n",
            "- Articles containing Ancient Greek (to 1453)-language text\n",
            "- Articles needing additional references from October 2024\n",
            "- All articles needing additional references\n",
            "- All articles with self-published sources\n",
            "- Articles with self-published sources from February 2020\n",
            "- All articles with unsourced statements\n",
            "- Articles with unsourced statements from January 2022\n",
            "- CS1: long volume value\n",
            "- Pages using Sister project links with wikidata mismatch\n",
            "- Pages using Sister project links with hidden wikidata\n",
            "- Webarchive template wayback links\n",
            "- Articles with Project Gutenberg links\n",
            "- Articles containing video clips\n",
            "\n",
            "- This page was last edited on 4 May 2025, at 16:34 (UTC).\n",
            "- Text is available under the Creative Commons Attribution-ShareAlike 4.0 License;\n",
            "additional terms may apply. By using this site, you agree to the Terms of Use and Privacy Policy. Wikipedia® is a registered trademark of the Wikimedia Foundation, Inc., a non-profit organization.\n",
            "\n",
            "- Privacy policy\n",
            "- About Wikipedia\n",
            "- Disclaimers\n",
            "- Contact Wikipedia\n",
            "- Code of Conduct\n",
            "- Developers\n",
            "- Statistics\n",
            "- Cookie statement\n",
            "- Mobile view\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from io import BytesIO\n",
        "from docling.datamodel.base_models import DocumentStream\n",
        "from docling.document_converter import DocumentConverter\n",
        "\n",
        "buf = BytesIO(your_binary_stream)\n",
        "source = DocumentStream(name=\"https://arxiv.org/pdf/2206.01062\", stream=buf)\n",
        "converter = DocumentConverter()\n",
        "result = converter.convert(source)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 211
        },
        "id": "Qo5a1qTfObf5",
        "outputId": "e4ceeaaa-45b9-41ba-adaf-757181a3c1ff"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "name 'your_binary_stream' is not defined",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-10-84d9bba207c9>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mdocling\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdocument_converter\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mDocumentConverter\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mbuf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mBytesIO\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0myour_binary_stream\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0msource\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mDocumentStream\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"https://arxiv.org/pdf/2206.01062\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstream\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbuf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0mconverter\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mDocumentConverter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mNameError\u001b[0m: name 'your_binary_stream' is not defined"
          ]
        }
      ]
    },
    {
      "source": [
        "from io import BytesIO\n",
        "from docling.datamodel.base_models import DocumentStream\n",
        "from docling.document_converter import DocumentConverter\n",
        "import requests\n",
        "\n",
        "# Assuming the PDF is accessible via a URL:\n",
        "pdf_url = \"https://arxiv.org/pdf/2206.01062\"  # Replace with your actual URL or local file path\n",
        "\n",
        "# Fetch the PDF content as binary data\n",
        "response = requests.get(pdf_url)\n",
        "response.raise_for_status()  # Raise an exception for bad responses\n",
        "\n",
        "buf = BytesIO(response.content)  # Use the downloaded content\n",
        "\n",
        "# If you have the PDF content as a local file, you can use:\n",
        "# with open(\"/path/to/your/pdf.pdf\", \"rb\") as f:\n",
        "#     buf = BytesIO(f.read())\n",
        "\n",
        "source = DocumentStream(name=\"/content/2408.09869v5.pdf\", stream=buf)\n",
        "converter = DocumentConverter()\n",
        "result = converter.convert(source)"
      ],
      "cell_type": "code",
      "metadata": {
        "id": "XuouX2jxPxJg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from pathlib import Path\n",
        "from docling.document_converter import DocumentConverter\n",
        "\n",
        "source = \"https://arxiv.org/pdf/2408.09869\"\n",
        "converter = DocumentConverter()\n",
        "result = converter.convert(source, max_num_pages=100, max_file_size=20971520)"
      ],
      "metadata": {
        "id": "LbqOm8emObj0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!docling-tools models download"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gxqLTa05QDXU",
        "outputId": "3c31dea9-a6ad-4435-a6ad-e090f3c12cdb"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 22:14:59.555999: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746915299.580337   40778 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746915299.587435   40778 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "\u001b[34mDownloading layout model\u001b[0m\u001b[33m...\u001b[0m                               \u001b]8;id=186836;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py\u001b\\\u001b[2mmodel_downloader.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=220657;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py#40\u001b\\\u001b[2m40\u001b[0m\u001b]8;;\u001b\\\n",
            "Fetching 11 files: 100% 11/11 [00:07<00:00,  1.49it/s]\n",
            "\u001b[34mDownloading tableformer model\u001b[0m\u001b[33m...\u001b[0m                          \u001b]8;id=558653;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py\u001b\\\u001b[2mmodel_downloader.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=142858;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py#48\u001b\\\u001b[2m48\u001b[0m\u001b]8;;\u001b\\\n",
            "Fetching 11 files: 100% 11/11 [00:03<00:00,  3.02it/s]\n",
            "\u001b[34mDownloading picture classifier model\u001b[0m\u001b[33m...\u001b[0m                   \u001b]8;id=306379;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py\u001b\\\u001b[2mmodel_downloader.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=492432;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py#56\u001b\\\u001b[2m56\u001b[0m\u001b]8;;\u001b\\\n",
            "Fetching 4 files:   0% 0/4 [00:00<?, ?it/s]\n",
            ".gitattributes: 100% 1.52k/1.52k [00:00<00:00, 7.64MB/s]\n",
            "\n",
            "README.md: 100% 1.11k/1.11k [00:00<00:00, 6.63MB/s]\n",
            "\n",
            "model.safetensors:   0% 0.00/16.3M [00:00<?, ?B/s]\u001b[A\n",
            "model.safetensors:  64% 10.5M/16.3M [00:00<00:00, 87.2MB/s]\u001b[A\n",
            "\n",
            "config.json: 100% 2.04k/2.04k [00:00<00:00, 10.7MB/s]\n",
            "model.safetensors: 100% 16.3M/16.3M [00:00<00:00, 70.6MB/s]\n",
            "Fetching 4 files: 100% 4/4 [00:00<00:00, 12.69it/s]\n",
            "\u001b[34mDownloading code formula model\u001b[0m\u001b[33m...\u001b[0m                         \u001b]8;id=834516;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py\u001b\\\u001b[2mmodel_downloader.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=503956;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py#64\u001b\\\u001b[2m64\u001b[0m\u001b]8;;\u001b\\\n",
            "Fetching 12 files:   0% 0/12 [00:00<?, ?it/s]\n",
            ".gitattributes: 100% 1.52k/1.52k [00:00<00:00, 10.9MB/s]\n",
            "\n",
            "config.json: 100% 996/996 [00:00<00:00, 7.10MB/s]\n",
            "\n",
            "README.md: 100% 2.32k/2.32k [00:00<00:00, 15.5MB/s]\n",
            "\n",
            "generation_config.json: 100% 132/132 [00:00<00:00, 943kB/s]\n",
            "\n",
            "merges.txt:   0% 0.00/456k [00:00<?, ?B/s]\u001b[A\n",
            "\n",
            "added_tokens.json: 100% 12.3k/12.3k [00:00<00:00, 37.7MB/s]\n",
            "\n",
            "\n",
            "special_tokens_map.json: 100% 10.1k/10.1k [00:00<00:00, 43.0MB/s]\n",
            "\n",
            "\n",
            "preprocessor_config.json: 100% 186/186 [00:00<00:00, 882kB/s]\n",
            "\n",
            "\n",
            "model.safetensors:   0% 0.00/525M [00:00<?, ?B/s]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "vocab.json:   0% 0.00/798k [00:00<?, ?B/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "tokenizer_config.json:   0% 0.00/109k [00:00<?, ?B/s]\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "tokenizer_config.json: 100% 109k/109k [00:00<00:00, 4.86MB/s]\n",
            "merges.txt: 100% 456k/456k [00:00<00:00, 5.19MB/s]\n",
            "vocab.json: 100% 798k/798k [00:00<00:00, 14.6MB/s]\n",
            "tokenizer.json: 100% 3.66M/3.66M [00:00<00:00, 30.6MB/s]\n",
            "\n",
            "\n",
            "model.safetensors:   2% 10.5M/525M [00:00<00:06, 75.4MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:   6% 31.5M/525M [00:00<00:04, 116MB/s] \u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  10% 52.4M/525M [00:00<00:03, 137MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  14% 73.4M/525M [00:02<00:24, 18.1MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  18% 94.4M/525M [00:03<00:15, 27.6MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  24% 126M/525M [00:03<00:08, 45.8MB/s] \u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  30% 157M/525M [00:03<00:05, 66.0MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  34% 178M/525M [00:03<00:04, 79.7MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  40% 210M/525M [00:03<00:03, 105MB/s] \u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  44% 231M/525M [00:03<00:02, 120MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  50% 262M/525M [00:03<00:01, 142MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  54% 283M/525M [00:03<00:01, 146MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  58% 304M/525M [00:04<00:01, 150MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  62% 325M/525M [00:04<00:01, 159MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  68% 357M/525M [00:04<00:00, 181MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  74% 388M/525M [00:04<00:00, 205MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  80% 419M/525M [00:04<00:00, 207MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  86% 451M/525M [00:04<00:00, 220MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors:  92% 482M/525M [00:04<00:00, 229MB/s]\u001b[A\u001b[A\n",
            "\n",
            "model.safetensors: 100% 525M/525M [00:05<00:00, 105MB/s]\n",
            "Fetching 12 files: 100% 12/12 [00:05<00:00,  2.35it/s]\n",
            "\u001b[34mDownloading easyocr models\u001b[0m\u001b[33m...\u001b[0m                             \u001b]8;id=168381;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py\u001b\\\u001b[2mmodel_downloader.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=621262;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py#90\u001b\\\u001b[2m90\u001b[0m\u001b]8;;\u001b\\\n",
            "100% 73.7M/73.7M [00:00<00:00, 258MB/s]\n",
            "100% 13.4M/13.4M [00:00<00:00, 236MB/s]\n",
            "100% 13.6M/13.6M [00:00<00:00, 271MB/s]\n",
            "\u001b[32m\n",
            "Models downloaded into: /root/.cache/docling/models.\u001b[0m\n",
            "\n",
            " Docling can now be configured for running offline using the local artifacts.\n",
            "\n",
            " Using the CLI: `docling --artifacts-\u001b[33mpath\u001b[0m=\u001b[35m/root/.cache/docling/\u001b[0m\u001b[95mmodels\u001b[0m FILE` \n",
            " Using Python: see the documentation at \n",
            "\u001b[1m<\u001b[0m\u001b[4;94mhttps:\u001b[0m\u001b[4;94m//docling-project.github.io/docling/usage\u001b[0m\u001b[1m>\u001b[0m.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install lxml beautifulsoup4"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ubkCsJ55QDaU",
        "outputId": "20e53a96-c238-44c5-a090-ae5ab29078e2"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: lxml in /usr/local/lib/python3.11/dist-packages (5.4.0)\n",
            "Requirement already satisfied: beautifulsoup4 in /usr/local/lib/python3.11/dist-packages (4.13.4)\n",
            "Requirement already satisfied: soupsieve>1.2 in /usr/local/lib/python3.11/dist-packages (from beautifulsoup4) (2.7)\n",
            "Requirement already satisfied: typing-extensions>=4.0.0 in /usr/local/lib/python3.11/dist-packages (from beautifulsoup4) (4.13.2)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import subprocess\n",
        "import os\n",
        "import shutil\n",
        "from PIL import Image\n",
        "from docx import Document\n",
        "from docx.shared import Inches\n",
        "from lxml import etree # لاستخدام lxml لتحليل hOCR\n",
        "import re # لاستخدام التعبيرات النمطية في تحليل title\n",
        "\n",
        "# --- إعدادات ---\n",
        "DOCLING_OUTPUT_DIR = \"docling_output\"\n",
        "\n",
        "def run_docling_ocr(image_path):\n",
        "    \"\"\"\n",
        "    ينفذ أمر docling ocr على الصورة ويخزن المخرجات.\n",
        "    يستخدم \"python -m docling\" لضمان التنفيذ الصحيح.\n",
        "    \"\"\"\n",
        "    if not os.path.exists(image_path):\n",
        "        print(f\"خطأ: ملف الصورة المدخل '{image_path}' غير موجود.\")\n",
        "        return None\n",
        "\n",
        "    if os.path.exists(DOCLING_OUTPUT_DIR):\n",
        "        shutil.rmtree(DOCLING_OUTPUT_DIR)\n",
        "    os.makedirs(DOCLING_OUTPUT_DIR, exist_ok=True)\n",
        "\n",
        "    cmd = [\"python\", \"-m\", \"docling\", \"ocr\", image_path, DOCLING_OUTPUT_DIR]\n",
        "\n",
        "    print(f\"تنفيذ أمر Docling: {' '.join(cmd)}\")\n",
        "    try:\n",
        "        process = subprocess.run(cmd, capture_output=True, text=True, check=False, encoding='utf-8', errors='ignore')\n",
        "\n",
        "        if process.returncode != 0:\n",
        "            print(f\"خطأ أثناء تنفيذ Docling (رمز الخروج: {process.returncode}):\")\n",
        "            if process.stdout: print(f\"STDOUT:\\n{process.stdout}\")\n",
        "            if process.stderr: print(f\"STDERR:\\n{process.stderr}\")\n",
        "            return None\n",
        "        else:\n",
        "            if process.stdout and process.stdout.strip(): print(f\"Docling STDOUT:\\n{process.stdout}\")\n",
        "            if process.stderr and process.stderr.strip(): print(f\"Docling STDERR (تحذيرات محتملة):\\n{process.stderr}\")\n",
        "            print(f\"Docling أنجز بنجاح. المخرجات في: {DOCLING_OUTPUT_DIR}\")\n",
        "\n",
        "            for fname in os.listdir(DOCLING_OUTPUT_DIR):\n",
        "                if fname.endswith(\".hocr\"):\n",
        "                    return os.path.join(DOCLING_OUTPUT_DIR, fname)\n",
        "            print(\"لم يتم العثور على ملف hOCR في المخرجات.\")\n",
        "            return None\n",
        "    except FileNotFoundError:\n",
        "        print(\"خطأ: لم يتم العثور على أمر 'python' أو أن وحدة 'docling' غير مثبتة بشكل صحيح.\")\n",
        "        print(\"تأكد من تثبيت docling في بيئة بايثون الحالية: pip install docling\")\n",
        "        return None\n",
        "    except Exception as e:\n",
        "        print(f\"حدث خطأ غير متوقع أثناء تشغيل Docling: {e}\")\n",
        "        return None\n",
        "\n",
        "def parse_hocr_with_lxml(hocr_file_path):\n",
        "    \"\"\"\n",
        "    يحلل ملف hOCR باستخدام lxml ويستخرج النصوص وإحداثياتها ومناطق الصور.\n",
        "    \"\"\"\n",
        "    if not hocr_file_path or not os.path.exists(hocr_file_path):\n",
        "        print(f\"ملف HOCR غير موجود: {hocr_file_path}\")\n",
        "        return [], None\n",
        "\n",
        "    extracted_elements = []\n",
        "    page_dimensions = None\n",
        "\n",
        "    try:\n",
        "        parser = etree.HTMLParser(recover=True, encoding='utf-8') # تحديد الترميز\n",
        "        tree = etree.parse(hocr_file_path, parser)\n",
        "        root = tree.getroot()\n",
        "\n",
        "        # محاولة العثور على عقدة تمثل الصفحة للحصول على الأبعاد الكلية\n",
        "        page_node = root.find(\".//div[@class='ocr_page']\")\n",
        "        if page_node is None: # محاولة بديلة إذا لم يتم العثور على ocr_page\n",
        "            body_node = root.find(\"body\")\n",
        "            if body_node is not None: page_node = body_node\n",
        "\n",
        "\n",
        "        if page_node is not None:\n",
        "            title_attr = page_node.get(\"title\", \"\")\n",
        "            # مثال على title: \"bbox 0 0 1200 1800; image \"path/to/image.png\"\"\n",
        "            bbox_search = re.search(r'bbox\\s+(\\d+)\\s+(\\d+)\\s+(\\d+)\\s+(\\d+)', title_attr)\n",
        "            if bbox_search:\n",
        "                x1, y1, x2, y2 = map(int, bbox_search.groups())\n",
        "                page_dimensions = (x2, y2) # استخدام x2, y2 كـ width, height مباشرة\n",
        "                                         # أو (x2-x1, y2-y1) إذا كانت x1,y1 ليست دائمًا 0\n",
        "\n",
        "        # البحث عن الكتل النصية الشائعة\n",
        "        # XPath للعثور على عناصر ocr_par (فقرة) أو ocr_carea (منطقة محتوى)\n",
        "        # أو ocr_line إذا لم تكن العناصر الأعلى مستوى متاحة أو مناسبة\n",
        "        # هذا قد يحتاج لتعديل بناءً على بنية ملف hOCR الفعلي\n",
        "        for node in root.xpath(\"//*[@class='ocr_par' or @class='ocr_carea' or @class='ocr_line']\"):\n",
        "            title = node.get(\"title\", \"\")\n",
        "            bbox_match = re.search(r'bbox\\s+(\\d+)\\s+(\\d+)\\s+(\\d+)\\s+(\\d+)', title)\n",
        "            if not bbox_match:\n",
        "                continue\n",
        "\n",
        "            bbox = tuple(map(int, bbox_match.groups()))\n",
        "\n",
        "            # استخلاص النص من العقدة الحالية وجميع العقد الفرعية النصية\n",
        "            # استخدام //text() للحصول على كل النصوص داخل العقدة\n",
        "            # ثم تجميعها وتنظيفها\n",
        "            text_nodes = node.xpath(\".//text()\")\n",
        "            raw_text = \" \".join(tn.strip() for tn in text_nodes if tn.strip())\n",
        "            clean_text = re.sub(r'\\s+', ' ', raw_text).strip() # إزالة المسافات الزائدة\n",
        "\n",
        "            if clean_text:\n",
        "                # التحقق مما إذا كانت هذه الكتلة جزءًا من كتلة أكبر تم استخلاصها بالفعل\n",
        "                # هذا لمنع التكرار إذا كنا نستخلص ocr_line داخل ocr_par\n",
        "                is_redundant = False\n",
        "                for el in extracted_elements:\n",
        "                    if el[\"type\"] == \"text\" and clean_text in el[\"text\"] and el[\"bbox\"] == bbox:\n",
        "                        is_redundant = True\n",
        "                        break\n",
        "                if not is_redundant:\n",
        "                    extracted_elements.append({\n",
        "                        \"type\": \"text\",\n",
        "                        \"text\": clean_text,\n",
        "                        \"bbox\": bbox\n",
        "                    })\n",
        "\n",
        "        # البحث عن الصور (إذا كانت مميزة بوسم <img> أو فئة معينة)\n",
        "        # هذا الجزء تخميني ويعتمد على كيفية تمييز docling للصور\n",
        "        for img_node in root.xpath(\"//img[contains(@class, 'ocr_image') or contains(@class, 'ocr_figure')] | //div[contains(@class, 'ocr_image_container') or contains(@class, 'ocr_figure')]\"):\n",
        "            title = img_node.get(\"title\", \"\")\n",
        "            bbox_match = re.search(r'bbox\\s+(\\d+)\\s+(\\d+)\\s+(\\d+)\\s+(\\d+)', title)\n",
        "            if bbox_match:\n",
        "                img_bbox = tuple(map(int, bbox_match.groups()))\n",
        "                # التحقق من عدم إضافة نفس منطقة الصورة أكثر من مرة\n",
        "                is_duplicate_img = any(el[\"type\"] == \"image_region\" and el[\"bbox\"] == img_bbox for el in extracted_elements)\n",
        "                if not is_duplicate_img:\n",
        "                    extracted_elements.append({\n",
        "                        \"type\": \"image_region\",\n",
        "                        \"bbox\": img_bbox,\n",
        "                        \"image_path\": None # سنعتمد على القص من الصورة الأصلية\n",
        "                    })\n",
        "\n",
        "        # ترتيب العناصر بناءً على موضعها الرأسي ثم الأفقي\n",
        "        extracted_elements.sort(key=lambda el: (el[\"bbox\"][1], el[\"bbox\"][0]))\n",
        "\n",
        "        # إذا لم يتم العثور على أبعاد الصفحة من hOCR، حاول تقديرها من أقصى إحداثيات العناصر\n",
        "        if not page_dimensions and extracted_elements:\n",
        "            max_x2 = 0\n",
        "            max_y2 = 0\n",
        "            for el in extracted_elements:\n",
        "                if el[\"bbox\"][2] > max_x2: max_x2 = el[\"bbox\"][2]\n",
        "                if el[\"bbox\"][3] > max_y2: max_y2 = el[\"bbox\"][3]\n",
        "            if max_x2 > 0 and max_y2 > 0:\n",
        "                page_dimensions = (max_x2, max_y2)\n",
        "\n",
        "        return extracted_elements, page_dimensions\n",
        "\n",
        "    except etree.XMLSyntaxError as e:\n",
        "        print(f\"خطأ في بناء جملة XML/HTML في ملف hOCR: {hocr_file_path} - {e}\")\n",
        "        try:\n",
        "            with open(hocr_file_path, 'r', encoding='utf-8') as f_err:\n",
        "                print(\"محتوى الملف (أول 1000 حرف):\"); print(f_err.read(1000))\n",
        "        except Exception: pass\n",
        "        return [], None\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ عام في تحليل ملف hOCR باستخدام lxml: {e}\")\n",
        "        import traceback\n",
        "        traceback.print_exc()\n",
        "        return [], None\n",
        "\n",
        "\n",
        "def create_word_from_elements(original_image_path, elements, page_dimensions, output_filename=\"docling_output.docx\"):\n",
        "    doc = Document()\n",
        "    doc.add_heading('مستند تم إنشاؤه بواسطة Docling', level=1)\n",
        "\n",
        "    img_width_px, img_height_px = 0, 0\n",
        "    if page_dimensions:\n",
        "        img_width_px, img_height_px = page_dimensions\n",
        "    else:\n",
        "        try:\n",
        "            with Image.open(original_image_path) as img:\n",
        "                img_width_px, img_height_px = img.size\n",
        "        except Exception as e_img_dim:\n",
        "            print(f\"تنبيه: لم يتمكن من الحصول على أبعاد الصورة الأصلية: {e_img_dim}\")\n",
        "\n",
        "    try:\n",
        "        doc.add_paragraph(\"الصورة الأصلية:\")\n",
        "        doc.add_picture(original_image_path, width=Inches(6.0))\n",
        "        doc.add_page_break()\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في إضافة الصورة الأصلية إلى الوورد: {e}\")\n",
        "\n",
        "    doc.add_heading('العناصر المستخرجة بواسطة Docling:', level=2)\n",
        "\n",
        "    if not elements:\n",
        "        doc.add_paragraph(\"لم يتم استخراج أي عناصر.\")\n",
        "\n",
        "    for i, element in enumerate(elements):\n",
        "        doc.add_paragraph(f\"--- عنصر {i+1} ---\")\n",
        "        doc.add_paragraph(f\"النوع: {element['type']}\")\n",
        "        doc.add_paragraph(f\"الإحداثيات (bbox): {element['bbox']}\")\n",
        "\n",
        "        if element[\"type\"] == \"text\":\n",
        "            doc.add_paragraph(\"النص:\")\n",
        "            doc.add_paragraph(element[\"text\"])\n",
        "        elif element[\"type\"] == \"image_region\":\n",
        "            doc.add_paragraph(\"منطقة صورة مكتشفة.\")\n",
        "            if img_width_px > 0 and img_height_px > 0:\n",
        "                try:\n",
        "                    with Image.open(original_image_path) as img_orig:\n",
        "                        bbox = element['bbox']\n",
        "                        # التأكد أن إحداثيات القص ضمن حدود الصورة\n",
        "                        crop_box = (\n",
        "                            max(0, bbox[0]),\n",
        "                            max(0, bbox[1]),\n",
        "                            min(img_width_px, bbox[2]),\n",
        "                            min(img_height_px, bbox[3])\n",
        "                        )\n",
        "                        if crop_box[0] < crop_box[2] and crop_box[1] < crop_box[3]: # التحقق من أن الصندوق صالح\n",
        "                            cropped_img = img_orig.crop(crop_box)\n",
        "                            # تحديد عرض مناسب للصورة المقصوصة في الوورد\n",
        "                            aspect_ratio = cropped_img.width / cropped_img.height if cropped_img.height > 0 else 1\n",
        "                            display_width = Inches(min(3.0, 6.0 * (cropped_img.width / img_width_px if img_width_px > 0 else 1))) # تقدير حجم العرض\n",
        "\n",
        "                            temp_cropped_path = os.path.join(DOCLING_OUTPUT_DIR, f\"temp_crop_word_{i}.png\")\n",
        "                            cropped_img.save(temp_cropped_path)\n",
        "                            doc.add_picture(temp_cropped_path, width=display_width)\n",
        "                            os.remove(temp_cropped_path)\n",
        "                        else:\n",
        "                             doc.add_paragraph(\"(إحداثيات المنطقة غير صالحة للقص أو خارج الحدود)\")\n",
        "                except Exception as e_crop:\n",
        "                     print(f\"خطأ في قص وإضافة منطقة الصورة {element['bbox']} من الأصلية: {e_crop}\")\n",
        "            else:\n",
        "                doc.add_paragraph(\"(لا يمكن قص المنطقة: أبعاد الصورة الأصلية غير متاحة أو صفرية)\")\n",
        "    try:\n",
        "        doc.save(output_filename)\n",
        "        print(f\"تم حفظ ملف الوورد: {output_filename}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في حفظ ملف الوورد: {e}\")\n",
        "\n",
        "def create_markdown_from_elements(original_image_path, elements, page_dimensions, output_filename=\"docling_output.md\"):\n",
        "    md_content = f\"# مستند تم إنشاؤه بواسطة Docling\\n\\n\"\n",
        "\n",
        "    original_image_filename_for_md = os.path.basename(original_image_path)\n",
        "    md_content += f\"## الصورة الأصلية\\n\"\n",
        "    md_content += f\"![الصورة الأصلية]({original_image_filename_for_md})\\n\\n\"\n",
        "\n",
        "    md_content += f\"## العناصر المستخرجة\\n\\n\"\n",
        "    if not elements:\n",
        "        md_content += \"لم يتم استخراج أي عناصر.\\n\"\n",
        "\n",
        "    img_width_px, img_height_px = 0,0\n",
        "    if page_dimensions: img_width_px, img_height_px = page_dimensions\n",
        "\n",
        "    for i, element in enumerate(elements):\n",
        "        md_content += f\"### عنصر {i+1}: {element['type']}\\n\"\n",
        "        md_content += f\"*الإحداثيات: `{element['bbox']}`*\\n\\n\"\n",
        "        if element[\"type\"] == \"text\":\n",
        "            md_content += f\"```text\\n{element['text']}\\n```\\n\\n\"\n",
        "        elif element[\"type\"] == \"image_region\":\n",
        "            md_content += \"منطقة صورة مكتشفة.\\n\"\n",
        "            # قص وحفظ الصورة للماركداون\n",
        "            if img_width_px > 0 and img_height_px > 0:\n",
        "                try:\n",
        "                    with Image.open(original_image_path) as img_orig:\n",
        "                        bbox = element['bbox']\n",
        "                        crop_box = (max(0, bbox[0]), max(0, bbox[1]), min(img_width_px, bbox[2]), min(img_height_px, bbox[3]))\n",
        "                        if crop_box[0] < crop_box[2] and crop_box[1] < crop_box[3]:\n",
        "                            cropped_img = img_orig.crop(crop_box)\n",
        "                            md_crop_filename = f\"md_crop_{i}_{os.path.basename(original_image_path)}\"\n",
        "                            md_crop_path = os.path.join(os.path.dirname(output_filename), md_crop_filename) # حفظ بجانب ملف الماركداون\n",
        "                            cropped_img.save(md_crop_path)\n",
        "                            md_content += f\"![منطقة صورة مقصوصة {i}]({md_crop_filename})\\n\\n\"\n",
        "                        else:\n",
        "                            md_content += \"(إحداثيات المنطقة غير صالحة للقص أو خارج الحدود)\\n\\n\"\n",
        "                except Exception as e_md_crop:\n",
        "                    print(f\"خطأ في قص وحفظ منطقة الصورة {element['bbox']} للماركداون: {e_md_crop}\")\n",
        "                    md_content += \"(خطأ أثناء محاولة قص الصورة للماركداون)\\n\\n\"\n",
        "            else:\n",
        "                 md_content += \"(لا يمكن قص المنطقة للماركداون: أبعاد الصورة الأصلية غير متاحة أو صفرية)\\n\\n\"\n",
        "\n",
        "        md_content += \"---\\n\\n\"\n",
        "\n",
        "    try:\n",
        "        with open(output_filename, \"w\", encoding=\"utf-8\") as f:\n",
        "            f.write(md_content)\n",
        "        print(f\"تم حفظ ملف الماركداون: {output_filename}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في حفظ ملف الماركداون: {e}\")\n",
        "\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    input_image_path = \"/content/PVC-Structure-PhotoRoom.png-PhotoRoom.png\" # <--- عدل المسار هنا إذا لزم الأمر\n",
        "\n",
        "    if not os.path.exists(input_image_path):\n",
        "        print(f\"خطأ: ملف الصورة '{input_image_path}' غير موجود.\")\n",
        "    else:\n",
        "        hocr_output_file = run_docling_ocr(input_image_path)\n",
        "\n",
        "        if hocr_output_file:\n",
        "            # استخدام الدالة الجديدة لتحليل hOCR\n",
        "            extracted_data, page_dims = parse_hocr_with_lxml(hocr_output_file)\n",
        "\n",
        "            if not extracted_data:\n",
        "                print(\"لم يتم استخلاص أية عناصر من ملف hOCR.\")\n",
        "            else:\n",
        "                print(f\"\\n--- تم استخلاص {len(extracted_data)} عنصرًا ---\")\n",
        "                if page_dims:\n",
        "                    print(f\"أبعاد الصفحة المقدرة: {page_dims}\")\n",
        "                for i_el, element in enumerate(extracted_data):\n",
        "                    print(f\"العنصر {i_el+1}: النوع={element['type']}, الإحداثيات={element['bbox']}\")\n",
        "                    if element['type'] == 'text':\n",
        "                        print(f\"  النص (مقتطف): {element['text'][:80].replace(chr(10), ' ')}...\")\n",
        "                print(\"------------------------------\\n\")\n",
        "\n",
        "                word_output_filename = \"output_from_docling.docx\"\n",
        "                create_word_from_elements(input_image_path, extracted_data, page_dims, word_output_filename)\n",
        "\n",
        "                markdown_dir = os.getcwd()\n",
        "                # نسخ الصورة الأصلية لملف الماركداون\n",
        "                try:\n",
        "                    if not os.path.exists(os.path.join(markdown_dir, os.path.basename(input_image_path))):\n",
        "                         shutil.copy(input_image_path, os.path.join(markdown_dir, os.path.basename(input_image_path)))\n",
        "                except Exception as e_copy:\n",
        "                    print(f\"تنبيه: لم يتم نسخ الصورة الأصلية لملف الماركداون: {e_copy}\")\n",
        "\n",
        "                markdown_output_filename = os.path.join(markdown_dir, \"output_from_docling.md\")\n",
        "                create_markdown_from_elements(input_image_path, extracted_data, page_dims, markdown_output_filename)\n",
        "        else:\n",
        "            print(\"فشل Docling OCR أو لم يتم إنتاج ملف hOCR.\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "De-kcHv8QDdK",
        "outputId": "dec85f73-04d1-40bc-e671-535053a5bf9e"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "تنفيذ أمر Docling: python -m docling ocr /content/PVC-Structure-PhotoRoom.png-PhotoRoom.png docling_output\n",
            "خطأ أثناء تنفيذ Docling (رمز الخروج: 1):\n",
            "STDERR:\n",
            "/usr/bin/python3: No module named docling.__main__; 'docling' is a package and cannot be directly executed\n",
            "\n",
            "فشل Docling OCR أو لم يتم إنتاج ملف hOCR.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import subprocess\n",
        "import os\n",
        "import shutil\n",
        "from PIL import Image\n",
        "from docx import Document\n",
        "from docx.shared import Inches\n",
        "import re # لا يزال مفيدًا لبعض الحالات\n",
        "\n",
        "# --- استيراد hocr-parser ---\n",
        "try:\n",
        "    # بناءً على بنية شائعة لمثل هذه المكتبات\n",
        "    from hocr_parser import HocrParser # هذا هو الاستيراد الصحيح بناءً على مصدر المكتبة\n",
        "except ImportError:\n",
        "    print(\"خطأ: لم يتم العثور على مكتبة hocr-parser. تأكد من تثبيتها بشكل صحيح.\")\n",
        "    print(\"يمكنك تثبيتها باستخدام: pip install git+https://github.com/jlieth/hocr-parser\")\n",
        "    HocrParser = None # لكي لا يفشل الكود إذا لم يتم التثبيت\n",
        "\n",
        "# --- إعدادات ---\n",
        "DOCLING_OUTPUT_DIR = \"docling_output\"\n",
        "\n",
        "def run_docling_ocr(image_path):\n",
        "    \"\"\"\n",
        "    ينفذ أمر docling ocr على الصورة ويخزن المخرجات.\n",
        "    يستخدم \"python -m docling\" لضمان التنفيذ الصحيح.\n",
        "    \"\"\"\n",
        "    if not os.path.exists(image_path):\n",
        "        print(f\"خطأ: ملف الصورة المدخل '{image_path}' غير موجود.\")\n",
        "        return None\n",
        "\n",
        "    if os.path.exists(DOCLING_OUTPUT_DIR):\n",
        "        shutil.rmtree(DOCLING_OUTPUT_DIR)\n",
        "    os.makedirs(DOCLING_OUTPUT_DIR, exist_ok=True)\n",
        "\n",
        "    cmd = [\"python\", \"-m\", \"docling\", \"ocr\", image_path, DOCLING_OUTPUT_DIR]\n",
        "\n",
        "    print(f\"تنفيذ أمر Docling: {' '.join(cmd)}\")\n",
        "    try:\n",
        "        process = subprocess.run(cmd, capture_output=True, text=True, check=False, encoding='utf-8', errors='ignore')\n",
        "\n",
        "        if process.returncode != 0:\n",
        "            print(f\"خطأ أثناء تنفيذ Docling (رمز الخروج: {process.returncode}):\")\n",
        "            if process.stdout: print(f\"STDOUT:\\n{process.stdout}\")\n",
        "            if process.stderr: print(f\"STDERR:\\n{process.stderr}\")\n",
        "            return None\n",
        "        else:\n",
        "            if process.stdout and process.stdout.strip(): print(f\"Docling STDOUT:\\n{process.stdout}\")\n",
        "            if process.stderr and process.stderr.strip(): print(f\"Docling STDERR (تحذيرات محتملة):\\n{process.stderr}\")\n",
        "            print(f\"Docling أنجز بنجاح. المخرجات في: {DOCLING_OUTPUT_DIR}\")\n",
        "\n",
        "            for fname in os.listdir(DOCLING_OUTPUT_DIR):\n",
        "                if fname.endswith(\".hocr\"):\n",
        "                    return os.path.join(DOCLING_OUTPUT_DIR, fname)\n",
        "            print(\"لم يتم العثور على ملف hOCR في المخرجات.\")\n",
        "            return None\n",
        "    except FileNotFoundError:\n",
        "        print(\"خطأ: لم يتم العثور على أمر 'python' أو أن وحدة 'docling' غير مثبتة بشكل صحيح.\")\n",
        "        return None\n",
        "    except Exception as e:\n",
        "        print(f\"حدث خطأ غير متوقع أثناء تشغيل Docling: {e}\")\n",
        "        return None\n",
        "\n",
        "def parse_hocr_with_hocr_library(hocr_file_path): # تم تغيير اسم الدالة\n",
        "    \"\"\"\n",
        "    يحلل ملف hOCR باستخدام مكتبة hocr-parser (jlieth/hocr-parser)\n",
        "    ويستخرج النصوص وإحداثياتها ومناطق الصور.\n",
        "    \"\"\"\n",
        "    if HocrParser is None: # التحقق مما إذا كان الاستيراد ناجحًا\n",
        "        print(\"مكتبة hocr-parser غير متوفرة. لا يمكن تحليل ملف hOCR.\")\n",
        "        return [], None\n",
        "\n",
        "    if not hocr_file_path or not os.path.exists(hocr_file_path):\n",
        "        print(f\"ملف HOCR غير موجود: {hocr_file_path}\")\n",
        "        return [], None\n",
        "\n",
        "    extracted_elements = []\n",
        "    page_dimensions = None\n",
        "\n",
        "    try:\n",
        "        # استخدام HocrParser من مكتبة jlieth/hocr-parser\n",
        "        parser = HocrParser(hocr_file_path) # تمرير مسار الملف مباشرة\n",
        "\n",
        "        # الحصول على أبعاد الصفحة (إذا كانت المكتبة توفرها بسهولة)\n",
        "        # عادة ما تكون على مستوى الصفحة الأولى\n",
        "        if parser.pages:\n",
        "            first_page = parser.pages[0]\n",
        "            # تعتمد أسماء الخصائص على كيفية تعريفها في المكتبة\n",
        "            # مثال: first_page.bbox قد يكون [x1, y1, x2, y2]\n",
        "            if hasattr(first_page, 'bbox') and first_page.bbox:\n",
        "                 x1, y1, x2, y2 = first_page.bbox\n",
        "                 page_dimensions = (x2 - x1, y2 - y1) # width, height\n",
        "            elif hasattr(first_page, 'width') and hasattr(first_page, 'height'):\n",
        "                 page_dimensions = (first_page.width, first_page.height)\n",
        "\n",
        "\n",
        "        for page in parser.pages:\n",
        "            # استخلاص كتل النص (عادةً ما تكون فقرات أو مناطق)\n",
        "            # قد توفر المكتبة وصولاً إلى ocr_area, ocr_par, ocr_line, ocrx_word\n",
        "            # سنحاول استهداف كتل نصية ذات معنى (مثل الفقرات)\n",
        "            if hasattr(page, 'paragraphs'): # إذا كانت المكتبة تميز الفقرات\n",
        "                for para in page.paragraphs:\n",
        "                    if hasattr(para, 'text') and hasattr(para, 'bbox') and para.text.strip():\n",
        "                        extracted_elements.append({\n",
        "                            \"type\": \"text\",\n",
        "                            \"text\": para.text.strip(),\n",
        "                            \"bbox\": para.bbox # يجب أن تكون الإحداثيات (x1, y1, x2, y2)\n",
        "                        })\n",
        "            elif hasattr(page, 'lines'): # إذا لم تكن الفقرات متاحة، ننتقل للأسطر\n",
        "                 current_paragraph_text = \"\"\n",
        "                 current_paragraph_bbox = None\n",
        "                 for line in page.lines:\n",
        "                    if hasattr(line, 'text') and hasattr(line, 'bbox') and line.text.strip():\n",
        "                        # محاولة تجميع الأسطر في فقرات إذا أمكن (تبسيط هنا)\n",
        "                        # سنضيف كل سطر كعنصر نصي منفصل حاليًا لتبسيط الأمر\n",
        "                        # أو يمكنك محاولة تجميعها إذا كانت متقاربة\n",
        "                        extracted_elements.append({\n",
        "                            \"type\": \"text\",\n",
        "                            \"text\": line.text.strip(),\n",
        "                            \"bbox\": line.bbox\n",
        "                        })\n",
        "            # يمكنك إضافة منطق أكثر تفصيلاً لاستخلاص ocr_carea أو عناصر أخرى إذا لزم الأمر\n",
        "            # بناءً على ما توفره المكتبة.\n",
        "\n",
        "            # استخلاص مناطق الصور (إذا كانت المكتبة تميزها)\n",
        "            # مثال: إذا كان هناك page.images أو شيء مشابه\n",
        "            if hasattr(page, 'images'): # هذا تخمين، تحقق من وثائق المكتبة\n",
        "                for img_element in page.images:\n",
        "                    if hasattr(img_element, 'bbox'):\n",
        "                        extracted_elements.append({\n",
        "                            \"type\": \"image_region\",\n",
        "                            \"bbox\": img_element.bbox,\n",
        "                            \"image_path\": None # سنعتمد على القص\n",
        "                        })\n",
        "            # طريقة أخرى قد تكون البحث عن عناصر لها class معين إذا لم يكن هناك تمييز مباشر للصور\n",
        "            # elif hasattr(page, 'elements_by_class'):\n",
        "            #     for img_element in page.elements_by_class('ocr_image') or page.elements_by_class('ocr_figure'):\n",
        "            #         if hasattr(img_element, 'bbox'):\n",
        "            #             extracted_elements.append({\n",
        "            #                 \"type\": \"image_region\",\n",
        "            #                 \"bbox\": img_element.bbox,\n",
        "            #                 \"image_path\": None\n",
        "            #             })\n",
        "\n",
        "        # ترتيب العناصر\n",
        "        extracted_elements.sort(key=lambda el: (el[\"bbox\"][1], el[\"bbox\"][0]))\n",
        "\n",
        "        # إذا لم يتم العثور على أبعاد الصفحة، حاول تقديرها\n",
        "        if not page_dimensions and extracted_elements:\n",
        "            max_x2 = 0\n",
        "            max_y2 = 0\n",
        "            for el in extracted_elements:\n",
        "                if el[\"bbox\"][2] > max_x2: max_x2 = el[\"bbox\"][2]\n",
        "                if el[\"bbox\"][3] > max_y2: max_y2 = el[\"bbox\"][3]\n",
        "            if max_x2 > 0 and max_y2 > 0:\n",
        "                page_dimensions = (max_x2, max_y2)\n",
        "\n",
        "        return extracted_elements, page_dimensions\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في تحليل ملف hOCR باستخدام مكتبة hocr-parser: {e}\")\n",
        "        import traceback\n",
        "        traceback.print_exc()\n",
        "        return [], None\n",
        "\n",
        "# --- الدوال الأخرى (create_word_from_elements, create_markdown_from_elements) تبقى كما هي ---\n",
        "# --- سأقوم بلصقها مرة أخرى للتأكد من أننا نستخدم النسخة الصحيحة ---\n",
        "\n",
        "def create_word_from_elements(original_image_path, elements, page_dimensions, output_filename=\"docling_output.docx\"):\n",
        "    doc = Document()\n",
        "    doc.add_heading('مستند تم إنشاؤه بواسطة Docling و hocr-parser', level=1)\n",
        "\n",
        "    img_width_px, img_height_px = 0, 0\n",
        "    if page_dimensions:\n",
        "        img_width_px, img_height_px = page_dimensions\n",
        "    else:\n",
        "        try:\n",
        "            with Image.open(original_image_path) as img:\n",
        "                img_width_px, img_height_px = img.size\n",
        "        except Exception as e_img_dim:\n",
        "            print(f\"تنبيه: لم يتمكن من الحصول على أبعاد الصورة الأصلية: {e_img_dim}\")\n",
        "\n",
        "    try:\n",
        "        doc.add_paragraph(\"الصورة الأصلية:\")\n",
        "        doc.add_picture(original_image_path, width=Inches(6.0))\n",
        "        doc.add_page_break()\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في إضافة الصورة الأصلية إلى الوورد: {e}\")\n",
        "\n",
        "    doc.add_heading('العناصر المستخرجة:', level=2)\n",
        "\n",
        "    if not elements:\n",
        "        doc.add_paragraph(\"لم يتم استخراج أي عناصر.\")\n",
        "\n",
        "    for i, element in enumerate(elements):\n",
        "        doc.add_paragraph(f\"--- عنصر {i+1} ---\")\n",
        "        doc.add_paragraph(f\"النوع: {element['type']}\")\n",
        "        doc.add_paragraph(f\"الإحداثيات (bbox): {element['bbox']}\")\n",
        "\n",
        "        if element[\"type\"] == \"text\":\n",
        "            doc.add_paragraph(\"النص:\")\n",
        "            doc.add_paragraph(element[\"text\"])\n",
        "        elif element[\"type\"] == \"image_region\":\n",
        "            doc.add_paragraph(\"منطقة صورة مكتشفة.\")\n",
        "            if img_width_px > 0 and img_height_px > 0:\n",
        "                try:\n",
        "                    with Image.open(original_image_path) as img_orig:\n",
        "                        bbox = element['bbox']\n",
        "                        crop_box = (\n",
        "                            max(0, int(bbox[0])), max(0, int(bbox[1])),\n",
        "                            min(img_width_px, int(bbox[2])), min(img_height_px, int(bbox[3]))\n",
        "                        )\n",
        "                        if crop_box[0] < crop_box[2] and crop_box[1] < crop_box[3]:\n",
        "                            cropped_img = img_orig.crop(crop_box)\n",
        "                            aspect_ratio = cropped_img.width / cropped_img.height if cropped_img.height > 0 else 1\n",
        "                            display_width_inches = min(3.0, 6.0 * (cropped_img.width / img_width_px if img_width_px > 0 else 1))\n",
        "                            display_width = Inches(display_width_inches if display_width_inches > 0.1 else 0.5) # Ensure minimum width\n",
        "\n",
        "                            temp_cropped_path = os.path.join(DOCLING_OUTPUT_DIR, f\"temp_crop_word_{i}.png\")\n",
        "                            cropped_img.save(temp_cropped_path)\n",
        "                            doc.add_picture(temp_cropped_path, width=display_width)\n",
        "                            os.remove(temp_cropped_path)\n",
        "                        else:\n",
        "                             doc.add_paragraph(\"(إحداثيات المنطقة غير صالحة للقص أو خارج الحدود)\")\n",
        "                except Exception as e_crop:\n",
        "                     print(f\"خطأ في قص وإضافة منطقة الصورة {element['bbox']} من الأصلية: {e_crop}\")\n",
        "            else:\n",
        "                doc.add_paragraph(\"(لا يمكن قص المنطقة: أبعاد الصورة الأصلية غير متاحة أو صفرية)\")\n",
        "    try:\n",
        "        doc.save(output_filename)\n",
        "        print(f\"تم حفظ ملف الوورد: {output_filename}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في حفظ ملف الوورد: {e}\")\n",
        "\n",
        "def create_markdown_from_elements(original_image_path, elements, page_dimensions, output_filename=\"docling_output.md\"):\n",
        "    md_content = f\"# مستند تم إنشاؤه بواسطة Docling و hocr-parser\\n\\n\"\n",
        "\n",
        "    original_image_filename_for_md = os.path.basename(original_image_path)\n",
        "    md_content += f\"## الصورة الأصلية\\n\"\n",
        "    md_content += f\"![الصورة الأصلية]({original_image_filename_for_md})\\n\\n\"\n",
        "\n",
        "    md_content += f\"## العناصر المستخرجة\\n\\n\"\n",
        "    if not elements:\n",
        "        md_content += \"لم يتم استخراج أي عناصر.\\n\"\n",
        "\n",
        "    img_width_px, img_height_px = 0,0\n",
        "    if page_dimensions: img_width_px, img_height_px = page_dimensions\n",
        "\n",
        "    for i, element in enumerate(elements):\n",
        "        md_content += f\"### عنصر {i+1}: {element['type']}\\n\"\n",
        "        md_content += f\"*الإحداثيات: `{element['bbox']}`*\\n\\n\"\n",
        "        if element[\"type\"] == \"text\":\n",
        "            md_content += f\"```text\\n{element['text']}\\n```\\n\\n\"\n",
        "        elif element[\"type\"] == \"image_region\":\n",
        "            md_content += \"منطقة صورة مكتشفة.\\n\"\n",
        "            if img_width_px > 0 and img_height_px > 0:\n",
        "                try:\n",
        "                    with Image.open(original_image_path) as img_orig:\n",
        "                        bbox = element['bbox']\n",
        "                        crop_box = (\n",
        "                            max(0, int(bbox[0])), max(0, int(bbox[1])),\n",
        "                            min(img_width_px, int(bbox[2])), min(img_height_px, int(bbox[3]))\n",
        "                        )\n",
        "                        if crop_box[0] < crop_box[2] and crop_box[1] < crop_box[3]:\n",
        "                            cropped_img = img_orig.crop(crop_box)\n",
        "                            md_crop_filename = f\"md_crop_{i}_{os.path.splitext(os.path.basename(original_image_path))[0]}.png\"\n",
        "                            md_crop_path = os.path.join(os.path.dirname(output_filename) or \".\", md_crop_filename)\n",
        "                            cropped_img.save(md_crop_path)\n",
        "                            md_content += f\"![منطقة صورة مقصوصة {i}]({md_crop_filename})\\n\\n\"\n",
        "                        else:\n",
        "                            md_content += \"(إحداثيات المنطقة غير صالحة للقص أو خارج الحدود)\\n\\n\"\n",
        "                except Exception as e_md_crop:\n",
        "                    print(f\"خطأ في قص وحفظ منطقة الصورة {element['bbox']} للماركداون: {e_md_crop}\")\n",
        "                    md_content += \"(خطأ أثناء محاولة قص الصورة للماركداون)\\n\\n\"\n",
        "            else:\n",
        "                 md_content += \"(لا يمكن قص المنطقة للماركداون: أبعاد الصورة الأصلية غير متاحة أو صفرية)\\n\\n\"\n",
        "        md_content += \"---\\n\\n\"\n",
        "\n",
        "    try:\n",
        "        # التأكد من أن مسار الإخراج للماركداون صحيح\n",
        "        if not os.path.dirname(output_filename) and output_filename: # إذا كان اسم ملف فقط\n",
        "            output_filename = os.path.join(os.getcwd(), output_filename)\n",
        "        elif not output_filename: # اسم ملف افتراضي إذا كان فارغًا\n",
        "             output_filename = os.path.join(os.getcwd(), \"output_from_docling.md\")\n",
        "\n",
        "        os.makedirs(os.path.dirname(output_filename), exist_ok=True) # التأكد من وجود المجلد\n",
        "        with open(output_filename, \"w\", encoding=\"utf-8\") as f:\n",
        "            f.write(md_content)\n",
        "        print(f\"تم حفظ ملف الماركداون: {output_filename}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في حفظ ملف الماركداون: {e}\")\n",
        "\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    input_image_path = \"/content/PVC-Structure-PhotoRoom.png-PhotoRoom.png\"\n",
        "\n",
        "    if not os.path.exists(input_image_path):\n",
        "        print(f\"خطأ: ملف الصورة '{input_image_path}' غير موجود.\")\n",
        "    else:\n",
        "        hocr_output_file = run_docling_ocr(input_image_path)\n",
        "\n",
        "        if hocr_output_file:\n",
        "            if HocrParser: # التحقق مرة أخرى قبل الاستخدام\n",
        "                extracted_data, page_dims = parse_hocr_with_hocr_library(hocr_output_file)\n",
        "            else:\n",
        "                print(\"لا يمكن المتابعة لأن HocrParser غير متاح.\")\n",
        "                extracted_data, page_dims = [], None\n",
        "\n",
        "            if not extracted_data:\n",
        "                print(\"لم يتم استخلاص أية عناصر من ملف hOCR.\")\n",
        "            else:\n",
        "                print(f\"\\n--- تم استخلاص {len(extracted_data)} عنصرًا ---\")\n",
        "                if page_dims: print(f\"أبعاد الصفحة المقدرة: {page_dims}\")\n",
        "                for i_el, element in enumerate(extracted_data):\n",
        "                    print(f\"العنصر {i_el+1}: النوع={element['type']}, الإحداثيات={element['bbox']}\")\n",
        "                    if element['type'] == 'text':\n",
        "                        print(f\"  النص (مقتطف): {element['text'][:80].replace(chr(10), ' ')}...\")\n",
        "                print(\"------------------------------\\n\")\n",
        "\n",
        "                word_output_filename = \"output_from_docling.docx\"\n",
        "                create_word_from_elements(input_image_path, extracted_data, page_dims, word_output_filename)\n",
        "\n",
        "                markdown_base_dir = os.getcwd()\n",
        "                markdown_output_filename = os.path.join(markdown_base_dir, \"output_from_docling.md\")\n",
        "\n",
        "                try:\n",
        "                    target_original_image_md = os.path.join(markdown_base_dir, os.path.basename(input_image_path))\n",
        "                    if not os.path.exists(target_original_image_md) or not os.path.samefile(input_image_path, target_original_image_md):\n",
        "                         shutil.copy(input_image_path, target_original_image_md)\n",
        "                except Exception as e_copy:\n",
        "                    print(f\"تنبيه: لم يتم نسخ الصورة الأصلية لملف الماركداون: {e_copy}\")\n",
        "\n",
        "                create_markdown_from_elements(input_image_path, extracted_data, page_dims, markdown_output_filename)\n",
        "        else:\n",
        "            print(\"فشل Docling OCR أو لم يتم إنتاج ملف hOCR.\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8JFGepN4RlTn",
        "outputId": "c62cfa78-0196-495f-b277-ea477d5ed917"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "خطأ: لم يتم العثور على مكتبة hocr-parser. تأكد من تثبيتها بشكل صحيح.\n",
            "يمكنك تثبيتها باستخدام: pip install git+https://github.com/jlieth/hocr-parser\n",
            "تنفيذ أمر Docling: python -m docling ocr /content/PVC-Structure-PhotoRoom.png-PhotoRoom.png docling_output\n",
            "خطأ أثناء تنفيذ Docling (رمز الخروج: 1):\n",
            "STDERR:\n",
            "/usr/bin/python3: No module named docling.__main__; 'docling' is a package and cannot be directly executed\n",
            "\n",
            "فشل Docling OCR أو لم يتم إنتاج ملف hOCR.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install git+https://github.com/jlieth/hocr-parser"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "erZSLF2rRlX9",
        "outputId": "284d6f64-8334-4f9f-f27e-dce2eec85473"
      },
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting git+https://github.com/jlieth/hocr-parser\n",
            "  Cloning https://github.com/jlieth/hocr-parser to /tmp/pip-req-build-98bwa66c\n",
            "  Running command git clone --filter=blob:none --quiet https://github.com/jlieth/hocr-parser /tmp/pip-req-build-98bwa66c\n",
            "  Resolved https://github.com/jlieth/hocr-parser to commit ed12231387bbc0e5841222152d996df4323931bc\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: lxml in /usr/local/lib/python3.11/dist-packages (from hocr-parser==0.3.0) (5.4.0)\n",
            "Requirement already satisfied: cssselect in /usr/local/lib/python3.11/dist-packages (from hocr-parser==0.3.0) (1.3.0)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "L1Bd0FzfRlbF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "https://github.com/docling-project/docling/blob/main/docs/v2.md"
      ],
      "metadata": {
        "id": "QsaBCOfWSAqg"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Convert a single file to Markdown (default)\n",
        "docling myfile.pdf\n",
        "\n",
        "# Convert a single file to Markdown and JSON, without OCR\n",
        "docling myfile.pdf --to json --to md --no-ocr\n",
        "\n",
        "# Convert PDF files in input directory to Markdown (default)\n",
        "docling ./input/dir --from pdf\n",
        "\n",
        "# Convert PDF and Word files in input directory to Markdown and JSON\n",
        "docling ./input/dir --from pdf --from docx --to md --to json --output ./scratch\n",
        "\n",
        "# Convert all supported files in input directory to Markdown, but abort on first error\n",
        "docling ./input/dir --output ./scratch --abort-on-error\n"
      ],
      "metadata": {
        "id": "hytot9dERldr"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import subprocess\n",
        "import os\n",
        "import shutil\n",
        "from PIL import Image\n",
        "from docx import Document\n",
        "from docx.shared import Inches\n",
        "import re # قد نحتاجه لاحقًا\n",
        "\n",
        "# --- محاولة استيراد HOCRParser من docling.parsers ---\n",
        "try:\n",
        "    from docling.parsers import HOCRParser\n",
        "    print(\"تم استيراد HOCRParser من docling.parsers بنجاح.\")\n",
        "except ImportError:\n",
        "    print(\"تحذير: لم يتم العثور على HOCRParser في docling.parsers.\")\n",
        "    print(\"إذا فشل تحليل hOCR، قد تحتاج إلى تثبيت مكتبة تحليل hOCR متخصصة مثل 'hocr-parser' (jlieth/hocr-parser)\")\n",
        "    print(\"أو استخدام lxml/BeautifulSoup لتحليل ملف hOCR يدويًا.\")\n",
        "    HOCRParser = None # لكي لا يفشل الكود فورًا إذا لم يتم العثور عليه\n",
        "\n",
        "# --- إعدادات ---\n",
        "DOCLING_OUTPUT_DIR = \"docling_output\"\n",
        "\n",
        "def run_docling_ocr(image_path): # <--- تأكد أن هذه الدالة محدثة\n",
        "    \"\"\"\n",
        "    ينفذ أمر docling ocr على الصورة ويخزن المخرجات.\n",
        "    يستدعي \"docling\" مباشرة.\n",
        "    \"\"\"\n",
        "    if not os.path.exists(image_path):\n",
        "        print(f\"خطأ: ملف الصورة المدخل '{image_path}' غير موجود.\")\n",
        "        return None\n",
        "\n",
        "    if os.path.exists(DOCLING_OUTPUT_DIR):\n",
        "        shutil.rmtree(DOCLING_OUTPUT_DIR)\n",
        "    os.makedirs(DOCLING_OUTPUT_DIR, exist_ok=True)\n",
        "\n",
        "    # --- التعديل الأساسي هنا ---\n",
        "    cmd = [\"docling\", \"ocr\", image_path, DOCLING_OUTPUT_DIR]\n",
        "    # ---------------------------\n",
        "\n",
        "    print(f\"تنفيذ أمر Docling: {' '.join(cmd)}\")\n",
        "    try:\n",
        "        process = subprocess.run(cmd, capture_output=True, text=True, check=False, encoding='utf-8', errors='ignore')\n",
        "\n",
        "        if process.returncode != 0:\n",
        "            print(f\"خطأ أثناء تنفيذ Docling (رمز الخروج: {process.returncode}):\")\n",
        "            if process.stdout: print(f\"STDOUT:\\n{process.stdout}\")\n",
        "            if process.stderr: print(f\"STDERR:\\n{process.stderr}\") # مهم جدًا هنا\n",
        "            return None\n",
        "        else:\n",
        "            if process.stdout and process.stdout.strip(): print(f\"Docling STDOUT:\\n{process.stdout}\")\n",
        "            if process.stderr and process.stderr.strip(): print(f\"Docling STDERR (تحذيرات محتملة):\\n{process.stderr}\")\n",
        "            print(f\"Docling أنجز بنجاح. المخرجات في: {DOCLING_OUTPUT_DIR}\")\n",
        "\n",
        "            for fname in os.listdir(DOCLING_OUTPUT_DIR):\n",
        "                if fname.endswith(\".hocr\"):\n",
        "                    return os.path.join(DOCLING_OUTPUT_DIR, fname)\n",
        "            print(\"لم يتم العثور على ملف hOCR في المخرجات.\")\n",
        "            return None\n",
        "    except FileNotFoundError:\n",
        "        print(\"خطأ: لم يتم العثور على أمر 'docling'. هل هو مثبت بشكل صحيح وفي PATH؟\")\n",
        "        return None\n",
        "    except Exception as e:\n",
        "        print(f\"حدث خطأ غير متوقع أثناء تشغيل Docling: {e}\")\n",
        "        return None\n",
        "\n",
        "def parse_hocr_with_docling_internal_parser(hocr_file_path):\n",
        "    \"\"\"\n",
        "    يحلل ملف hOCR باستخدام HOCRParser المفترض وجوده في docling.parsers\n",
        "    ويستخرج النصوص وإحداثياتها ومناطق الصور.\n",
        "    \"\"\"\n",
        "    if HOCRParser is None:\n",
        "        print(\"HOCRParser من docling.parsers غير متاح. لا يمكن تحليل الملف بهذه الطريقة.\")\n",
        "        return [], None\n",
        "\n",
        "    if not hocr_file_path or not os.path.exists(hocr_file_path):\n",
        "        print(f\"ملف HOCR غير موجود: {hocr_file_path}\")\n",
        "        return [], None\n",
        "\n",
        "    extracted_elements = []\n",
        "    page_dimensions = None\n",
        "\n",
        "    try:\n",
        "        parser = HOCRParser()\n",
        "        doc_layout = parser.parse(hocr_file_path)\n",
        "\n",
        "        if hasattr(doc_layout, 'pages') and doc_layout.pages:\n",
        "            first_page = doc_layout.pages[0]\n",
        "            if hasattr(first_page, 'bbox') and first_page.bbox:\n",
        "                 x1, y1, x2, y2 = first_page.bbox\n",
        "                 page_dimensions = (x2 - x1, y2 - y1)\n",
        "            elif hasattr(first_page, 'width') and hasattr(first_page, 'height'):\n",
        "                 page_dimensions = (first_page.width, first_page.height)\n",
        "\n",
        "        for page in doc_layout.pages:\n",
        "            for block in page.blocks:\n",
        "                if block.block_type == \"text\":\n",
        "                    block_text_content = \"\"\n",
        "                    for line in block.lines:\n",
        "                        word_texts = [word.text for word in line.words if hasattr(word, 'text') and word.text]\n",
        "                        if word_texts:\n",
        "                            block_text_content += \" \".join(word_texts) + \"\\n\"\n",
        "\n",
        "                    if block_text_content.strip() and hasattr(block, 'bbox'):\n",
        "                        extracted_elements.append({\n",
        "                            \"type\": \"text\",\n",
        "                            \"text\": block_text_content.strip(),\n",
        "                            \"bbox\": block.bbox\n",
        "                        })\n",
        "                elif block.block_type in [\"image\", \"figure\", \"picture\"] and hasattr(block, 'bbox'):\n",
        "                    extracted_elements.append({\n",
        "                        \"type\": \"image_region\",\n",
        "                        \"bbox\": block.bbox,\n",
        "                        \"image_path\": None\n",
        "                    })\n",
        "\n",
        "        extracted_elements.sort(key=lambda el: (el[\"bbox\"][1], el[\"bbox\"][0]))\n",
        "        return extracted_elements, page_dimensions\n",
        "\n",
        "    except AttributeError as e_attr:\n",
        "        print(f\"خطأ سمة (AttributeError) أثناء استخدام HOCRParser من docling: {e_attr}\")\n",
        "        print(\"راجع وثائق docling أو استخدم dir() لفحص الكائنات.\")\n",
        "        return [], None\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في تحليل ملف hOCR باستخدام HOCRParser من docling: {e}\")\n",
        "        import traceback\n",
        "        traceback.print_exc()\n",
        "        return [], None\n",
        "\n",
        "def create_word_from_elements(original_image_path, elements, page_dimensions, output_filename=\"docling_output.docx\"):\n",
        "    doc = Document()\n",
        "    doc.add_heading('مستند تم إنشاؤه بواسطة Docling', level=1)\n",
        "\n",
        "    img_width_px, img_height_px = 0, 0\n",
        "    if page_dimensions:\n",
        "        img_width_px, img_height_px = page_dimensions\n",
        "    else:\n",
        "        try:\n",
        "            with Image.open(original_image_path) as img:\n",
        "                img_width_px, img_height_px = img.size\n",
        "        except Exception as e_img_dim:\n",
        "            print(f\"تنبيه: لم يتمكن من الحصول على أبعاد الصورة الأصلية: {e_img_dim}\")\n",
        "\n",
        "    try:\n",
        "        doc.add_paragraph(\"الصورة الأصلية:\")\n",
        "        doc.add_picture(original_image_path, width=Inches(6.0))\n",
        "        doc.add_page_break()\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في إضافة الصورة الأصلية إلى الوورد: {e}\")\n",
        "\n",
        "    doc.add_heading('العناصر المستخرجة بواسطة Docling:', level=2)\n",
        "\n",
        "    if not elements:\n",
        "        doc.add_paragraph(\"لم يتم استخراج أي عناصر.\")\n",
        "\n",
        "    for i, element in enumerate(elements):\n",
        "        doc.add_paragraph(f\"--- عنصر {i+1} ---\")\n",
        "        doc.add_paragraph(f\"النوع: {element['type']}\")\n",
        "        doc.add_paragraph(f\"الإحداثيات (bbox): {element['bbox']}\")\n",
        "\n",
        "        if element[\"type\"] == \"text\":\n",
        "            doc.add_paragraph(\"النص:\")\n",
        "            doc.add_paragraph(element[\"text\"])\n",
        "        elif element[\"type\"] == \"image_region\":\n",
        "            doc.add_paragraph(\"منطقة صورة مكتشفة.\")\n",
        "            if img_width_px > 0 and img_height_px > 0 and element.get('bbox'):\n",
        "                try:\n",
        "                    with Image.open(original_image_path) as img_orig:\n",
        "                        bbox = element['bbox']\n",
        "                        crop_box = (\n",
        "                            max(0, int(bbox[0])), max(0, int(bbox[1])),\n",
        "                            min(img_width_px, int(bbox[2])), min(img_height_px, int(bbox[3]))\n",
        "                        )\n",
        "                        if crop_box[0] < crop_box[2] and crop_box[1] < crop_box[3]:\n",
        "                            cropped_img = img_orig.crop(crop_box)\n",
        "                            aspect_ratio = cropped_img.width / cropped_img.height if cropped_img.height > 0 else 1\n",
        "                            display_width_inches = min(3.0, 6.0 * (cropped_img.width / img_width_px if img_width_px > 0 else 1))\n",
        "                            display_width = Inches(display_width_inches if display_width_inches > 0.1 else 0.5)\n",
        "\n",
        "                            temp_cropped_path = os.path.join(DOCLING_OUTPUT_DIR, f\"temp_crop_word_{i}.png\")\n",
        "                            cropped_img.save(temp_cropped_path)\n",
        "                            doc.add_picture(temp_cropped_path, width=display_width)\n",
        "                            os.remove(temp_cropped_path)\n",
        "                        else:\n",
        "                             doc.add_paragraph(\"(إحداثيات المنطقة غير صالحة للقص أو خارج الحدود)\")\n",
        "                except Exception as e_crop:\n",
        "                     print(f\"خطأ في قص وإضافة منطقة الصورة {element['bbox']} من الأصلية: {e_crop}\")\n",
        "            elif not element.get('bbox'):\n",
        "                doc.add_paragraph(\"(لم يتم توفير إحداثيات للمنطقة)\")\n",
        "            else:\n",
        "                doc.add_paragraph(\"(لا يمكن قص المنطقة: أبعاد الصورة الأصلية غير متاحة أو صفرية)\")\n",
        "    try:\n",
        "        doc.save(output_filename)\n",
        "        print(f\"تم حفظ ملف الوورد: {output_filename}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في حفظ ملف الوورد: {e}\")\n",
        "\n",
        "def create_markdown_from_elements(original_image_path, elements, page_dimensions, output_filename=\"docling_output.md\"):\n",
        "    md_content = f\"# مستند تم إنشاؤه بواسطة Docling\\n\\n\"\n",
        "\n",
        "    original_image_filename_for_md = os.path.basename(original_image_path)\n",
        "    md_content += f\"## الصورة الأصلية\\n\"\n",
        "    md_content += f\"![الصورة الأصلية]({original_image_filename_for_md})\\n\\n\"\n",
        "\n",
        "    md_content += f\"## العناصر المستخرجة\\n\\n\"\n",
        "    if not elements:\n",
        "        md_content += \"لم يتم استخراج أي عناصر.\\n\"\n",
        "\n",
        "    img_width_px, img_height_px = 0,0\n",
        "    if page_dimensions: img_width_px, img_height_px = page_dimensions\n",
        "    else:\n",
        "        try:\n",
        "            with Image.open(original_image_path) as img:\n",
        "                img_width_px, img_height_px = img.size\n",
        "        except Exception: pass\n",
        "\n",
        "    for i, element in enumerate(elements):\n",
        "        md_content += f\"### عنصر {i+1}: {element['type']}\\n\"\n",
        "        md_content += f\"*الإحداثيات: `{element['bbox']}`*\\n\\n\"\n",
        "        if element[\"type\"] == \"text\":\n",
        "            md_content += f\"```text\\n{element['text']}\\n```\\n\\n\"\n",
        "        elif element[\"type\"] == \"image_region\":\n",
        "            md_content += \"منطقة صورة مكتشفة.\\n\"\n",
        "            if img_width_px > 0 and img_height_px > 0 and element.get('bbox'):\n",
        "                try:\n",
        "                    with Image.open(original_image_path) as img_orig:\n",
        "                        bbox = element['bbox']\n",
        "                        crop_box = (\n",
        "                            max(0, int(bbox[0])), max(0, int(bbox[1])),\n",
        "                            min(img_width_px, int(bbox[2])), min(img_height_px, int(bbox[3]))\n",
        "                        )\n",
        "                        if crop_box[0] < crop_box[2] and crop_box[1] < crop_box[3]:\n",
        "                            cropped_img = img_orig.crop(crop_box)\n",
        "                            md_crop_filename = f\"md_crop_{i}_{os.path.splitext(os.path.basename(original_image_path))[0]}.png\"\n",
        "                            output_dir_md = os.path.dirname(output_filename) or os.getcwd()\n",
        "                            md_crop_path = os.path.join(output_dir_md, md_crop_filename)\n",
        "\n",
        "                            cropped_img.save(md_crop_path)\n",
        "                            md_content += f\"![منطقة صورة مقصوصة {i}]({md_crop_filename})\\n\\n\"\n",
        "                        else:\n",
        "                            md_content += \"(إحداثيات المنطقة غير صالحة للقص أو خارج الحدود)\\n\\n\"\n",
        "                except Exception as e_md_crop:\n",
        "                    print(f\"خطأ في قص وحفظ منطقة الصورة {element['bbox']} للماركداون: {e_md_crop}\")\n",
        "                    md_content += \"(خطأ أثناء محاولة قص الصورة للماركداون)\\n\\n\"\n",
        "            elif not element.get('bbox'):\n",
        "                md_content += \"(لم يتم توفير إحداثيات للمنطقة)\\n\\n\"\n",
        "            else:\n",
        "                 md_content += \"(لا يمكن قص المنطقة للماركداون: أبعاد الصورة الأصلية غير متاحة أو صفرية)\\n\\n\"\n",
        "        md_content += \"---\\n\\n\"\n",
        "\n",
        "    try:\n",
        "        output_dir_md_final = os.path.dirname(output_filename) or os.getcwd()\n",
        "        os.makedirs(output_dir_md_final, exist_ok=True)\n",
        "        final_md_path = os.path.join(output_dir_md_final, os.path.basename(output_filename))\n",
        "\n",
        "        with open(final_md_path, \"w\", encoding=\"utf-8\") as f:\n",
        "            f.write(md_content)\n",
        "        print(f\"تم حفظ ملف الماركداون: {final_md_path}\")\n",
        "    except Exception as e:\n",
        "        print(f\"خطأ في حفظ ملف الماركداون: {e}\")\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    input_image_path = \"/content/PVC-Structure-PhotoRoom.png-PhotoRoom.png\"\n",
        "\n",
        "    if not os.path.exists(input_image_path):\n",
        "        print(f\"خطأ: ملف الصورة '{input_image_path}' غير موجود.\")\n",
        "    else:\n",
        "        hocr_output_file = run_docling_ocr(input_image_path)\n",
        "\n",
        "        if hocr_output_file:\n",
        "            extracted_data, page_dims = parse_hocr_with_docling_internal_parser(hocr_output_file)\n",
        "\n",
        "            if not extracted_data:\n",
        "                print(\"لم يتم استخلاص أية عناصر من ملف hOCR.\")\n",
        "            else:\n",
        "                print(f\"\\n--- تم استخلاص {len(extracted_data)} عنصرًا ---\")\n",
        "                if page_dims: print(f\"أبعاد الصفحة المقدرة من HOCRParser: {page_dims}\")\n",
        "                for i_el, element in enumerate(extracted_data):\n",
        "                    print(f\"العنصر {i_el+1}: النوع={element['type']}, الإحداثيات={element['bbox']}\")\n",
        "                    if element['type'] == 'text':\n",
        "                        print(f\"  النص (مقتطف): {element['text'][:80].replace(chr(10), ' ')}...\")\n",
        "                print(\"------------------------------\\n\")\n",
        "\n",
        "                word_output_filename = \"output_from_docling.docx\"\n",
        "                create_word_from_elements(input_image_path, extracted_data, page_dims, word_output_filename)\n",
        "\n",
        "                markdown_base_dir = os.getcwd()\n",
        "                markdown_output_filename = os.path.join(markdown_base_dir, \"output_from_docling.md\")\n",
        "\n",
        "                try:\n",
        "                    target_original_image_md = os.path.join(markdown_base_dir, os.path.basename(input_image_path))\n",
        "                    if not os.path.exists(target_original_image_md) or not os.path.samefile(input_image_path, target_original_image_md):\n",
        "                         shutil.copy(input_image_path, target_original_image_md)\n",
        "                except Exception as e_copy:\n",
        "                    print(f\"تنبيه: لم يتم نسخ الصورة الأصلية لملف الماركداون: {e_copy}\")\n",
        "\n",
        "                create_markdown_from_elements(input_image_path, extracted_data, page_dims, markdown_output_filename)\n",
        "        else:\n",
        "            print(\"فشل Docling OCR أو لم يتم إنتاج ملف hOCR.\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XbeEyl1iRlhF",
        "outputId": "204c3c47-af68-44e7-907e-a45bb69162a8"
      },
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "تحذير: لم يتم العثور على HOCRParser في docling.parsers.\n",
            "إذا فشل تحليل hOCR، قد تحتاج إلى تثبيت مكتبة تحليل hOCR متخصصة مثل 'hocr-parser' (jlieth/hocr-parser)\n",
            "أو استخدام lxml/BeautifulSoup لتحليل ملف hOCR يدويًا.\n",
            "تنفيذ أمر Docling: docling ocr /content/PVC-Structure-PhotoRoom.png-PhotoRoom.png docling_output\n",
            "خطأ أثناء تنفيذ Docling (رمز الخروج: 1):\n",
            "STDERR:\n",
            "2025-05-10 22:21:08.379331: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746915668.423567   42375 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746915668.430794   42375 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "Error: The input file ocr does not exist.\n",
            "Aborted.\n",
            "\n",
            "فشل Docling OCR أو لم يتم إنتاج ملف hOCR.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "4yld58HYObpO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "BvNNs0yxM9zn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "j6VZjOrPM9v_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "DYPBPDdkM9sm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "fUfKAuj4M9pL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "0Hn6mKbdSaEl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "L8mUMa02SaIY"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "NkxYAOqXSaLU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "-rlLIOaVSaN5"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "YOtrkI6dSaQ_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!docling /content/PVC-Structure-PhotoRoom.png-PhotoRoom.png --from image --to md --output /content"
      ],
      "metadata": {
        "id": "-q7kGgHMJs9m"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!docling --help"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "uq_VKO9MJVOI",
        "outputId": "423f9d2d-ee00-46b0-d7a1-5f8cdc61f00c"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 21:42:10.031501: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746913330.056228   32726 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746913330.063543   32726 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "2025-05-10 21:42:10.089399: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
            "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
            "\u001b[1m                                                                                \u001b[0m\n",
            "\u001b[1m \u001b[0m\u001b[1;33mUsage: \u001b[0m\u001b[1mdocling [OPTIONS] source\u001b[0m\u001b[1m                                               \u001b[0m\u001b[1m \u001b[0m\n",
            "\u001b[1m                                                                                \u001b[0m\n",
            "\u001b[2m╭─\u001b[0m\u001b[2m Arguments \u001b[0m\u001b[2m─────────────────────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[31m*\u001b[0m    input_sources      \u001b[1;33msource\u001b[0m  PDF files to convert. Can be local file /    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                 directory paths or URL.                      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                 \u001b[2m[default: None]                             \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                 \u001b[2;31m[required]                                  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
            "\u001b[2m╭─\u001b[0m\u001b[2m Options \u001b[0m\u001b[2m───────────────────────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-from\u001b[0m                                    \u001b[1;2;33m[\u001b[0m\u001b[1;33mdocx\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mpptx\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mhtml\u001b[0m  Specify input     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;2;33m|\u001b[0m\u001b[1;33mimage\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mpdf\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33masci\u001b[0m  formats to        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33midoc\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mmd\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mcsv\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mxls\u001b[0m  convert from.     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mx\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mxml_uspto\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mxml\u001b[0m  Defaults to all   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33m_jats\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mjson_docl\u001b[0m  formats.          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33ming\u001b[0m\u001b[1;2;33m]\u001b[0m\u001b[1;33m           \u001b[0m  \u001b[2m[default: None]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-to\u001b[0m                                      \u001b[1;2;33m[\u001b[0m\u001b[1;33mmd\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mjson\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mhtml\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mh\u001b[0m  Specify output    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mtml_split_page\u001b[0m\u001b[1;2;33m|\u001b[0m  formats. Defaults \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mtext\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mdoctags\u001b[0m\u001b[1;2;33m]\u001b[0m\u001b[1;33m  \u001b[0m  to Markdown.      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: None]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-headers\u001b[0m                                 \u001b[1;33mTEXT           \u001b[0m  Specify http      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            request headers   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            used when         \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            fetching url      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            input sources in  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            the form of a     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            JSON string       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: None]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-image\u001b[0m\u001b[1;36m-export-…\u001b[0m                          \u001b[1;2;33m[\u001b[0m\u001b[1;33mplaceholder\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mem\u001b[0m  Image export mode \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mbedded\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mreferenc\u001b[0m  for the document  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33med\u001b[0m\u001b[1;2;33m]\u001b[0m\u001b[1;33m            \u001b[0m  (only in case of  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            JSON, Markdown or \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            HTML). With       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            `placeholder`,    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            only the position \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            of the image is   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            marked in the     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            output. In        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            `embedded` mode,  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            the image is      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            embedded as       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            base64 encoded    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            string. In        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            `referenced`      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            mode, the image   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            is exported in    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            PNG format and    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            referenced from   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            the main exported \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            document.         \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2membedded]        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-pipeline\u001b[0m                                \u001b[1;2;33m[\u001b[0m\u001b[1;33mstandard\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mvlm\u001b[0m\u001b[1;2;33m]\u001b[0m\u001b[1;33m \u001b[0m  Choose the        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            pipeline to       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            process PDF or    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            image files.      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mstandard]        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-vlm\u001b[0m\u001b[1;36m-model\u001b[0m                               \u001b[1;2;33m[\u001b[0m\u001b[1;33msmoldocling\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mgr\u001b[0m  Choose the VLM    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33manite_vision\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mgr\u001b[0m  model to use with \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33manite_vision_ol\u001b[0m  PDF or image      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mlama\u001b[0m\u001b[1;2;33m]\u001b[0m\u001b[1;33m          \u001b[0m  files.            \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2msmoldocling]     \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-ocr\u001b[0m                 \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-ocr\u001b[0m            \u001b[1;33m               \u001b[0m  If enabled, the   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            bitmap content    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            will be processed \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            using OCR.        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: ocr]   \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-force\u001b[0m\u001b[1;36m-ocr\u001b[0m           \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-force-ocr\u001b[0m      \u001b[1;33m               \u001b[0m  Replace any       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            existing text     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            with OCR          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            generated text    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            over the full     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            content.          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-force-ocr]    \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-ocr\u001b[0m\u001b[1;36m-engine\u001b[0m                              \u001b[1;33mTEXT           \u001b[0m  The OCR engine to \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            use. When         \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[1;36m-\u001b[0m\u001b[1;36m-allow\u001b[0m\u001b[1;36m-external…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            is *not* set, the \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            available values  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            are: easyocr,     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            ocrmac, rapidocr, \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            tesserocr,        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            tesseract. Use    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            the option        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[1;36m-\u001b[0m\u001b[1;36m-show\u001b[0m\u001b[1;36m-external-…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            to see the        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            options allowed   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            with external     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            plugins.          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2measyocr]         \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-ocr\u001b[0m\u001b[1;36m-lang\u001b[0m                                \u001b[1;33mTEXT           \u001b[0m  Provide a         \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            comma-separated   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            list of languages \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            used by the OCR   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            engine. Note that \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            each OCR engine   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            has different     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            values for the    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            language names.   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: None]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-pdf\u001b[0m\u001b[1;36m-backend\u001b[0m                             \u001b[1;2;33m[\u001b[0m\u001b[1;33mpypdfium2\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mdlpa\u001b[0m  The PDF backend   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mrse_v1\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mdlparse_\u001b[0m  to use.           \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mv2\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mdlparse_v4\u001b[0m\u001b[1;2;33m]\u001b[0m\u001b[1;33m \u001b[0m  \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mdlparse_v2]      \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-table\u001b[0m\u001b[1;36m-mode\u001b[0m                              \u001b[1;2;33m[\u001b[0m\u001b[1;33mfast\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33maccurate\u001b[0m\u001b[1;2;33m]\u001b[0m  The mode to use   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            in the table      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            structure model.  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2maccurate]        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-enrich\u001b[0m\u001b[1;36m-code\u001b[0m         \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-enrich-code\u001b[0m    \u001b[1;33m               \u001b[0m  Enable the code   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            enrichment model  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            in the pipeline.  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-enrich-code]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-enrich\u001b[0m\u001b[1;36m-formula\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-enrich-for…\u001b[0m    \u001b[1;33m               \u001b[0m  Enable the        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            formula           \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            enrichment model  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            in the pipeline.  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-enrich-formul…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-enrich\u001b[0m\u001b[1;36m-pictur…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-enrich-pic…\u001b[0m    \u001b[1;33m               \u001b[0m  Enable the        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            picture           \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            classification    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            enrichment model  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            in the pipeline.  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-enrich-pictur…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-enrich\u001b[0m\u001b[1;36m-pictur…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-enrich-pic…\u001b[0m    \u001b[1;33m               \u001b[0m  Enable the        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            picture           \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            description model \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            in the pipeline.  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-enrich-pictur…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-artifacts\u001b[0m\u001b[1;36m-path\u001b[0m                          \u001b[1;33mPATH           \u001b[0m  If provided, the  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            location of the   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            model artifacts.  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: None]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-enable\u001b[0m\u001b[1;36m-remote…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-enable-rem…\u001b[0m    \u001b[1;33m               \u001b[0m  Must be enabled   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            when using models \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            connecting to     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            remote services.  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-enable-remote…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-allow\u001b[0m\u001b[1;36m-externa…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-allow-exte…\u001b[0m    \u001b[1;33m               \u001b[0m  Must be enabled   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            for loading       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            modules from      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            third-party       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            plugins.          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-allow-externa…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-show\u001b[0m\u001b[1;36m-external…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-show-exter…\u001b[0m    \u001b[1;33m               \u001b[0m  List the          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            third-party       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            plugins which are \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            available when    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            the option        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[1;36m-\u001b[0m\u001b[1;36m-allow\u001b[0m\u001b[1;36m-external…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            is set.           \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-show-external…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-abort\u001b[0m\u001b[1;36m-on-error\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-abort-on-e…\u001b[0m    \u001b[1;33m               \u001b[0m  If enabled, the   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            processing will   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            be aborted when   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            the first error   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            is encountered.   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-abort-on-erro…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-output\u001b[0m                                  \u001b[1;33mPATH           \u001b[0m  Output directory  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            where results are \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            saved.            \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: .]     \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-verbose\u001b[0m         \u001b[1;32m-v\u001b[0m                      \u001b[1;33mINTEGER        \u001b[0m  Set the verbosity \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            level. \u001b[1;32m-v\u001b[0m for     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            info logging, \u001b[1;32m-vv\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            for debug         \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            logging.          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: 0]     \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-debug\u001b[0m\u001b[1;36m-visuali…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-debug-visu…\u001b[0m    \u001b[1;33m               \u001b[0m  Enable debug      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            output which      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            visualizes the    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            PDF cells         \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-debug-visuali…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-debug\u001b[0m\u001b[1;36m-visuali…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-debug-visu…\u001b[0m    \u001b[1;33m               \u001b[0m  Enable debug      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            output which      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            visualizes the    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            OCR cells         \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-debug-visuali…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-debug\u001b[0m\u001b[1;36m-visuali…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-debug-visu…\u001b[0m    \u001b[1;33m               \u001b[0m  Enable debug      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            output which      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            visualizes the    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            layour clusters   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-debug-visuali…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-debug\u001b[0m\u001b[1;36m-visuali…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-debug-visu…\u001b[0m    \u001b[1;33m               \u001b[0m  Enable debug      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            output which      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            visualizes the    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            table cells       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-debug-visuali…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-version\u001b[0m                                 \u001b[1;33m               \u001b[0m  Show version      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            information.      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-document\u001b[0m\u001b[1;36m-time…\u001b[0m                          \u001b[1;33mFLOAT          \u001b[0m  The timeout for   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            processing each   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            document, in      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            seconds.          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: None]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-num\u001b[0m\u001b[1;36m-threads\u001b[0m                             \u001b[1;33mINTEGER        \u001b[0m  Number of threads \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: 4]     \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-device\u001b[0m                                  \u001b[1;2;33m[\u001b[0m\u001b[1;33mauto\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mcpu\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mcuda\u001b[0m\u001b[1;2;33m|\u001b[0m  Accelerator       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mmps\u001b[0m\u001b[1;2;33m]\u001b[0m\u001b[1;33m           \u001b[0m  device            \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: auto]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-logo\u001b[0m                                    \u001b[1;33m               \u001b[0m  Docling logo      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-help\u001b[0m                                    \u001b[1;33m               \u001b[0m  Show this message \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            and exit.         \u001b[2m│\u001b[0m\n",
            "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### شغال حول صورة بها نص وشكل الى ماركدون يتعدل عليه بنفس الل ف الصورة"
      ],
      "metadata": {
        "id": "s1ebj9gGK_Vu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!docling /content/PVC-Structure-PhotoRoom.png-PhotoRoom.png --from image --to md --output /content"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "kFKxpWGRJeFW",
        "outputId": "07410967-9160-4a6d-b786-f2611bc4623e"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 21:45:34.626973: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746913534.666310   33554 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746913534.674095   33554 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "2025-05-10 21:45:34.699067: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
            "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!docling /content/2408.09869v5.pdf --from pdf --to md --output /content"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WmtPMJ0cKQWY",
        "outputId": "3241efc7-ba8e-435f-ee65-496976ab5e46"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 21:50:43.240720: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746913843.265026   34801 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746913843.272125   34801 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "2025-05-10 21:50:43.296504: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
            "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "GrbIQFFUUBSo"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "KDREX2gGUBPY"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!docling /content/wik2.pdf --from pdf --to md --output /content --ocr-lang eng"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ni1QOj6WUBLw",
        "outputId": "470f28a8-1e4e-411d-9e7b-3aebff268cb3"
      },
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 22:37:37.665231: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746916657.690892   46352 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746916657.698428   46352 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/bin/docling\", line 8, in <module>\n",
            "    sys.exit(app())\n",
            "             ^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/typer/main.py\", line 340, in __call__\n",
            "    raise e\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/typer/main.py\", line 323, in __call__\n",
            "    return get_command(self)(*args, **kwargs)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/click/core.py\", line 1161, in __call__\n",
            "    return self.main(*args, **kwargs)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/typer/core.py\", line 677, in main\n",
            "    return _main(\n",
            "           ^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/typer/core.py\", line 195, in _main\n",
            "    rv = self.invoke(ctx)\n",
            "         ^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/click/core.py\", line 1443, in invoke\n",
            "    return ctx.invoke(self.callback, **ctx.params)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/click/core.py\", line 788, in invoke\n",
            "    return __callback(*args, **kwargs)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/typer/main.py\", line 698, in wrapper\n",
            "    return callback(**use_params)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/cli/main.py\", line 593, in convert\n",
            "    export_documents(\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/cli/main.py\", line 167, in export_documents\n",
            "    for conv_res in conv_results:\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/document_converter.py\", line 243, in convert_all\n",
            "    for conv_res in conv_res_iter:\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/document_converter.py\", line 278, in _convert\n",
            "    for item in map(\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/document_converter.py\", line 324, in _process_document\n",
            "    conv_res = self._execute_pipeline(in_doc, raises_on_error=raises_on_error)\n",
            "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/document_converter.py\", line 345, in _execute_pipeline\n",
            "    pipeline = self._get_pipeline(in_doc.format)\n",
            "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/document_converter.py\", line 307, in _get_pipeline\n",
            "    self.initialized_pipelines[cache_key] = pipeline_class(\n",
            "                                            ^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/pipeline/standard_pdf_pipeline.py\", line 65, in __init__\n",
            "    ocr_model = self.get_ocr_model(artifacts_path=artifacts_path)\n",
            "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/pipeline/standard_pdf_pipeline.py\", line 152, in get_ocr_model\n",
            "    return factory.create_instance(\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/models/factories/base_factory.py\", line 57, in create_instance\n",
            "    return _cls(options=options, **kwargs)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/models/easyocr_model.py\", line 82, in __init__\n",
            "    self.reader = easyocr.Reader(\n",
            "                  ^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/easyocr/easyocr.py\", line 109, in __init__\n",
            "    raise ValueError(unknown_lang, 'is not supported')\n",
            "ValueError: ({'eng'}, 'is not supported')\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "--ocr-lang"
      ],
      "metadata": {
        "id": "c05kSibfWKdp"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!docling /content/wik2.pdf --from pdf --to md --output /content --ocr-lang en"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Oq9JZHb-WkAa",
        "outputId": "56e615cc-e5de-4b56-aa79-4b1d8f7c2696"
      },
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 22:40:03.271191: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746916803.295769   46952 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746916803.302864   46952 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "WARNING:easyocr.easyocr:Downloading recognition model, please wait. This may take several minutes depending upon your network connection.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!docling --ocr-lang"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jvKuHbOTWl6p",
        "outputId": "a74189e3-4b30-4baa-cc27-31f0d3d53a4a"
      },
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 22:39:35.710897: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746916775.735680   46832 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746916775.743015   46832 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "\u001b[31m╭─\u001b[0m\u001b[31m Error \u001b[0m\u001b[31m─────────────────────────────────────────────────────────────────────\u001b[0m\u001b[31m─╮\u001b[0m\n",
            "\u001b[31m│\u001b[0m Option '\u001b[1;36m-\u001b[0m\u001b[1;36m-ocr\u001b[0m\u001b[1;36m-lang\u001b[0m' requires an argument.                                    \u001b[31m│\u001b[0m\n",
            "\u001b[31m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!docling /content/wiki.pdf --from pdf --to md --output /content --ocr-engine --ocr-lang"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DHDgYM48WnZR",
        "outputId": "619e2429-2904-45c9-afcb-913691b6f752"
      },
      "execution_count": 26,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 22:50:44.452288: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746917444.478645   49544 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746917444.486236   49544 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/bin/docling\", line 8, in <module>\n",
            "    sys.exit(app())\n",
            "             ^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/typer/main.py\", line 340, in __call__\n",
            "    raise e\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/typer/main.py\", line 323, in __call__\n",
            "    return get_command(self)(*args, **kwargs)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/click/core.py\", line 1161, in __call__\n",
            "    return self.main(*args, **kwargs)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/typer/core.py\", line 677, in main\n",
            "    return _main(\n",
            "           ^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/typer/core.py\", line 195, in _main\n",
            "    rv = self.invoke(ctx)\n",
            "         ^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/click/core.py\", line 1443, in invoke\n",
            "    return ctx.invoke(self.callback, **ctx.params)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/click/core.py\", line 788, in invoke\n",
            "    return __callback(*args, **kwargs)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/typer/main.py\", line 698, in wrapper\n",
            "    return callback(**use_params)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/cli/main.py\", line 490, in convert\n",
            "    ocr_options: OcrOptions = ocr_factory.create_options(  # type: ignore\n",
            "                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/models/factories/base_factory.py\", line 65, in create_options\n",
            "    raise RuntimeError(self._err_msg_on_class_not_found(kind))\n",
            "RuntimeError: No class found with the name '--ocr-lang', known classes are:\n",
            "\t'easyocr' => <class 'docling.models.easyocr_model.EasyOcrModel'>\n",
            "\t'ocrmac' => <class 'docling.models.ocr_mac_model.OcrMacModel'>\n",
            "\t'rapidocr' => <class 'docling.models.rapid_ocr_model.RapidOcrModel'>\n",
            "\t'tesserocr' => <class 'docling.models.tesseract_ocr_model.TesseractOcrModel'>\n",
            "\t'tesseract' => <class 'docling.models.tesseract_ocr_cli_model.TesseractOcrCliModel'>\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!docling-tools models download"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9G0kOgfGYtnp",
        "outputId": "92371d79-3a45-416a-8469-b07f5516934c"
      },
      "execution_count": 27,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 22:52:47.808197: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746917567.839478   50036 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746917567.851794   50036 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "\u001b[34mDownloading layout model\u001b[0m\u001b[33m...\u001b[0m                               \u001b]8;id=924146;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py\u001b\\\u001b[2mmodel_downloader.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=818318;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py#40\u001b\\\u001b[2m40\u001b[0m\u001b]8;;\u001b\\\n",
            "Fetching 11 files: 100% 11/11 [00:00<00:00, 14.00it/s]\n",
            "\u001b[34mDownloading tableformer model\u001b[0m\u001b[33m...\u001b[0m                          \u001b]8;id=557265;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py\u001b\\\u001b[2mmodel_downloader.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=450227;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py#48\u001b\\\u001b[2m48\u001b[0m\u001b]8;;\u001b\\\n",
            "Fetching 11 files: 100% 11/11 [00:06<00:00,  1.73it/s]\n",
            "\u001b[34mDownloading picture classifier model\u001b[0m\u001b[33m...\u001b[0m                   \u001b]8;id=608653;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py\u001b\\\u001b[2mmodel_downloader.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=575915;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py#56\u001b\\\u001b[2m56\u001b[0m\u001b]8;;\u001b\\\n",
            "Fetching 4 files: 100% 4/4 [00:00<00:00, 1348.33it/s]\n",
            "\u001b[34mDownloading code formula model\u001b[0m\u001b[33m...\u001b[0m                         \u001b]8;id=938750;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py\u001b\\\u001b[2mmodel_downloader.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=175219;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py#64\u001b\\\u001b[2m64\u001b[0m\u001b]8;;\u001b\\\n",
            "Fetching 12 files: 100% 12/12 [00:00<00:00, 1566.01it/s]\n",
            "\u001b[34mDownloading easyocr models\u001b[0m\u001b[33m...\u001b[0m                             \u001b]8;id=636908;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py\u001b\\\u001b[2mmodel_downloader.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=965938;file:///usr/local/lib/python3.11/dist-packages/docling/utils/model_downloader.py#90\u001b\\\u001b[2m90\u001b[0m\u001b]8;;\u001b\\\n",
            "100% 73.7M/73.7M [00:00<00:00, 317MB/s]\n",
            "100% 13.4M/13.4M [00:00<00:00, 257MB/s]\n",
            "100% 13.6M/13.6M [00:00<00:00, 280MB/s]\n",
            "\u001b[32m\n",
            "Models downloaded into: /root/.cache/docling/models.\u001b[0m\n",
            "\n",
            " Docling can now be configured for running offline using the local artifacts.\n",
            "\n",
            " Using the CLI: `docling --artifacts-\u001b[33mpath\u001b[0m=\u001b[35m/root/.cache/docling/\u001b[0m\u001b[95mmodels\u001b[0m FILE` \n",
            " Using Python: see the documentation at \n",
            "\u001b[1m<\u001b[0m\u001b[4;94mhttps:\u001b[0m\u001b[4;94m//docling-project.github.io/docling/usage\u001b[0m\u001b[1m>\u001b[0m.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!docling /content/wiki.pdf --from pdf --to md --output /content --ocr-engine"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Ep7pNme8ZoqO",
        "outputId": "ecc2aa14-c836-4525-a3d6-6828c5d64250"
      },
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 22:53:41.899585: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746917621.923813   50298 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746917621.930862   50298 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "\u001b[31m╭─\u001b[0m\u001b[31m Error \u001b[0m\u001b[31m─────────────────────────────────────────────────────────────────────\u001b[0m\u001b[31m─╮\u001b[0m\n",
            "\u001b[31m│\u001b[0m Option '\u001b[1;36m-\u001b[0m\u001b[1;36m-ocr\u001b[0m\u001b[1;36m-engine\u001b[0m' requires an argument.                                  \u001b[31m│\u001b[0m\n",
            "\u001b[31m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!docling --help"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xtSRed-uZ8-T",
        "outputId": "82d79523-7367-4d59-e052-0995c120b454"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 22:55:01.199834: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746917701.322736   50612 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746917701.347221   50612 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "2025-05-10 22:55:01.437509: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
            "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
            "\u001b[1m                                                                                \u001b[0m\n",
            "\u001b[1m \u001b[0m\u001b[1;33mUsage: \u001b[0m\u001b[1mdocling [OPTIONS] source\u001b[0m\u001b[1m                                               \u001b[0m\u001b[1m \u001b[0m\n",
            "\u001b[1m                                                                                \u001b[0m\n",
            "\u001b[2m╭─\u001b[0m\u001b[2m Arguments \u001b[0m\u001b[2m─────────────────────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[31m*\u001b[0m    input_sources      \u001b[1;33msource\u001b[0m  PDF files to convert. Can be local file /    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                 directory paths or URL.                      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                 \u001b[2m[default: None]                             \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                 \u001b[2;31m[required]                                  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
            "\u001b[2m╭─\u001b[0m\u001b[2m Options \u001b[0m\u001b[2m───────────────────────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-from\u001b[0m                                    \u001b[1;2;33m[\u001b[0m\u001b[1;33mdocx\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mpptx\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mhtml\u001b[0m  Specify input     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;2;33m|\u001b[0m\u001b[1;33mimage\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mpdf\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33masci\u001b[0m  formats to        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33midoc\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mmd\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mcsv\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mxls\u001b[0m  convert from.     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mx\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mxml_uspto\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mxml\u001b[0m  Defaults to all   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33m_jats\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mjson_docl\u001b[0m  formats.          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33ming\u001b[0m\u001b[1;2;33m]\u001b[0m\u001b[1;33m           \u001b[0m  \u001b[2m[default: None]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-to\u001b[0m                                      \u001b[1;2;33m[\u001b[0m\u001b[1;33mmd\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mjson\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mhtml\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mh\u001b[0m  Specify output    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mtml_split_page\u001b[0m\u001b[1;2;33m|\u001b[0m  formats. Defaults \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mtext\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mdoctags\u001b[0m\u001b[1;2;33m]\u001b[0m\u001b[1;33m  \u001b[0m  to Markdown.      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: None]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-headers\u001b[0m                                 \u001b[1;33mTEXT           \u001b[0m  Specify http      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            request headers   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            used when         \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            fetching url      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            input sources in  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            the form of a     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            JSON string       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: None]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-image\u001b[0m\u001b[1;36m-export-…\u001b[0m                          \u001b[1;2;33m[\u001b[0m\u001b[1;33mplaceholder\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mem\u001b[0m  Image export mode \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mbedded\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mreferenc\u001b[0m  for the document  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33med\u001b[0m\u001b[1;2;33m]\u001b[0m\u001b[1;33m            \u001b[0m  (only in case of  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            JSON, Markdown or \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            HTML). With       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            `placeholder`,    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            only the position \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            of the image is   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            marked in the     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            output. In        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            `embedded` mode,  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            the image is      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            embedded as       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            base64 encoded    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            string. In        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            `referenced`      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            mode, the image   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            is exported in    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            PNG format and    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            referenced from   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            the main exported \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            document.         \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2membedded]        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-pipeline\u001b[0m                                \u001b[1;2;33m[\u001b[0m\u001b[1;33mstandard\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mvlm\u001b[0m\u001b[1;2;33m]\u001b[0m\u001b[1;33m \u001b[0m  Choose the        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            pipeline to       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            process PDF or    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            image files.      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mstandard]        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-vlm\u001b[0m\u001b[1;36m-model\u001b[0m                               \u001b[1;2;33m[\u001b[0m\u001b[1;33msmoldocling\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mgr\u001b[0m  Choose the VLM    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33manite_vision\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mgr\u001b[0m  model to use with \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33manite_vision_ol\u001b[0m  PDF or image      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mlama\u001b[0m\u001b[1;2;33m]\u001b[0m\u001b[1;33m          \u001b[0m  files.            \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2msmoldocling]     \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-ocr\u001b[0m                 \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-ocr\u001b[0m            \u001b[1;33m               \u001b[0m  If enabled, the   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            bitmap content    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            will be processed \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            using OCR.        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: ocr]   \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-force\u001b[0m\u001b[1;36m-ocr\u001b[0m           \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-force-ocr\u001b[0m      \u001b[1;33m               \u001b[0m  Replace any       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            existing text     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            with OCR          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            generated text    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            over the full     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            content.          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-force-ocr]    \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-ocr\u001b[0m\u001b[1;36m-engine\u001b[0m                              \u001b[1;33mTEXT           \u001b[0m  The OCR engine to \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            use. When         \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[1;36m-\u001b[0m\u001b[1;36m-allow\u001b[0m\u001b[1;36m-external…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            is *not* set, the \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            available values  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            are: easyocr,     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            ocrmac, rapidocr, \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            tesserocr,        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            tesseract. Use    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            the option        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[1;36m-\u001b[0m\u001b[1;36m-show\u001b[0m\u001b[1;36m-external-…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            to see the        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            options allowed   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            with external     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            plugins.          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2measyocr]         \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-ocr\u001b[0m\u001b[1;36m-lang\u001b[0m                                \u001b[1;33mTEXT           \u001b[0m  Provide a         \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            comma-separated   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            list of languages \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            used by the OCR   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            engine. Note that \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            each OCR engine   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            has different     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            values for the    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            language names.   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: None]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-pdf\u001b[0m\u001b[1;36m-backend\u001b[0m                             \u001b[1;2;33m[\u001b[0m\u001b[1;33mpypdfium2\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mdlpa\u001b[0m  The PDF backend   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mrse_v1\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mdlparse_\u001b[0m  to use.           \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mv2\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mdlparse_v4\u001b[0m\u001b[1;2;33m]\u001b[0m\u001b[1;33m \u001b[0m  \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mdlparse_v2]      \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-table\u001b[0m\u001b[1;36m-mode\u001b[0m                              \u001b[1;2;33m[\u001b[0m\u001b[1;33mfast\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33maccurate\u001b[0m\u001b[1;2;33m]\u001b[0m  The mode to use   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            in the table      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            structure model.  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2maccurate]        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-enrich\u001b[0m\u001b[1;36m-code\u001b[0m         \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-enrich-code\u001b[0m    \u001b[1;33m               \u001b[0m  Enable the code   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            enrichment model  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            in the pipeline.  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-enrich-code]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-enrich\u001b[0m\u001b[1;36m-formula\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-enrich-for…\u001b[0m    \u001b[1;33m               \u001b[0m  Enable the        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            formula           \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            enrichment model  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            in the pipeline.  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-enrich-formul…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-enrich\u001b[0m\u001b[1;36m-pictur…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-enrich-pic…\u001b[0m    \u001b[1;33m               \u001b[0m  Enable the        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            picture           \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            classification    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            enrichment model  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            in the pipeline.  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-enrich-pictur…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-enrich\u001b[0m\u001b[1;36m-pictur…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-enrich-pic…\u001b[0m    \u001b[1;33m               \u001b[0m  Enable the        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            picture           \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            description model \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            in the pipeline.  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-enrich-pictur…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-artifacts\u001b[0m\u001b[1;36m-path\u001b[0m                          \u001b[1;33mPATH           \u001b[0m  If provided, the  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            location of the   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            model artifacts.  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: None]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-enable\u001b[0m\u001b[1;36m-remote…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-enable-rem…\u001b[0m    \u001b[1;33m               \u001b[0m  Must be enabled   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            when using models \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            connecting to     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            remote services.  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-enable-remote…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-allow\u001b[0m\u001b[1;36m-externa…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-allow-exte…\u001b[0m    \u001b[1;33m               \u001b[0m  Must be enabled   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            for loading       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            modules from      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            third-party       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            plugins.          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-allow-externa…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-show\u001b[0m\u001b[1;36m-external…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-show-exter…\u001b[0m    \u001b[1;33m               \u001b[0m  List the          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            third-party       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            plugins which are \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            available when    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            the option        \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[1;36m-\u001b[0m\u001b[1;36m-allow\u001b[0m\u001b[1;36m-external…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            is set.           \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-show-external…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-abort\u001b[0m\u001b[1;36m-on-error\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-abort-on-e…\u001b[0m    \u001b[1;33m               \u001b[0m  If enabled, the   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            processing will   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            be aborted when   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            the first error   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            is encountered.   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-abort-on-erro…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-output\u001b[0m                                  \u001b[1;33mPATH           \u001b[0m  Output directory  \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            where results are \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            saved.            \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: .]     \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-verbose\u001b[0m         \u001b[1;32m-v\u001b[0m                      \u001b[1;33mINTEGER        \u001b[0m  Set the verbosity \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            level. \u001b[1;32m-v\u001b[0m for     \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            info logging, \u001b[1;32m-vv\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            for debug         \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            logging.          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: 0]     \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-debug\u001b[0m\u001b[1;36m-visuali…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-debug-visu…\u001b[0m    \u001b[1;33m               \u001b[0m  Enable debug      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            output which      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            visualizes the    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            PDF cells         \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-debug-visuali…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-debug\u001b[0m\u001b[1;36m-visuali…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-debug-visu…\u001b[0m    \u001b[1;33m               \u001b[0m  Enable debug      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            output which      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            visualizes the    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            OCR cells         \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-debug-visuali…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-debug\u001b[0m\u001b[1;36m-visuali…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-debug-visu…\u001b[0m    \u001b[1;33m               \u001b[0m  Enable debug      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            output which      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            visualizes the    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            layour clusters   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-debug-visuali…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-debug\u001b[0m\u001b[1;36m-visuali…\u001b[0m      \u001b[1;35m-\u001b[0m\u001b[1;35m-no\u001b[0m\u001b[1;35m-debug-visu…\u001b[0m    \u001b[1;33m               \u001b[0m  Enable debug      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            output which      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            visualizes the    \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            table cells       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default:        \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2mno-debug-visuali…\u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-version\u001b[0m                                 \u001b[1;33m               \u001b[0m  Show version      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            information.      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-document\u001b[0m\u001b[1;36m-time…\u001b[0m                          \u001b[1;33mFLOAT          \u001b[0m  The timeout for   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            processing each   \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            document, in      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            seconds.          \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: None]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-num\u001b[0m\u001b[1;36m-threads\u001b[0m                             \u001b[1;33mINTEGER        \u001b[0m  Number of threads \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: 4]     \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-device\u001b[0m                                  \u001b[1;2;33m[\u001b[0m\u001b[1;33mauto\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mcpu\u001b[0m\u001b[1;2;33m|\u001b[0m\u001b[1;33mcuda\u001b[0m\u001b[1;2;33m|\u001b[0m  Accelerator       \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                           \u001b[1;33mmps\u001b[0m\u001b[1;2;33m]\u001b[0m\u001b[1;33m           \u001b[0m  device            \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            \u001b[2m[default: auto]  \u001b[0m \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-logo\u001b[0m                                    \u001b[1;33m               \u001b[0m  Docling logo      \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m \u001b[1;36m-\u001b[0m\u001b[1;36m-help\u001b[0m                                    \u001b[1;33m               \u001b[0m  Show this message \u001b[2m│\u001b[0m\n",
            "\u001b[2m│\u001b[0m                                                            and exit.         \u001b[2m│\u001b[0m\n",
            "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "w-obJDftZ-Y6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "cOV7gKwLa44c"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "VeK_6xwRa470"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!docling /content/z.png --from image --ocr-engine tesseract --ocr-lang ara --to md --output docling_output"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Skt1ZNDpa4-1",
        "outputId": "a8ab2ae8-4b9e-4024-b0ad-e7d0b1b448e4"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 22:59:43.784928: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746917983.809659   51822 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746917983.816859   51822 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "2025-05-10 22:59:43.841510: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
            "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
            "WARNING:docling.pipeline.base_pipeline:Encountered an error during conversion of document eb3cfaf02e135a533d5e12758cd9079bcd02d15f108198cf5b28166fbe475f62:\n",
            "Traceback (most recent call last):\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/pipeline/base_pipeline.py\", line 160, in _build_document\n",
            "    for p in pipeline_pages:  # Must exhaust!\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/pipeline/base_pipeline.py\", line 126, in _apply_on_pages\n",
            "    yield from page_batch\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/models/page_assemble_model.py\", line 69, in __call__\n",
            "    for page in page_batch:\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/models/table_structure_model.py\", line 181, in __call__\n",
            "    for page in page_batch:\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/models/layout_model.py\", line 146, in __call__\n",
            "    for page in page_batch:\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/models/tesseract_ocr_cli_model.py\", line 229, in __call__\n",
            "    df_result = self._run_tesseract(fname)\n",
            "                ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/models/tesseract_ocr_cli_model.py\", line 128, in _run_tesseract\n",
            "    df_result = pd.read_csv(\n",
            "                ^^^^^^^^^^^^\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/pandas/io/parsers/readers.py\", line 1026, in read_csv\n",
            "    return _read(filepath_or_buffer, kwds)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/pandas/io/parsers/readers.py\", line 620, in _read\n",
            "    parser = TextFileReader(filepath_or_buffer, **kwds)\n",
            "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/pandas/io/parsers/readers.py\", line 1620, in __init__\n",
            "    self._engine = self._make_engine(f, self.engine)\n",
            "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/pandas/io/parsers/readers.py\", line 1898, in _make_engine\n",
            "    return mapping[engine](f, **self.options)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/pandas/io/parsers/c_parser_wrapper.py\", line 93, in __init__\n",
            "    self._reader = parsers.TextReader(src, **kwds)\n",
            "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "\n",
            "  File \"parsers.pyx\", line 581, in pandas._libs.parsers.TextReader.__cinit__\n",
            "\n",
            "pandas.errors.EmptyDataError: No columns to parse from file\n",
            "\n",
            "WARNING:docling.cli.main:Document /tmp/tmp5o1oznww/z.png failed to convert.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!docling /content/z.png --from image --ocr-engine tesseract --ocr-lang ara --to html --output docling_output"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "P8jLd-UjbOKy",
        "outputId": "41fbe0a4-ffd3-41ef-fc58-2c6c194a3ec0"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 23:01:06.589982: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746918066.616139   52185 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746918066.623326   52185 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "2025-05-10 23:01:06.647406: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
            "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
            "WARNING:docling.pipeline.base_pipeline:Encountered an error during conversion of document eb3cfaf02e135a533d5e12758cd9079bcd02d15f108198cf5b28166fbe475f62:\n",
            "Traceback (most recent call last):\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/pipeline/base_pipeline.py\", line 160, in _build_document\n",
            "    for p in pipeline_pages:  # Must exhaust!\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/pipeline/base_pipeline.py\", line 126, in _apply_on_pages\n",
            "    yield from page_batch\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/models/page_assemble_model.py\", line 69, in __call__\n",
            "    for page in page_batch:\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/models/table_structure_model.py\", line 181, in __call__\n",
            "    for page in page_batch:\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/models/layout_model.py\", line 146, in __call__\n",
            "    for page in page_batch:\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/models/tesseract_ocr_cli_model.py\", line 229, in __call__\n",
            "    df_result = self._run_tesseract(fname)\n",
            "                ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/models/tesseract_ocr_cli_model.py\", line 128, in _run_tesseract\n",
            "    df_result = pd.read_csv(\n",
            "                ^^^^^^^^^^^^\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/pandas/io/parsers/readers.py\", line 1026, in read_csv\n",
            "    return _read(filepath_or_buffer, kwds)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/pandas/io/parsers/readers.py\", line 620, in _read\n",
            "    parser = TextFileReader(filepath_or_buffer, **kwds)\n",
            "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/pandas/io/parsers/readers.py\", line 1620, in __init__\n",
            "    self._engine = self._make_engine(f, self.engine)\n",
            "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/pandas/io/parsers/readers.py\", line 1898, in _make_engine\n",
            "    return mapping[engine](f, **self.options)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/pandas/io/parsers/c_parser_wrapper.py\", line 93, in __init__\n",
            "    self._reader = parsers.TextReader(src, **kwds)\n",
            "                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "\n",
            "  File \"parsers.pyx\", line 581, in pandas._libs.parsers.TextReader.__cinit__\n",
            "\n",
            "pandas.errors.EmptyDataError: No columns to parse from file\n",
            "\n",
            "WARNING:docling.cli.main:Document /tmp/tmprv03nrt8/z.png failed to convert.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!docling /content/z.png --from image --ocr-engine easyocr --ocr-lang ara --to md --output docling_output"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3D9FyIfFb1lb",
        "outputId": "9be76471-1fdb-445b-bb51-560b1de02fd4"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 23:02:31.222697: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746918151.248123   52548 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746918151.255577   52548 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "2025-05-10 23:02:31.280582: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
            "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
            "Traceback (most recent call last):\n",
            "  File \"/usr/local/bin/docling\", line 8, in <module>\n",
            "    sys.exit(app())\n",
            "             ^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/typer/main.py\", line 340, in __call__\n",
            "    raise e\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/typer/main.py\", line 323, in __call__\n",
            "    return get_command(self)(*args, **kwargs)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/click/core.py\", line 1161, in __call__\n",
            "    return self.main(*args, **kwargs)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/typer/core.py\", line 677, in main\n",
            "    return _main(\n",
            "           ^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/typer/core.py\", line 195, in _main\n",
            "    rv = self.invoke(ctx)\n",
            "         ^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/click/core.py\", line 1443, in invoke\n",
            "    return ctx.invoke(self.callback, **ctx.params)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/click/core.py\", line 788, in invoke\n",
            "    return __callback(*args, **kwargs)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/typer/main.py\", line 698, in wrapper\n",
            "    return callback(**use_params)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/cli/main.py\", line 593, in convert\n",
            "    export_documents(\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/cli/main.py\", line 167, in export_documents\n",
            "    for conv_res in conv_results:\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/document_converter.py\", line 243, in convert_all\n",
            "    for conv_res in conv_res_iter:\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/document_converter.py\", line 278, in _convert\n",
            "    for item in map(\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/document_converter.py\", line 324, in _process_document\n",
            "    conv_res = self._execute_pipeline(in_doc, raises_on_error=raises_on_error)\n",
            "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/document_converter.py\", line 345, in _execute_pipeline\n",
            "    pipeline = self._get_pipeline(in_doc.format)\n",
            "               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/document_converter.py\", line 307, in _get_pipeline\n",
            "    self.initialized_pipelines[cache_key] = pipeline_class(\n",
            "                                            ^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/pipeline/standard_pdf_pipeline.py\", line 65, in __init__\n",
            "    ocr_model = self.get_ocr_model(artifacts_path=artifacts_path)\n",
            "                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/pipeline/standard_pdf_pipeline.py\", line 152, in get_ocr_model\n",
            "    return factory.create_instance(\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/models/factories/base_factory.py\", line 57, in create_instance\n",
            "    return _cls(options=options, **kwargs)\n",
            "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/docling/models/easyocr_model.py\", line 82, in __init__\n",
            "    self.reader = easyocr.Reader(\n",
            "                  ^^^^^^^^^^^^^^^\n",
            "  File \"/usr/local/lib/python3.11/dist-packages/easyocr/easyocr.py\", line 109, in __init__\n",
            "    raise ValueError(unknown_lang, 'is not supported')\n",
            "ValueError: ({'ara'}, 'is not supported')\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "https://github.com/docling-project/docling/issues/1304"
      ],
      "metadata": {
        "id": "gG3ueRsNc7Vj"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import httpx\n",
        "\n",
        "async_client = httpx.AsyncClient(timeout=6000.0)\n",
        "url = \"http://localhost:6001/v1alpha/convert/source\"\n",
        "payload = {\n",
        "\"options\": {\n",
        "\"from_formats\": [\"pdf\",\"image\"],\n",
        "\"to_formats\": [\"md\"],\n",
        "\"image_export_mode\": \"placeholder\",\n",
        "\"do_ocr\": True,\n",
        "\"pipeline\": \"standard\",\n",
        "\"force_ocr\": False,\n",
        "\"ocr_engine\": \"easyocr\",\n",
        "\"ocr_lang\": [\"en\"],\n",
        "\"pdf_backend\": \"dlparse_v2\",\n",
        "\"table_mode\": \"fast\",\n",
        "\"abort_on_error\": False,\n",
        "\"return_as_file\": False\n",
        "},\n",
        "\"file_sources\": [],\n",
        "\"http_sources\": [\n",
        "{\"url\": \"https://ftp.idu.ac.id/wp-content/uploads/ebook/tdg/DESIGN%20SISTEM%20DAYA%20GERAK/Fundamentals%20of%20Rocket%20Propulsion.pdf\"}\n",
        "]\n",
        "}\n",
        "\n",
        "response = await async_client.post(url, json=payload)\n",
        "\n",
        "data = response.json()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 408
        },
        "id": "RphLzag1cvgr",
        "outputId": "ef5b4e94-d439-4dd7-a71b-80579cd89621"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "error",
          "ename": "ConnectError",
          "evalue": "All connection attempts failed",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mConnectError\u001b[0m                              Traceback (most recent call last)",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpx/_transports/default.py\u001b[0m in \u001b[0;36mmap_httpcore_exceptions\u001b[0;34m()\u001b[0m\n\u001b[1;32m    100\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 101\u001b[0;31m         \u001b[0;32myield\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    102\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpx/_transports/default.py\u001b[0m in \u001b[0;36mhandle_async_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    393\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mmap_httpcore_exceptions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 394\u001b[0;31m             \u001b[0mresp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mawait\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_pool\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandle_async_request\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    395\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpcore/_async/connection_pool.py\u001b[0m in \u001b[0;36mhandle_async_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    255\u001b[0m             \u001b[0;32mawait\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_close_connections\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mclosing\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 256\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mexc\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    257\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpcore/_async/connection_pool.py\u001b[0m in \u001b[0;36mhandle_async_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    235\u001b[0m                     \u001b[0;31m# Send the request on the assigned connection.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 236\u001b[0;31m                     response = await connection.handle_async_request(\n\u001b[0m\u001b[1;32m    237\u001b[0m                         \u001b[0mpool_request\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpcore/_async/connection.py\u001b[0m in \u001b[0;36mhandle_async_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    100\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_connect_failed\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 101\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    102\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpcore/_async/connection.py\u001b[0m in \u001b[0;36mhandle_async_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m     77\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_connection\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 78\u001b[0;31m                     \u001b[0mstream\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mawait\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_connect\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     79\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpcore/_async/connection.py\u001b[0m in \u001b[0;36m_connect\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    123\u001b[0m                     \u001b[0;32masync\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mTrace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"connect_tcp\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlogger\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mtrace\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 124\u001b[0;31m                         \u001b[0mstream\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mawait\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_network_backend\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconnect_tcp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    125\u001b[0m                         \u001b[0mtrace\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturn_value\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstream\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpcore/_backends/auto.py\u001b[0m in \u001b[0;36mconnect_tcp\u001b[0;34m(self, host, port, timeout, local_address, socket_options)\u001b[0m\n\u001b[1;32m     30\u001b[0m         \u001b[0;32mawait\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_init_backend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 31\u001b[0;31m         return await self._backend.connect_tcp(\n\u001b[0m\u001b[1;32m     32\u001b[0m             \u001b[0mhost\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpcore/_backends/anyio.py\u001b[0m in \u001b[0;36mconnect_tcp\u001b[0;34m(self, host, port, timeout, local_address, socket_options)\u001b[0m\n\u001b[1;32m    112\u001b[0m         }\n\u001b[0;32m--> 113\u001b[0;31m         \u001b[0;32mwith\u001b[0m \u001b[0mmap_exceptions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexc_map\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    114\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0manyio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfail_after\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/lib/python3.11/contextlib.py\u001b[0m in \u001b[0;36m__exit__\u001b[0;34m(self, typ, value, traceback)\u001b[0m\n\u001b[1;32m    157\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 158\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgen\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mthrow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtyp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtraceback\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    159\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpcore/_exceptions.py\u001b[0m in \u001b[0;36mmap_exceptions\u001b[0;34m(map)\u001b[0m\n\u001b[1;32m     13\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfrom_exc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mto_exc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexc\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m         \u001b[0;32mraise\u001b[0m  \u001b[0;31m# pragma: nocover\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mConnectError\u001b[0m: All connection attempts failed",
            "\nThe above exception was the direct cause of the following exception:\n",
            "\u001b[0;31mConnectError\u001b[0m                              Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-7-bfe2c2643005>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     24\u001b[0m }\n\u001b[1;32m     25\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 26\u001b[0;31m \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mawait\u001b[0m \u001b[0masync_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpost\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpayload\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     27\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mresponse\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjson\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpx/_client.py\u001b[0m in \u001b[0;36mpost\u001b[0;34m(self, url, content, data, files, json, params, headers, cookies, auth, follow_redirects, timeout, extensions)\u001b[0m\n\u001b[1;32m   1857\u001b[0m         \u001b[0;34m**\u001b[0m\u001b[0mParameters\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mSee\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m`\u001b[0m\u001b[0mhttpx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;31m`\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1858\u001b[0m         \"\"\"\n\u001b[0;32m-> 1859\u001b[0;31m         return await self.request(\n\u001b[0m\u001b[1;32m   1860\u001b[0m             \u001b[0;34m\"POST\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1861\u001b[0m             \u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpx/_client.py\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(self, method, url, content, data, files, json, params, headers, cookies, auth, follow_redirects, timeout, extensions)\u001b[0m\n\u001b[1;32m   1538\u001b[0m             \u001b[0mextensions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mextensions\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1539\u001b[0m         )\n\u001b[0;32m-> 1540\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0;32mawait\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mauth\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mauth\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfollow_redirects\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfollow_redirects\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1541\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1542\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0masynccontextmanager\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpx/_client.py\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, stream, auth, follow_redirects)\u001b[0m\n\u001b[1;32m   1627\u001b[0m         \u001b[0mauth\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_build_request_auth\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mauth\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1628\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1629\u001b[0;31m         response = await self._send_handling_auth(\n\u001b[0m\u001b[1;32m   1630\u001b[0m             \u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1631\u001b[0m             \u001b[0mauth\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mauth\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpx/_client.py\u001b[0m in \u001b[0;36m_send_handling_auth\u001b[0;34m(self, request, auth, follow_redirects, history)\u001b[0m\n\u001b[1;32m   1655\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1656\u001b[0m             \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1657\u001b[0;31m                 response = await self._send_handling_redirects(\n\u001b[0m\u001b[1;32m   1658\u001b[0m                     \u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1659\u001b[0m                     \u001b[0mfollow_redirects\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfollow_redirects\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpx/_client.py\u001b[0m in \u001b[0;36m_send_handling_redirects\u001b[0;34m(self, request, follow_redirects, history)\u001b[0m\n\u001b[1;32m   1692\u001b[0m                 \u001b[0;32mawait\u001b[0m \u001b[0mhook\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1693\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1694\u001b[0;31m             \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mawait\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_send_single_request\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1695\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1696\u001b[0m                 \u001b[0;32mfor\u001b[0m \u001b[0mhook\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_event_hooks\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"response\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpx/_client.py\u001b[0m in \u001b[0;36m_send_single_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m   1728\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1729\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mrequest_context\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1730\u001b[0;31m             \u001b[0mresponse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mawait\u001b[0m \u001b[0mtransport\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandle_async_request\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1731\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1732\u001b[0m         \u001b[0;32massert\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresponse\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstream\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mAsyncByteStream\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpx/_transports/default.py\u001b[0m in \u001b[0;36mhandle_async_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    391\u001b[0m             \u001b[0mextensions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextensions\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    392\u001b[0m         )\n\u001b[0;32m--> 393\u001b[0;31m         \u001b[0;32mwith\u001b[0m \u001b[0mmap_httpcore_exceptions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    394\u001b[0m             \u001b[0mresp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mawait\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_pool\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandle_async_request\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    395\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/lib/python3.11/contextlib.py\u001b[0m in \u001b[0;36m__exit__\u001b[0;34m(self, typ, value, traceback)\u001b[0m\n\u001b[1;32m    156\u001b[0m                 \u001b[0mvalue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtyp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    157\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 158\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgen\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mthrow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtyp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtraceback\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    159\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    160\u001b[0m                 \u001b[0;31m# Suppress StopIteration *unless* it's the same exception that\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/httpx/_transports/default.py\u001b[0m in \u001b[0;36mmap_httpcore_exceptions\u001b[0;34m()\u001b[0m\n\u001b[1;32m    116\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    117\u001b[0m         \u001b[0mmessage\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 118\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mmapped_exc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmessage\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    119\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    120\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mConnectError\u001b[0m: All connection attempts failed"
          ]
        }
      ]
    },
    {
      "source": [
        "!nc -zv localhost 6001"
      ],
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8Cth5ox0c9Ck",
        "outputId": "0d001e28-b9f2-4d43-c194-b5be837eccdd"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/bin/bash: line 1: nc: command not found\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "O5aYSDc9dOG6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "source": [
        "#!pip install docling --upgrade  # Make sure docling is installed and up-to-date\n",
        "\n",
        "from docling.document_converter import DocumentConverter\n",
        "\n",
        "def convert_document(source_url, output_format=\"md\"):\n",
        "  \"\"\"Converts a document from a URL to the specified output format.\n",
        "\n",
        "  Args:\n",
        "      source_url: The URL of the document to convert.\n",
        "      output_format: The desired output format (e.g., \"md\", \"html\"). Defaults to \"md\".\n",
        "\n",
        "  Returns:\n",
        "      The converted document content as a string.\n",
        "  \"\"\"\n",
        "  try:\n",
        "    converter = DocumentConverter()\n",
        "    result = converter.convert(source_url)\n",
        "    if output_format == \"md\":\n",
        "      return result.document.export_to_markdown()\n",
        "    elif output_format == \"html\":\n",
        "      return result.document.export_to_html()\n",
        "    else:\n",
        "      print(f\"Unsupported output format: {output_format}\")\n",
        "      return None\n",
        "  except Exception as e:\n",
        "    print(f\"Error converting document: {e}\")\n",
        "    return None\n",
        "\n",
        "# Example usage:\n",
        "source_url = \"https://ftp.idu.ac.id/wp-content/uploads/ebook/tdg/DESIGN%20SISTEM%20DAYA%20GERAK/Fundamentals%20of%20Rocket%20Propulsion.pdf\"\n",
        "output_format = \"md\"  # Or \"html\"\n",
        "\n",
        "converted_content = convert_document(source_url, output_format)\n",
        "\n",
        "if converted_content:\n",
        "  print(converted_content)\n",
        "  # You can further save the converted content to a file if needed\n",
        "  with open(\"output.md\", \"w\", encoding=\"utf-8\") as f:\n",
        "    f.write(converted_content)"
      ],
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jkyv6gB6dPDz",
        "outputId": "760d8d43-1427-4c6e-b29f-5666d2870d76"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Error converting document: HTTPSConnectionPool(host='ftp.idu.ac.id', port=443): Max retries exceeded with url: /wp-content/uploads/ebook/tdg/DESIGN%20SISTEM%20DAYA%20GERAK/Fundamentals%20of%20Rocket%20Propulsion.pdf (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7efea212df50>, 'Connection to ftp.idu.ac.id timed out. (connect timeout=None)'))\n"
          ]
        }
      ]
    },
    {
      "source": [
        "source_url = \"/content/Fundamentals of Rocket Propulsion.pdf\"  # Replace with the actual path"
      ],
      "cell_type": "code",
      "metadata": {
        "id": "DAZwgR9OeShT"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "BQ2V3g0LecnX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "g1Uz7XKPeckn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "BTrU89t9ecgt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "EHlNFRcOeccz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### شغال من صورة الى ماركدون عربى"
      ],
      "metadata": {
        "id": "YjJM6DtQf3dc"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!docling /content/z.png --from image --ocr-engine easyocr --ocr-lang ar --to md --output docling_output"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "htcFBJ4BdQKT",
        "outputId": "7e6bb2e7-531a-4eb7-9a80-60bb98728baf"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 23:11:51.047911: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746918711.072893   54787 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746918711.080536   54787 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "WARNING:easyocr.easyocr:Downloading recognition model, please wait. This may take several minutes depending upon your network connection.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!docling /content/wiki.pdf --from pdf --ocr-engine easyocr --ocr-lang ar --to md --output docling_output"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "VjGUQlpbejcD",
        "outputId": "49d1ee72-bae1-4de6-a5f0-690bcc9786fd"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 23:21:28.986083: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746919289.028074   57118 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746919289.040676   57118 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!docling /content/wiki.pdf --from pdf --ocr-engine easyocr --ocr-lang ar --to md --output docling_output"
      ],
      "metadata": {
        "id": "L5xMiIKMejYa"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "aشغال حلو جدا"
      ],
      "metadata": {
        "id": "srLcwClGvhih"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!time docling /content/wik2.pdf --pdf-backend dlparse_v4   --no-enrich-formula --force-ocr  --no-enrich-code  --ocr   --ocr-engine easyocr   --ocr-lang en   --device auto  --num-threads 8   --to md"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EW0zg2gDejVs",
        "outputId": "421f44a3-1455-4eaf-d4ea-e8e63715d2c6"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2025-05-10 23:27:41.612403: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
            "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
            "E0000 00:00:1746919661.649408   58626 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
            "E0000 00:00:1746919661.661488   58626 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
            "\n",
            "real\t58m58.351s\n",
            "user\t78m36.757s\n",
            "sys\t14m11.729s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "tDUzKuNrejSV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "https://github.com/docling-project/docling/issues/1236"
      ],
      "metadata": {
        "id": "ChxSqwH-ejya"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "time docling ./test01.pdf --pdf-backend dlparse_v4   --no-enrich-formula --force-ocr  --no-enrich-code  --ocr   --ocr-engine easyocr   --ocr-lang en   --device auto  --num-threads 8   --to md"
      ],
      "metadata": {
        "id": "x6GyuuR7ecEd"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "https://github.com/docling-project/docling/issues/1456"
      ],
      "metadata": {
        "id": "DDPbBdezjTaX"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "docling --show-external-plugins"
      ],
      "metadata": {
        "id": "nn0YRyYtecCC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!docling --output \"[your output path]\" -v --pdf-backend dlparse_v2 --table-mode accurate --image-export-mode referenced  --artifacts-path=~/.cache/docling/models --enrich-picture-classes  [input file name] --ocr-engine tesseract --ocr-lang [lang code name]"
      ],
      "metadata": {
        "id": "Hy14XY9seb7U"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "https://github.com/felixdittrich92/docling-OCR-OnnxTR"
      ],
      "metadata": {
        "id": "TWU6SZ99jotc"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "MWfYR6uoeb07"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "https://docling-project.github.io/docling/installation/"
      ],
      "metadata": {
        "id": "IuYSW2iykxXt"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "For testing, I used the notebook from https://ds4sd.github.io/docling/examples/custom_convert/ and set the OCR language to German by configuring pipeline_options.ocr_options.lang = [\"de\"]"
      ],
      "metadata": {
        "id": "2dtEPLm_ebzJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "jFfrUrLYebu3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "1x9JCP5Nebqw"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "6YKq7XnRebnM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install easyocr"
      ],
      "metadata": {
        "id": "bo-hQqMQeDyb"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "source": [
        "!pip install docling --upgrade  # Make sure docling is installed and up-to-date\n",
        "\n",
        "from docling.document_converter import DocumentConverter\n",
        "\n",
        "def convert_document(source_url, output_format=\"md\"):\n",
        "  \"\"\"Converts a document from a URL to the specified output format.\n",
        "\n",
        "  Args:\n",
        "      source_url: The URL of the document to convert.\n",
        "      output_format: The desired output format (e.g., \"md\", \"html\"). Defaults to \"md\".\n",
        "\n",
        "  Returns:\n",
        "      The converted document content as a string.\n",
        "  \"\"\"\n",
        "  try:\n",
        "    converter = DocumentConverter()\n",
        "    result = converter.convert(source_url)\n",
        "    if output_format == \"md\":\n",
        "      return result.document.export_to_markdown()\n",
        "    elif output_format == \"html\":\n",
        "      return result.document.export_to_html()\n",
        "    else:\n",
        "      print(f\"Unsupported output format: {output_format}\")\n",
        "      return None\n",
        "  except Exception as e:\n",
        "    print(f\"Error converting document: {e}\")\n",
        "    return None\n",
        "\n",
        "# Example usage:\n",
        "source_url = \"https://ftp.idu.ac.id/wp-content/uploads/ebook/tdg/DESIGN%20SISTEM%20DAYA%20GERAK/Fundamentals%20of%20Rocket%20Propulsion.pdf\"\n",
        "output_format = \"md\"  # Or \"html\"\n",
        "\n",
        "converted_content = convert_document(source_url, output_format)\n",
        "\n",
        "if converted_content:\n",
        "  print(converted_content)\n",
        "  # You can further save the converted content to a file if needed\n",
        "  with open(\"output.md\", \"w\", encoding=\"utf-8\") as f:\n",
        "    f.write(converted_content)"
      ],
      "cell_type": "code",
      "metadata": {
        "id": "rMCQAOPXdsKr"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from pathlib import Path\n",
        "\n",
        "from docling.datamodel.base_models import InputFormat\n",
        "from docling.datamodel.pipeline_options import (\n",
        "    PdfPipelineOptions,\n",
        "    TesseractCliOcrOptions,\n",
        ")\n",
        "from docling.document_converter import DocumentConverter, PdfFormatOption\n",
        "\n",
        "inputData = \"/content/wiki.pdf\"\n",
        "outputData = \"/content/wiki.md\"\n",
        "\n",
        "ocr_options = TesseractCliOcrOptions(lang=[\"auto\"])\n",
        "pipeline_options = PdfPipelineOptions(\n",
        "        do_ocr=True, force_full_page_ocr=True, ocr_options=ocr_options\n",
        "    )\n",
        "converter = DocumentConverter(\n",
        "        format_options={\n",
        "            InputFormat.PDF: PdfFormatOption(\n",
        "                pipeline_options=pipeline_options,\n",
        "            )\n",
        "        }\n",
        "    )\n",
        "\n",
        "print(\"beginning data conversion...\")\n",
        "\n",
        "# ingest all pdfs in data\n",
        "input_path = Path(inputData)\n",
        "pathlist = input_path.rglob('*.pdf')\n",
        "for input_file in pathlist:\n",
        "    print(\"loading \" + input_file.name + \"...\")\n",
        "    doc = converter.convert(str(input_file)).document\n",
        "    md = doc.export_to_markdown()\n",
        "\n",
        "    outfileName = outputData + str(input_file.stem).strip() + \".md\"\n",
        "    print(\"written to \" + outfileName)\n",
        "    outfile = open(outfileName, \"x\")\n",
        "    outfile.write(md)\n",
        "    outfile.close()\n",
        "\n",
        "print(\"data conversion complete\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1dw2nxPowOAp",
        "outputId": "809c9e2c-8f89-49f0-c92e-c56de7c877fd"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "beginning data conversion...\n",
            "data conversion complete\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "nuLxyVAcwwtA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "source": [
        "from pathlib import Path\n",
        "from docling.datamodel.base_models import InputFormat\n",
        "from docling.datamodel.pipeline_options import (\n",
        "    PdfPipelineOptions,\n",
        "    TesseractCliOcrOptions,\n",
        ")\n",
        "from docling.document_converter import DocumentConverter, PdfFormatOption\n",
        "\n",
        "inputData = \"/content/wiki.pdf\"  # Make sure wiki.pdf is uploaded in Colab\n",
        "outputData = \"/content/\"\n",
        "\n",
        "ocr_options = TesseractCliOcrOptions(lang=[\"auto\"])\n",
        "pipeline_options = PdfPipelineOptions(\n",
        "    do_ocr=True, force_full_page_ocr=True, ocr_options=ocr_options\n",
        ")\n",
        "converter = DocumentConverter(\n",
        "    format_options={\n",
        "        InputFormat.PDF: PdfFormatOption(\n",
        "            pipeline_options=pipeline_options,\n",
        "        )\n",
        "    }\n",
        ")\n",
        "\n",
        "print(\"beginning data conversion...\")\n",
        "\n",
        "input_path = Path(inputData)\n",
        "pathlist = input_path.rglob('*.pdf')\n",
        "for input_file in pathlist:\n",
        "    print(\"loading \" + input_file.name + \"...\")\n",
        "    print(\"Before conversion\")  # Debugging print\n",
        "    doc = converter.convert(str(input_file)).document\n",
        "    print(\"After conversion\")  # Debugging print\n",
        "    md = doc.export_to_markdown()\n",
        "\n",
        "    outfileName = outputData + str(input_file.stem).strip() + \".md\"\n",
        "    print(\"written to \" + outfileName)\n",
        "    outfile = open(outfileName, \"w\", encoding='utf-8')  # Changed to \"w\" mode\n",
        "    outfile.write(md)\n",
        "    outfile.close()\n",
        "\n",
        "print(\"data conversion complete\")"
      ],
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4y8LuGcHw079",
        "outputId": "0b9e8062-14d5-4dc0-f54b-404dbb6b2518"
      },
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "beginning data conversion...\n",
            "data conversion complete\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "https://github.com/docling-project/docling/issues/1448"
      ],
      "metadata": {
        "id": "TPSAWU3FxRvq"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "y7UwPo8BxSbp"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "source": [
        "from pathlib import Path\n",
        "from docling.datamodel.base_models import InputFormat\n",
        "from docling.datamodel.pipeline_options import (\n",
        "    PdfPipelineOptions,\n",
        "    TesseractCliOcrOptions,\n",
        ")\n",
        "from docling.document_converter import DocumentConverter, PdfFormatOption\n",
        "\n",
        "inputData = \"/content/BofA_InterestChecking_en_ADA.pdf\"  # Make sure wiki.pdf is uploaded in Colab\n",
        "outputData = \"/content/\"\n",
        "\n",
        "# ... (Rest of the code is the same)\n",
        "\n",
        "print(\"beginning data conversion...\")\n",
        "\n",
        "input_path = Path(inputData)\n",
        "pathlist = input_path.rglob('*.pdf')\n",
        "print(list(pathlist)) # Debugging print for pathlist\n",
        "for input_file in pathlist:\n",
        "    print(input_file)  # Debugging print for input_file\n",
        "    print(\"loading \" + input_file.name + \"...\")\n",
        "    print(\"Before conversion\")\n",
        "    try:\n",
        "        doc = converter.convert(str(input_file)).document\n",
        "        print(\"Conversion successful\")\n",
        "    except Exception as e:\n",
        "        print(f\"Error during conversion: {e}\")\n",
        "    print(\"After conversion\")\n",
        "    md = doc.export_to_markdown()\n",
        "\n",
        "    # ... (Rest of the code is the same)\n",
        "\n",
        "print(\"data conversion complete\")"
      ],
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "VeY-OpnWxVQy",
        "outputId": "99068b6e-3ba8-4d09-9b02-184e00b49cd6"
      },
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "beginning data conversion...\n",
            "[]\n",
            "data conversion complete\n"
          ]
        }
      ]
    },
    {
      "source": [
        "from pathlib import Path\n",
        "from docling.datamodel.base_models import InputFormat\n",
        "from docling.datamodel.pipeline_options import (\n",
        "    PdfPipelineOptions,\n",
        "    TesseractCliOcrOptions,\n",
        ")\n",
        "from docling.document_converter import DocumentConverter, PdfFormatOption\n",
        "\n",
        "inputData = \"/content/BofA_InterestChecking_en_ADA.pdf\"  # Make sure wiki.pdf is uploaded in Colab\n",
        "outputData = \"/content/\"\n",
        "\n",
        "# ... (Rest of the code is the same)\n",
        "\n",
        "print(\"beginning data conversion...\")\n",
        "\n",
        "input_path = Path('/content/') # Updated to search in /content/\n",
        "pathlist = input_path.glob('*.pdf')  # Or use rglob to search subdirectories\n",
        "for input_file in pathlist:\n",
        "    print(input_file) # Debugging print for input_file\n",
        "    print(\"loading \" + input_file.name + \"...\")\n",
        "    print(\"Before conversion\")\n",
        "    try:\n",
        "        doc = converter.convert(str(input_file)).document\n",
        "        print(\"Conversion successful\")\n",
        "    except Exception as e:\n",
        "        print(f\"Error during conversion: {e}\")\n",
        "    print(\"After conversion\")\n",
        "    md = doc.export_to_markdown()\n",
        "\n",
        "    # ... (Rest of the code is the same)\n",
        "\n",
        "print(\"data conversion complete\")"
      ],
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "oiz6VdnxyF8L",
        "outputId": "caf23f39-2740-4870-94e0-c58e6e892112"
      },
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "beginning data conversion...\n",
            "/content/BofA_InterestChecking_en_ADA.pdf\n",
            "loading BofA_InterestChecking_en_ADA.pdf...\n",
            "Before conversion\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: \n",
            "The secret `HF_TOKEN` does not exist in your Colab secrets.\n",
            "To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.\n",
            "You will be able to reuse this secret in all of your notebooks.\n",
            "Please note that authentication is recommended but still optional to access public models or datasets.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Conversion successful\n",
            "After conversion\n",
            "/content/wiki.pdf\n",
            "loading wiki.pdf...\n",
            "Before conversion\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Cyrillic' and language 'Cyrillic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Arabic' and language 'Arabic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Arabic' and language 'Arabic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Conversion successful\n",
            "After conversion\n",
            "/content/q.pdf\n",
            "loading q.pdf...\n",
            "Before conversion\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Arabic' and language 'Arabic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Arabic' and language 'Arabic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Arabic' and language 'Arabic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Arabic' and language 'Arabic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Arabic' and language 'Arabic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Arabic' and language 'Arabic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Arabic' and language 'Arabic'. However this language is not installed in your system and will be ignored.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Conversion successful\n",
            "After conversion\n",
            "/content/wik2.pdf\n",
            "loading wik2.pdf...\n",
            "Before conversion\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Cyrillic' and language 'Cyrillic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Japanese' and language 'Japanese'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Cyrillic' and language 'Cyrillic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Arabic' and language 'Arabic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Arabic' and language 'Arabic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Japanese' and language 'Japanese'. However this language is not installed in your system and will be ignored.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Conversion successful\n",
            "After conversion\n",
            "/content/1.pdf\n",
            "loading 1.pdf...\n",
            "Before conversion\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Cyrillic' and language 'Cyrillic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Cyrillic' and language 'Cyrillic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Cyrillic' and language 'Cyrillic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Cyrillic' and language 'Cyrillic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract cannot detect the script of the page\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Cyrillic' and language 'Cyrillic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Conversion successful\n",
            "After conversion\n",
            "data conversion complete\n"
          ]
        }
      ]
    },
    {
      "source": [
        "try:\n",
        "    print(\"Attempting conversion...\")\n",
        "    doc = converter.convert(str(input_file))  # Remove .document to see potential intermediate errors\n",
        "    print(\"Conversion successful\")\n",
        "    print(\"Attempting to get document...\")  # Add debug print\n",
        "    doc = doc.document  # Access .document separately\n",
        "    print(\"Got document successfully!\")  # Add debug print\n",
        "    md = doc.export_to_markdown()\n",
        "    print(\"Markdown extracted successfully!\")  # Add debug print\n",
        "except Exception as e:\n",
        "    print(f\"Error during conversion: {e}\")"
      ],
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 515
        },
        "id": "QCUHvzJj_cxx",
        "outputId": "cb11f4ba-c401-425e-ce5a-aee76d1aa513"
      },
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Attempting conversion...\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Cyrillic' and language 'Cyrillic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Cyrillic' and language 'Cyrillic'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n",
            "WARNING:docling.models.tesseract_ocr_cli_model:Tesseract detected the script 'Latin' and language 'Latin'. However this language is not installed in your system and will be ignored.\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "KeyboardInterrupt",
          "evalue": "",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-23-fe909fb19b83>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Attempting conversion...\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0mdoc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconverter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconvert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# Remove .document to see potential intermediate errors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Conversion successful\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Attempting to get document...\"\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# Add debug print\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/pydantic/_internal/_validate_call.py\u001b[0m in \u001b[0;36mwrapper_function\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     37\u001b[0m         \u001b[0;34m@\u001b[0m\u001b[0mfunctools\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwraps\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwrapped\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mwrapper_function\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 39\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     40\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m     \u001b[0;31m# We need to manually update this because `partial` object has no `__name__` and `__qualname__`.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/pydantic/_internal/_validate_call.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    134\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_create_validators\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    135\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 136\u001b[0;31m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__pydantic_validator__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalidate_python\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpydantic_core\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mArgsKwargs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    137\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__return_pydantic_validator__\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    138\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__return_pydantic_validator__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mres\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/docling/document_converter.py\u001b[0m in \u001b[0;36mconvert\u001b[0;34m(self, source, headers, raises_on_error, max_num_pages, max_file_size, page_range)\u001b[0m\n\u001b[1;32m    218\u001b[0m             \u001b[0mpage_range\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpage_range\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    219\u001b[0m         )\n\u001b[0;32m--> 220\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mall_res\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    221\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    222\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mvalidate_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconfig\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mConfigDict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstrict\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/docling/document_converter.py\u001b[0m in \u001b[0;36mconvert_all\u001b[0;34m(self, source, headers, raises_on_error, max_num_pages, max_file_size, page_range)\u001b[0m\n\u001b[1;32m    241\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    242\u001b[0m         \u001b[0mhad_result\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 243\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0mconv_res\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mconv_res_iter\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    244\u001b[0m             \u001b[0mhad_result\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    245\u001b[0m             if raises_on_error and conv_res.status not in {\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/docling/document_converter.py\u001b[0m in \u001b[0;36m_convert\u001b[0;34m(self, conv_input, raises_on_error)\u001b[0m\n\u001b[1;32m    276\u001b[0m             \u001b[0;31m# Note: PDF backends are not thread-safe, thread pool usage was disabled.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    277\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 278\u001b[0;31m             for item in map(\n\u001b[0m\u001b[1;32m    279\u001b[0m                 \u001b[0mpartial\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_process_document\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mraises_on_error\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mraises_on_error\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    280\u001b[0m                 \u001b[0minput_batch\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/docling/document_converter.py\u001b[0m in \u001b[0;36m_process_document\u001b[0;34m(self, in_doc, raises_on_error)\u001b[0m\n\u001b[1;32m    322\u001b[0m         )\n\u001b[1;32m    323\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mvalid\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 324\u001b[0;31m             \u001b[0mconv_res\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_execute_pipeline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0min_doc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mraises_on_error\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mraises_on_error\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    325\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    326\u001b[0m             \u001b[0merror_message\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34mf\"File format not allowed: {in_doc.file}\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/docling/document_converter.py\u001b[0m in \u001b[0;36m_execute_pipeline\u001b[0;34m(self, in_doc, raises_on_error)\u001b[0m\n\u001b[1;32m    345\u001b[0m             \u001b[0mpipeline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_pipeline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0min_doc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    346\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mpipeline\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 347\u001b[0;31m                 \u001b[0mconv_res\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpipeline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0min_doc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mraises_on_error\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mraises_on_error\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    348\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    349\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mraises_on_error\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/docling/pipeline/base_pipeline.py\u001b[0m in \u001b[0;36mexecute\u001b[0;34m(self, in_doc, raises_on_error)\u001b[0m\n\u001b[1;32m     44\u001b[0m                 \u001b[0;31m# These steps are building and assembling the structure of the\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     45\u001b[0m                 \u001b[0;31m# output DoclingDocument.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 46\u001b[0;31m                 \u001b[0mconv_res\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_build_document\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconv_res\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     47\u001b[0m                 \u001b[0mconv_res\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_assemble_document\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconv_res\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     48\u001b[0m                 \u001b[0;31m# From this stage, all operations should rely only on conv_res.output\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/docling/pipeline/base_pipeline.py\u001b[0m in \u001b[0;36m_build_document\u001b[0;34m(self, conv_res)\u001b[0m\n\u001b[1;32m    158\u001b[0m                     \u001b[0mpipeline_pages\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_apply_on_pages\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconv_res\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minit_pages\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    159\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 160\u001b[0;31m                     \u001b[0;32mfor\u001b[0m \u001b[0mp\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mpipeline_pages\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# Must exhaust!\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    161\u001b[0m                         \u001b[0;31m# Cleanup cached images\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    162\u001b[0m                         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeep_images\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/docling/pipeline/base_pipeline.py\u001b[0m in \u001b[0;36m_apply_on_pages\u001b[0;34m(self, conv_res, page_batch)\u001b[0m\n\u001b[1;32m    124\u001b[0m             \u001b[0mpage_batch\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mconv_res\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpage_batch\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    125\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 126\u001b[0;31m         \u001b[0;32myield\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mpage_batch\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    127\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    128\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_build_document\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconv_res\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mConversionResult\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mConversionResult\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/docling/models/page_assemble_model.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, conv_res, page_batch)\u001b[0m\n\u001b[1;32m     67\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconv_res\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mConversionResult\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpage_batch\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mIterable\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mPage\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     68\u001b[0m     ) -> Iterable[Page]:\n\u001b[0;32m---> 69\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0mpage\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mpage_batch\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     70\u001b[0m             \u001b[0;32massert\u001b[0m \u001b[0mpage\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_backend\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     71\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mpage\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_backend\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_valid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/docling/models/table_structure_model.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, conv_res, page_batch)\u001b[0m\n\u001b[1;32m    179\u001b[0m             \u001b[0;32mreturn\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    180\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 181\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0mpage\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mpage_batch\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    182\u001b[0m             \u001b[0;32massert\u001b[0m \u001b[0mpage\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_backend\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    183\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mpage\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_backend\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_valid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/docling/models/layout_model.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, conv_res, page_batch)\u001b[0m\n\u001b[1;32m    144\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconv_res\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mConversionResult\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpage_batch\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mIterable\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mPage\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    145\u001b[0m     ) -> Iterable[Page]:\n\u001b[0;32m--> 146\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0mpage\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mpage_batch\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    147\u001b[0m             \u001b[0;32massert\u001b[0m \u001b[0mpage\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_backend\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    148\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mpage\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_backend\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_valid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/docling/models/tesseract_ocr_cli_model.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, conv_res, page_batch)\u001b[0m\n\u001b[1;32m    227\u001b[0m                                 \u001b[0mhigh_res_image\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msave\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mimage_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    228\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 229\u001b[0;31m                             \u001b[0mdf_result\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_run_tesseract\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    230\u001b[0m                         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    231\u001b[0m                             \u001b[0;32mif\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.11/dist-packages/docling/models/tesseract_ocr_cli_model.py\u001b[0m in \u001b[0;36m_run_tesseract\u001b[0;34m(self, ifilename)\u001b[0m\n\u001b[1;32m    117\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    118\u001b[0m         \u001b[0mproc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstdout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mPIPE\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mDEVNULL\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 119\u001b[0;31m         \u001b[0moutput\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mproc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcommunicate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    120\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    121\u001b[0m         \u001b[0;31m# _log.info(output)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/lib/python3.11/subprocess.py\u001b[0m in \u001b[0;36mcommunicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1194\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_stdin_write\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1195\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1196\u001b[0;31m                 \u001b[0mstdout\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1197\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1198\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from pathlib import Path\n",
        "\n",
        "from docling.datamodel.base_models import InputFormat\n",
        "from docling.datamodel.pipeline_options import (\n",
        "    PdfPipelineOptions,\n",
        "    TesseractCliOcrOptions,\n",
        ")\n",
        "from docling.document_converter import DocumentConverter, PdfFormatOption\n",
        "\n",
        "inputData = \"/content/BofA_InterestChecking_en_ADA.pdf\"\n",
        "outputData = \"/content/BofA_InterestChecking_en_ADA.md\"\n",
        "\n",
        "ocr_options = TesseractCliOcrOptions(lang=[\"en\"])\n",
        "pipeline_options = PdfPipelineOptions(\n",
        "        do_ocr=True, force_full_page_ocr=True, ocr_options=ocr_options\n",
        "    )\n",
        "converter = DocumentConverter(\n",
        "        format_options={\n",
        "            InputFormat.PDF: PdfFormatOption(\n",
        "                pipeline_options=pipeline_options,\n",
        "            )\n",
        "        }\n",
        "    )\n",
        "\n",
        "print(\"beginning data conversion...\")\n",
        "\n",
        "# ingest all pdfs in data\n",
        "input_path = Path(inputData)\n",
        "pathlist = input_path.rglob('*.pdf')\n",
        "for input_file in pathlist:\n",
        "    print(\"loading \" + input_file.name + \"...\")\n",
        "    doc = converter.convert(str(input_file)).document\n",
        "    md = doc.export_to_markdown()\n",
        "\n",
        "    outfileName = outputData + str(input_file.stem).strip() + \".md\"\n",
        "    print(\"written to \" + outfileName)\n",
        "    outfile = open(outfileName, \"x\")\n",
        "    outfile.write(md)\n",
        "    outfile.close()\n",
        "\n",
        "print(\"data conversion complete\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rxXAgkATyIHg",
        "outputId": "85f75026-55f3-45c8-fd38-f559a83c191a"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "beginning data conversion...\n",
            "data conversion complete\n"
          ]
        }
      ]
    }
  ]
}